# Preparing data sets {#dataset}
## Case I: 10x data (Stewart et al., Nat Cancer, 2020)
```{r, eval = FALSE}
source("R/function_general.R")
```

The data were obtained from NCBI repository with accession number [GSE138474](https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE138474).
From the above, we downloaded scRNA-seq data with sample accessions numbers
GSM4104164 (for MDA-SC68 with vehicle treatment) and
GSM4104163 (for MDA-SC68 with cisplatin treatment).

Due to limits of file size, we prepared only the following ASURAT objects:
```{r, eval = FALSE}
# ----------------------------------------
# GSM4104164
# ----------------------------------------
sc68_vehi <- readRDS(file = "backup/01_001_sc68_vehi_raw.rds")
# ----------------------------------------
# GSM4104163
# ----------------------------------------
sc68_cisp <- readRDS(file = "backup/02_001_sc68_cisp_raw.rds")
```



## Case II: PDAC-A ST and inDrop (Moncada et al., Nat Biotechnol, 2020)
The data were obtained from NCBI repository with accession number [GSE111672](https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE111672).



### PDAC-A ST
From [GSE111672](https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE111672),
we downloaded spatial transcriptome data with sample accession number
GSM3036911 (for PDAC-A ST).

The spatial coordinate information was obtained from the authors of
[DOI:10.1038/s41587-019-0392-8](https://doi.org/10.1038/s41587-019-0392-8) and
[DOI:10.1093/nar/gkab043](https://academic.oup.com/nar/article/49/9/e50/6129341).

Due to limits of file size, we prepared only the following ASURAT object:
```{r, eval = FALSE}
pdac_ast1 <- readRDS(file = "backup/05_001_pdac_ast1_raw.rds")
```



### PDAC-A inDrop
From [GSE111672](https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE111672),
we downloaded spatial transcriptome data with sample accession numbers
GSM3036909, GSM3036910, GSM3405527, GSM3405528, GSM3405529, and GSM3405530
(for PDAC-A inDrop1-inDrop6).

Due to limits of file size, we prepared only the following ASURAT object:
```{r, eval = FALSE}
pdac_arna <- readRDS(file = "backup/06_001_pdac_arna_raw.rds")
```



## Case III: input 10x PBMC data
```{r, eval = FALSE}
source("R/function_general.R")
```

The data (peripheral blood mononuclear cells from a healthy donor) were
obtained from 10x Genomics repository (PBMC 4k and 6k).

Due to limits of file size, we prepared only the following ASURAT objects:
```{r, eval = FALSE}
# ----------------------------------------
# PBMC 4k
# ----------------------------------------
pbmc_4000 <- readRDS(file = "backup/10_001_pbmc_4000_raw.rds")
# ----------------------------------------
# PBMC 6k
# ----------------------------------------
pbmc_6000 <- readRDS(file = "backup/11_001_pbmc_6000_raw.rds")
```
