# Collecting databases (optional) {#collectdb}
In this section, the following databases (DBs) are downloaded and properly
formatted.
Since all of the results in this section have already been prepared for human
and mouse datasets in December 2020 (see `data/2020_001_databases/`),
users may be able to skip this section.

1. DB for Disease Ontology (DO) for human disease
2. DB for Cell Ontology (CO)
3. DB for Gene Ontology (GO)
4. DB for Kyoto Encyclopedia of Genes and Genomes (KEGG)
5. DB for Reactome



## Disease Ontology DB
Load the data.
```{r, eval = FALSE}
rm(list = ls())
source("R/function_DiseaseOntology.R")
```

The following functions `collect_DO()`, `format_DO()`, and `compute_IC_DO()`
(i) collects a database for Disease Ontology,
(ii) formats the results, and
(iii) prepares a list of information content (IC), respectively.
The definition of IC is reintroduced in Mistry and Pavlidis,
BMC Bioinformatics, 2008, but note that the values depend on the data set.
In fact, the ICs in `IC_DO` may not equal to those computed in `doSim()`
in `DOSE` package.
```{r, eval = FALSE}
#
# Note again that the following result has already been prepared for human and
# mouse datasets in December 2020 (see `data/2020_001_databases/`).
#
library(DO.db)
dict_DO <- collect_DO()
tidy_DO <- format_DO(dict = dict_DO)
IC_DO <- compute_IC_DO(tidy = tidy_DO)
#
# Save the result.
#
saveRDS(dict_DO, file = "data/2020_001_databases/YYYYMMDD_dict_DO_human.rds")
saveRDS(tidy_DO, file = "data/2020_001_databases/YYYYMMDD_tidy_DO_human.rds")
saveRDS(IC_DO, file = "data/2020_001_databases/YYYYMMDD_IC_DO_human.rds")
```



## Cell Ontology DB
Load the data.
```{r, eval = FALSE}
rm(list = ls())
source("R/function_CellOntology.R")
```

The following functions `collect_CO()`, `format_CO()`, `make_treeTable_CO()`,
and `compute_IC_CO()`
(i) collects a database for Cell Ontology,
(ii) formats the results,
(iii) prepares a corresponding table between child and parent CO IDs, and
(iv) prepares a list of information content (IC), respectively.

**Tips:**
Currently, in December 2020, the following Cell Ontology DB seems to be immature
because, for example, the well-known marker genes for pancreatic beta cell,
Ins1 and Ins2, are not registered in the gene slot of `type B pancreatic cell`
with CO ID `CL:0000169`.
In addition, the IC seems to be meaningless.
Nevertheless, we computed the IC for future direction.

```{r, eval = FALSE}
#
# Note again that the following result has already been prepared for human and
# mouse datasets in December 2020 (see `data/2020_001_databases/`).
#
dict_CO <- collect_CO(orgdb = org.Hs.eg.db)
tidy_CO <- format_CO(dict = dict_CO)
treeTable_CO <- make_treeTable_CO(tidy = tidy_CO)
IC_CO <- compute_IC_CO(tidy = tidy_CO, treeTable = treeTable_CO)
#
# Save the result.
#
saveRDS(dict_CO, file = "data/2020_001_databases//YYYYMMDD_dict_CO_human.rds")
saveRDS(tidy_CO, file = "data/2020_001_databases//YYYYMMDD_tidy_CO_human.rds")
saveRDS(treeTable_CO,
        file = "data/2020_001_databases//YYYYMMDD_treeTable_CO_human.rds")
saveRDS(IC_CO, file = "data/2020_001_databases//YYYYMMDD_IC_CO_human.rds")
```



## Gene Ontology DB
Load the data.
```{r, eval = FALSE}
rm(list = ls())
source("R/function_GeneOntology.R")
```

The following functions `collect_GO()`, `format_GO()`, and `compute_IC_GO()`
(i) collects a database for Gene Ontology,
(ii) formats the results, and
(iii) prepares a list of information content (IC), respectively.
The arguments of `collect_GO()` are `orgdb`
(annotation data package such as `org.Hs.eg.db` and `org.Mm.eg.db`) and
`all_geneIDs` (all Entrez IDs of a specific organism).
The definition of IC is reintroduced in Mistry and Pavlidis,
BMC Bioinformatics, 2008, but note that the values depend on the data set.
In fact, the ICs in `IC_GO` may not equal to those computed in `mgoSim()` in
`GOSemSim` package.
```{r, eval = FALSE}
#
# Note again that the following result has already been prepared for human and
# mouse datasets in December 2020 (see `data/2020_001_databases/`).
#
library(GO.db)
all_geneIDs <- as.list(org.Hs.egGO)
all_geneIDs <- names(all_geneIDs[!is.na(all_geneIDs)])
dict_GO <- collect_GO(orgdb = org.Hs.eg.db, all_geneIDs = all_geneIDs)
tidy_GO <- format_GO(dict = dict_GO)
IC_GO <- compute_IC_GO(tidy = tidy_GO)
#
# Save the result.
#
saveRDS(dict_DO, file = "data/2020_001_databases/YYYYMMDD_dict_GO_human.rds")
saveRDS(tidy_DO, file = "data/2020_001_databases/YYYYMMDD_tidy_GO_human.rds")
saveRDS(IC_DO, file = "data/2020_001_databases/YYYYMMDD_IC_GO_human.rds")
```



## KEGG
Load the data.
```{r, eval = FALSE}
rm(list = ls())
source("R/function_KEGG.R")
```

The following functions `collect_KEGG()` and `format_KEGG()`
(i) collects a database for KEGG and (ii) formats the results, respectively.
In the current version, KEGG PATHWAY (`"pathway"`) and MODULE (`"module"`)
can be input into `collect_KEGG()` (cf. `KEGGREST` function `listDatabases()`).
The arguments of `collect_KEGG()` are `organism` (such as Homo sapiens `"hsa"`,
obeying the naming rule of [KEGG](http://rest.kegg.jp/list/organism)) and
`category_names` (such as `"pathway"` and `"module"`).
This function requires an internet connection.
```{r, eval = FALSE}
#
# Note again that the following result has already been prepared for human and
# mouse datasets in December 2020 (see `data/2020_001_databases/`).
#
# Note that `collect_KEGG` uses `KEGGREST` function `keggGet` with internet
# connection, which may produce both successful and unsuccessful results.
#
dict_KEGG_raw <- collect_KEGG(organism = "hsa", category_names = c("pathway"))
dict_KEGG <- list(pathway = dict_KEGG_raw[["pathway"]][["success"]])
tidy_KEGG <- format_KEGG(dict = dict_KEGG)
#
# Save the result.
#
saveRDS(dict_KEGG_raw,
        file = "data/2020_001_databases/YYYYMMDD_dict_KEGG_raw_human.rds")
saveRDS(dict_KEGG,
        file = "data/2020_001_databases/YYYYMMDD_dict_KEGG_human.rds")
saveRDS(tidy_KEGG,
        file = "data/2020_001_databases/YYYYMMDD_tidy_KEGG_human.rds")
```



## Reactome
Load the data.
```{r, eval = FALSE}
rm(list = ls())
source("R/function_Reactome.R")
```

The following functions `collect_Reactome()` and `format_Reactome()`
(i) collects a database for Reactome and
(ii) formats the results, respectively.
In the current version, Reactome pathway (`"PATHID"`) can be input into
`collect_Reactome()` (cf. `reactome.db` functions `ls("package:reactome.db")`
and `keytypes(reactome.db)`).
The arguments are `organism` (such as Homo sapiens `Homo sapiens`, obeying the
naming rule of `reactome.db`) and `category_names` (such as `PATHID`).
```{r, eval = FALSE}
#
# Note again that the following result has already been prepared for human and
# mouse datasets in December 2020 (see `data/2020_001_databases/`).
#
# Note that `collect_Reactome` may produce both successful and unsuccessful
# results.
#
dict_Reactome_raw <- collect_Reactome(organism = "Homo sapiens",
                                      category_names = c("PATHID"))
dict_Reactome <- list(pathid = dict_Reactome_raw[["PATHID"]][["success"]])
tidy_Reactome <- format_Reactome(dict = dict_Reactome)
#
# Save the result.
#
saveRDS(dict_Reactome_raw,
        file = "data/2020_001_databases/YYYYMMDD_dict_Reactome_raw_human.rds")
saveRDS(dict_Reactome,
        file = "data/2020_001_databases/YYYYMMDD_dict_Reactome_human.rds")
saveRDS(tidy_Reactome,
        file = "data/2020_001_databases/YYYYMMDD_tidy_Reactome_human.rds")
```
