# ASURAT using Disease Ontology database
In this section, we apply ASURAT on human cancer transcriptome.
ASURAT transforms a single-cell transcriptome to cell-type or functional
spectrum matrix by extracting the system of signs on each of the thousands of
biological terms and the associated gene sets obtained from public databases.
The resulting matrices termed sign-by-sample matrices (SSMs) can be supplied
to the subsequent unsupervised clusterings.

**Tips:**
Disease Ontology database is basically used for analyses of human disease
sample.
Users should select proper databases depending on their data.



## Creating sign-by-sample matrices
Load the data.
```{r, eval = FALSE}
source("R/plot.R")
source("R/function_sign.R")
tidy_DO <- readRDS(file = "data/2020_001_databases/20201213_tidy_DO_human.rds")
IC_DO <- readRDS(file = "data/2020_001_databases/20201213_IC_DO_human.rds")

day1_norm <- readRDS(file = "backup/01_005_day1_norm_normalized.rds")
day7_hypo <- readRDS(file = "backup/02_005_day7_hypo_normalized.rds")
sc68_vehi <- readRDS(file = "backup/03_005_sc68_vehi_normalized.rds")
sc68_cisp <- readRDS(file = "backup/04_005_sc68_cisp_normalized.rds")

day1_norm_cor <- readRDS(file = "backup/01_006_day1_norm_correlation.rds")
day7_hypo_cor <- readRDS(file = "backup/02_006_day7_hypo_correlation.rds")
sc68_vehi_cor <- readRDS(file = "backup/03_006_sc68_vehi_correlation.rds")
sc68_cisp_cor <- readRDS(file = "backup/04_006_sc68_cisp_correlation.rds")
```

To make the object small, unused items are removed.
```{r, eval = FALSE}
myfunc <- function(obj){
  obj[["data"]][["raw"]] <- NULL
  obj[["data"]][["bayNorm"]] <- NULL
  return(obj)
}

day1_norm <- myfunc(day1_norm)
day7_hypo <- myfunc(day7_hypo)
sc68_vehi <- myfunc(sc68_vehi)
sc68_cisp <- myfunc(sc68_cisp)
```

Input a formatted Disease Ontology database `tidy_DO` into the ASURAT object.
```{r, eval = FALSE}
day1_norm[["sign"]][["DO"]] <- tidy_DO
day7_hypo[["sign"]][["DO"]] <- tidy_DO
sc68_vehi[["sign"]][["DO"]] <- tidy_DO
sc68_cisp[["sign"]][["DO"]] <- tidy_DO
```

The following function `do_quickQC_sign()` takes an intersection of genes
between user's data and `tidy_DO`, and then excludes the IDs including too
few or too many genes by setting thresholds `min_ngenes` (minimal number of
genes which must be greater than one) and `max_ngenes` (maximal number of genes
which must be greater than one).
Here, the values of `min_ngenes` and `max_ngenes` were set as 2 and 1000,
respectively, as default values.
```{r, eval = FALSE}
day1_norm <- do_quickQC_sign(obj = day1_norm, data_type = "DO",
                             min_ngenes = 2, max_ngenes = 1000)
day7_hypo <- do_quickQC_sign(obj = day7_hypo, data_type = "DO",
                             min_ngenes = 2, max_ngenes = 1000)
sc68_vehi <- do_quickQC_sign(obj = sc68_vehi, data_type = "DO",
                             min_ngenes = 2, max_ngenes = 1000)
sc68_cisp <- do_quickQC_sign(obj = sc68_cisp, data_type = "DO",
                             min_ngenes = 2, max_ngenes = 1000)
```
The results are stored in `obj[["sign"]][["DO"]][["disease"]]`.

The following function `separate_variables_sign()` classifies genes annotated
with each biological description into three groups, i.e., strongly correlated
gene set (SCG), variably correlated gene set (VCG), and weakly correlated gene
set (WCG).
The arguments are `obj` (ASURAT object), `obj_cor`, `data_type`,
`method = "spearman"` (one of the arguments input in `do_cor_variables()`),
`th_posi`, and `th_nega` (threshold values of positive and negative correlation
coefficients, respectively).

**Tips:**
This decomposition sometimes produces unexpected results.
In fact the smaller values of `th_posi` and `th_nega` do not always result in
the larger sizes of SCG and VCG.
Hence we recommend users to carefully tune the parameters, checking if the
biological terms (signs) of interest are included in the results.
```{r, eval = FALSE}
day1_norm_red <- separate_variables_sign(obj = day1_norm,
                                         obj_cor = day1_norm_cor,
                                         data_type = "DO",
                                         method = "spearman",
                                         th_posi = 0.40, th_nega = -0.20)
day7_hypo_red <- separate_variables_sign(obj = day7_hypo,
                                         obj_cor = day7_hypo_cor,
                                         data_type = "DO",
                                         method = "spearman",
                                         th_posi = 0.43, th_nega = -0.24)
sc68_vehi_red <- separate_variables_sign(obj = sc68_vehi,
                                         obj_cor = sc68_vehi_cor,
                                         data_type = "DO",
                                         method = "spearman",
                                         th_posi = 0.28, th_nega = -0.24)
sc68_cisp_red <- separate_variables_sign(obj = sc68_cisp,
                                         obj_cor = sc68_cisp_cor,
                                         data_type = "DO",
                                         method = "spearman",
                                         th_posi = 0.28, th_nega = -0.22)
```
The result is stored in `obj[["sign"]][["DO"]][["disease"]]`.

The following function `select_sign()` selects the signs by a user-defined
criteria, which has been preset as follows:

1. sum of the number of genes in SCG and VCG is greater than or equal to
`min_cnt`, and
2. the number of genes in WCG is greater than or equal to `min_cnt_weak`
(the default value is 2).

Users can modify the programming code indicated by "User-defined criteria"
in the file `R/function_sign.R`.
Note that the signs having at most one gene will be removed when creating
sign-by-sample matrices.
For example, given a certain parent sign having SCG = \{GENE_A, GENE_B\} and
VCG = \{GENE_C\}, the VCG will be removed from the downstream analysis.

**Tips:**
The larger the value of `min_cnt` is, the more reliable signs users can obtain.
However, one should pay attention to several bias problems such as annotation
bias, that is some biological terms are associated with many genes, while
others with few.
```{r, eval = FALSE}
day1_norm_red <- select_sign(obj = day1_norm_red, data_type = "DO",
                             min_cnt = 2, min_cnt_weak = 2)
day7_hypo_red <- select_sign(obj = day7_hypo_red, data_type = "DO",
                             min_cnt = 3, min_cnt_weak = 2)
sc68_vehi_red <- select_sign(obj = sc68_vehi_red, data_type = "DO",
                             min_cnt = 2, min_cnt_weak = 2)
sc68_cisp_red <- select_sign(obj = sc68_cisp_red, data_type = "DO",
                             min_cnt = 3, min_cnt_weak = 2)
```
The result is stored in `obj[["sign"]][["DO"]][["disease"]]`.

Save the objects.
```{r, eval = FALSE}
saveRDS(day1_norm_red, file = "backup/01_100_day1_norm_select_DO.rds")
saveRDS(day7_hypo_red, file = "backup/02_100_day7_hypo_select_DO.rds")
saveRDS(sc68_vehi_red, file = "backup/03_100_sc68_vehi_select_DO.rds")
saveRDS(sc68_cisp_red, file = "backup/04_100_sc68_cisp_select_DO.rds")
```

Load the objects.
```{r, eval = FALSE}
day1_norm_red <- readRDS(file = "backup/01_100_day1_norm_select_DO.rds")
day7_hypo_red <- readRDS(file = "backup/02_100_day7_hypo_select_DO.rds")
sc68_vehi_red <- readRDS(file = "backup/03_100_sc68_vehi_select_DO.rds")
sc68_cisp_red <- readRDS(file = "backup/04_100_sc68_cisp_select_DO.rds")
```

The following function `compute_SemSim_sign()` computes pairwise semantic
similarities between biological descriptions.
The arguments are `obj` (ASURAT object), `data_type`
(the name of database used), `measure` (argument of `doSim()` in `DOSE` package:
one of `"Wang"`, `"Resnik"`, `"Rel"`, `"Jiang"`, and `"Lin"`), `orgdb = NULL`,
`treeTable = NULL`, and `IC = IC_DO`.
Note that the results are stored into `obj[["sign"]][["DO_Sim"]]`, but the row
and column names are in ascending order with respect to the values of
information content (IC).

**Tips:**
The similarity between biological terms are computed based on `measure`, and
the optimal choice depends on the ontology structure.
```{r, eval = FALSE}
day1_norm_red <- compute_SemSim_sign(obj = day1_norm_red, data_type = "DO",
                                     measure = "Jiang", orgdb = NULL,
                                    treeTable = NULL, IC = IC_DO)
day7_hypo_red <- compute_SemSim_sign(obj = day7_hypo_red, data_type = "DO",
                                     measure = "Jiang", orgdb = NULL,
                                     treeTable = NULL, IC = IC_DO)
sc68_vehi_red <- compute_SemSim_sign(obj = sc68_vehi_red, data_type = "DO",
                                     measure = "Jiang", orgdb = NULL,
                                     treeTable = NULL, IC = IC_DO)
sc68_cisp_red <- compute_SemSim_sign(obj = sc68_cisp_red, data_type = "DO",
                                     measure = "Jiang", orgdb = NULL,
                                     treeTable = NULL, IC = IC_DO)
```
The result is stored in `obj[["sign"]][["DO"]][["disease"]]`.

The following function `reduce_sign()` removes the biological terms based on the
results of `compute_SemSim_sign()`.
The arguments are `obj`, `data_type`, `threshold` (threshold for regarding
two biological terms as similar ones), and `keep_rareID` (if `"TRUE"`, then the
biological terms with larger ICs are kept).
Note that the larger ICs are, the deeper the biological terms tend to be.
Users can access a correspondence table in `obj[["sign"]][["DO_Sim"]]`, where
SCG, VCG, and WCG are separated by `||`.

**Tips:**
The optimal value of `threshold` depends on the ontology structure as well as
the measure in `compute_SemSim_sign()`.
The logical argument `keep_rareID` should be carefully determined by the
research objective.
For example, if users will use DO for detailed sample annotation or
classification, it might be `keep_rareID = TRUE`.
Note that the smaller ICs of DO terms are, the more genes tend to be associated
with.
```{r, eval = FALSE}
day1_norm_red <- reduce_sign(obj = day1_norm_red, data_type = "DO",
                             threshold = 0.85, keep_rareID = FALSE)
day7_hypo_red <- reduce_sign(obj = day7_hypo_red, data_type = "DO",
                             threshold = 0.85, keep_rareID = TRUE)
sc68_vehi_red <- reduce_sign(obj = sc68_vehi_red, data_type = "DO",
                             threshold = 0.80, keep_rareID = TRUE)
sc68_cisp_red <- reduce_sign(obj = sc68_cisp_red, data_type = "DO",
                             threshold = 0.85, keep_rareID = TRUE)
```
The result is stored in `obj[["sign"]][["DO_Sim"]][["disease"]][["report"]]`.

The following function `manual_curation_sign()` removes biological terms by
specifying the IDs or keywords.
The argument are `obj`, `data_type`, and `keywords` (list of keywords
separated by `|`).

**Tips:**
Users may take into account the following points:

1. correlation coefficients (small value means that the sign has low relevance)
2. information contents (ICs) (small value means that the biological term
is less specific)
3. ratios of mitochondrial genes for SCG and VCG
4. biological terms which are obviously unrelated to the research objective

```{r, eval = FALSE}
keywords <- "foo|foofoo"
day1_norm_red <- manual_curation_sign(obj = day1_norm_red, data_type = "DO",
                                      keywords = keywords)
keywords <- "foo|foofoo"
day7_hypo_red <- manual_curation_sign(obj = day7_hypo_red, data_type = "DO",
                                      keywords = keywords)
keywords <- "foo|foofoo"
sc68_vehi_red <- manual_curation_sign(obj = sc68_vehi_red, data_type = "DO",
                                      keywords = keywords)
keywords <- "foo|foofoo"
sc68_cisp_red <- manual_curation_sign(obj = sc68_cisp_red, data_type = "DO",
                                      keywords = keywords)
```
The result is stored in `obj[["sign"]][["DO"]][["disease"]]`.

The following function `make_signxsample_matrix()` makes sign-by-sample
matrices.
The arguments are `obj`, `data_type`, `weight_strg` (weight parameter for SCG:
the default value is 0.5) and `weight_vari` (weight parameter for VCG:
the default value is 0.5).
```{r, eval = FALSE}
day1_norm_red <- make_signxsample_matrix(obj = day1_norm_red, data_type = "DO",
                                         weight_strg = 0.5, weight_vari = 0.5)
day7_hypo_red <- make_signxsample_matrix(obj = day7_hypo_red, data_type = "DO",
                                         weight_strg = 0.5, weight_vari = 0.5)
sc68_vehi_red <- make_signxsample_matrix(obj = sc68_vehi_red, data_type = "DO",
                                         weight_strg = 0.5, weight_vari = 0.5)
sc68_cisp_red <- make_signxsample_matrix(obj = sc68_cisp_red, data_type = "DO",
                                         weight_strg = 0.5, weight_vari = 0.5)
```
The result is stored in `obj[["sign"]][["DOxSample"]][["disease"]]`.

The following command `plot_Heatmap_SignxSamp()` plots a heat map by using
`Heatmap()` in `ComplexHeatmap` package applied to the sign-by-sample matrices.
The arguments are `obj`, `data_type`, `category`, `algo_name = NULL`,
`method` (agglomeration method to be used such as `complete`, `ward.D2`,
`average`, etc.), and `show_nReads` (if `TRUE`, the number of reads is shown).
```{r, eval = FALSE}
# ----------------------------------------
# day1_norm
# ----------------------------------------
filename <- "figures/figure_01_0100.png"
png(file = filename, height = 1000, width = 1150, res = 300)
plot_Heatmap_SignxSamp(obj = day1_norm_red,
                       data_type = "DO", category = "disease",
                       algo_name = NULL, method = "ward.D2",
                       show_nReads = TRUE, title = "day1_norm (DO: disease)",
                       name = "DOxSamp", show_rownames_sign = FALSE,
                       show_rownames_label = FALSE,
                       show_rownames_nReads = FALSE, default_color = TRUE)
dev.off()
```

```{r, eval = FALSE, echo = FALSE, results = "hide"}
# ----------------------------------------
# day7_hypo
# ----------------------------------------
filename <- "figures/figure_02_0100.png"
png(file = filename, height = 1000, width = 1150, res = 300)
plot_Heatmap_SignxSamp(obj = day7_hypo_red,
                       data_type = "DO", category = "disease",
                       algo_name = NULL, method = "ward.D2",
                       show_nReads = TRUE, title = "day7_hypo (DO: disease)",
                       name = "DOxSamp", show_rownames_sign = FALSE,
                       show_rownames_label = FALSE,
                       show_rownames_nReads = FALSE, default_color = TRUE)
dev.off()
# ----------------------------------------
# sc68_vehi
# ----------------------------------------
filename <- "figures/figure_03_0100.png"
png(file = filename, height = 1000, width = 1150, res = 300)
plot_Heatmap_SignxSamp(obj = sc68_vehi_red,
                       data_type = "DO", category = "disease",
                       algo_name = NULL, method = "ward.D2",
                       show_nReads = TRUE, title = "sc68_vehi (DO: disease)",
                       name = "DOxSamp", show_rownames_sign = FALSE,
                       show_rownames_label = FALSE,
                       show_rownames_nReads = FALSE, default_color = TRUE)
dev.off()
# ----------------------------------------
# sc68_cisp
# ----------------------------------------
filename <- "figures/figure_04_0100.png"
png(file = filename, height = 1000, width = 1150, res = 300)
plot_Heatmap_SignxSamp(obj = sc68_cisp_red,
                       data_type = "DO", category = "disease",
                       algo_name = NULL, method = "ward.D2",
                       show_nReads = TRUE, title = "sc68_cisp (DO: disease)",
                       name = "DOxSamp", show_rownames_sign = FALSE,
                       show_rownames_label = FALSE,
                       show_rownames_nReads = FALSE, default_color = TRUE)
dev.off()
```

<img src="figures/figure_01_0100.png" width="300px">
<img src="figures/figure_02_0100.png" width="300px">

<img src="figures/figure_03_0100.png" width="300px">
<img src="figures/figure_04_0100.png" width="300px">

Save the objects.
```{r, eval = FALSE}
saveRDS(day1_norm_red, file = "backup/01_101_day1_norm_matrix_DO.rds")
saveRDS(day7_hypo_red, file = "backup/02_101_day7_hypo_matrix_DO.rds")
saveRDS(sc68_vehi_red, file = "backup/03_101_sc68_vehi_matrix_DO.rds")
saveRDS(sc68_cisp_red, file = "backup/04_101_sc68_cisp_matrix_DO.rds")
```



## Dimensionality reduction
Load the objects.
```{r, eval = FALSE}
rm(list = ls())
source("R/plot.R")
source("R/function_sign.R")
day1_norm <- readRDS(file = "backup/01_101_day1_norm_matrix_DO.rds")
day7_hypo <- readRDS(file = "backup/02_101_day7_hypo_matrix_DO.rds")
sc68_vehi <- readRDS(file = "backup/03_101_sc68_vehi_matrix_DO.rds")
sc68_cisp <- readRDS(file = "backup/04_101_sc68_cisp_matrix_DO.rds")
```

The following function `do_pca_sign()` performs principal component analysis
based on the sign-by-sample matrices of the input `data_type` and `category`.
```{r, eval = FALSE}
day1_norm <- do_pca_sign(obj = day1_norm,
                         data_type = "DO", category = "disease")
day7_hypo <- do_pca_sign(obj = day7_hypo,
                         data_type = "DO", category = "disease")
sc68_vehi <- do_pca_sign(obj = sc68_vehi,
                         data_type = "DO", category = "disease")
sc68_cisp <- do_pca_sign(obj = sc68_cisp,
                         data_type = "DO", category = "disease")
```
The result is stored in `obj[["reduction"]][["pca"]][["DO"]][["disease"]]`.

The following function `do_tsne_sign()` performs `Rtsne()` in `Rtsne` package
based on the sign-by-sample matrices of the input `data_type` and `category`.
The arguments are `obj` (ASURAT object), `data_type`, `category`,
`pca_dim` (dimension of principal component used for t-SNE, in which `NULL` is
accepted if users would like to compute t-SNE from the original sample-by-sign
matrices), and `tsne_dim` (t-SNE dimension: either `2` or `3`).
```{r, eval = FALSE}
day1_norm <- do_tsne_sign(obj = day1_norm, data_type = "DO",
                          category = "disease", pca_dim = NULL, tsne_dim = 2)
day7_hypo <- do_tsne_sign(obj = day7_hypo, data_type = "DO",
                          category = "disease", pca_dim = NULL, tsne_dim = 2)
sc68_vehi <- do_tsne_sign(obj = sc68_vehi, data_type = "DO",
                          category = "disease", pca_dim = NULL, tsne_dim = 2)
sc68_cisp <- do_tsne_sign(obj = sc68_cisp, data_type = "DO",
                          category = "disease", pca_dim = NULL, tsne_dim = 2)
```
The result is stored in `obj[["reduction"]][["tsne"]][["DO"]][["disease"]]`.

The following function `plot_tsne_sign()` shows the data in a t-SNE space,
in which the space dimension is automatically given by `tsne_dim` used in
`do_tsne_sign()`.
The arguments are `obj`, `data_type`, `category`, `algo_name = NULL`, `theta`
(angle of view for 3D plots), and `phi` (angle of view for 3D plots).
```{r, eval = FALSE}
# ----------------------------------------
# day1_norm
# ----------------------------------------
p <- plot_tsne_sign(obj = day1_norm, data_type = "DO", category = "disease",
                    algo_name = NULL, theta = NULL, phi = NULL,
                    title = "day1_norm (DO: disease)", title_size = 18,
                    xlabel = "tSNE_1", ylabel = "tSNE_2", zlabel = NULL,
                    default_color = TRUE)
filename <- "figures/figure_01_0105.png"
ggsave(file = filename, plot = p, dpi = 300, width = 4.3, height = 4.5)
```

```{r, eval = FALSE, echo = FALSE, results = "hide"}
# ----------------------------------------
# day7_hypo
# ----------------------------------------
p <- plot_tsne_sign(obj = day7_hypo, data_type = "DO", category = "disease",
                    algo_name = NULL, theta = NULL, phi = NULL,
                    title = "day7_hypo (DO: disease)", title_size = 18,
                    xlabel = "tSNE_1", ylabel = "tSNE_2", zlabel = NULL,
                    default_color = TRUE)
filename <- "figures/figure_02_0105.png"
ggsave(file = filename, plot = p, dpi = 300, width = 4.3, height = 4.5)
# ----------------------------------------
# sc68_vehi
# ----------------------------------------
p <- plot_tsne_sign(obj = sc68_vehi, data_type = "DO", category = "disease",
                    algo_name = NULL, theta = NULL, phi = NULL,
                    title = "sc68_vehi (DO: disease)", title_size = 18,
                    xlabel = "tSNE_1", ylabel = "tSNE_2", zlabel = NULL,
                    default_color = TRUE)
filename <- "figures/figure_03_0105.png"
ggsave(file = filename, plot = p, dpi = 300, width = 4.3, height = 4.5)
# ----------------------------------------
# sc68_cisp
# ----------------------------------------
p <- plot_tsne_sign(obj = sc68_cisp, data_type = "DO", category = "disease",
                    algo_name = NULL, theta = NULL, phi = NULL,
                    title = "sc68_cisp (DO: disease)", title_size = 18,
                    xlabel = "tSNE_1", ylabel = "tSNE_2", zlabel = NULL,
                    default_color = TRUE)
filename <- "figures/figure_04_0105.png"
ggsave(file = filename, plot = p, dpi = 300, width = 4.3, height = 4.5)
```

<img src="figures/figure_01_0105.png" width="150px">
<img src="figures/figure_02_0105.png" width="150px">
<img src="figures/figure_03_0105.png" width="150px">
<img src="figures/figure_04_0105.png" width="150px">

The following function `do_umap_sign()` performs `umap()` in `umap` package
based on the sign-by-sample matrices of the input `data_type` and `category`.
The arguments are `obj`, `data_type`, `category`, `pca_dim` (dimension of
principal component used for UMAP, in which `NULL` is accepted if users would
like to compute UMAP from the original sample-by-sign matrices), and
`umap_dim` (UMAP dimension: either `2` or `3`).
```{r, eval = FALSE}
day1_norm <- do_umap_sign(obj = day1_norm, data_type = "DO",
                          category = "disease", pca_dim = NULL, umap_dim = 2)
day7_hypo <- do_umap_sign(obj = day7_hypo, data_type = "DO",
                          category = "disease", pca_dim = NULL, umap_dim = 2)
sc68_vehi <- do_umap_sign(obj = sc68_vehi, data_type = "DO",
                          category = "disease", pca_dim = NULL, umap_dim = 2)
sc68_cisp <- do_umap_sign(obj = sc68_cisp, data_type = "DO",
                          category = "disease", pca_dim = NULL, umap_dim = 2)
```
The result is stored in `obj[["reduction"]][["umap"]][["DO"]][["disease"]]`.

The following function `plot_umap_sign()` shows the data in a UMAP space,
in which the space dimension is automatically given by `umap_dim` used in
`do_umap_sign()`.
The arguments are `obj`, `data_type`, `category`, `algo_name = NULL`, `theta`
(angle of view for 3D plots), and `phi` (angle of view for 3D plots).
```{r, eval = FALSE}
# ----------------------------------------
# day1_norm
# ----------------------------------------
p <- plot_umap_sign(obj = day1_norm, data_type = "DO", category = "disease",
                    algo_name = NULL, theta = NULL, phi = NULL,
                    title = "day1_norm (DO: disease)", title_size = 18,
                    xlabel = "UMAP_1", ylabel = "UMAP_2", zlabel = NULL,
                    default_color = FALSE)
filename <- "figures/figure_01_0106.png"
ggsave(file = filename, plot = p, dpi = 300, width = 4.3, height = 4.5)
```

```{r, eval = FALSE, echo = FALSE, results = "hide"}
# ----------------------------------------
# day7_hypo
# ----------------------------------------
p <- plot_umap_sign(obj = day7_hypo, data_type = "DO", category = "disease",
                    algo_name = NULL, theta = NULL, phi = NULL,
                    title = "day7_hypo (DO: disease)", title_size = 18,
                    xlabel = "UMAP_1", ylabel = "UMAP_2", zlabel = NULL,
                    default_color = FALSE)
filename <- "figures/figure_02_0106.png"
ggsave(file = filename, plot = p, dpi = 300, width = 4.3, height = 4.5)
# ----------------------------------------
# sc68_vehi
# ----------------------------------------
p <- plot_umap_sign(obj = sc68_vehi, data_type = "DO", category = "disease",
                    algo_name = NULL, theta = NULL, phi = NULL,
                    title = "sc68_vehi (DO: disease)", title_size = 18,
                    xlabel = "UMAP_1", ylabel = "UMAP_2", zlabel = NULL,
                    default_color = FALSE)
filename <- "figures/figure_03_0106.png"
ggsave(file = filename, plot = p, dpi = 300, width = 4.3, height = 4.5)
# ----------------------------------------
# sc68_cisp
# ----------------------------------------
p <- plot_umap_sign(obj = sc68_cisp, data_type = "DO", category = "disease",
                    algo_name = NULL, theta = NULL, phi = NULL,
                    title = "sc68_cisp (DO: disease)", title_size = 18,
                    xlabel = "UMAP_1", ylabel = "UMAP_2", zlabel = NULL,
                    default_color = FALSE)
filename <- "figures/figure_04_0106.png"
ggsave(file = filename, plot = p, dpi = 300, width = 4.3, height = 4.5)
```

<img src="figures/figure_01_0106.png" width="150px">
<img src="figures/figure_02_0106.png" width="150px">
<img src="figures/figure_03_0106.png" width="150px">
<img src="figures/figure_04_0106.png" width="150px">

The following function `do_DiffusionMap_sign()` performs `DiffusionMap()` using
`destiny` package, where the arguments are preset as `sigma = "local"` and
`distance = "euclidean"`.
The arguments are `obj`, `data_type`, `category`, `sigma`
(argument of `DiffusionMap()`), `distance` (argument of `DiffusionMap()`), and
`pca_dim` (an integer: dimension of principal component used for UMAP, in which
`NULL` is accepted if users would like to compute UMAP from the original
sample-by-sign matrices).

**Tips:**
If the resulting diffusion map is distorted by outliers, tuning of `pca_dim`
will be required.
```{r, eval = FALSE}
day1_norm <- do_DiffusionMap_sign(obj = day1_norm, data_type = "DO",
                                  category = "disease", pca_dim = NULL)
day7_hypo <- do_DiffusionMap_sign(obj = day7_hypo, data_type = "DO",
                                  category = "disease", pca_dim = NULL)
sc68_vehi <- do_DiffusionMap_sign(obj = sc68_vehi, data_type = "DO",
                                  category = "disease", pca_dim = NULL)
sc68_cisp <- do_DiffusionMap_sign(obj = sc68_cisp, data_type = "DO",
                                  category = "disease", pca_dim = NULL)
```
The result is stored in `obj[["reduction"]][["dmap"]][["DO"]][["disease"]]`.

The following function `plot_DiffusionMap_sign()` shows the data in a diffusion
space.
The arguments are `obj`, `data_type`, `category`, `algo_name = NULL`, `dims`
(dimension vector for the plot), `theta` (angle of view), and
`phi` (angle of view).
```{r, eval = FALSE}
# ----------------------------------------
# day7_hypo
# ----------------------------------------
filename <- "figures/figure_02_0110.png"
png(file = filename, height = 1200, width = 1200, res = 250)
plot_DiffusionMap_sign(obj = day7_hypo, data_type = "DO", category = "disease",
                       algo_name = NULL, dims = 1:3, theta = 90, phi = 30,
                       title = "day7_hypo (DO: disease)", title_size = 1.2,
                       xlabel = "DC_1", ylabel = "DC_2", zlabel = "DC_3",
                       default_color = TRUE)
dev.off()
#
# Below is a memo for using plot_ly()
#
# library(plotly)
# x <- day7_hypo[["reduction"]][["dmap"]][["DO"]][["disease"]]@eigenvectors
# df <- as.data.frame(x)
# plot_ly(df, x = df[,1], y = df[,2], z = df[,3], mode = "markers",
# type = "scatter3d", marker = list(opacity = 1, showlegend = T, size = 2))
```

```{r, eval = FALSE, echo = FALSE, results = "hide"}
# ----------------------------------------
# sc68_cisp
# ----------------------------------------
filename <- "figures/figure_04_0110.png"
png(file = filename, height = 1200, width = 1200, res = 250)
plot_DiffusionMap_sign(obj = sc68_cisp, data_type = "DO", category = "disease",
                       algo_name = NULL, dims = 1:3, theta = 0, phi = 30,
                       title = "sc68_cisp (DO: disease)", title_size = 1.2,
                       xlabel = "DC_1", ylabel = "DC_2", zlabel = "DC_3",
                       default_color = TRUE)
dev.off()
```

<img src="figures/figure_02_0110.png" width="250px">
<img src="figures/figure_04_0110.png" width="250px">

Save the objects.
```{r, eval = FALSE}
saveRDS(day1_norm, file = "backup/01_102_day1_norm_reduction_DO.rds")
saveRDS(day7_hypo, file = "backup/02_102_day7_hypo_reduction_DO.rds")
saveRDS(sc68_vehi, file = "backup/03_102_sc68_vehi_reduction_DO.rds")
saveRDS(sc68_cisp, file = "backup/04_102_sc68_cisp_reduction_DO.rds")
```

Load the objects.
```{r, eval = FALSE}
day1_norm <- readRDS(file = "backup/01_102_day1_norm_reduction_DO.rds")
day7_hypo <- readRDS(file = "backup/02_102_day7_hypo_reduction_DO.rds")
sc68_vehi <- readRDS(file = "backup/03_102_sc68_vehi_reduction_DO.rds")
sc68_cisp <- readRDS(file = "backup/04_102_sc68_cisp_reduction_DO.rds")
```



## Cell clustering
Load the objects.
```{r, eval = FALSE}
source("R/plot.R")
source("R/function_sign.R")
day1_norm <- readRDS(file = "backup/01_102_day1_norm_reduction_DO.rds")
day7_hypo <- readRDS(file = "backup/02_102_day7_hypo_reduction_DO.rds")
sc68_vehi <- readRDS(file = "backup/03_102_sc68_vehi_reduction_DO.rds")
sc68_cisp <- readRDS(file = "backup/04_102_sc68_cisp_reduction_DO.rds")
```

Here users can choose one of the following unsupervised clustering methods.

**Tips:**
A medoid-and-flat type of unsupervised clustering such as k-medoids method is
the most intuitive way, but their classification results do not always match to
the clusters in the low dimensional space.
On the other hand, though Louvain algorithm with k-nearest neighbor graph
(see below) is a more flexible way, it might be better to try first using more
simple, intuitive, and translatable method.

**1. Partitioning around medoids (PAM)**

The following function `classify_samples_pam_sign()` performs `pam()` in
`cluster` package.
The arguments are `obj`, `data_type`, `category`, and `k` (argument of `pam()`).
```{r, eval = FALSE}
# day1_norm <- classify_samples_pam_sign(
#   obj = day1_norm, data_type = "DO",category = "disease", k = 4)

#
# Below is a memo for plotting the result.
#
# mat <- day1_norm[["reduction"]][["tsne"]][["DO"]][["disease"]]$Y
# lab <- day1_norm[["sample"]][["pam_DO_disease"]]
# col <- rainbow(length(unique(lab)))
# plot(mat, pch=21, bg=col[lab])
# legend("bottomright", legend=unique(sort(lab)), pch=21, pt.bg=col)
```
If the above is conducted, the results are stored in
`obj[["classification"]][["pam"]][["DO"]][["disease"]]` and
`obj[["sample"]][["pam_DO_disease"]]`.

**2. Hierarchical clustering and cutting tree**

The following function `classify_samples_hclustCutree_sign()` performs
`hclust()` and `cutree()` in `stats` package.
The arguments are `obj`, `data_type`, `category`, `method`
(argument of `hclust()`) and `k` (argument of `cutree()`).
```{r, eval = FALSE}
# day1_norm <- classify_samples_hclustCutree_sign(
#   obj = day1_norm, data_type = "DO", category = "disease",
#   method = "ward.D2", k = 3)

#
# Below is a memo for plotting the result.
#
# mat <- day1_norm[["reduction"]][["tsne"]][["DO"]][["disease"]]$Y
# lab <- day1_norm[["sample"]][["hclustCutree_DO_disease"]]
# col <- rainbow(length(unique(lab)))
# plot(mat, pch=21, bg=col[lab])
# legend("bottomright", legend=unique(sort(lab)), pch=21, pt.bg=col)
```
The results are stored in
`obj[["classification"]][["hclustCutree"]][["DO"]][["disease"]]` and
`obj[["sample"]][["hclustCutree_DO_disease"]]`.

**3. Community detection algorithm with k-nearest neighbor (KNN) graph**

The following function `classify_samples_seuratFindClusters_sign()` performs
`CreateSeuratObject()`, `ScaleData()`, `RunPCA()`, `FindNeighbors()`, and
`FindClusters()` in `Seurat` package.
The arguments are `obj`, `data_type`, `category`, `reduction`, `dims`,
`k.param`, `prune.SNN`, `resolution`, and `algorithm`, where the latter six
parameters are used in `FindNeighbors()` and `FindClusters()`.
```{r, eval = FALSE}
day1_norm <- classify_samples_seuratFindClusters_sign(
  obj = day1_norm, data_type = "DO", category = "disease", reduction = "pca",
  dims = 1:20, k.param = 10, prune.SNN = 1./15, resolution = 0.20,
  algorithm = 1)

sc68_vehi <- classify_samples_seuratFindClusters_sign(
  obj = sc68_vehi, data_type = "DO", category = "disease", reduction = "pca",
  dims = 1:20, k.param = 10, prune.SNN = 1./15, resolution = 0.20,
  algorithm = 1)

#
# Below is a memo for plotting the result.
#
# mat <- day1_norm[["reduction"]][["tsne"]][["DO"]][["disease"]]$Y
# lab <- day1_norm[["sample"]][["seuratFindClusters_DO_disease"]]
# col <- rainbow(length(unique(lab)))
# plot(mat, pch=21, bg=col[lab])
# legend("bottomright", legend=unique(sort(lab)), pch=21, pt.bg=col)
```
The results are stored in
`obj[["sample"]][["seuratFindClusters_DO_disease"]]` and
`obj[["classification"]][["seuratFindClusters"]][["DO"]][["disease"]]`.

**4. Clustering by branch detection along pseudotime**

The following function `do_CalculateScaffoldTree_dmap()` computes a scaffold
tree by using `CalculateScaffoldTree()` in `merlot` package, which is based on
the result of diffusion map.
The arguments are `obj`, `data_type`, `category`, `dims` (dimensions of the
diffusion map: only two or three dimensional vector is accepted), `NEndpoints`
(number of end points of the scaffold tree), and `random_seed` (random seed to
be passed on to Python to ensure reproducibility).
```{r, eval = FALSE}
day7_hypo <- do_CalculateScaffoldTree_dmap(
  obj = day7_hypo, data_type = "DO", category = "disease",
  dims = 1:3, NEndpoints = 3, random_seed = 8)

sc68_cisp <- do_CalculateScaffoldTree_dmap(
  obj = sc68_cisp, data_type = "DO", category = "disease",
  dims = 1:3, NEndpoints = 3, random_seed = 8)
```
The result is stored in
`obj[["classification"]][["merlot"]][["DO"]][["disease"]][["ScaffoldTree"]]`.

The following function `do_CalculateElasticTree_dmap()` computes an elastic
principal tree by using `CalculateElasticTree()` in `merlot` package,
which is based on the result of two-dimensional diffusion map.
The arguments are `obj`, `data_type`, `category`, and `N_yk`
(number of support nodes).
```{r, eval = FALSE}
day7_hypo <- do_CalculateElasticTree_dmap(
  obj = day7_hypo, data_type = "DO", category = "disease", N_yk = 50)

sc68_cisp <- do_CalculateElasticTree_dmap(
  obj = sc68_cisp, data_type = "DO", category = "disease", N_yk = 30)
```
The result is stored in
`obj[["classification"]][["merlot"]][["DO"]][["disease"]][["ElasticTree"]]`.

The following function `plot_ElasticTree_dmap()` visualizes the result of
`do_CalculateElasticTree_dmap()`.
```{r, eval = FALSE}
# ----------------------------------------
# day7_hypo
# ----------------------------------------
filename <- "figures/figure_02_0115.png"
png(file = filename, height = 1200, width = 1200, res = 250)
plot_ElasticTree_dmap(obj = day7_hypo, data_type = "DO", category = "disease",
                      theta = 90, phi = 30,
                      title = "day7_hypo (DO: disease)", title_size = 1.2,
                      xlabel = "DC_1", ylabel = "DC_2", zlabel = "DC_3",
                      default_color = TRUE)
dev.off()
```

```{r, eval = FALSE, echo = FALSE, results = "hide"}
# ----------------------------------------
# sc68_cisp
# ----------------------------------------
filename <- "figures/figure_04_0115.png"
png(file = filename, height = 1200, width = 1200, res = 250)
plot_ElasticTree_dmap(obj = sc68_cisp, data_type = "DO", category = "disease",
                      theta = 0, phi = 30,
                      title = "sc68_cisp (DO: disease)", title_size = 1.2,
                      xlabel = "DC_1", ylabel = "DC_2", zlabel = "DC_3",
                      default_color = TRUE)
dev.off()
```

<img src="figures/figure_02_0115.png" width="250px">
<img src="figures/figure_04_0115.png" width="250px">

The following function `do_SignSpaceEmbedding_dmap()` embeds the elastic
principal tree into the original space spanned by the row vectors of
sign-by-sample matrices by using `GenesSpaceEmbedding()` in `merlot` package,
in which Gene is read as Sign here.
```{r, eval = FALSE}
day7_hypo <- do_SignSpaceEmbedding_dmap(
  obj = day7_hypo, data_type = "DO", category = "disease")

sc68_cisp <- do_SignSpaceEmbedding_dmap(
  obj = sc68_cisp, data_type = "DO", category = "disease")
```
The result is stored in
`obj[["classification"]][["merlot"]][["DO"]][["disease"]][["EmbeddedTree"]]`.

The following function `do_CalculatePseudotimes_dmap()` computes a pseudotime
for each cell by using `CalculatePseudotimes()` in `merlot` package.
The arguments are `obj`, `data_type`, `category`, and `T0` (endpoint used at
time zero).
```{r, eval = FALSE}
day7_hypo <- do_CalculatePseudotimes_dmap(
  obj = day7_hypo, data_type = "DO", category = "disease", T0 = 2)

sc68_cisp <- do_CalculatePseudotimes_dmap(
  obj = sc68_cisp, data_type = "DO", category = "disease", T0 = 1)
```
The result is stored in
`obj[["classification"]][["merlot"]][["DO"]][["disease"]][["Pseudotimes"]]`.

The following function `plot_Pseudotime_dmap()` shows the result of
`do_CalculatePseudotimes_dmap()` in a two dimensional space.
```{r, eval = FALSE}
# ----------------------------------------
# day7_hypo
# ----------------------------------------
filename <- "figures/figure_02_0120.png"
png(file = filename, height = 1200, width = 1200, res = 250)
plot_Pseudotime_dmap(obj = day7_hypo, data_type = "DO", category = "disease",
                     theta = 90, phi = 30,
                     title = "day7_hypo (DO: disease)", title_size = 1.2,
                     xlabel = "DC_1", ylabel = "DC_2", zlabel = "DC_3")
dev.off()
```

```{r, eval = FALSE, echo = FALSE, results = "hide"}
# ----------------------------------------
# sc68_cisp
# ----------------------------------------
filename <- "figures/figure_04_0120.png"
png(file = filename, height = 1200, width = 1200, res = 250)
plot_Pseudotime_dmap(obj = sc68_cisp, data_type = "DO", category = "disease",
                     theta = 0, phi = 30,
                     title = "sc68_cisp (DO: disease)", title_size = 1.2,
                     xlabel = "DC_1", ylabel = "DC_2", zlabel = "DC_3")
dev.off()
```

<img src="figures/figure_02_0120.png" width="250px">
<img src="figures/figure_04_0120.png" width="250px">

The following function `classify_samples_merlot_sign()` classifies samples
according to the labels of branches.
```{r, eval = FALSE}
day7_hypo <- classify_samples_merlot_sign(
  obj = day7_hypo, data_type = "DO", category = "disease")

sc68_cisp <- classify_samples_merlot_sign(
  obj = sc68_cisp, data_type = "DO", category = "disease")
```
The result is stored in `obj[["sample"]][["merlot_DO_disease"]]`.

Save the objects.
```{r, eval = FALSE}
saveRDS(day1_norm, file = "backup/01_103_day1_norm_classification_DO.rds")
saveRDS(day7_hypo, file = "backup/02_103_day7_hypo_classification_DO.rds")
saveRDS(sc68_vehi, file = "backup/03_103_sc68_vehi_classification_DO.rds")
saveRDS(sc68_cisp, file = "backup/04_103_sc68_cisp_classification_DO.rds")
```

Load the objects.
```{r, eval = FALSE}
day1_norm <- readRDS(file = "backup/01_103_day1_norm_classification_DO.rds")
day7_hypo <- readRDS(file = "backup/02_103_day7_hypo_classification_DO.rds")
sc68_vehi <- readRDS(file = "backup/03_103_sc68_vehi_classification_DO.rds")
sc68_cisp <- readRDS(file = "backup/04_103_sc68_cisp_classification_DO.rds")
```



## Visualization
In this subsection, the clustering results are visualized.

The following function `plot_Heatmap_SignxSamp()` visualizes the result of
sample clustering by identifying `algo_name` (in the current version,
one of `"pam"`, `"hclustCutree"`, `"seuratFindClusters"`, and `"merlot"`).
```{r, eval = FALSE}
# ----------------------------------------
# For `day1_norm`
# ----------------------------------------
filename <- "figures/figure_01_0130.png"
png(file = filename, height = 1000, width = 1250, res = 300)
plot_Heatmap_SignxSamp(obj = day1_norm, data_type = "DO", category = "disease",
                       algo_name = "seuratFindClusters", method = "ward.D2",
                       show_nReads = FALSE, title = "day1_norm (DO: disease)",
                       name = "DOxSamp",
                       show_rownames_sign = FALSE, show_rownames_label = FALSE,
                       show_rownames_nReads = FALSE, default_color = TRUE)
dev.off()
```

```{r, eval = FALSE, echo = FALSE, results = "hide"}
# ----------------------------------------
# For `day7_hypo`
# ----------------------------------------
filename <- "figures/figure_02_0130.png"
png(file = filename, height = 1000, width = 1250, res = 300)
plot_Heatmap_SignxSamp(obj = day7_hypo, data_type = "DO", category = "disease",
                       algo_name = "merlot", method = "ward.D2",
                       show_nReads = FALSE, title = "day7_hypo (DO: disease)",
                       name = "DOxSamp",
                       show_rownames_sign = FALSE, show_rownames_label = FALSE,
                       show_rownames_nReads = FALSE, default_color = TRUE)
dev.off()
# ----------------------------------------
# For `sc68_vehi`
# ----------------------------------------
filename <- "figures/figure_03_0130.png"
png(file = filename, height = 1000, width = 1250, res = 300)
plot_Heatmap_SignxSamp(obj = sc68_vehi, data_type = "DO", category = "disease",
                       algo_name = "seuratFindClusters", method = "ward.D2",
                       show_nReads = FALSE, title = "sc68_vehi (DO: disease)",
                       name = "DOxSamp",
                       show_rownames_sign = FALSE, show_rownames_label = FALSE,
                       show_rownames_nReads = FALSE, default_color = TRUE)
dev.off()
# ----------------------------------------
# For `sc68_cisp`
# ----------------------------------------
filename <- "figures/figure_04_0130.png"
png(file = filename, height = 1000, width = 1250, res = 300)
plot_Heatmap_SignxSamp(obj = sc68_cisp, data_type = "DO", category = "disease",
                       algo_name = "merlot", method = "ward.D2",
                       show_nReads = FALSE, title = "sc68_cisp (DO: disease)",
                       name = "DOxSamp",
                       show_rownames_sign = FALSE, show_rownames_label = FALSE,
                       show_rownames_nReads = FALSE, default_color = TRUE)
dev.off()
```

<img src="figures/figure_01_0130.png" width="300px">
<img src="figures/figure_02_0130.png" width="300px">

<img src="figures/figure_03_0130.png" width="300px">
<img src="figures/figure_04_0130.png" width="300px">

```{r, eval = FALSE}
# ----------------------------------------
# For `day1_norm`
# ----------------------------------------
p <- plot_tsne_sign(obj = day1_norm, data_type = "DO", category = "disease",
                    algo_name = "seuratFindClusters",
                    theta = NULL, phi = NULL,
                    title = "day1_norm (DO: disease)", title_size = 16,
                    xlabel = "tSNE_1", ylabel = "tSNE_2", zlabel = NULL,
                    default_color = TRUE)
filename <- "figures/figure_01_0131.png"
ggsave(file = filename, plot = p, dpi = 250, width = 4.9, height = 4)
```

```{r, eval = FALSE, echo = FALSE, results = "hide"}
# ----------------------------------------
# For `day7_hypo`
# ----------------------------------------
filename <- "figures/figure_02_0131.png"
png(file = filename, height = 1200, width = 1200, res = 300)
plot_DiffusionMap_sign(obj = day7_hypo, data_type = "DO", category = "disease",
                       algo_name = "merlot",
                       dims = 1:3, theta = 90, phi = 30,
                       title = "day7_hypo (DO: disease)", title_size = 1.2,
                       xlabel = "DC_1", ylabel = "DC_2", zlabel = "DC_3",
                       default_color = TRUE)
dev.off()
# ----------------------------------------
# For `sc68_vehi`
# ----------------------------------------
p <- plot_tsne_sign(obj = sc68_vehi, data_type = "DO", category = "disease",
                    algo_name = "seuratFindClusters",
                    theta = NULL, phi = NULL,
                    title = "sc68_vehi (DO: disease)", title_size = 16,
                    xlabel = "tSNE_1", ylabel = "tSNE_2", zlabel = NULL,
                    default_color = TRUE)
filename <- "figures/figure_03_0131.png"
ggsave(file = filename, plot = p, dpi = 250, width = 4.9, height = 4)
# ----------------------------------------
# For `sc68_cisp`
# ----------------------------------------
filename <- "figures/figure_04_0131.png"
png(file = filename, height = 1200, width = 1200, res = 300)
plot_DiffusionMap_sign(obj = sc68_cisp, data_type = "DO", category = "disease",
                       algo_name = "merlot",
                       dims = 1:3, theta = 0, phi = 30,
                       title = "sc68_cisp (DO: disease)", title_size = 1.2,
                       xlabel = "DC_1", ylabel = "DC_2", zlabel = "DC_3",
                       default_color = TRUE)
dev.off()
```

<img src="figures/figure_01_0131.png" width="250px">
<img src="figures/figure_02_0131.png" width="250px">

<img src="figures/figure_03_0131.png" width="250px">
<img src="figures/figure_04_0131.png" width="250px">

The following command shows the number of cells per cluster visualized
using bar graph.
```{r, eval = FALSE}
# ----------------------------------------
# For `day1_norm`
# ----------------------------------------
p <- plot_bargraph_sign(obj = day1_norm, data_type = "DO", category = "disease",
                        algo_name = "seuratFindClusters",
                        title = "day1_norm", title_size = 20,
                        xlabel = "Label (DO: disease)",
                        ylabel = "Number of samples", ymax = 2000)
filename <- "figures/figure_01_0135.png"
ggsave(file = filename, plot = p, dpi = 300, width = 4.3, height = 4)
```

```{r, eval = FALSE, echo = FALSE, results = "hide"}
# ----------------------------------------
# For `day7_hypo`
# ----------------------------------------
p <- plot_bargraph_sign(obj = day7_hypo, data_type = "DO", category = "disease",
                        algo_name = "merlot",
                        title = "day7_hypo", title_size = 20,
                        xlabel = "Label (DO: disease)",
                        ylabel = "Number of samples", ymax = 1500)
filename <- "figures/figure_02_0135.png"
ggsave(file = filename, plot = p, dpi = 300, width = 4.3, height = 4)
# ----------------------------------------
# For `sc68_vehi`
# ----------------------------------------
p <- plot_bargraph_sign(obj = sc68_vehi, data_type = "DO", category = "disease",
                        algo_name = "seuratFindClusters",
                        title = "sc68_vehi", title_size = 20,
                        xlabel = "Label (DO: disease)",
                        ylabel = "Number of samples", ymax = 3500)
filename <- "figures/figure_03_0135.png"
ggsave(file = filename, plot = p, dpi = 300, width = 4.3, height = 4)
# ----------------------------------------
# For `sc68_cisp`
# ----------------------------------------
p <- plot_bargraph_sign(obj = sc68_cisp, data_type = "DO", category = "disease",
                        algo_name = "merlot",
                        title = "sc68_cisp", title_size = 20,
                        xlabel = "Label (DO: disease)",
                        ylabel = "Number of samples", ymax = 1000)
filename <- "figures/figure_04_0135.png"
ggsave(file = filename, plot = p, dpi = 300, width = 4.3, height = 4)
```

<img src="figures/figure_01_0135.png" width="150px">
<img src="figures/figure_02_0135.png" width="150px">
<img src="figures/figure_03_0135.png" width="150px">
<img src="figures/figure_04_0135.png" width="150px">



## Significant signs
In this subsection, signs of which the sign scores are specifically upregulated
in some cluster are detected by computing the separation indices.
The concept of significant sign is similar to that of differentially expressed
genes (DEGs).

Load the objects.
```{r, eval = FALSE}
source("R/plot.R")
source("R/function_sign.R")
day1_norm <- readRDS(file = "backup/01_103_day1_norm_classification_DO.rds")
day7_hypo <- readRDS(file = "backup/02_103_day7_hypo_classification_DO.rds")
sc68_vehi <- readRDS(file = "backup/03_103_sc68_vehi_classification_DO.rds")
sc68_cisp <- readRDS(file = "backup/04_103_sc68_cisp_classification_DO.rds")
```

The following function `auto_find_marker_sign()` computes separation indices
(i.e., non-parametric measure of degree in separation of two random variable
sets) of signs of interest for each subpopulation.
The arguments are `obj`, `data_type_for_label` (`data_type` used for sample
classification), `category_for_label` (`category` used for sample
classification), `algo_name_for_label` (`algo_name` used for sample
classification), `data_type_for_expr` (`data_type` of interest for
investigating gene expression), `category_for_expr` (`category` of interest for
investigating gene expression), and `label_1` and `label_2` (cluster labels to
be compared: both integer and vector forms are accepted).
This function helps users find positive and negative marker signs of `label_1`
with the positive and negative values of the indices, respectively.
```{r, eval = FALSE}
day1_norm <- auto_find_marker_sign(obj = day1_norm, data_type_for_label = "DO",
                                   category_for_label = "disease",
                                   algo_name_for_label = "seuratFindClusters",
                                   data_type_for_expr = "DO",
                                   category_for_expr = "disease")
day7_hypo <- auto_find_marker_sign(obj = day7_hypo, data_type_for_label = "DO",
                                   category_for_label = "disease",
                                   algo_name_for_label = "merlot",
                                   data_type_for_expr = "DO",
                                   category_for_expr = "disease")
sc68_vehi <- auto_find_marker_sign(obj = sc68_vehi, data_type_for_label = "DO",
                                   category_for_label = "disease",
                                   algo_name_for_label = "seuratFindClusters",
                                   data_type_for_expr = "DO",
                                   category_for_expr = "disease")
sc68_cisp <- auto_find_marker_sign(obj = sc68_cisp, data_type_for_label = "DO",
                                   category_for_label = "disease",
                                   algo_name_for_label = "merlot",
                                   data_type_for_expr = "DO",
                                   category_for_expr = "disease")
```
The results are stored in
`obj[["marker"]][["DO"]][["disease"]][["Label_DO_disease_1_vs_2/3"]]`,
`obj[["marker"]][["DO"]][["disease"]][["Label_DO_disease_2_vs_1/3"]]`,
`obj[["marker"]][["DO"]][["disease"]][["Label_DO_disease_3_vs_1/2"]]`, and
`obj[["marker"]][["DO"]][["disease"]][["all"]]`.
Here `sep_I` stands for the separation index of the sign, which quantifies the
extent of separation between two different sets of random variables.
Note that `auto_find_marker_sign()` computes `sep_I` for the given group versus
all other cells.
```{r, eval = FALSE}
datatable(day1_norm[["marker"]][["DO"]][["disease"]][["all"]], rownames = FALSE)
datatable(day7_hypo[["marker"]][["DO"]][["disease"]][["all"]], rownames = FALSE)
datatable(sc68_vehi[["marker"]][["DO"]][["disease"]][["all"]], rownames = FALSE)
datatable(sc68_cisp[["marker"]][["DO"]][["disease"]][["all"]], rownames = FALSE)
```

* `day1_norm`

<iframe src="figures/day1_norm_DO_disease.html" style="width:100%; height:300px;"></iframe>

<br><br><br><br><br><br><br><br><br><br>

The following function `plot_violin_signScore()` shows the distribution of
expression levels of a given sign across subpopulations by violin plots.
The arguments are `obj`, `sign_name`, `data_type_for_label` (`data_type` used
for sample classification), `category_for_label` (`category` used for sample
classification), `algo_name_for_label` (`algo_name` used for sample
classification), `data_type_for_expr` (`data_type` of interest for investigating
gene expression), and `category_for_expr` (`category` of interest for
investigating gene expression).
```{r, eval = FALSE}
# ----------------------------------------
# For `day1_norm`
# ----------------------------------------
p <- plot_violin_signScore(
  obj = day1_norm, sign_name = "DOID:28_V",
  data_type_for_label = "DO", category_for_label = "disease",
  algo_name_for_label = "seuratFindClusters",
  data_type_for_expr = "DO", category_for_expr = "disease",
  title = "DOID:28_V\nEndocrine system disease\n(HMGB1/CHGA/..)",
  title_size = 16, default_color = TRUE)
filename <- "figures/figure_01_0140.png"
ggsave(file = filename, plot = p, dpi = 300, width = 4, height = 4)

p <- plot_violin_signScore(
  obj = day1_norm, sign_name = "DOID:299_S",
  data_type_for_label = "DO", category_for_label = "disease",
  algo_name_for_label = "seuratFindClusters",
  data_type_for_expr = "DO", category_for_expr = "disease",
  title = "DOID:299_S\nAdenocarcinoma\n(KRT19/KRT18/..)",
  title_size = 16, default_color = TRUE)
filename <- "figures/figure_01_0141.png"
ggsave(file = filename, plot = p, dpi = 300, width = 4, height = 4)
```

```{r, eval = FALSE, echo = FALSE, results = "hide"}
# ----------------------------------------
# For `day7_hypo`
# ----------------------------------------
p <- plot_violin_signScore(
  obj = day7_hypo, sign_name = "DOID:299_S",
  data_type_for_label = "DO", category_for_label = "disease",
  algo_name_for_label = "merlot",
  data_type_for_expr = "DO", category_for_expr = "disease",
  title = "DOID:299_S\nAdenocarcinoma\n(KRT19/KRT18/..)",
  title_size = 16, default_color = TRUE)
filename <- "figures/figure_02_0140.png"
ggsave(file = filename, plot = p, dpi = 300, width = 4, height = 4)

p <- plot_violin_signScore(
  obj = day7_hypo, sign_name = "DOID:1749_V",
  data_type_for_label = "DO", category_for_label = "disease",
  algo_name_for_label = "merlot",
  data_type_for_expr = "DO", category_for_expr = "disease",
  title = "DOID:1749_V\nSquamous cell carcinoma\n(SST/VIM/..)",
  title_size = 16, default_color = TRUE)
filename <- "figures/figure_02_0141.png"
ggsave(file = filename, plot = p, dpi = 300, width = 4, height = 4)
# ----------------------------------------
# For `sc68_vehi`
# ----------------------------------------
p <- plot_violin_signScore(
  obj = sc68_vehi, sign_name = "DOID:5409_S",
  data_type_for_label = "DO", category_for_label = "disease",
  algo_name_for_label = "seuratFindClusters",
  data_type_for_expr = "DO", category_for_expr = "disease",
  title = "DOID:5409_S\nLung small cell carcinoma\n(ASCL1/GRP/..)",
  title_size = 16, default_color = TRUE)
filename <- "figures/figure_03_0140.png"
ggsave(file = filename, plot = p, dpi = 300, width = 4, height = 4)

p <- plot_violin_signScore(
  obj = sc68_vehi, sign_name = "DOID:5409_V",
  data_type_for_label = "DO", category_for_label = "disease",
  algo_name_for_label = "seuratFindClusters",
  data_type_for_expr = "DO", category_for_expr = "disease",
  title = "DOID:5409_V\nLung small cell carcinoma\n(MKI67/BIRC5/..)",
  title_size = 16, default_color = TRUE)
filename <- "figures/figure_03_0141.png"
ggsave(file = filename, plot = p, dpi = 300, width = 4, height = 4)
# ----------------------------------------
# For `sc68_cisp`
# ----------------------------------------
p <- plot_violin_signScore(
  obj = sc68_cisp, sign_name = "DOID:5409_S",
  data_type_for_label = "DO", category_for_label = "disease",
  algo_name_for_label = "merlot",
  data_type_for_expr = "DO", category_for_expr = "disease",
  title = "DOID:5409_S\nLung small cell carcinoma\n(ASCL1/GRP/..)",
  title_size = 16, default_color = TRUE)
filename <- "figures/figure_04_0140.png"
ggsave(file = filename, plot = p, dpi = 300, width = 4, height = 4)

p <- plot_violin_signScore(
  obj = sc68_cisp, sign_name = "DOID:5409_V",
  data_type_for_label = "DO", category_for_label = "disease",
  algo_name_for_label = "merlot",
  data_type_for_expr = "DO", category_for_expr = "disease",
  title = "DOID:5409_V\nLung small cell carcinoma\n(BIRC5/MKI67..)",
  title_size = 16, default_color = TRUE)
filename <- "figures/figure_04_0141.png"
ggsave(file = filename, plot = p, dpi = 300, width = 4, height = 4)

p <- plot_violin_signScore(
  obj = sc68_cisp, sign_name = "DOID:654_S",
  data_type_for_label = "DO", category_for_label = "disease",
  algo_name_for_label = "merlot",
  data_type_for_expr = "DO", category_for_expr = "disease",
  title = "DOID:654_S\nOvernutrition\n(ATF3/DUSP1/..)",
  title_size = 16, default_color = TRUE)
filename <- "figures/figure_04_0142.png"
ggsave(file = filename, plot = p, dpi = 300, width = 4, height = 4)
```

* `day1_norm`

<img src="figures/figure_01_0140.png" width="230px">
<img src="figures/figure_01_0141.png" width="230px">

* `day7_hypo`

<img src="figures/figure_02_0140.png" width="230px">
<img src="figures/figure_02_0141.png" width="230px">

* `sc68_vehi`

<img src="figures/figure_03_0140.png" width="230px">
<img src="figures/figure_03_0141.png" width="230px">

* `sc68_cisp`

<img src="figures/figure_04_0140.png" width="230px">
<img src="figures/figure_04_0141.png" width="230px">
<img src="figures/figure_04_0142.png" width="230px">

The functions `plot_tsne_signScore()`, `plot_umap_signScore()`, and
`plot_dmap_signScore()` show the sign score of a given sign on the t-SNE,
UMAP, and diffusion map spaces, respectively.
```{r, eval = FALSE}
# ----------------------------------------
# For `day1_norm`
# ----------------------------------------
p <- plot_tsne_signScore(
  obj = day1_norm, sign_name = "DOID:28_V",
  data_type_for_tsne = "DO", category_for_tsne = "disease",
  data_type_for_expr = "DO", category_for_expr = "disease",
  theta = NULL, phi = NULL,
  title = "DOID:28_V\nEndocrine system disease\n(CHGA/HMGB1/..)",
  title_size = 16, label_name = "Sign score",
  xlabel = "tSNE_1", ylabel = "tSNE_2", zlabel = NULL)
filename <- "figures/figure_01_0150.png"
ggsave(file = filename, plot = p, dpi = 300, width = 5.6, height = 4.4)
```

```{r, eval = FALSE, echo = FALSE, results = "hide"}
# ----------------------------------------
# For `day1_norm`
# ----------------------------------------
p <- plot_tsne_signScore(
  obj = day1_norm, sign_name = "DOID:299_S",
  data_type_for_tsne = "DO", category_for_tsne = "disease",
  data_type_for_expr = "DO", category_for_expr = "disease",
  theta = NULL, phi = NULL,
  title = "DOID:299_S\nAdenocarcinoma\n(KRT19/KRT8/..)",
  title_size = 16, label_name = "Sign score",
  xlabel = "tSNE_1", ylabel = "tSNE_2", zlabel = NULL)
filename <- "figures/figure_01_0151.png"
ggsave(file = filename, plot = p, dpi = 300, width = 5.6, height = 4.4)

p <- plot_umap_signScore(
  obj = day1_norm, sign_name = "DOID:28_V",
  data_type_for_umap = "DO", category_for_umap = "disease",
  data_type_for_expr = "DO", category_for_expr = "disease",
  theta = NULL, phi = NULL,
  title = "DOID:28_V\nEndocrine system disease\n(CHGA/HMGB1/..)",
  title_size = 16, label_name = "Sign score",
  xlabel = "UMAP_1", ylabel = "UMAP_2", zlabel = NULL)
filename <- "figures/figure_01_0152.png"
ggsave(file = filename, plot = p, dpi = 300, width = 5.6, height = 4.4)

p <- plot_umap_signScore(
  obj = day1_norm, sign_name = "DOID:299_S",
  data_type_for_umap = "DO", category_for_umap = "disease",
  data_type_for_expr = "DO", category_for_expr = "disease",
  theta = NULL, phi = NULL,
  title = "DOID:299_S\nAdenocarcinoma\n(KRT19/KRT8/..)",
  title_size = 16, label_name = "Sign score",
  xlabel = "UMAP_1", ylabel = "UMAP_2", zlabel = NULL)
filename <- "figures/figure_01_0153.png"
ggsave(file = filename, plot = p, dpi = 300, width = 5.6, height = 4.4)
```

```{r, eval = FALSE}
# ----------------------------------------
# For `day7_hypo`
# ----------------------------------------
filename <- "figures/figure_02_0150.png"
png(file = filename, height = 1200, width = 1200, res = 300)
plot_dmap_signScore(
  obj = day7_hypo, sign_name = "DOID:299_S",
  data_type_for_dmap = "DO", category_for_dmap = "disease",
  data_type_for_expr = "DO", category_for_expr = "disease",
  dim_dmap = 1:3, theta = 90, phi = 30,
  title = "DOID:299_S\nAdenocarcinoma\n(KRT19/KRT8/..)",
  title_size = 1.0, label_name = "Sign score",
  xlabel = "DC_1", ylabel = "DC_2", zlabel = "DC_3")
dev.off()
```

```{r, eval = FALSE, echo = FALSE, results = "hide"}
# ----------------------------------------
# For `day7_hypo`
# ----------------------------------------
filename <- "figures/figure_02_0151.png"
png(file = filename, height = 1200, width = 1200, res = 300)
plot_dmap_signScore(
  obj = day7_hypo, sign_name = "DOID:1749_V",
  data_type_for_dmap = "DO", category_for_dmap = "disease",
  data_type_for_expr = "DO", category_for_expr = "disease",
  dim_dmap = 1:3, theta = 90, phi = 30,
  title = "DOID:1749_V\nSquamous cell carcinoma\n(SST/VIM/..)",
  title_size = 1.0, label_name = "Sign score",
  xlabel = "DC_1", ylabel = "DC_2", zlabel = "DC_3")
dev.off()
# ----------------------------------------
# For `sc68_vehi`
# ----------------------------------------
p <- plot_tsne_signScore(
  obj = sc68_vehi, sign_name = "DOID:5409_S",
  data_type_for_tsne = "DO", category_for_tsne = "disease",
  data_type_for_expr = "DO", category_for_expr = "disease",
  theta = NULL, phi = NULL,
  title = "DOID:5409_S\nLung small cell carcinoma\n(ASCL1/GRP/..)",
  title_size = 16, label_name = "Sign score",
  xlabel = "tSNE_1", ylabel = "tSNE_2", zlabel = NULL)
filename <- "figures/figure_03_0150.png"
ggsave(file = filename, plot = p, dpi = 300, width = 5.6, height = 4.4)

p <- plot_tsne_signScore(
  obj = sc68_vehi, sign_name = "DOID:5409_V",
  data_type_for_tsne = "DO", category_for_tsne = "disease",
  data_type_for_expr = "DO", category_for_expr = "disease",
  theta = NULL, phi = NULL,
  title = "DOID:5409_V\nLung small cell carcinoma\n(MKI67/BIRC5/..)",
  title_size = 16, label_name = "Sign score",
  xlabel = "tSNE_1", ylabel = "tSNE_2", zlabel = NULL)
filename <- "figures/figure_03_0151.png"
ggsave(file = filename, plot = p, dpi = 300, width = 5.6, height = 4.4)
# ----------------------------------------
# For `sc68_cisp`
# ----------------------------------------
filename <- "figures/figure_04_0150.png"
png(file = filename, height = 1200, width = 1200, res = 300)
plot_dmap_signScore(
  obj = sc68_cisp, sign_name = "DOID:5409_S",
  data_type_for_dmap = "DO", category_for_dmap = "disease",
  data_type_for_expr = "DO", category_for_expr = "disease",
  dim_dmap = 1:3, theta = 0, phi = 30,
  title = "DOID:5409_S\nLung small cell carcinoma\n(ASCL1/GRP/..)",
  title_size = 1.0, label_name = "Sign score",
  xlabel = "DC_1", ylabel = "DC_2", zlabel = "DC_3")
dev.off()

filename <- "figures/figure_04_0151.png"
png(file = filename, height = 1200, width = 1200, res = 300)
plot_dmap_signScore(
  obj = sc68_cisp, sign_name = "DOID:5409_V",
  data_type_for_dmap = "DO", category_for_dmap = "disease",
  data_type_for_expr = "DO", category_for_expr = "disease",
  dim_dmap = 1:3, theta = 0, phi = 30,
  title = "DOID:5409_V\nLung small cell carcinoma\n(BIRC5/MKI67..)",
  title_size = 1.0, label_name = "Sign score",
  xlabel = "DC_1", ylabel = "DC_2", zlabel = "DC_3")
dev.off()

filename <- "figures/figure_04_0152.png"
png(file = filename, height = 1200, width = 1200, res = 300)
plot_dmap_signScore(
  obj = sc68_cisp, sign_name = "DOID:654_S",
  data_type_for_dmap = "DO", category_for_dmap = "disease",
  data_type_for_expr = "DO", category_for_expr = "disease",
  dim_dmap = 1:3, theta = 0, phi = 30,
  title = "DOID:654_S\nOvernutrition\n(ATF3/DUSP1/..)",
  title_size = 1.0, label_name = "Sign score",
  xlabel = "DC_1", ylabel = "DC_2", zlabel = "DC_3")
dev.off()
```

* `day1_norm`

<img src="figures/figure_01_0150.png" width="250px">
<img src="figures/figure_01_0151.png" width="250px">

<img src="figures/figure_01_0152.png" width="250px">
<img src="figures/figure_01_0153.png" width="250px">

* `day7_hypo`

<img src="figures/figure_02_0150.png" width="250px">
<img src="figures/figure_02_0151.png" width="250px">

* `sc68_vehi`

<img src="figures/figure_03_0150.png" width="250px">
<img src="figures/figure_03_0151.png" width="250px">

* `sc68_cisp`

<img src="figures/figure_04_0150.png" width="250px">
<img src="figures/figure_04_0151.png" width="250px">
<img src="figures/figure_04_0152.png" width="250px">

The functions `plot_tsne_geneExpression()`, `plot_umap_geneExpression()`, and
`plot_dmap_geneExpression()` show the mean of normalized gene expression levels
(if `zscore = TRUE`, the expression levels are converted into z-score) for
the input genes (`gene_names`) on the low-dimensional space.
```{r, eval = FALSE}
# ----------------------------------------
# For `day1_norm`
# ----------------------------------------
p <- plot_tsne_geneExpression(
  obj = day1_norm, gene_name = "CHGA", zscore = FALSE,
  data_type_for_tsne = "DO", category_for_tsne = "disease",
  theta = 0, phi = 0,
  title = "day1_norm (DO: disease)", title_size = 16,
  label_name = "CHGA",
  xlabel = "tSNE_1", ylabel = "tSNE_2", zlabel = NULL)
filename <- "figures/figure_01_0160.png"
ggsave(file = filename, plot = p, dpi = 300, width = 5.2, height = 4)

p <- plot_tsne_geneExpression(
  obj = day1_norm, gene_name = "KRT19", zscore = FALSE,
  data_type_for_tsne = "DO", category_for_tsne = "disease",
  theta = 0, phi = 0,
  title = "day1_norm (DO: disease)", title_size = 16,
  label_name = "KRT19",
  xlabel = "tSNE_1", ylabel = "tSNE_2", zlabel = NULL)
filename <- "figures/figure_01_0161.png"
ggsave(file = filename, plot = p, dpi = 300, width = 5.2, height = 4)
# ----------------------------------------
# For `day7_hypo`
# ----------------------------------------
filename <- "figures/figure_02_0160.png"
png(file = filename, height = 1200, width = 1200, res = 300)
plot_dmap_geneExpression(
  obj = day7_hypo, gene_name = "CHGA", zscore = FALSE,
  data_type_for_dmap = "DO", category_for_dmap = "disease",
  dim_dmap = 1:3, theta = 90, phi = 30,
  title = "day7_hypo (DO: disease)", title_size = 1.2,
  label_name = "CHGA",
  xlabel = "DC_1", ylabel = "DC_2", zlabel = "DC_3")
dev.off()

filename <- "figures/figure_02_0161.png"
png(file = filename, height = 1200, width = 1200, res = 300)
plot_dmap_geneExpression(
  obj = day7_hypo, gene_name = "KRT19", zscore = FALSE,
  data_type_for_dmap = "DO", category_for_dmap = "disease",
  dim_dmap = 1:3, theta = 90, phi = 30,
  title = "day7_hypo (DO: disease)", title_size = 1.2,
  label_name = "KRT19",
  xlabel = "DC_1", ylabel = "DC_2", zlabel = "DC_3")
dev.off()
```

```{r, eval = FALSE, echo = FALSE, results = "hide"}
# ----------------------------------------
# For `sc68_vehi`
# ----------------------------------------
p <- plot_tsne_geneExpression(
  obj = sc68_vehi, gene_name = "ASCL1", zscore = FALSE,
  data_type_for_tsne = "DO", category_for_tsne = "disease",
  theta = 0, phi = 0,
  title = "sc68_vehi (DO: disease)", title_size = 16,
  label_name = "ASCL1",
  xlabel = "tSNE_1", ylabel = "tSNE_2", zlabel = NULL)
filename <- "figures/figure_03_0160.png"
ggsave(file = filename, plot = p, dpi = 300, width = 5.2, height = 4)

p <- plot_tsne_geneExpression(
  obj = sc68_vehi, gene_name = "BIRC5", zscore = FALSE,
  data_type_for_tsne = "DO", category_for_tsne = "disease",
  theta = 0, phi = 0,
  title = "sc68_vehi (DO: disease)", title_size = 16,
  label_name = "BIRC5",
  xlabel = "tSNE_1", ylabel = "tSNE_2", zlabel = NULL)
filename <- "figures/figure_03_0161.png"
ggsave(file = filename, plot = p, dpi = 300, width = 5.1, height = 4)
# ----------------------------------------
# For `sc68_cisp`
# ----------------------------------------
filename <- "figures/figure_04_0160.png"
png(file = filename, height = 1200, width = 1200, res = 300)
plot_dmap_geneExpression(
  obj = sc68_cisp, gene_name = "ASCL1", zscore = FALSE,
  data_type_for_dmap = "DO", category_for_dmap = "disease",
  dim_dmap = 1:3, theta = 0, phi = 30,
  title = "sc68_cisp (DO: disease)", title_size = 1.2,
  label_name = "ASCL1",
  xlabel = "DC_1", ylabel = "DC_2", zlabel = "DC_3")
dev.off()

filename <- "figures/figure_04_0161.png"
png(file = filename, height = 1200, width = 1200, res = 300)
plot_dmap_geneExpression(
  obj = sc68_cisp, gene_name = "BIRC5", zscore = FALSE,
  data_type_for_dmap = "DO", category_for_dmap = "disease",
  dim_dmap = 1:3, theta = 0, phi = 30,
  title = "sc68_cisp (DO: disease)", title_size = 1.2,
  label_name = "BIRC5",
  xlabel = "DC_1", ylabel = "DC_2", zlabel = "DC_3")
dev.off()
```

* `day1_norm`

<img src="figures/figure_01_0160.png" width="250px">
<img src="figures/figure_01_0161.png" width="250px">

* `day7_hypo`

<img src="figures/figure_02_0160.png" width="250px">
<img src="figures/figure_02_0161.png" width="250px">

* `sc68_vehi`

<img src="figures/figure_03_0160.png" width="250px">
<img src="figures/figure_03_0161.png" width="250px">

* `sc68_cisp`

<img src="figures/figure_04_0160.png" width="250px">
<img src="figures/figure_04_0161.png" width="250px">

The following function `plot_pseudotime_vs_signscore()` plots expression
levels versus pseudotime for a given sign, in which filled circles, bold lines,
and the shade regions stand for the expression values of samples, mean
expression levels, and standard deviations (this value is 0 if there are less
than three samples), respectively.
The arguments `data_type_for_tree` and `category_for_tree` are the data type
and category used for the branch-based classification of samples (i.e., the
labels assigned to the branches), while `data_type_for_expr` and
`category_for_expr` for the expression levels of signs of interest.
```{r, eval = FALSE}
# ----------------------------------------
# For `day7_hypo`
# ----------------------------------------
p <- plot_pseudotime_vs_signscore(
  obj = day7_hypo, sign_name = "DOID:299_S",
  data_type_for_tree = "DO", category_for_tree = "disease",
  data_type_for_expr = "DO", category_for_expr = "disease",
  title = "DOID:299_S\nAdenocarcionma\n(KRT19/KRT8/..)", title_size = 22,
  label_name = "Label\n(DO: disease)",
  xlabel = "Pseudotime (DO: disease)", ylabel = "Sign score")
filename <- "figures/figure_02_0170.png"
ggsave(file = filename, plot = p, dpi = 300, width = 7.5, height = 5)
```

```{r, eval = FALSE, echo = FALSE, results = "hide"}
# ----------------------------------------
# For `day7_hypo`
# ----------------------------------------
p <- plot_pseudotime_vs_signscore(
  obj = day7_hypo, sign_name = "DOID:1749_V",
  data_type_for_tree = "DO", category_for_tree = "disease",
  data_type_for_expr = "DO", category_for_expr = "disease",
  title = "DOID:1749_V\nSquamous cell carcinoma\n(SST/VIM/..)", title_size = 22,
  label_name = "Label\n(DO: disease)",
  xlabel = "Pseudotime (DO: disease)", ylabel = "Sign score")
filename <- "figures/figure_02_0171.png"
ggsave(file = filename, plot = p, dpi = 300, width = 7.5, height = 5)
# ----------------------------------------
# For `sc68_cisp`
# ----------------------------------------
p <- plot_pseudotime_vs_signscore(
  obj = sc68_cisp, sign_name = "DOID:5409_S",
  data_type_for_tree = "DO", category_for_tree = "disease",
  data_type_for_expr = "DO", category_for_expr = "disease",
  title = "DOID:5409_S\nLung small cell carcinoma\n(ASCL1/GRP/..)",
  title_size = 22,
  label_name = "Label\n(DO: disease)",
  xlabel = "Pseudotime (DO: disease)", ylabel = "Sign score")
filename <- "figures/figure_04_0170.png"
ggsave(file = filename, plot = p, dpi = 300, width = 7.5, height = 5)

p <- plot_pseudotime_vs_signscore(
  obj = sc68_cisp, sign_name = "DOID:5409_V",
  data_type_for_tree = "DO", category_for_tree = "disease",
  data_type_for_expr = "DO", category_for_expr = "disease",
  title = "DOID:5409_V\nLung small cell carcinoma\n(BIRC5/MKI67..)",
  title_size = 22,
  label_name = "Label\n(DO: disease)",
  xlabel = "Pseudotime (DO: disease)", ylabel = "Sign score")
filename <- "figures/figure_04_0171.png"
ggsave(file = filename, plot = p, dpi = 300, width = 7.5, height = 5)

p <- plot_pseudotime_vs_signscore(
  obj = sc68_cisp, sign_name = "DOID:654_S",
  data_type_for_tree = "DO", category_for_tree = "disease",
  data_type_for_expr = "DO", category_for_expr = "disease",
  title = "DOID:654_S\nOvernutrition\n(ATF3/DUSP1/..)",
  title_size = 22,
  label_name = "Label\n(DO: disease)",
  xlabel = "Pseudotime (DO: disease)", ylabel = "Sign score")
filename <- "figures/figure_04_0172.png"
ggsave(file = filename, plot = p, dpi = 300, width = 7.5, height = 5)
```

* `day7_hypo`

<img src="figures/figure_02_0170.png" width="250px">
<img src="figures/figure_02_0171.png" width="250px">

* `sc68_cisp`

<img src="figures/figure_04_0170.png" width="250px">
<img src="figures/figure_04_0171.png" width="250px">
<img src="figures/figure_04_0172.png" width="250px">

Save the objects.
```{r, eval = FALSE}
saveRDS(day1_norm, file = "backup/01_104_day1_norm_DES_DO.rds")
saveRDS(day7_hypo, file = "backup/02_104_day7_hypo_DES_DO.rds")
saveRDS(sc68_vehi, file = "backup/03_104_sc68_vehi_DES_DO.rds")
saveRDS(sc68_cisp, file = "backup/04_104_sc68_cisp_DES_DO.rds")
```
