# Appendix B: automatically tuning ASURATâ€™s parameters
In this section, we demonstrate tuning the parameters of ASURAT for creating
an interpretable system of signs as well as a sign-by-sample matrix (SSM) for
Disease Ontology (DO) by using a single-cell RNA sequencing data `day1_norm`
(cervical cancer data under normoxic condition).

Below are the R scripts to produce a series of plots of UMAP of the SSMs for DO.
Please note that the R codes in `function_movie.R` is almost the same with
those in `function_sign.R` except that `function_movie.R` always requires
an argument `category` as input and it does not record the parameters in
`obj[["history"]]` for faster computing.
```{r, eval = FALSE}
# ----------------------------------------
# Installation
# ----------------------------------------
# General
library(tidyverse)       # For efficient handling of data.frame
# Below is a memo for attaching org.Hs.eg.db
# options(connectionObserver = NULL)
library(org.Hs.eg.db)    # For using human genome annotation package

# Plot
library(ggrepel)         # For graphics with labeling

# Utilities
library(Rtsne)           # For using Rtsne
library(umap)            # For using umap
library(destiny)         # For using diffusion map
library(cluster)         # For using `pam`

# Databases
library(DOSE)            # For using DB for Disease Ontology
library(clusterProfiler) # For using DB for Gene Ontology
library(GOSemSim)        # For using godata in GOSemSim package
# ----------------------------------------
# Reading files
# ----------------------------------------
source("R/function_movie.R")
tidy_DO <- readRDS(file = "data/2020_001_databases/20201213_tidy_DO_human.rds")
IC_DO <- readRDS(file = "data/2020_001_databases/20201213_IC_DO_human.rds")
day1_norm <- readRDS(file = "backup/01_005_day1_norm_normalized.rds")
day1_norm_cor <- readRDS(file = "backup/01_006_day1_norm_correlation.rds")
# ----------------------------------------
# Definition
# ----------------------------------------
tidy_DB <- tidy_DO
obj <- day1_norm
obj_cor <- day1_norm_cor
data_type <- "DO"
category <- "disease"
title <- paste("day1_norm (", data_type, ": ", category, ")", sep = "")
dirname <- "movie"
do_tsne <- function(obj_red){
  do_tsne_sign_movie(obj = obj_red, data_type = data_type, category = category,
                     pca_dim = NULL, tsne_dim = 2)
}
do_umap <- function(obj_red){
  do_umap_sign_movie(obj = obj_red, data_type = data_type, category = category,
                     pca_dim = NULL, umap_dim = 2)
}
do_dmap <- function(obj_red){
  do_DiffusionMap_movie(obj = obj_red,
                        data_type = data_type, category = category,
                        sigma = "local", distance = "euclidean")
}
# ----------------------------------------
# Setting the global parameters
# ----------------------------------------
min_ngenes <- 2
max_ngenes <- 1000
threshold <- 0.8
keep_rareID <- TRUE
plot_type <- "tsne"
xlabel <- "tSNE_1"
ylabel <- "tSNE_2"
# ----------------------------------------
# Preparing
# ----------------------------------------
obj[["sign"]][[data_type]] <- tidy_DB
obj <- do_quickQC_sign_movie(obj = obj,
                             data_type = data_type, category = category,
                             min_ngenes = min_ngenes, max_ngenes = max_ngenes)
# ----------------------------------------
# Loop
# ----------------------------------------
cnt <- 1
for(min_cnt in 2:2){
  for(i in 0:20){
    th_posi = 0.20 + i * 0.01
    for(j in 0:10){
      th_nega = -(0.2 + j * 0.02)
      
      obj_red <- separate_variables_sign_movie(
        obj = obj, obj_cor = obj_cor,
        data_type = data_type, category = category,
        method = "spearman", th_posi = th_posi, th_nega = th_nega)
      
      obj_red <- select_sign_movie(
        obj = obj_red, data_type = data_type, category = category,
        min_cnt = min_cnt, min_cnt_weak = 2)
      
      if(data_type == "DO"){
        obj_red <- compute_SemSim_sign_movie(
          obj = obj_red, data_type = data_type, category = category,
          measure = "Jiang", orgdb = NULL, treeTable = NULL, IC = IC_DO)
        obj_red <- reduce_sign_movie(
          obj = obj_red, data_type = data_type, category = category,
          threshold = threshold, keep_rareID = keep_rareID)
      }else if(data_type == "GO"){
        obj_red <- compute_SemSim_sign_movie(
          obj = obj_red, data_type = data_type, category = category,
          measure = "Jiang", orgdb = org.Hs.eg.db, treeTable = NULL, IC = IC_GO)
        obj_red <- reduce_sign_movie(
          obj = obj_red, data_type = data_type, category = category,
          threshold = threshold, keep_rareID = keep_rareID)
      }
      
      obj_red <- make_signxsample_matrix_movie(
        obj = obj_red, data_type = data_type, category = category,
        weight_strg = 0.5, weight_vari = 0.5)
      
      if(plot_type == "tsne"){
        obj_red <- try(do_tsne(obj = obj_red), silent = TRUE)
      }else if(plot_type == "umap"){
        obj_red <- try(do_umap(obj = obj_red), silent = TRUE)
      }else if(plot_type == "dmap"){
        obj_red <- try(do_dmap(obj = obj_red), silent = TRUE)
      }

      if(class(obj_red) == "try-error"){
        print("Skip this parameter set.")
        next
      }

      if(plot_type == "tsne"){
        plot_tsne_sign_movie(
          obj = obj_red, data_type = data_type, category = category,
          p1 = th_posi, p2 = th_nega, p3 = min_cnt, title = title,
          xlabel = xlabel, ylabel = ylabel, dirname = dirname, cnt = cnt)
      }else if(plot_type == "umap"){
        plot_umap_sign_movie(
          obj = obj_red, data_type = data_type, category = category,
          p1 = th_posi, p2 = th_nega, p3 = min_cnt, title = title,
          xlabel = xlabel, ylabel = ylabel, dirname = dirname, cnt = cnt)
      }else if(plot_type == "dmap"){
        plot_DiffusionMap_movie(
          obj = obj_red, data_type = data_type, category = category,
          p1 = th_posi, p2 = th_nega, p3 = min_cnt, title = title,
          xlabel = xlabel, ylabel = ylabel, dirname = dirname, cnt = cnt)
      }
      
      cnt <- cnt + 1
    }
  }
}
```
