--- 
title: "Introduction to ASURAT (version 0.0.0.9001)"
author: "Keita Iida"
date: "`r Sys.Date()`"
output: pdf_document
documentclass: book
biblio-style: apalike
link-citations: yes
description: This is a developmental version of ASURAT.
site: bookdown::bookdown_site
---

```{r, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>"
)
```

```{r include = FALSE}
# automatically create a bib database for R packages
knitr::write_bib(c(.packages(), 'bookdown', 'knitr', 'rmarkdown'),
                 'packages.bib')
```



# Overview
ASURAT is an analysis pipeline, which simultaneously performs
unsupervised clustering and biological interpretation in terms of
cell type, disease, biological process, and signaling pathway activity.
Inputting a single-cell RNA-seq (scRNA-seq) data and knowledge-based databases,
ASURAT transforms the read count table into an original multivariate table,
termed a sign-by-sample matrix (SSM).
By analyzing SSMs, users can cluster cells to aid their interpretation.

<br>

<img src="figures/figure_00_0000.png" width="400px">



## License and description
```
Package: ASURAT
Title: Functional annotation-driven unsupervised clustering of
  single-cell transcriptomes
Version: 0.0.0.9000
Authors@R: person("Keita", "Iida", email = "kiida@protein.osaka-u.ac.jp",
  role = c("aut", "cre"))
Description: In single-cell transcriptomics ASURAT assists users to cluster
    cells in biologically explainable manners and find various types of
    biological terms specifically upregulated in each cluster.
    Using ASURAT, users can create and analyze a special type of matrix termed
    "sign-by-sample matrix", of which the rows and columns stand for
    biological terms and samples, respectively.
License: GPL-3
Encoding: UTF-8
LazyData: true
Imports: Rcpp (>= 1.0.6)
LinkingTo: Rcpp
Suggests: 
    rmarkdown,
    knitr,
    Seurat (>= 3.0.0)
RoxygenNote: 7.1.1
Depends: 
    R (>= 2.10)
VignetteBuilder: knitr
```



## Computational environment
MacBook Pro (Big Sur, 16-inch, 2019), Processor (2.4 GHz 8-Core Intel Core i9),
Memory (64 GB 2667 MHz DDR4).



## Installation
The following libraries are used almost everywhere.
```{r, eval = FALSE}
# General
library(tidyverse)       # For efficient handling of data.frame
library(Matrix)          # For using `readMM`
library(DT)              # For using datatable

# Plot
library(ggExtra)         # For graphics
library(ggrepel)         # For graphics with labeling
library(plot3D)          # For 3-dimensional scatter plot
library(ComplexHeatmap)  # For ComplexHeatmap
library(circlize)        # For color mapping function in ComplexHeatmap
library(viridis)         # For using color palette `plasma` in pseudotime anal

# Normalization
library(bayNorm)         # For using bayNorm

# Utilities
library(Rtsne)           # For using Rtsne()
library(umap)            # For using umap()
library(Rcpp)            # For using external C++ codes
library(cluster)         # For using pam()
library(Seurat)          # For using Seurat
library(destiny)         # For using DiffusionMap()
library(merlot)          # For detecting branches along pseudotime
library(exactRankTests)  # For using wilcox.exact()
library(Spaniel)         # For using Spaniel

# Databases
library(org.Hs.eg.db)    # For using human genome annotation package
library(DOSE)            # For using DB for Disease Ontology
library(ontoProc)        # For using DB for Cell Ontology
library(clusterProfiler) # For using DB for Gene Ontology
library(GOSemSim)        # For using godata() in GOSemSim package
library(KEGGREST)        # For using KEGG
```

```{r}
sessionInfo()
```

```
R version 4.0.4 (2021-02-15)
Platform: x86_64-apple-darwin17.0 (64-bit)
Running under: macOS Big Sur 11.1

Matrix products: default
LAPACK: /Library/Frameworks/R.framework/Versions/4.0/Resources/lib/libRlapack.dylib

locale:
[1] en_US.UTF-8/en_US.UTF-8/en_US.UTF-8/C/en_US.UTF-8/en_US.UTF-8

attached base packages:
 [1] parallel  stats4    grid      stats     graphics  grDevices utils     datasets  methods   base     

other attached packages:
 [1] KEGGREST_1.30.1             GOSemSim_2.16.1             clusterProfiler_3.18.1      ontoProc_1.12.0            
 [5] ontologyIndex_2.7           DOSE_3.16.0                 org.Hs.eg.db_3.12.0         AnnotationDbi_1.52.0       
 [9] Spaniel_1.4.0               SingleCellExperiment_1.12.0 SummarizedExperiment_1.20.0 Biobase_2.50.0             
[13] GenomicRanges_1.42.0        GenomeInfoDb_1.26.7         IRanges_2.24.1              S4Vectors_0.28.1           
[17] BiocGenerics_0.36.1         MatrixGenerics_1.2.1        matrixStats_0.60.0          exactRankTests_0.8-32      
[21] merlot_0.2.2                destiny_3.4.0               SeuratObject_4.0.2          Seurat_4.0.3               
[25] cluster_2.1.2               Rcpp_1.0.7                  umap_0.2.7.0                Rtsne_0.15                 
[29] bayNorm_1.8.0               viridis_0.6.1               viridisLite_0.4.0           circlize_0.4.13            
[33] ComplexHeatmap_2.6.2        plot3D_1.4                  ggrepel_0.9.1               ggExtra_0.9                
[37] DT_0.18                     Matrix_1.3-4                forcats_0.5.1               stringr_1.4.0              
[41] dplyr_1.0.7                 purrr_0.3.4                 readr_2.0.1                 tidyr_1.1.3                
[45] tibble_3.1.3                ggplot2_3.3.5               tidyverse_1.3.1            

loaded via a namespace (and not attached):
  [1] rgl_0.107.10            vcd_1.4-8               ica_1.0-2               class_7.3-19           
  [5] foreach_1.5.1           lmtest_0.9-38           crayon_1.4.1            laeken_0.5.1           
  [9] spatstat.core_2.3-0     MASS_7.3-54             nlme_3.1-152            backports_1.2.1        
 [13] reprex_2.0.1            rlang_0.4.11            XVector_0.30.0          ROCR_1.0-11            
 [17] readxl_1.3.1            irlba_2.3.3             smoother_1.1            BiocParallel_1.24.1    
 [21] rjson_0.2.20            bit64_4.0.5             glue_1.4.2              sctransform_0.3.2      
 [25] spatstat.sparse_2.0-0   dotCall64_1.0-1         tcltk_4.0.4             spatstat.geom_2.2-2    
 [29] haven_2.4.3             tidyselect_1.1.1        rio_0.5.27              fitdistrplus_1.1-5     
 [33] zoo_1.8-9               xtable_1.8-4            RcppHNSW_0.3.0          magrittr_2.0.1         
 [37] evaluate_0.14           cli_3.0.1               zlibbioc_1.36.0         rstudioapi_0.13        
 [41] miniUI_0.1.1.1          sp_1.4-5                rpart_4.1-15            fastmatch_1.1-3        
 [45] RcppEigen_0.3.3.9.1     maps_3.3.0              fields_12.5             shiny_1.6.0            
 [49] xfun_0.25               askpass_1.1             clue_0.3-59             tidygraph_1.2.0        
 [53] pcaMethods_1.82.0       doSNOW_1.0.19           ElPiGraph.R_1.0.0       listenv_0.8.0          
 [57] Biostrings_2.58.0       png_0.1-7               future_1.21.0           withr_2.4.2            
 [61] bitops_1.0-7            ggforce_0.3.3           ranger_0.13.1           plyr_1.8.6             
 [65] cellranger_1.1.0        e1071_1.7-8             pillar_1.6.2            GlobalOptions_0.1.2    
 [69] cachem_1.0.5            fs_1.5.0                scatterplot3d_0.3-41    TTR_0.24.2             
 [73] GetoptLong_1.0.5        xts_0.12.1              vctrs_0.3.8             ellipsis_0.3.2         
 [77] generics_0.1.0          tools_4.0.4             foreign_0.8-81          munsell_0.5.0          
 [81] tweenr_1.0.2            fgsea_1.16.0            proxy_0.4-26            DelayedArray_0.16.3    
 [85] fastmap_1.1.0           compiler_4.0.4          paintmap_1.0            abind_1.4-5            
 [89] httpuv_1.6.1            plotly_4.9.4.1          GenomeInfoDbData_1.2.4  gridExtra_2.3          
 [93] lattice_0.20-44         deldir_0.2-10           snow_0.4-3              utf8_1.2.2             
 [97] later_1.2.0             BiocFileCache_1.14.0    jsonlite_1.7.2          ggplot.multistats_1.0.0
[101] scales_1.1.1            graph_1.68.0            pbapply_1.4-3           carData_3.0-4          
[105] lazyeval_0.2.2          promises_1.2.0.1        car_3.0-11              goftest_1.2-2          
[109] spatstat.utils_2.2-0    reticulate_1.20         rmarkdown_2.10          openxlsx_4.2.4         
[113] cowplot_1.1.1           downloader_0.4          uwot_0.1.10             igraph_1.2.6           
[117] survival_3.2-12         yaml_2.2.1              htmltools_0.5.1.1       memoise_2.0.0          
[121] locfit_1.5-9.4          graphlayouts_0.7.1      digest_0.6.27           assertthat_0.2.1       
[125] mime_0.11               rappdirs_0.3.3          spam_2.7-0              RSQLite_2.2.7          
[129] LSD_4.1-0               future.apply_1.8.1      misc3d_0.9-0            data.table_1.14.0      
[133] blob_1.2.2              splines_4.0.4           Cairo_1.5-12.2          RCurl_1.98-1.3         
[137] broom_0.7.9             hms_1.1.0               modelr_0.1.8            colorspace_2.0-2       
[141] BiocManager_1.30.16     shape_1.4.6             nnet_7.3-16             RANN_2.6.1             
[145] enrichplot_1.10.2       fansi_0.5.0             tzdb_0.1.2              VIM_6.1.1              
[149] parallelly_1.27.0       R6_2.5.0                ggridges_0.5.3          lifecycle_1.0.0        
[153] zip_2.2.0               curl_4.3.2              leiden_0.3.9            robustbase_0.93-8      
[157] DO.db_2.9               qvalue_2.22.0           RcppAnnoy_0.0.19        RColorBrewer_1.1-2     
[161] iterators_1.0.13        htmlwidgets_1.5.3       polyclip_1.10-0         crosstalk_1.1.1        
[165] shadowtext_0.0.8        rvest_1.0.1             mgcv_1.8-36             globals_0.14.0         
[169] openssl_1.4.4           patchwork_1.1.1         codetools_0.2-18        lubridate_1.7.10       
[173] GO.db_3.12.1            dbplyr_2.1.1            RSpectra_0.16-0         gtable_0.3.0           
[177] DBI_1.1.1               tensor_1.5              httr_1.4.2              KernSmooth_2.23-20     
[181] stringi_1.7.3           reshape2_1.4.4          farver_2.1.0            ontologyPlot_1.6       
[185] ggthemes_4.2.4          hexbin_1.28.2           Rgraphviz_2.34.0        xml2_1.3.2             
[189] rvcheck_0.1.8           boot_1.3-28             scattermore_0.7         DEoptimR_1.0-9         
[193] bit_4.0.4               scatterpie_0.1.6        spatstat.data_2.1-0     ggraph_2.0.5           
[197] pkgconfig_2.0.3         knitr_1.33 
```
