<!DOCTYPE html>
<html lang="" xml:lang="">
<head>

  <meta charset="utf-8" />
  <meta http-equiv="X-UA-Compatible" content="IE=edge" />
  <title>Chapter 14 Appendix A: comparing performances of ASURAT and existing methods | Introduction to ASURAT (version 0.0.0.9000)</title>
  <meta name="description" content="This is a developmental version of ASURAT." />
  <meta name="generator" content="bookdown 0.21 and GitBook 2.6.7" />

  <meta property="og:title" content="Chapter 14 Appendix A: comparing performances of ASURAT and existing methods | Introduction to ASURAT (version 0.0.0.9000)" />
  <meta property="og:type" content="book" />
  
  
  <meta property="og:description" content="This is a developmental version of ASURAT." />
  

  <meta name="twitter:card" content="summary" />
  <meta name="twitter:title" content="Chapter 14 Appendix A: comparing performances of ASURAT and existing methods | Introduction to ASURAT (version 0.0.0.9000)" />
  
  <meta name="twitter:description" content="This is a developmental version of ASURAT." />
  

<meta name="author" content="Keita Iida" />


<meta name="date" content="2021-06-04" />

  <meta name="viewport" content="width=device-width, initial-scale=1" />
  <meta name="apple-mobile-web-app-capable" content="yes" />
  <meta name="apple-mobile-web-app-status-bar-style" content="black" />
  
  
<link rel="prev" href="multiple-sign-analysis-by-concatenating-do-co-go-kegg-and-reactome.html"/>
<link rel="next" href="appendix-b-automatically-tuning-the-asurats-parameters.html"/>
<script src="libs/jquery-2.2.3/jquery.min.js"></script>
<link href="libs/gitbook-2.6.7/css/style.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-table.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-bookdown.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-highlight.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-search.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-fontsettings.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-clipboard.css" rel="stylesheet" />









<script src="libs/accessible-code-block-0.0.1/empty-anchor.js"></script>


<style type="text/css">
code.sourceCode > span { display: inline-block; line-height: 1.25; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode { white-space: pre; position: relative; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
code.sourceCode { white-space: pre-wrap; }
code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } /* Alert */
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code span.at { color: #7d9029; } /* Attribute */
code span.bn { color: #40a070; } /* BaseN */
code span.bu { } /* BuiltIn */
code span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code span.ch { color: #4070a0; } /* Char */
code span.cn { color: #880000; } /* Constant */
code span.co { color: #60a0b0; font-style: italic; } /* Comment */
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code span.do { color: #ba2121; font-style: italic; } /* Documentation */
code span.dt { color: #902000; } /* DataType */
code span.dv { color: #40a070; } /* DecVal */
code span.er { color: #ff0000; font-weight: bold; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #40a070; } /* Float */
code span.fu { color: #06287e; } /* Function */
code span.im { } /* Import */
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
code span.kw { color: #007020; font-weight: bold; } /* Keyword */
code span.op { color: #666666; } /* Operator */
code span.ot { color: #007020; } /* Other */
code span.pp { color: #bc7a00; } /* Preprocessor */
code span.sc { color: #4070a0; } /* SpecialChar */
code span.ss { color: #bb6688; } /* SpecialString */
code span.st { color: #4070a0; } /* String */
code span.va { color: #19177c; } /* Variable */
code span.vs { color: #4070a0; } /* VerbatimString */
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */
</style>

<link rel="stylesheet" href="style.css" type="text/css" />
</head>

<body>



  <div class="book without-animation with-summary font-size-2 font-family-1" data-basepath=".">

    <div class="book-summary">
      <nav role="navigation">

<ul class="summary">
<li><a href="./">A Minimal Book Example</a></li>

<li class="divider"></li>
<li class="chapter" data-level="1" data-path="index.html"><a href="index.html"><i class="fa fa-check"></i><b>1</b> Overview</a><ul>
<li class="chapter" data-level="1.1" data-path="index.html"><a href="index.html#license-and-description"><i class="fa fa-check"></i><b>1.1</b> License and description</a></li>
<li class="chapter" data-level="1.2" data-path="index.html"><a href="index.html#computational-environment"><i class="fa fa-check"></i><b>1.2</b> Computational environment</a></li>
<li class="chapter" data-level="1.3" data-path="index.html"><a href="index.html#installation"><i class="fa fa-check"></i><b>1.3</b> Installation</a></li>
</ul></li>
<li class="chapter" data-level="2" data-path="dataset.html"><a href="dataset.html"><i class="fa fa-check"></i><b>2</b> Preparing data sets</a><ul>
<li class="chapter" data-level="2.1" data-path="dataset.html"><a href="dataset.html#case-i-input-gene-expression-matrix-with-row-and-column-names"><i class="fa fa-check"></i><b>2.1</b> Case I: input gene expression matrix with row and column names</a></li>
<li class="chapter" data-level="2.2" data-path="dataset.html"><a href="dataset.html#case-ii-input-10x-data"><i class="fa fa-check"></i><b>2.2</b> Case II: input 10x data</a></li>
<li class="chapter" data-level="2.3" data-path="dataset.html"><a href="dataset.html#case-iii-input-10x-pbmc-data"><i class="fa fa-check"></i><b>2.3</b> Case III: input 10x PBMC data</a></li>
</ul></li>
<li class="chapter" data-level="3" data-path="dataqc.html"><a href="dataqc.html"><i class="fa fa-check"></i><b>3</b> Data quality control (QC)</a><ul>
<li class="chapter" data-level="3.1" data-path="dataqc.html"><a href="dataqc.html#quick-qc"><i class="fa fa-check"></i><b>3.1</b> Quick QC</a></li>
<li class="chapter" data-level="3.2" data-path="dataqc.html"><a href="dataqc.html#sample-qc"><i class="fa fa-check"></i><b>3.2</b> Sample QC</a></li>
<li class="chapter" data-level="3.3" data-path="dataqc.html"><a href="dataqc.html#variable-qc"><i class="fa fa-check"></i><b>3.3</b> Variable QC</a></li>
</ul></li>
<li class="chapter" data-level="4" data-path="normalization.html"><a href="normalization.html"><i class="fa fa-check"></i><b>4</b> Normalizing and centering data</a></li>
<li class="chapter" data-level="5" data-path="computcorr.html"><a href="computcorr.html"><i class="fa fa-check"></i><b>5</b> Computing correlations among genes</a></li>
<li class="chapter" data-level="6" data-path="markergenes.html"><a href="markergenes.html"><i class="fa fa-check"></i><b>6</b> Checking expression profiles of marker genes</a></li>
<li class="chapter" data-level="7" data-path="collectdb.html"><a href="collectdb.html"><i class="fa fa-check"></i><b>7</b> Collecting databases (optional)</a><ul>
<li class="chapter" data-level="7.1" data-path="collectdb.html"><a href="collectdb.html#disease-ontology-db"><i class="fa fa-check"></i><b>7.1</b> Disease Ontology DB</a></li>
<li class="chapter" data-level="7.2" data-path="collectdb.html"><a href="collectdb.html#cell-ontology-db"><i class="fa fa-check"></i><b>7.2</b> Cell Ontology DB</a></li>
<li class="chapter" data-level="7.3" data-path="collectdb.html"><a href="collectdb.html#gene-ontology-db"><i class="fa fa-check"></i><b>7.3</b> Gene Ontology DB</a></li>
<li class="chapter" data-level="7.4" data-path="collectdb.html"><a href="collectdb.html#kegg"><i class="fa fa-check"></i><b>7.4</b> KEGG</a></li>
<li class="chapter" data-level="7.5" data-path="collectdb.html"><a href="collectdb.html#reactome"><i class="fa fa-check"></i><b>7.5</b> Reactome</a></li>
</ul></li>
<li class="chapter" data-level="8" data-path="asurat-using-disease-ontology-database.html"><a href="asurat-using-disease-ontology-database.html"><i class="fa fa-check"></i><b>8</b> ASURAT using Disease Ontology database</a><ul>
<li class="chapter" data-level="8.1" data-path="asurat-using-disease-ontology-database.html"><a href="asurat-using-disease-ontology-database.html#creating-sign-by-sample-matrices"><i class="fa fa-check"></i><b>8.1</b> Creating sign-by-sample matrices</a></li>
<li class="chapter" data-level="8.2" data-path="asurat-using-disease-ontology-database.html"><a href="asurat-using-disease-ontology-database.html#dimension-reduction"><i class="fa fa-check"></i><b>8.2</b> Dimension reduction</a></li>
<li class="chapter" data-level="8.3" data-path="asurat-using-disease-ontology-database.html"><a href="asurat-using-disease-ontology-database.html#sample-clustering"><i class="fa fa-check"></i><b>8.3</b> Sample clustering</a></li>
<li class="chapter" data-level="8.4" data-path="asurat-using-disease-ontology-database.html"><a href="asurat-using-disease-ontology-database.html#visualizing"><i class="fa fa-check"></i><b>8.4</b> Visualizing</a></li>
<li class="chapter" data-level="8.5" data-path="asurat-using-disease-ontology-database.html"><a href="asurat-using-disease-ontology-database.html#significant-signs"><i class="fa fa-check"></i><b>8.5</b> Significant signs</a></li>
</ul></li>
<li class="chapter" data-level="9" data-path="asurat-using-cell-ontology-database-optional.html"><a href="asurat-using-cell-ontology-database-optional.html"><i class="fa fa-check"></i><b>9</b> ASURAT using Cell Ontology database (optional)</a><ul>
<li class="chapter" data-level="9.1" data-path="asurat-using-cell-ontology-database-optional.html"><a href="asurat-using-cell-ontology-database-optional.html#creating-sign-by-sample-matrices-1"><i class="fa fa-check"></i><b>9.1</b> Creating sign-by-sample matrices</a></li>
<li class="chapter" data-level="9.2" data-path="asurat-using-cell-ontology-database-optional.html"><a href="asurat-using-cell-ontology-database-optional.html#dimension-reduction-1"><i class="fa fa-check"></i><b>9.2</b> Dimension reduction</a></li>
<li class="chapter" data-level="9.3" data-path="asurat-using-cell-ontology-database-optional.html"><a href="asurat-using-cell-ontology-database-optional.html#sample-clustering-1"><i class="fa fa-check"></i><b>9.3</b> Sample clustering</a></li>
<li class="chapter" data-level="9.4" data-path="asurat-using-cell-ontology-database-optional.html"><a href="asurat-using-cell-ontology-database-optional.html#visualizing-1"><i class="fa fa-check"></i><b>9.4</b> Visualizing</a></li>
<li class="chapter" data-level="9.5" data-path="asurat-using-cell-ontology-database-optional.html"><a href="asurat-using-cell-ontology-database-optional.html#significant-signs-1"><i class="fa fa-check"></i><b>9.5</b> Significant signs</a></li>
</ul></li>
<li class="chapter" data-level="10" data-path="asurat-using-gene-ontology-database-optional.html"><a href="asurat-using-gene-ontology-database-optional.html"><i class="fa fa-check"></i><b>10</b> ASURAT using Gene Ontology database (optional)</a><ul>
<li class="chapter" data-level="10.1" data-path="asurat-using-gene-ontology-database-optional.html"><a href="asurat-using-gene-ontology-database-optional.html#creating-sign-by-sample-matrices-2"><i class="fa fa-check"></i><b>10.1</b> Creating sign-by-sample matrices</a></li>
<li class="chapter" data-level="10.2" data-path="asurat-using-gene-ontology-database-optional.html"><a href="asurat-using-gene-ontology-database-optional.html#dimension-reduction-2"><i class="fa fa-check"></i><b>10.2</b> Dimension reduction</a></li>
<li class="chapter" data-level="10.3" data-path="asurat-using-gene-ontology-database-optional.html"><a href="asurat-using-gene-ontology-database-optional.html#sample-clustering-2"><i class="fa fa-check"></i><b>10.3</b> Sample clustering</a></li>
<li class="chapter" data-level="10.4" data-path="asurat-using-gene-ontology-database-optional.html"><a href="asurat-using-gene-ontology-database-optional.html#visualizing-2"><i class="fa fa-check"></i><b>10.4</b> Visualizing</a></li>
<li class="chapter" data-level="10.5" data-path="asurat-using-gene-ontology-database-optional.html"><a href="asurat-using-gene-ontology-database-optional.html#significant-signs-2"><i class="fa fa-check"></i><b>10.5</b> Significant signs</a></li>
</ul></li>
<li class="chapter" data-level="11" data-path="asurat-using-kegg-optional.html"><a href="asurat-using-kegg-optional.html"><i class="fa fa-check"></i><b>11</b> ASURAT using KEGG (optional)</a><ul>
<li class="chapter" data-level="11.1" data-path="asurat-using-kegg-optional.html"><a href="asurat-using-kegg-optional.html#creating-sign-by-sample-matrices-3"><i class="fa fa-check"></i><b>11.1</b> Creating sign-by-sample matrices</a></li>
<li class="chapter" data-level="11.2" data-path="asurat-using-kegg-optional.html"><a href="asurat-using-kegg-optional.html#dimension-reduction-3"><i class="fa fa-check"></i><b>11.2</b> Dimension reduction</a></li>
<li class="chapter" data-level="11.3" data-path="asurat-using-kegg-optional.html"><a href="asurat-using-kegg-optional.html#sample-clustering-3"><i class="fa fa-check"></i><b>11.3</b> Sample clustering</a></li>
<li class="chapter" data-level="11.4" data-path="asurat-using-kegg-optional.html"><a href="asurat-using-kegg-optional.html#visualizing-3"><i class="fa fa-check"></i><b>11.4</b> Visualizing</a></li>
<li class="chapter" data-level="11.5" data-path="asurat-using-kegg-optional.html"><a href="asurat-using-kegg-optional.html#significant-signs-3"><i class="fa fa-check"></i><b>11.5</b> Significant signs</a></li>
</ul></li>
<li class="chapter" data-level="12" data-path="asurat-using-reactome-optional.html"><a href="asurat-using-reactome-optional.html"><i class="fa fa-check"></i><b>12</b> ASURAT using Reactome (optional)</a><ul>
<li class="chapter" data-level="12.1" data-path="asurat-using-reactome-optional.html"><a href="asurat-using-reactome-optional.html#creating-sign-by-sample-matrices-4"><i class="fa fa-check"></i><b>12.1</b> Creating sign-by-sample matrices</a></li>
<li class="chapter" data-level="12.2" data-path="asurat-using-reactome-optional.html"><a href="asurat-using-reactome-optional.html#dimension-reduction-4"><i class="fa fa-check"></i><b>12.2</b> Dimension reduction</a></li>
<li class="chapter" data-level="12.3" data-path="asurat-using-reactome-optional.html"><a href="asurat-using-reactome-optional.html#sample-clustering-4"><i class="fa fa-check"></i><b>12.3</b> Sample clustering</a></li>
</ul></li>
<li class="chapter" data-level="13" data-path="multiple-sign-analysis-by-concatenating-do-co-go-kegg-and-reactome.html"><a href="multiple-sign-analysis-by-concatenating-do-co-go-kegg-and-reactome.html"><i class="fa fa-check"></i><b>13</b> Multiple sign analysis by concatenating DO, CO, GO, KEGG, and Reactome</a></li>
<li class="chapter" data-level="14" data-path="appendix-a-comparing-performances-of-asurat-and-existing-methods.html"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html"><i class="fa fa-check"></i><b>14</b> Appendix A: comparing performances of ASURAT and existing methods</a><ul>
<li class="chapter" data-level="14.1" data-path="appendix-a-comparing-performances-of-asurat-and-existing-methods.html"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#installation-1"><i class="fa fa-check"></i><b>14.1</b> Installation</a></li>
<li class="chapter" data-level="14.2" data-path="appendix-a-comparing-performances-of-asurat-and-existing-methods.html"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#removing-the-outlier-cells-detected-by-asurat"><i class="fa fa-check"></i><b>14.2</b> Removing the outlier cells detected by ASURAT</a></li>
<li class="chapter" data-level="14.3" data-path="appendix-a-comparing-performances-of-asurat-and-existing-methods.html"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#assessing-clustering-performances"><i class="fa fa-check"></i><b>14.3</b> Assessing clustering performances</a><ul>
<li class="chapter" data-level="14.3.1" data-path="appendix-a-comparing-performances-of-asurat-and-existing-methods.html"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#seurat"><i class="fa fa-check"></i><b>14.3.1</b> Seurat</a></li>
<li class="chapter" data-level="14.3.2" data-path="appendix-a-comparing-performances-of-asurat-and-existing-methods.html"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#monocle-3"><i class="fa fa-check"></i><b>14.3.2</b> Monocle 3</a></li>
<li class="chapter" data-level="14.3.3" data-path="appendix-a-comparing-performances-of-asurat-and-existing-methods.html"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#sc3"><i class="fa fa-check"></i><b>14.3.3</b> SC3</a></li>
<li class="chapter" data-level="14.3.4" data-path="appendix-a-comparing-performances-of-asurat-and-existing-methods.html"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#principal-component-analysis-pca"><i class="fa fa-check"></i><b>14.3.4</b> Principal component analysis (PCA)</a></li>
<li class="chapter" data-level="14.3.5" data-path="appendix-a-comparing-performances-of-asurat-and-existing-methods.html"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#asurat"><i class="fa fa-check"></i><b>14.3.5</b> ASURAT</a></li>
<li class="chapter" data-level="14.3.6" data-path="appendix-a-comparing-performances-of-asurat-and-existing-methods.html"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#comparing-clustering-validity-indices"><i class="fa fa-check"></i><b>14.3.6</b> Comparing clustering validity indices</a></li>
</ul></li>
<li class="chapter" data-level="14.4" data-path="appendix-a-comparing-performances-of-asurat-and-existing-methods.html"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cell-type-inference-in-pbmc_4000-and-pbmc_6000"><i class="fa fa-check"></i><b>14.4</b> Cell-type inference in <code>pbmc_4000</code> and <code>pbmc_6000</code></a><ul>
<li class="chapter" data-level="14.4.1" data-path="appendix-a-comparing-performances-of-asurat-and-existing-methods.html"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#seurat-1"><i class="fa fa-check"></i><b>14.4.1</b> Seurat</a></li>
<li class="chapter" data-level="14.4.2" data-path="appendix-a-comparing-performances-of-asurat-and-existing-methods.html"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#monocle-3-1"><i class="fa fa-check"></i><b>14.4.2</b> Monocle 3</a></li>
<li class="chapter" data-level="14.4.3" data-path="appendix-a-comparing-performances-of-asurat-and-existing-methods.html"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#sc3-1"><i class="fa fa-check"></i><b>14.4.3</b> SC3</a></li>
<li class="chapter" data-level="14.4.4" data-path="appendix-a-comparing-performances-of-asurat-and-existing-methods.html"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#asurat-1"><i class="fa fa-check"></i><b>14.4.4</b> ASURAT</a></li>
</ul></li>
<li class="chapter" data-level="14.5" data-path="appendix-a-comparing-performances-of-asurat-and-existing-methods.html"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#identifying-cell-types-in-pbmc_4000-and-pbmc_6000"><i class="fa fa-check"></i><b>14.5</b> Identifying cell types in <code>pbmc_4000</code> and <code>pbmc_6000</code></a></li>
</ul></li>
<li class="chapter" data-level="15" data-path="appendix-b-automatically-tuning-the-asurats-parameters.html"><a href="appendix-b-automatically-tuning-the-asurats-parameters.html"><i class="fa fa-check"></i><b>15</b> Appendix B: automatically tuning the ASURAT’s parameters</a></li>
<li class="divider"></li>
<li><a href="https://github.com/rstudio/bookdown" target="blank">Published with bookdown</a></li>

</ul>

      </nav>
    </div>

    <div class="book-body">
      <div class="body-inner">
        <div class="book-header" role="navigation">
          <h1>
            <i class="fa fa-circle-o-notch fa-spin"></i><a href="./">Introduction to ASURAT (version 0.0.0.9000)</a>
          </h1>
        </div>

        <div class="page-wrapper" tabindex="-1" role="main">
          <div class="page-inner">

            <section class="normal" id="section-">
<div id="appendix-a-comparing-performances-of-asurat-and-existing-methods" class="section level1">
<h1><span class="header-section-number">Chapter 14</span> Appendix A: comparing performances of ASURAT and existing methods</h1>
<p>In this section, we assess the clustering performances of existing algorithms
and ASURAT, which are listed below:</p>
<ul>
<li>Seurat 4 (version 4.0.1)</li>
<li>Monocle 3 (version 0.2.3.0)</li>
<li>SC3 (version 1.18.0)</li>
<li>Principal component analysis (PCA)</li>
<li>ASURAT (version 0.0.0.9000)</li>
</ul>
<p>Our strategy to compare the clustering performances consists of the following
steps:</p>
<ol style="list-style-type: decimal">
<li><p>prepare cancer patient- and healthy donor-derived single-cell RNA sequencing
datasets, which are symbolized by <code>day1_norm</code>, <code>day7_hypo</code>, <code>sc68_vehi</code>,
<code>sc68_cisp</code>, <code>pbmc_4000</code>, and <code>pbmc_6000</code>.</p></li>
<li><p>remove the outlier cells from <code>sc68_vehi</code>, which is detected by ASURAT;</p></li>
<li><p>obtain the low-dimensional representations by means of t-SNE and UMAP
except for SC3 because SC3 uses distance matrices for sample clustering;</p></li>
<li><p>perform k-means clusterings upon the results of the low-dimensional
representations;</p></li>
<li><p>compute several indices measuring clustering quality by using <code>NbClust</code>;</p></li>
<li><p>compare the indices across the above clustering methods; and</p></li>
<li><p>infer the cell types in the datasets, <code>pmbc_4000</code> and <code>pbmc_6000</code>, by using
Seurat, Monocle3 plus Garnett, SC3, and ASURAT.</p></li>
</ol>
<div id="installation-1" class="section level2">
<h2><span class="header-section-number">14.1</span> Installation</h2>
<p>The following libraries are used.</p>
<div class="sourceCode" id="cb1"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb1-1"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb1-1"></a><span class="co"># General</span></span>
<span id="cb1-2"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb1-2"></a><span class="kw">library</span>(tidyverse)       <span class="co"># For efficient handling of data.frame</span></span>
<span id="cb1-3"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb1-3"></a><span class="co"># Below is a memo for attaching org.Hs.eg.db</span></span>
<span id="cb1-4"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb1-4"></a><span class="co"># options(connectionObserver = NULL)</span></span>
<span id="cb1-5"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb1-5"></a><span class="kw">library</span>(org.Hs.eg.db)    <span class="co"># For using human genome annotation package</span></span>
<span id="cb1-6"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb1-6"></a></span>
<span id="cb1-7"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb1-7"></a><span class="co"># Plot</span></span>
<span id="cb1-8"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb1-8"></a><span class="kw">library</span>(ggExtra)         <span class="co"># For graphics</span></span>
<span id="cb1-9"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb1-9"></a><span class="kw">library</span>(ggrepel)         <span class="co"># For graphics with labeling</span></span>
<span id="cb1-10"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb1-10"></a><span class="kw">library</span>(plot3D)          <span class="co"># For 3-dimensional scatter plot</span></span>
<span id="cb1-11"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb1-11"></a><span class="kw">library</span>(ComplexHeatmap)  <span class="co"># For ComplexHeatmap</span></span>
<span id="cb1-12"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb1-12"></a><span class="kw">library</span>(circlize)        <span class="co"># For color mapping function in ComplexHeatmap</span></span>
<span id="cb1-13"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb1-13"></a><span class="kw">library</span>(DT)              <span class="co"># For using datatable</span></span>
<span id="cb1-14"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb1-14"></a><span class="kw">library</span>(viridis)         <span class="co"># For using color palette `plasma`</span></span>
<span id="cb1-15"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb1-15"></a><span class="kw">library</span>(RColorBrewer)    <span class="co"># For using color palette</span></span>
<span id="cb1-16"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb1-16"></a></span>
<span id="cb1-17"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb1-17"></a><span class="co"># Utilities</span></span>
<span id="cb1-18"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb1-18"></a><span class="kw">library</span>(NbClust)         <span class="co"># For using NbClust</span></span>
<span id="cb1-19"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb1-19"></a><span class="kw">library</span>(Rtsne)           <span class="co"># For using Rtsne</span></span>
<span id="cb1-20"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb1-20"></a><span class="kw">library</span>(umap)            <span class="co"># For using umap</span></span>
<span id="cb1-21"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb1-21"></a><span class="kw">library</span>(Rcpp)            <span class="co"># For using external C++ codes</span></span>
<span id="cb1-22"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb1-22"></a><span class="kw">library</span>(cluster)         <span class="co"># For using pam</span></span>
<span id="cb1-23"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb1-23"></a><span class="kw">library</span>(Seurat)          <span class="co"># For using Seurat</span></span>
<span id="cb1-24"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb1-24"></a><span class="kw">library</span>(garnett)         <span class="co"># For using Garnett along with monocle 3</span></span>
<span id="cb1-25"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb1-25"></a><span class="kw">library</span>(monocle3)        <span class="co"># For using Monocle</span></span>
<span id="cb1-26"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb1-26"></a><span class="kw">library</span>(SC3)                    <span class="co"># For using SC3</span></span>
<span id="cb1-27"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb1-27"></a><span class="kw">library</span>(scater)                 <span class="co"># For using SC3</span></span>
<span id="cb1-28"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb1-28"></a><span class="kw">library</span>(SingleCellExperiment)   <span class="co"># For using SC3</span></span>
<span id="cb1-29"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb1-29"></a><span class="kw">library</span>(openxlsx)        <span class="co"># For outputting excel file with multiple sheets</span></span>
<span id="cb1-30"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb1-30"></a></span>
<span id="cb1-31"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb1-31"></a><span class="co"># Databases</span></span>
<span id="cb1-32"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb1-32"></a><span class="kw">library</span>(DOSE)            <span class="co"># For using DB for Disease Ontology</span></span>
<span id="cb1-33"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb1-33"></a><span class="kw">library</span>(ontoProc)        <span class="co"># For using DB for Cell Ontology</span></span>
<span id="cb1-34"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb1-34"></a><span class="kw">library</span>(clusterProfiler) <span class="co"># For using DB for Gene Ontology</span></span>
<span id="cb1-35"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb1-35"></a><span class="kw">library</span>(GOSemSim)        <span class="co"># For using godata in GOSemSim package</span></span></code></pre></div>
<div class="sourceCode" id="cb2"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb2-1"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb2-1"></a><span class="kw">packageVersion</span>(<span class="st">&quot;Seurat&quot;</span>)</span></code></pre></div>
<pre><code>## [1] &#39;4.0.1&#39;</code></pre>
<div class="sourceCode" id="cb4"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb4-1"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb4-1"></a><span class="kw">packageVersion</span>(<span class="st">&quot;monocle3&quot;</span>)</span></code></pre></div>
<pre><code>## [1] &#39;0.2.3.0&#39;</code></pre>
<div class="sourceCode" id="cb6"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb6-1"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb6-1"></a><span class="kw">packageVersion</span>(<span class="st">&quot;garnett&quot;</span>)</span></code></pre></div>
<pre><code>## [1] &#39;0.2.17&#39;</code></pre>
<div class="sourceCode" id="cb8"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb8-1"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb8-1"></a><span class="kw">packageVersion</span>(<span class="st">&quot;SC3&quot;</span>)</span></code></pre></div>
<pre><code>## [1] &#39;1.18.0&#39;</code></pre>
</div>
<div id="removing-the-outlier-cells-detected-by-asurat" class="section level2">
<h2><span class="header-section-number">14.2</span> Removing the outlier cells detected by ASURAT</h2>
<p>Below are the R codes to obtain the low-dimensional representation of the data.
Please see Chapter 8 “ASURAT using Disease Ontology DB” for the details.</p>
<div class="sourceCode" id="cb10"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb10-1"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb10-1"></a><span class="kw">rm</span>(<span class="dt">list =</span> <span class="kw">ls</span>())</span>
<span id="cb10-2"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb10-2"></a><span class="kw">source</span>(<span class="st">&quot;R/plot.R&quot;</span>)</span>
<span id="cb10-3"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb10-3"></a><span class="kw">source</span>(<span class="st">&quot;R/function_sign.R&quot;</span>)</span>
<span id="cb10-4"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb10-4"></a></span>
<span id="cb10-5"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb10-5"></a>sc68_vehi &lt;-<span class="st"> </span><span class="kw">readRDS</span>(<span class="dt">file =</span> <span class="st">&quot;backup/03_104_sc68_vehi_DES_DO.rds&quot;</span>)</span>
<span id="cb10-6"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb10-6"></a>sc68_vehi &lt;-<span class="st"> </span><span class="kw">do_umap_sign</span>(<span class="dt">obj =</span> sc68_vehi,</span>
<span id="cb10-7"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb10-7"></a>                          <span class="dt">data_type =</span> <span class="st">&quot;DO&quot;</span>, <span class="dt">category =</span> <span class="st">&quot;disease&quot;</span>,</span>
<span id="cb10-8"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb10-8"></a>                          <span class="dt">pca_dim =</span> <span class="ot">NULL</span>, <span class="dt">umap_dim =</span> <span class="dv">2</span>)</span>
<span id="cb10-9"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb10-9"></a><span class="co"># ----------------------------------------</span></span>
<span id="cb10-10"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb10-10"></a><span class="co"># For `sc68_vehi`</span></span>
<span id="cb10-11"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb10-11"></a><span class="co"># ----------------------------------------</span></span>
<span id="cb10-12"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb10-12"></a>p &lt;-<span class="st"> </span><span class="kw">plot_umap_sign</span>(<span class="dt">obj =</span> sc68_vehi,</span>
<span id="cb10-13"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb10-13"></a>                    <span class="dt">data_type =</span> <span class="st">&quot;DO&quot;</span>, <span class="dt">category =</span> <span class="st">&quot;disease&quot;</span>,</span>
<span id="cb10-14"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb10-14"></a>                    <span class="dt">algo_name =</span> <span class="ot">NULL</span>, <span class="dt">theta =</span> <span class="ot">NULL</span>, <span class="dt">phi =</span> <span class="ot">NULL</span>,</span>
<span id="cb10-15"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb10-15"></a>                    <span class="dt">title =</span> <span class="st">&quot;sc68_vehi (DO: disease)&quot;</span>, <span class="dt">title_size =</span> <span class="dv">16</span>,</span>
<span id="cb10-16"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb10-16"></a>                    <span class="dt">xlabel =</span> <span class="st">&quot;UMAP_1&quot;</span>, <span class="dt">ylabel =</span> <span class="st">&quot;UMAP_2&quot;</span>, <span class="dt">zlabel =</span> <span class="ot">NULL</span>)</span>
<span id="cb10-17"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb10-17"></a>filename &lt;-<span class="st"> &quot;figures/figure_13_0010.png&quot;</span></span>
<span id="cb10-18"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb10-18"></a><span class="kw">ggsave</span>(<span class="dt">file =</span> filename, <span class="dt">plot =</span> p, <span class="dt">dpi =</span> <span class="dv">300</span>, <span class="dt">width =</span> <span class="dv">4</span>, <span class="dt">height =</span> <span class="fl">4.2</span>)</span></code></pre></div>
<p><img src="figures/figure_13_0010.png" width="230px"></p>
<p>As can be seen, there exist non-negligible outliers, referred to as <code>blacklist</code>.
Indeed, if the outliers are included in the downstream analysis, we can obtain
very large ASWs depending on the parameter settings, which seems to be unfair.
Hence, we remove these outliers.</p>
<div class="sourceCode" id="cb11"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb11-1"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb11-1"></a><span class="co"># ----------------------------------------</span></span>
<span id="cb11-2"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb11-2"></a><span class="co"># For `sc68_vehi`</span></span>
<span id="cb11-3"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb11-3"></a><span class="co"># ----------------------------------------</span></span>
<span id="cb11-4"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb11-4"></a>mat &lt;-<span class="st"> </span>sc68_vehi[[<span class="st">&quot;reduction&quot;</span>]][[<span class="st">&quot;umap&quot;</span>]][[<span class="st">&quot;DO&quot;</span>]][[<span class="st">&quot;disease&quot;</span>]][[<span class="st">&quot;layout&quot;</span>]]</span>
<span id="cb11-5"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb11-5"></a>tmp &lt;-<span class="st"> </span>mat[<span class="kw">order</span>(mat[,<span class="dv">1</span>], <span class="dt">decreasing =</span> <span class="ot">TRUE</span>),]</span>
<span id="cb11-6"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb11-6"></a>blacklist &lt;-<span class="st"> </span><span class="kw">setdiff</span>(<span class="kw">rownames</span>(tmp), <span class="kw">rownames</span>(tmp[<span class="op">-</span>(<span class="dv">1</span><span class="op">:</span><span class="dv">22</span>),]))</span>
<span id="cb11-7"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb11-7"></a><span class="kw">saveRDS</span>(blacklist, <span class="dt">file =</span> <span class="st">&quot;backup/13_001_sc68_vehi_blacklist.rds&quot;</span>)</span></code></pre></div>
</div>
<div id="assessing-clustering-performances" class="section level2">
<h2><span class="header-section-number">14.3</span> Assessing clustering performances</h2>
<div id="seurat" class="section level3">
<h3><span class="header-section-number">14.3.1</span> Seurat</h3>
<p>Load the data.</p>
<div class="sourceCode" id="cb12"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb12-1"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb12-1"></a><span class="kw">rm</span>(<span class="dt">list =</span> <span class="kw">ls</span>())</span>
<span id="cb12-2"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb12-2"></a><span class="kw">source</span>(<span class="st">&quot;R/function_seurat.R&quot;</span>)</span>
<span id="cb12-3"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb12-3"></a>day1_norm &lt;-<span class="st"> </span><span class="kw">readRDS</span>(<span class="dt">file =</span> <span class="st">&quot;backup/01_005_day1_norm_normalized.rds&quot;</span>)</span>
<span id="cb12-4"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb12-4"></a>day7_hypo &lt;-<span class="st"> </span><span class="kw">readRDS</span>(<span class="dt">file =</span> <span class="st">&quot;backup/02_005_day7_hypo_normalized.rds&quot;</span>)</span>
<span id="cb12-5"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb12-5"></a>sc68_vehi &lt;-<span class="st"> </span><span class="kw">readRDS</span>(<span class="dt">file =</span> <span class="st">&quot;backup/03_005_sc68_vehi_normalized.rds&quot;</span>)</span>
<span id="cb12-6"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb12-6"></a>sc68_cisp &lt;-<span class="st"> </span><span class="kw">readRDS</span>(<span class="dt">file =</span> <span class="st">&quot;backup/04_005_sc68_cisp_normalized.rds&quot;</span>)</span>
<span id="cb12-7"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb12-7"></a>pbmc_<span class="dv">4000</span> &lt;-<span class="st"> </span><span class="kw">readRDS</span>(<span class="dt">file =</span> <span class="st">&quot;backup/05_005_pbmc_4000_normalized.rds&quot;</span>)</span>
<span id="cb12-8"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb12-8"></a>pbmc_<span class="dv">6000</span> &lt;-<span class="st"> </span><span class="kw">readRDS</span>(<span class="dt">file =</span> <span class="st">&quot;backup/06_005_pbmc_6000_normalized.rds&quot;</span>)</span>
<span id="cb12-9"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb12-9"></a>sc68_vehi_bl &lt;-<span class="st"> </span><span class="kw">readRDS</span>(<span class="dt">file =</span> <span class="st">&quot;backup/13_001_sc68_vehi_blacklist.rds&quot;</span>)</span></code></pre></div>
<p>Remove the outlier cells from <code>sc68_vehi</code>.</p>
<div class="sourceCode" id="cb13"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb13-1"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb13-1"></a><span class="co"># ----------------------------------------</span></span>
<span id="cb13-2"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb13-2"></a><span class="co"># For `sc68_vehi`</span></span>
<span id="cb13-3"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb13-3"></a><span class="co"># ----------------------------------------</span></span>
<span id="cb13-4"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb13-4"></a>mat &lt;-<span class="st"> </span>sc68_vehi[[<span class="st">&quot;data&quot;</span>]][[<span class="st">&quot;bayNorm&quot;</span>]][[<span class="st">&quot;Bay_out&quot;</span>]]</span>
<span id="cb13-5"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb13-5"></a>inds &lt;-<span class="st"> </span><span class="kw">which</span>(<span class="kw">colnames</span>(mat) <span class="op">%in%</span><span class="st"> </span>sc68_vehi_bl)</span>
<span id="cb13-6"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb13-6"></a>sc68_vehi[[<span class="st">&quot;data&quot;</span>]][[<span class="st">&quot;bayNorm&quot;</span>]][[<span class="st">&quot;Bay_out&quot;</span>]] &lt;-<span class="st"> </span>mat[, <span class="op">-</span>inds]</span></code></pre></div>
<p>Create Seurat objects.</p>
<div class="sourceCode" id="cb14"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb14-1"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb14-1"></a>day1_norm &lt;-<span class="st"> </span><span class="kw">CreateSeuratObject</span>(</span>
<span id="cb14-2"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb14-2"></a>  <span class="dt">counts =</span> day1_norm[[<span class="st">&quot;data&quot;</span>]][[<span class="st">&quot;bayNorm&quot;</span>]][[<span class="st">&quot;Bay_out&quot;</span>]],</span>
<span id="cb14-3"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb14-3"></a>  <span class="dt">project =</span> day1_norm[[<span class="st">&quot;history&quot;</span>]][[<span class="st">&quot;make_asurat_obj&quot;</span>]][[<span class="st">&quot;obj_name&quot;</span>]])</span>
<span id="cb14-4"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb14-4"></a>day7_hypo &lt;-<span class="st"> </span><span class="kw">CreateSeuratObject</span>(</span>
<span id="cb14-5"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb14-5"></a>  <span class="dt">counts =</span> day7_hypo[[<span class="st">&quot;data&quot;</span>]][[<span class="st">&quot;bayNorm&quot;</span>]][[<span class="st">&quot;Bay_out&quot;</span>]],</span>
<span id="cb14-6"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb14-6"></a>  <span class="dt">project =</span> day7_hypo[[<span class="st">&quot;history&quot;</span>]][[<span class="st">&quot;make_asurat_obj&quot;</span>]][[<span class="st">&quot;obj_name&quot;</span>]])</span>
<span id="cb14-7"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb14-7"></a>sc68_vehi &lt;-<span class="st"> </span><span class="kw">CreateSeuratObject</span>(</span>
<span id="cb14-8"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb14-8"></a>  <span class="dt">counts =</span> sc68_vehi[[<span class="st">&quot;data&quot;</span>]][[<span class="st">&quot;bayNorm&quot;</span>]][[<span class="st">&quot;Bay_out&quot;</span>]],</span>
<span id="cb14-9"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb14-9"></a>  <span class="dt">project =</span> sc68_vehi[[<span class="st">&quot;history&quot;</span>]][[<span class="st">&quot;make_asurat_obj&quot;</span>]][[<span class="st">&quot;obj_name&quot;</span>]])</span>
<span id="cb14-10"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb14-10"></a>sc68_cisp &lt;-<span class="st"> </span><span class="kw">CreateSeuratObject</span>(</span>
<span id="cb14-11"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb14-11"></a>  <span class="dt">counts =</span> sc68_cisp[[<span class="st">&quot;data&quot;</span>]][[<span class="st">&quot;bayNorm&quot;</span>]][[<span class="st">&quot;Bay_out&quot;</span>]],</span>
<span id="cb14-12"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb14-12"></a>  <span class="dt">project =</span> sc68_cisp[[<span class="st">&quot;history&quot;</span>]][[<span class="st">&quot;make_asurat_obj&quot;</span>]][[<span class="st">&quot;obj_name&quot;</span>]])</span>
<span id="cb14-13"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb14-13"></a>pbmc_<span class="dv">4000</span> &lt;-<span class="st"> </span><span class="kw">CreateSeuratObject</span>(</span>
<span id="cb14-14"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb14-14"></a>  <span class="dt">counts =</span> pbmc_<span class="dv">4000</span>[[<span class="st">&quot;data&quot;</span>]][[<span class="st">&quot;bayNorm&quot;</span>]][[<span class="st">&quot;Bay_out&quot;</span>]],</span>
<span id="cb14-15"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb14-15"></a>  <span class="dt">project =</span> pbmc_<span class="dv">4000</span>[[<span class="st">&quot;history&quot;</span>]][[<span class="st">&quot;make_asurat_obj&quot;</span>]][[<span class="st">&quot;obj_name&quot;</span>]])</span>
<span id="cb14-16"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb14-16"></a>pbmc_<span class="dv">6000</span> &lt;-<span class="st"> </span><span class="kw">CreateSeuratObject</span>(</span>
<span id="cb14-17"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb14-17"></a>  <span class="dt">counts =</span> pbmc_<span class="dv">6000</span>[[<span class="st">&quot;data&quot;</span>]][[<span class="st">&quot;bayNorm&quot;</span>]][[<span class="st">&quot;Bay_out&quot;</span>]],</span>
<span id="cb14-18"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb14-18"></a>  <span class="dt">project =</span> pbmc_<span class="dv">6000</span>[[<span class="st">&quot;history&quot;</span>]][[<span class="st">&quot;make_asurat_obj&quot;</span>]][[<span class="st">&quot;obj_name&quot;</span>]])</span></code></pre></div>
<p>The following function <code>process_001_seurat()</code> (1) normalizes the data by using
<code>NormalizeData()</code> in Seurat package with the argument
<code>normalization.method = LogNormalize</code>, (2) performs variance stabilizing
transform (VST) by using <code>FindVariableFeatures()</code> setting the number of
variable feature <code>nfeatures</code>, (3) scales the data, and (4) reduce the dimension
by principal component analysis (PCA), using <code>RunPCA()</code> with the argument
<code>features = VariableFeatures(obj)</code>.</p>
<p><strong>Tips:</strong>
As mentioned in Cruz and Wishart, Cancer Inform. 2, 59-77 (2006),
<code>nfeatures</code> was set as the sample-per-variable feature ratio being 5:1.</p>
<div class="sourceCode" id="cb15"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb15-1"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb15-1"></a>day1_norm &lt;-<span class="st"> </span><span class="kw">process_001_seurat</span>(</span>
<span id="cb15-2"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb15-2"></a>  <span class="dt">obj =</span> day1_norm,</span>
<span id="cb15-3"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb15-3"></a>  <span class="dt">nfeatures =</span> <span class="kw">round</span>(<span class="fl">0.2</span> <span class="op">*</span><span class="st"> </span>day1_norm<span class="op">@</span>assays[[<span class="st">&quot;RNA&quot;</span>]]<span class="op">@</span>counts<span class="op">@</span>Dim[<span class="dv">1</span>]))</span>
<span id="cb15-4"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb15-4"></a>day7_hypo &lt;-<span class="st"> </span><span class="kw">process_001_seurat</span>(</span>
<span id="cb15-5"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb15-5"></a>  <span class="dt">obj =</span> day7_hypo,</span>
<span id="cb15-6"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb15-6"></a>  <span class="dt">nfeatures =</span> <span class="kw">round</span>(<span class="fl">0.2</span> <span class="op">*</span><span class="st"> </span>day7_hypo<span class="op">@</span>assays[[<span class="st">&quot;RNA&quot;</span>]]<span class="op">@</span>counts<span class="op">@</span>Dim[<span class="dv">1</span>]))</span>
<span id="cb15-7"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb15-7"></a>sc68_vehi &lt;-<span class="st"> </span><span class="kw">process_001_seurat</span>(</span>
<span id="cb15-8"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb15-8"></a>  <span class="dt">obj =</span> sc68_vehi,</span>
<span id="cb15-9"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb15-9"></a>  <span class="dt">nfeatures =</span> <span class="kw">round</span>(<span class="fl">0.2</span> <span class="op">*</span><span class="st"> </span>sc68_vehi<span class="op">@</span>assays[[<span class="st">&quot;RNA&quot;</span>]]<span class="op">@</span>counts<span class="op">@</span>Dim[<span class="dv">1</span>]))</span>
<span id="cb15-10"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb15-10"></a>sc68_cisp &lt;-<span class="st"> </span><span class="kw">process_001_seurat</span>(</span>
<span id="cb15-11"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb15-11"></a>  <span class="dt">obj =</span> sc68_cisp,</span>
<span id="cb15-12"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb15-12"></a>  <span class="dt">nfeatures =</span> <span class="kw">round</span>(<span class="fl">0.2</span> <span class="op">*</span><span class="st"> </span>sc68_cisp<span class="op">@</span>assays[[<span class="st">&quot;RNA&quot;</span>]]<span class="op">@</span>counts<span class="op">@</span>Dim[<span class="dv">1</span>]))</span>
<span id="cb15-13"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb15-13"></a>pbmc_<span class="dv">4000</span> &lt;-<span class="st"> </span><span class="kw">process_001_seurat</span>(</span>
<span id="cb15-14"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb15-14"></a>  <span class="dt">obj =</span> pbmc_<span class="dv">4000</span>,</span>
<span id="cb15-15"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb15-15"></a>  <span class="dt">nfeatures =</span> <span class="kw">round</span>(<span class="fl">0.2</span> <span class="op">*</span><span class="st"> </span>pbmc_<span class="dv">4000</span><span class="op">@</span>assays[[<span class="st">&quot;RNA&quot;</span>]]<span class="op">@</span>counts<span class="op">@</span>Dim[<span class="dv">1</span>]))</span>
<span id="cb15-16"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb15-16"></a>pbmc_<span class="dv">6000</span> &lt;-<span class="st"> </span><span class="kw">process_001_seurat</span>(</span>
<span id="cb15-17"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb15-17"></a>  <span class="dt">obj =</span> pbmc_<span class="dv">6000</span>,</span>
<span id="cb15-18"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb15-18"></a>  <span class="dt">nfeatures =</span> <span class="kw">round</span>(<span class="fl">0.2</span> <span class="op">*</span><span class="st"> </span>pbmc_<span class="dv">6000</span><span class="op">@</span>assays[[<span class="st">&quot;RNA&quot;</span>]]<span class="op">@</span>counts<span class="op">@</span>Dim[<span class="dv">1</span>]))</span></code></pre></div>
<p>The following command can calculate the cumulative sum over whole sum.
These results are used to determine the dimensions of the principal components.
In what follows, we use the PCs whose contribution ratio is greater than
or equal to 0.9 for the first time.</p>
<div class="sourceCode" id="cb16"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb16-1"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb16-1"></a>day1_norm_pc &lt;-<span class="st"> </span><span class="kw">which</span>(<span class="kw">cumsum</span>(day1_norm<span class="op">@</span>reductions[[<span class="st">&quot;pca&quot;</span>]]<span class="op">@</span>stdev) <span class="op">/</span></span>
<span id="cb16-2"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb16-2"></a><span class="st">                        </span><span class="kw">sum</span>(day1_norm<span class="op">@</span>reductions[[<span class="st">&quot;pca&quot;</span>]]<span class="op">@</span>stdev) <span class="op">&gt;</span><span class="st"> </span><span class="fl">0.9</span>)[<span class="dv">1</span>]</span>
<span id="cb16-3"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb16-3"></a>day7_hypo_pc &lt;-<span class="st"> </span><span class="kw">which</span>(<span class="kw">cumsum</span>(day7_hypo<span class="op">@</span>reductions[[<span class="st">&quot;pca&quot;</span>]]<span class="op">@</span>stdev) <span class="op">/</span></span>
<span id="cb16-4"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb16-4"></a><span class="st">                        </span><span class="kw">sum</span>(day7_hypo<span class="op">@</span>reductions[[<span class="st">&quot;pca&quot;</span>]]<span class="op">@</span>stdev) <span class="op">&gt;</span><span class="st"> </span><span class="fl">0.9</span>)[<span class="dv">1</span>]</span>
<span id="cb16-5"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb16-5"></a>sc68_vehi_pc &lt;-<span class="st"> </span><span class="kw">which</span>(<span class="kw">cumsum</span>(sc68_vehi<span class="op">@</span>reductions[[<span class="st">&quot;pca&quot;</span>]]<span class="op">@</span>stdev) <span class="op">/</span></span>
<span id="cb16-6"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb16-6"></a><span class="st">                        </span><span class="kw">sum</span>(sc68_vehi<span class="op">@</span>reductions[[<span class="st">&quot;pca&quot;</span>]]<span class="op">@</span>stdev) <span class="op">&gt;</span><span class="st"> </span><span class="fl">0.9</span>)[<span class="dv">1</span>]</span>
<span id="cb16-7"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb16-7"></a>sc68_cisp_pc &lt;-<span class="st"> </span><span class="kw">which</span>(<span class="kw">cumsum</span>(sc68_cisp<span class="op">@</span>reductions[[<span class="st">&quot;pca&quot;</span>]]<span class="op">@</span>stdev) <span class="op">/</span></span>
<span id="cb16-8"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb16-8"></a><span class="st">                        </span><span class="kw">sum</span>(sc68_cisp<span class="op">@</span>reductions[[<span class="st">&quot;pca&quot;</span>]]<span class="op">@</span>stdev) <span class="op">&gt;</span><span class="st"> </span><span class="fl">0.9</span>)[<span class="dv">1</span>]</span>
<span id="cb16-9"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb16-9"></a>pbmc_<span class="dv">4000</span>_pc &lt;-<span class="st"> </span><span class="kw">which</span>(<span class="kw">cumsum</span>(pbmc_<span class="dv">4000</span><span class="op">@</span>reductions[[<span class="st">&quot;pca&quot;</span>]]<span class="op">@</span>stdev) <span class="op">/</span></span>
<span id="cb16-10"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb16-10"></a><span class="st">                        </span><span class="kw">sum</span>(pbmc_<span class="dv">4000</span><span class="op">@</span>reductions[[<span class="st">&quot;pca&quot;</span>]]<span class="op">@</span>stdev) <span class="op">&gt;</span><span class="st"> </span><span class="fl">0.9</span>)[<span class="dv">1</span>]</span>
<span id="cb16-11"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb16-11"></a>pbmc_<span class="dv">6000</span>_pc &lt;-<span class="st"> </span><span class="kw">which</span>(<span class="kw">cumsum</span>(pbmc_<span class="dv">6000</span><span class="op">@</span>reductions[[<span class="st">&quot;pca&quot;</span>]]<span class="op">@</span>stdev) <span class="op">/</span></span>
<span id="cb16-12"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb16-12"></a><span class="st">                        </span><span class="kw">sum</span>(pbmc_<span class="dv">6000</span><span class="op">@</span>reductions[[<span class="st">&quot;pca&quot;</span>]]<span class="op">@</span>stdev) <span class="op">&gt;</span><span class="st"> </span><span class="fl">0.9</span>)[<span class="dv">1</span>]</span></code></pre></div>
<p>The following functions computes t-SNE and UMAP.</p>
<div class="sourceCode" id="cb17"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb17-1"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb17-1"></a>day1_norm &lt;-<span class="st"> </span><span class="kw">FindNeighbors</span>(day1_norm, <span class="dt">reduction =</span> <span class="st">&quot;pca&quot;</span>, <span class="dt">dim =</span> <span class="dv">1</span><span class="op">:</span>day1_norm_pc)</span>
<span id="cb17-2"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb17-2"></a>day7_hypo &lt;-<span class="st"> </span><span class="kw">FindNeighbors</span>(day7_hypo, <span class="dt">reduction =</span> <span class="st">&quot;pca&quot;</span>, <span class="dt">dim =</span> <span class="dv">1</span><span class="op">:</span>day7_hypo_pc)</span>
<span id="cb17-3"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb17-3"></a>sc68_vehi &lt;-<span class="st"> </span><span class="kw">FindNeighbors</span>(sc68_vehi, <span class="dt">reduction =</span> <span class="st">&quot;pca&quot;</span>, <span class="dt">dim =</span> <span class="dv">1</span><span class="op">:</span>sc68_vehi_pc)</span>
<span id="cb17-4"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb17-4"></a>sc68_cisp &lt;-<span class="st"> </span><span class="kw">FindNeighbors</span>(sc68_cisp, <span class="dt">reduction =</span> <span class="st">&quot;pca&quot;</span>, <span class="dt">dim =</span> <span class="dv">1</span><span class="op">:</span>sc68_cisp_pc)</span>
<span id="cb17-5"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb17-5"></a>pbmc_<span class="dv">4000</span> &lt;-<span class="st"> </span><span class="kw">FindNeighbors</span>(pbmc_<span class="dv">4000</span>, <span class="dt">reduction =</span> <span class="st">&quot;pca&quot;</span>, <span class="dt">dim =</span> <span class="dv">1</span><span class="op">:</span>pbmc_<span class="dv">4000</span>_pc)</span>
<span id="cb17-6"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb17-6"></a>pbmc_<span class="dv">6000</span> &lt;-<span class="st"> </span><span class="kw">FindNeighbors</span>(pbmc_<span class="dv">6000</span>, <span class="dt">reduction =</span> <span class="st">&quot;pca&quot;</span>, <span class="dt">dim =</span> <span class="dv">1</span><span class="op">:</span>pbmc_<span class="dv">6000</span>_pc)</span>
<span id="cb17-7"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb17-7"></a><span class="co"># ----------------------------------------</span></span>
<span id="cb17-8"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb17-8"></a><span class="co"># t-SNE</span></span>
<span id="cb17-9"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb17-9"></a><span class="co"># ----------------------------------------</span></span>
<span id="cb17-10"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb17-10"></a>day1_norm &lt;-<span class="st"> </span><span class="kw">RunTSNE</span>(day1_norm, <span class="dt">dims.use =</span> <span class="dv">1</span><span class="op">:</span><span class="dv">2</span>, <span class="dt">reduction =</span> <span class="st">&quot;pca&quot;</span>,</span>
<span id="cb17-11"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb17-11"></a>                     <span class="dt">dims =</span> <span class="dv">1</span><span class="op">:</span>day1_norm_pc, <span class="dt">do.fast =</span> <span class="ot">FALSE</span>, <span class="dt">perplexity =</span> <span class="dv">30</span>)</span>
<span id="cb17-12"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb17-12"></a>day7_hypo &lt;-<span class="st"> </span><span class="kw">RunTSNE</span>(day7_hypo, <span class="dt">dims.use =</span> <span class="dv">1</span><span class="op">:</span><span class="dv">2</span>, <span class="dt">reduction =</span> <span class="st">&quot;pca&quot;</span>,</span>
<span id="cb17-13"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb17-13"></a>                     <span class="dt">dims =</span> <span class="dv">1</span><span class="op">:</span>day7_hypo_pc, <span class="dt">do.fast =</span> <span class="ot">FALSE</span>, <span class="dt">perplexity =</span> <span class="dv">30</span>)</span>
<span id="cb17-14"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb17-14"></a>sc68_vehi &lt;-<span class="st"> </span><span class="kw">RunTSNE</span>(sc68_vehi, <span class="dt">dims.use =</span> <span class="dv">1</span><span class="op">:</span><span class="dv">2</span>, <span class="dt">reduction =</span> <span class="st">&quot;pca&quot;</span>,</span>
<span id="cb17-15"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb17-15"></a>                     <span class="dt">dims =</span> <span class="dv">1</span><span class="op">:</span>sc68_vehi_pc, <span class="dt">do.fast =</span> <span class="ot">FALSE</span>, <span class="dt">perplexity =</span> <span class="dv">30</span>)</span>
<span id="cb17-16"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb17-16"></a>sc68_cisp &lt;-<span class="st"> </span><span class="kw">RunTSNE</span>(sc68_cisp, <span class="dt">dims.use =</span> <span class="dv">1</span><span class="op">:</span><span class="dv">2</span>, <span class="dt">reduction =</span> <span class="st">&quot;pca&quot;</span>,</span>
<span id="cb17-17"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb17-17"></a>                     <span class="dt">dims =</span> <span class="dv">1</span><span class="op">:</span>sc68_cisp_pc, <span class="dt">do.fast =</span> <span class="ot">FALSE</span>, <span class="dt">perplexity =</span> <span class="dv">30</span>)</span>
<span id="cb17-18"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb17-18"></a>pbmc_<span class="dv">4000</span> &lt;-<span class="st"> </span><span class="kw">RunTSNE</span>(pbmc_<span class="dv">4000</span>, <span class="dt">dims.use =</span> <span class="dv">1</span><span class="op">:</span><span class="dv">2</span>, <span class="dt">reduction =</span> <span class="st">&quot;pca&quot;</span>,</span>
<span id="cb17-19"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb17-19"></a>                     <span class="dt">dims =</span> <span class="dv">1</span><span class="op">:</span>pbmc_<span class="dv">4000</span>_pc, <span class="dt">do.fast =</span> <span class="ot">FALSE</span>, <span class="dt">perplexity =</span> <span class="dv">30</span>)</span>
<span id="cb17-20"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb17-20"></a>pbmc_<span class="dv">6000</span> &lt;-<span class="st"> </span><span class="kw">RunTSNE</span>(pbmc_<span class="dv">6000</span>, <span class="dt">dims.use =</span> <span class="dv">1</span><span class="op">:</span><span class="dv">2</span>, <span class="dt">reduction =</span> <span class="st">&quot;pca&quot;</span>,</span>
<span id="cb17-21"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb17-21"></a>                     <span class="dt">dims =</span> <span class="dv">1</span><span class="op">:</span>pbmc_<span class="dv">6000</span>_pc, <span class="dt">do.fast =</span> <span class="ot">FALSE</span>, <span class="dt">perplexity =</span> <span class="dv">30</span>)</span>
<span id="cb17-22"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb17-22"></a><span class="co"># ----------------------------------------</span></span>
<span id="cb17-23"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb17-23"></a><span class="co"># UMAP</span></span>
<span id="cb17-24"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb17-24"></a><span class="co"># ----------------------------------------</span></span>
<span id="cb17-25"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb17-25"></a>day1_norm &lt;-<span class="st"> </span><span class="kw">RunUMAP</span>(day1_norm, <span class="dt">dims =</span> <span class="dv">1</span><span class="op">:</span>day1_norm_pc)</span>
<span id="cb17-26"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb17-26"></a>day7_hypo &lt;-<span class="st"> </span><span class="kw">RunUMAP</span>(day7_hypo, <span class="dt">dims =</span> <span class="dv">1</span><span class="op">:</span>day7_hypo_pc)</span>
<span id="cb17-27"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb17-27"></a>sc68_vehi &lt;-<span class="st"> </span><span class="kw">RunUMAP</span>(sc68_vehi, <span class="dt">dims =</span> <span class="dv">1</span><span class="op">:</span>sc68_vehi_pc)</span>
<span id="cb17-28"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb17-28"></a>sc68_cisp &lt;-<span class="st"> </span><span class="kw">RunUMAP</span>(sc68_cisp, <span class="dt">dims =</span> <span class="dv">1</span><span class="op">:</span>sc68_cisp_pc)</span>
<span id="cb17-29"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb17-29"></a>pbmc_<span class="dv">4000</span> &lt;-<span class="st"> </span><span class="kw">RunUMAP</span>(pbmc_<span class="dv">4000</span>, <span class="dt">dims =</span> <span class="dv">1</span><span class="op">:</span>pbmc_<span class="dv">4000</span>_pc)</span>
<span id="cb17-30"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb17-30"></a>pbmc_<span class="dv">6000</span> &lt;-<span class="st"> </span><span class="kw">RunUMAP</span>(pbmc_<span class="dv">6000</span>, <span class="dt">dims =</span> <span class="dv">1</span><span class="op">:</span>pbmc_<span class="dv">6000</span>_pc)</span></code></pre></div>
<p>Save the objects.</p>
<div class="sourceCode" id="cb18"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb18-1"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb18-1"></a><span class="kw">saveRDS</span>(day1_norm, <span class="dt">file =</span> <span class="st">&quot;backup/11_020_day1_norm_seurat.rds&quot;</span>)</span>
<span id="cb18-2"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb18-2"></a><span class="kw">saveRDS</span>(day7_hypo, <span class="dt">file =</span> <span class="st">&quot;backup/12_020_day7_hypo_seurat.rds&quot;</span>)</span>
<span id="cb18-3"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb18-3"></a><span class="kw">saveRDS</span>(sc68_vehi, <span class="dt">file =</span> <span class="st">&quot;backup/13_020_sc68_vehi_seurat.rds&quot;</span>)</span>
<span id="cb18-4"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb18-4"></a><span class="kw">saveRDS</span>(sc68_cisp, <span class="dt">file =</span> <span class="st">&quot;backup/14_020_sc68_cisp_seurat.rds&quot;</span>)</span>
<span id="cb18-5"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb18-5"></a><span class="kw">saveRDS</span>(pbmc_<span class="dv">4000</span>, <span class="dt">file =</span> <span class="st">&quot;backup/15_020_pbmc_4000_seurat.rds&quot;</span>)</span>
<span id="cb18-6"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb18-6"></a><span class="kw">saveRDS</span>(pbmc_<span class="dv">6000</span>, <span class="dt">file =</span> <span class="st">&quot;backup/16_020_pbmc_6000_seurat.rds&quot;</span>)</span></code></pre></div>
<p>Load the objects.</p>
<div class="sourceCode" id="cb19"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb19-1"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb19-1"></a>day1_norm &lt;-<span class="st"> </span><span class="kw">readRDS</span>(<span class="dt">file =</span> <span class="st">&quot;backup/11_020_day1_norm_seurat.rds&quot;</span>)</span>
<span id="cb19-2"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb19-2"></a>day7_hypo &lt;-<span class="st"> </span><span class="kw">readRDS</span>(<span class="dt">file =</span> <span class="st">&quot;backup/12_020_day7_hypo_seurat.rds&quot;</span>)</span>
<span id="cb19-3"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb19-3"></a>sc68_vehi &lt;-<span class="st"> </span><span class="kw">readRDS</span>(<span class="dt">file =</span> <span class="st">&quot;backup/13_020_sc68_vehi_seurat.rds&quot;</span>)</span>
<span id="cb19-4"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb19-4"></a>sc68_cisp &lt;-<span class="st"> </span><span class="kw">readRDS</span>(<span class="dt">file =</span> <span class="st">&quot;backup/14_020_sc68_cisp_seurat.rds&quot;</span>)</span>
<span id="cb19-5"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb19-5"></a>pbmc_<span class="dv">4000</span> &lt;-<span class="st"> </span><span class="kw">readRDS</span>(<span class="dt">file =</span> <span class="st">&quot;backup/15_020_pbmc_4000_seurat.rds&quot;</span>)</span>
<span id="cb19-6"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb19-6"></a>pbmc_<span class="dv">6000</span> &lt;-<span class="st"> </span><span class="kw">readRDS</span>(<span class="dt">file =</span> <span class="st">&quot;backup/16_020_pbmc_6000_seurat.rds&quot;</span>)</span></code></pre></div>
<p>The following function <code>plot_manifold()</code> shows sample states on t-SNE or UMAP
space.</p>
<div class="sourceCode" id="cb20"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb20-1"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb20-1"></a>plot_manifold &lt;-<span class="st"> </span><span class="cf">function</span>(obj, plot_type, title, title_size, xlabel, ylabel){</span>
<span id="cb20-2"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb20-2"></a>  <span class="cf">if</span>(plot_type <span class="op">==</span><span class="st"> &quot;umap&quot;</span>){</span>
<span id="cb20-3"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb20-3"></a>    df &lt;-<span class="st"> </span>obj<span class="op">@</span>reductions[[<span class="st">&quot;umap&quot;</span>]]<span class="op">@</span>cell.embeddings</span>
<span id="cb20-4"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb20-4"></a>  }<span class="cf">else</span> <span class="cf">if</span>(plot_type <span class="op">==</span><span class="st"> &quot;tsne&quot;</span>){</span>
<span id="cb20-5"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb20-5"></a>    df &lt;-<span class="st"> </span>obj<span class="op">@</span>reductions[[<span class="st">&quot;tsne&quot;</span>]]<span class="op">@</span>cell.embeddings</span>
<span id="cb20-6"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb20-6"></a>  }</span>
<span id="cb20-7"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb20-7"></a>  p &lt;-<span class="st"> </span><span class="kw">ggplot</span>() <span class="op">+</span></span>
<span id="cb20-8"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb20-8"></a><span class="st">    </span><span class="kw">geom_point</span>(<span class="kw">aes</span>(<span class="dt">x=</span>df[,<span class="dv">1</span>], <span class="dt">y=</span>df[,<span class="dv">2</span>]), <span class="dt">color=</span><span class="st">&quot;black&quot;</span>, <span class="dt">size=</span><span class="fl">0.5</span>, <span class="dt">alpha=</span><span class="fl">1.0</span>) <span class="op">+</span></span>
<span id="cb20-9"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb20-9"></a><span class="st">    </span><span class="kw">labs</span>(<span class="dt">title=</span>title, <span class="dt">x=</span>xlabel, <span class="dt">y=</span>ylabel) <span class="op">+</span></span>
<span id="cb20-10"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb20-10"></a><span class="st">    </span><span class="kw">theme_classic</span>(<span class="dt">base_size=</span><span class="dv">20</span>, <span class="dt">base_family=</span><span class="st">&quot;Helvetica&quot;</span>) <span class="op">+</span></span>
<span id="cb20-11"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb20-11"></a><span class="st">    </span><span class="kw">theme</span>(<span class="dt">plot.title=</span><span class="kw">element_text</span>(<span class="dt">hjust=</span><span class="fl">0.5</span>, <span class="dt">size=</span>title_size),</span>
<span id="cb20-12"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb20-12"></a>      <span class="dt">plot.margin=</span><span class="kw">unit</span>(<span class="kw">c</span>(<span class="fl">1.5</span>,<span class="fl">1.5</span>,<span class="fl">1.5</span>,<span class="fl">1.5</span>), <span class="st">&quot;lines&quot;</span>), <span class="dt">legend.position=</span><span class="st">&quot;right&quot;</span>)</span>
<span id="cb20-13"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb20-13"></a>  <span class="kw">return</span>(p)</span>
<span id="cb20-14"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb20-14"></a>}</span>
<span id="cb20-15"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb20-15"></a></span>
<span id="cb20-16"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb20-16"></a>data &lt;-<span class="st"> </span><span class="kw">list</span>(day1_norm, day7_hypo, sc68_vehi, sc68_cisp, pbmc_<span class="dv">4000</span>, pbmc_<span class="dv">6000</span>)</span>
<span id="cb20-17"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb20-17"></a>names &lt;-<span class="st"> </span><span class="kw">c</span>(<span class="st">&quot;day1_norm&quot;</span>, <span class="st">&quot;day7_hypo&quot;</span>, <span class="st">&quot;sc68_vehi&quot;</span>, <span class="st">&quot;sc68_cisp&quot;</span>,</span>
<span id="cb20-18"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb20-18"></a>           <span class="st">&quot;pbmc_4000&quot;</span>, <span class="st">&quot;pbmc_6000&quot;</span>)</span>
<span id="cb20-19"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb20-19"></a><span class="cf">for</span>(i <span class="cf">in</span> <span class="dv">1</span><span class="op">:</span><span class="kw">length</span>(data)){</span>
<span id="cb20-20"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb20-20"></a>  p &lt;-<span class="st"> </span><span class="kw">plot_manifold</span>(<span class="dt">obj =</span> data[[i]], <span class="dt">plot_type =</span> <span class="st">&quot;umap&quot;</span>,</span>
<span id="cb20-21"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb20-21"></a>                     <span class="dt">title =</span> <span class="kw">paste</span>(names[i], <span class="st">&quot; (Seurat: UMAP)&quot;</span>, <span class="dt">sep =</span> <span class="st">&quot;&quot;</span>),</span>
<span id="cb20-22"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb20-22"></a>                     <span class="dt">title_size =</span> <span class="dv">16</span>, <span class="dt">xlabel =</span> <span class="st">&quot;UMAP_1&quot;</span>, <span class="dt">ylabel =</span> <span class="st">&quot;UMAP_2&quot;</span>)</span>
<span id="cb20-23"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb20-23"></a>  filename &lt;-<span class="st"> </span><span class="kw">paste</span>(<span class="st">&quot;figures/figure_&quot;</span>, <span class="kw">sprintf</span>(<span class="st">&quot;%02d&quot;</span>, <span class="dv">10</span> <span class="op">+</span><span class="st"> </span>i), <span class="st">&quot;_0020.png&quot;</span>,</span>
<span id="cb20-24"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb20-24"></a>                    <span class="dt">sep =</span> <span class="st">&quot;&quot;</span>)</span>
<span id="cb20-25"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb20-25"></a>  <span class="kw">ggsave</span>(<span class="dt">file =</span> filename, <span class="dt">plot =</span> p, <span class="dt">dpi =</span> <span class="dv">300</span>, <span class="dt">width =</span> <span class="dv">4</span>, <span class="dt">height =</span> <span class="fl">4.2</span>)</span>
<span id="cb20-26"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb20-26"></a>}</span>
<span id="cb20-27"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb20-27"></a><span class="cf">for</span>(i <span class="cf">in</span> <span class="dv">1</span><span class="op">:</span><span class="kw">length</span>(data)){</span>
<span id="cb20-28"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb20-28"></a>  p &lt;-<span class="st"> </span><span class="kw">plot_manifold</span>(<span class="dt">obj =</span> data[[i]], <span class="dt">plot_type =</span> <span class="st">&quot;tsne&quot;</span>,</span>
<span id="cb20-29"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb20-29"></a>                     <span class="dt">title =</span> <span class="kw">paste</span>(names[i], <span class="st">&quot; (Seurat: tSNE)&quot;</span>, <span class="dt">sep =</span> <span class="st">&quot;&quot;</span>),</span>
<span id="cb20-30"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb20-30"></a>                     <span class="dt">title_size =</span> <span class="dv">16</span>, <span class="dt">xlabel =</span> <span class="st">&quot;tSNE_1&quot;</span>, <span class="dt">ylabel =</span> <span class="st">&quot;tSNE_2&quot;</span>)</span>
<span id="cb20-31"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb20-31"></a>  filename &lt;-<span class="st"> </span><span class="kw">paste</span>(<span class="st">&quot;figures/figure_&quot;</span>, <span class="kw">sprintf</span>(<span class="st">&quot;%02d&quot;</span>, <span class="dv">10</span> <span class="op">+</span><span class="st"> </span>i), <span class="st">&quot;_0021.png&quot;</span>,</span>
<span id="cb20-32"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb20-32"></a>                    <span class="dt">sep =</span> <span class="st">&quot;&quot;</span>)</span>
<span id="cb20-33"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb20-33"></a>  <span class="kw">ggsave</span>(<span class="dt">file =</span> filename, <span class="dt">plot =</span> p, <span class="dt">dpi =</span> <span class="dv">300</span>, <span class="dt">width =</span> <span class="dv">4</span>, <span class="dt">height =</span> <span class="fl">4.2</span>)</span>
<span id="cb20-34"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb20-34"></a>}</span></code></pre></div>
<p><img src="figures/figure_11_0020.png" width="150px">
<img src="figures/figure_12_0020.png" width="150px">
<img src="figures/figure_13_0020.png" width="150px">
<img src="figures/figure_14_0020.png" width="150px">
<img src="figures/figure_15_0020.png" width="150px">
<img src="figures/figure_16_0020.png" width="150px"></p>
<p><img src="figures/figure_11_0021.png" width="150px">
<img src="figures/figure_12_0021.png" width="150px">
<img src="figures/figure_13_0021.png" width="150px">
<img src="figures/figure_14_0021.png" width="150px">
<img src="figures/figure_15_0021.png" width="150px">
<img src="figures/figure_16_0021.png" width="150px"></p>
<p>Perform k-means clusterings and compute several clustering validity scores by
using <code>NbClust()</code>.</p>
<div class="sourceCode" id="cb21"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb21-1"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb21-1"></a>compute_nbclust &lt;-<span class="st"> </span><span class="cf">function</span>(obj, plot_type){</span>
<span id="cb21-2"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb21-2"></a>  <span class="cf">if</span>(plot_type <span class="op">==</span><span class="st"> &quot;umap&quot;</span>){</span>
<span id="cb21-3"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb21-3"></a>    mat &lt;-<span class="st"> </span>obj<span class="op">@</span>reductions[[<span class="st">&quot;umap&quot;</span>]]<span class="op">@</span>cell.embeddings</span>
<span id="cb21-4"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb21-4"></a>  }<span class="cf">else</span> <span class="cf">if</span>(plot_type <span class="op">==</span><span class="st"> &quot;tsne&quot;</span>){</span>
<span id="cb21-5"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb21-5"></a>    mat &lt;-<span class="st"> </span>obj<span class="op">@</span>reductions[[<span class="st">&quot;tsne&quot;</span>]]<span class="op">@</span>cell.embeddings</span>
<span id="cb21-6"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb21-6"></a>  }</span>
<span id="cb21-7"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb21-7"></a>  <span class="kw">set.seed</span>(<span class="dv">8</span>)</span>
<span id="cb21-8"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb21-8"></a>  res &lt;-<span class="st"> </span><span class="kw">NbClust</span>(mat, <span class="dt">distance =</span> <span class="st">&quot;euclidean&quot;</span>, <span class="dt">min.nc =</span> <span class="dv">2</span>, <span class="dt">max.nc =</span> <span class="dv">7</span>,</span>
<span id="cb21-9"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb21-9"></a>                 <span class="dt">method =</span> <span class="st">&quot;kmeans&quot;</span>, <span class="dt">index =</span> <span class="st">&quot;all&quot;</span>)</span>
<span id="cb21-10"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb21-10"></a>  <span class="kw">return</span>(res)</span>
<span id="cb21-11"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb21-11"></a>}</span>
<span id="cb21-12"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb21-12"></a></span>
<span id="cb21-13"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb21-13"></a>data &lt;-<span class="st"> </span><span class="kw">list</span>(day1_norm, day7_hypo, sc68_vehi, sc68_cisp, pbmc_<span class="dv">4000</span>, pbmc_<span class="dv">6000</span>)</span>
<span id="cb21-14"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb21-14"></a>names &lt;-<span class="st"> </span><span class="kw">c</span>(<span class="st">&quot;day1_norm&quot;</span>, <span class="st">&quot;day7_hypo&quot;</span>, <span class="st">&quot;sc68_vehi&quot;</span>, <span class="st">&quot;sc68_cisp&quot;</span>, <span class="st">&quot;pbmc_4000&quot;</span>,</span>
<span id="cb21-15"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb21-15"></a>           <span class="st">&quot;pbmc_6000&quot;</span>)</span>
<span id="cb21-16"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb21-16"></a>nbclust_umap &lt;-<span class="st"> </span><span class="kw">list</span>()</span>
<span id="cb21-17"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb21-17"></a>nbclust_tsne &lt;-<span class="st"> </span><span class="kw">list</span>()</span>
<span id="cb21-18"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb21-18"></a><span class="cf">for</span>(i <span class="cf">in</span> <span class="dv">1</span><span class="op">:</span><span class="kw">length</span>(data)){</span>
<span id="cb21-19"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb21-19"></a>  nbclust_umap[[i]] &lt;-<span class="st"> </span><span class="kw">compute_nbclust</span>(<span class="dt">obj =</span> data[[i]], <span class="dt">plot_type =</span> <span class="st">&quot;umap&quot;</span>)</span>
<span id="cb21-20"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb21-20"></a>}</span>
<span id="cb21-21"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb21-21"></a><span class="cf">for</span>(i <span class="cf">in</span> <span class="dv">1</span><span class="op">:</span><span class="kw">length</span>(data)){</span>
<span id="cb21-22"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb21-22"></a>  nbclust_tsne[[i]] &lt;-<span class="st"> </span><span class="kw">compute_nbclust</span>(<span class="dt">obj =</span> data[[i]], <span class="dt">plot_type =</span> <span class="st">&quot;tsne&quot;</span>)</span>
<span id="cb21-23"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb21-23"></a>}</span>
<span id="cb21-24"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb21-24"></a><span class="kw">names</span>(nbclust_umap) &lt;-<span class="st"> </span>names</span>
<span id="cb21-25"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb21-25"></a><span class="kw">names</span>(nbclust_tsne) &lt;-<span class="st"> </span>names</span></code></pre></div>
<p>Save the objects.</p>
<div class="sourceCode" id="cb22"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb22-1"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb22-1"></a><span class="kw">saveRDS</span>(nbclust_umap, <span class="dt">file =</span> <span class="st">&quot;backup/20_001_nbclust_umap_seurat.rds&quot;</span>)</span>
<span id="cb22-2"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb22-2"></a><span class="kw">saveRDS</span>(nbclust_tsne, <span class="dt">file =</span> <span class="st">&quot;backup/20_002_nbclust_tsne_seurat.rds&quot;</span>)</span>
<span id="cb22-3"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb22-3"></a><span class="co"># ----------------------------------------</span></span>
<span id="cb22-4"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb22-4"></a><span class="co"># Output an excel file</span></span>
<span id="cb22-5"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb22-5"></a><span class="co"># ----------------------------------------</span></span>
<span id="cb22-6"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb22-6"></a>dnames &lt;-<span class="st"> </span><span class="kw">c</span>(<span class="st">&quot;day1_norm&quot;</span>, <span class="st">&quot;day7_hypo&quot;</span>, <span class="st">&quot;sc68_vehi&quot;</span>, <span class="st">&quot;sc68_cisp&quot;</span>,</span>
<span id="cb22-7"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb22-7"></a>            <span class="st">&quot;pbmc_4000&quot;</span>, <span class="st">&quot;pbmc_6000&quot;</span>)</span>
<span id="cb22-8"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb22-8"></a>dlist &lt;-<span class="st"> </span><span class="kw">list</span>()</span>
<span id="cb22-9"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb22-9"></a><span class="cf">for</span>(i <span class="cf">in</span> <span class="dv">1</span><span class="op">:</span><span class="kw">length</span>(dnames)){</span>
<span id="cb22-10"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb22-10"></a>  res &lt;-<span class="st"> </span><span class="kw">rbind</span>(nbclust_umap[[dnames[i]]][[<span class="st">&quot;All.index&quot;</span>]], <span class="st">&quot;&quot;</span>)</span>
<span id="cb22-11"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb22-11"></a>  res &lt;-<span class="st"> </span><span class="kw">rbind</span>(res, nbclust_umap[[dnames[i]]][[<span class="st">&quot;Best.nc&quot;</span>]])</span>
<span id="cb22-12"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb22-12"></a>  dlist &lt;-<span class="st"> </span><span class="kw">c</span>(dlist, <span class="kw">list</span>(res))</span>
<span id="cb22-13"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb22-13"></a>}</span>
<span id="cb22-14"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb22-14"></a><span class="kw">names</span>(dlist) &lt;-<span class="st"> </span>dnames</span>
<span id="cb22-15"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb22-15"></a>filename &lt;-<span class="st"> &quot;backup/nbclust_umap_seurat.xlsx&quot;</span></span>
<span id="cb22-16"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb22-16"></a><span class="kw">write.xlsx</span>(dlist, <span class="dt">file =</span> filename, <span class="dt">row.names =</span> <span class="ot">TRUE</span>)</span></code></pre></div>
<p>Load the objects.</p>
<div class="sourceCode" id="cb23"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb23-1"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb23-1"></a>nbclust_umap &lt;-<span class="st"> </span><span class="kw">readRDS</span>(<span class="dt">file =</span> <span class="st">&quot;backup/20_001_nbclust_umap_seurat.rds&quot;</span>)</span>
<span id="cb23-2"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb23-2"></a>nbclust_tsne &lt;-<span class="st"> </span><span class="kw">readRDS</span>(<span class="dt">file =</span> <span class="st">&quot;backup/20_002_nbclust_tsne_seurat.rds&quot;</span>)</span></code></pre></div>
</div>
<div id="monocle-3" class="section level3">
<h3><span class="header-section-number">14.3.2</span> Monocle 3</h3>
<p>Load the data.</p>
<div class="sourceCode" id="cb24"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb24-1"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb24-1"></a><span class="kw">rm</span>(<span class="dt">list =</span> <span class="kw">ls</span>())</span>
<span id="cb24-2"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb24-2"></a>day1_norm &lt;-<span class="st"> </span><span class="kw">readRDS</span>(<span class="dt">file =</span> <span class="st">&quot;backup/01_005_day1_norm_normalized.rds&quot;</span>)</span>
<span id="cb24-3"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb24-3"></a>day7_hypo &lt;-<span class="st"> </span><span class="kw">readRDS</span>(<span class="dt">file =</span> <span class="st">&quot;backup/02_005_day7_hypo_normalized.rds&quot;</span>)</span>
<span id="cb24-4"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb24-4"></a>sc68_vehi &lt;-<span class="st"> </span><span class="kw">readRDS</span>(<span class="dt">file =</span> <span class="st">&quot;backup/03_005_sc68_vehi_normalized.rds&quot;</span>)</span>
<span id="cb24-5"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb24-5"></a>sc68_cisp &lt;-<span class="st"> </span><span class="kw">readRDS</span>(<span class="dt">file =</span> <span class="st">&quot;backup/04_005_sc68_cisp_normalized.rds&quot;</span>)</span>
<span id="cb24-6"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb24-6"></a>pbmc_<span class="dv">4000</span> &lt;-<span class="st"> </span><span class="kw">readRDS</span>(<span class="dt">file =</span> <span class="st">&quot;backup/05_005_pbmc_4000_normalized.rds&quot;</span>)</span>
<span id="cb24-7"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb24-7"></a>pbmc_<span class="dv">6000</span> &lt;-<span class="st"> </span><span class="kw">readRDS</span>(<span class="dt">file =</span> <span class="st">&quot;backup/06_005_pbmc_6000_normalized.rds&quot;</span>)</span>
<span id="cb24-8"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb24-8"></a>sc68_vehi_bl &lt;-<span class="st"> </span><span class="kw">readRDS</span>(<span class="dt">file =</span> <span class="st">&quot;backup/13_001_sc68_vehi_blacklist.rds&quot;</span>)</span></code></pre></div>
<p>Remove the outlier cells from <code>sc68_vehi</code>.</p>
<div class="sourceCode" id="cb25"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb25-1"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb25-1"></a><span class="co"># ----------------------------------------</span></span>
<span id="cb25-2"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb25-2"></a><span class="co"># For `sc68_vehi`</span></span>
<span id="cb25-3"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb25-3"></a><span class="co"># ----------------------------------------</span></span>
<span id="cb25-4"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb25-4"></a>mat &lt;-<span class="st"> </span>sc68_vehi[[<span class="st">&quot;data&quot;</span>]][[<span class="st">&quot;bayNorm&quot;</span>]][[<span class="st">&quot;Bay_out&quot;</span>]]</span>
<span id="cb25-5"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb25-5"></a>inds &lt;-<span class="st"> </span><span class="kw">which</span>(<span class="kw">colnames</span>(mat) <span class="op">%in%</span><span class="st"> </span>sc68_vehi_bl)</span>
<span id="cb25-6"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb25-6"></a>sc68_vehi[[<span class="st">&quot;data&quot;</span>]][[<span class="st">&quot;bayNorm&quot;</span>]][[<span class="st">&quot;Bay_out&quot;</span>]] &lt;-<span class="st"> </span>mat[, <span class="op">-</span>inds]</span>
<span id="cb25-7"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb25-7"></a>sc68_vehi[[<span class="st">&quot;sample&quot;</span>]] &lt;-<span class="st"> </span>sc68_vehi[[<span class="st">&quot;sample&quot;</span>]][<span class="op">-</span>inds,]</span></code></pre></div>
<p>Create Monocle 3 objects (cell_data_set (CDS)).</p>
<div class="sourceCode" id="cb26"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb26-1"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb26-1"></a>create_monocle_obj &lt;-<span class="st"> </span><span class="cf">function</span>(obj){</span>
<span id="cb26-2"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb26-2"></a>  cells &lt;-<span class="st"> </span>obj[[<span class="st">&quot;sample&quot;</span>]][[<span class="st">&quot;barcode&quot;</span>]]</span>
<span id="cb26-3"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb26-3"></a>  <span class="kw">names</span>(cells) &lt;-<span class="st"> </span>obj[[<span class="st">&quot;sample&quot;</span>]][[<span class="st">&quot;barcode&quot;</span>]]</span>
<span id="cb26-4"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb26-4"></a>  genes &lt;-<span class="st"> </span>obj[[<span class="st">&quot;variable&quot;</span>]][[<span class="st">&quot;symbol&quot;</span>]]</span>
<span id="cb26-5"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb26-5"></a>  <span class="kw">names</span>(genes) &lt;-<span class="st"> </span>obj[[<span class="st">&quot;variable&quot;</span>]][[<span class="st">&quot;symbol&quot;</span>]]</span>
<span id="cb26-6"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb26-6"></a>  mat &lt;-<span class="st"> </span>obj[[<span class="st">&quot;data&quot;</span>]][[<span class="st">&quot;bayNorm&quot;</span>]][[<span class="st">&quot;Bay_out&quot;</span>]]</span>
<span id="cb26-7"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb26-7"></a>  obj &lt;-<span class="st"> </span><span class="kw">new_cell_data_set</span>(</span>
<span id="cb26-8"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb26-8"></a>    <span class="dt">expression_data =</span> <span class="kw">as.matrix</span>(mat),</span>
<span id="cb26-9"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb26-9"></a>    <span class="dt">cell_metadata =</span> <span class="kw">as.data.frame</span>(cells),</span>
<span id="cb26-10"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb26-10"></a>    <span class="dt">gene_metadata =</span> <span class="kw">data.frame</span>(<span class="dt">gene_short_name=</span>genes))</span>
<span id="cb26-11"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb26-11"></a>  <span class="kw">return</span>(obj)</span>
<span id="cb26-12"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb26-12"></a>}</span>
<span id="cb26-13"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb26-13"></a></span>
<span id="cb26-14"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb26-14"></a>day1_norm &lt;-<span class="st"> </span><span class="kw">create_monocle_obj</span>(<span class="dt">obj =</span> day1_norm)</span>
<span id="cb26-15"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb26-15"></a>day7_hypo &lt;-<span class="st"> </span><span class="kw">create_monocle_obj</span>(<span class="dt">obj =</span> day7_hypo)</span>
<span id="cb26-16"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb26-16"></a>sc68_vehi &lt;-<span class="st"> </span><span class="kw">create_monocle_obj</span>(<span class="dt">obj =</span> sc68_vehi)</span>
<span id="cb26-17"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb26-17"></a>sc68_cisp &lt;-<span class="st"> </span><span class="kw">create_monocle_obj</span>(<span class="dt">obj =</span> sc68_cisp)</span>
<span id="cb26-18"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb26-18"></a>pbmc_<span class="dv">4000</span> &lt;-<span class="st"> </span><span class="kw">create_monocle_obj</span>(<span class="dt">obj =</span> pbmc_<span class="dv">4000</span>)</span>
<span id="cb26-19"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb26-19"></a>pbmc_<span class="dv">6000</span> &lt;-<span class="st"> </span><span class="kw">create_monocle_obj</span>(<span class="dt">obj =</span> pbmc_<span class="dv">6000</span>)</span></code></pre></div>
<p>Preprocess the data under the default settings of Monocle 3 (version 0.2.3.0),
e.g., <code>num_dim = 50</code> and <code>norm_method = "log"</code>.</p>
<div class="sourceCode" id="cb27"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb27-1"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb27-1"></a>day1_norm &lt;-<span class="st"> </span><span class="kw">preprocess_cds</span>(day1_norm)</span>
<span id="cb27-2"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb27-2"></a>day7_hypo &lt;-<span class="st"> </span><span class="kw">preprocess_cds</span>(day7_hypo)</span>
<span id="cb27-3"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb27-3"></a>sc68_vehi &lt;-<span class="st"> </span><span class="kw">preprocess_cds</span>(sc68_vehi)</span>
<span id="cb27-4"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb27-4"></a>sc68_cisp &lt;-<span class="st"> </span><span class="kw">preprocess_cds</span>(sc68_cisp)</span>
<span id="cb27-5"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb27-5"></a>pbmc_<span class="dv">4000</span> &lt;-<span class="st"> </span><span class="kw">preprocess_cds</span>(pbmc_<span class="dv">4000</span>)</span>
<span id="cb27-6"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb27-6"></a>pbmc_<span class="dv">6000</span> &lt;-<span class="st"> </span><span class="kw">preprocess_cds</span>(pbmc_<span class="dv">6000</span>)</span></code></pre></div>
<p>Users can visualize the results of PCA by the following command:</p>
<div class="sourceCode" id="cb28"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb28-1"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb28-1"></a><span class="kw">plot_pc_variance_explained</span>(day1_norm)</span>
<span id="cb28-2"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb28-2"></a><span class="kw">plot_pc_variance_explained</span>(day7_hypo)</span>
<span id="cb28-3"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb28-3"></a><span class="kw">plot_pc_variance_explained</span>(sc68_vehi)</span>
<span id="cb28-4"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb28-4"></a><span class="kw">plot_pc_variance_explained</span>(sc68_cisp)</span>
<span id="cb28-5"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb28-5"></a><span class="kw">plot_pc_variance_explained</span>(pbmc_<span class="dv">4000</span>)</span>
<span id="cb28-6"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb28-6"></a><span class="kw">plot_pc_variance_explained</span>(pbmc_<span class="dv">6000</span>)</span></code></pre></div>
<p>Reduce the dimensions and visualize the results.</p>
<div class="sourceCode" id="cb29"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb29-1"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb29-1"></a>day1_norm &lt;-<span class="st"> </span><span class="kw">reduce_dimension</span>(day1_norm, <span class="dt">reduction_method =</span> <span class="st">&quot;UMAP&quot;</span>,</span>
<span id="cb29-2"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb29-2"></a>                              <span class="dt">preprocess_method =</span> <span class="st">&quot;PCA&quot;</span>)</span>
<span id="cb29-3"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb29-3"></a>day7_hypo &lt;-<span class="st"> </span><span class="kw">reduce_dimension</span>(day7_hypo, <span class="dt">reduction_method =</span> <span class="st">&quot;UMAP&quot;</span>,</span>
<span id="cb29-4"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb29-4"></a>                              <span class="dt">preprocess_method =</span> <span class="st">&quot;PCA&quot;</span>)</span>
<span id="cb29-5"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb29-5"></a>sc68_vehi &lt;-<span class="st"> </span><span class="kw">reduce_dimension</span>(sc68_vehi, <span class="dt">reduction_method =</span> <span class="st">&quot;UMAP&quot;</span>,</span>
<span id="cb29-6"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb29-6"></a>                              <span class="dt">preprocess_method =</span> <span class="st">&quot;PCA&quot;</span>)</span>
<span id="cb29-7"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb29-7"></a>sc68_cisp &lt;-<span class="st"> </span><span class="kw">reduce_dimension</span>(sc68_cisp, <span class="dt">reduction_method =</span> <span class="st">&quot;UMAP&quot;</span>,</span>
<span id="cb29-8"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb29-8"></a>                              <span class="dt">preprocess_method =</span> <span class="st">&quot;PCA&quot;</span>)</span>
<span id="cb29-9"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb29-9"></a>pbmc_<span class="dv">4000</span> &lt;-<span class="st"> </span><span class="kw">reduce_dimension</span>(pbmc_<span class="dv">4000</span>, <span class="dt">reduction_method =</span> <span class="st">&quot;UMAP&quot;</span>,</span>
<span id="cb29-10"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb29-10"></a>                              <span class="dt">preprocess_method =</span> <span class="st">&quot;PCA&quot;</span>)</span>
<span id="cb29-11"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb29-11"></a>pbmc_<span class="dv">6000</span> &lt;-<span class="st"> </span><span class="kw">reduce_dimension</span>(pbmc_<span class="dv">6000</span>, <span class="dt">reduction_method =</span> <span class="st">&quot;UMAP&quot;</span>,</span>
<span id="cb29-12"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb29-12"></a>                              <span class="dt">preprocess_method =</span> <span class="st">&quot;PCA&quot;</span>)</span>
<span id="cb29-13"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb29-13"></a>day1_norm &lt;-<span class="st"> </span><span class="kw">reduce_dimension</span>(day1_norm, <span class="dt">reduction_method =</span> <span class="st">&quot;tSNE&quot;</span>,</span>
<span id="cb29-14"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb29-14"></a>                              <span class="dt">preprocess_method =</span> <span class="st">&quot;PCA&quot;</span>)</span>
<span id="cb29-15"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb29-15"></a>day7_hypo &lt;-<span class="st"> </span><span class="kw">reduce_dimension</span>(day7_hypo, <span class="dt">reduction_method =</span> <span class="st">&quot;tSNE&quot;</span>,</span>
<span id="cb29-16"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb29-16"></a>                              <span class="dt">preprocess_method =</span> <span class="st">&quot;PCA&quot;</span>)</span>
<span id="cb29-17"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb29-17"></a>sc68_vehi &lt;-<span class="st"> </span><span class="kw">reduce_dimension</span>(sc68_vehi, <span class="dt">reduction_method =</span> <span class="st">&quot;tSNE&quot;</span>,</span>
<span id="cb29-18"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb29-18"></a>                              <span class="dt">preprocess_method =</span> <span class="st">&quot;PCA&quot;</span>)</span>
<span id="cb29-19"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb29-19"></a>sc68_cisp &lt;-<span class="st"> </span><span class="kw">reduce_dimension</span>(sc68_cisp, <span class="dt">reduction_method =</span> <span class="st">&quot;tSNE&quot;</span>,</span>
<span id="cb29-20"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb29-20"></a>                              <span class="dt">preprocess_method =</span> <span class="st">&quot;PCA&quot;</span>)</span>
<span id="cb29-21"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb29-21"></a>pbmc_<span class="dv">4000</span> &lt;-<span class="st"> </span><span class="kw">reduce_dimension</span>(pbmc_<span class="dv">4000</span>, <span class="dt">reduction_method =</span> <span class="st">&quot;tSNE&quot;</span>,</span>
<span id="cb29-22"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb29-22"></a>                              <span class="dt">preprocess_method =</span> <span class="st">&quot;PCA&quot;</span>)</span>
<span id="cb29-23"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb29-23"></a>pbmc_<span class="dv">6000</span> &lt;-<span class="st"> </span><span class="kw">reduce_dimension</span>(pbmc_<span class="dv">6000</span>, <span class="dt">reduction_method =</span> <span class="st">&quot;tSNE&quot;</span>,</span>
<span id="cb29-24"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb29-24"></a>                              <span class="dt">preprocess_method =</span> <span class="st">&quot;PCA&quot;</span>)</span></code></pre></div>
<p>The following function <code>plot_manifold()</code> shows sample states on t-SNE or UMAP
space.</p>
<div class="sourceCode" id="cb30"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb30-1"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb30-1"></a>plot_manifold &lt;-<span class="st"> </span><span class="cf">function</span>(obj, plot_type, title, title_size, xlabel, ylabel){</span>
<span id="cb30-2"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb30-2"></a>  <span class="cf">if</span>(plot_type <span class="op">==</span><span class="st"> &quot;umap&quot;</span>){</span>
<span id="cb30-3"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb30-3"></a>    df &lt;-<span class="st"> </span>obj<span class="op">@</span>int_colData<span class="op">@</span>listData[[<span class="st">&quot;reducedDims&quot;</span>]]<span class="op">@</span>listData[[<span class="st">&quot;UMAP&quot;</span>]]</span>
<span id="cb30-4"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb30-4"></a>  }<span class="cf">else</span> <span class="cf">if</span>(plot_type <span class="op">==</span><span class="st"> &quot;tsne&quot;</span>){</span>
<span id="cb30-5"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb30-5"></a>    df &lt;-<span class="st"> </span>obj<span class="op">@</span>int_colData<span class="op">@</span>listData[[<span class="st">&quot;reducedDims&quot;</span>]]<span class="op">@</span>listData[[<span class="st">&quot;tSNE&quot;</span>]]</span>
<span id="cb30-6"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb30-6"></a>  }</span>
<span id="cb30-7"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb30-7"></a>  p &lt;-<span class="st"> </span><span class="kw">ggplot</span>() <span class="op">+</span></span>
<span id="cb30-8"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb30-8"></a><span class="st">    </span><span class="kw">geom_point</span>(<span class="kw">aes</span>(<span class="dt">x=</span>df[,<span class="dv">1</span>], <span class="dt">y=</span>df[,<span class="dv">2</span>]), <span class="dt">color=</span><span class="st">&quot;black&quot;</span>, <span class="dt">size=</span><span class="fl">0.5</span>, <span class="dt">alpha=</span><span class="fl">1.0</span>) <span class="op">+</span></span>
<span id="cb30-9"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb30-9"></a><span class="st">    </span><span class="kw">labs</span>(<span class="dt">title=</span>title, <span class="dt">x=</span>xlabel, <span class="dt">y=</span>ylabel) <span class="op">+</span></span>
<span id="cb30-10"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb30-10"></a><span class="st">    </span><span class="kw">theme_classic</span>(<span class="dt">base_size=</span><span class="dv">20</span>, <span class="dt">base_family=</span><span class="st">&quot;Helvetica&quot;</span>) <span class="op">+</span></span>
<span id="cb30-11"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb30-11"></a><span class="st">    </span><span class="kw">theme</span>(<span class="dt">plot.title=</span><span class="kw">element_text</span>(<span class="dt">hjust=</span><span class="fl">0.5</span>, <span class="dt">size=</span>title_size),</span>
<span id="cb30-12"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb30-12"></a>      <span class="dt">plot.margin=</span><span class="kw">unit</span>(<span class="kw">c</span>(<span class="fl">1.5</span>,<span class="fl">1.5</span>,<span class="fl">1.5</span>,<span class="fl">1.5</span>), <span class="st">&quot;lines&quot;</span>), <span class="dt">legend.position=</span><span class="st">&quot;right&quot;</span>)</span>
<span id="cb30-13"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb30-13"></a>  <span class="kw">return</span>(p)</span>
<span id="cb30-14"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb30-14"></a>}</span>
<span id="cb30-15"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb30-15"></a></span>
<span id="cb30-16"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb30-16"></a>data &lt;-<span class="st"> </span><span class="kw">list</span>(day1_norm, day7_hypo, sc68_vehi, sc68_cisp, pbmc_<span class="dv">4000</span>, pbmc_<span class="dv">6000</span>)</span>
<span id="cb30-17"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb30-17"></a>names &lt;-<span class="st"> </span><span class="kw">c</span>(<span class="st">&quot;day1_norm&quot;</span>, <span class="st">&quot;day7_hypo&quot;</span>, <span class="st">&quot;sc68_vehi&quot;</span>, <span class="st">&quot;sc68_cisp&quot;</span>,</span>
<span id="cb30-18"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb30-18"></a>           <span class="st">&quot;pbmc_4000&quot;</span>, <span class="st">&quot;pbmc_6000&quot;</span>)</span>
<span id="cb30-19"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb30-19"></a><span class="cf">for</span>(i <span class="cf">in</span> <span class="dv">1</span><span class="op">:</span><span class="kw">length</span>(data)){</span>
<span id="cb30-20"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb30-20"></a>  p &lt;-<span class="st"> </span><span class="kw">plot_manifold</span>(<span class="dt">obj =</span> data[[i]], <span class="dt">plot_type =</span> <span class="st">&quot;umap&quot;</span>,</span>
<span id="cb30-21"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb30-21"></a>                     <span class="dt">title =</span> <span class="kw">paste</span>(names[i], <span class="st">&quot; (Monocle 3: UMAP)&quot;</span>, <span class="dt">sep =</span> <span class="st">&quot;&quot;</span>),</span>
<span id="cb30-22"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb30-22"></a>                     <span class="dt">title_size =</span> <span class="dv">16</span>, <span class="dt">xlabel =</span> <span class="st">&quot;UMAP_1&quot;</span>, <span class="dt">ylabel =</span> <span class="st">&quot;UMAP_2&quot;</span>)</span>
<span id="cb30-23"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb30-23"></a>  filename &lt;-<span class="st"> </span><span class="kw">paste</span>(<span class="st">&quot;figures/figure_&quot;</span>, <span class="kw">sprintf</span>(<span class="st">&quot;%02d&quot;</span>, <span class="dv">10</span> <span class="op">+</span><span class="st"> </span>i), <span class="st">&quot;_0030.png&quot;</span>,</span>
<span id="cb30-24"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb30-24"></a>                    <span class="dt">sep =</span> <span class="st">&quot;&quot;</span>)</span>
<span id="cb30-25"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb30-25"></a>  <span class="kw">ggsave</span>(<span class="dt">file =</span> filename, <span class="dt">plot =</span> p, <span class="dt">dpi =</span> <span class="dv">300</span>, <span class="dt">width =</span> <span class="dv">4</span>, <span class="dt">height =</span> <span class="fl">4.2</span>)</span>
<span id="cb30-26"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb30-26"></a>}</span>
<span id="cb30-27"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb30-27"></a><span class="cf">for</span>(i <span class="cf">in</span> <span class="dv">1</span><span class="op">:</span><span class="kw">length</span>(data)){</span>
<span id="cb30-28"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb30-28"></a>  p &lt;-<span class="st"> </span><span class="kw">plot_manifold</span>(<span class="dt">obj =</span> data[[i]], <span class="dt">plot_type =</span> <span class="st">&quot;tsne&quot;</span>,</span>
<span id="cb30-29"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb30-29"></a>                     <span class="dt">title =</span> <span class="kw">paste</span>(names[i], <span class="st">&quot; (Monocle 3: tSNE)&quot;</span>, <span class="dt">sep =</span> <span class="st">&quot;&quot;</span>),</span>
<span id="cb30-30"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb30-30"></a>                     <span class="dt">title_size =</span> <span class="dv">16</span>, <span class="dt">xlabel =</span> <span class="st">&quot;tSNE_1&quot;</span>, <span class="dt">ylabel =</span> <span class="st">&quot;tSNE_2&quot;</span>)</span>
<span id="cb30-31"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb30-31"></a>  filename &lt;-<span class="st"> </span><span class="kw">paste</span>(<span class="st">&quot;figures/figure_&quot;</span>, <span class="kw">sprintf</span>(<span class="st">&quot;%02d&quot;</span>, <span class="dv">10</span> <span class="op">+</span><span class="st"> </span>i), <span class="st">&quot;_0031.png&quot;</span>,</span>
<span id="cb30-32"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb30-32"></a>                    <span class="dt">sep =</span> <span class="st">&quot;&quot;</span>)</span>
<span id="cb30-33"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb30-33"></a>  <span class="kw">ggsave</span>(<span class="dt">file =</span> filename, <span class="dt">plot =</span> p, <span class="dt">dpi =</span> <span class="dv">300</span>, <span class="dt">width =</span> <span class="dv">4</span>, <span class="dt">height =</span> <span class="fl">4.2</span>)</span>
<span id="cb30-34"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb30-34"></a>}</span></code></pre></div>
<p><img src="figures/figure_11_0030.png" width="150px">
<img src="figures/figure_12_0030.png" width="150px">
<img src="figures/figure_13_0030.png" width="150px">
<img src="figures/figure_14_0030.png" width="150px">
<img src="figures/figure_15_0030.png" width="150px">
<img src="figures/figure_16_0030.png" width="150px"></p>
<p><img src="figures/figure_11_0031.png" width="150px">
<img src="figures/figure_12_0031.png" width="150px">
<img src="figures/figure_13_0031.png" width="150px">
<img src="figures/figure_14_0031.png" width="150px">
<img src="figures/figure_15_0031.png" width="150px">
<img src="figures/figure_16_0031.png" width="150px"></p>
<p>Save the objects.</p>
<div class="sourceCode" id="cb31"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb31-1"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb31-1"></a><span class="kw">saveRDS</span>(day1_norm, <span class="dt">file =</span> <span class="st">&quot;backup/11_030_day1_norm_monocle3.rds&quot;</span>)</span>
<span id="cb31-2"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb31-2"></a><span class="kw">saveRDS</span>(day7_hypo, <span class="dt">file =</span> <span class="st">&quot;backup/12_030_day7_hypo_monocle3.rds&quot;</span>)</span>
<span id="cb31-3"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb31-3"></a><span class="kw">saveRDS</span>(sc68_vehi, <span class="dt">file =</span> <span class="st">&quot;backup/13_030_sc68_vehi_monocle3.rds&quot;</span>)</span>
<span id="cb31-4"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb31-4"></a><span class="kw">saveRDS</span>(sc68_cisp, <span class="dt">file =</span> <span class="st">&quot;backup/14_030_sc68_cisp_monocle3.rds&quot;</span>)</span>
<span id="cb31-5"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb31-5"></a><span class="kw">saveRDS</span>(pbmc_<span class="dv">4000</span>, <span class="dt">file =</span> <span class="st">&quot;backup/15_030_pbmc_4000_monocle3.rds&quot;</span>)</span>
<span id="cb31-6"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb31-6"></a><span class="kw">saveRDS</span>(pbmc_<span class="dv">6000</span>, <span class="dt">file =</span> <span class="st">&quot;backup/16_030_pbmc_6000_monocle3.rds&quot;</span>)</span></code></pre></div>
<p>Load the objects.</p>
<div class="sourceCode" id="cb32"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb32-1"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb32-1"></a>day1_norm &lt;-<span class="st"> </span><span class="kw">readRDS</span>(<span class="dt">file =</span> <span class="st">&quot;backup/11_030_day1_norm_monocle3.rds&quot;</span>)</span>
<span id="cb32-2"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb32-2"></a>day7_hypo &lt;-<span class="st"> </span><span class="kw">readRDS</span>(<span class="dt">file =</span> <span class="st">&quot;backup/12_030_day7_hypo_monocle3.rds&quot;</span>)</span>
<span id="cb32-3"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb32-3"></a>sc68_vehi &lt;-<span class="st"> </span><span class="kw">readRDS</span>(<span class="dt">file =</span> <span class="st">&quot;backup/13_030_sc68_vehi_monocle3.rds&quot;</span>)</span>
<span id="cb32-4"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb32-4"></a>sc68_cisp &lt;-<span class="st"> </span><span class="kw">readRDS</span>(<span class="dt">file =</span> <span class="st">&quot;backup/14_030_sc68_cisp_monocle3.rds&quot;</span>)</span>
<span id="cb32-5"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb32-5"></a>pbmc_<span class="dv">4000</span> &lt;-<span class="st"> </span><span class="kw">readRDS</span>(<span class="dt">file =</span> <span class="st">&quot;backup/15_030_pbmc_4000_monocle3.rds&quot;</span>)</span>
<span id="cb32-6"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb32-6"></a>pbmc_<span class="dv">6000</span> &lt;-<span class="st"> </span><span class="kw">readRDS</span>(<span class="dt">file =</span> <span class="st">&quot;backup/16_030_pbmc_6000_monocle3.rds&quot;</span>)</span></code></pre></div>
<p>Perform k-means clusterings and compute several clustering validity scores by
using <code>NbClust()</code>.</p>
<div class="sourceCode" id="cb33"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb33-1"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb33-1"></a>compute_nbclust &lt;-<span class="st"> </span><span class="cf">function</span>(obj, plot_type){</span>
<span id="cb33-2"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb33-2"></a>  <span class="cf">if</span>(plot_type <span class="op">==</span><span class="st"> &quot;umap&quot;</span>){</span>
<span id="cb33-3"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb33-3"></a>    mat &lt;-<span class="st"> </span>obj<span class="op">@</span>int_colData<span class="op">@</span>listData[[<span class="st">&quot;reducedDims&quot;</span>]]<span class="op">@</span>listData[[<span class="st">&quot;UMAP&quot;</span>]]</span>
<span id="cb33-4"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb33-4"></a>  }<span class="cf">else</span> <span class="cf">if</span>(plot_type <span class="op">==</span><span class="st"> &quot;tsne&quot;</span>){</span>
<span id="cb33-5"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb33-5"></a>    mat &lt;-<span class="st"> </span>obj<span class="op">@</span>int_colData<span class="op">@</span>listData[[<span class="st">&quot;reducedDims&quot;</span>]]<span class="op">@</span>listData[[<span class="st">&quot;tSNE&quot;</span>]]</span>
<span id="cb33-6"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb33-6"></a>  }</span>
<span id="cb33-7"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb33-7"></a>  <span class="kw">set.seed</span>(<span class="dv">8</span>)</span>
<span id="cb33-8"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb33-8"></a>  res &lt;-<span class="st"> </span><span class="kw">NbClust</span>(mat, <span class="dt">distance =</span> <span class="st">&quot;euclidean&quot;</span>, <span class="dt">min.nc =</span> <span class="dv">2</span>, <span class="dt">max.nc =</span> <span class="dv">7</span>,</span>
<span id="cb33-9"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb33-9"></a>                 <span class="dt">method =</span> <span class="st">&quot;kmeans&quot;</span>, <span class="dt">index =</span> <span class="st">&quot;all&quot;</span>)</span>
<span id="cb33-10"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb33-10"></a>  <span class="kw">return</span>(res)</span>
<span id="cb33-11"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb33-11"></a>}</span>
<span id="cb33-12"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb33-12"></a></span>
<span id="cb33-13"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb33-13"></a>data &lt;-<span class="st"> </span><span class="kw">list</span>(day1_norm, day7_hypo, sc68_vehi, sc68_cisp, pbmc_<span class="dv">4000</span>, pbmc_<span class="dv">6000</span>)</span>
<span id="cb33-14"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb33-14"></a>names &lt;-<span class="st"> </span><span class="kw">c</span>(<span class="st">&quot;day1_norm&quot;</span>, <span class="st">&quot;day7_hypo&quot;</span>, <span class="st">&quot;sc68_vehi&quot;</span>, <span class="st">&quot;sc68_cisp&quot;</span>, <span class="st">&quot;pbmc_4000&quot;</span>,</span>
<span id="cb33-15"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb33-15"></a>           <span class="st">&quot;pbmc_6000&quot;</span>)</span>
<span id="cb33-16"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb33-16"></a>nbclust_umap &lt;-<span class="st"> </span><span class="kw">list</span>()</span>
<span id="cb33-17"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb33-17"></a>nbclust_tsne &lt;-<span class="st"> </span><span class="kw">list</span>()</span>
<span id="cb33-18"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb33-18"></a><span class="cf">for</span>(i <span class="cf">in</span> <span class="dv">1</span><span class="op">:</span><span class="kw">length</span>(data)){</span>
<span id="cb33-19"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb33-19"></a>  nbclust_umap[[i]] &lt;-<span class="st"> </span><span class="kw">compute_nbclust</span>(<span class="dt">obj =</span> data[[i]], <span class="dt">plot_type =</span> <span class="st">&quot;umap&quot;</span>)</span>
<span id="cb33-20"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb33-20"></a>}</span>
<span id="cb33-21"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb33-21"></a><span class="cf">for</span>(i <span class="cf">in</span> <span class="dv">1</span><span class="op">:</span><span class="kw">length</span>(data)){</span>
<span id="cb33-22"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb33-22"></a>  nbclust_tsne[[i]] &lt;-<span class="st"> </span><span class="kw">compute_nbclust</span>(<span class="dt">obj =</span> data[[i]], <span class="dt">plot_type =</span> <span class="st">&quot;tsne&quot;</span>)</span>
<span id="cb33-23"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb33-23"></a>}</span>
<span id="cb33-24"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb33-24"></a><span class="kw">names</span>(nbclust_umap) &lt;-<span class="st"> </span>names</span>
<span id="cb33-25"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb33-25"></a><span class="kw">names</span>(nbclust_tsne) &lt;-<span class="st"> </span>names</span></code></pre></div>
<p>Save the objects.</p>
<div class="sourceCode" id="cb34"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb34-1"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb34-1"></a><span class="kw">saveRDS</span>(nbclust_umap, <span class="dt">file =</span> <span class="st">&quot;backup/30_001_nbclust_umap_monocle3.rds&quot;</span>)</span>
<span id="cb34-2"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb34-2"></a><span class="kw">saveRDS</span>(nbclust_tsne, <span class="dt">file =</span> <span class="st">&quot;backup/30_002_nbclust_tsne_monocle3.rds&quot;</span>)</span>
<span id="cb34-3"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb34-3"></a><span class="co"># ----------------------------------------</span></span>
<span id="cb34-4"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb34-4"></a><span class="co"># Output an excel file</span></span>
<span id="cb34-5"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb34-5"></a><span class="co"># ----------------------------------------</span></span>
<span id="cb34-6"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb34-6"></a>dnames &lt;-<span class="st"> </span><span class="kw">c</span>(<span class="st">&quot;day1_norm&quot;</span>, <span class="st">&quot;day7_hypo&quot;</span>, <span class="st">&quot;sc68_vehi&quot;</span>, <span class="st">&quot;sc68_cisp&quot;</span>,</span>
<span id="cb34-7"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb34-7"></a>            <span class="st">&quot;pbmc_4000&quot;</span>, <span class="st">&quot;pbmc_6000&quot;</span>)</span>
<span id="cb34-8"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb34-8"></a>dlist &lt;-<span class="st"> </span><span class="kw">list</span>()</span>
<span id="cb34-9"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb34-9"></a><span class="cf">for</span>(i <span class="cf">in</span> <span class="dv">1</span><span class="op">:</span><span class="kw">length</span>(dnames)){</span>
<span id="cb34-10"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb34-10"></a>  res &lt;-<span class="st"> </span><span class="kw">rbind</span>(nbclust_umap[[dnames[i]]][[<span class="st">&quot;All.index&quot;</span>]], <span class="st">&quot;&quot;</span>)</span>
<span id="cb34-11"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb34-11"></a>  res &lt;-<span class="st"> </span><span class="kw">rbind</span>(res, nbclust_umap[[dnames[i]]][[<span class="st">&quot;Best.nc&quot;</span>]])</span>
<span id="cb34-12"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb34-12"></a>  dlist &lt;-<span class="st"> </span><span class="kw">c</span>(dlist, <span class="kw">list</span>(res))</span>
<span id="cb34-13"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb34-13"></a>}</span>
<span id="cb34-14"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb34-14"></a><span class="kw">names</span>(dlist) &lt;-<span class="st"> </span>dnames</span>
<span id="cb34-15"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb34-15"></a>filename &lt;-<span class="st"> &quot;backup/nbclust_umap_monocle3.xlsx&quot;</span></span>
<span id="cb34-16"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb34-16"></a><span class="kw">write.xlsx</span>(dlist, <span class="dt">file =</span> filename, <span class="dt">row.names =</span> <span class="ot">TRUE</span>)</span></code></pre></div>
<p>Load the objects.</p>
<div class="sourceCode" id="cb35"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb35-1"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb35-1"></a>nbclust_umap &lt;-<span class="st"> </span><span class="kw">readRDS</span>(<span class="dt">file =</span> <span class="st">&quot;backup/30_001_nbclust_umap_monocle3.rds&quot;</span>)</span>
<span id="cb35-2"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb35-2"></a>nbclust_tsne &lt;-<span class="st"> </span><span class="kw">readRDS</span>(<span class="dt">file =</span> <span class="st">&quot;backup/30_002_nbclust_tsne_monocle3.rds&quot;</span>)</span></code></pre></div>
</div>
<div id="sc3" class="section level3">
<h3><span class="header-section-number">14.3.3</span> SC3</h3>
<p>Load the data.</p>
<div class="sourceCode" id="cb36"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb36-1"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb36-1"></a><span class="kw">rm</span>(<span class="dt">list =</span> <span class="kw">ls</span>())</span>
<span id="cb36-2"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb36-2"></a>day1_norm &lt;-<span class="st"> </span><span class="kw">readRDS</span>(<span class="dt">file =</span> <span class="st">&quot;backup/01_005_day1_norm_normalized.rds&quot;</span>)</span>
<span id="cb36-3"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb36-3"></a>day7_hypo &lt;-<span class="st"> </span><span class="kw">readRDS</span>(<span class="dt">file =</span> <span class="st">&quot;backup/02_005_day7_hypo_normalized.rds&quot;</span>)</span>
<span id="cb36-4"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb36-4"></a>sc68_vehi &lt;-<span class="st"> </span><span class="kw">readRDS</span>(<span class="dt">file =</span> <span class="st">&quot;backup/03_005_sc68_vehi_normalized.rds&quot;</span>)</span>
<span id="cb36-5"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb36-5"></a>sc68_cisp &lt;-<span class="st"> </span><span class="kw">readRDS</span>(<span class="dt">file =</span> <span class="st">&quot;backup/04_005_sc68_cisp_normalized.rds&quot;</span>)</span>
<span id="cb36-6"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb36-6"></a>pbmc_<span class="dv">4000</span> &lt;-<span class="st"> </span><span class="kw">readRDS</span>(<span class="dt">file =</span> <span class="st">&quot;backup/05_005_pbmc_4000_normalized.rds&quot;</span>)</span>
<span id="cb36-7"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb36-7"></a>pbmc_<span class="dv">6000</span> &lt;-<span class="st"> </span><span class="kw">readRDS</span>(<span class="dt">file =</span> <span class="st">&quot;backup/06_005_pbmc_6000_normalized.rds&quot;</span>)</span>
<span id="cb36-8"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb36-8"></a>sc68_vehi_bl &lt;-<span class="st"> </span><span class="kw">readRDS</span>(<span class="dt">file =</span> <span class="st">&quot;backup/13_001_sc68_vehi_blacklist.rds&quot;</span>)</span></code></pre></div>
<p>Remove the outlier cells from <code>sc68_vehi</code>.</p>
<div class="sourceCode" id="cb37"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb37-1"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb37-1"></a><span class="co"># ----------------------------------------</span></span>
<span id="cb37-2"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb37-2"></a><span class="co"># For `sc68_vehi`</span></span>
<span id="cb37-3"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb37-3"></a><span class="co"># ----------------------------------------</span></span>
<span id="cb37-4"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb37-4"></a>mat &lt;-<span class="st"> </span>sc68_vehi[[<span class="st">&quot;data&quot;</span>]][[<span class="st">&quot;bayNorm&quot;</span>]][[<span class="st">&quot;Bay_out&quot;</span>]]</span>
<span id="cb37-5"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb37-5"></a>inds &lt;-<span class="st"> </span><span class="kw">which</span>(<span class="kw">colnames</span>(mat) <span class="op">%in%</span><span class="st"> </span>sc68_vehi_bl)</span>
<span id="cb37-6"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb37-6"></a>sc68_vehi[[<span class="st">&quot;data&quot;</span>]][[<span class="st">&quot;bayNorm&quot;</span>]][[<span class="st">&quot;Bay_out&quot;</span>]] &lt;-<span class="st"> </span>mat[, <span class="op">-</span>inds]</span>
<span id="cb37-7"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb37-7"></a></span>
<span id="cb37-8"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb37-8"></a>mat &lt;-<span class="st"> </span>sc68_vehi[[<span class="st">&quot;data&quot;</span>]][[<span class="st">&quot;log1p&quot;</span>]]</span>
<span id="cb37-9"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb37-9"></a>sc68_vehi[[<span class="st">&quot;data&quot;</span>]][[<span class="st">&quot;log1p&quot;</span>]] &lt;-<span class="st"> </span>mat[, <span class="op">-</span>inds]</span>
<span id="cb37-10"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb37-10"></a>sc68_vehi[[<span class="st">&quot;sample&quot;</span>]] &lt;-<span class="st"> </span>sc68_vehi[[<span class="st">&quot;sample&quot;</span>]][<span class="op">-</span>inds,]</span></code></pre></div>
<p>Create SC3 objects (<code>SingleCellExperiment</code>).</p>
<div class="sourceCode" id="cb38"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb38-1"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb38-1"></a>create_sce_object &lt;-<span class="st"> </span><span class="cf">function</span>(obj){</span>
<span id="cb38-2"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb38-2"></a>  obj_sce &lt;-<span class="st"> </span><span class="kw">SingleCellExperiment</span>(</span>
<span id="cb38-3"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb38-3"></a>    <span class="dt">assay =</span> <span class="kw">list</span>(</span>
<span id="cb38-4"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb38-4"></a>      <span class="dt">counts =</span> <span class="kw">as.matrix</span>(obj[[<span class="st">&quot;data&quot;</span>]][[<span class="st">&quot;bayNorm&quot;</span>]][[<span class="st">&quot;Bay_out&quot;</span>]]),</span>
<span id="cb38-5"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb38-5"></a>      <span class="dt">logcounts =</span> <span class="kw">as.matrix</span>(obj[[<span class="st">&quot;data&quot;</span>]][[<span class="st">&quot;log1p&quot;</span>]])</span>
<span id="cb38-6"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb38-6"></a>    ),</span>
<span id="cb38-7"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb38-7"></a>    <span class="dt">colData =</span> <span class="kw">list</span>(<span class="dt">barcode =</span> obj[[<span class="st">&quot;sample&quot;</span>]][[<span class="st">&quot;barcode&quot;</span>]])</span>
<span id="cb38-8"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb38-8"></a>  )</span>
<span id="cb38-9"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb38-9"></a>  <span class="kw">rowData</span>(obj_sce)<span class="op">$</span>feature_symbol &lt;-<span class="st"> </span>obj[[<span class="st">&quot;variable&quot;</span>]][[<span class="st">&quot;symbol&quot;</span>]]</span>
<span id="cb38-10"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb38-10"></a>  obj_sce &lt;-<span class="st"> </span>obj_sce[<span class="op">!</span><span class="kw">duplicated</span>(<span class="kw">rowData</span>(obj_sce)<span class="op">$</span>feature_symbol),]</span>
<span id="cb38-11"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb38-11"></a>  </span>
<span id="cb38-12"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb38-12"></a>  <span class="kw">return</span>(obj_sce)</span>
<span id="cb38-13"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb38-13"></a>}</span>
<span id="cb38-14"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb38-14"></a></span>
<span id="cb38-15"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb38-15"></a>day1_norm &lt;-<span class="st"> </span><span class="kw">create_sce_object</span>(<span class="dt">obj =</span> day1_norm)</span>
<span id="cb38-16"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb38-16"></a>day7_hypo &lt;-<span class="st"> </span><span class="kw">create_sce_object</span>(<span class="dt">obj =</span> day7_hypo)</span>
<span id="cb38-17"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb38-17"></a>sc68_vehi &lt;-<span class="st"> </span><span class="kw">create_sce_object</span>(<span class="dt">obj =</span> sc68_vehi)</span>
<span id="cb38-18"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb38-18"></a>sc68_cisp &lt;-<span class="st"> </span><span class="kw">create_sce_object</span>(<span class="dt">obj =</span> sc68_cisp)</span>
<span id="cb38-19"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb38-19"></a>pbmc_<span class="dv">4000</span> &lt;-<span class="st"> </span><span class="kw">create_sce_object</span>(<span class="dt">obj =</span> pbmc_<span class="dv">4000</span>)</span>
<span id="cb38-20"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb38-20"></a>pbmc_<span class="dv">6000</span> &lt;-<span class="st"> </span><span class="kw">create_sce_object</span>(<span class="dt">obj =</span> pbmc_<span class="dv">6000</span>)</span></code></pre></div>
<p>Run PCA.</p>
<div class="sourceCode" id="cb39"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb39-1"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb39-1"></a>day1_norm &lt;-<span class="st"> </span><span class="kw">runPCA</span>(day1_norm)</span>
<span id="cb39-2"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb39-2"></a>day7_hypo &lt;-<span class="st"> </span><span class="kw">runPCA</span>(day7_hypo)</span>
<span id="cb39-3"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb39-3"></a>sc68_vehi &lt;-<span class="st"> </span><span class="kw">runPCA</span>(sc68_vehi)</span>
<span id="cb39-4"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb39-4"></a>sc68_cisp &lt;-<span class="st"> </span><span class="kw">runPCA</span>(sc68_cisp)</span>
<span id="cb39-5"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb39-5"></a>pbmc_<span class="dv">4000</span> &lt;-<span class="st"> </span><span class="kw">runPCA</span>(pbmc_<span class="dv">4000</span>)</span>
<span id="cb39-6"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb39-6"></a>pbmc_<span class="dv">6000</span> &lt;-<span class="st"> </span><span class="kw">runPCA</span>(pbmc_<span class="dv">6000</span>)</span></code></pre></div>
<p>Run SC3, in which the parameter <code>ks</code> is subjectively determined by considering
the biological background of the data.
Here, SC3 could not compute for <code>sc68_vehi</code> and <code>sc68_cisp</code>.</p>
<div class="sourceCode" id="cb40"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb40-1"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb40-1"></a>day1_norm &lt;-<span class="st"> </span><span class="kw">sc3</span>(day1_norm, <span class="dt">ks =</span> <span class="dv">2</span><span class="op">:</span><span class="dv">7</span>, <span class="dt">biology =</span> <span class="ot">TRUE</span>)</span>
<span id="cb40-2"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb40-2"></a>day7_hypo &lt;-<span class="st"> </span><span class="kw">sc3</span>(day7_hypo, <span class="dt">ks =</span> <span class="dv">2</span><span class="op">:</span><span class="dv">7</span>, <span class="dt">biology =</span> <span class="ot">TRUE</span>)</span>
<span id="cb40-3"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb40-3"></a>pbmc_<span class="dv">4000</span> &lt;-<span class="st"> </span><span class="kw">sc3</span>(pbmc_<span class="dv">4000</span>, <span class="dt">ks =</span> <span class="dv">2</span><span class="op">:</span><span class="dv">7</span>, <span class="dt">biology =</span> <span class="ot">TRUE</span>)</span>
<span id="cb40-4"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb40-4"></a>pbmc_<span class="dv">6000</span> &lt;-<span class="st"> </span><span class="kw">sc3</span>(pbmc_<span class="dv">6000</span>, <span class="dt">ks =</span> <span class="dv">2</span><span class="op">:</span><span class="dv">7</span>, <span class="dt">biology =</span> <span class="ot">TRUE</span>)</span>
<span id="cb40-5"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb40-5"></a>sc68_vehi &lt;-<span class="st"> </span><span class="kw">sc3</span>(sc68_vehi, <span class="dt">ks =</span> <span class="dv">2</span><span class="op">:</span><span class="dv">7</span>, <span class="dt">biology =</span> <span class="ot">TRUE</span>)  <span class="co"># Error</span></span>
<span id="cb40-6"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb40-6"></a>sc68_cisp &lt;-<span class="st"> </span><span class="kw">sc3</span>(sc68_cisp, <span class="dt">ks =</span> <span class="dv">2</span><span class="op">:</span><span class="dv">7</span>, <span class="dt">biology =</span> <span class="ot">TRUE</span>)  <span class="co"># Error</span></span></code></pre></div>
<p>Save the objects.</p>
<div class="sourceCode" id="cb41"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb41-1"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb41-1"></a><span class="kw">saveRDS</span>(day1_norm, <span class="dt">file =</span> <span class="st">&quot;backup/11_040_day1_norm_sc3.rds&quot;</span>)</span>
<span id="cb41-2"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb41-2"></a><span class="kw">saveRDS</span>(day7_hypo, <span class="dt">file =</span> <span class="st">&quot;backup/12_040_day7_hypo_sc3.rds&quot;</span>)</span>
<span id="cb41-3"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb41-3"></a><span class="co"># saveRDS(sc68_vehi, file = &quot;backup/13_040_sc68_vehi_sc3.rds&quot;)</span></span>
<span id="cb41-4"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb41-4"></a><span class="co"># saveRDS(sc68_cisp, file = &quot;backup/14_040_sc68_cisp_sc3.rds&quot;)</span></span>
<span id="cb41-5"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb41-5"></a><span class="kw">saveRDS</span>(pbmc_<span class="dv">4000</span>, <span class="dt">file =</span> <span class="st">&quot;backup/15_040_pbmc_4000_sc3.rds&quot;</span>)</span>
<span id="cb41-6"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb41-6"></a><span class="kw">saveRDS</span>(pbmc_<span class="dv">6000</span>, <span class="dt">file =</span> <span class="st">&quot;backup/16_040_pbmc_6000_sc3.rds&quot;</span>)</span></code></pre></div>
<p>Load the objects.</p>
<div class="sourceCode" id="cb42"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb42-1"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb42-1"></a>day1_norm &lt;-<span class="st"> </span><span class="kw">readRDS</span>(<span class="dt">file =</span> <span class="st">&quot;backup/11_040_day1_norm_sc3.rds&quot;</span>)</span>
<span id="cb42-2"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb42-2"></a>day7_hypo &lt;-<span class="st"> </span><span class="kw">readRDS</span>(<span class="dt">file =</span> <span class="st">&quot;backup/12_040_day7_hypo_sc3.rds&quot;</span>)</span>
<span id="cb42-3"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb42-3"></a><span class="co"># sc68_vehi &lt;- readRDS(file = &quot;backup/13_040_sc68_vehi_sc3.rds&quot;)</span></span>
<span id="cb42-4"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb42-4"></a><span class="co"># sc68_cisp &lt;- readRDS(file = &quot;backup/14_040_sc68_cisp_sc3.rds&quot;)</span></span>
<span id="cb42-5"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb42-5"></a>pbmc_<span class="dv">4000</span> &lt;-<span class="st"> </span><span class="kw">readRDS</span>(<span class="dt">file =</span> <span class="st">&quot;backup/15_040_pbmc_4000_sc3.rds&quot;</span>)</span>
<span id="cb42-6"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb42-6"></a>pbmc_<span class="dv">6000</span> &lt;-<span class="st"> </span><span class="kw">readRDS</span>(<span class="dt">file =</span> <span class="st">&quot;backup/16_040_pbmc_6000_sc3.rds&quot;</span>)</span></code></pre></div>
<p>Compute average silhouette widths.</p>
<div class="sourceCode" id="cb43"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb43-1"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb43-1"></a>compute_asw &lt;-<span class="st"> </span><span class="cf">function</span>(obj){</span>
<span id="cb43-2"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb43-2"></a>  <span class="cf">if</span>(<span class="op">!</span><span class="kw">is.null</span>(obj)){</span>
<span id="cb43-3"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb43-3"></a>    res &lt;-<span class="st"> </span><span class="kw">list</span>()</span>
<span id="cb43-4"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb43-4"></a>    knames &lt;-<span class="st"> </span><span class="kw">c</span>()</span>
<span id="cb43-5"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb43-5"></a>    <span class="cf">for</span>(k <span class="cf">in</span> <span class="dv">1</span><span class="op">:</span><span class="dv">6</span>){</span>
<span id="cb43-6"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb43-6"></a>      n &lt;-<span class="st"> </span><span class="kw">as.character</span>(k <span class="op">+</span><span class="st"> </span><span class="dv">1</span>)</span>
<span id="cb43-7"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb43-7"></a><span class="co">#      mat &lt;- obj@metadata[[&quot;sc3&quot;]][[&quot;consensus&quot;]][[n]][[&quot;consensus&quot;]]</span></span>
<span id="cb43-8"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb43-8"></a><span class="co">#      indices &lt;- c(&quot;kl&quot;, &quot;ch&quot;, &quot;hartigan&quot;, &quot;ccc&quot;, &quot;scott&quot;, &quot;marriot&quot;, &quot;trcovw&quot;,</span></span>
<span id="cb43-9"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb43-9"></a><span class="co">#                   &quot;tracew&quot;, &quot;friedman&quot;, &quot;rubin&quot;, &quot;cindex&quot;, &quot;db&quot;, &quot;silhouette&quot;,</span></span>
<span id="cb43-10"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb43-10"></a><span class="co">#                   &quot;duda&quot;, &quot;pseudot2&quot;, &quot;beale&quot;, &quot;ratkowsky&quot;, &quot;ball&quot;,</span></span>
<span id="cb43-11"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb43-11"></a><span class="co">#                   &quot;ptbiserial&quot;, &quot;gap&quot;, &quot;frey&quot;, &quot;mcclain&quot;, &quot;gamma&quot;, &quot;gplus&quot;,</span></span>
<span id="cb43-12"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb43-12"></a><span class="co">#                   &quot;tau&quot;, &quot;dunn&quot;, &quot;hubert&quot;, &quot;sdindex&quot;, &quot;dindex&quot;, &quot;sdbw&quot;)</span></span>
<span id="cb43-13"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb43-13"></a><span class="co">#      res[[k]] &lt;- NbClust(mat, method = &quot;ward.D2&quot;, index = indices)</span></span>
<span id="cb43-14"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb43-14"></a>      asw &lt;-<span class="st"> </span><span class="kw">mean</span>(obj<span class="op">@</span>metadata[[<span class="st">&quot;sc3&quot;</span>]][[<span class="st">&quot;consensus&quot;</span>]][[n]][[<span class="st">&quot;silhouette&quot;</span>]][,<span class="dv">3</span>])</span>
<span id="cb43-15"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb43-15"></a>      res[[k]] &lt;-<span class="st"> </span>asw</span>
<span id="cb43-16"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb43-16"></a>      knames[k] &lt;-<span class="st"> </span><span class="kw">paste</span>(<span class="st">&quot;k_&quot;</span>, k <span class="op">+</span><span class="st"> </span><span class="dv">1</span>, <span class="dt">sep =</span> <span class="st">&quot;&quot;</span>)</span>
<span id="cb43-17"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb43-17"></a>    }</span>
<span id="cb43-18"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb43-18"></a>    <span class="kw">names</span>(res) &lt;-<span class="st"> </span>knames</span>
<span id="cb43-19"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb43-19"></a>    <span class="kw">return</span>(res)</span>
<span id="cb43-20"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb43-20"></a>  }<span class="cf">else</span>{</span>
<span id="cb43-21"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb43-21"></a>    <span class="kw">return</span>(<span class="ot">NULL</span>)</span>
<span id="cb43-22"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb43-22"></a>  }</span>
<span id="cb43-23"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb43-23"></a>}</span>
<span id="cb43-24"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb43-24"></a></span>
<span id="cb43-25"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb43-25"></a>data &lt;-<span class="st"> </span><span class="kw">list</span>(day1_norm, day7_hypo, <span class="ot">NULL</span>, <span class="ot">NULL</span>, pbmc_<span class="dv">4000</span>, pbmc_<span class="dv">6000</span>)</span>
<span id="cb43-26"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb43-26"></a>names &lt;-<span class="st"> </span><span class="kw">c</span>(<span class="st">&quot;day1_norm&quot;</span>, <span class="st">&quot;day7_hypo&quot;</span>, <span class="st">&quot;NULL&quot;</span>, <span class="st">&quot;NULL&quot;</span>, <span class="st">&quot;pbmc_4000&quot;</span>, <span class="st">&quot;pbmc_6000&quot;</span>)</span>
<span id="cb43-27"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb43-27"></a>asw_sc3 &lt;-<span class="st"> </span><span class="kw">list</span>()</span>
<span id="cb43-28"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb43-28"></a><span class="cf">for</span>(i <span class="cf">in</span> <span class="dv">1</span><span class="op">:</span><span class="kw">length</span>(data)){</span>
<span id="cb43-29"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb43-29"></a>  asw_sc3[[i]] &lt;-<span class="st"> </span><span class="kw">compute_nbclust</span>(<span class="dt">obj =</span> data[[i]])</span>
<span id="cb43-30"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb43-30"></a>}</span>
<span id="cb43-31"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb43-31"></a><span class="kw">names</span>(asw_sc3) &lt;-<span class="st"> </span>names</span></code></pre></div>
<p>Save the objects.</p>
<div class="sourceCode" id="cb44"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb44-1"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb44-1"></a><span class="kw">saveRDS</span>(asw_sc3, <span class="dt">file =</span> <span class="st">&quot;backup/40_001_asw_sc3.rds&quot;</span>)</span>
<span id="cb44-2"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb44-2"></a><span class="co"># ----------------------------------------</span></span>
<span id="cb44-3"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb44-3"></a><span class="co"># Output an excel file</span></span>
<span id="cb44-4"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb44-4"></a><span class="co"># ----------------------------------------</span></span>
<span id="cb44-5"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb44-5"></a>dnames &lt;-<span class="st"> </span><span class="kw">c</span>(<span class="st">&quot;day1_norm&quot;</span>, <span class="st">&quot;day7_hypo&quot;</span>, <span class="st">&quot;sc68_vehi&quot;</span>, <span class="st">&quot;sc68_cisp&quot;</span>,</span>
<span id="cb44-6"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb44-6"></a>            <span class="st">&quot;pbmc_4000&quot;</span>, <span class="st">&quot;pbmc_6000&quot;</span>)</span>
<span id="cb44-7"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb44-7"></a>res &lt;-<span class="st"> </span><span class="kw">c</span>()</span>
<span id="cb44-8"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb44-8"></a><span class="cf">for</span>(n <span class="cf">in</span> <span class="dv">1</span><span class="op">:</span><span class="kw">length</span>(dnames)){</span>
<span id="cb44-9"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb44-9"></a>  <span class="cf">if</span>(<span class="op">!</span><span class="kw">is.null</span>(asw_sc3[[n]])){</span>
<span id="cb44-10"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb44-10"></a>    tmp &lt;-<span class="st"> </span><span class="kw">t</span>(<span class="kw">unlist</span>(asw_sc3[[n]]))</span>
<span id="cb44-11"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb44-11"></a>  }<span class="cf">else</span>{</span>
<span id="cb44-12"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb44-12"></a>    tmp &lt;-<span class="st"> </span><span class="kw">matrix</span>(<span class="dt">ncol =</span> <span class="kw">length</span>(dnames), <span class="dt">nrow =</span> <span class="dv">1</span>, <span class="ot">NA</span>)</span>
<span id="cb44-13"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb44-13"></a>  }</span>
<span id="cb44-14"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb44-14"></a>  <span class="kw">rownames</span>(tmp) &lt;-<span class="st"> </span>dnames[n]</span>
<span id="cb44-15"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb44-15"></a>  res &lt;-<span class="st"> </span><span class="kw">rbind</span>(res, tmp)</span>
<span id="cb44-16"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb44-16"></a>}</span>
<span id="cb44-17"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb44-17"></a>filename &lt;-<span class="st"> &quot;backup/asw_sc3.xlsx&quot;</span></span>
<span id="cb44-18"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb44-18"></a><span class="kw">write.xlsx</span>(res, <span class="dt">file =</span> filename, <span class="dt">row.names =</span> <span class="ot">TRUE</span>)</span></code></pre></div>
<p>Load the objects.</p>
<div class="sourceCode" id="cb45"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb45-1"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb45-1"></a>asw_sc3 &lt;-<span class="st"> </span><span class="kw">readRDS</span>(<span class="dt">file =</span> <span class="st">&quot;backup/40_001_asw_sc3.rds&quot;</span>)</span></code></pre></div>
</div>
<div id="principal-component-analysis-pca" class="section level3">
<h3><span class="header-section-number">14.3.4</span> Principal component analysis (PCA)</h3>
<p>Load the data.</p>
<div class="sourceCode" id="cb46"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb46-1"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb46-1"></a><span class="kw">rm</span>(<span class="dt">list =</span> <span class="kw">ls</span>())</span>
<span id="cb46-2"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb46-2"></a>day1_norm &lt;-<span class="st"> </span><span class="kw">readRDS</span>(<span class="dt">file =</span> <span class="st">&quot;backup/01_005_day1_norm_normalized.rds&quot;</span>)</span>
<span id="cb46-3"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb46-3"></a>day7_hypo &lt;-<span class="st"> </span><span class="kw">readRDS</span>(<span class="dt">file =</span> <span class="st">&quot;backup/02_005_day7_hypo_normalized.rds&quot;</span>)</span>
<span id="cb46-4"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb46-4"></a>sc68_vehi &lt;-<span class="st"> </span><span class="kw">readRDS</span>(<span class="dt">file =</span> <span class="st">&quot;backup/03_005_sc68_vehi_normalized.rds&quot;</span>)</span>
<span id="cb46-5"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb46-5"></a>sc68_cisp &lt;-<span class="st"> </span><span class="kw">readRDS</span>(<span class="dt">file =</span> <span class="st">&quot;backup/04_005_sc68_cisp_normalized.rds&quot;</span>)</span>
<span id="cb46-6"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb46-6"></a>pbmc_<span class="dv">4000</span> &lt;-<span class="st"> </span><span class="kw">readRDS</span>(<span class="dt">file =</span> <span class="st">&quot;backup/05_005_pbmc_4000_normalized.rds&quot;</span>)</span>
<span id="cb46-7"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb46-7"></a>pbmc_<span class="dv">6000</span> &lt;-<span class="st"> </span><span class="kw">readRDS</span>(<span class="dt">file =</span> <span class="st">&quot;backup/06_005_pbmc_6000_normalized.rds&quot;</span>)</span>
<span id="cb46-8"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb46-8"></a>sc68_vehi_bl &lt;-<span class="st"> </span><span class="kw">readRDS</span>(<span class="dt">file =</span> <span class="st">&quot;backup/13_001_sc68_vehi_blacklist.rds&quot;</span>)</span></code></pre></div>
<p>Remove the outlier cells from <code>sc68_vehi</code>.</p>
<div class="sourceCode" id="cb47"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb47-1"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb47-1"></a><span class="co"># ----------------------------------------</span></span>
<span id="cb47-2"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb47-2"></a><span class="co"># For `sc68_vehi`</span></span>
<span id="cb47-3"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb47-3"></a><span class="co"># ----------------------------------------</span></span>
<span id="cb47-4"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb47-4"></a>mat &lt;-<span class="st"> </span>sc68_vehi[[<span class="st">&quot;data&quot;</span>]][[<span class="st">&quot;log1p&quot;</span>]]</span>
<span id="cb47-5"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb47-5"></a>inds &lt;-<span class="st"> </span><span class="kw">which</span>(<span class="kw">colnames</span>(mat) <span class="op">%in%</span><span class="st"> </span>sc68_vehi_bl)</span>
<span id="cb47-6"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb47-6"></a>sc68_vehi[[<span class="st">&quot;data&quot;</span>]][[<span class="st">&quot;log1p&quot;</span>]] &lt;-<span class="st"> </span>mat[, <span class="op">-</span>inds]</span></code></pre></div>
<p>Run PCA for the z-score matrix of gene expression.</p>
<div class="sourceCode" id="cb48"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb48-1"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb48-1"></a>do_pca &lt;-<span class="st"> </span><span class="cf">function</span>(obj){</span>
<span id="cb48-2"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb48-2"></a>  mat &lt;-<span class="st"> </span><span class="kw">t</span>(obj[[<span class="st">&quot;data&quot;</span>]][[<span class="st">&quot;log1p&quot;</span>]])</span>
<span id="cb48-3"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb48-3"></a>  <span class="kw">set.seed</span>(<span class="dv">8</span>)</span>
<span id="cb48-4"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb48-4"></a>  obj[[<span class="st">&quot;reduction&quot;</span>]][[<span class="st">&quot;pca&quot;</span>]] &lt;-<span class="st"> </span><span class="kw">prcomp</span>(mat, <span class="dt">scale =</span> <span class="ot">TRUE</span>)</span>
<span id="cb48-5"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb48-5"></a>  <span class="kw">return</span>(obj)</span>
<span id="cb48-6"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb48-6"></a>}</span>
<span id="cb48-7"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb48-7"></a></span>
<span id="cb48-8"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb48-8"></a>day1_norm &lt;-<span class="st"> </span><span class="kw">do_pca</span>(day1_norm)</span>
<span id="cb48-9"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb48-9"></a>day7_hypo &lt;-<span class="st"> </span><span class="kw">do_pca</span>(day7_hypo)</span>
<span id="cb48-10"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb48-10"></a>sc68_vehi &lt;-<span class="st"> </span><span class="kw">do_pca</span>(sc68_vehi)</span>
<span id="cb48-11"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb48-11"></a>sc68_cisp &lt;-<span class="st"> </span><span class="kw">do_pca</span>(sc68_cisp)</span>
<span id="cb48-12"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb48-12"></a>pbmc_<span class="dv">4000</span> &lt;-<span class="st"> </span><span class="kw">do_pca</span>(pbmc_<span class="dv">4000</span>)</span>
<span id="cb48-13"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb48-13"></a>pbmc_<span class="dv">6000</span> &lt;-<span class="st"> </span><span class="kw">do_pca</span>(pbmc_<span class="dv">6000</span>)</span></code></pre></div>
<p>The following command can calculate the cumulative sum over whole sum.
These results are used to determine the dimensions of the principal components.
In what follows, we use the PC whose contribution ratio is greater than
or equal to 0.9 for the first time.</p>
<div class="sourceCode" id="cb49"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb49-1"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb49-1"></a>day1_norm_pc &lt;-<span class="st"> </span><span class="kw">which</span>(<span class="kw">cumsum</span>(day1_norm[[<span class="st">&quot;reduction&quot;</span>]][[<span class="st">&quot;pca&quot;</span>]][[<span class="st">&quot;sdev&quot;</span>]]) <span class="op">/</span></span>
<span id="cb49-2"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb49-2"></a><span class="st">                      </span><span class="kw">sum</span>(day1_norm[[<span class="st">&quot;reduction&quot;</span>]][[<span class="st">&quot;pca&quot;</span>]][[<span class="st">&quot;sdev&quot;</span>]]) <span class="op">&gt;</span><span class="st"> </span><span class="fl">0.9</span>)[<span class="dv">1</span>]</span>
<span id="cb49-3"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb49-3"></a>day7_hypo_pc &lt;-<span class="st"> </span><span class="kw">which</span>(<span class="kw">cumsum</span>(day7_hypo[[<span class="st">&quot;reduction&quot;</span>]][[<span class="st">&quot;pca&quot;</span>]][[<span class="st">&quot;sdev&quot;</span>]]) <span class="op">/</span></span>
<span id="cb49-4"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb49-4"></a><span class="st">                      </span><span class="kw">sum</span>(day7_hypo[[<span class="st">&quot;reduction&quot;</span>]][[<span class="st">&quot;pca&quot;</span>]][[<span class="st">&quot;sdev&quot;</span>]]) <span class="op">&gt;</span><span class="st"> </span><span class="fl">0.9</span>)[<span class="dv">1</span>]</span>
<span id="cb49-5"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb49-5"></a>sc68_vehi_pc &lt;-<span class="st"> </span><span class="kw">which</span>(<span class="kw">cumsum</span>(sc68_vehi[[<span class="st">&quot;reduction&quot;</span>]][[<span class="st">&quot;pca&quot;</span>]][[<span class="st">&quot;sdev&quot;</span>]]) <span class="op">/</span></span>
<span id="cb49-6"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb49-6"></a><span class="st">                      </span><span class="kw">sum</span>(sc68_vehi[[<span class="st">&quot;reduction&quot;</span>]][[<span class="st">&quot;pca&quot;</span>]][[<span class="st">&quot;sdev&quot;</span>]]) <span class="op">&gt;</span><span class="st"> </span><span class="fl">0.9</span>)[<span class="dv">1</span>]</span>
<span id="cb49-7"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb49-7"></a>sc68_cisp_pc &lt;-<span class="st"> </span><span class="kw">which</span>(<span class="kw">cumsum</span>(sc68_cisp[[<span class="st">&quot;reduction&quot;</span>]][[<span class="st">&quot;pca&quot;</span>]][[<span class="st">&quot;sdev&quot;</span>]]) <span class="op">/</span></span>
<span id="cb49-8"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb49-8"></a><span class="st">                      </span><span class="kw">sum</span>(sc68_cisp[[<span class="st">&quot;reduction&quot;</span>]][[<span class="st">&quot;pca&quot;</span>]][[<span class="st">&quot;sdev&quot;</span>]]) <span class="op">&gt;</span><span class="st"> </span><span class="fl">0.9</span>)[<span class="dv">1</span>]</span>
<span id="cb49-9"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb49-9"></a>pbmc_<span class="dv">4000</span>_pc &lt;-<span class="st"> </span><span class="kw">which</span>(<span class="kw">cumsum</span>(pbmc_<span class="dv">4000</span>[[<span class="st">&quot;reduction&quot;</span>]][[<span class="st">&quot;pca&quot;</span>]][[<span class="st">&quot;sdev&quot;</span>]]) <span class="op">/</span></span>
<span id="cb49-10"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb49-10"></a><span class="st">                      </span><span class="kw">sum</span>(pbmc_<span class="dv">4000</span>[[<span class="st">&quot;reduction&quot;</span>]][[<span class="st">&quot;pca&quot;</span>]][[<span class="st">&quot;sdev&quot;</span>]]) <span class="op">&gt;</span><span class="st"> </span><span class="fl">0.9</span>)[<span class="dv">1</span>]</span>
<span id="cb49-11"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb49-11"></a>pbmc_<span class="dv">6000</span>_pc &lt;-<span class="st"> </span><span class="kw">which</span>(<span class="kw">cumsum</span>(pbmc_<span class="dv">6000</span>[[<span class="st">&quot;reduction&quot;</span>]][[<span class="st">&quot;pca&quot;</span>]][[<span class="st">&quot;sdev&quot;</span>]]) <span class="op">/</span></span>
<span id="cb49-12"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb49-12"></a><span class="st">                      </span><span class="kw">sum</span>(pbmc_<span class="dv">6000</span>[[<span class="st">&quot;reduction&quot;</span>]][[<span class="st">&quot;pca&quot;</span>]][[<span class="st">&quot;sdev&quot;</span>]]) <span class="op">&gt;</span><span class="st"> </span><span class="fl">0.9</span>)[<span class="dv">1</span>]</span></code></pre></div>
<p>The following functions computes t-SNE and UMAP.</p>
<div class="sourceCode" id="cb50"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb50-1"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb50-1"></a>do_tsne &lt;-<span class="st"> </span><span class="cf">function</span>(obj, pca_dim, tsne_dim){</span>
<span id="cb50-2"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb50-2"></a>  mat &lt;-<span class="st"> </span><span class="kw">as.matrix</span>(obj[[<span class="st">&quot;reduction&quot;</span>]][[<span class="st">&quot;pca&quot;</span>]][[<span class="st">&quot;x&quot;</span>]][, <span class="dv">1</span><span class="op">:</span>pca_dim])</span>
<span id="cb50-3"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb50-3"></a>  <span class="kw">set.seed</span>(<span class="dv">8</span>)</span>
<span id="cb50-4"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb50-4"></a>  res &lt;-<span class="st"> </span><span class="kw">Rtsne</span>(mat, <span class="dt">dim =</span> tsne_dim, <span class="dt">pca =</span> <span class="ot">FALSE</span>)</span>
<span id="cb50-5"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb50-5"></a>  obj[[<span class="st">&quot;reduction&quot;</span>]][[<span class="st">&quot;tsne&quot;</span>]] &lt;-<span class="st"> </span>res</span>
<span id="cb50-6"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb50-6"></a>  <span class="kw">return</span>(obj)</span>
<span id="cb50-7"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb50-7"></a>}</span>
<span id="cb50-8"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb50-8"></a></span>
<span id="cb50-9"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb50-9"></a>do_umap &lt;-<span class="st"> </span><span class="cf">function</span>(obj, pca_dim, umap_dim){</span>
<span id="cb50-10"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb50-10"></a>  mat &lt;-<span class="st"> </span><span class="kw">as.matrix</span>(obj[[<span class="st">&quot;reduction&quot;</span>]][[<span class="st">&quot;pca&quot;</span>]][[<span class="st">&quot;x&quot;</span>]][, <span class="dv">1</span><span class="op">:</span>pca_dim])</span>
<span id="cb50-11"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb50-11"></a>  <span class="kw">set.seed</span>(<span class="dv">8</span>)</span>
<span id="cb50-12"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb50-12"></a>  res &lt;-<span class="st"> </span><span class="kw">umap</span>(mat, <span class="dt">n_components =</span> umap_dim)</span>
<span id="cb50-13"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb50-13"></a>  obj[[<span class="st">&quot;reduction&quot;</span>]][[<span class="st">&quot;umap&quot;</span>]] &lt;-<span class="st"> </span>res</span>
<span id="cb50-14"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb50-14"></a>  <span class="kw">return</span>(obj)</span>
<span id="cb50-15"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb50-15"></a>}</span>
<span id="cb50-16"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb50-16"></a></span>
<span id="cb50-17"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb50-17"></a>day1_norm &lt;-<span class="st"> </span><span class="kw">do_umap</span>(<span class="dt">obj =</span> day1_norm, <span class="dt">pca_dim =</span> day1_norm_pc, <span class="dt">umap_dim =</span> <span class="dv">2</span>)</span>
<span id="cb50-18"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb50-18"></a>day7_hypo &lt;-<span class="st"> </span><span class="kw">do_umap</span>(<span class="dt">obj =</span> day7_hypo, <span class="dt">pca_dim =</span> day7_hypo_pc, <span class="dt">umap_dim =</span> <span class="dv">2</span>)</span>
<span id="cb50-19"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb50-19"></a>sc68_vehi &lt;-<span class="st"> </span><span class="kw">do_umap</span>(<span class="dt">obj =</span> sc68_vehi, <span class="dt">pca_dim =</span> sc68_vehi_pc, <span class="dt">umap_dim =</span> <span class="dv">2</span>)</span>
<span id="cb50-20"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb50-20"></a>sc68_cisp &lt;-<span class="st"> </span><span class="kw">do_umap</span>(<span class="dt">obj =</span> sc68_cisp, <span class="dt">pca_dim =</span> sc68_cisp_pc, <span class="dt">umap_dim =</span> <span class="dv">2</span>)</span>
<span id="cb50-21"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb50-21"></a>pbmc_<span class="dv">4000</span> &lt;-<span class="st"> </span><span class="kw">do_umap</span>(<span class="dt">obj =</span> pbmc_<span class="dv">4000</span>, <span class="dt">pca_dim =</span> pbmc_<span class="dv">4000</span>_pc, <span class="dt">umap_dim =</span> <span class="dv">2</span>)</span>
<span id="cb50-22"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb50-22"></a>pbmc_<span class="dv">6000</span> &lt;-<span class="st"> </span><span class="kw">do_umap</span>(<span class="dt">obj =</span> pbmc_<span class="dv">6000</span>, <span class="dt">pca_dim =</span> pbmc_<span class="dv">6000</span>_pc, <span class="dt">umap_dim =</span> <span class="dv">2</span>)</span>
<span id="cb50-23"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb50-23"></a></span>
<span id="cb50-24"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb50-24"></a>day1_norm &lt;-<span class="st"> </span><span class="kw">do_tsne</span>(<span class="dt">obj =</span> day1_norm, <span class="dt">pca_dim =</span> day1_norm_pc, <span class="dt">tsne_dim =</span> <span class="dv">2</span>)</span>
<span id="cb50-25"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb50-25"></a>day7_hypo &lt;-<span class="st"> </span><span class="kw">do_tsne</span>(<span class="dt">obj =</span> day7_hypo, <span class="dt">pca_dim =</span> day7_hypo_pc, <span class="dt">tsne_dim =</span> <span class="dv">2</span>)</span>
<span id="cb50-26"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb50-26"></a>sc68_vehi &lt;-<span class="st"> </span><span class="kw">do_tsne</span>(<span class="dt">obj =</span> sc68_vehi, <span class="dt">pca_dim =</span> sc68_vehi_pc, <span class="dt">tsne_dim =</span> <span class="dv">2</span>)</span>
<span id="cb50-27"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb50-27"></a>sc68_cisp &lt;-<span class="st"> </span><span class="kw">do_tsne</span>(<span class="dt">obj =</span> sc68_cisp, <span class="dt">pca_dim =</span> sc68_cisp_pc, <span class="dt">tsne_dim =</span> <span class="dv">2</span>)</span>
<span id="cb50-28"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb50-28"></a>pbmc_<span class="dv">4000</span> &lt;-<span class="st"> </span><span class="kw">do_tsne</span>(<span class="dt">obj =</span> pbmc_<span class="dv">4000</span>, <span class="dt">pca_dim =</span> pbmc_<span class="dv">4000</span>_pc, <span class="dt">tsne_dim =</span> <span class="dv">2</span>)</span>
<span id="cb50-29"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb50-29"></a>pbmc_<span class="dv">6000</span> &lt;-<span class="st"> </span><span class="kw">do_tsne</span>(<span class="dt">obj =</span> pbmc_<span class="dv">6000</span>, <span class="dt">pca_dim =</span> pbmc_<span class="dv">6000</span>_pc, <span class="dt">tsne_dim =</span> <span class="dv">2</span>)</span></code></pre></div>
<p>Save the objects.</p>
<div class="sourceCode" id="cb51"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb51-1"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb51-1"></a><span class="kw">saveRDS</span>(day1_norm, <span class="dt">file =</span> <span class="st">&quot;backup/11_050_day1_norm_pca.rds&quot;</span>)</span>
<span id="cb51-2"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb51-2"></a><span class="kw">saveRDS</span>(day7_hypo, <span class="dt">file =</span> <span class="st">&quot;backup/12_050_day7_hypo_pca.rds&quot;</span>)</span>
<span id="cb51-3"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb51-3"></a><span class="kw">saveRDS</span>(sc68_vehi, <span class="dt">file =</span> <span class="st">&quot;backup/13_050_sc68_vehi_pca.rds&quot;</span>)</span>
<span id="cb51-4"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb51-4"></a><span class="kw">saveRDS</span>(sc68_cisp, <span class="dt">file =</span> <span class="st">&quot;backup/14_050_sc68_cisp_pca.rds&quot;</span>)</span>
<span id="cb51-5"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb51-5"></a><span class="kw">saveRDS</span>(pbmc_<span class="dv">4000</span>, <span class="dt">file =</span> <span class="st">&quot;backup/15_050_pbmc_4000_pca.rds&quot;</span>)</span>
<span id="cb51-6"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb51-6"></a><span class="kw">saveRDS</span>(pbmc_<span class="dv">6000</span>, <span class="dt">file =</span> <span class="st">&quot;backup/16_050_pbmc_6000_pca.rds&quot;</span>)</span></code></pre></div>
<p>Load the objects.</p>
<div class="sourceCode" id="cb52"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb52-1"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb52-1"></a>day1_norm &lt;-<span class="st"> </span><span class="kw">readRDS</span>(<span class="dt">file =</span> <span class="st">&quot;backup/11_050_day1_norm_pca.rds&quot;</span>)</span>
<span id="cb52-2"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb52-2"></a>day7_hypo &lt;-<span class="st"> </span><span class="kw">readRDS</span>(<span class="dt">file =</span> <span class="st">&quot;backup/12_050_day7_hypo_pca.rds&quot;</span>)</span>
<span id="cb52-3"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb52-3"></a>sc68_vehi &lt;-<span class="st"> </span><span class="kw">readRDS</span>(<span class="dt">file =</span> <span class="st">&quot;backup/13_050_sc68_vehi_pca.rds&quot;</span>)</span>
<span id="cb52-4"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb52-4"></a>sc68_cisp &lt;-<span class="st"> </span><span class="kw">readRDS</span>(<span class="dt">file =</span> <span class="st">&quot;backup/14_050_sc68_cisp_pca.rds&quot;</span>)</span>
<span id="cb52-5"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb52-5"></a>pbmc_<span class="dv">4000</span> &lt;-<span class="st"> </span><span class="kw">readRDS</span>(<span class="dt">file =</span> <span class="st">&quot;backup/15_050_pbmc_4000_pca.rds&quot;</span>)</span>
<span id="cb52-6"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb52-6"></a>pbmc_<span class="dv">6000</span> &lt;-<span class="st"> </span><span class="kw">readRDS</span>(<span class="dt">file =</span> <span class="st">&quot;backup/16_050_pbmc_6000_pca.rds&quot;</span>)</span></code></pre></div>
<p>The following function <code>plot_manifold()</code> shows sample states on t-SNE or UMAP
space.</p>
<div class="sourceCode" id="cb53"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb53-1"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb53-1"></a>plot_manifold &lt;-<span class="st"> </span><span class="cf">function</span>(obj, plot_type, title, title_size, xlabel, ylabel){</span>
<span id="cb53-2"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb53-2"></a>  <span class="cf">if</span>(plot_type <span class="op">==</span><span class="st"> &quot;umap&quot;</span>){</span>
<span id="cb53-3"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb53-3"></a>    df &lt;-<span class="st"> </span>obj[[<span class="st">&quot;reduction&quot;</span>]][[<span class="st">&quot;umap&quot;</span>]][[<span class="st">&quot;layout&quot;</span>]]</span>
<span id="cb53-4"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb53-4"></a>  }<span class="cf">else</span> <span class="cf">if</span>(plot_type <span class="op">==</span><span class="st"> &quot;tsne&quot;</span>){</span>
<span id="cb53-5"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb53-5"></a>    df &lt;-<span class="st"> </span>obj[[<span class="st">&quot;reduction&quot;</span>]][[<span class="st">&quot;tsne&quot;</span>]][[<span class="st">&quot;Y&quot;</span>]]</span>
<span id="cb53-6"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb53-6"></a>  }</span>
<span id="cb53-7"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb53-7"></a>  p &lt;-<span class="st"> </span><span class="kw">ggplot</span>() <span class="op">+</span></span>
<span id="cb53-8"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb53-8"></a><span class="st">    </span><span class="kw">geom_point</span>(<span class="kw">aes</span>(<span class="dt">x=</span>df[,<span class="dv">1</span>], <span class="dt">y=</span>df[,<span class="dv">2</span>]), <span class="dt">color=</span><span class="st">&quot;black&quot;</span>, <span class="dt">size=</span><span class="fl">0.5</span>, <span class="dt">alpha=</span><span class="fl">1.0</span>) <span class="op">+</span></span>
<span id="cb53-9"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb53-9"></a><span class="st">    </span><span class="kw">labs</span>(<span class="dt">title=</span>title, <span class="dt">x=</span>xlabel, <span class="dt">y=</span>ylabel) <span class="op">+</span></span>
<span id="cb53-10"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb53-10"></a><span class="st">    </span><span class="kw">theme_classic</span>(<span class="dt">base_size=</span><span class="dv">20</span>, <span class="dt">base_family=</span><span class="st">&quot;Helvetica&quot;</span>) <span class="op">+</span></span>
<span id="cb53-11"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb53-11"></a><span class="st">    </span><span class="kw">theme</span>(<span class="dt">plot.title=</span><span class="kw">element_text</span>(<span class="dt">hjust=</span><span class="fl">0.5</span>, <span class="dt">size=</span>title_size),</span>
<span id="cb53-12"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb53-12"></a>      <span class="dt">plot.margin=</span><span class="kw">unit</span>(<span class="kw">c</span>(<span class="fl">1.5</span>,<span class="fl">1.5</span>,<span class="fl">1.5</span>,<span class="fl">1.5</span>), <span class="st">&quot;lines&quot;</span>), <span class="dt">legend.position=</span><span class="st">&quot;right&quot;</span>)</span>
<span id="cb53-13"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb53-13"></a>  <span class="kw">return</span>(p)</span>
<span id="cb53-14"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb53-14"></a>}</span>
<span id="cb53-15"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb53-15"></a></span>
<span id="cb53-16"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb53-16"></a>data &lt;-<span class="st"> </span><span class="kw">list</span>(day1_norm, day7_hypo, sc68_vehi, sc68_cisp, pbmc_<span class="dv">4000</span>, pbmc_<span class="dv">6000</span>)</span>
<span id="cb53-17"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb53-17"></a>names &lt;-<span class="st"> </span><span class="kw">c</span>(<span class="st">&quot;day1_norm&quot;</span>, <span class="st">&quot;day7_hypo&quot;</span>, <span class="st">&quot;sc68_vehi&quot;</span>, <span class="st">&quot;sc68_cisp&quot;</span>,</span>
<span id="cb53-18"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb53-18"></a>           <span class="st">&quot;pbmc_4000&quot;</span>, <span class="st">&quot;pbmc_6000&quot;</span>)</span>
<span id="cb53-19"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb53-19"></a><span class="cf">for</span>(i <span class="cf">in</span> <span class="dv">1</span><span class="op">:</span><span class="kw">length</span>(data)){</span>
<span id="cb53-20"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb53-20"></a>  p &lt;-<span class="st"> </span><span class="kw">plot_manifold</span>(<span class="dt">obj =</span> data[[i]], <span class="dt">plot_type =</span> <span class="st">&quot;umap&quot;</span>,</span>
<span id="cb53-21"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb53-21"></a>                     <span class="dt">title =</span> <span class="kw">paste</span>(names[i], <span class="st">&quot; (PCA + UMAP)&quot;</span>, <span class="dt">sep =</span> <span class="st">&quot;&quot;</span>),</span>
<span id="cb53-22"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb53-22"></a>                     <span class="dt">title_size =</span> <span class="dv">16</span>, <span class="dt">xlabel =</span> <span class="st">&quot;UMAP_1&quot;</span>, <span class="dt">ylabel =</span> <span class="st">&quot;UMAP_2&quot;</span>)</span>
<span id="cb53-23"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb53-23"></a>  filename &lt;-<span class="st"> </span><span class="kw">paste</span>(<span class="st">&quot;figures/figure_&quot;</span>, <span class="kw">sprintf</span>(<span class="st">&quot;%02d&quot;</span>, <span class="dv">10</span> <span class="op">+</span><span class="st"> </span>i), <span class="st">&quot;_0050.png&quot;</span>,</span>
<span id="cb53-24"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb53-24"></a>                    <span class="dt">sep =</span> <span class="st">&quot;&quot;</span>)</span>
<span id="cb53-25"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb53-25"></a>  <span class="kw">ggsave</span>(<span class="dt">file =</span> filename, <span class="dt">plot =</span> p, <span class="dt">dpi =</span> <span class="dv">300</span>, <span class="dt">width =</span> <span class="dv">4</span>, <span class="dt">height =</span> <span class="fl">4.2</span>)</span>
<span id="cb53-26"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb53-26"></a>}</span>
<span id="cb53-27"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb53-27"></a><span class="cf">for</span>(i <span class="cf">in</span> <span class="dv">1</span><span class="op">:</span><span class="kw">length</span>(data)){</span>
<span id="cb53-28"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb53-28"></a>  p &lt;-<span class="st"> </span><span class="kw">plot_manifold</span>(<span class="dt">obj =</span> data[[i]], <span class="dt">plot_type =</span> <span class="st">&quot;tsne&quot;</span>,</span>
<span id="cb53-29"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb53-29"></a>                     <span class="dt">title =</span> <span class="kw">paste</span>(names[i], <span class="st">&quot; (PCA + tSNE)&quot;</span>, <span class="dt">sep =</span> <span class="st">&quot;&quot;</span>),</span>
<span id="cb53-30"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb53-30"></a>                     <span class="dt">title_size =</span> <span class="dv">16</span>, <span class="dt">xlabel =</span> <span class="st">&quot;tSNE_1&quot;</span>, <span class="dt">ylabel =</span> <span class="st">&quot;tSNE_2&quot;</span>)</span>
<span id="cb53-31"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb53-31"></a>  filename &lt;-<span class="st"> </span><span class="kw">paste</span>(<span class="st">&quot;figures/figure_&quot;</span>, <span class="kw">sprintf</span>(<span class="st">&quot;%02d&quot;</span>, <span class="dv">10</span> <span class="op">+</span><span class="st"> </span>i), <span class="st">&quot;_0051.png&quot;</span>,</span>
<span id="cb53-32"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb53-32"></a>                    <span class="dt">sep =</span> <span class="st">&quot;&quot;</span>)</span>
<span id="cb53-33"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb53-33"></a>  <span class="kw">ggsave</span>(<span class="dt">file =</span> filename, <span class="dt">plot =</span> p, <span class="dt">dpi =</span> <span class="dv">300</span>, <span class="dt">width =</span> <span class="dv">4</span>, <span class="dt">height =</span> <span class="fl">4.2</span>)</span>
<span id="cb53-34"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb53-34"></a>}</span></code></pre></div>
<p><img src="figures/figure_11_0050.png" width="150px">
<img src="figures/figure_12_0050.png" width="150px">
<img src="figures/figure_13_0050.png" width="150px">
<img src="figures/figure_14_0050.png" width="150px">
<img src="figures/figure_15_0050.png" width="150px">
<img src="figures/figure_16_0050.png" width="150px"></p>
<p><img src="figures/figure_11_0051.png" width="150px">
<img src="figures/figure_12_0051.png" width="150px">
<img src="figures/figure_13_0051.png" width="150px">
<img src="figures/figure_14_0051.png" width="150px">
<img src="figures/figure_15_0051.png" width="150px">
<img src="figures/figure_16_0051.png" width="150px"></p>
<p>Perform k-means clusterings and compute several clustering validity scores by
using <code>NbClust()</code>.</p>
<div class="sourceCode" id="cb54"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb54-1"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb54-1"></a>compute_nbclust &lt;-<span class="st"> </span><span class="cf">function</span>(obj, plot_type){</span>
<span id="cb54-2"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb54-2"></a>  <span class="cf">if</span>(plot_type <span class="op">==</span><span class="st"> &quot;umap&quot;</span>){</span>
<span id="cb54-3"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb54-3"></a>    mat &lt;-<span class="st"> </span>obj[[<span class="st">&quot;reduction&quot;</span>]][[<span class="st">&quot;umap&quot;</span>]][[<span class="st">&quot;layout&quot;</span>]]</span>
<span id="cb54-4"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb54-4"></a>  }<span class="cf">else</span> <span class="cf">if</span>(plot_type <span class="op">==</span><span class="st"> &quot;tsne&quot;</span>){</span>
<span id="cb54-5"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb54-5"></a>    mat &lt;-<span class="st"> </span>obj[[<span class="st">&quot;reduction&quot;</span>]][[<span class="st">&quot;tsne&quot;</span>]][[<span class="st">&quot;Y&quot;</span>]]</span>
<span id="cb54-6"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb54-6"></a>  }</span>
<span id="cb54-7"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb54-7"></a>  <span class="kw">set.seed</span>(<span class="dv">8</span>)</span>
<span id="cb54-8"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb54-8"></a>  res &lt;-<span class="st"> </span><span class="kw">NbClust</span>(mat, <span class="dt">distance =</span> <span class="st">&quot;euclidean&quot;</span>, <span class="dt">min.nc =</span> <span class="dv">2</span>, <span class="dt">max.nc =</span> <span class="dv">7</span>,</span>
<span id="cb54-9"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb54-9"></a>                 <span class="dt">method =</span> <span class="st">&quot;kmeans&quot;</span>, <span class="dt">index =</span> <span class="st">&quot;all&quot;</span>)</span>
<span id="cb54-10"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb54-10"></a>  <span class="kw">return</span>(res)</span>
<span id="cb54-11"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb54-11"></a>}</span>
<span id="cb54-12"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb54-12"></a></span>
<span id="cb54-13"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb54-13"></a>data &lt;-<span class="st"> </span><span class="kw">list</span>(day1_norm, day7_hypo, sc68_vehi, sc68_cisp, pbmc_<span class="dv">4000</span>, pbmc_<span class="dv">6000</span>)</span>
<span id="cb54-14"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb54-14"></a>names &lt;-<span class="st"> </span><span class="kw">c</span>(<span class="st">&quot;day1_norm&quot;</span>, <span class="st">&quot;day7_hypo&quot;</span>, <span class="st">&quot;sc68_vehi&quot;</span>, <span class="st">&quot;sc68_cisp&quot;</span>, <span class="st">&quot;pbmc_4000&quot;</span>,</span>
<span id="cb54-15"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb54-15"></a>           <span class="st">&quot;pbmc_6000&quot;</span>)</span>
<span id="cb54-16"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb54-16"></a>nbclust_umap &lt;-<span class="st"> </span><span class="kw">list</span>()</span>
<span id="cb54-17"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb54-17"></a>nbclust_tsne &lt;-<span class="st"> </span><span class="kw">list</span>()</span>
<span id="cb54-18"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb54-18"></a><span class="cf">for</span>(i <span class="cf">in</span> <span class="dv">1</span><span class="op">:</span><span class="kw">length</span>(data)){</span>
<span id="cb54-19"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb54-19"></a>  nbclust_umap[[i]] &lt;-<span class="st"> </span><span class="kw">compute_nbclust</span>(<span class="dt">obj =</span> data[[i]], <span class="dt">plot_type =</span> <span class="st">&quot;umap&quot;</span>)</span>
<span id="cb54-20"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb54-20"></a>}</span>
<span id="cb54-21"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb54-21"></a><span class="cf">for</span>(i <span class="cf">in</span> <span class="dv">1</span><span class="op">:</span><span class="kw">length</span>(data)){</span>
<span id="cb54-22"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb54-22"></a>  nbclust_tsne[[i]] &lt;-<span class="st"> </span><span class="kw">compute_nbclust</span>(<span class="dt">obj =</span> data[[i]], <span class="dt">plot_type =</span> <span class="st">&quot;tsne&quot;</span>)</span>
<span id="cb54-23"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb54-23"></a>}</span>
<span id="cb54-24"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb54-24"></a><span class="kw">names</span>(nbclust_umap) &lt;-<span class="st"> </span>names</span>
<span id="cb54-25"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb54-25"></a><span class="kw">names</span>(nbclust_tsne) &lt;-<span class="st"> </span>names</span></code></pre></div>
<p>Save the objects.</p>
<div class="sourceCode" id="cb55"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb55-1"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb55-1"></a><span class="kw">saveRDS</span>(nbclust_umap, <span class="dt">file =</span> <span class="st">&quot;backup/50_001_nbclust_umap_pca.rds&quot;</span>)</span>
<span id="cb55-2"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb55-2"></a><span class="kw">saveRDS</span>(nbclust_tsne, <span class="dt">file =</span> <span class="st">&quot;backup/50_002_nbclust_tsne_pca.rds&quot;</span>)</span>
<span id="cb55-3"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb55-3"></a><span class="co"># ----------------------------------------</span></span>
<span id="cb55-4"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb55-4"></a><span class="co"># Output an excel file</span></span>
<span id="cb55-5"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb55-5"></a><span class="co"># ----------------------------------------</span></span>
<span id="cb55-6"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb55-6"></a>dnames &lt;-<span class="st"> </span><span class="kw">c</span>(<span class="st">&quot;day1_norm&quot;</span>, <span class="st">&quot;day7_hypo&quot;</span>, <span class="st">&quot;sc68_vehi&quot;</span>, <span class="st">&quot;sc68_cisp&quot;</span>,</span>
<span id="cb55-7"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb55-7"></a>            <span class="st">&quot;pbmc_4000&quot;</span>, <span class="st">&quot;pbmc_6000&quot;</span>)</span>
<span id="cb55-8"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb55-8"></a>dlist &lt;-<span class="st"> </span><span class="kw">list</span>()</span>
<span id="cb55-9"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb55-9"></a><span class="cf">for</span>(i <span class="cf">in</span> <span class="dv">1</span><span class="op">:</span><span class="kw">length</span>(dnames)){</span>
<span id="cb55-10"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb55-10"></a>  res &lt;-<span class="st"> </span><span class="kw">rbind</span>(nbclust_umap[[dnames[i]]][[<span class="st">&quot;All.index&quot;</span>]], <span class="st">&quot;&quot;</span>)</span>
<span id="cb55-11"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb55-11"></a>  res &lt;-<span class="st"> </span><span class="kw">rbind</span>(res, nbclust_umap[[dnames[i]]][[<span class="st">&quot;Best.nc&quot;</span>]])</span>
<span id="cb55-12"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb55-12"></a>  dlist &lt;-<span class="st"> </span><span class="kw">c</span>(dlist, <span class="kw">list</span>(res))</span>
<span id="cb55-13"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb55-13"></a>}</span>
<span id="cb55-14"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb55-14"></a><span class="kw">names</span>(dlist) &lt;-<span class="st"> </span>dnames</span>
<span id="cb55-15"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb55-15"></a>filename &lt;-<span class="st"> &quot;backup/nbclust_umap_pca.xlsx&quot;</span></span>
<span id="cb55-16"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb55-16"></a><span class="kw">write.xlsx</span>(dlist, <span class="dt">file =</span> filename, <span class="dt">row.names =</span> <span class="ot">TRUE</span>)</span></code></pre></div>
<p>Load the objects.</p>
<div class="sourceCode" id="cb56"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb56-1"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb56-1"></a>nbclust_umap &lt;-<span class="st"> </span><span class="kw">readRDS</span>(<span class="dt">file =</span> <span class="st">&quot;backup/50_001_nbclust_umap_pca.rds&quot;</span>)</span>
<span id="cb56-2"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb56-2"></a>nbclust_tsne &lt;-<span class="st"> </span><span class="kw">readRDS</span>(<span class="dt">file =</span> <span class="st">&quot;backup/50_002_nbclust_tsne_pca.rds&quot;</span>)</span></code></pre></div>
</div>
<div id="asurat" class="section level3">
<h3><span class="header-section-number">14.3.5</span> ASURAT</h3>
<p>Note again that ASURAT can cluster cells from the viewpoints of cell type
(by using DO and CO) and biological functions (by using GO, etc.).
In this section, we choose proper databases for producing the greater ASWs.</p>
<p>Load the data.</p>
<div class="sourceCode" id="cb57"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb57-1"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb57-1"></a><span class="kw">rm</span>(<span class="dt">list =</span> <span class="kw">ls</span>())</span>
<span id="cb57-2"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb57-2"></a><span class="kw">source</span>(<span class="st">&quot;R/function_general.R&quot;</span>)</span>
<span id="cb57-3"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb57-3"></a><span class="kw">source</span>(<span class="st">&quot;R/function_sign.R&quot;</span>)</span>
<span id="cb57-4"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb57-4"></a><span class="kw">source</span>(<span class="st">&quot;R/plot.R&quot;</span>)</span>
<span id="cb57-5"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb57-5"></a>day1_norm &lt;-<span class="st"> </span><span class="kw">readRDS</span>(<span class="dt">file =</span> <span class="st">&quot;backup/01_500_day1_norm_all.rds&quot;</span>)</span>
<span id="cb57-6"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb57-6"></a>day7_hypo &lt;-<span class="st"> </span><span class="kw">readRDS</span>(<span class="dt">file =</span> <span class="st">&quot;backup/02_500_day7_hypo_all.rds&quot;</span>)</span>
<span id="cb57-7"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb57-7"></a>sc68_vehi &lt;-<span class="st"> </span><span class="kw">readRDS</span>(<span class="dt">file =</span> <span class="st">&quot;backup/03_005_sc68_vehi_normalized.rds&quot;</span>)</span>
<span id="cb57-8"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb57-8"></a>sc68_cisp &lt;-<span class="st"> </span><span class="kw">readRDS</span>(<span class="dt">file =</span> <span class="st">&quot;backup/04_005_sc68_cisp_normalized.rds&quot;</span>)</span>
<span id="cb57-9"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb57-9"></a>pbmc_<span class="dv">4000</span> &lt;-<span class="st"> </span><span class="kw">readRDS</span>(<span class="dt">file =</span> <span class="st">&quot;backup/05_500_pbmc_4000_all.rds&quot;</span>)</span>
<span id="cb57-10"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb57-10"></a>pbmc_<span class="dv">6000</span> &lt;-<span class="st"> </span><span class="kw">readRDS</span>(<span class="dt">file =</span> <span class="st">&quot;backup/06_500_pbmc_6000_all.rds&quot;</span>)</span>
<span id="cb57-11"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb57-11"></a>sc68_vehi_bl &lt;-<span class="st"> </span><span class="kw">readRDS</span>(<span class="dt">file =</span> <span class="st">&quot;backup/13_001_sc68_vehi_blacklist.rds&quot;</span>)</span></code></pre></div>
<p>Remove the outlier cells from <code>sc68_vehi</code>.</p>
<div class="sourceCode" id="cb58"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb58-1"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb58-1"></a><span class="co"># ----------------------------------------</span></span>
<span id="cb58-2"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb58-2"></a><span class="co"># For `sc68_vehi`</span></span>
<span id="cb58-3"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb58-3"></a><span class="co"># ----------------------------------------</span></span>
<span id="cb58-4"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb58-4"></a>mat &lt;-<span class="st"> </span>sc68_vehi[[<span class="st">&quot;data&quot;</span>]][[<span class="st">&quot;bayNorm&quot;</span>]][[<span class="st">&quot;Bay_out&quot;</span>]]</span>
<span id="cb58-5"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb58-5"></a>inds &lt;-<span class="st"> </span><span class="kw">which</span>(<span class="kw">colnames</span>(mat) <span class="op">%in%</span><span class="st"> </span>sc68_vehi_bl)</span>
<span id="cb58-6"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb58-6"></a>sc68_vehi[[<span class="st">&quot;data&quot;</span>]][[<span class="st">&quot;bayNorm&quot;</span>]][[<span class="st">&quot;Bay_out&quot;</span>]] &lt;-<span class="st"> </span>mat[, <span class="op">-</span>inds]</span>
<span id="cb58-7"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb58-7"></a></span>
<span id="cb58-8"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb58-8"></a>mat &lt;-<span class="st"> </span>sc68_vehi[[<span class="st">&quot;data&quot;</span>]][[<span class="st">&quot;log1p&quot;</span>]]</span>
<span id="cb58-9"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb58-9"></a>sc68_vehi[[<span class="st">&quot;data&quot;</span>]][[<span class="st">&quot;log1p&quot;</span>]] &lt;-<span class="st"> </span>mat[, <span class="op">-</span>inds]</span>
<span id="cb58-10"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb58-10"></a></span>
<span id="cb58-11"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb58-11"></a>mat &lt;-<span class="st"> </span>sc68_vehi[[<span class="st">&quot;data&quot;</span>]][[<span class="st">&quot;centered&quot;</span>]]</span>
<span id="cb58-12"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb58-12"></a>sc68_vehi[[<span class="st">&quot;data&quot;</span>]][[<span class="st">&quot;centered&quot;</span>]] &lt;-<span class="st"> </span>mat[, <span class="op">-</span>inds]</span>
<span id="cb58-13"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb58-13"></a>sc68_vehi[[<span class="st">&quot;sample&quot;</span>]] &lt;-<span class="st"> </span>sc68_vehi[[<span class="st">&quot;sample&quot;</span>]][<span class="op">-</span>inds,]</span></code></pre></div>
<p>Compute the correlation matrix for <code>sc68_vehi</code> and save the result.</p>
<div class="sourceCode" id="cb59"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb59-1"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb59-1"></a>do_cor_variables &lt;-<span class="st"> </span><span class="cf">function</span>(obj, method){</span>
<span id="cb59-2"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb59-2"></a>  res &lt;-<span class="st"> </span><span class="kw">list</span>()</span>
<span id="cb59-3"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb59-3"></a>  tmat &lt;-<span class="st"> </span><span class="kw">t</span>(obj[[<span class="st">&quot;data&quot;</span>]][[<span class="st">&quot;log1p&quot;</span>]])</span>
<span id="cb59-4"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb59-4"></a>  <span class="cf">for</span>(m <span class="cf">in</span> method){</span>
<span id="cb59-5"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb59-5"></a>    res &lt;-<span class="st"> </span><span class="kw">c</span>(res, <span class="kw">list</span>(<span class="kw">cor</span>(tmat, <span class="dt">method =</span> m)))</span>
<span id="cb59-6"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb59-6"></a>  }</span>
<span id="cb59-7"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb59-7"></a>  <span class="kw">names</span>(res) &lt;-<span class="st"> </span>method</span>
<span id="cb59-8"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb59-8"></a>  <span class="kw">return</span>(res)</span>
<span id="cb59-9"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb59-9"></a>}</span>
<span id="cb59-10"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb59-10"></a>sc68_vehi_cor &lt;-<span class="st"> </span><span class="kw">do_cor_variables</span>(<span class="dt">obj =</span> sc68_vehi, <span class="dt">method =</span> <span class="kw">c</span>(<span class="st">&quot;spearman&quot;</span>))</span>
<span id="cb59-11"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb59-11"></a></span>
<span id="cb59-12"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb59-12"></a><span class="co"># Save the correlation matrix.</span></span>
<span id="cb59-13"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb59-13"></a><span class="kw">saveRDS</span>(sc68_vehi_cor, <span class="dt">file =</span> <span class="st">&quot;backup/13_002_sc68_vehi_correlation_wo_bl.rds&quot;</span>)</span>
<span id="cb59-14"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb59-14"></a></span>
<span id="cb59-15"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb59-15"></a><span class="co"># Load the correlation matrix.</span></span>
<span id="cb59-16"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb59-16"></a>sc68_vehi_cor &lt;-<span class="st"> </span><span class="kw">readRDS</span>(<span class="dt">file =</span> <span class="st">&quot;backup/13_002_sc68_vehi_correlation_wo_bl.rds&quot;</span>)</span></code></pre></div>
<p>Below are the R codes to obtain the low-dimensional representations for
<code>sc68_vehi</code> and <code>sc68_cisp</code>.
Please see Chapter 8 “ASURAT using Disease Ontology DB” for the details.</p>
<div class="sourceCode" id="cb60"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb60-1"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb60-1"></a>sc68_vehi_cor &lt;-<span class="st"> </span><span class="kw">readRDS</span>(<span class="dt">file =</span> <span class="st">&quot;backup/13_002_sc68_vehi_correlation_wo_bl.rds&quot;</span>)</span>
<span id="cb60-2"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb60-2"></a>sc68_cisp_cor &lt;-<span class="st"> </span><span class="kw">readRDS</span>(<span class="dt">file =</span> <span class="st">&quot;backup/04_006_sc68_cisp_correlation.rds&quot;</span>)</span>
<span id="cb60-3"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb60-3"></a>tidy_DO &lt;-<span class="st"> </span><span class="kw">readRDS</span>(<span class="dt">file =</span> <span class="st">&quot;data/2020_001_databases/20201213_tidy_DO_human.rds&quot;</span>)</span>
<span id="cb60-4"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb60-4"></a>IC_DO &lt;-<span class="st"> </span><span class="kw">readRDS</span>(<span class="dt">file =</span> <span class="st">&quot;data/2020_001_databases/20201213_IC_DO_human.rds&quot;</span>)</span>
<span id="cb60-5"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb60-5"></a>sc68_vehi[[<span class="st">&quot;sign&quot;</span>]][[<span class="st">&quot;DO&quot;</span>]] &lt;-<span class="st"> </span>tidy_DO</span>
<span id="cb60-6"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb60-6"></a>sc68_cisp[[<span class="st">&quot;sign&quot;</span>]][[<span class="st">&quot;DO&quot;</span>]] &lt;-<span class="st"> </span>tidy_DO</span>
<span id="cb60-7"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb60-7"></a>sc68_vehi &lt;-<span class="st"> </span><span class="kw">do_quickQC_sign</span>(<span class="dt">obj =</span> sc68_vehi, <span class="dt">data_type =</span> <span class="st">&quot;DO&quot;</span>,</span>
<span id="cb60-8"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb60-8"></a>                             <span class="dt">min_ngenes =</span> <span class="dv">2</span>, <span class="dt">max_ngenes =</span> <span class="dv">1000</span>)</span>
<span id="cb60-9"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb60-9"></a>sc68_cisp &lt;-<span class="st"> </span><span class="kw">do_quickQC_sign</span>(<span class="dt">obj =</span> sc68_cisp, <span class="dt">data_type =</span> <span class="st">&quot;DO&quot;</span>,</span>
<span id="cb60-10"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb60-10"></a>                             <span class="dt">min_ngenes =</span> <span class="dv">2</span>, <span class="dt">max_ngenes =</span> <span class="dv">1000</span>)</span>
<span id="cb60-11"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb60-11"></a>sc68_vehi &lt;-<span class="st"> </span><span class="kw">separate_variables_sign</span>(<span class="dt">obj =</span> sc68_vehi, <span class="dt">obj_cor =</span> sc68_vehi_cor,</span>
<span id="cb60-12"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb60-12"></a>                                     <span class="dt">data_type =</span> <span class="st">&quot;DO&quot;</span>, <span class="dt">method =</span> <span class="st">&quot;spearman&quot;</span>,</span>
<span id="cb60-13"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb60-13"></a>                                     <span class="dt">th_posi =</span> <span class="fl">0.36</span>, <span class="dt">th_nega =</span> <span class="fl">-0.22</span>)</span>
<span id="cb60-14"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb60-14"></a>sc68_cisp &lt;-<span class="st"> </span><span class="kw">separate_variables_sign</span>(<span class="dt">obj =</span> sc68_cisp, <span class="dt">obj_cor =</span> sc68_cisp_cor,</span>
<span id="cb60-15"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb60-15"></a>                                     <span class="dt">data_type =</span> <span class="st">&quot;DO&quot;</span>, <span class="dt">method =</span> <span class="st">&quot;spearman&quot;</span>,</span>
<span id="cb60-16"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb60-16"></a>                                     <span class="dt">th_posi =</span> <span class="fl">0.38</span>, <span class="dt">th_nega =</span> <span class="fl">-0.26</span>)</span>
<span id="cb60-17"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb60-17"></a>sc68_vehi &lt;-<span class="st"> </span><span class="kw">select_sign</span>(<span class="dt">obj =</span> sc68_vehi, <span class="dt">data_type =</span> <span class="st">&quot;DO&quot;</span>,</span>
<span id="cb60-18"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb60-18"></a>                         <span class="dt">min_cnt =</span> <span class="dv">2</span>, <span class="dt">min_cnt_weak =</span> <span class="dv">2</span>)</span>
<span id="cb60-19"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb60-19"></a>sc68_cisp &lt;-<span class="st"> </span><span class="kw">select_sign</span>(<span class="dt">obj =</span> sc68_cisp, <span class="dt">data_type =</span> <span class="st">&quot;DO&quot;</span>,</span>
<span id="cb60-20"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb60-20"></a>                         <span class="dt">min_cnt =</span> <span class="dv">2</span>, <span class="dt">min_cnt_weak =</span> <span class="dv">2</span>)</span>
<span id="cb60-21"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb60-21"></a>sc68_vehi &lt;-<span class="st"> </span><span class="kw">compute_SemSim_sign</span>(<span class="dt">obj =</span> sc68_vehi, <span class="dt">data_type =</span> <span class="st">&quot;DO&quot;</span>,</span>
<span id="cb60-22"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb60-22"></a>                                 <span class="dt">measure =</span> <span class="st">&quot;Jiang&quot;</span>, <span class="dt">orgdb =</span> <span class="ot">NULL</span>,</span>
<span id="cb60-23"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb60-23"></a>                                 <span class="dt">treeTable =</span> <span class="ot">NULL</span>, <span class="dt">IC =</span> IC_DO)</span>
<span id="cb60-24"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb60-24"></a>sc68_cisp &lt;-<span class="st"> </span><span class="kw">compute_SemSim_sign</span>(<span class="dt">obj =</span> sc68_cisp, <span class="dt">data_type =</span> <span class="st">&quot;DO&quot;</span>,</span>
<span id="cb60-25"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb60-25"></a>                                 <span class="dt">measure =</span> <span class="st">&quot;Jiang&quot;</span>, <span class="dt">orgdb =</span> <span class="ot">NULL</span>,</span>
<span id="cb60-26"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb60-26"></a>                                 <span class="dt">treeTable =</span> <span class="ot">NULL</span>, <span class="dt">IC =</span> IC_DO)</span>
<span id="cb60-27"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb60-27"></a>sc68_vehi &lt;-<span class="st"> </span><span class="kw">reduce_sign</span>(<span class="dt">obj =</span> sc68_vehi, <span class="dt">data_type =</span> <span class="st">&quot;DO&quot;</span>,</span>
<span id="cb60-28"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb60-28"></a>                         <span class="dt">threshold =</span> <span class="fl">0.80</span>, <span class="dt">keep_rareID =</span> <span class="ot">TRUE</span>)</span>
<span id="cb60-29"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb60-29"></a>sc68_cisp &lt;-<span class="st"> </span><span class="kw">reduce_sign</span>(<span class="dt">obj =</span> sc68_cisp, <span class="dt">data_type =</span> <span class="st">&quot;DO&quot;</span>,</span>
<span id="cb60-30"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb60-30"></a>                         <span class="dt">threshold =</span> <span class="fl">0.80</span>, <span class="dt">keep_rareID =</span> <span class="ot">TRUE</span>)</span>
<span id="cb60-31"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb60-31"></a>sc68_vehi &lt;-<span class="st"> </span><span class="kw">make_signxsample_matrix</span>(<span class="dt">obj =</span> sc68_vehi, <span class="dt">data_type =</span> <span class="st">&quot;DO&quot;</span>,</span>
<span id="cb60-32"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb60-32"></a>                                     <span class="dt">weight_strg =</span> <span class="fl">0.5</span>, <span class="dt">weight_vari =</span> <span class="fl">0.5</span>)</span>
<span id="cb60-33"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb60-33"></a>sc68_cisp &lt;-<span class="st"> </span><span class="kw">make_signxsample_matrix</span>(<span class="dt">obj =</span> sc68_cisp, <span class="dt">data_type =</span> <span class="st">&quot;DO&quot;</span>,</span>
<span id="cb60-34"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb60-34"></a>                                     <span class="dt">weight_strg =</span> <span class="fl">0.5</span>, <span class="dt">weight_vari =</span> <span class="fl">0.5</span>)</span>
<span id="cb60-35"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb60-35"></a>sc68_vehi &lt;-<span class="st"> </span><span class="kw">do_tsne_sign</span>(<span class="dt">obj =</span> sc68_vehi, <span class="dt">data_type =</span> <span class="st">&quot;DO&quot;</span>,</span>
<span id="cb60-36"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb60-36"></a>                          <span class="dt">category =</span> <span class="st">&quot;disease&quot;</span>, <span class="dt">pca_dim =</span> <span class="ot">NULL</span>, <span class="dt">tsne_dim =</span> <span class="dv">2</span>)</span>
<span id="cb60-37"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb60-37"></a>sc68_cisp &lt;-<span class="st"> </span><span class="kw">do_tsne_sign</span>(<span class="dt">obj =</span> sc68_cisp, <span class="dt">data_type =</span> <span class="st">&quot;DO&quot;</span>,</span>
<span id="cb60-38"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb60-38"></a>                          <span class="dt">category =</span> <span class="st">&quot;disease&quot;</span>, <span class="dt">pca_dim =</span> <span class="ot">NULL</span>, <span class="dt">tsne_dim =</span> <span class="dv">2</span>)</span></code></pre></div>
<p>Run <code>umap</code> for all the data.</p>
<div class="sourceCode" id="cb61"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb61-1"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb61-1"></a>day1_norm &lt;-<span class="st"> </span><span class="kw">do_umap_sign</span>(<span class="dt">obj =</span> day1_norm, <span class="dt">data_type =</span> <span class="st">&quot;GO&quot;</span>,</span>
<span id="cb61-2"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb61-2"></a>                          <span class="dt">category =</span> <span class="st">&quot;BP&quot;</span>, <span class="dt">pca_dim =</span> <span class="ot">NULL</span>, <span class="dt">umap_dim =</span> <span class="dv">2</span>)</span>
<span id="cb61-3"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb61-3"></a>day7_hypo &lt;-<span class="st"> </span><span class="kw">do_umap_sign</span>(<span class="dt">obj =</span> day7_hypo, <span class="dt">data_type =</span> <span class="st">&quot;GO&quot;</span>,</span>
<span id="cb61-4"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb61-4"></a>                          <span class="dt">category =</span> <span class="st">&quot;BP&quot;</span>, <span class="dt">pca_dim =</span> <span class="ot">NULL</span>, <span class="dt">umap_dim =</span> <span class="dv">2</span>)</span>
<span id="cb61-5"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb61-5"></a>sc68_vehi &lt;-<span class="st"> </span><span class="kw">do_umap_sign</span>(<span class="dt">obj =</span> sc68_vehi, <span class="dt">data_type =</span> <span class="st">&quot;DO&quot;</span>,</span>
<span id="cb61-6"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb61-6"></a>                          <span class="dt">category =</span> <span class="st">&quot;disease&quot;</span>, <span class="dt">pca_dim =</span> <span class="ot">NULL</span>, <span class="dt">umap_dim =</span> <span class="dv">2</span>)</span>
<span id="cb61-7"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb61-7"></a>sc68_cisp &lt;-<span class="st"> </span><span class="kw">do_umap_sign</span>(<span class="dt">obj =</span> sc68_cisp, <span class="dt">data_type =</span> <span class="st">&quot;DO&quot;</span>,</span>
<span id="cb61-8"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb61-8"></a>                          <span class="dt">category =</span> <span class="st">&quot;disease&quot;</span>, <span class="dt">pca_dim =</span> <span class="ot">NULL</span>, <span class="dt">umap_dim =</span> <span class="dv">2</span>)</span>
<span id="cb61-9"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb61-9"></a>pbmc_<span class="dv">4000</span> &lt;-<span class="st"> </span><span class="kw">do_umap_sign</span>(<span class="dt">obj =</span> pbmc_<span class="dv">4000</span>, <span class="dt">data_type =</span> <span class="st">&quot;CO&quot;</span>,</span>
<span id="cb61-10"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb61-10"></a>                          <span class="dt">category =</span> <span class="st">&quot;cell&quot;</span>, <span class="dt">pca_dim =</span> <span class="ot">NULL</span>, <span class="dt">umap_dim =</span> <span class="dv">2</span>)</span>
<span id="cb61-11"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb61-11"></a>pbmc_<span class="dv">6000</span> &lt;-<span class="st"> </span><span class="kw">do_umap_sign</span>(<span class="dt">obj =</span> pbmc_<span class="dv">6000</span>, <span class="dt">data_type =</span> <span class="st">&quot;GO&quot;</span>,</span>
<span id="cb61-12"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb61-12"></a>                          <span class="dt">category =</span> <span class="st">&quot;BP&quot;</span>, <span class="dt">pca_dim =</span> <span class="ot">NULL</span>, <span class="dt">umap_dim =</span> <span class="dv">2</span>)</span></code></pre></div>
<p>Visualize the results.</p>
<div class="sourceCode" id="cb62"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb62-1"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb62-1"></a><span class="co"># ----------------------------------------</span></span>
<span id="cb62-2"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb62-2"></a><span class="co"># UMAP</span></span>
<span id="cb62-3"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb62-3"></a><span class="co"># ----------------------------------------</span></span>
<span id="cb62-4"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb62-4"></a>p &lt;-<span class="st"> </span><span class="kw">plot_umap_sign</span>(<span class="dt">obj =</span> day1_norm, <span class="dt">data_type =</span> <span class="st">&quot;GO&quot;</span>, <span class="dt">category =</span> <span class="st">&quot;BP&quot;</span>,</span>
<span id="cb62-5"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb62-5"></a>                    <span class="dt">algo_name =</span> <span class="ot">NULL</span>, <span class="dt">theta =</span> <span class="ot">NULL</span>, <span class="dt">phi =</span> <span class="ot">NULL</span>,</span>
<span id="cb62-6"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb62-6"></a>                    <span class="dt">title =</span> <span class="st">&quot;day1_norm (GO: BP)&quot;</span>, <span class="dt">title_size =</span> <span class="dv">16</span>,</span>
<span id="cb62-7"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb62-7"></a>                    <span class="dt">xlabel =</span> <span class="st">&quot;UMAP_1&quot;</span>, <span class="dt">ylabel =</span> <span class="st">&quot;UMAP_2&quot;</span>, <span class="dt">zlabel =</span> <span class="ot">NULL</span>,</span>
<span id="cb62-8"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb62-8"></a>                    <span class="dt">default_color =</span> <span class="ot">FALSE</span>)</span>
<span id="cb62-9"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb62-9"></a>filename &lt;-<span class="st"> &quot;figures/figure_11_0060.png&quot;</span></span>
<span id="cb62-10"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb62-10"></a><span class="kw">ggsave</span>(<span class="dt">file =</span> filename, <span class="dt">plot =</span> p, <span class="dt">dpi =</span> <span class="dv">300</span>, <span class="dt">width =</span> <span class="dv">4</span>, <span class="dt">height =</span> <span class="fl">4.2</span>)</span></code></pre></div>
<p><img src="figures/figure_11_0060.png" width="150px">
<img src="figures/figure_12_0060.png" width="150px">
<img src="figures/figure_13_0060.png" width="150px">
<img src="figures/figure_14_0060.png" width="150px">
<img src="figures/figure_15_0060.png" width="150px">
<img src="figures/figure_16_0060.png" width="150px"></p>
<p><img src="figures/figure_11_0061.png" width="150px">
<img src="figures/figure_12_0061.png" width="150px">
<img src="figures/figure_13_0061.png" width="150px">
<img src="figures/figure_14_0061.png" width="150px">
<img src="figures/figure_15_0061.png" width="150px">
<img src="figures/figure_16_0061.png" width="150px"></p>
<p>Save the objects.</p>
<div class="sourceCode" id="cb63"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb63-1"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb63-1"></a><span class="kw">saveRDS</span>(day1_norm, <span class="dt">file =</span> <span class="st">&quot;backup/11_060_day1_norm_asurat.rds&quot;</span>)</span>
<span id="cb63-2"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb63-2"></a><span class="kw">saveRDS</span>(day7_hypo, <span class="dt">file =</span> <span class="st">&quot;backup/12_060_day7_hypo_asurat.rds&quot;</span>)</span>
<span id="cb63-3"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb63-3"></a><span class="kw">saveRDS</span>(sc68_vehi, <span class="dt">file =</span> <span class="st">&quot;backup/13_060_sc68_vehi_asurat.rds&quot;</span>)</span>
<span id="cb63-4"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb63-4"></a><span class="kw">saveRDS</span>(sc68_cisp, <span class="dt">file =</span> <span class="st">&quot;backup/14_060_sc68_cisp_asurat.rds&quot;</span>)</span>
<span id="cb63-5"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb63-5"></a><span class="kw">saveRDS</span>(pbmc_<span class="dv">4000</span>, <span class="dt">file =</span> <span class="st">&quot;backup/15_060_pbmc_4000_asurat.rds&quot;</span>)</span>
<span id="cb63-6"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb63-6"></a><span class="kw">saveRDS</span>(pbmc_<span class="dv">6000</span>, <span class="dt">file =</span> <span class="st">&quot;backup/16_060_pbmc_6000_asurat.rds&quot;</span>)</span></code></pre></div>
<p>Load the objects.</p>
<div class="sourceCode" id="cb64"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb64-1"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb64-1"></a>day1_norm &lt;-<span class="st"> </span><span class="kw">readRDS</span>(<span class="dt">file =</span> <span class="st">&quot;backup/11_060_day1_norm_asurat.rds&quot;</span>)</span>
<span id="cb64-2"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb64-2"></a>day7_hypo &lt;-<span class="st"> </span><span class="kw">readRDS</span>(<span class="dt">file =</span> <span class="st">&quot;backup/12_060_day7_hypo_asurat.rds&quot;</span>)</span>
<span id="cb64-3"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb64-3"></a>sc68_vehi &lt;-<span class="st"> </span><span class="kw">readRDS</span>(<span class="dt">file =</span> <span class="st">&quot;backup/13_060_sc68_vehi_asurat.rds&quot;</span>)</span>
<span id="cb64-4"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb64-4"></a>sc68_cisp &lt;-<span class="st"> </span><span class="kw">readRDS</span>(<span class="dt">file =</span> <span class="st">&quot;backup/14_060_sc68_cisp_asurat.rds&quot;</span>)</span>
<span id="cb64-5"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb64-5"></a>pbmc_<span class="dv">4000</span> &lt;-<span class="st"> </span><span class="kw">readRDS</span>(<span class="dt">file =</span> <span class="st">&quot;backup/15_060_pbmc_4000_asurat.rds&quot;</span>)</span>
<span id="cb64-6"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb64-6"></a>pbmc_<span class="dv">6000</span> &lt;-<span class="st"> </span><span class="kw">readRDS</span>(<span class="dt">file =</span> <span class="st">&quot;backup/16_060_pbmc_6000_asurat.rds&quot;</span>)</span></code></pre></div>
<p>Perform k-means clusterings and compute several clustering validity scores by
using <code>NbClust()</code>.</p>
<div class="sourceCode" id="cb65"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb65-1"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb65-1"></a>compute_nbclust &lt;-<span class="st"> </span><span class="cf">function</span>(obj, plot_type, data_type, category){</span>
<span id="cb65-2"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb65-2"></a>  <span class="cf">if</span>(plot_type <span class="op">==</span><span class="st"> &quot;umap&quot;</span>){</span>
<span id="cb65-3"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb65-3"></a>    mat &lt;-<span class="st"> </span>obj[[<span class="st">&quot;reduction&quot;</span>]][[<span class="st">&quot;umap&quot;</span>]][[data_type]][[category]][[<span class="st">&quot;layout&quot;</span>]]</span>
<span id="cb65-4"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb65-4"></a>  }<span class="cf">else</span> <span class="cf">if</span>(plot_type <span class="op">==</span><span class="st"> &quot;tsne&quot;</span>){</span>
<span id="cb65-5"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb65-5"></a>    mat &lt;-<span class="st"> </span>obj[[<span class="st">&quot;reduction&quot;</span>]][[<span class="st">&quot;tsne&quot;</span>]][[data_type]][[category]][[<span class="st">&quot;Y&quot;</span>]]</span>
<span id="cb65-6"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb65-6"></a>  }</span>
<span id="cb65-7"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb65-7"></a>  <span class="kw">set.seed</span>(<span class="dv">8</span>)</span>
<span id="cb65-8"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb65-8"></a>  res &lt;-<span class="st"> </span><span class="kw">NbClust</span>(mat, <span class="dt">distance =</span> <span class="st">&quot;euclidean&quot;</span>, <span class="dt">min.nc =</span> <span class="dv">2</span>, <span class="dt">max.nc =</span> <span class="dv">7</span>,</span>
<span id="cb65-9"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb65-9"></a>                 <span class="dt">method =</span> <span class="st">&quot;kmeans&quot;</span>, <span class="dt">index =</span> <span class="st">&quot;all&quot;</span>)</span>
<span id="cb65-10"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb65-10"></a>  <span class="kw">return</span>(res)</span>
<span id="cb65-11"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb65-11"></a>}</span>
<span id="cb65-12"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb65-12"></a></span>
<span id="cb65-13"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb65-13"></a>data &lt;-<span class="st"> </span><span class="kw">list</span>(day1_norm, day7_hypo, sc68_vehi, sc68_cisp, pbmc_<span class="dv">4000</span>, pbmc_<span class="dv">6000</span>)</span>
<span id="cb65-14"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb65-14"></a>names &lt;-<span class="st"> </span><span class="kw">c</span>(<span class="st">&quot;day1_norm&quot;</span>, <span class="st">&quot;day7_hypo&quot;</span>, <span class="st">&quot;sc68_vehi&quot;</span>, <span class="st">&quot;sc68_cisp&quot;</span>, <span class="st">&quot;pbmc_4000&quot;</span>,</span>
<span id="cb65-15"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb65-15"></a>           <span class="st">&quot;pbmc_6000&quot;</span>)</span>
<span id="cb65-16"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb65-16"></a>data_types &lt;-<span class="st"> </span><span class="kw">c</span>(<span class="st">&quot;GO&quot;</span>, <span class="st">&quot;GO&quot;</span>, <span class="st">&quot;DO&quot;</span>, <span class="st">&quot;DO&quot;</span>, <span class="st">&quot;CO&quot;</span>, <span class="st">&quot;GO&quot;</span>)</span>
<span id="cb65-17"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb65-17"></a>categories &lt;-<span class="st"> </span><span class="kw">c</span>(<span class="st">&quot;BP&quot;</span>, <span class="st">&quot;BP&quot;</span>, <span class="st">&quot;disease&quot;</span>, <span class="st">&quot;disease&quot;</span>, <span class="st">&quot;cell&quot;</span>, <span class="st">&quot;BP&quot;</span>)</span>
<span id="cb65-18"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb65-18"></a>nbclust_umap &lt;-<span class="st"> </span><span class="kw">list</span>()</span>
<span id="cb65-19"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb65-19"></a>nbclust_tsne &lt;-<span class="st"> </span><span class="kw">list</span>()</span>
<span id="cb65-20"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb65-20"></a><span class="cf">for</span>(i <span class="cf">in</span> <span class="dv">1</span><span class="op">:</span><span class="kw">length</span>(data)){</span>
<span id="cb65-21"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb65-21"></a>  nbclust_umap[[i]] &lt;-<span class="st"> </span><span class="kw">compute_nbclust</span>(<span class="dt">obj =</span> data[[i]], <span class="dt">plot_type =</span> <span class="st">&quot;umap&quot;</span>,</span>
<span id="cb65-22"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb65-22"></a>                                       data_types[i], categories[i])</span>
<span id="cb65-23"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb65-23"></a>}</span>
<span id="cb65-24"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb65-24"></a><span class="cf">for</span>(i <span class="cf">in</span> <span class="dv">1</span><span class="op">:</span><span class="kw">length</span>(data)){</span>
<span id="cb65-25"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb65-25"></a>  nbclust_tsne[[i]] &lt;-<span class="st"> </span><span class="kw">compute_nbclust</span>(<span class="dt">obj =</span> data[[i]], <span class="dt">plot_type =</span> <span class="st">&quot;tsne&quot;</span>,</span>
<span id="cb65-26"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb65-26"></a>                                       data_types[i], categories[i])</span>
<span id="cb65-27"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb65-27"></a>}</span>
<span id="cb65-28"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb65-28"></a><span class="kw">names</span>(nbclust_umap) &lt;-<span class="st"> </span>names</span>
<span id="cb65-29"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb65-29"></a><span class="kw">names</span>(nbclust_tsne) &lt;-<span class="st"> </span>names</span></code></pre></div>
<p>Save the objects.</p>
<div class="sourceCode" id="cb66"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb66-1"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb66-1"></a><span class="kw">saveRDS</span>(nbclust_umap, <span class="dt">file =</span> <span class="st">&quot;backup/60_001_nbclust_umap_asurat.rds&quot;</span>)</span>
<span id="cb66-2"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb66-2"></a><span class="kw">saveRDS</span>(nbclust_tsne, <span class="dt">file =</span> <span class="st">&quot;backup/60_002_nbclust_tsne_asurat.rds&quot;</span>)</span>
<span id="cb66-3"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb66-3"></a><span class="co"># ----------------------------------------</span></span>
<span id="cb66-4"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb66-4"></a><span class="co"># Output an excel file</span></span>
<span id="cb66-5"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb66-5"></a><span class="co"># ----------------------------------------</span></span>
<span id="cb66-6"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb66-6"></a>dnames &lt;-<span class="st"> </span><span class="kw">c</span>(<span class="st">&quot;day1_norm&quot;</span>, <span class="st">&quot;day7_hypo&quot;</span>, <span class="st">&quot;sc68_vehi&quot;</span>, <span class="st">&quot;sc68_cisp&quot;</span>,</span>
<span id="cb66-7"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb66-7"></a>            <span class="st">&quot;pbmc_4000&quot;</span>, <span class="st">&quot;pbmc_6000&quot;</span>)</span>
<span id="cb66-8"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb66-8"></a>dlist &lt;-<span class="st"> </span><span class="kw">list</span>()</span>
<span id="cb66-9"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb66-9"></a><span class="cf">for</span>(i <span class="cf">in</span> <span class="dv">1</span><span class="op">:</span><span class="kw">length</span>(dnames)){</span>
<span id="cb66-10"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb66-10"></a>  res &lt;-<span class="st"> </span><span class="kw">rbind</span>(nbclust_umap[[dnames[i]]][[<span class="st">&quot;All.index&quot;</span>]], <span class="st">&quot;&quot;</span>)</span>
<span id="cb66-11"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb66-11"></a>  res &lt;-<span class="st"> </span><span class="kw">rbind</span>(res, nbclust_umap[[dnames[i]]][[<span class="st">&quot;Best.nc&quot;</span>]])</span>
<span id="cb66-12"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb66-12"></a>  dlist &lt;-<span class="st"> </span><span class="kw">c</span>(dlist, <span class="kw">list</span>(res))</span>
<span id="cb66-13"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb66-13"></a>}</span>
<span id="cb66-14"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb66-14"></a><span class="kw">names</span>(dlist) &lt;-<span class="st"> </span>dnames</span>
<span id="cb66-15"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb66-15"></a>filename &lt;-<span class="st"> &quot;backup/nbclust_umap_asurat.xlsx&quot;</span></span>
<span id="cb66-16"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb66-16"></a><span class="kw">write.xlsx</span>(dlist, <span class="dt">file =</span> filename, <span class="dt">row.names =</span> <span class="ot">TRUE</span>)</span></code></pre></div>
<p>Load the objects.</p>
<div class="sourceCode" id="cb67"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb67-1"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb67-1"></a>nbclust_umap &lt;-<span class="st"> </span><span class="kw">readRDS</span>(<span class="dt">file =</span> <span class="st">&quot;backup/60_001_nbclust_umap_asurat.rds&quot;</span>)</span>
<span id="cb67-2"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb67-2"></a>nbclust_tsne &lt;-<span class="st"> </span><span class="kw">readRDS</span>(<span class="dt">file =</span> <span class="st">&quot;backup/60_002_nbclust_tsne_asurat.rds&quot;</span>)</span></code></pre></div>
</div>
<div id="comparing-clustering-validity-indices" class="section level3">
<h3><span class="header-section-number">14.3.6</span> Comparing clustering validity indices</h3>
<p>Load the objects.</p>
<div class="sourceCode" id="cb68"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb68-1"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb68-1"></a><span class="kw">rm</span>(<span class="dt">list =</span> <span class="kw">ls</span>())</span>
<span id="cb68-2"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb68-2"></a>nbclust_umap_seurat   &lt;-<span class="st"> </span><span class="kw">readRDS</span>(<span class="dt">file =</span> <span class="st">&quot;backup/20_001_nbclust_umap_seurat.rds&quot;</span>)</span>
<span id="cb68-3"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb68-3"></a>nbclust_umap_monocle3 &lt;-<span class="st"> </span><span class="kw">readRDS</span>(</span>
<span id="cb68-4"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb68-4"></a>  <span class="dt">file =</span> <span class="st">&quot;backup/30_001_nbclust_umap_monocle3.rds&quot;</span>)</span>
<span id="cb68-5"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb68-5"></a>asw_sc3               &lt;-<span class="st"> </span><span class="kw">readRDS</span>(<span class="dt">file =</span> <span class="st">&quot;backup/40_001_asw_sc3.rds&quot;</span>)</span>
<span id="cb68-6"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb68-6"></a>nbclust_umap_pca      &lt;-<span class="st"> </span><span class="kw">readRDS</span>(<span class="dt">file =</span> <span class="st">&quot;backup/50_001_nbclust_umap_pca.rds&quot;</span>)</span>
<span id="cb68-7"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb68-7"></a>nbclust_umap_asurat   &lt;-<span class="st"> </span><span class="kw">readRDS</span>(<span class="dt">file =</span> <span class="st">&quot;backup/60_001_nbclust_umap_asurat.rds&quot;</span>)</span>
<span id="cb68-8"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb68-8"></a></span>
<span id="cb68-9"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb68-9"></a>nbclust_tsne_seurat   &lt;-<span class="st"> </span><span class="kw">readRDS</span>(<span class="dt">file =</span> <span class="st">&quot;backup/20_002_nbclust_tsne_seurat.rds&quot;</span>)</span>
<span id="cb68-10"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb68-10"></a>nbclust_tsne_monocle3 &lt;-<span class="st"> </span><span class="kw">readRDS</span>(</span>
<span id="cb68-11"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb68-11"></a>  <span class="dt">file =</span> <span class="st">&quot;backup/30_002_nbclust_tsne_monocle3.rds&quot;</span>)</span>
<span id="cb68-12"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb68-12"></a>nbclust_tsne_pca      &lt;-<span class="st"> </span><span class="kw">readRDS</span>(<span class="dt">file =</span> <span class="st">&quot;backup/50_002_nbclust_tsne_pca.rds&quot;</span>)</span>
<span id="cb68-13"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb68-13"></a>nbclust_tsne_asurat   &lt;-<span class="st"> </span><span class="kw">readRDS</span>(<span class="dt">file =</span> <span class="st">&quot;backup/60_002_nbclust_tsne_asurat.rds&quot;</span>)</span></code></pre></div>
<p>Plot the average silhouette widths across all the methods.</p>
<div class="sourceCode" id="cb69"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb69-1"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb69-1"></a>method_names &lt;-<span class="st"> </span><span class="kw">c</span>(<span class="st">&quot;Seurat&quot;</span>, <span class="st">&quot;Monocle 3&quot;</span>, <span class="st">&quot;SC3&quot;</span>, <span class="st">&quot;PCA&quot;</span>, <span class="st">&quot;ASURAT&quot;</span>)</span>
<span id="cb69-2"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb69-2"></a>max_k &lt;-<span class="st"> </span><span class="dv">6</span>  <span class="co"># ASWs are computed for k = 2, 3, ..., (max_k + 1),</span></span>
<span id="cb69-3"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb69-3"></a>            <span class="co"># where k is the number of clusters</span></span>
<span id="cb69-4"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb69-4"></a>xlabel &lt;-<span class="st"> &quot;Number of clusters&quot;</span></span>
<span id="cb69-5"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb69-5"></a>ylabel &lt;-<span class="st"> &quot;Average silhouette width&quot;</span></span>
<span id="cb69-6"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb69-6"></a><span class="co"># ========================================</span></span>
<span id="cb69-7"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb69-7"></a><span class="co"># UMAP</span></span>
<span id="cb69-8"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb69-8"></a><span class="co"># ========================================</span></span>
<span id="cb69-9"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb69-9"></a>names &lt;-<span class="st"> </span><span class="kw">c</span>(<span class="st">&quot;day1_norm (UMAP)&quot;</span>, <span class="st">&quot;day7_hypo (UMAP)&quot;</span>, <span class="st">&quot;sc68_vehi (UMAP)&quot;</span>,</span>
<span id="cb69-10"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb69-10"></a>           <span class="st">&quot;sc68_cisp (UMAP)&quot;</span>, <span class="st">&quot;pbmc_4000 (UMAP)&quot;</span>, <span class="st">&quot;pbmc_6000 (UMAP)&quot;</span>)</span>
<span id="cb69-11"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb69-11"></a>ymin &lt;-<span class="st"> </span><span class="fl">0.3</span></span>
<span id="cb69-12"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb69-12"></a>ymax &lt;-<span class="st"> </span><span class="fl">0.905</span></span>
<span id="cb69-13"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb69-13"></a><span class="cf">for</span>(i <span class="cf">in</span> <span class="dv">1</span><span class="op">:</span><span class="dv">6</span>){</span>
<span id="cb69-14"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb69-14"></a>  <span class="co"># ----------------------------------------</span></span>
<span id="cb69-15"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb69-15"></a>  <span class="co"># From day1_norm (i = 1) to pbmc_6000 (i = 6)</span></span>
<span id="cb69-16"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb69-16"></a>  <span class="co"># ----------------------------------------</span></span>
<span id="cb69-17"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb69-17"></a>  si &lt;-<span class="st"> </span><span class="kw">list</span>(</span>
<span id="cb69-18"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb69-18"></a>    <span class="kw">as.data.frame</span>(nbclust_umap_seurat[[i]][[<span class="st">&quot;All.index&quot;</span>]])<span class="op">$</span>Silhouette,</span>
<span id="cb69-19"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb69-19"></a>    <span class="kw">as.data.frame</span>(nbclust_umap_monocle3[[i]][[<span class="st">&quot;All.index&quot;</span>]])<span class="op">$</span>Silhouette,</span>
<span id="cb69-20"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb69-20"></a>    <span class="kw">as.numeric</span>(<span class="kw">unlist</span>(asw_sc3[[i]])),</span>
<span id="cb69-21"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb69-21"></a>    <span class="kw">as.data.frame</span>(nbclust_umap_pca[[i]][[<span class="st">&quot;All.index&quot;</span>]])<span class="op">$</span>Silhouette,</span>
<span id="cb69-22"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb69-22"></a>    <span class="kw">as.data.frame</span>(nbclust_umap_asurat[[i]][[<span class="st">&quot;All.index&quot;</span>]])<span class="op">$</span>Silhouette)</span>
<span id="cb69-23"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb69-23"></a>  res &lt;-<span class="st"> </span><span class="kw">c</span>()</span>
<span id="cb69-24"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb69-24"></a>  <span class="cf">for</span>(l <span class="cf">in</span> <span class="dv">1</span><span class="op">:</span><span class="kw">length</span>(si)){</span>
<span id="cb69-25"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb69-25"></a>    <span class="cf">for</span>(k <span class="cf">in</span> <span class="dv">1</span><span class="op">:</span>max_k){</span>
<span id="cb69-26"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb69-26"></a>      <span class="cf">if</span>(l <span class="op">!=</span><span class="st"> </span><span class="dv">3</span>){    <span class="co"># For other than SC3</span></span>
<span id="cb69-27"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb69-27"></a>        res &lt;-<span class="st"> </span><span class="kw">rbind</span>(res, <span class="kw">c</span>(k <span class="op">+</span><span class="st"> </span><span class="dv">1</span>, si[[l]][k], l, method_names[l]))</span>
<span id="cb69-28"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb69-28"></a>      }<span class="cf">else</span>{</span>
<span id="cb69-29"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb69-29"></a>        <span class="cf">if</span>((i <span class="op">!=</span><span class="st"> </span><span class="dv">3</span>) <span class="op">&amp;</span><span class="st"> </span>(i <span class="op">!=</span><span class="st"> </span><span class="dv">4</span>)){    <span class="co"># For other than `sc68_vehi` and `sc68_cisp`</span></span>
<span id="cb69-30"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb69-30"></a>          res &lt;-<span class="st"> </span><span class="kw">rbind</span>(res, <span class="kw">c</span>(k <span class="op">+</span><span class="st"> </span><span class="dv">1</span>, si[[l]][k], l, method_names[l]))</span>
<span id="cb69-31"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb69-31"></a>        }<span class="cf">else</span>{</span>
<span id="cb69-32"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb69-32"></a>          res &lt;-<span class="st"> </span><span class="kw">rbind</span>(res, <span class="kw">c</span>(k <span class="op">+</span><span class="st"> </span><span class="dv">1</span>, <span class="kw">log</span>(<span class="dv">0</span>), l, method_names[l]))</span>
<span id="cb69-33"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb69-33"></a>        }</span>
<span id="cb69-34"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb69-34"></a>      }</span>
<span id="cb69-35"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb69-35"></a>    }</span>
<span id="cb69-36"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb69-36"></a>  }</span>
<span id="cb69-37"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb69-37"></a>  <span class="kw">colnames</span>(res) &lt;-<span class="st"> </span><span class="kw">c</span>(<span class="st">&quot;identity&quot;</span>, <span class="st">&quot;asw&quot;</span>, <span class="st">&quot;method_id&quot;</span>, <span class="st">&quot;method_name&quot;</span>)</span>
<span id="cb69-38"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb69-38"></a>  df &lt;-<span class="st"> </span><span class="kw">as.data.frame</span>(res)</span>
<span id="cb69-39"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb69-39"></a>  my_colors &lt;-<span class="st"> </span><span class="kw">brewer.pal</span>(<span class="dv">5</span>, <span class="dt">name =</span> <span class="st">&quot;Set1&quot;</span>)</span>
<span id="cb69-40"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb69-40"></a>  <span class="co"># ----------------------------------------</span></span>
<span id="cb69-41"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb69-41"></a>  <span class="co"># For annotations</span></span>
<span id="cb69-42"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb69-42"></a>  <span class="co"># ----------------------------------------</span></span>
<span id="cb69-43"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb69-43"></a>  <span class="cf">if</span>((i <span class="op">==</span><span class="st"> </span><span class="dv">5</span>) <span class="op">||</span><span class="st"> </span>(i <span class="op">==</span><span class="st"> </span><span class="dv">6</span>)){</span>
<span id="cb69-44"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb69-44"></a>    anno_xmin =<span class="st"> </span><span class="fl">3.7</span></span>
<span id="cb69-45"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb69-45"></a>    anno_xmax =<span class="st"> </span><span class="fl">6.3</span></span>
<span id="cb69-46"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb69-46"></a>  }<span class="cf">else</span>{</span>
<span id="cb69-47"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb69-47"></a>    anno_xmin =<span class="st"> </span><span class="dv">0</span></span>
<span id="cb69-48"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb69-48"></a>    anno_xmax =<span class="st"> </span><span class="dv">0</span></span>
<span id="cb69-49"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb69-49"></a>  }</span>
<span id="cb69-50"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb69-50"></a>  p &lt;-<span class="st"> </span><span class="kw">ggplot</span>() <span class="op">+</span></span>
<span id="cb69-51"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb69-51"></a><span class="st">    </span><span class="kw">geom_hline</span>(<span class="dt">yintercept =</span> <span class="fl">.6</span>, <span class="dt">linetype =</span> <span class="st">&quot;dashed&quot;</span>, <span class="dt">col =</span> <span class="st">&quot;black&quot;</span>, <span class="dt">size =</span> <span class="fl">.8</span>) <span class="op">+</span></span>
<span id="cb69-52"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb69-52"></a><span class="st">    </span><span class="kw">annotate</span>(<span class="st">&quot;rect&quot;</span>, <span class="dt">xmin =</span> anno_xmin, <span class="dt">xmax =</span> anno_xmax,</span>
<span id="cb69-53"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb69-53"></a>             <span class="dt">ymin =</span> ymin, <span class="dt">ymax =</span> ymax, <span class="dt">alpha =</span> <span class="fl">0.3</span>,</span>
<span id="cb69-54"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb69-54"></a>             <span class="dt">color =</span> <span class="st">&quot;black&quot;</span>, <span class="dt">fill =</span> <span class="st">&quot;gray&quot;</span>) <span class="op">+</span></span>
<span id="cb69-55"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb69-55"></a><span class="st">    </span><span class="kw">geom_point</span>(<span class="kw">aes</span>(<span class="dt">x =</span> <span class="kw">as.integer</span>(df[,<span class="dv">1</span>]), <span class="dt">y =</span> <span class="kw">as.numeric</span>(df[,<span class="dv">2</span>]),</span>
<span id="cb69-56"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb69-56"></a>                   <span class="dt">color =</span> <span class="kw">as.factor</span>(df[,<span class="dv">3</span>])), <span class="dt">size =</span> <span class="dv">5</span>, <span class="dt">alpha =</span> <span class="fl">0.8</span>) <span class="op">+</span></span>
<span id="cb69-57"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb69-57"></a><span class="st">    </span><span class="kw">labs</span>(<span class="dt">title =</span> names[i], <span class="dt">x =</span> <span class="st">&quot;Number of clusters&quot;</span>,</span>
<span id="cb69-58"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb69-58"></a>         <span class="dt">y =</span> <span class="st">&quot;Average silhouette width&quot;</span>, <span class="dt">colour =</span> <span class="st">&quot;&quot;</span>) <span class="op">+</span></span>
<span id="cb69-59"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb69-59"></a><span class="st">    </span><span class="kw">xlim</span>(<span class="fl">1.7</span>, (max_k<span class="op">+</span><span class="dv">1</span><span class="fl">+0.3</span>)) <span class="op">+</span><span class="st"> </span><span class="kw">ylim</span>(ymin, ymax) <span class="op">+</span></span>
<span id="cb69-60"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb69-60"></a><span class="st">    </span><span class="kw">scale_colour_manual</span>(<span class="dt">values =</span> my_colors, <span class="dt">name =</span> <span class="st">&quot;Method&quot;</span>,</span>
<span id="cb69-61"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb69-61"></a>                        <span class="dt">labels =</span> <span class="kw">unique</span>(df[,<span class="dv">4</span>])) <span class="op">+</span></span>
<span id="cb69-62"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb69-62"></a><span class="st">    </span><span class="kw">theme_classic</span>(<span class="dt">base_size =</span> <span class="dv">22</span>, <span class="dt">base_family =</span> <span class="st">&quot;Helvetica&quot;</span>) <span class="op">+</span></span>
<span id="cb69-63"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb69-63"></a><span class="st">    </span><span class="kw">theme</span>(<span class="dt">plot.title =</span> <span class="kw">element_text</span>(<span class="dt">hjust =</span> <span class="fl">0.5</span>, <span class="dt">size =</span> <span class="dv">16</span>),</span>
<span id="cb69-64"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb69-64"></a>          <span class="dt">plot.margin =</span> <span class="kw">unit</span>(<span class="kw">c</span>(<span class="fl">1.5</span>,<span class="fl">1.5</span>,<span class="fl">1.5</span>,<span class="fl">1.5</span>), <span class="st">&quot;lines&quot;</span>),</span>
<span id="cb69-65"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb69-65"></a>          <span class="dt">legend.position =</span> <span class="st">&quot;right&quot;</span>) <span class="op">+</span></span>
<span id="cb69-66"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb69-66"></a><span class="st">    </span><span class="kw">coord_cartesian</span>(<span class="dt">clip =</span> <span class="st">&#39;off&#39;</span>)</span>
<span id="cb69-67"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb69-67"></a>  filename &lt;-<span class="st"> </span><span class="kw">paste</span>(<span class="st">&quot;figures/figure_90_&quot;</span>, <span class="kw">sprintf</span>(<span class="st">&quot;%04d&quot;</span>, i), <span class="st">&quot;.png&quot;</span>, <span class="dt">sep =</span> <span class="st">&quot;&quot;</span>)</span>
<span id="cb69-68"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb69-68"></a>  <span class="kw">ggsave</span>(<span class="dt">file =</span> filename, <span class="dt">plot =</span> p, <span class="dt">dpi =</span> <span class="dv">300</span>, <span class="dt">width =</span> <span class="fl">6.7</span>, <span class="dt">height =</span> <span class="fl">4.5</span>)</span>
<span id="cb69-69"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb69-69"></a>}</span>
<span id="cb69-70"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb69-70"></a><span class="co"># ========================================</span></span>
<span id="cb69-71"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb69-71"></a><span class="co"># t-SNE</span></span>
<span id="cb69-72"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb69-72"></a><span class="co"># ========================================</span></span>
<span id="cb69-73"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb69-73"></a>names &lt;-<span class="st"> </span><span class="kw">c</span>(<span class="st">&quot;day1_norm (t-SNE)&quot;</span>, <span class="st">&quot;day7_hypo (t-SNE)&quot;</span>, <span class="st">&quot;sc68_vehi (t-SNE)&quot;</span>,</span>
<span id="cb69-74"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb69-74"></a>           <span class="st">&quot;sc68_cisp (t-SNE)&quot;</span>, <span class="st">&quot;pbmc_4000 (t-SNE)&quot;</span>, <span class="st">&quot;pbmc_6000 (t-SNE)&quot;</span>)</span>
<span id="cb69-75"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb69-75"></a>ymin &lt;-<span class="st"> </span><span class="fl">0.3</span></span>
<span id="cb69-76"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb69-76"></a>ymax &lt;-<span class="st"> </span><span class="fl">0.67</span></span>
<span id="cb69-77"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb69-77"></a><span class="cf">for</span>(i <span class="cf">in</span> <span class="dv">1</span><span class="op">:</span><span class="dv">6</span>){</span>
<span id="cb69-78"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb69-78"></a>  <span class="co"># ----------------------------------------</span></span>
<span id="cb69-79"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb69-79"></a>  <span class="co"># From day1_norm (i = 1) to pbmc_6000 (i = 6)</span></span>
<span id="cb69-80"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb69-80"></a>  <span class="co"># ----------------------------------------</span></span>
<span id="cb69-81"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb69-81"></a>  si &lt;-<span class="st"> </span><span class="kw">list</span>(</span>
<span id="cb69-82"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb69-82"></a>    <span class="kw">as.data.frame</span>(nbclust_tsne_seurat[[i]][[<span class="st">&quot;All.index&quot;</span>]])<span class="op">$</span>Silhouette,</span>
<span id="cb69-83"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb69-83"></a>    <span class="kw">as.data.frame</span>(nbclust_tsne_monocle3[[i]][[<span class="st">&quot;All.index&quot;</span>]])<span class="op">$</span>Silhouette,</span>
<span id="cb69-84"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb69-84"></a>    <span class="ot">NA</span>,</span>
<span id="cb69-85"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb69-85"></a>    <span class="kw">as.data.frame</span>(nbclust_tsne_pca[[i]][[<span class="st">&quot;All.index&quot;</span>]])<span class="op">$</span>Silhouette,</span>
<span id="cb69-86"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb69-86"></a>    <span class="kw">as.data.frame</span>(nbclust_tsne_asurat[[i]][[<span class="st">&quot;All.index&quot;</span>]])<span class="op">$</span>Silhouette)</span>
<span id="cb69-87"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb69-87"></a>  res &lt;-<span class="st"> </span><span class="kw">c</span>()</span>
<span id="cb69-88"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb69-88"></a>  <span class="cf">for</span>(l <span class="cf">in</span> <span class="dv">1</span><span class="op">:</span><span class="kw">length</span>(si)){</span>
<span id="cb69-89"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb69-89"></a>    <span class="cf">for</span>(k <span class="cf">in</span> <span class="dv">1</span><span class="op">:</span>max_k){</span>
<span id="cb69-90"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb69-90"></a>      <span class="cf">if</span>(<span class="op">!</span><span class="kw">is.na</span>(si[[l]])){</span>
<span id="cb69-91"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb69-91"></a>        res &lt;-<span class="st"> </span><span class="kw">rbind</span>(res, <span class="kw">c</span>(k <span class="op">+</span><span class="st"> </span><span class="dv">1</span>, si[[l]][k], l, method_names[l]))</span>
<span id="cb69-92"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb69-92"></a>      }<span class="cf">else</span>{</span>
<span id="cb69-93"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb69-93"></a>        <span class="cf">next</span></span>
<span id="cb69-94"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb69-94"></a>      }</span>
<span id="cb69-95"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb69-95"></a>    }</span>
<span id="cb69-96"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb69-96"></a>  }</span>
<span id="cb69-97"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb69-97"></a>  <span class="kw">colnames</span>(res) &lt;-<span class="st"> </span><span class="kw">c</span>(<span class="st">&quot;identity&quot;</span>, <span class="st">&quot;asw&quot;</span>, <span class="st">&quot;method_id&quot;</span>, <span class="st">&quot;method_name&quot;</span>)</span>
<span id="cb69-98"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb69-98"></a>  df &lt;-<span class="st"> </span><span class="kw">as.data.frame</span>(res)</span>
<span id="cb69-99"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb69-99"></a>  my_colors &lt;-<span class="st"> </span><span class="kw">brewer.pal</span>(<span class="dv">5</span>, <span class="dt">name =</span> <span class="st">&quot;Set1&quot;</span>)</span>
<span id="cb69-100"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb69-100"></a>  my_colors[<span class="dv">3</span>] &lt;-<span class="st"> </span>my_colors[<span class="dv">4</span>]</span>
<span id="cb69-101"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb69-101"></a>  my_colors[<span class="dv">4</span>] &lt;-<span class="st"> </span>my_colors[<span class="dv">5</span>]</span>
<span id="cb69-102"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb69-102"></a>  <span class="co"># ----------------------------------------</span></span>
<span id="cb69-103"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb69-103"></a>  <span class="co"># For annotations</span></span>
<span id="cb69-104"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb69-104"></a>  <span class="co"># ----------------------------------------</span></span>
<span id="cb69-105"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb69-105"></a>  <span class="cf">if</span>((i <span class="op">==</span><span class="st"> </span><span class="dv">5</span>) <span class="op">||</span><span class="st"> </span>(i <span class="op">==</span><span class="st"> </span><span class="dv">6</span>)){</span>
<span id="cb69-106"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb69-106"></a>    anno_xmin =<span class="st"> </span><span class="fl">3.7</span></span>
<span id="cb69-107"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb69-107"></a>    anno_xmax =<span class="st"> </span><span class="fl">6.3</span></span>
<span id="cb69-108"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb69-108"></a>  }<span class="cf">else</span>{</span>
<span id="cb69-109"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb69-109"></a>    anno_xmin =<span class="st"> </span><span class="dv">0</span></span>
<span id="cb69-110"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb69-110"></a>    anno_xmax =<span class="st"> </span><span class="dv">0</span></span>
<span id="cb69-111"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb69-111"></a>  }</span>
<span id="cb69-112"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb69-112"></a>  p &lt;-<span class="st"> </span><span class="kw">ggplot</span>() <span class="op">+</span></span>
<span id="cb69-113"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb69-113"></a><span class="st">    </span><span class="kw">annotate</span>(<span class="st">&quot;rect&quot;</span>, <span class="dt">xmin =</span> anno_xmin, <span class="dt">xmax =</span> anno_xmax,</span>
<span id="cb69-114"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb69-114"></a>             <span class="dt">ymin =</span> ymin, <span class="dt">ymax =</span> ymax, <span class="dt">alpha =</span> <span class="fl">0.3</span>,</span>
<span id="cb69-115"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb69-115"></a>             <span class="dt">color =</span> <span class="st">&quot;black&quot;</span>, <span class="dt">fill =</span> <span class="st">&quot;gray&quot;</span>) <span class="op">+</span></span>
<span id="cb69-116"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb69-116"></a><span class="st">    </span><span class="kw">geom_point</span>(<span class="kw">aes</span>(<span class="dt">x =</span> <span class="kw">as.integer</span>(df[,<span class="dv">1</span>]), <span class="dt">y =</span> <span class="kw">as.numeric</span>(df[,<span class="dv">2</span>]),</span>
<span id="cb69-117"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb69-117"></a>                   <span class="dt">color =</span> <span class="kw">as.factor</span>(df[,<span class="dv">3</span>])), <span class="dt">size =</span> <span class="dv">5</span>, <span class="dt">alpha =</span> <span class="fl">0.8</span>) <span class="op">+</span></span>
<span id="cb69-118"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb69-118"></a><span class="st">    </span><span class="kw">labs</span>(<span class="dt">title =</span> names[i], <span class="dt">x =</span> <span class="st">&quot;Number of clusters&quot;</span>,</span>
<span id="cb69-119"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb69-119"></a>         <span class="dt">y =</span> <span class="st">&quot;Average silhouette width&quot;</span>, <span class="dt">colour =</span> <span class="st">&quot;&quot;</span>) <span class="op">+</span></span>
<span id="cb69-120"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb69-120"></a><span class="st">    </span><span class="kw">xlim</span>(<span class="fl">1.7</span>, (max_k<span class="op">+</span><span class="dv">1</span><span class="fl">+0.3</span>)) <span class="op">+</span><span class="st"> </span><span class="kw">ylim</span>(ymin, ymax) <span class="op">+</span></span>
<span id="cb69-121"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb69-121"></a><span class="st">    </span><span class="kw">scale_colour_manual</span>(<span class="dt">values =</span> my_colors, <span class="dt">name =</span> <span class="st">&quot;Method&quot;</span>,</span>
<span id="cb69-122"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb69-122"></a>                        <span class="dt">labels =</span> <span class="kw">unique</span>(df[,<span class="dv">4</span>])) <span class="op">+</span></span>
<span id="cb69-123"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb69-123"></a><span class="st">    </span><span class="kw">theme_classic</span>(<span class="dt">base_size =</span> <span class="dv">22</span>, <span class="dt">base_family =</span> <span class="st">&quot;Helvetica&quot;</span>) <span class="op">+</span></span>
<span id="cb69-124"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb69-124"></a><span class="st">    </span><span class="kw">theme</span>(<span class="dt">plot.title =</span> <span class="kw">element_text</span>(<span class="dt">hjust =</span> <span class="fl">0.5</span>, <span class="dt">size =</span> <span class="dv">16</span>),</span>
<span id="cb69-125"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb69-125"></a>          <span class="dt">plot.margin =</span> <span class="kw">unit</span>(<span class="kw">c</span>(<span class="fl">1.5</span>,<span class="fl">1.5</span>,<span class="fl">1.5</span>,<span class="fl">1.5</span>), <span class="st">&quot;lines&quot;</span>),</span>
<span id="cb69-126"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb69-126"></a>          <span class="dt">legend.position =</span> <span class="st">&quot;right&quot;</span>) <span class="op">+</span></span>
<span id="cb69-127"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb69-127"></a><span class="st">    </span><span class="kw">coord_cartesian</span>(<span class="dt">clip =</span> <span class="st">&#39;off&#39;</span>)</span>
<span id="cb69-128"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb69-128"></a>  filename &lt;-<span class="st"> </span><span class="kw">paste</span>(<span class="st">&quot;figures/figure_91_&quot;</span>, <span class="kw">sprintf</span>(<span class="st">&quot;%04d&quot;</span>, i), <span class="st">&quot;.png&quot;</span>, <span class="dt">sep =</span> <span class="st">&quot;&quot;</span>)</span>
<span id="cb69-129"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb69-129"></a>  <span class="kw">ggsave</span>(<span class="dt">file =</span> filename, <span class="dt">plot =</span> p, <span class="dt">dpi =</span> <span class="dv">300</span>, <span class="dt">width =</span> <span class="fl">6.7</span>, <span class="dt">height =</span> <span class="fl">4.5</span>)</span>
<span id="cb69-130"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb69-130"></a>}</span></code></pre></div>
<p><img src="figures/figure_90_0001.png" width="250px">
<img src="figures/figure_90_0002.png" width="250px">
<img src="figures/figure_90_0003.png" width="250px">
<img src="figures/figure_90_0004.png" width="250px">
<img src="figures/figure_90_0005.png" width="250px">
<img src="figures/figure_90_0006.png" width="250px"></p>
<p><img src="figures/figure_91_0001.png" width="250px">
<img src="figures/figure_91_0002.png" width="250px">
<img src="figures/figure_91_0003.png" width="250px">
<img src="figures/figure_91_0004.png" width="250px">
<img src="figures/figure_91_0005.png" width="250px">
<img src="figures/figure_91_0006.png" width="250px"></p>
</div>
</div>
<div id="cell-type-inference-in-pbmc_4000-and-pbmc_6000" class="section level2">
<h2><span class="header-section-number">14.4</span> Cell-type inference in <code>pbmc_4000</code> and <code>pbmc_6000</code></h2>
<p>In this section, we infer the cell types existing in the PBMC datasets
by using Seurat, Monocle 3 plus Garnett, SC3, and ASURAT.</p>
<div id="seurat-1" class="section level3">
<h3><span class="header-section-number">14.4.1</span> Seurat</h3>
<p>Load the objects.</p>
<div class="sourceCode" id="cb70"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb70-1"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb70-1"></a><span class="kw">rm</span>(<span class="dt">list =</span> <span class="kw">ls</span>())</span>
<span id="cb70-2"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb70-2"></a><span class="kw">source</span>(<span class="st">&quot;R/function_seurat.R&quot;</span>)</span>
<span id="cb70-3"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb70-3"></a>pbmc_<span class="dv">4000</span> &lt;-<span class="st"> </span><span class="kw">readRDS</span>(<span class="dt">file =</span> <span class="st">&quot;backup/15_020_pbmc_4000_seurat.rds&quot;</span>)</span>
<span id="cb70-4"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb70-4"></a>pbmc_<span class="dv">6000</span> &lt;-<span class="st"> </span><span class="kw">readRDS</span>(<span class="dt">file =</span> <span class="st">&quot;backup/16_020_pbmc_6000_seurat.rds&quot;</span>)</span></code></pre></div>
<p>Run <code>FindClusters()</code> for clustering cells.</p>
<div class="sourceCode" id="cb71"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb71-1"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb71-1"></a>pbmc_<span class="dv">4000</span> &lt;-<span class="st"> </span><span class="kw">FindClusters</span>(pbmc_<span class="dv">4000</span>, <span class="dt">resolution =</span> <span class="fl">0.08</span>)</span>
<span id="cb71-2"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb71-2"></a>pbmc_<span class="dv">6000</span> &lt;-<span class="st"> </span><span class="kw">FindClusters</span>(pbmc_<span class="dv">6000</span>, <span class="dt">resolution =</span> <span class="fl">0.15</span>)</span></code></pre></div>
<p>The following function <code>plot_umap_seurat()</code> shows cell states on a UMAP space.
Note that the labels of groups are added by 1 from the original Seurat’s
labeling (0, 1, 2, …).
The arguments are <code>obj</code>, <code>title</code>, <code>title_size, ``xlabel</code>, <code>ylabel</code>, and
<code>default_color</code> (if <code>TRUE</code>, the default color of ggplot is used,
otherwise rainbow).</p>
<div class="sourceCode" id="cb72"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb72-1"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb72-1"></a><span class="co"># ----------------------------------------</span></span>
<span id="cb72-2"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb72-2"></a><span class="co"># For `pbmc_4000`</span></span>
<span id="cb72-3"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb72-3"></a><span class="co"># ----------------------------------------</span></span>
<span id="cb72-4"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb72-4"></a>p &lt;-<span class="st"> </span><span class="kw">plot_umap_seurat</span>(<span class="dt">obj =</span> pbmc_<span class="dv">4000</span>, <span class="dt">title =</span> <span class="st">&quot;pbmc_4000&quot;</span>, <span class="dt">title_size =</span> <span class="dv">24</span>,</span>
<span id="cb72-5"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb72-5"></a>                      <span class="dt">xlabel =</span> <span class="st">&quot;UMAP_1&quot;</span>, <span class="dt">ylabel =</span> <span class="st">&quot;UMAP_2&quot;</span>,</span>
<span id="cb72-6"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb72-6"></a>                      <span class="dt">default_color =</span> <span class="ot">TRUE</span>)</span>
<span id="cb72-7"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb72-7"></a>filename &lt;-<span class="st"> &quot;figures/figure_15_0025.png&quot;</span></span>
<span id="cb72-8"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb72-8"></a><span class="kw">ggsave</span>(<span class="dt">file =</span> filename, <span class="dt">plot =</span> p, <span class="dt">dpi =</span> <span class="dv">300</span>, <span class="dt">width =</span> <span class="dv">5</span>, <span class="dt">height =</span> <span class="dv">4</span>)</span></code></pre></div>
<p><img src="figures/figure_15_0025.png" width="250px">
<img src="figures/figure_16_0025.png" width="250px"></p>
<p>Find differentially expressed genes using Seurat’s function and stores the
results into <code>obj@misc[["markers"]]</code>.</p>
<div class="sourceCode" id="cb73"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb73-1"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb73-1"></a>pbmc_<span class="dv">4000</span><span class="op">@</span>misc[[<span class="st">&quot;markers&quot;</span>]] &lt;-<span class="st"> </span><span class="kw">FindAllMarkers</span>(</span>
<span id="cb73-2"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb73-2"></a>  pbmc_<span class="dv">4000</span>, <span class="dt">only.pos =</span> T, <span class="dt">min.pct =</span> <span class="fl">0.25</span>, <span class="dt">logfc.threshold =</span> <span class="fl">0.25</span>)</span>
<span id="cb73-3"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb73-3"></a>pbmc_<span class="dv">6000</span><span class="op">@</span>misc[[<span class="st">&quot;markers&quot;</span>]] &lt;-<span class="st"> </span><span class="kw">FindAllMarkers</span>(</span>
<span id="cb73-4"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb73-4"></a>  pbmc_<span class="dv">6000</span>, <span class="dt">only.pos =</span> T, <span class="dt">min.pct =</span> <span class="fl">0.25</span>, <span class="dt">logfc.threshold =</span> <span class="fl">0.25</span>)</span></code></pre></div>
<p>Save the objects.</p>
<div class="sourceCode" id="cb74"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb74-1"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb74-1"></a><span class="kw">saveRDS</span>(pbmc_<span class="dv">4000</span>, <span class="dt">file =</span> <span class="st">&quot;backup/15_021_pbmc_4000_marker_seurat.rds&quot;</span>)</span>
<span id="cb74-2"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb74-2"></a><span class="kw">saveRDS</span>(pbmc_<span class="dv">6000</span>, <span class="dt">file =</span> <span class="st">&quot;backup/16_021_pbmc_6000_marker_seurat.rds&quot;</span>)</span></code></pre></div>
<p>Load the objects.</p>
<div class="sourceCode" id="cb75"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb75-1"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb75-1"></a>pbmc_<span class="dv">4000</span> &lt;-<span class="st"> </span><span class="kw">readRDS</span>(<span class="dt">file =</span> <span class="st">&quot;backup/15_021_pbmc_4000_marker_seurat.rds&quot;</span>)</span>
<span id="cb75-2"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb75-2"></a>pbmc_<span class="dv">6000</span> &lt;-<span class="st"> </span><span class="kw">readRDS</span>(<span class="dt">file =</span> <span class="st">&quot;backup/16_021_pbmc_6000_marker_seurat.rds&quot;</span>)</span></code></pre></div>
<p>Users can check the marker genes of each cluster by the following command:</p>
<div class="sourceCode" id="cb76"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb76-1"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb76-1"></a><span class="kw">datatable</span>(pbmc_<span class="dv">4000</span><span class="op">@</span>misc[[<span class="st">&quot;markers&quot;</span>]], <span class="dt">rownames =</span> <span class="ot">FALSE</span>)</span>
<span id="cb76-2"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb76-2"></a><span class="kw">datatable</span>(pbmc_<span class="dv">6000</span><span class="op">@</span>misc[[<span class="st">&quot;markers&quot;</span>]], <span class="dt">rownames =</span> <span class="ot">FALSE</span>)</span></code></pre></div>
<ul>
<li><code>pbmc_4000</code></li>
</ul>
<iframe src="figures/pbmc_4000_marker_seurat.html" style="width:100%; height:530px;">
</iframe>
<ul>
<li><code>pbmc_6000</code></li>
</ul>
<iframe src="figures/pbmc_6000_marker_seurat.html" style="width:100%; height:530px;">
</iframe>
<p>Based on the above results and the manual investigation using GeneCards,
we annotated each group as follows:</p>
<ul>
<li><code>pbmc_4000</code></li>
</ul>
<pre><code>0: T cell         # CD3D, IL32, TRAC, CD3E
1: MC             # S100A8, LYZ, CD14
2: B cell         # CD79A, MS4A1, IGHM
3: NK/NKT cell    # NKG7, CD160
4: NK/NKT cell    # KLRF1, GZMB, CD160
5: Immune cell    # PTCRA, JCHAIN, MZB1, GZMB, ..</code></pre>
<ul>
<li><code>pbmc_6000</code></li>
</ul>
<pre><code>0: T cell         # CD3D, TCF7, CD27, CD3E
1: MC             # S100A8, LYZ, CD14
2: NK/NKT cell    # NKG7, GZMA, FGFBP2, GNLY
3: B cell         # CD79A, VPREB3, BANK1</code></pre>
<p>Identify the cell types.</p>
<div class="sourceCode" id="cb79"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb79-1"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb79-1"></a><span class="co"># ----------------------------------------</span></span>
<span id="cb79-2"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb79-2"></a><span class="co"># For `pbmc_4000`</span></span>
<span id="cb79-3"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb79-3"></a><span class="co"># ----------------------------------------</span></span>
<span id="cb79-4"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb79-4"></a>tmp &lt;-<span class="st"> </span><span class="kw">as.integer</span>(<span class="kw">as.character</span>(pbmc_<span class="dv">4000</span><span class="op">@</span>meta.data[[<span class="st">&quot;seurat_clusters&quot;</span>]]))</span>
<span id="cb79-5"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb79-5"></a>tmp[tmp <span class="op">==</span><span class="st"> </span><span class="dv">0</span>] &lt;-<span class="st"> &quot;T&quot;</span></span>
<span id="cb79-6"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb79-6"></a>tmp[tmp <span class="op">==</span><span class="st"> </span><span class="dv">1</span>] &lt;-<span class="st"> &quot;MC&quot;</span></span>
<span id="cb79-7"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb79-7"></a>tmp[tmp <span class="op">==</span><span class="st"> </span><span class="dv">2</span>] &lt;-<span class="st"> &quot;B&quot;</span></span>
<span id="cb79-8"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb79-8"></a>tmp[tmp <span class="op">==</span><span class="st"> </span><span class="dv">3</span>] &lt;-<span class="st"> &quot;NK/NKT&quot;</span></span>
<span id="cb79-9"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb79-9"></a>tmp[tmp <span class="op">==</span><span class="st"> </span><span class="dv">4</span>] &lt;-<span class="st"> &quot;NK/NKT&quot;</span></span>
<span id="cb79-10"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb79-10"></a>tmp[tmp <span class="op">==</span><span class="st"> </span><span class="dv">5</span>] &lt;-<span class="st"> &quot;u.s.&quot;</span>   <span class="co"># Unspecified</span></span>
<span id="cb79-11"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb79-11"></a>tmp &lt;-<span class="st"> </span><span class="kw">factor</span>(tmp, <span class="dt">levels =</span> <span class="kw">c</span>(<span class="st">&quot;T&quot;</span>, <span class="st">&quot;MC&quot;</span>, <span class="st">&quot;B&quot;</span>, <span class="st">&quot;NK/NKT&quot;</span>, <span class="st">&quot;u.s.&quot;</span>))</span>
<span id="cb79-12"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb79-12"></a>pbmc_<span class="dv">4000</span><span class="op">@</span>meta.data[[<span class="st">&quot;mylabel&quot;</span>]] &lt;-<span class="st"> </span>tmp</span>
<span id="cb79-13"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb79-13"></a><span class="co"># ----------------------------------------</span></span>
<span id="cb79-14"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb79-14"></a><span class="co"># For `pbmc_6000`</span></span>
<span id="cb79-15"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb79-15"></a><span class="co"># ----------------------------------------</span></span>
<span id="cb79-16"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb79-16"></a>tmp &lt;-<span class="st"> </span><span class="kw">as.integer</span>(<span class="kw">as.character</span>(pbmc_<span class="dv">6000</span><span class="op">@</span>meta.data[[<span class="st">&quot;seurat_clusters&quot;</span>]]))</span>
<span id="cb79-17"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb79-17"></a>tmp[tmp <span class="op">==</span><span class="st"> </span><span class="dv">0</span>] &lt;-<span class="st"> &quot;T&quot;</span></span>
<span id="cb79-18"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb79-18"></a>tmp[tmp <span class="op">==</span><span class="st"> </span><span class="dv">1</span>] &lt;-<span class="st"> &quot;MC&quot;</span></span>
<span id="cb79-19"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb79-19"></a>tmp[tmp <span class="op">==</span><span class="st"> </span><span class="dv">2</span>] &lt;-<span class="st"> &quot;NK/NKT&quot;</span></span>
<span id="cb79-20"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb79-20"></a>tmp[tmp <span class="op">==</span><span class="st"> </span><span class="dv">3</span>] &lt;-<span class="st"> &quot;B&quot;</span></span>
<span id="cb79-21"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb79-21"></a>tmp &lt;-<span class="st"> </span><span class="kw">factor</span>(tmp, <span class="dt">levels =</span> <span class="kw">c</span>(<span class="st">&quot;T&quot;</span>, <span class="st">&quot;MC&quot;</span>, <span class="st">&quot;B&quot;</span>, <span class="st">&quot;NK/NKT&quot;</span>))</span>
<span id="cb79-22"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb79-22"></a>pbmc_<span class="dv">6000</span><span class="op">@</span>meta.data[[<span class="st">&quot;mylabel&quot;</span>]] &lt;-<span class="st"> </span>tmp</span></code></pre></div>
<p>The following function <code>plot_manifold()</code> shows sample states on UMAP space.</p>
<div class="sourceCode" id="cb80"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb80-1"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb80-1"></a>plot_manifold &lt;-<span class="st"> </span><span class="cf">function</span>(obj, title, title_size, xlabel, ylabel){</span>
<span id="cb80-2"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb80-2"></a>  <span class="co"># ----------------------------------------</span></span>
<span id="cb80-3"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb80-3"></a>  <span class="co"># Definitions</span></span>
<span id="cb80-4"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb80-4"></a>  <span class="co"># ----------------------------------------</span></span>
<span id="cb80-5"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb80-5"></a>  df &lt;-<span class="st"> </span><span class="kw">data.frame</span>(</span>
<span id="cb80-6"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb80-6"></a>    <span class="dt">x =</span> obj<span class="op">@</span>reductions[[<span class="st">&quot;umap&quot;</span>]]<span class="op">@</span>cell.embeddings[,<span class="dv">1</span>],</span>
<span id="cb80-7"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb80-7"></a>    <span class="dt">y =</span> obj<span class="op">@</span>reductions[[<span class="st">&quot;umap&quot;</span>]]<span class="op">@</span>cell.embeddings[,<span class="dv">2</span>],</span>
<span id="cb80-8"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb80-8"></a>    <span class="dt">label =</span> obj[[<span class="st">&quot;seurat_clusters&quot;</span>]],</span>
<span id="cb80-9"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb80-9"></a>    <span class="dt">anno =</span> obj[[<span class="st">&quot;mylabel&quot;</span>]]</span>
<span id="cb80-10"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb80-10"></a>  )</span>
<span id="cb80-11"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb80-11"></a>  <span class="kw">colnames</span>(df) &lt;-<span class="st"> </span><span class="kw">c</span>(<span class="st">&quot;x&quot;</span>, <span class="st">&quot;y&quot;</span>, <span class="st">&quot;label&quot;</span>, <span class="st">&quot;anno&quot;</span>)</span>
<span id="cb80-12"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb80-12"></a>  n_groups &lt;-<span class="st"> </span><span class="kw">length</span>(<span class="kw">unique</span>(obj<span class="op">$</span>seurat_clusters))</span>
<span id="cb80-13"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb80-13"></a>  <span class="co"># ----------------------------------------</span></span>
<span id="cb80-14"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb80-14"></a>  <span class="co"># Color</span></span>
<span id="cb80-15"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb80-15"></a>  <span class="co"># ----------------------------------------</span></span>
<span id="cb80-16"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb80-16"></a>  ggColorHue &lt;-<span class="st"> </span><span class="cf">function</span>(n, <span class="dt">l=</span><span class="dv">65</span>){   <span class="co"># To produce default colors of ggplot2</span></span>
<span id="cb80-17"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb80-17"></a>    hues &lt;-<span class="st"> </span><span class="kw">seq</span>(<span class="dv">15</span>, <span class="dv">375</span>, <span class="dt">length=</span>n<span class="op">+</span><span class="dv">1</span>)</span>
<span id="cb80-18"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb80-18"></a>    <span class="kw">hcl</span>(<span class="dt">h=</span>hues, <span class="dt">l=</span>l, <span class="dt">c=</span><span class="dv">100</span>)[<span class="dv">1</span><span class="op">:</span>n]</span>
<span id="cb80-19"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb80-19"></a>  }</span>
<span id="cb80-20"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb80-20"></a>  my_colors &lt;-<span class="st"> </span><span class="kw">ggColorHue</span>(<span class="dv">6</span>)</span>
<span id="cb80-21"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb80-21"></a>  tmp &lt;-<span class="st"> </span><span class="kw">as.character</span>(df<span class="op">$</span>anno)</span>
<span id="cb80-22"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb80-22"></a>  tmp[tmp <span class="op">==</span><span class="st"> &quot;T&quot;</span>] &lt;-<span class="st"> </span>my_colors[<span class="dv">1</span>]</span>
<span id="cb80-23"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb80-23"></a>  tmp[tmp <span class="op">==</span><span class="st"> &quot;MC&quot;</span>] &lt;-<span class="st"> </span>my_colors[<span class="dv">2</span>]</span>
<span id="cb80-24"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb80-24"></a>  tmp[tmp <span class="op">==</span><span class="st"> &quot;B&quot;</span>] &lt;-<span class="st"> </span>my_colors[<span class="dv">3</span>]</span>
<span id="cb80-25"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb80-25"></a>  tmp[tmp <span class="op">==</span><span class="st"> &quot;NK/NKT&quot;</span>] &lt;-<span class="st"> </span>my_colors[<span class="dv">4</span>]</span>
<span id="cb80-26"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb80-26"></a>  tmp[tmp <span class="op">==</span><span class="st"> &quot;u.s.&quot;</span>] &lt;-<span class="st"> </span>my_colors[<span class="dv">6</span>]</span>
<span id="cb80-27"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb80-27"></a>  tmp &lt;-<span class="st"> </span><span class="kw">factor</span>(tmp, <span class="dt">levels =</span> my_colors)</span>
<span id="cb80-28"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb80-28"></a>  df<span class="op">$</span>col &lt;-<span class="st"> </span>tmp</span>
<span id="cb80-29"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb80-29"></a>  <span class="co"># ----------------------------------------</span></span>
<span id="cb80-30"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb80-30"></a>  <span class="co"># Positions of labels</span></span>
<span id="cb80-31"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb80-31"></a>  <span class="co"># ----------------------------------------</span></span>
<span id="cb80-32"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb80-32"></a>  cluster_num &lt;-<span class="st"> </span><span class="kw">levels</span>(df<span class="op">$</span>label)</span>
<span id="cb80-33"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb80-33"></a>  cog &lt;-<span class="st"> </span><span class="kw">c</span>()  <span class="co"># Center of gravity</span></span>
<span id="cb80-34"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb80-34"></a>  <span class="cf">for</span>(i <span class="cf">in</span> cluster_num){</span>
<span id="cb80-35"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb80-35"></a>    cog &lt;-<span class="st"> </span><span class="kw">rbind</span>(cog, <span class="kw">c</span>(</span>
<span id="cb80-36"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb80-36"></a>      <span class="kw">mean</span>(df[<span class="kw">which</span>(df<span class="op">$</span>label <span class="op">==</span><span class="st"> </span>i),]<span class="op">$</span>x),</span>
<span id="cb80-37"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb80-37"></a>      <span class="kw">mean</span>(df[<span class="kw">which</span>(df<span class="op">$</span>label <span class="op">==</span><span class="st"> </span>i),]<span class="op">$</span>y),</span>
<span id="cb80-38"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb80-38"></a>      <span class="kw">as.character</span>(<span class="kw">unique</span>(df[<span class="kw">which</span>(df<span class="op">$</span>label <span class="op">==</span><span class="st"> </span>i),]<span class="op">$</span>anno))</span>
<span id="cb80-39"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb80-39"></a>    ))</span>
<span id="cb80-40"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb80-40"></a>  }</span>
<span id="cb80-41"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb80-41"></a>  <span class="co"># ----------------------------------------</span></span>
<span id="cb80-42"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb80-42"></a>  <span class="co"># Plot</span></span>
<span id="cb80-43"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb80-43"></a>  <span class="co"># ----------------------------------------</span></span>
<span id="cb80-44"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb80-44"></a>  p &lt;-<span class="st"> </span><span class="kw">ggplot</span>() <span class="op">+</span></span>
<span id="cb80-45"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb80-45"></a><span class="st">    </span><span class="kw">geom_point</span>(<span class="kw">aes</span>(<span class="dt">x=</span>df[,<span class="dv">1</span>], <span class="dt">y=</span>df[,<span class="dv">2</span>], <span class="dt">color=</span>df[,<span class="dv">5</span>]), <span class="dt">size=</span><span class="fl">0.5</span>, <span class="dt">alpha=</span><span class="fl">1.0</span>) <span class="op">+</span></span>
<span id="cb80-46"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb80-46"></a><span class="st">    </span><span class="kw">labs</span>(<span class="dt">title=</span>title, <span class="dt">x=</span>xlabel, <span class="dt">y=</span>ylabel, <span class="dt">colour=</span><span class="st">&quot;&quot;</span>) <span class="op">+</span></span>
<span id="cb80-47"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb80-47"></a><span class="st">    </span><span class="kw">scale_colour_manual</span>(<span class="dt">values=</span>my_colors) <span class="op">+</span></span>
<span id="cb80-48"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb80-48"></a><span class="st">    </span><span class="kw">theme_classic</span>(<span class="dt">base_size=</span><span class="dv">20</span>, <span class="dt">base_family=</span><span class="st">&quot;Helvetica&quot;</span>) <span class="op">+</span></span>
<span id="cb80-49"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb80-49"></a><span class="st">    </span><span class="kw">theme</span>(<span class="dt">plot.title=</span><span class="kw">element_text</span>(<span class="dt">hjust=</span><span class="fl">0.5</span>, <span class="dt">size=</span>title_size),</span>
<span id="cb80-50"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb80-50"></a>      <span class="dt">plot.margin=</span><span class="kw">unit</span>(<span class="kw">c</span>(<span class="fl">0.5</span>,<span class="dv">2</span>,<span class="fl">0.5</span>,<span class="fl">0.5</span>), <span class="st">&quot;lines&quot;</span>), <span class="dt">legend.position=</span><span class="st">&quot;none&quot;</span>) <span class="op">+</span></span>
<span id="cb80-51"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb80-51"></a><span class="st">    </span><span class="kw">geom_text</span>(<span class="kw">aes</span>(<span class="dt">x=</span><span class="kw">as.numeric</span>(cog[,<span class="dv">1</span>]), <span class="dt">y=</span><span class="kw">as.numeric</span>(cog[,<span class="dv">2</span>]), <span class="dt">label=</span>cog[,<span class="dv">3</span>]),</span>
<span id="cb80-52"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb80-52"></a>      <span class="dt">size=</span><span class="dv">8</span>, <span class="dt">colour=</span><span class="st">&quot;black&quot;</span>)</span>
<span id="cb80-53"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb80-53"></a>  </span>
<span id="cb80-54"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb80-54"></a>  <span class="kw">return</span>(p)</span>
<span id="cb80-55"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb80-55"></a>}</span>
<span id="cb80-56"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb80-56"></a></span>
<span id="cb80-57"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb80-57"></a><span class="co"># ----------------------------------------</span></span>
<span id="cb80-58"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb80-58"></a><span class="co"># For `pbmc_4000`</span></span>
<span id="cb80-59"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb80-59"></a><span class="co"># ----------------------------------------</span></span>
<span id="cb80-60"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb80-60"></a>p &lt;-<span class="st"> </span><span class="kw">plot_manifold</span>(<span class="dt">obj =</span> pbmc_<span class="dv">4000</span>, <span class="dt">title =</span> <span class="st">&quot;pbmc_4000 (Seurat)&quot;</span>,</span>
<span id="cb80-61"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb80-61"></a>                   <span class="dt">title_size =</span> <span class="dv">24</span>, <span class="dt">xlabel =</span> <span class="st">&quot;UMAP_1&quot;</span>, <span class="dt">ylabel =</span> <span class="st">&quot;UMAP_2&quot;</span>)</span>
<span id="cb80-62"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb80-62"></a>filename &lt;-<span class="st"> &quot;figures/figure_15_0026.png&quot;</span></span>
<span id="cb80-63"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb80-63"></a><span class="kw">ggsave</span>(<span class="dt">file =</span> filename, <span class="dt">plot =</span> p, <span class="dt">dpi =</span> <span class="dv">300</span>, <span class="dt">width =</span> <span class="fl">4.9</span>, <span class="dt">height =</span> <span class="dv">5</span>)</span></code></pre></div>
<p><img src="figures/figure_15_0026.png" width="250px">
<img src="figures/figure_16_0026.png" width="250px"></p>
<p>The numbers of cells are stored in a table.</p>
<div class="sourceCode" id="cb81"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb81-1"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb81-1"></a>identify_cell &lt;-<span class="st"> </span><span class="cf">function</span>(obj){</span>
<span id="cb81-2"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb81-2"></a>  tmp &lt;-<span class="st"> </span>obj<span class="op">@</span>meta.data[[<span class="st">&quot;mylabel&quot;</span>]]</span>
<span id="cb81-3"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb81-3"></a>  cells &lt;-<span class="st"> </span><span class="kw">levels</span>(tmp)</span>
<span id="cb81-4"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb81-4"></a>  df &lt;-<span class="st"> </span><span class="kw">c</span>()</span>
<span id="cb81-5"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb81-5"></a>  <span class="cf">for</span>(cell <span class="cf">in</span> cells){</span>
<span id="cb81-6"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb81-6"></a>    df &lt;-<span class="st"> </span><span class="kw">rbind</span>(df, <span class="kw">c</span>(cell, <span class="kw">length</span>(tmp[tmp<span class="op">==</span>cell])))</span>
<span id="cb81-7"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb81-7"></a>  }</span>
<span id="cb81-8"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb81-8"></a>  df &lt;-<span class="st"> </span><span class="kw">as.data.frame</span>(df)</span>
<span id="cb81-9"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb81-9"></a>  <span class="kw">colnames</span>(df) &lt;-<span class="st"> </span><span class="kw">c</span>(<span class="st">&quot;cell_type&quot;</span>, <span class="st">&quot;n&quot;</span>)</span>
<span id="cb81-10"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb81-10"></a>  <span class="kw">return</span>(df)</span>
<span id="cb81-11"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb81-11"></a>}</span>
<span id="cb81-12"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb81-12"></a></span>
<span id="cb81-13"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb81-13"></a>pbmc_<span class="dv">4000</span>_cell &lt;-<span class="st"> </span><span class="kw">identify_cell</span>(<span class="dt">obj =</span> pbmc_<span class="dv">4000</span>)</span>
<span id="cb81-14"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb81-14"></a>pbmc_<span class="dv">6000</span>_cell &lt;-<span class="st"> </span><span class="kw">identify_cell</span>(<span class="dt">obj =</span> pbmc_<span class="dv">6000</span>)</span></code></pre></div>
<p>Save the objects.</p>
<div class="sourceCode" id="cb82"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb82-1"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb82-1"></a><span class="kw">saveRDS</span>(pbmc_<span class="dv">4000</span>_cell, <span class="dt">file =</span> <span class="st">&quot;backup/05_025_pbmc_4000_cell_seurat.rds&quot;</span>)</span>
<span id="cb82-2"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb82-2"></a><span class="kw">saveRDS</span>(pbmc_<span class="dv">6000</span>_cell, <span class="dt">file =</span> <span class="st">&quot;backup/06_025_pbmc_6000_cell_seurat.rds&quot;</span>)</span></code></pre></div>
<p>Load the objects.</p>
<div class="sourceCode" id="cb83"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb83-1"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb83-1"></a>pbmc_<span class="dv">4000</span>_cell &lt;-<span class="st"> </span><span class="kw">readRDS</span>(<span class="dt">file =</span> <span class="st">&quot;backup/05_025_pbmc_4000_cell_seurat.rds&quot;</span>)</span>
<span id="cb83-2"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb83-2"></a>pbmc_<span class="dv">6000</span>_cell &lt;-<span class="st"> </span><span class="kw">readRDS</span>(<span class="dt">file =</span> <span class="st">&quot;backup/06_025_pbmc_6000_cell_seurat.rds&quot;</span>)</span></code></pre></div>
</div>
<div id="monocle-3-1" class="section level3">
<h3><span class="header-section-number">14.4.2</span> Monocle 3</h3>
<p>Load the objects.</p>
<div class="sourceCode" id="cb84"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb84-1"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb84-1"></a><span class="kw">rm</span>(<span class="dt">list =</span> <span class="kw">ls</span>())</span>
<span id="cb84-2"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb84-2"></a>pbmc_<span class="dv">4000</span> &lt;-<span class="st"> </span><span class="kw">readRDS</span>(<span class="dt">file =</span> <span class="st">&quot;backup/15_030_pbmc_4000_monocle3.rds&quot;</span>)</span>
<span id="cb84-3"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb84-3"></a>pbmc_<span class="dv">6000</span> &lt;-<span class="st"> </span><span class="kw">readRDS</span>(<span class="dt">file =</span> <span class="st">&quot;backup/16_030_pbmc_6000_monocle3.rds&quot;</span>)</span></code></pre></div>
<p>Group cells into clusters by using a community detection algorithm.</p>
<div class="sourceCode" id="cb85"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb85-1"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb85-1"></a>pbmc_<span class="dv">4000</span> &lt;-<span class="st"> </span><span class="kw">cluster_cells</span>(pbmc_<span class="dv">4000</span>, <span class="dt">resolution =</span> <span class="fl">1e-5</span>)</span>
<span id="cb85-2"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb85-2"></a>pbmc_<span class="dv">6000</span> &lt;-<span class="st"> </span><span class="kw">cluster_cells</span>(pbmc_<span class="dv">6000</span>, <span class="dt">resolution =</span> <span class="fl">1e-5</span>)</span></code></pre></div>
<p>Save the objects.</p>
<div class="sourceCode" id="cb86"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb86-1"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb86-1"></a><span class="kw">saveRDS</span>(pbmc_<span class="dv">4000</span>, <span class="dt">file =</span> <span class="st">&quot;backup/15_031_pbmc_4000_monocle3.rds&quot;</span>)</span>
<span id="cb86-2"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb86-2"></a><span class="kw">saveRDS</span>(pbmc_<span class="dv">6000</span>, <span class="dt">file =</span> <span class="st">&quot;backup/16_031_pbmc_6000_monocle3.rds&quot;</span>)</span></code></pre></div>
<p>Load the objects.</p>
<div class="sourceCode" id="cb87"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb87-1"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb87-1"></a>pbmc_<span class="dv">4000</span> &lt;-<span class="st"> </span><span class="kw">readRDS</span>(<span class="dt">file =</span> <span class="st">&quot;backup/15_031_pbmc_4000_monocle3.rds&quot;</span>)</span>
<span id="cb87-2"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb87-2"></a>pbmc_<span class="dv">6000</span> &lt;-<span class="st"> </span><span class="kw">readRDS</span>(<span class="dt">file =</span> <span class="st">&quot;backup/16_031_pbmc_6000_monocle3.rds&quot;</span>)</span></code></pre></div>
<p>Visualize the cell states on a UMAP space.</p>
<div class="sourceCode" id="cb88"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb88-1"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb88-1"></a>plot_manifold &lt;-<span class="st"> </span><span class="cf">function</span>(obj, plot_type, title, title_size, xlabel, ylabel){</span>
<span id="cb88-2"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb88-2"></a>  <span class="co"># ----------------------------------------</span></span>
<span id="cb88-3"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb88-3"></a>  <span class="co"># Definitions</span></span>
<span id="cb88-4"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb88-4"></a>  <span class="co"># ----------------------------------------</span></span>
<span id="cb88-5"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb88-5"></a>  df &lt;-<span class="st"> </span><span class="kw">data.frame</span>(</span>
<span id="cb88-6"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb88-6"></a>    <span class="dt">x =</span> <span class="kw">reducedDims</span>(obj)<span class="op">$</span>UMAP[,<span class="dv">1</span>],</span>
<span id="cb88-7"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb88-7"></a>    <span class="dt">y =</span> <span class="kw">reducedDims</span>(obj)<span class="op">$</span>UMAP[,<span class="dv">2</span>],</span>
<span id="cb88-8"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb88-8"></a>    <span class="dt">label =</span> obj<span class="op">@</span>clusters<span class="op">@</span>listData[[<span class="st">&quot;UMAP&quot;</span>]][[<span class="st">&quot;clusters&quot;</span>]]</span>
<span id="cb88-9"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb88-9"></a>  )</span>
<span id="cb88-10"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb88-10"></a>  <span class="kw">colnames</span>(df) &lt;-<span class="st"> </span><span class="kw">c</span>(<span class="st">&quot;x&quot;</span>, <span class="st">&quot;y&quot;</span>, <span class="st">&quot;label&quot;</span>)</span>
<span id="cb88-11"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb88-11"></a>  n_groups &lt;-<span class="st"> </span><span class="kw">length</span>(<span class="kw">unique</span>(df<span class="op">$</span>label))</span>
<span id="cb88-12"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb88-12"></a>  <span class="co"># ----------------------------------------</span></span>
<span id="cb88-13"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb88-13"></a>  <span class="co"># Color</span></span>
<span id="cb88-14"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb88-14"></a>  <span class="co"># ----------------------------------------</span></span>
<span id="cb88-15"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb88-15"></a>  ggColorHue &lt;-<span class="st"> </span><span class="cf">function</span>(n, <span class="dt">l=</span><span class="dv">65</span>){   <span class="co"># To produce default colors of ggplot2</span></span>
<span id="cb88-16"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb88-16"></a>    hues &lt;-<span class="st"> </span><span class="kw">seq</span>(<span class="dv">15</span>, <span class="dv">375</span>, <span class="dt">length=</span>n<span class="op">+</span><span class="dv">1</span>)</span>
<span id="cb88-17"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb88-17"></a>    <span class="kw">hcl</span>(<span class="dt">h=</span>hues, <span class="dt">l=</span>l, <span class="dt">c=</span><span class="dv">100</span>)[<span class="dv">1</span><span class="op">:</span>n]</span>
<span id="cb88-18"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb88-18"></a>  }</span>
<span id="cb88-19"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb88-19"></a>  my_colors &lt;-<span class="st"> </span><span class="kw">ggColorHue</span>(n_groups)</span>
<span id="cb88-20"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb88-20"></a>  <span class="co"># ----------------------------------------</span></span>
<span id="cb88-21"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb88-21"></a>  <span class="co"># Positions of labels</span></span>
<span id="cb88-22"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb88-22"></a>  <span class="co"># ----------------------------------------</span></span>
<span id="cb88-23"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb88-23"></a>  cluster_num &lt;-<span class="st"> </span><span class="kw">levels</span>(df<span class="op">$</span>label)</span>
<span id="cb88-24"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb88-24"></a>  cog &lt;-<span class="st"> </span><span class="kw">c</span>()  <span class="co"># Center of gravity</span></span>
<span id="cb88-25"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb88-25"></a>  <span class="cf">for</span>(i <span class="cf">in</span> cluster_num){</span>
<span id="cb88-26"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb88-26"></a>    cog &lt;-<span class="st"> </span><span class="kw">rbind</span>(cog, <span class="kw">c</span>(</span>
<span id="cb88-27"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb88-27"></a>      <span class="kw">mean</span>(df[<span class="kw">which</span>(df<span class="op">$</span>label <span class="op">==</span><span class="st"> </span>i),]<span class="op">$</span>x),</span>
<span id="cb88-28"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb88-28"></a>      <span class="kw">mean</span>(df[<span class="kw">which</span>(df<span class="op">$</span>label <span class="op">==</span><span class="st"> </span>i),]<span class="op">$</span>y),</span>
<span id="cb88-29"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb88-29"></a>      <span class="kw">as.character</span>(<span class="kw">unique</span>(df[<span class="kw">which</span>(df<span class="op">$</span>label <span class="op">==</span><span class="st"> </span>i),]<span class="op">$</span>label))</span>
<span id="cb88-30"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb88-30"></a>    ))</span>
<span id="cb88-31"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb88-31"></a>  }</span>
<span id="cb88-32"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb88-32"></a>  <span class="co"># ----------------------------------------</span></span>
<span id="cb88-33"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb88-33"></a>  <span class="co"># Plot</span></span>
<span id="cb88-34"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb88-34"></a>  <span class="co"># ----------------------------------------</span></span>
<span id="cb88-35"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb88-35"></a>  p &lt;-<span class="st"> </span><span class="kw">ggplot</span>() <span class="op">+</span></span>
<span id="cb88-36"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb88-36"></a><span class="st">    </span><span class="kw">geom_point</span>(<span class="kw">aes</span>(<span class="dt">x=</span>df[,<span class="dv">1</span>], <span class="dt">y=</span>df[,<span class="dv">2</span>], <span class="dt">color=</span>df[,<span class="dv">3</span>]), <span class="dt">size=</span><span class="fl">0.5</span>, <span class="dt">alpha=</span><span class="fl">1.0</span>) <span class="op">+</span></span>
<span id="cb88-37"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb88-37"></a><span class="st">    </span><span class="kw">labs</span>(<span class="dt">title=</span>title, <span class="dt">x=</span>xlabel, <span class="dt">y=</span>ylabel, <span class="dt">colour=</span><span class="st">&quot;&quot;</span>) <span class="op">+</span></span>
<span id="cb88-38"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb88-38"></a><span class="st">    </span><span class="kw">scale_colour_manual</span>(<span class="dt">values=</span>my_colors) <span class="op">+</span></span>
<span id="cb88-39"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb88-39"></a><span class="st">    </span><span class="kw">theme_classic</span>(<span class="dt">base_size=</span><span class="dv">20</span>, <span class="dt">base_family=</span><span class="st">&quot;Helvetica&quot;</span>) <span class="op">+</span></span>
<span id="cb88-40"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb88-40"></a><span class="st">    </span><span class="kw">theme</span>(<span class="dt">plot.title=</span><span class="kw">element_text</span>(<span class="dt">hjust=</span><span class="fl">0.5</span>, <span class="dt">size=</span>title_size),</span>
<span id="cb88-41"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb88-41"></a>      <span class="dt">plot.margin=</span><span class="kw">unit</span>(<span class="kw">c</span>(<span class="fl">0.5</span>,<span class="dv">2</span>,<span class="fl">0.5</span>,<span class="fl">0.5</span>), <span class="st">&quot;lines&quot;</span>), <span class="dt">legend.position=</span><span class="st">&quot;right&quot;</span>) <span class="op">+</span></span>
<span id="cb88-42"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb88-42"></a><span class="st">    </span><span class="kw">geom_text</span>(<span class="kw">aes</span>(<span class="dt">x=</span><span class="kw">as.numeric</span>(cog[,<span class="dv">1</span>]), <span class="dt">y=</span><span class="kw">as.numeric</span>(cog[,<span class="dv">2</span>]), <span class="dt">label=</span>cog[,<span class="dv">3</span>]),</span>
<span id="cb88-43"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb88-43"></a>      <span class="dt">size=</span><span class="dv">8</span>, <span class="dt">colour=</span><span class="st">&quot;black&quot;</span>)</span>
<span id="cb88-44"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb88-44"></a>  </span>
<span id="cb88-45"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb88-45"></a>  <span class="kw">return</span>(p)</span>
<span id="cb88-46"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb88-46"></a>}</span>
<span id="cb88-47"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb88-47"></a></span>
<span id="cb88-48"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb88-48"></a><span class="co"># ----------------------------------------</span></span>
<span id="cb88-49"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb88-49"></a><span class="co"># For `pbmc_4000`</span></span>
<span id="cb88-50"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb88-50"></a><span class="co"># ----------------------------------------</span></span>
<span id="cb88-51"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb88-51"></a>p &lt;-<span class="st"> </span><span class="kw">plot_manifold</span>(<span class="dt">obj =</span> pbmc_<span class="dv">4000</span>, <span class="dt">title =</span> <span class="st">&quot;pbmc_4000 (Monocle 3)&quot;</span>,</span>
<span id="cb88-52"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb88-52"></a>                   <span class="dt">title_size =</span> <span class="dv">24</span>, <span class="dt">xlabel =</span> <span class="st">&quot;UMAP_1&quot;</span>, <span class="dt">ylabel =</span> <span class="st">&quot;UMAP_2&quot;</span>)</span>
<span id="cb88-53"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb88-53"></a>filename &lt;-<span class="st"> &quot;figures/figure_15_0035.png&quot;</span></span>
<span id="cb88-54"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb88-54"></a><span class="kw">ggsave</span>(<span class="dt">file =</span> filename, <span class="dt">plot =</span> p, <span class="dt">dpi =</span> <span class="dv">300</span>, <span class="dt">width =</span> <span class="dv">5</span>, <span class="dt">height =</span> <span class="dv">4</span>)</span></code></pre></div>
<p><img src="figures/figure_15_0035.png" width="250px">
<img src="figures/figure_16_0035.png" width="250px"></p>
<p>Investigate the marker genes for each cluster.</p>
<div class="sourceCode" id="cb89"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb89-1"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb89-1"></a>pbmc_<span class="dv">4000</span>_marker &lt;-<span class="st"> </span><span class="kw">top_markers</span>(pbmc_<span class="dv">4000</span>, <span class="dt">group_cells_by =</span> <span class="st">&quot;cluster&quot;</span>,</span>
<span id="cb89-2"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb89-2"></a>                                <span class="dt">reference_cells =</span> <span class="dv">1000</span>, <span class="dt">cores =</span> <span class="dv">2</span>)</span>
<span id="cb89-3"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb89-3"></a>pbmc_<span class="dv">6000</span>_marker &lt;-<span class="st"> </span><span class="kw">top_markers</span>(pbmc_<span class="dv">6000</span>, <span class="dt">group_cells_by =</span> <span class="st">&quot;cluster&quot;</span>,</span>
<span id="cb89-4"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb89-4"></a>                                <span class="dt">reference_cells =</span> <span class="dv">1000</span>, <span class="dt">cores =</span> <span class="dv">2</span>)</span></code></pre></div>
<p>Find marker genes by ranking them according to the argument <code>pseudo_R2</code>.</p>
<div class="sourceCode" id="cb90"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb90-1"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb90-1"></a>pbmc_<span class="dv">4000</span>_topmarker &lt;-<span class="st"> </span>pbmc_<span class="dv">4000</span>_marker <span class="op">%&gt;%</span></span>
<span id="cb90-2"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb90-2"></a><span class="st">  </span><span class="kw">filter</span>(fraction_expressing <span class="op">&gt;=</span><span class="st"> </span><span class="fl">0.10</span>) <span class="op">%&gt;%</span><span class="st"> </span><span class="kw">group_by</span>(cell_group) <span class="op">%&gt;%</span></span>
<span id="cb90-3"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb90-3"></a><span class="st">  </span><span class="kw">top_n</span>(<span class="dv">50</span>, pseudo_R2)</span>
<span id="cb90-4"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb90-4"></a></span>
<span id="cb90-5"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb90-5"></a>pbmc_<span class="dv">6000</span>_topmarker &lt;-<span class="st"> </span>pbmc_<span class="dv">6000</span>_marker <span class="op">%&gt;%</span></span>
<span id="cb90-6"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb90-6"></a><span class="st">  </span><span class="kw">filter</span>(fraction_expressing <span class="op">&gt;=</span><span class="st"> </span><span class="fl">0.10</span>) <span class="op">%&gt;%</span><span class="st"> </span><span class="kw">group_by</span>(cell_group) <span class="op">%&gt;%</span></span>
<span id="cb90-7"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb90-7"></a><span class="st">  </span><span class="kw">top_n</span>(<span class="dv">50</span>, pseudo_R2)</span></code></pre></div>
<p>The following Monocle 3’s function <code>plot_genes_by_group()</code> visualizes the
expression levels of genes and fraction of cells.</p>
<div class="sourceCode" id="cb91"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb91-1"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb91-1"></a>top_genes &lt;-<span class="st"> </span><span class="kw">unique</span>(pbmc_<span class="dv">4000</span>_topmarker <span class="op">%&gt;%</span><span class="st"> </span><span class="kw">pull</span>(gene_id))</span>
<span id="cb91-2"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb91-2"></a>p &lt;-<span class="st"> </span><span class="kw">plot_genes_by_group</span>(pbmc_<span class="dv">4000</span>, top_genes, <span class="dt">group_cells_by =</span> <span class="st">&quot;cluster&quot;</span>,</span>
<span id="cb91-3"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb91-3"></a>                         <span class="dt">ordering_type =</span> <span class="st">&quot;cluster_row_col&quot;</span>, <span class="dt">max.size =</span> <span class="dv">3</span>)</span>
<span id="cb91-4"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb91-4"></a>filename &lt;-<span class="st"> &quot;figures/figure_15_0036.png&quot;</span></span>
<span id="cb91-5"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb91-5"></a><span class="kw">ggsave</span>(<span class="dt">file =</span> filename, <span class="dt">plot =</span> p, <span class="dt">dpi =</span> <span class="dv">300</span>, <span class="dt">width =</span> <span class="dv">5</span>, <span class="dt">height =</span> <span class="dv">14</span>)</span>
<span id="cb91-6"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb91-6"></a></span>
<span id="cb91-7"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb91-7"></a>top_genes &lt;-<span class="st"> </span><span class="kw">unique</span>(pbmc_<span class="dv">6000</span>_topmarker <span class="op">%&gt;%</span><span class="st"> </span><span class="kw">pull</span>(gene_id))</span>
<span id="cb91-8"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb91-8"></a>p &lt;-<span class="st"> </span><span class="kw">plot_genes_by_group</span>(pbmc_<span class="dv">6000</span>, top_genes, <span class="dt">group_cells_by =</span> <span class="st">&quot;cluster&quot;</span>,</span>
<span id="cb91-9"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb91-9"></a>                         <span class="dt">ordering_type =</span> <span class="st">&quot;cluster_row_col&quot;</span>, <span class="dt">max.size =</span> <span class="dv">3</span>)</span>
<span id="cb91-10"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb91-10"></a>filename &lt;-<span class="st"> &quot;figures/figure_16_0036.png&quot;</span></span>
<span id="cb91-11"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb91-11"></a><span class="kw">ggsave</span>(<span class="dt">file =</span> filename, <span class="dt">plot =</span> p, <span class="dt">dpi =</span> <span class="dv">300</span>, <span class="dt">width =</span> <span class="dv">5</span>, <span class="dt">height =</span> <span class="dv">14</span>)</span></code></pre></div>
<p><img src="figures/figure_15_0036.png" width="250px">
<img src="figures/figure_16_0036.png" width="250px"></p>
<p>Below are the process for using <code>Garnett</code>, which may help users infer cell
types.</p>
<div class="sourceCode" id="cb92"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb92-1"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb92-1"></a>pbmc_<span class="dv">4000</span>_garnett_markers &lt;-<span class="st"> </span>pbmc_<span class="dv">4000</span>_topmarker <span class="op">%&gt;%</span></span>
<span id="cb92-2"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb92-2"></a><span class="st">  </span><span class="kw">filter</span>(marker_test_q_value <span class="op">&lt;</span><span class="st"> </span><span class="fl">0.01</span> <span class="op">&amp;</span><span class="st"> </span>specificity <span class="op">&gt;=</span><span class="st"> </span><span class="fl">0.5</span>) <span class="op">%&gt;%</span></span>
<span id="cb92-3"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb92-3"></a><span class="st">  </span><span class="kw">group_by</span>(cell_group) <span class="op">%&gt;%</span><span class="st"> </span><span class="kw">top_n</span>(<span class="dv">50</span>, marker_score)</span>
<span id="cb92-4"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb92-4"></a>pbmc_<span class="dv">4000</span>_garnett_markers &lt;-<span class="st"> </span>pbmc_<span class="dv">4000</span>_garnett_markers <span class="op">%&gt;%</span></span>
<span id="cb92-5"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb92-5"></a><span class="st">  </span><span class="kw">group_by</span>(gene_short_name) <span class="op">%&gt;%</span><span class="st"> </span><span class="kw">filter</span>(<span class="kw">n</span>() <span class="op">==</span><span class="st"> </span><span class="dv">1</span>)</span>
<span id="cb92-6"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb92-6"></a></span>
<span id="cb92-7"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb92-7"></a>pbmc_<span class="dv">6000</span>_garnett_markers &lt;-<span class="st"> </span>pbmc_<span class="dv">6000</span>_topmarker <span class="op">%&gt;%</span></span>
<span id="cb92-8"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb92-8"></a><span class="st">  </span><span class="kw">filter</span>(marker_test_q_value <span class="op">&lt;</span><span class="st"> </span><span class="fl">0.01</span> <span class="op">&amp;</span><span class="st"> </span>specificity <span class="op">&gt;=</span><span class="st"> </span><span class="fl">0.5</span>) <span class="op">%&gt;%</span></span>
<span id="cb92-9"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb92-9"></a><span class="st">  </span><span class="kw">group_by</span>(cell_group) <span class="op">%&gt;%</span><span class="st"> </span><span class="kw">top_n</span>(<span class="dv">50</span>, marker_score)</span>
<span id="cb92-10"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb92-10"></a>pbmc_<span class="dv">6000</span>_garnett_markers &lt;-<span class="st"> </span>pbmc_<span class="dv">6000</span>_garnett_markers <span class="op">%&gt;%</span></span>
<span id="cb92-11"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb92-11"></a><span class="st">  </span><span class="kw">group_by</span>(gene_short_name) <span class="op">%&gt;%</span><span class="st"> </span><span class="kw">filter</span>(<span class="kw">n</span>() <span class="op">==</span><span class="st"> </span><span class="dv">1</span>)</span></code></pre></div>
<p>Users can check the marker genes of each cluster by the following command:</p>
<div class="sourceCode" id="cb93"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb93-1"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb93-1"></a><span class="kw">datatable</span>(pbmc_<span class="dv">4000</span>_topmarker, <span class="dt">rownames =</span> <span class="ot">FALSE</span>)</span>
<span id="cb93-2"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb93-2"></a><span class="kw">datatable</span>(pbmc_<span class="dv">6000</span>_topmarker, <span class="dt">rownames =</span> <span class="ot">FALSE</span>)</span>
<span id="cb93-3"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb93-3"></a><span class="kw">datatable</span>(pbmc_<span class="dv">4000</span>_garnett_markers, <span class="dt">rownames =</span> <span class="ot">FALSE</span>)</span>
<span id="cb93-4"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb93-4"></a><span class="kw">datatable</span>(pbmc_<span class="dv">6000</span>_garnett_markers, <span class="dt">rownames =</span> <span class="ot">FALSE</span>)</span></code></pre></div>
<ul>
<li><code>pbmc_4000</code> (Monocle 3)</li>
</ul>
<iframe src="figures/pbmc_4000_marker_monocle3.html" style="width:100%; height:300px;">
</iframe>
<ul>
<li><code>pbmc_4000</code> (Monocle 3 plus Garnett)</li>
</ul>
<iframe src="figures/pbmc_4000_marker_monocle3_garnett.html" style="width:100%; height:300px;">
</iframe>
<ul>
<li><code>pbmc_6000</code> (Monocle 3)</li>
</ul>
<iframe src="figures/pbmc_6000_marker_monocle3.html" style="width:100%; height:300px;">
</iframe>
<ul>
<li><code>pbmc_6000</code> (Monocle 3 plus Garnett)</li>
</ul>
<iframe src="figures/pbmc_6000_marker_monocle3_garnett.html" style="width:100%; height:300px;">
</iframe>
<p>Based on the above results and the manual investigation using GeneCards,
we annotated each group as follows:</p>
<ul>
<li>pbmc_4000</li>
</ul>
<pre><code>1: T cell         # CD3D, IL32, TRAC, CD3E
2: MC             # S100A8, LYZ
3: B cell         # CD79A, MS4A1, IGHM
4: NK/NKT cell    # NKG7, GNLY, CD160
5: B cell         # BCL11A, MZB1</code></pre>
<ul>
<li>pbmc_6000</li>
</ul>
<pre><code>1: T cell         # CD3D, TCF7, CD27, CD3E
2: MC             # S100A8, LYZ, CD14
3: B cell         # CD79B, VPREB3, BANK1
4: NK/NKT cell    # NKG7, GZMA, FGFBP2, GNLY
5: Immune cell    # MZB1, CD27</code></pre>
<p>Identify the cell types.</p>
<div class="sourceCode" id="cb96"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb96-1"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb96-1"></a><span class="co"># ----------------------------------------</span></span>
<span id="cb96-2"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb96-2"></a><span class="co"># For `pbmc_4000`</span></span>
<span id="cb96-3"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb96-3"></a><span class="co"># ----------------------------------------</span></span>
<span id="cb96-4"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb96-4"></a>tmp &lt;-<span class="st"> </span><span class="kw">as.integer</span>(<span class="kw">as.character</span>(<span class="kw">clusters</span>(pbmc_<span class="dv">4000</span>)))</span>
<span id="cb96-5"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb96-5"></a>tmp[tmp <span class="op">==</span><span class="st"> </span><span class="dv">1</span>] &lt;-<span class="st"> &quot;T&quot;</span></span>
<span id="cb96-6"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb96-6"></a>tmp[tmp <span class="op">==</span><span class="st"> </span><span class="dv">2</span>] &lt;-<span class="st"> &quot;MC&quot;</span></span>
<span id="cb96-7"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb96-7"></a>tmp[tmp <span class="op">==</span><span class="st"> </span><span class="dv">3</span>] &lt;-<span class="st"> &quot;B&quot;</span></span>
<span id="cb96-8"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb96-8"></a>tmp[tmp <span class="op">==</span><span class="st"> </span><span class="dv">4</span>] &lt;-<span class="st"> &quot;NK/NKT&quot;</span></span>
<span id="cb96-9"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb96-9"></a>tmp[tmp <span class="op">==</span><span class="st"> </span><span class="dv">5</span>] &lt;-<span class="st"> &quot;B&quot;</span></span>
<span id="cb96-10"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb96-10"></a>tmp &lt;-<span class="st"> </span><span class="kw">factor</span>(tmp, <span class="dt">levels =</span> <span class="kw">c</span>(<span class="st">&quot;T&quot;</span>, <span class="st">&quot;MC&quot;</span>, <span class="st">&quot;B&quot;</span>, <span class="st">&quot;NK/NKT&quot;</span>))</span>
<span id="cb96-11"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb96-11"></a><span class="kw">colData</span>(pbmc_<span class="dv">4000</span>)<span class="op">$</span>assigned_cell_type &lt;-<span class="st"> </span>tmp</span>
<span id="cb96-12"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb96-12"></a><span class="co"># ----------------------------------------</span></span>
<span id="cb96-13"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb96-13"></a><span class="co"># For `pbmc_6000`</span></span>
<span id="cb96-14"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb96-14"></a><span class="co"># ----------------------------------------</span></span>
<span id="cb96-15"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb96-15"></a>tmp &lt;-<span class="st"> </span><span class="kw">as.integer</span>(<span class="kw">as.character</span>(<span class="kw">clusters</span>(pbmc_<span class="dv">6000</span>)))</span>
<span id="cb96-16"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb96-16"></a>tmp[tmp <span class="op">==</span><span class="st"> </span><span class="dv">1</span>] &lt;-<span class="st"> &quot;T&quot;</span></span>
<span id="cb96-17"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb96-17"></a>tmp[tmp <span class="op">==</span><span class="st"> </span><span class="dv">2</span>] &lt;-<span class="st"> &quot;MC&quot;</span></span>
<span id="cb96-18"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb96-18"></a>tmp[tmp <span class="op">==</span><span class="st"> </span><span class="dv">3</span>] &lt;-<span class="st"> &quot;B&quot;</span></span>
<span id="cb96-19"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb96-19"></a>tmp[tmp <span class="op">==</span><span class="st"> </span><span class="dv">4</span>] &lt;-<span class="st"> &quot;NK/NKT&quot;</span></span>
<span id="cb96-20"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb96-20"></a>tmp[tmp <span class="op">==</span><span class="st"> </span><span class="dv">5</span>] &lt;-<span class="st"> &quot;u.s.&quot;</span></span>
<span id="cb96-21"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb96-21"></a>tmp &lt;-<span class="st"> </span><span class="kw">factor</span>(tmp, <span class="dt">levels =</span> <span class="kw">c</span>(<span class="st">&quot;T&quot;</span>, <span class="st">&quot;MC&quot;</span>, <span class="st">&quot;B&quot;</span>, <span class="st">&quot;NK/NKT&quot;</span>, <span class="st">&quot;u.s.&quot;</span>))</span>
<span id="cb96-22"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb96-22"></a><span class="kw">colData</span>(pbmc_<span class="dv">6000</span>)<span class="op">$</span>assigned_cell_type &lt;-<span class="st"> </span>tmp</span></code></pre></div>
<p>The following function <code>plot_manifold()</code> shows sample states on UMAP space.</p>
<div class="sourceCode" id="cb97"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb97-1"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb97-1"></a>plot_manifold &lt;-<span class="st"> </span><span class="cf">function</span>(obj, title, title_size, xlabel, ylabel){</span>
<span id="cb97-2"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb97-2"></a>  <span class="co"># ----------------------------------------</span></span>
<span id="cb97-3"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb97-3"></a>  <span class="co"># Definitions</span></span>
<span id="cb97-4"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb97-4"></a>  <span class="co"># ----------------------------------------</span></span>
<span id="cb97-5"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb97-5"></a>  df &lt;-<span class="st"> </span><span class="kw">data.frame</span>(</span>
<span id="cb97-6"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb97-6"></a>    <span class="dt">x =</span> <span class="kw">reducedDims</span>(obj)<span class="op">$</span>UMAP[,<span class="dv">1</span>],</span>
<span id="cb97-7"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb97-7"></a>    <span class="dt">y =</span> <span class="kw">reducedDims</span>(obj)<span class="op">$</span>UMAP[,<span class="dv">2</span>],</span>
<span id="cb97-8"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb97-8"></a>    <span class="dt">label =</span> obj<span class="op">@</span>clusters<span class="op">@</span>listData[[<span class="st">&quot;UMAP&quot;</span>]][[<span class="st">&quot;clusters&quot;</span>]],</span>
<span id="cb97-9"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb97-9"></a>    <span class="dt">anno =</span> obj<span class="op">@</span>colData<span class="op">@</span>listData[[<span class="st">&quot;assigned_cell_type&quot;</span>]]</span>
<span id="cb97-10"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb97-10"></a>  )</span>
<span id="cb97-11"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb97-11"></a>  <span class="kw">colnames</span>(df) &lt;-<span class="st"> </span><span class="kw">c</span>(<span class="st">&quot;x&quot;</span>, <span class="st">&quot;y&quot;</span>, <span class="st">&quot;label&quot;</span>, <span class="st">&quot;anno&quot;</span>)</span>
<span id="cb97-12"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb97-12"></a>  n_groups &lt;-<span class="st"> </span><span class="kw">length</span>(<span class="kw">unique</span>(obj<span class="op">$</span>seurat_clusters))</span>
<span id="cb97-13"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb97-13"></a>  <span class="co"># ----------------------------------------</span></span>
<span id="cb97-14"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb97-14"></a>  <span class="co"># Color</span></span>
<span id="cb97-15"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb97-15"></a>  <span class="co"># ----------------------------------------</span></span>
<span id="cb97-16"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb97-16"></a>  ggColorHue &lt;-<span class="st"> </span><span class="cf">function</span>(n, <span class="dt">l=</span><span class="dv">65</span>){   <span class="co"># To produce default colors of ggplot2</span></span>
<span id="cb97-17"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb97-17"></a>    hues &lt;-<span class="st"> </span><span class="kw">seq</span>(<span class="dv">15</span>, <span class="dv">375</span>, <span class="dt">length=</span>n<span class="op">+</span><span class="dv">1</span>)</span>
<span id="cb97-18"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb97-18"></a>    <span class="kw">hcl</span>(<span class="dt">h=</span>hues, <span class="dt">l=</span>l, <span class="dt">c=</span><span class="dv">100</span>)[<span class="dv">1</span><span class="op">:</span>n]</span>
<span id="cb97-19"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb97-19"></a>  }</span>
<span id="cb97-20"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb97-20"></a>  my_colors &lt;-<span class="st"> </span><span class="kw">ggColorHue</span>(<span class="dv">6</span>)</span>
<span id="cb97-21"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb97-21"></a>  tmp &lt;-<span class="st"> </span><span class="kw">as.character</span>(df<span class="op">$</span>anno)</span>
<span id="cb97-22"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb97-22"></a>  tmp[tmp <span class="op">==</span><span class="st"> &quot;T&quot;</span>] &lt;-<span class="st"> </span>my_colors[<span class="dv">1</span>]</span>
<span id="cb97-23"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb97-23"></a>  tmp[tmp <span class="op">==</span><span class="st"> &quot;MC&quot;</span>] &lt;-<span class="st"> </span>my_colors[<span class="dv">2</span>]</span>
<span id="cb97-24"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb97-24"></a>  tmp[tmp <span class="op">==</span><span class="st"> &quot;B&quot;</span>] &lt;-<span class="st"> </span>my_colors[<span class="dv">3</span>]</span>
<span id="cb97-25"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb97-25"></a>  tmp[tmp <span class="op">==</span><span class="st"> &quot;NK/NKT&quot;</span>] &lt;-<span class="st"> </span>my_colors[<span class="dv">4</span>]</span>
<span id="cb97-26"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb97-26"></a>  tmp[tmp <span class="op">==</span><span class="st"> &quot;u.s.&quot;</span>] &lt;-<span class="st"> </span>my_colors[<span class="dv">6</span>]</span>
<span id="cb97-27"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb97-27"></a>  tmp &lt;-<span class="st"> </span><span class="kw">factor</span>(tmp, <span class="dt">levels =</span> my_colors)</span>
<span id="cb97-28"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb97-28"></a>  df<span class="op">$</span>col &lt;-<span class="st"> </span>tmp</span>
<span id="cb97-29"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb97-29"></a>  <span class="co"># ----------------------------------------</span></span>
<span id="cb97-30"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb97-30"></a>  <span class="co"># Positions of labels</span></span>
<span id="cb97-31"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb97-31"></a>  <span class="co"># ----------------------------------------</span></span>
<span id="cb97-32"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb97-32"></a>  cluster_num &lt;-<span class="st"> </span><span class="kw">levels</span>(df<span class="op">$</span>label)</span>
<span id="cb97-33"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb97-33"></a>  cog &lt;-<span class="st"> </span><span class="kw">c</span>()  <span class="co"># Center of gravity</span></span>
<span id="cb97-34"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb97-34"></a>  <span class="cf">for</span>(i <span class="cf">in</span> cluster_num){</span>
<span id="cb97-35"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb97-35"></a>    cog &lt;-<span class="st"> </span><span class="kw">rbind</span>(cog, <span class="kw">c</span>(</span>
<span id="cb97-36"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb97-36"></a>      <span class="kw">mean</span>(df[<span class="kw">which</span>(df<span class="op">$</span>label <span class="op">==</span><span class="st"> </span>i),]<span class="op">$</span>x),</span>
<span id="cb97-37"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb97-37"></a>      <span class="kw">mean</span>(df[<span class="kw">which</span>(df<span class="op">$</span>label <span class="op">==</span><span class="st"> </span>i),]<span class="op">$</span>y),</span>
<span id="cb97-38"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb97-38"></a>      <span class="kw">as.character</span>(<span class="kw">unique</span>(df[<span class="kw">which</span>(df<span class="op">$</span>label <span class="op">==</span><span class="st"> </span>i),]<span class="op">$</span>anno))</span>
<span id="cb97-39"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb97-39"></a>    ))</span>
<span id="cb97-40"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb97-40"></a>  }</span>
<span id="cb97-41"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb97-41"></a>  <span class="co"># ----------------------------------------</span></span>
<span id="cb97-42"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb97-42"></a>  <span class="co"># Plot</span></span>
<span id="cb97-43"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb97-43"></a>  <span class="co"># ----------------------------------------</span></span>
<span id="cb97-44"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb97-44"></a>  p &lt;-<span class="st"> </span><span class="kw">ggplot</span>() <span class="op">+</span></span>
<span id="cb97-45"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb97-45"></a><span class="st">    </span><span class="kw">geom_point</span>(<span class="kw">aes</span>(<span class="dt">x=</span>df[,<span class="dv">1</span>], <span class="dt">y=</span>df[,<span class="dv">2</span>], <span class="dt">color=</span>df[,<span class="dv">5</span>]), <span class="dt">size=</span><span class="fl">0.5</span>, <span class="dt">alpha=</span><span class="fl">1.0</span>) <span class="op">+</span></span>
<span id="cb97-46"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb97-46"></a><span class="st">    </span><span class="kw">labs</span>(<span class="dt">title=</span>title, <span class="dt">x=</span>xlabel, <span class="dt">y=</span>ylabel, <span class="dt">colour=</span><span class="st">&quot;&quot;</span>) <span class="op">+</span></span>
<span id="cb97-47"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb97-47"></a><span class="st">    </span><span class="kw">scale_colour_manual</span>(<span class="dt">values=</span>my_colors) <span class="op">+</span></span>
<span id="cb97-48"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb97-48"></a><span class="st">    </span><span class="kw">theme_classic</span>(<span class="dt">base_size=</span><span class="dv">20</span>, <span class="dt">base_family=</span><span class="st">&quot;Helvetica&quot;</span>) <span class="op">+</span></span>
<span id="cb97-49"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb97-49"></a><span class="st">    </span><span class="kw">theme</span>(<span class="dt">plot.title=</span><span class="kw">element_text</span>(<span class="dt">hjust=</span><span class="fl">0.5</span>, <span class="dt">size=</span>title_size),</span>
<span id="cb97-50"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb97-50"></a>      <span class="dt">plot.margin=</span><span class="kw">unit</span>(<span class="kw">c</span>(<span class="fl">0.5</span>,<span class="dv">2</span>,<span class="fl">0.5</span>,<span class="fl">0.5</span>), <span class="st">&quot;lines&quot;</span>), <span class="dt">legend.position=</span><span class="st">&quot;none&quot;</span>) <span class="op">+</span></span>
<span id="cb97-51"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb97-51"></a><span class="st">    </span><span class="kw">geom_text</span>(<span class="kw">aes</span>(<span class="dt">x=</span><span class="kw">as.numeric</span>(cog[,<span class="dv">1</span>]), <span class="dt">y=</span><span class="kw">as.numeric</span>(cog[,<span class="dv">2</span>]), <span class="dt">label=</span>cog[,<span class="dv">3</span>]),</span>
<span id="cb97-52"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb97-52"></a>      <span class="dt">size=</span><span class="dv">8</span>, <span class="dt">colour=</span><span class="st">&quot;black&quot;</span>)</span>
<span id="cb97-53"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb97-53"></a>  </span>
<span id="cb97-54"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb97-54"></a>  <span class="kw">return</span>(p)</span>
<span id="cb97-55"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb97-55"></a>}</span>
<span id="cb97-56"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb97-56"></a></span>
<span id="cb97-57"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb97-57"></a><span class="co"># ----------------------------------------</span></span>
<span id="cb97-58"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb97-58"></a><span class="co"># For `pbmc_4000`</span></span>
<span id="cb97-59"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb97-59"></a><span class="co"># ----------------------------------------</span></span>
<span id="cb97-60"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb97-60"></a>p &lt;-<span class="st"> </span><span class="kw">plot_manifold</span>(<span class="dt">obj =</span> pbmc_<span class="dv">4000</span>, <span class="dt">title =</span> <span class="st">&quot;pbmc_4000 (Monocle 3)&quot;</span>,</span>
<span id="cb97-61"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb97-61"></a>                   <span class="dt">title_size =</span> <span class="dv">24</span>, <span class="dt">xlabel =</span> <span class="st">&quot;UMAP_1&quot;</span>, <span class="dt">ylabel =</span> <span class="st">&quot;UMAP_2&quot;</span>)</span>
<span id="cb97-62"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb97-62"></a>filename &lt;-<span class="st"> &quot;figures/figure_15_0037.png&quot;</span></span>
<span id="cb97-63"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb97-63"></a><span class="kw">ggsave</span>(<span class="dt">file =</span> filename, <span class="dt">plot =</span> p, <span class="dt">dpi =</span> <span class="dv">300</span>, <span class="dt">width =</span> <span class="fl">4.9</span>, <span class="dt">height =</span> <span class="dv">5</span>)</span></code></pre></div>
<p><img src="figures/figure_15_0037.png" width="250px">
<img src="figures/figure_16_0037.png" width="250px"></p>
<p>The numbers of cells are stored in a table.</p>
<div class="sourceCode" id="cb98"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb98-1"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb98-1"></a>identify_cell &lt;-<span class="st"> </span><span class="cf">function</span>(obj){</span>
<span id="cb98-2"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb98-2"></a>  tmp &lt;-<span class="st"> </span>obj<span class="op">@</span>colData<span class="op">@</span>listData[[<span class="st">&quot;assigned_cell_type&quot;</span>]]</span>
<span id="cb98-3"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb98-3"></a>  cells &lt;-<span class="st"> </span><span class="kw">levels</span>(tmp)</span>
<span id="cb98-4"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb98-4"></a>  df &lt;-<span class="st"> </span><span class="kw">c</span>()</span>
<span id="cb98-5"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb98-5"></a>  <span class="cf">for</span>(cell <span class="cf">in</span> cells){</span>
<span id="cb98-6"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb98-6"></a>    df &lt;-<span class="st"> </span><span class="kw">rbind</span>(df, <span class="kw">c</span>(cell, <span class="kw">length</span>(tmp[tmp<span class="op">==</span>cell])))</span>
<span id="cb98-7"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb98-7"></a>  }</span>
<span id="cb98-8"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb98-8"></a>  df &lt;-<span class="st"> </span><span class="kw">as.data.frame</span>(df)</span>
<span id="cb98-9"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb98-9"></a>  <span class="kw">colnames</span>(df) &lt;-<span class="st"> </span><span class="kw">c</span>(<span class="st">&quot;cell_type&quot;</span>, <span class="st">&quot;n&quot;</span>)</span>
<span id="cb98-10"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb98-10"></a>  <span class="kw">return</span>(df)</span>
<span id="cb98-11"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb98-11"></a>}</span>
<span id="cb98-12"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb98-12"></a></span>
<span id="cb98-13"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb98-13"></a>pbmc_<span class="dv">4000</span>_cell &lt;-<span class="st"> </span><span class="kw">identify_cell</span>(<span class="dt">obj =</span> pbmc_<span class="dv">4000</span>)</span>
<span id="cb98-14"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb98-14"></a>pbmc_<span class="dv">6000</span>_cell &lt;-<span class="st"> </span><span class="kw">identify_cell</span>(<span class="dt">obj =</span> pbmc_<span class="dv">6000</span>)</span></code></pre></div>
<p>Save the objects.</p>
<div class="sourceCode" id="cb99"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb99-1"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb99-1"></a><span class="kw">saveRDS</span>(pbmc_<span class="dv">4000</span>_cell, <span class="dt">file =</span> <span class="st">&quot;backup/05_035_pbmc_4000_cell_monocle3.rds&quot;</span>)</span>
<span id="cb99-2"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb99-2"></a><span class="kw">saveRDS</span>(pbmc_<span class="dv">6000</span>_cell, <span class="dt">file =</span> <span class="st">&quot;backup/06_035_pbmc_6000_cell_monocle3.rds&quot;</span>)</span></code></pre></div>
<p>Load the objects.</p>
<div class="sourceCode" id="cb100"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb100-1"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb100-1"></a>pbmc_<span class="dv">4000</span>_cell &lt;-<span class="st"> </span><span class="kw">readRDS</span>(<span class="dt">file =</span> <span class="st">&quot;backup/05_035_pbmc_4000_cell_monocle3.rds&quot;</span>)</span>
<span id="cb100-2"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb100-2"></a>pbmc_<span class="dv">6000</span>_cell &lt;-<span class="st"> </span><span class="kw">readRDS</span>(<span class="dt">file =</span> <span class="st">&quot;backup/06_035_pbmc_6000_cell_monocle3.rds&quot;</span>)</span></code></pre></div>
</div>
<div id="sc3-1" class="section level3">
<h3><span class="header-section-number">14.4.3</span> SC3</h3>
<p>Load the objects.</p>
<div class="sourceCode" id="cb101"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb101-1"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb101-1"></a><span class="kw">rm</span>(<span class="dt">list =</span> <span class="kw">ls</span>())</span>
<span id="cb101-2"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb101-2"></a>pbmc_<span class="dv">4000</span> &lt;-<span class="st"> </span><span class="kw">readRDS</span>(<span class="dt">file =</span> <span class="st">&quot;backup/15_040_pbmc_4000_sc3.rds&quot;</span>)</span>
<span id="cb101-3"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb101-3"></a>pbmc_<span class="dv">6000</span> &lt;-<span class="st"> </span><span class="kw">readRDS</span>(<span class="dt">file =</span> <span class="st">&quot;backup/16_040_pbmc_6000_sc3.rds&quot;</span>)</span></code></pre></div>
<p>Differentially expressed genes are calculated using non-parametric
Kruskal-Wallis test in <code>SC3</code> package.</p>
<div class="sourceCode" id="cb102"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb102-1"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb102-1"></a>p &lt;-<span class="st"> </span><span class="kw">sc3_plot_de_genes</span>(pbmc_<span class="dv">4000</span>, <span class="dt">k =</span> <span class="dv">5</span>)</span>
<span id="cb102-2"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb102-2"></a>filename &lt;-<span class="st"> &quot;figures/figure_15_0040.png&quot;</span></span>
<span id="cb102-3"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb102-3"></a><span class="kw">ggsave</span>(<span class="dt">file =</span> filename, <span class="dt">plot =</span> p, <span class="dt">dpi =</span> <span class="dv">300</span>, <span class="dt">width =</span> <span class="fl">8.5</span>, <span class="dt">height =</span> <span class="fl">8.5</span>)</span>
<span id="cb102-4"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb102-4"></a></span>
<span id="cb102-5"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb102-5"></a>p &lt;-<span class="st"> </span><span class="kw">sc3_plot_de_genes</span>(pbmc_<span class="dv">6000</span>, <span class="dt">k =</span> <span class="dv">5</span>)</span>
<span id="cb102-6"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb102-6"></a>filename &lt;-<span class="st"> &quot;figures/figure_16_0040.png&quot;</span></span>
<span id="cb102-7"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb102-7"></a><span class="kw">ggsave</span>(<span class="dt">file =</span> filename, <span class="dt">plot =</span> p, <span class="dt">dpi =</span> <span class="dv">300</span>, <span class="dt">width =</span> <span class="fl">8.5</span>, <span class="dt">height =</span> <span class="fl">8.5</span>)</span></code></pre></div>
<p><img src="figures/figure_15_0040.png" width="300px">
<img src="figures/figure_16_0040.png" width="300px"></p>
<p>Find marker genes by using <code>sc3_plot_markers()</code> in <code>SC3</code> package.
Although we changed the argument <code>k</code> from 2 to 8, the following values seemed
to be the best of all.</p>
<div class="sourceCode" id="cb103"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb103-1"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb103-1"></a>p &lt;-<span class="st"> </span><span class="kw">sc3_plot_markers</span>(pbmc_<span class="dv">4000</span>, <span class="dt">k =</span> <span class="dv">8</span>)</span>
<span id="cb103-2"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb103-2"></a>filename &lt;-<span class="st"> &quot;figures/figure_15_0041.png&quot;</span></span>
<span id="cb103-3"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb103-3"></a><span class="kw">ggsave</span>(<span class="dt">file =</span> filename, <span class="dt">plot =</span> p, <span class="dt">dpi =</span> <span class="dv">300</span>, <span class="dt">width =</span> <span class="fl">8.5</span>, <span class="dt">height =</span> <span class="fl">8.5</span>)</span>
<span id="cb103-4"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb103-4"></a></span>
<span id="cb103-5"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb103-5"></a>p &lt;-<span class="st"> </span><span class="kw">sc3_plot_markers</span>(pbmc_<span class="dv">6000</span>, <span class="dt">k =</span> <span class="dv">6</span>)</span>
<span id="cb103-6"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb103-6"></a>filename &lt;-<span class="st"> &quot;figures/figure_16_0041.png&quot;</span></span>
<span id="cb103-7"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb103-7"></a><span class="kw">ggsave</span>(<span class="dt">file =</span> filename, <span class="dt">plot =</span> p, <span class="dt">dpi =</span> <span class="dv">300</span>, <span class="dt">width =</span> <span class="fl">8.5</span>, <span class="dt">height =</span> <span class="fl">8.5</span>)</span></code></pre></div>
<p><img src="figures/figure_15_0041.png" width="300px">
<img src="figures/figure_16_0041.png" width="300px"></p>
<p>Based on the above results and the manual investigation using GeneCards,
we annotated each group as follows:</p>
<ul>
<li>pbmc_4000</li>
</ul>
<pre><code>1: NK/NKT cell    # GZMH, (TPD52L2)
2: u.s.
3: u.s.
4: u.s.
5: u.s.
6: u.s.
7: u.s.
8: u.s.</code></pre>
<ul>
<li>pbmc_6000</li>
</ul>
<pre><code>1: u.s.
2: u.s.
3: B cell         # CD79A, MS4A1
4: NK/NKT cell    # GZMK
5: Immune cell    # FCGR3A, MS4A7
6: NK/NKT cell    # GZMB, GZMA, GZMH</code></pre>
<p>Identify the cell types.</p>
<div class="sourceCode" id="cb106"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb106-1"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb106-1"></a><span class="co"># ----------------------------------------</span></span>
<span id="cb106-2"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb106-2"></a><span class="co"># For `pbmc_4000`</span></span>
<span id="cb106-3"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb106-3"></a><span class="co"># ----------------------------------------</span></span>
<span id="cb106-4"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb106-4"></a>tmp &lt;-<span class="st"> </span><span class="kw">as.integer</span>(<span class="kw">as.character</span>(pbmc_<span class="dv">4000</span><span class="op">@</span>colData<span class="op">@</span>listData[[<span class="st">&quot;sc3_8_clusters&quot;</span>]]))</span>
<span id="cb106-5"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb106-5"></a>tmp[tmp <span class="op">==</span><span class="st"> </span><span class="dv">1</span>] &lt;-<span class="st"> &quot;NK/NKT&quot;</span></span>
<span id="cb106-6"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb106-6"></a>tmp[tmp <span class="op">==</span><span class="st"> </span><span class="dv">2</span>] &lt;-<span class="st"> &quot;u.s.&quot;</span></span>
<span id="cb106-7"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb106-7"></a>tmp[tmp <span class="op">==</span><span class="st"> </span><span class="dv">3</span>] &lt;-<span class="st"> &quot;u.s.&quot;</span></span>
<span id="cb106-8"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb106-8"></a>tmp[tmp <span class="op">==</span><span class="st"> </span><span class="dv">4</span>] &lt;-<span class="st"> &quot;u.s.&quot;</span></span>
<span id="cb106-9"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb106-9"></a>tmp[tmp <span class="op">==</span><span class="st"> </span><span class="dv">5</span>] &lt;-<span class="st"> &quot;u.s.&quot;</span></span>
<span id="cb106-10"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb106-10"></a>tmp[tmp <span class="op">==</span><span class="st"> </span><span class="dv">6</span>] &lt;-<span class="st"> &quot;u.s.&quot;</span></span>
<span id="cb106-11"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb106-11"></a>tmp[tmp <span class="op">==</span><span class="st"> </span><span class="dv">7</span>] &lt;-<span class="st"> &quot;u.s.&quot;</span></span>
<span id="cb106-12"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb106-12"></a>tmp[tmp <span class="op">==</span><span class="st"> </span><span class="dv">8</span>] &lt;-<span class="st"> &quot;u.s.&quot;</span></span>
<span id="cb106-13"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb106-13"></a>tmp &lt;-<span class="st"> </span><span class="kw">factor</span>(tmp, <span class="dt">levels =</span> <span class="kw">c</span>(<span class="st">&quot;NK/NKT&quot;</span>, <span class="st">&quot;u.s.&quot;</span>))</span>
<span id="cb106-14"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb106-14"></a><span class="kw">colData</span>(pbmc_<span class="dv">4000</span>)<span class="op">$</span>assigned_cell_type &lt;-<span class="st"> </span>tmp</span>
<span id="cb106-15"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb106-15"></a><span class="co"># ----------------------------------------</span></span>
<span id="cb106-16"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb106-16"></a><span class="co"># For `pbmc_6000`</span></span>
<span id="cb106-17"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb106-17"></a><span class="co"># ----------------------------------------</span></span>
<span id="cb106-18"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb106-18"></a>tmp &lt;-<span class="st"> </span><span class="kw">as.integer</span>(<span class="kw">as.character</span>(pbmc_<span class="dv">6000</span><span class="op">@</span>colData<span class="op">@</span>listData[[<span class="st">&quot;sc3_6_clusters&quot;</span>]]))</span>
<span id="cb106-19"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb106-19"></a>tmp[tmp <span class="op">==</span><span class="st"> </span><span class="dv">1</span>] &lt;-<span class="st"> &quot;u.s.&quot;</span></span>
<span id="cb106-20"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb106-20"></a>tmp[tmp <span class="op">==</span><span class="st"> </span><span class="dv">2</span>] &lt;-<span class="st"> &quot;u.s.&quot;</span></span>
<span id="cb106-21"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb106-21"></a>tmp[tmp <span class="op">==</span><span class="st"> </span><span class="dv">3</span>] &lt;-<span class="st"> &quot;B&quot;</span></span>
<span id="cb106-22"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb106-22"></a>tmp[tmp <span class="op">==</span><span class="st"> </span><span class="dv">4</span>] &lt;-<span class="st"> &quot;NK/NKT&quot;</span></span>
<span id="cb106-23"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb106-23"></a>tmp[tmp <span class="op">==</span><span class="st"> </span><span class="dv">5</span>] &lt;-<span class="st"> &quot;u.s.&quot;</span></span>
<span id="cb106-24"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb106-24"></a>tmp[tmp <span class="op">==</span><span class="st"> </span><span class="dv">6</span>] &lt;-<span class="st"> &quot;NK/NKT&quot;</span></span>
<span id="cb106-25"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb106-25"></a>tmp &lt;-<span class="st"> </span><span class="kw">factor</span>(tmp, <span class="dt">levels =</span> <span class="kw">c</span>(<span class="st">&quot;B&quot;</span>, <span class="st">&quot;NK/NKT&quot;</span>, <span class="st">&quot;u.s.&quot;</span>))</span>
<span id="cb106-26"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb106-26"></a><span class="kw">colData</span>(pbmc_<span class="dv">6000</span>)<span class="op">$</span>assigned_cell_type &lt;-<span class="st"> </span>tmp</span></code></pre></div>
<p>Save the objects.</p>
<div class="sourceCode" id="cb107"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb107-1"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb107-1"></a><span class="kw">saveRDS</span>(pbmc_<span class="dv">4000</span>, <span class="dt">file =</span> <span class="st">&quot;backup/15_041_pbmc_4000_sc3.rds&quot;</span>)</span>
<span id="cb107-2"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb107-2"></a><span class="kw">saveRDS</span>(pbmc_<span class="dv">6000</span>, <span class="dt">file =</span> <span class="st">&quot;backup/16_041_pbmc_6000_sc3.rds&quot;</span>)</span></code></pre></div>
<p>Load the objects.</p>
<div class="sourceCode" id="cb108"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb108-1"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb108-1"></a>pbmc_<span class="dv">4000</span> &lt;-<span class="st"> </span><span class="kw">readRDS</span>(<span class="dt">file =</span> <span class="st">&quot;backup/15_041_pbmc_4000_sc3.rds&quot;</span>)</span>
<span id="cb108-2"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb108-2"></a>pbmc_<span class="dv">6000</span> &lt;-<span class="st"> </span><span class="kw">readRDS</span>(<span class="dt">file =</span> <span class="st">&quot;backup/16_041_pbmc_6000_sc3.rds&quot;</span>)</span></code></pre></div>
<p>The numbers of cells are stored in a table.</p>
<div class="sourceCode" id="cb109"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb109-1"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb109-1"></a>identify_cell &lt;-<span class="st"> </span><span class="cf">function</span>(obj){</span>
<span id="cb109-2"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb109-2"></a>  tmp &lt;-<span class="st"> </span><span class="kw">colData</span>(obj)<span class="op">$</span>assigned_cell_type</span>
<span id="cb109-3"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb109-3"></a>  cells &lt;-<span class="st"> </span><span class="kw">levels</span>(tmp)</span>
<span id="cb109-4"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb109-4"></a>  df &lt;-<span class="st"> </span><span class="kw">c</span>()</span>
<span id="cb109-5"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb109-5"></a>  <span class="cf">for</span>(cell <span class="cf">in</span> cells){</span>
<span id="cb109-6"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb109-6"></a>    df &lt;-<span class="st"> </span><span class="kw">rbind</span>(df, <span class="kw">c</span>(cell, <span class="kw">length</span>(tmp[tmp<span class="op">==</span>cell])))</span>
<span id="cb109-7"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb109-7"></a>  }</span>
<span id="cb109-8"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb109-8"></a>  df &lt;-<span class="st"> </span><span class="kw">as.data.frame</span>(df)</span>
<span id="cb109-9"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb109-9"></a>  <span class="kw">colnames</span>(df) &lt;-<span class="st"> </span><span class="kw">c</span>(<span class="st">&quot;cell_type&quot;</span>, <span class="st">&quot;n&quot;</span>)</span>
<span id="cb109-10"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb109-10"></a>  <span class="kw">return</span>(df)</span>
<span id="cb109-11"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb109-11"></a>}</span>
<span id="cb109-12"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb109-12"></a></span>
<span id="cb109-13"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb109-13"></a>pbmc_<span class="dv">4000</span>_cell &lt;-<span class="st"> </span><span class="kw">identify_cell</span>(<span class="dt">obj =</span> pbmc_<span class="dv">4000</span>)</span>
<span id="cb109-14"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb109-14"></a>pbmc_<span class="dv">6000</span>_cell &lt;-<span class="st"> </span><span class="kw">identify_cell</span>(<span class="dt">obj =</span> pbmc_<span class="dv">6000</span>)</span></code></pre></div>
<p>Save the objects.</p>
<div class="sourceCode" id="cb110"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb110-1"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb110-1"></a><span class="kw">saveRDS</span>(pbmc_<span class="dv">4000</span>_cell, <span class="dt">file =</span> <span class="st">&quot;backup/05_045_pbmc_4000_cell_sc3.rds&quot;</span>)</span>
<span id="cb110-2"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb110-2"></a><span class="kw">saveRDS</span>(pbmc_<span class="dv">6000</span>_cell, <span class="dt">file =</span> <span class="st">&quot;backup/06_045_pbmc_6000_cell_sc3.rds&quot;</span>)</span></code></pre></div>
<p>Load the objects.</p>
<div class="sourceCode" id="cb111"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb111-1"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb111-1"></a>pbmc_<span class="dv">4000</span>_cell &lt;-<span class="st"> </span><span class="kw">readRDS</span>(<span class="dt">file =</span> <span class="st">&quot;backup/05_045_pbmc_4000_cell_sc3.rds&quot;</span>)</span>
<span id="cb111-2"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb111-2"></a>pbmc_<span class="dv">6000</span>_cell &lt;-<span class="st"> </span><span class="kw">readRDS</span>(<span class="dt">file =</span> <span class="st">&quot;backup/06_045_pbmc_6000_cell_sc3.rds&quot;</span>)</span></code></pre></div>
</div>
<div id="asurat-1" class="section level3">
<h3><span class="header-section-number">14.4.4</span> ASURAT</h3>
<p>Load the objects.</p>
<div class="sourceCode" id="cb112"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb112-1"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb112-1"></a><span class="kw">rm</span>(<span class="dt">list =</span> <span class="kw">ls</span>())</span>
<span id="cb112-2"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb112-2"></a>pbmc_<span class="dv">4000</span> &lt;-<span class="st"> </span><span class="kw">readRDS</span>(<span class="dt">file =</span> <span class="st">&quot;backup/05_500_pbmc_4000_all.rds&quot;</span>)</span>
<span id="cb112-3"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb112-3"></a>pbmc_<span class="dv">6000</span> &lt;-<span class="st"> </span><span class="kw">readRDS</span>(<span class="dt">file =</span> <span class="st">&quot;backup/06_500_pbmc_6000_all.rds&quot;</span>)</span></code></pre></div>
<p>Visualize the sign scores upregulated in specific cell clusters, using violin
plots.
Please see Chapter 9 “ASURAT using Cell Ontology DB” for the details.</p>
<ul>
<li><code>pbmc_4000</code></li>
</ul>
<p><img src="figures/figure_05_0140.png" width="150px">
<img src="figures/figure_05_0141.png" width="150px">
<img src="figures/figure_05_0142.png" width="150px"></p>
<p><img src="figures/figure_05_0242.png" width="150px">
<img src="figures/figure_05_0243.png" width="150px"></p>
<p><img src="figures/figure_05_0340.png" width="150px">
<img src="figures/figure_05_0342.png" width="150px"></p>
<ul>
<li><code>pbmc_6000</code></li>
</ul>
<p><img src="figures/figure_06_0140.png" width="150px">
<img src="figures/figure_06_0141.png" width="150px">
<img src="figures/figure_06_0142.png" width="150px">
<img src="figures/figure_06_0143.png" width="150px"></p>
<p><img src="figures/figure_06_0240.png" width="150px">
<img src="figures/figure_06_0241.png" width="150px">
<img src="figures/figure_06_0243.png" width="150px"></p>
<p><img src="figures/figure_06_0340.png" width="150px">
<img src="figures/figure_06_0341.png" width="150px">
<img src="figures/figure_06_0342.png" width="150px"></p>
<p>Show the population sizes by two-variable bar graphs.</p>
<p><img src="figures/figure_05_0501.png" width="230px">
<img src="figures/figure_05_0502.png" width="230px"></p>
<p><img src="figures/figure_06_0501.png" width="230px">
<img src="figures/figure_06_0502.png" width="230px"></p>
<p>Based on the above results and the manual investigation using GeneCards,
we annotated each group as follows:</p>
<ul>
<li>pbmc_4000</li>
</ul>
<pre><code>1: T cell
2: Monocyte
3: B cell
4: NK/NKT cell
5: NK/NKT cell</code></pre>
<ul>
<li>pbmc_6000</li>
</ul>
<pre><code>1: T cell
2: Monocyte
3: NK/NKT cell
4: B cell
5: Dendritic cell</code></pre>
<p>Identify the cell types.</p>
<div class="sourceCode" id="cb115"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb115-1"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb115-1"></a><span class="co"># ----------------------------------------</span></span>
<span id="cb115-2"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb115-2"></a><span class="co"># For `pbmc_4000`</span></span>
<span id="cb115-3"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb115-3"></a><span class="co"># ----------------------------------------</span></span>
<span id="cb115-4"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb115-4"></a>tmp &lt;-<span class="st"> </span>pbmc_<span class="dv">4000</span>[[<span class="st">&quot;sample&quot;</span>]][[<span class="st">&quot;seuratFindClusters_CO_cell&quot;</span>]]</span>
<span id="cb115-5"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb115-5"></a>tmp[tmp <span class="op">==</span><span class="st"> </span><span class="dv">1</span>] &lt;-<span class="st"> &quot;T&quot;</span></span>
<span id="cb115-6"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb115-6"></a>tmp[tmp <span class="op">==</span><span class="st"> </span><span class="dv">2</span>] &lt;-<span class="st"> &quot;MC&quot;</span></span>
<span id="cb115-7"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb115-7"></a>tmp[tmp <span class="op">==</span><span class="st"> </span><span class="dv">3</span>] &lt;-<span class="st"> &quot;B&quot;</span></span>
<span id="cb115-8"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb115-8"></a>tmp[tmp <span class="op">==</span><span class="st"> </span><span class="dv">4</span>] &lt;-<span class="st"> &quot;NK/NKT&quot;</span></span>
<span id="cb115-9"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb115-9"></a>tmp[tmp <span class="op">==</span><span class="st"> </span><span class="dv">5</span>] &lt;-<span class="st"> &quot;NK/NKT&quot;</span></span>
<span id="cb115-10"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb115-10"></a>tmp &lt;-<span class="st"> </span><span class="kw">factor</span>(tmp, <span class="dt">levels =</span> <span class="kw">c</span>(<span class="st">&quot;T&quot;</span>, <span class="st">&quot;MC&quot;</span>, <span class="st">&quot;B&quot;</span>, <span class="st">&quot;NK/NKT&quot;</span>))</span>
<span id="cb115-11"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb115-11"></a>pbmc_<span class="dv">4000</span>[[<span class="st">&quot;sample&quot;</span>]][[<span class="st">&quot;assigned_cell_type&quot;</span>]] &lt;-<span class="st"> </span>tmp</span>
<span id="cb115-12"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb115-12"></a><span class="co"># ----------------------------------------</span></span>
<span id="cb115-13"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb115-13"></a><span class="co"># For `pbmc_6000`</span></span>
<span id="cb115-14"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb115-14"></a><span class="co"># ----------------------------------------</span></span>
<span id="cb115-15"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb115-15"></a>tmp &lt;-<span class="st"> </span>pbmc_<span class="dv">6000</span>[[<span class="st">&quot;sample&quot;</span>]][[<span class="st">&quot;seuratFindClusters_CO_cell&quot;</span>]]</span>
<span id="cb115-16"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb115-16"></a>tmp[tmp <span class="op">==</span><span class="st"> </span><span class="dv">1</span>] &lt;-<span class="st"> &quot;T&quot;</span></span>
<span id="cb115-17"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb115-17"></a>tmp[tmp <span class="op">==</span><span class="st"> </span><span class="dv">2</span>] &lt;-<span class="st"> &quot;MC&quot;</span></span>
<span id="cb115-18"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb115-18"></a>tmp[tmp <span class="op">==</span><span class="st"> </span><span class="dv">3</span>] &lt;-<span class="st"> &quot;B&quot;</span></span>
<span id="cb115-19"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb115-19"></a>tmp[tmp <span class="op">==</span><span class="st"> </span><span class="dv">4</span>] &lt;-<span class="st"> &quot;NK/NKT&quot;</span></span>
<span id="cb115-20"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb115-20"></a>tmp[tmp <span class="op">==</span><span class="st"> </span><span class="dv">5</span>] &lt;-<span class="st"> &quot;DC&quot;</span></span>
<span id="cb115-21"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb115-21"></a>tmp &lt;-<span class="st"> </span><span class="kw">factor</span>(tmp, <span class="dt">levels =</span> <span class="kw">c</span>(<span class="st">&quot;T&quot;</span>, <span class="st">&quot;MC&quot;</span>, <span class="st">&quot;B&quot;</span>, <span class="st">&quot;NK/NKT&quot;</span>, <span class="st">&quot;DC&quot;</span>))</span>
<span id="cb115-22"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb115-22"></a>pbmc_<span class="dv">6000</span>[[<span class="st">&quot;sample&quot;</span>]][[<span class="st">&quot;assigned_cell_type&quot;</span>]] &lt;-<span class="st"> </span>tmp</span></code></pre></div>
<p>The following function <code>plot_manifold()</code> shows sample states on UMAP space.</p>
<div class="sourceCode" id="cb116"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb116-1"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb116-1"></a>plot_manifold &lt;-<span class="st"> </span><span class="cf">function</span>(obj, title, title_size, xlabel, ylabel){</span>
<span id="cb116-2"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb116-2"></a>  <span class="co"># ----------------------------------------</span></span>
<span id="cb116-3"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb116-3"></a>  <span class="co"># Definitions</span></span>
<span id="cb116-4"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb116-4"></a>  <span class="co"># ----------------------------------------</span></span>
<span id="cb116-5"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb116-5"></a>  df &lt;-<span class="st"> </span><span class="kw">data.frame</span>(</span>
<span id="cb116-6"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb116-6"></a>    <span class="dt">x =</span> obj[[<span class="st">&quot;reduction&quot;</span>]][[<span class="st">&quot;umap&quot;</span>]][[<span class="st">&quot;CO&quot;</span>]][[<span class="st">&quot;cell&quot;</span>]][[<span class="st">&quot;layout&quot;</span>]][,<span class="dv">1</span>],</span>
<span id="cb116-7"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb116-7"></a>    <span class="dt">y =</span> obj[[<span class="st">&quot;reduction&quot;</span>]][[<span class="st">&quot;umap&quot;</span>]][[<span class="st">&quot;CO&quot;</span>]][[<span class="st">&quot;cell&quot;</span>]][[<span class="st">&quot;layout&quot;</span>]][,<span class="dv">2</span>],</span>
<span id="cb116-8"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb116-8"></a>    <span class="dt">label =</span> obj[[<span class="st">&quot;sample&quot;</span>]][[<span class="st">&quot;seuratFindClusters_CO_cell&quot;</span>]],</span>
<span id="cb116-9"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb116-9"></a>    <span class="dt">anno =</span> obj[[<span class="st">&quot;sample&quot;</span>]][[<span class="st">&quot;assigned_cell_type&quot;</span>]]</span>
<span id="cb116-10"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb116-10"></a>  )</span>
<span id="cb116-11"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb116-11"></a>  <span class="kw">colnames</span>(df) &lt;-<span class="st"> </span><span class="kw">c</span>(<span class="st">&quot;x&quot;</span>, <span class="st">&quot;y&quot;</span>, <span class="st">&quot;label&quot;</span>, <span class="st">&quot;anno&quot;</span>)</span>
<span id="cb116-12"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb116-12"></a>  n_groups &lt;-<span class="st"> </span><span class="kw">length</span>(<span class="kw">unique</span>(df<span class="op">$</span>label))</span>
<span id="cb116-13"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb116-13"></a>  <span class="co"># ----------------------------------------</span></span>
<span id="cb116-14"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb116-14"></a>  <span class="co"># Color</span></span>
<span id="cb116-15"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb116-15"></a>  <span class="co"># ----------------------------------------</span></span>
<span id="cb116-16"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb116-16"></a>  ggColorHue &lt;-<span class="st"> </span><span class="cf">function</span>(n, <span class="dt">l=</span><span class="dv">65</span>){   <span class="co"># To produce default colors of ggplot2</span></span>
<span id="cb116-17"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb116-17"></a>    hues &lt;-<span class="st"> </span><span class="kw">seq</span>(<span class="dv">15</span>, <span class="dv">375</span>, <span class="dt">length=</span>n<span class="op">+</span><span class="dv">1</span>)</span>
<span id="cb116-18"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb116-18"></a>    <span class="kw">hcl</span>(<span class="dt">h=</span>hues, <span class="dt">l=</span>l, <span class="dt">c=</span><span class="dv">100</span>)[<span class="dv">1</span><span class="op">:</span>n]</span>
<span id="cb116-19"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb116-19"></a>  }</span>
<span id="cb116-20"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb116-20"></a>  my_colors &lt;-<span class="st"> </span><span class="kw">ggColorHue</span>(<span class="dv">6</span>)</span>
<span id="cb116-21"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb116-21"></a>  tmp &lt;-<span class="st"> </span><span class="kw">as.character</span>(df<span class="op">$</span>anno)</span>
<span id="cb116-22"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb116-22"></a>  tmp[tmp <span class="op">==</span><span class="st"> &quot;T&quot;</span>] &lt;-<span class="st"> </span>my_colors[<span class="dv">1</span>]</span>
<span id="cb116-23"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb116-23"></a>  tmp[tmp <span class="op">==</span><span class="st"> &quot;MC&quot;</span>] &lt;-<span class="st"> </span>my_colors[<span class="dv">2</span>]</span>
<span id="cb116-24"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb116-24"></a>  tmp[tmp <span class="op">==</span><span class="st"> &quot;B&quot;</span>] &lt;-<span class="st"> </span>my_colors[<span class="dv">3</span>]</span>
<span id="cb116-25"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb116-25"></a>  tmp[tmp <span class="op">==</span><span class="st"> &quot;NK/NKT&quot;</span>] &lt;-<span class="st"> </span>my_colors[<span class="dv">4</span>]</span>
<span id="cb116-26"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb116-26"></a>  <span class="cf">if</span>(<span class="kw">length</span>(tmp[tmp <span class="op">==</span><span class="st"> &quot;DC&quot;</span>]) <span class="op">!=</span><span class="st"> </span><span class="dv">0</span>){</span>
<span id="cb116-27"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb116-27"></a>    tmp[tmp <span class="op">==</span><span class="st"> &quot;DC&quot;</span>] &lt;-<span class="st"> </span>my_colors[<span class="dv">5</span>]</span>
<span id="cb116-28"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb116-28"></a>  }</span>
<span id="cb116-29"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb116-29"></a>  tmp &lt;-<span class="st"> </span><span class="kw">factor</span>(tmp, <span class="dt">levels =</span> my_colors)</span>
<span id="cb116-30"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb116-30"></a>  df<span class="op">$</span>col &lt;-<span class="st"> </span>tmp</span>
<span id="cb116-31"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb116-31"></a>  <span class="co"># ----------------------------------------</span></span>
<span id="cb116-32"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb116-32"></a>  <span class="co"># Positions of labels</span></span>
<span id="cb116-33"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb116-33"></a>  <span class="co"># ----------------------------------------</span></span>
<span id="cb116-34"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb116-34"></a>  cluster_num &lt;-<span class="st"> </span><span class="kw">unique</span>(<span class="kw">sort</span>(df<span class="op">$</span>label))</span>
<span id="cb116-35"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb116-35"></a>  cog &lt;-<span class="st"> </span><span class="kw">c</span>()  <span class="co"># Center of gravity</span></span>
<span id="cb116-36"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb116-36"></a>  <span class="cf">for</span>(i <span class="cf">in</span> cluster_num){</span>
<span id="cb116-37"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb116-37"></a>    cog &lt;-<span class="st"> </span><span class="kw">rbind</span>(cog, <span class="kw">c</span>(</span>
<span id="cb116-38"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb116-38"></a>      <span class="kw">mean</span>(df[<span class="kw">which</span>(df<span class="op">$</span>label <span class="op">==</span><span class="st"> </span>i),]<span class="op">$</span>x),</span>
<span id="cb116-39"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb116-39"></a>      <span class="kw">mean</span>(df[<span class="kw">which</span>(df<span class="op">$</span>label <span class="op">==</span><span class="st"> </span>i),]<span class="op">$</span>y),</span>
<span id="cb116-40"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb116-40"></a>      <span class="kw">as.character</span>(<span class="kw">unique</span>(df[<span class="kw">which</span>(df<span class="op">$</span>label <span class="op">==</span><span class="st"> </span>i),]<span class="op">$</span>anno))</span>
<span id="cb116-41"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb116-41"></a>    ))</span>
<span id="cb116-42"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb116-42"></a>  }</span>
<span id="cb116-43"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb116-43"></a>  <span class="co"># ----------------------------------------</span></span>
<span id="cb116-44"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb116-44"></a>  <span class="co"># Plot</span></span>
<span id="cb116-45"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb116-45"></a>  <span class="co"># ----------------------------------------</span></span>
<span id="cb116-46"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb116-46"></a>  p &lt;-<span class="st"> </span><span class="kw">ggplot</span>() <span class="op">+</span></span>
<span id="cb116-47"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb116-47"></a><span class="st">    </span><span class="kw">geom_point</span>(<span class="kw">aes</span>(<span class="dt">x=</span>df[,<span class="dv">1</span>], <span class="dt">y=</span>df[,<span class="dv">2</span>], <span class="dt">color=</span>df[,<span class="dv">5</span>]), <span class="dt">size=</span><span class="fl">0.5</span>, <span class="dt">alpha=</span><span class="fl">1.0</span>) <span class="op">+</span></span>
<span id="cb116-48"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb116-48"></a><span class="st">    </span><span class="kw">labs</span>(<span class="dt">title=</span>title, <span class="dt">x=</span>xlabel, <span class="dt">y=</span>ylabel, <span class="dt">colour=</span><span class="st">&quot;&quot;</span>) <span class="op">+</span></span>
<span id="cb116-49"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb116-49"></a><span class="st">    </span><span class="kw">scale_colour_manual</span>(<span class="dt">values=</span>my_colors) <span class="op">+</span></span>
<span id="cb116-50"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb116-50"></a><span class="st">    </span><span class="kw">theme_classic</span>(<span class="dt">base_size=</span><span class="dv">20</span>, <span class="dt">base_family=</span><span class="st">&quot;Helvetica&quot;</span>) <span class="op">+</span></span>
<span id="cb116-51"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb116-51"></a><span class="st">    </span><span class="kw">theme</span>(<span class="dt">plot.title=</span><span class="kw">element_text</span>(<span class="dt">hjust=</span><span class="fl">0.5</span>, <span class="dt">size=</span>title_size),</span>
<span id="cb116-52"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb116-52"></a>      <span class="dt">plot.margin=</span><span class="kw">unit</span>(<span class="kw">c</span>(<span class="fl">0.5</span>,<span class="dv">2</span>,<span class="fl">0.5</span>,<span class="fl">0.5</span>), <span class="st">&quot;lines&quot;</span>), <span class="dt">legend.position=</span><span class="st">&quot;none&quot;</span>) <span class="op">+</span></span>
<span id="cb116-53"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb116-53"></a><span class="st">    </span><span class="kw">geom_text</span>(<span class="kw">aes</span>(<span class="dt">x=</span><span class="kw">as.numeric</span>(cog[,<span class="dv">1</span>]), <span class="dt">y=</span><span class="kw">as.numeric</span>(cog[,<span class="dv">2</span>]), <span class="dt">label=</span>cog[,<span class="dv">3</span>]),</span>
<span id="cb116-54"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb116-54"></a>      <span class="dt">size=</span><span class="dv">8</span>, <span class="dt">colour=</span><span class="st">&quot;black&quot;</span>)</span>
<span id="cb116-55"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb116-55"></a>  </span>
<span id="cb116-56"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb116-56"></a>  <span class="kw">return</span>(p)</span>
<span id="cb116-57"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb116-57"></a>}</span>
<span id="cb116-58"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb116-58"></a></span>
<span id="cb116-59"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb116-59"></a><span class="co"># ----------------------------------------</span></span>
<span id="cb116-60"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb116-60"></a><span class="co"># For `pbmc_4000`</span></span>
<span id="cb116-61"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb116-61"></a><span class="co"># ----------------------------------------</span></span>
<span id="cb116-62"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb116-62"></a>p &lt;-<span class="st"> </span><span class="kw">plot_manifold</span>(<span class="dt">obj =</span> pbmc_<span class="dv">4000</span>, <span class="dt">title =</span> <span class="st">&quot;pbmc_4000 (CO: cell)&quot;</span>,</span>
<span id="cb116-63"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb116-63"></a>                   <span class="dt">title_size =</span> <span class="dv">24</span>, <span class="dt">xlabel =</span> <span class="st">&quot;UMAP_1&quot;</span>, <span class="dt">ylabel =</span> <span class="st">&quot;UMAP_2&quot;</span>)</span>
<span id="cb116-64"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb116-64"></a>filename &lt;-<span class="st"> &quot;figures/figure_15_0065.png&quot;</span></span>
<span id="cb116-65"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb116-65"></a><span class="kw">ggsave</span>(<span class="dt">file =</span> filename, <span class="dt">plot =</span> p, <span class="dt">dpi =</span> <span class="dv">300</span>, <span class="dt">width =</span> <span class="fl">4.9</span>, <span class="dt">height =</span> <span class="dv">5</span>)</span></code></pre></div>
<p><img src="figures/figure_15_0065.png" width="250px">
<img src="figures/figure_16_0065.png" width="250px"></p>
<p>The numbers of cells are stored in a table.</p>
<div class="sourceCode" id="cb117"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb117-1"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb117-1"></a>identify_cell &lt;-<span class="st"> </span><span class="cf">function</span>(obj){</span>
<span id="cb117-2"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb117-2"></a>  tmp &lt;-<span class="st"> </span>obj[[<span class="st">&quot;sample&quot;</span>]][[<span class="st">&quot;assigned_cell_type&quot;</span>]]</span>
<span id="cb117-3"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb117-3"></a>  cells &lt;-<span class="st"> </span><span class="kw">levels</span>(tmp)</span>
<span id="cb117-4"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb117-4"></a>  df &lt;-<span class="st"> </span><span class="kw">c</span>()</span>
<span id="cb117-5"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb117-5"></a>  <span class="cf">for</span>(cell <span class="cf">in</span> cells){</span>
<span id="cb117-6"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb117-6"></a>    df &lt;-<span class="st"> </span><span class="kw">rbind</span>(df, <span class="kw">c</span>(cell, <span class="kw">length</span>(tmp[tmp<span class="op">==</span>cell])))</span>
<span id="cb117-7"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb117-7"></a>  }</span>
<span id="cb117-8"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb117-8"></a>  df &lt;-<span class="st"> </span><span class="kw">as.data.frame</span>(df)</span>
<span id="cb117-9"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb117-9"></a>  <span class="kw">colnames</span>(df) &lt;-<span class="st"> </span><span class="kw">c</span>(<span class="st">&quot;cell_type&quot;</span>, <span class="st">&quot;n&quot;</span>)</span>
<span id="cb117-10"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb117-10"></a>  <span class="kw">return</span>(df)</span>
<span id="cb117-11"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb117-11"></a>}</span>
<span id="cb117-12"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb117-12"></a></span>
<span id="cb117-13"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb117-13"></a>pbmc_<span class="dv">4000</span>_cell &lt;-<span class="st"> </span><span class="kw">identify_cell</span>(<span class="dt">obj =</span> pbmc_<span class="dv">4000</span>)</span>
<span id="cb117-14"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb117-14"></a>pbmc_<span class="dv">6000</span>_cell &lt;-<span class="st"> </span><span class="kw">identify_cell</span>(<span class="dt">obj =</span> pbmc_<span class="dv">6000</span>)</span></code></pre></div>
<p>Save the objects.</p>
<div class="sourceCode" id="cb118"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb118-1"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb118-1"></a><span class="kw">saveRDS</span>(pbmc_<span class="dv">4000</span>_cell, <span class="dt">file =</span> <span class="st">&quot;backup/05_065_pbmc_4000_cell_asurat.rds&quot;</span>)</span>
<span id="cb118-2"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb118-2"></a><span class="kw">saveRDS</span>(pbmc_<span class="dv">6000</span>_cell, <span class="dt">file =</span> <span class="st">&quot;backup/06_065_pbmc_6000_cell_asurat.rds&quot;</span>)</span></code></pre></div>
<p>Load the objects.</p>
<div class="sourceCode" id="cb119"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb119-1"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb119-1"></a>pbmc_<span class="dv">4000</span>_cell &lt;-<span class="st"> </span><span class="kw">readRDS</span>(<span class="dt">file =</span> <span class="st">&quot;backup/05_065_pbmc_4000_cell_asurat.rds&quot;</span>)</span>
<span id="cb119-2"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb119-2"></a>pbmc_<span class="dv">6000</span>_cell &lt;-<span class="st"> </span><span class="kw">readRDS</span>(<span class="dt">file =</span> <span class="st">&quot;backup/06_065_pbmc_6000_cell_asurat.rds&quot;</span>)</span></code></pre></div>
</div>
</div>
<div id="identifying-cell-types-in-pbmc_4000-and-pbmc_6000" class="section level2">
<h2><span class="header-section-number">14.5</span> Identifying cell types in <code>pbmc_4000</code> and <code>pbmc_6000</code></h2>
<p>Load the objects.</p>
<div class="sourceCode" id="cb120"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb120-1"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb120-1"></a><span class="kw">rm</span>(<span class="dt">list =</span> <span class="kw">ls</span>())</span>
<span id="cb120-2"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb120-2"></a>pbmc_<span class="dv">4000</span>_seurat &lt;-<span class="st"> </span><span class="kw">readRDS</span>(<span class="dt">file =</span> <span class="st">&quot;backup/05_025_pbmc_4000_cell_seurat.rds&quot;</span>)</span>
<span id="cb120-3"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb120-3"></a>pbmc_<span class="dv">6000</span>_seurat &lt;-<span class="st"> </span><span class="kw">readRDS</span>(<span class="dt">file =</span> <span class="st">&quot;backup/06_025_pbmc_6000_cell_seurat.rds&quot;</span>)</span>
<span id="cb120-4"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb120-4"></a>pbmc_<span class="dv">4000</span>_monocle &lt;-<span class="st"> </span><span class="kw">readRDS</span>(<span class="dt">file =</span> <span class="st">&quot;backup/05_035_pbmc_4000_cell_monocle3.rds&quot;</span>)</span>
<span id="cb120-5"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb120-5"></a>pbmc_<span class="dv">6000</span>_monocle &lt;-<span class="st"> </span><span class="kw">readRDS</span>(<span class="dt">file =</span> <span class="st">&quot;backup/06_035_pbmc_6000_cell_monocle3.rds&quot;</span>)</span>
<span id="cb120-6"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb120-6"></a>pbmc_<span class="dv">4000</span>_sc3 &lt;-<span class="st"> </span><span class="kw">readRDS</span>(<span class="dt">file =</span> <span class="st">&quot;backup/05_045_pbmc_4000_cell_sc3.rds&quot;</span>)</span>
<span id="cb120-7"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb120-7"></a>pbmc_<span class="dv">6000</span>_sc3 &lt;-<span class="st"> </span><span class="kw">readRDS</span>(<span class="dt">file =</span> <span class="st">&quot;backup/06_045_pbmc_6000_cell_sc3.rds&quot;</span>)</span>
<span id="cb120-8"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb120-8"></a>pbmc_<span class="dv">4000</span>_asurat &lt;-<span class="st"> </span><span class="kw">readRDS</span>(<span class="dt">file =</span> <span class="st">&quot;backup/05_065_pbmc_4000_cell_asurat.rds&quot;</span>)</span>
<span id="cb120-9"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb120-9"></a>pbmc_<span class="dv">6000</span>_asurat &lt;-<span class="st"> </span><span class="kw">readRDS</span>(<span class="dt">file =</span> <span class="st">&quot;backup/06_065_pbmc_6000_cell_asurat.rds&quot;</span>)</span></code></pre></div>
<p>The hypothetical results can be obtained from Cao et al., Front. Genet. (2020).</p>
<div class="sourceCode" id="cb121"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb121-1"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb121-1"></a>pbmc_<span class="dv">4000</span>_cao &lt;-<span class="st"> </span><span class="kw">data.frame</span>(</span>
<span id="cb121-2"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb121-2"></a>  <span class="dt">cell_type =</span> <span class="kw">c</span>(<span class="st">&quot;T&quot;</span>, <span class="st">&quot;MC&quot;</span>, <span class="st">&quot;B&quot;</span>, <span class="st">&quot;NK/NKT&quot;</span>, <span class="st">&quot;Megakaryocyte&quot;</span>),</span>
<span id="cb121-3"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb121-3"></a>  <span class="dt">n =</span> <span class="kw">c</span>(<span class="fl">0.4226</span>, <span class="fl">0.2707</span>, <span class="fl">0.1509</span>, <span class="fl">0.1546</span>, <span class="fl">0.0012</span>)</span>
<span id="cb121-4"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb121-4"></a>)</span>
<span id="cb121-5"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb121-5"></a>pbmc_<span class="dv">6000</span>_cao &lt;-<span class="st"> </span><span class="kw">data.frame</span>(</span>
<span id="cb121-6"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb121-6"></a>  <span class="dt">cell_type =</span> <span class="kw">c</span>(<span class="st">&quot;T&quot;</span>, <span class="st">&quot;MC&quot;</span>, <span class="st">&quot;B&quot;</span>, <span class="st">&quot;NK/NKT&quot;</span>, <span class="st">&quot;Megakaryocyte&quot;</span>),</span>
<span id="cb121-7"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb121-7"></a>  <span class="dt">n =</span> <span class="kw">c</span>(<span class="fl">0.4920</span>, <span class="fl">0.2652</span>, <span class="fl">0.1292</span>, <span class="fl">0.1102</span>, <span class="fl">0.0035</span>)</span>
<span id="cb121-8"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb121-8"></a>)</span></code></pre></div>
<p>Change the numbers of cells into ratios.</p>
<div class="sourceCode" id="cb122"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb122-1"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb122-1"></a>take_ratio &lt;-<span class="st"> </span><span class="cf">function</span>(df){</span>
<span id="cb122-2"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb122-2"></a>  df<span class="op">$</span>n &lt;-<span class="st"> </span><span class="kw">as.numeric</span>(df<span class="op">$</span>n) <span class="op">/</span><span class="st"> </span><span class="kw">sum</span>(<span class="kw">as.numeric</span>(df<span class="op">$</span>n))</span>
<span id="cb122-3"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb122-3"></a>  <span class="kw">return</span>(df)</span>
<span id="cb122-4"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb122-4"></a>}</span>
<span id="cb122-5"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb122-5"></a>pbmc_<span class="dv">4000</span>_seurat &lt;-<span class="st"> </span><span class="kw">take_ratio</span>(pbmc_<span class="dv">4000</span>_seurat)</span>
<span id="cb122-6"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb122-6"></a>pbmc_<span class="dv">4000</span>_monocle &lt;-<span class="st"> </span><span class="kw">take_ratio</span>(pbmc_<span class="dv">4000</span>_monocle)</span>
<span id="cb122-7"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb122-7"></a>pbmc_<span class="dv">4000</span>_sc3 &lt;-<span class="st"> </span><span class="kw">take_ratio</span>(pbmc_<span class="dv">4000</span>_sc3)</span>
<span id="cb122-8"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb122-8"></a>pbmc_<span class="dv">4000</span>_asurat &lt;-<span class="st"> </span><span class="kw">take_ratio</span>(pbmc_<span class="dv">4000</span>_asurat)</span>
<span id="cb122-9"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb122-9"></a></span>
<span id="cb122-10"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb122-10"></a>pbmc_<span class="dv">6000</span>_seurat &lt;-<span class="st"> </span><span class="kw">take_ratio</span>(pbmc_<span class="dv">6000</span>_seurat)</span>
<span id="cb122-11"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb122-11"></a>pbmc_<span class="dv">6000</span>_monocle &lt;-<span class="st"> </span><span class="kw">take_ratio</span>(pbmc_<span class="dv">6000</span>_monocle)</span>
<span id="cb122-12"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb122-12"></a>pbmc_<span class="dv">6000</span>_sc3 &lt;-<span class="st"> </span><span class="kw">take_ratio</span>(pbmc_<span class="dv">6000</span>_sc3)</span>
<span id="cb122-13"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb122-13"></a>pbmc_<span class="dv">6000</span>_asurat &lt;-<span class="st"> </span><span class="kw">take_ratio</span>(pbmc_<span class="dv">6000</span>_asurat)</span></code></pre></div>
<p>Add the names of algorithms.</p>
<div class="sourceCode" id="cb123"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb123-1"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb123-1"></a>pbmc_<span class="dv">4000</span>_seurat<span class="op">$</span>algorithm &lt;-<span class="st"> &quot;Seurat&quot;</span></span>
<span id="cb123-2"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb123-2"></a>pbmc_<span class="dv">4000</span>_monocle<span class="op">$</span>algorithm &lt;-<span class="st"> &quot;Monocle 3&quot;</span></span>
<span id="cb123-3"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb123-3"></a>pbmc_<span class="dv">4000</span>_sc3<span class="op">$</span>algorithm &lt;-<span class="st"> &quot;SC3&quot;</span></span>
<span id="cb123-4"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb123-4"></a>pbmc_<span class="dv">4000</span>_cao<span class="op">$</span>algorithm &lt;-<span class="st"> &quot;SCSA&quot;</span></span>
<span id="cb123-5"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb123-5"></a>pbmc_<span class="dv">4000</span>_asurat<span class="op">$</span>algorithm &lt;-<span class="st"> &quot;ASURAT&quot;</span></span>
<span id="cb123-6"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb123-6"></a></span>
<span id="cb123-7"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb123-7"></a>pbmc_<span class="dv">6000</span>_seurat<span class="op">$</span>algorithm &lt;-<span class="st"> &quot;Seurat&quot;</span></span>
<span id="cb123-8"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb123-8"></a>pbmc_<span class="dv">6000</span>_monocle<span class="op">$</span>algorithm &lt;-<span class="st"> &quot;Monocle 3&quot;</span></span>
<span id="cb123-9"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb123-9"></a>pbmc_<span class="dv">6000</span>_sc3<span class="op">$</span>algorithm &lt;-<span class="st"> &quot;SC3&quot;</span></span>
<span id="cb123-10"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb123-10"></a>pbmc_<span class="dv">6000</span>_cao<span class="op">$</span>algorithm &lt;-<span class="st"> &quot;SCSA&quot;</span></span>
<span id="cb123-11"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb123-11"></a>pbmc_<span class="dv">6000</span>_asurat<span class="op">$</span>algorithm &lt;-<span class="st"> &quot;ASURAT&quot;</span></span></code></pre></div>
<p>Concatenate all of the data.</p>
<div class="sourceCode" id="cb124"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb124-1"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb124-1"></a><span class="kw">colnames</span>(pbmc_<span class="dv">4000</span>_seurat)[<span class="dv">1</span>] &lt;-<span class="st"> &quot;cell_type&quot;</span></span>
<span id="cb124-2"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb124-2"></a><span class="kw">colnames</span>(pbmc_<span class="dv">6000</span>_seurat)[<span class="dv">1</span>] &lt;-<span class="st"> &quot;cell_type&quot;</span></span>
<span id="cb124-3"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb124-3"></a></span>
<span id="cb124-4"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb124-4"></a>df &lt;-<span class="st"> </span><span class="kw">rbind</span>(<span class="kw">rbind</span>(<span class="kw">rbind</span>(<span class="kw">rbind</span>(<span class="kw">rbind</span>(<span class="kw">rbind</span>(<span class="kw">rbind</span>(</span>
<span id="cb124-5"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb124-5"></a>  pbmc_<span class="dv">4000</span>_asurat, pbmc_<span class="dv">4000</span>_seurat), pbmc_<span class="dv">4000</span>_monocle), pbmc_<span class="dv">4000</span>_sc3),</span>
<span id="cb124-6"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb124-6"></a>  pbmc_<span class="dv">6000</span>_asurat), pbmc_<span class="dv">6000</span>_seurat), pbmc_<span class="dv">6000</span>_monocle), pbmc_<span class="dv">6000</span>_sc3)</span></code></pre></div>
<p>Rename the names of cell types.</p>
<div class="sourceCode" id="cb125"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb125-1"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb125-1"></a>df[<span class="kw">which</span>(df<span class="op">$</span>cell_type <span class="op">==</span><span class="st"> &quot;T&quot;</span>),]<span class="op">$</span>cell_type &lt;-<span class="st"> &quot;T cell&quot;</span></span>
<span id="cb125-2"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb125-2"></a>df[<span class="kw">which</span>(df<span class="op">$</span>cell_type <span class="op">==</span><span class="st"> &quot;MC&quot;</span>),]<span class="op">$</span>cell_type &lt;-<span class="st"> &quot;Monocyte&quot;</span></span>
<span id="cb125-3"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb125-3"></a>df[<span class="kw">which</span>(df<span class="op">$</span>cell_type <span class="op">==</span><span class="st"> &quot;B&quot;</span>),]<span class="op">$</span>cell_type &lt;-<span class="st"> &quot;B cell&quot;</span></span>
<span id="cb125-4"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb125-4"></a>df[<span class="kw">which</span>(df<span class="op">$</span>cell_type <span class="op">==</span><span class="st"> &quot;NK/NKT&quot;</span>),]<span class="op">$</span>cell_type &lt;-<span class="st"> &quot;NK/NKT cell&quot;</span></span>
<span id="cb125-5"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb125-5"></a>df[<span class="kw">which</span>(df<span class="op">$</span>cell_type <span class="op">==</span><span class="st"> &quot;DC&quot;</span>),]<span class="op">$</span>cell_type &lt;-<span class="st"> &quot;Dendritic cell&quot;</span></span>
<span id="cb125-6"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb125-6"></a>df[<span class="kw">which</span>(df<span class="op">$</span>cell_type <span class="op">==</span><span class="st"> &quot;u.s.&quot;</span>),]<span class="op">$</span>cell_type &lt;-<span class="st"> &quot;Unspecified&quot;</span></span>
<span id="cb125-7"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb125-7"></a><span class="co"># ----------------------------------------</span></span>
<span id="cb125-8"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb125-8"></a><span class="co"># Seurat</span></span>
<span id="cb125-9"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb125-9"></a><span class="co"># ----------------------------------------</span></span>
<span id="cb125-10"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb125-10"></a>dg &lt;-<span class="st"> </span>pbmc_<span class="dv">4000</span>_seurat</span>
<span id="cb125-11"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb125-11"></a>dg[<span class="kw">which</span>(dg<span class="op">$</span>cell_type <span class="op">==</span><span class="st"> &quot;T&quot;</span>),]<span class="op">$</span>cell_type &lt;-<span class="st"> &quot;T cell&quot;</span></span>
<span id="cb125-12"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb125-12"></a>dg[<span class="kw">which</span>(dg<span class="op">$</span>cell_type <span class="op">==</span><span class="st"> &quot;MC&quot;</span>),]<span class="op">$</span>cell_type &lt;-<span class="st"> &quot;Monocyte&quot;</span></span>
<span id="cb125-13"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb125-13"></a>dg[<span class="kw">which</span>(dg<span class="op">$</span>cell_type <span class="op">==</span><span class="st"> &quot;B&quot;</span>),]<span class="op">$</span>cell_type &lt;-<span class="st"> &quot;B cell&quot;</span></span>
<span id="cb125-14"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb125-14"></a>dg[<span class="kw">which</span>(dg<span class="op">$</span>cell_type <span class="op">==</span><span class="st"> &quot;NK/NKT&quot;</span>),]<span class="op">$</span>cell_type &lt;-<span class="st"> &quot;NK/NKT cell&quot;</span></span>
<span id="cb125-15"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb125-15"></a>dg[<span class="kw">which</span>(dg<span class="op">$</span>cell_type <span class="op">==</span><span class="st"> &quot;u.s.&quot;</span>),]<span class="op">$</span>cell_type &lt;-<span class="st"> &quot;Unspecified&quot;</span></span>
<span id="cb125-16"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb125-16"></a>pbmc_<span class="dv">4000</span>_seurat &lt;-<span class="st"> </span>dg</span>
<span id="cb125-17"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb125-17"></a></span>
<span id="cb125-18"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb125-18"></a>dg &lt;-<span class="st"> </span>pbmc_<span class="dv">6000</span>_seurat</span>
<span id="cb125-19"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb125-19"></a>dg[<span class="kw">which</span>(dg<span class="op">$</span>cell_type <span class="op">==</span><span class="st"> &quot;T&quot;</span>),]<span class="op">$</span>cell_type &lt;-<span class="st"> &quot;T cell&quot;</span></span>
<span id="cb125-20"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb125-20"></a>dg[<span class="kw">which</span>(dg<span class="op">$</span>cell_type <span class="op">==</span><span class="st"> &quot;MC&quot;</span>),]<span class="op">$</span>cell_type &lt;-<span class="st"> &quot;Monocyte&quot;</span></span>
<span id="cb125-21"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb125-21"></a>dg[<span class="kw">which</span>(dg<span class="op">$</span>cell_type <span class="op">==</span><span class="st"> &quot;B&quot;</span>),]<span class="op">$</span>cell_type &lt;-<span class="st"> &quot;B cell&quot;</span></span>
<span id="cb125-22"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb125-22"></a>dg[<span class="kw">which</span>(dg<span class="op">$</span>cell_type <span class="op">==</span><span class="st"> &quot;NK/NKT&quot;</span>),]<span class="op">$</span>cell_type &lt;-<span class="st"> &quot;NK/NKT cell&quot;</span></span>
<span id="cb125-23"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb125-23"></a>pbmc_<span class="dv">6000</span>_seurat &lt;-<span class="st"> </span>dg</span>
<span id="cb125-24"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb125-24"></a><span class="co"># ----------------------------------------</span></span>
<span id="cb125-25"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb125-25"></a><span class="co"># Monocle 3</span></span>
<span id="cb125-26"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb125-26"></a><span class="co"># ----------------------------------------</span></span>
<span id="cb125-27"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb125-27"></a>dg &lt;-<span class="st"> </span>pbmc_<span class="dv">4000</span>_monocle</span>
<span id="cb125-28"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb125-28"></a>dg[<span class="kw">which</span>(dg<span class="op">$</span>cell_type <span class="op">==</span><span class="st"> &quot;T&quot;</span>),]<span class="op">$</span>cell_type &lt;-<span class="st"> &quot;T cell&quot;</span></span>
<span id="cb125-29"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb125-29"></a>dg[<span class="kw">which</span>(dg<span class="op">$</span>cell_type <span class="op">==</span><span class="st"> &quot;MC&quot;</span>),]<span class="op">$</span>cell_type &lt;-<span class="st"> &quot;Monocyte&quot;</span></span>
<span id="cb125-30"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb125-30"></a>dg[<span class="kw">which</span>(dg<span class="op">$</span>cell_type <span class="op">==</span><span class="st"> &quot;B&quot;</span>),]<span class="op">$</span>cell_type &lt;-<span class="st"> &quot;B cell&quot;</span></span>
<span id="cb125-31"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb125-31"></a>dg[<span class="kw">which</span>(dg<span class="op">$</span>cell_type <span class="op">==</span><span class="st"> &quot;NK/NKT&quot;</span>),]<span class="op">$</span>cell_type &lt;-<span class="st"> &quot;NK/NKT cell&quot;</span></span>
<span id="cb125-32"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb125-32"></a>pbmc_<span class="dv">4000</span>_monocle &lt;-<span class="st"> </span>dg</span>
<span id="cb125-33"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb125-33"></a></span>
<span id="cb125-34"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb125-34"></a>dg &lt;-<span class="st"> </span>pbmc_<span class="dv">6000</span>_monocle</span>
<span id="cb125-35"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb125-35"></a>dg[<span class="kw">which</span>(dg<span class="op">$</span>cell_type <span class="op">==</span><span class="st"> &quot;T&quot;</span>),]<span class="op">$</span>cell_type &lt;-<span class="st"> &quot;T cell&quot;</span></span>
<span id="cb125-36"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb125-36"></a>dg[<span class="kw">which</span>(dg<span class="op">$</span>cell_type <span class="op">==</span><span class="st"> &quot;MC&quot;</span>),]<span class="op">$</span>cell_type &lt;-<span class="st"> &quot;Monocyte&quot;</span></span>
<span id="cb125-37"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb125-37"></a>dg[<span class="kw">which</span>(dg<span class="op">$</span>cell_type <span class="op">==</span><span class="st"> &quot;B&quot;</span>),]<span class="op">$</span>cell_type &lt;-<span class="st"> &quot;B cell&quot;</span></span>
<span id="cb125-38"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb125-38"></a>dg[<span class="kw">which</span>(dg<span class="op">$</span>cell_type <span class="op">==</span><span class="st"> &quot;NK/NKT&quot;</span>),]<span class="op">$</span>cell_type &lt;-<span class="st"> &quot;NK/NKT cell&quot;</span></span>
<span id="cb125-39"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb125-39"></a>dg[<span class="kw">which</span>(dg<span class="op">$</span>cell_type <span class="op">==</span><span class="st"> &quot;u.s.&quot;</span>),]<span class="op">$</span>cell_type &lt;-<span class="st"> &quot;Unspecified&quot;</span></span>
<span id="cb125-40"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb125-40"></a>pbmc_<span class="dv">6000</span>_monocle &lt;-<span class="st"> </span>dg</span>
<span id="cb125-41"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb125-41"></a><span class="co"># ----------------------------------------</span></span>
<span id="cb125-42"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb125-42"></a><span class="co"># SC3</span></span>
<span id="cb125-43"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb125-43"></a><span class="co"># ----------------------------------------</span></span>
<span id="cb125-44"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb125-44"></a>dg &lt;-<span class="st"> </span>pbmc_<span class="dv">4000</span>_sc3</span>
<span id="cb125-45"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb125-45"></a>dg[<span class="kw">which</span>(dg<span class="op">$</span>cell_type <span class="op">==</span><span class="st"> &quot;NK/NKT&quot;</span>),]<span class="op">$</span>cell_type &lt;-<span class="st"> &quot;NK/NKT cell&quot;</span></span>
<span id="cb125-46"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb125-46"></a>dg[<span class="kw">which</span>(dg<span class="op">$</span>cell_type <span class="op">==</span><span class="st"> &quot;u.s.&quot;</span>),]<span class="op">$</span>cell_type &lt;-<span class="st"> &quot;Unspecified&quot;</span></span>
<span id="cb125-47"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb125-47"></a>pbmc_<span class="dv">4000</span>_sc3 &lt;-<span class="st"> </span>dg</span>
<span id="cb125-48"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb125-48"></a></span>
<span id="cb125-49"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb125-49"></a>dg &lt;-<span class="st"> </span>pbmc_<span class="dv">6000</span>_sc3</span>
<span id="cb125-50"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb125-50"></a>dg[<span class="kw">which</span>(dg<span class="op">$</span>cell_type <span class="op">==</span><span class="st"> &quot;B&quot;</span>),]<span class="op">$</span>cell_type &lt;-<span class="st"> &quot;B cell&quot;</span></span>
<span id="cb125-51"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb125-51"></a>dg[<span class="kw">which</span>(dg<span class="op">$</span>cell_type <span class="op">==</span><span class="st"> &quot;NK/NKT&quot;</span>),]<span class="op">$</span>cell_type &lt;-<span class="st"> &quot;NK/NKT cell&quot;</span></span>
<span id="cb125-52"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb125-52"></a>dg[<span class="kw">which</span>(dg<span class="op">$</span>cell_type <span class="op">==</span><span class="st"> &quot;u.s.&quot;</span>),]<span class="op">$</span>cell_type &lt;-<span class="st"> &quot;Unspecified&quot;</span></span>
<span id="cb125-53"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb125-53"></a>pbmc_<span class="dv">6000</span>_sc3 &lt;-<span class="st"> </span>dg</span>
<span id="cb125-54"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb125-54"></a><span class="co"># ----------------------------------------</span></span>
<span id="cb125-55"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb125-55"></a><span class="co"># ASURAT</span></span>
<span id="cb125-56"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb125-56"></a><span class="co"># ----------------------------------------</span></span>
<span id="cb125-57"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb125-57"></a>dg &lt;-<span class="st"> </span>pbmc_<span class="dv">4000</span>_asurat</span>
<span id="cb125-58"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb125-58"></a>dg[<span class="kw">which</span>(dg<span class="op">$</span>cell_type <span class="op">==</span><span class="st"> &quot;T&quot;</span>),]<span class="op">$</span>cell_type &lt;-<span class="st"> &quot;T cell&quot;</span></span>
<span id="cb125-59"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb125-59"></a>dg[<span class="kw">which</span>(dg<span class="op">$</span>cell_type <span class="op">==</span><span class="st"> &quot;MC&quot;</span>),]<span class="op">$</span>cell_type &lt;-<span class="st"> &quot;Monocyte&quot;</span></span>
<span id="cb125-60"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb125-60"></a>dg[<span class="kw">which</span>(dg<span class="op">$</span>cell_type <span class="op">==</span><span class="st"> &quot;B&quot;</span>),]<span class="op">$</span>cell_type &lt;-<span class="st"> &quot;B cell&quot;</span></span>
<span id="cb125-61"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb125-61"></a>dg[<span class="kw">which</span>(dg<span class="op">$</span>cell_type <span class="op">==</span><span class="st"> &quot;NK/NKT&quot;</span>),]<span class="op">$</span>cell_type &lt;-<span class="st"> &quot;NK/NKT cell&quot;</span></span>
<span id="cb125-62"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb125-62"></a>pbmc_<span class="dv">4000</span>_asurat &lt;-<span class="st"> </span>dg</span>
<span id="cb125-63"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb125-63"></a></span>
<span id="cb125-64"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb125-64"></a>dg &lt;-<span class="st"> </span>pbmc_<span class="dv">6000</span>_asurat</span>
<span id="cb125-65"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb125-65"></a>dg[<span class="kw">which</span>(dg<span class="op">$</span>cell_type <span class="op">==</span><span class="st"> &quot;T&quot;</span>),]<span class="op">$</span>cell_type &lt;-<span class="st"> &quot;T cell&quot;</span></span>
<span id="cb125-66"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb125-66"></a>dg[<span class="kw">which</span>(dg<span class="op">$</span>cell_type <span class="op">==</span><span class="st"> &quot;MC&quot;</span>),]<span class="op">$</span>cell_type &lt;-<span class="st"> &quot;Monocyte&quot;</span></span>
<span id="cb125-67"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb125-67"></a>dg[<span class="kw">which</span>(dg<span class="op">$</span>cell_type <span class="op">==</span><span class="st"> &quot;B&quot;</span>),]<span class="op">$</span>cell_type &lt;-<span class="st"> &quot;B cell&quot;</span></span>
<span id="cb125-68"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb125-68"></a>dg[<span class="kw">which</span>(dg<span class="op">$</span>cell_type <span class="op">==</span><span class="st"> &quot;NK/NKT&quot;</span>),]<span class="op">$</span>cell_type &lt;-<span class="st"> &quot;NK/NKT cell&quot;</span></span>
<span id="cb125-69"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb125-69"></a>dg[<span class="kw">which</span>(dg<span class="op">$</span>cell_type <span class="op">==</span><span class="st"> &quot;DC&quot;</span>),]<span class="op">$</span>cell_type &lt;-<span class="st"> &quot;Dendritic cell&quot;</span></span>
<span id="cb125-70"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb125-70"></a>pbmc_<span class="dv">6000</span>_asurat &lt;-<span class="st"> </span>dg</span>
<span id="cb125-71"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb125-71"></a><span class="co"># ----------------------------------------</span></span>
<span id="cb125-72"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb125-72"></a><span class="co"># Cao</span></span>
<span id="cb125-73"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb125-73"></a><span class="co"># ----------------------------------------</span></span>
<span id="cb125-74"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb125-74"></a>dg &lt;-<span class="st"> </span>pbmc_<span class="dv">4000</span>_cao</span>
<span id="cb125-75"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb125-75"></a>dg[<span class="kw">which</span>(dg<span class="op">$</span>cell_type <span class="op">==</span><span class="st"> &quot;T&quot;</span>),]<span class="op">$</span>cell_type &lt;-<span class="st"> &quot;T cell&quot;</span></span>
<span id="cb125-76"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb125-76"></a>dg[<span class="kw">which</span>(dg<span class="op">$</span>cell_type <span class="op">==</span><span class="st"> &quot;MC&quot;</span>),]<span class="op">$</span>cell_type &lt;-<span class="st"> &quot;Monocyte&quot;</span></span>
<span id="cb125-77"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb125-77"></a>dg[<span class="kw">which</span>(dg<span class="op">$</span>cell_type <span class="op">==</span><span class="st"> &quot;B&quot;</span>),]<span class="op">$</span>cell_type &lt;-<span class="st"> &quot;B cell&quot;</span></span>
<span id="cb125-78"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb125-78"></a>dg[<span class="kw">which</span>(dg<span class="op">$</span>cell_type <span class="op">==</span><span class="st"> &quot;NK/NKT&quot;</span>),]<span class="op">$</span>cell_type &lt;-<span class="st"> &quot;NK/NKT cell&quot;</span></span>
<span id="cb125-79"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb125-79"></a>pbmc_<span class="dv">4000</span>_cao &lt;-<span class="st"> </span>dg</span>
<span id="cb125-80"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb125-80"></a></span>
<span id="cb125-81"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb125-81"></a>dg &lt;-<span class="st"> </span>pbmc_<span class="dv">6000</span>_cao</span>
<span id="cb125-82"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb125-82"></a>dg[<span class="kw">which</span>(dg<span class="op">$</span>cell_type <span class="op">==</span><span class="st"> &quot;T&quot;</span>),]<span class="op">$</span>cell_type &lt;-<span class="st"> &quot;T cell&quot;</span></span>
<span id="cb125-83"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb125-83"></a>dg[<span class="kw">which</span>(dg<span class="op">$</span>cell_type <span class="op">==</span><span class="st"> &quot;MC&quot;</span>),]<span class="op">$</span>cell_type &lt;-<span class="st"> &quot;Monocyte&quot;</span></span>
<span id="cb125-84"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb125-84"></a>dg[<span class="kw">which</span>(dg<span class="op">$</span>cell_type <span class="op">==</span><span class="st"> &quot;B&quot;</span>),]<span class="op">$</span>cell_type &lt;-<span class="st"> &quot;B cell&quot;</span></span>
<span id="cb125-85"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb125-85"></a>dg[<span class="kw">which</span>(dg<span class="op">$</span>cell_type <span class="op">==</span><span class="st"> &quot;NK/NKT&quot;</span>),]<span class="op">$</span>cell_type &lt;-<span class="st"> &quot;NK/NKT cell&quot;</span></span>
<span id="cb125-86"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb125-86"></a>pbmc_<span class="dv">6000</span>_cao &lt;-<span class="st"> </span>dg</span></code></pre></div>
<p>Show the annotation results.</p>
<div class="sourceCode" id="cb126"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb126-1"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb126-1"></a>xlabel =<span class="st"> &quot;&quot;</span></span>
<span id="cb126-2"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb126-2"></a>ylabel =<span class="st"> &quot;Ratio&quot;</span></span>
<span id="cb126-3"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb126-3"></a>my_colors &lt;-<span class="st"> </span><span class="kw">brewer.pal</span>(<span class="dv">5</span>, <span class="dt">name=</span><span class="st">&quot;Set1&quot;</span>)</span>
<span id="cb126-4"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb126-4"></a>size =<span class="st"> </span><span class="fl">4.5</span></span>
<span id="cb126-5"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb126-5"></a>alpha =<span class="st"> </span><span class="fl">0.85</span></span>
<span id="cb126-6"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb126-6"></a>jitter_size =<span class="st"> </span><span class="fl">0.2</span></span>
<span id="cb126-7"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb126-7"></a><span class="kw">set.seed</span>(<span class="dv">29</span>)</span>
<span id="cb126-8"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb126-8"></a>p &lt;-<span class="st"> </span><span class="kw">ggplot</span>() <span class="op">+</span></span>
<span id="cb126-9"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb126-9"></a><span class="st">  </span><span class="kw">geom_boxplot</span>(<span class="kw">aes</span>(<span class="dt">x=</span>df<span class="op">$</span>cell_type, <span class="dt">y=</span>df<span class="op">$</span>n), <span class="dt">color=</span><span class="st">&quot;white&quot;</span>, <span class="dt">alpha=</span><span class="dv">0</span>, <span class="dt">size=</span><span class="dv">0</span>,</span>
<span id="cb126-10"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb126-10"></a>               <span class="dt">coef =</span> <span class="fl">1e10</span>) <span class="op">+</span></span>
<span id="cb126-11"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb126-11"></a><span class="st">  </span><span class="co"># ----------------------------------------</span></span>
<span id="cb126-12"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb126-12"></a><span class="st">  </span><span class="co"># Cao et al., 2020.</span></span>
<span id="cb126-13"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb126-13"></a><span class="st">  </span><span class="co"># ----------------------------------------</span></span>
<span id="cb126-14"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb126-14"></a><span class="st">  </span><span class="kw">geom_jitter</span>(<span class="kw">aes</span>(<span class="dt">x=</span>pbmc_<span class="dv">4000</span>_cao<span class="op">$</span>cell_type, <span class="dt">y=</span>pbmc_<span class="dv">4000</span>_cao<span class="op">$</span>n),</span>
<span id="cb126-15"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb126-15"></a>              <span class="dt">color=</span><span class="st">&quot;black&quot;</span>, <span class="dt">size=</span>size, <span class="dt">shape=</span><span class="dv">15</span>, <span class="dt">alpha =</span> <span class="dv">1</span>,</span>
<span id="cb126-16"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb126-16"></a>              <span class="dt">position=</span><span class="kw">position_jitter</span>(<span class="dt">w =</span> jitter_size, <span class="dt">h =</span> <span class="dv">0</span>)) <span class="op">+</span></span>
<span id="cb126-17"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb126-17"></a><span class="st">  </span><span class="kw">geom_jitter</span>(<span class="kw">aes</span>(<span class="dt">x=</span>pbmc_<span class="dv">6000</span>_cao<span class="op">$</span>cell_type, <span class="dt">y=</span>pbmc_<span class="dv">6000</span>_cao<span class="op">$</span>n),</span>
<span id="cb126-18"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb126-18"></a>              <span class="dt">color=</span><span class="st">&quot;black&quot;</span>, <span class="dt">size=</span>size, <span class="dt">shape=</span><span class="dv">15</span>, <span class="dt">alpha =</span> <span class="dv">1</span>,</span>
<span id="cb126-19"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb126-19"></a>              <span class="dt">position=</span><span class="kw">position_jitter</span>(<span class="dt">w =</span> jitter_size, <span class="dt">h =</span> <span class="dv">0</span>)) <span class="op">+</span></span>
<span id="cb126-20"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb126-20"></a><span class="st">  </span><span class="co"># ----------------------------------------</span></span>
<span id="cb126-21"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb126-21"></a><span class="st">  </span><span class="co"># pbmc_4000</span></span>
<span id="cb126-22"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb126-22"></a><span class="st">  </span><span class="co"># ----------------------------------------</span></span>
<span id="cb126-23"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb126-23"></a><span class="st">  </span><span class="kw">geom_jitter</span>(<span class="kw">aes</span>(<span class="dt">x=</span>pbmc_<span class="dv">4000</span>_seurat<span class="op">$</span>cell_type, <span class="dt">y=</span>pbmc_<span class="dv">4000</span>_seurat<span class="op">$</span>n),</span>
<span id="cb126-24"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb126-24"></a>              <span class="dt">color=</span>my_colors[<span class="dv">1</span>], <span class="dt">size=</span>size, <span class="dt">shape=</span><span class="dv">16</span>, <span class="dt">alpha =</span> alpha,</span>
<span id="cb126-25"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb126-25"></a>              <span class="dt">position=</span><span class="kw">position_jitter</span>(<span class="dt">w =</span> jitter_size, <span class="dt">h =</span> <span class="dv">0</span>)) <span class="op">+</span></span>
<span id="cb126-26"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb126-26"></a><span class="st">  </span><span class="kw">geom_jitter</span>(<span class="kw">aes</span>(<span class="dt">x=</span>pbmc_<span class="dv">4000</span>_monocle<span class="op">$</span>cell_type, <span class="dt">y=</span>pbmc_<span class="dv">4000</span>_monocle<span class="op">$</span>n),</span>
<span id="cb126-27"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb126-27"></a>              <span class="dt">color=</span>my_colors[<span class="dv">2</span>], <span class="dt">size=</span>size, <span class="dt">shape=</span><span class="dv">16</span>, <span class="dt">alpha =</span> alpha,</span>
<span id="cb126-28"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb126-28"></a>              <span class="dt">position=</span><span class="kw">position_jitter</span>(<span class="dt">w =</span> jitter_size, <span class="dt">h =</span> <span class="dv">0</span>)) <span class="op">+</span></span>
<span id="cb126-29"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb126-29"></a><span class="st">  </span><span class="kw">geom_jitter</span>(<span class="kw">aes</span>(<span class="dt">x=</span>pbmc_<span class="dv">4000</span>_sc3<span class="op">$</span>cell_type, <span class="dt">y=</span>pbmc_<span class="dv">4000</span>_sc3<span class="op">$</span>n),</span>
<span id="cb126-30"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb126-30"></a>              <span class="dt">color=</span>my_colors[<span class="dv">3</span>], <span class="dt">size=</span>size, <span class="dt">shape=</span><span class="dv">16</span>, <span class="dt">alpha =</span> alpha,</span>
<span id="cb126-31"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb126-31"></a>              <span class="dt">position=</span><span class="kw">position_jitter</span>(<span class="dt">w =</span> jitter_size, <span class="dt">h =</span> <span class="dv">0</span>)) <span class="op">+</span></span>
<span id="cb126-32"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb126-32"></a><span class="st">  </span><span class="kw">geom_jitter</span>(<span class="kw">aes</span>(<span class="dt">x=</span>pbmc_<span class="dv">4000</span>_asurat<span class="op">$</span>cell_type, <span class="dt">y=</span>pbmc_<span class="dv">4000</span>_asurat<span class="op">$</span>n),</span>
<span id="cb126-33"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb126-33"></a>              <span class="dt">color=</span>my_colors[<span class="dv">5</span>], <span class="dt">size=</span>size, <span class="dt">shape=</span><span class="dv">16</span>, <span class="dt">alpha =</span> alpha,</span>
<span id="cb126-34"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb126-34"></a>              <span class="dt">position=</span><span class="kw">position_jitter</span>(<span class="dt">w =</span> jitter_size, <span class="dt">h =</span> <span class="dv">0</span>)) <span class="op">+</span></span>
<span id="cb126-35"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb126-35"></a><span class="st">  </span><span class="co"># ----------------------------------------</span></span>
<span id="cb126-36"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb126-36"></a><span class="st">  </span><span class="co"># pbmc_6000</span></span>
<span id="cb126-37"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb126-37"></a><span class="st">  </span><span class="co"># ----------------------------------------</span></span>
<span id="cb126-38"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb126-38"></a><span class="st">  </span><span class="kw">geom_jitter</span>(<span class="kw">aes</span>(<span class="dt">x=</span>pbmc_<span class="dv">6000</span>_seurat<span class="op">$</span>cell_type, <span class="dt">y=</span>pbmc_<span class="dv">6000</span>_seurat<span class="op">$</span>n),</span>
<span id="cb126-39"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb126-39"></a>              <span class="dt">color=</span>my_colors[<span class="dv">1</span>], <span class="dt">size=</span>size, <span class="dt">shape=</span><span class="dv">16</span>, <span class="dt">alpha =</span> alpha,</span>
<span id="cb126-40"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb126-40"></a>              <span class="dt">position=</span><span class="kw">position_jitter</span>(<span class="dt">w =</span> jitter_size, <span class="dt">h =</span> <span class="dv">0</span>)) <span class="op">+</span></span>
<span id="cb126-41"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb126-41"></a><span class="st">  </span><span class="kw">geom_jitter</span>(<span class="kw">aes</span>(<span class="dt">x=</span>pbmc_<span class="dv">6000</span>_monocle<span class="op">$</span>cell_type, <span class="dt">y=</span>pbmc_<span class="dv">6000</span>_monocle<span class="op">$</span>n),</span>
<span id="cb126-42"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb126-42"></a>              <span class="dt">color=</span>my_colors[<span class="dv">2</span>], <span class="dt">size=</span>size, <span class="dt">shape=</span><span class="dv">16</span>, <span class="dt">alpha =</span> alpha,</span>
<span id="cb126-43"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb126-43"></a>              <span class="dt">position=</span><span class="kw">position_jitter</span>(<span class="dt">w =</span> jitter_size, <span class="dt">h =</span> <span class="dv">0</span>)) <span class="op">+</span></span>
<span id="cb126-44"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb126-44"></a><span class="st">  </span><span class="kw">geom_jitter</span>(<span class="kw">aes</span>(<span class="dt">x=</span>pbmc_<span class="dv">6000</span>_sc3<span class="op">$</span>cell_type, <span class="dt">y=</span>pbmc_<span class="dv">6000</span>_sc3<span class="op">$</span>n),</span>
<span id="cb126-45"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb126-45"></a>              <span class="dt">color=</span>my_colors[<span class="dv">3</span>], <span class="dt">size=</span>size, <span class="dt">shape=</span><span class="dv">16</span>, <span class="dt">alpha =</span> alpha,</span>
<span id="cb126-46"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb126-46"></a>              <span class="dt">position=</span><span class="kw">position_jitter</span>(<span class="dt">w =</span> jitter_size, <span class="dt">h =</span> <span class="dv">0</span>)) <span class="op">+</span></span>
<span id="cb126-47"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb126-47"></a><span class="st">  </span><span class="kw">geom_jitter</span>(<span class="kw">aes</span>(<span class="dt">x=</span>pbmc_<span class="dv">6000</span>_asurat<span class="op">$</span>cell_type, <span class="dt">y=</span>pbmc_<span class="dv">6000</span>_asurat<span class="op">$</span>n),</span>
<span id="cb126-48"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb126-48"></a>              <span class="dt">color=</span>my_colors[<span class="dv">5</span>], <span class="dt">size=</span>size, <span class="dt">shape=</span><span class="dv">16</span>, <span class="dt">alpha =</span> alpha,</span>
<span id="cb126-49"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb126-49"></a>              <span class="dt">position=</span><span class="kw">position_jitter</span>(<span class="dt">w =</span> jitter_size, <span class="dt">h =</span> <span class="dv">0</span>)) <span class="op">+</span></span>
<span id="cb126-50"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb126-50"></a><span class="st">  </span><span class="kw">ylim</span>(<span class="dv">0</span>, <span class="fl">0.9</span>) <span class="op">+</span></span>
<span id="cb126-51"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb126-51"></a><span class="st">  </span><span class="kw">scale_x_discrete</span>(<span class="dt">guide=</span><span class="kw">guide_axis</span>(<span class="dt">angle=</span><span class="dv">45</span>)) <span class="op">+</span></span>
<span id="cb126-52"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb126-52"></a><span class="st">  </span><span class="kw">labs</span>(<span class="dt">title=</span><span class="st">&quot;&quot;</span>, <span class="dt">x=</span>xlabel, <span class="dt">y=</span>ylabel, <span class="dt">colour=</span><span class="st">&quot;&quot;</span>) <span class="op">+</span></span>
<span id="cb126-53"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb126-53"></a><span class="st">  </span><span class="kw">theme_minimal</span>(<span class="dt">base_size=</span><span class="dv">16</span>, <span class="dt">base_family=</span><span class="st">&quot;Helvetica&quot;</span>) <span class="op">+</span></span>
<span id="cb126-54"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb126-54"></a><span class="st">  </span><span class="kw">theme</span>(<span class="dt">plot.title=</span><span class="kw">element_text</span>(<span class="dt">hjust=</span><span class="fl">0.5</span>),</span>
<span id="cb126-55"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb126-55"></a>        <span class="dt">panel.grid=</span><span class="kw">element_line</span>(<span class="dt">color=</span><span class="st">&quot;gray80&quot;</span>),</span>
<span id="cb126-56"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb126-56"></a>        <span class="dt">plot.margin=</span><span class="kw">unit</span>(<span class="kw">c</span>(<span class="fl">0.5</span>,<span class="dv">2</span>,<span class="fl">0.5</span>,<span class="fl">0.5</span>), <span class="st">&quot;lines&quot;</span>), <span class="dt">legend.position=</span><span class="st">&quot;right&quot;</span>)</span>
<span id="cb126-57"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb126-57"></a></span>
<span id="cb126-58"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb126-58"></a>filename &lt;-<span class="st"> &quot;figures/figure_95_0001.png&quot;</span></span>
<span id="cb126-59"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cb126-59"></a><span class="kw">ggsave</span>(<span class="dt">file =</span> filename, <span class="dt">plot =</span> p, <span class="dt">dpi =</span> <span class="dv">300</span>, <span class="dt">width =</span> <span class="fl">4.4</span>, <span class="dt">height =</span> <span class="fl">4.5</span>)</span></code></pre></div>
<p><img src="figures/figure_95_0001.png" width="400px"></p>

</div>
</div>
            </section>

          </div>
        </div>
      </div>
<a href="multiple-sign-analysis-by-concatenating-do-co-go-kegg-and-reactome.html" class="navigation navigation-prev " aria-label="Previous page"><i class="fa fa-angle-left"></i></a>
<a href="appendix-b-automatically-tuning-the-asurats-parameters.html" class="navigation navigation-next " aria-label="Next page"><i class="fa fa-angle-right"></i></a>
    </div>
  </div>
<script src="libs/gitbook-2.6.7/js/app.min.js"></script>
<script src="libs/gitbook-2.6.7/js/lunr.js"></script>
<script src="libs/gitbook-2.6.7/js/clipboard.min.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-search.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-sharing.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-fontsettings.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-bookdown.js"></script>
<script src="libs/gitbook-2.6.7/js/jquery.highlight.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-clipboard.js"></script>
<script>
gitbook.require(["gitbook"], function(gitbook) {
gitbook.start({
"sharing": {
"github": false,
"facebook": true,
"twitter": true,
"linkedin": false,
"weibo": false,
"instapaper": false,
"vk": false,
"all": ["facebook", "twitter", "linkedin", "weibo", "instapaper"]
},
"fontsettings": {
"theme": "white",
"family": "sans",
"size": 2
},
"edit": {
"link": null,
"text": null
},
"history": {
"link": null,
"text": null
},
"view": {
"link": null,
"text": null
},
"download": ["Introduction to ASURAT (ver. 0.0.0.9000).pdf", "Introduction to ASURAT (ver. 0.0.0.9000).epub"],
"toc": {
"collapse": "subsection"
}
});
});
</script>

</body>

</html>
