<!DOCTYPE html>
<html lang="" xml:lang="">
<head>

  <meta charset="utf-8" />
  <meta http-equiv="X-UA-Compatible" content="IE=edge" />
  <title>Chapter 18 Appendix F: creating figures | Introduction to ASURAT (version 0.0.0.9001)</title>
  <meta name="description" content="This is a developmental version of ASURAT." />
  <meta name="generator" content="bookdown 0.23 and GitBook 2.6.7" />

  <meta property="og:title" content="Chapter 18 Appendix F: creating figures | Introduction to ASURAT (version 0.0.0.9001)" />
  <meta property="og:type" content="book" />
  
  
  <meta property="og:description" content="This is a developmental version of ASURAT." />
  

  <meta name="twitter:card" content="summary" />
  <meta name="twitter:title" content="Chapter 18 Appendix F: creating figures | Introduction to ASURAT (version 0.0.0.9001)" />
  
  <meta name="twitter:description" content="This is a developmental version of ASURAT." />
  

<meta name="author" content="Keita Iida" />


<meta name="date" content="2021-09-09" />

  <meta name="viewport" content="width=device-width, initial-scale=1" />
  <meta name="apple-mobile-web-app-capable" content="yes" />
  <meta name="apple-mobile-web-app-status-bar-style" content="black" />
  
  
<link rel="prev" href="appendix-e-analyzing-scrna-seq-data-using-asurat.html"/>

<script src="libs/header-attrs-2.10/header-attrs.js"></script>
<script src="libs/jquery-3.6.0/jquery-3.6.0.min.js"></script>
<link href="libs/gitbook-2.6.7/css/style.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-table.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-bookdown.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-highlight.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-search.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-fontsettings.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-clipboard.css" rel="stylesheet" />









<link href="libs/anchor-sections-1.0.1/anchor-sections.css" rel="stylesheet" />
<script src="libs/anchor-sections-1.0.1/anchor-sections.js"></script>


<style type="text/css">
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } /* Alert */
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code span.at { color: #7d9029; } /* Attribute */
code span.bn { color: #40a070; } /* BaseN */
code span.bu { } /* BuiltIn */
code span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code span.ch { color: #4070a0; } /* Char */
code span.cn { color: #880000; } /* Constant */
code span.co { color: #60a0b0; font-style: italic; } /* Comment */
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code span.do { color: #ba2121; font-style: italic; } /* Documentation */
code span.dt { color: #902000; } /* DataType */
code span.dv { color: #40a070; } /* DecVal */
code span.er { color: #ff0000; font-weight: bold; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #40a070; } /* Float */
code span.fu { color: #06287e; } /* Function */
code span.im { } /* Import */
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
code span.kw { color: #007020; font-weight: bold; } /* Keyword */
code span.op { color: #666666; } /* Operator */
code span.ot { color: #007020; } /* Other */
code span.pp { color: #bc7a00; } /* Preprocessor */
code span.sc { color: #4070a0; } /* SpecialChar */
code span.ss { color: #bb6688; } /* SpecialString */
code span.st { color: #4070a0; } /* String */
code span.va { color: #19177c; } /* Variable */
code span.vs { color: #4070a0; } /* VerbatimString */
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */
</style>


<link rel="stylesheet" href="style.css" type="text/css" />
</head>

<body>



  <div class="book without-animation with-summary font-size-2 font-family-1" data-basepath=".">

    <div class="book-summary">
      <nav role="navigation">

<ul class="summary">
<li><a href="./">A Minimal Book Example</a></li>

<li class="divider"></li>
<li class="chapter" data-level="1" data-path="index.html"><a href="index.html"><i class="fa fa-check"></i><b>1</b> Overview</a>
<ul>
<li class="chapter" data-level="1.1" data-path="index.html"><a href="index.html#license-and-description"><i class="fa fa-check"></i><b>1.1</b> License and description</a></li>
<li class="chapter" data-level="1.2" data-path="index.html"><a href="index.html#computational-environment"><i class="fa fa-check"></i><b>1.2</b> Computational environment</a></li>
<li class="chapter" data-level="1.3" data-path="index.html"><a href="index.html#installation"><i class="fa fa-check"></i><b>1.3</b> Installation</a></li>
</ul></li>
<li class="chapter" data-level="2" data-path="dataset.html"><a href="dataset.html"><i class="fa fa-check"></i><b>2</b> Preparing data sets</a>
<ul>
<li class="chapter" data-level="2.1" data-path="dataset.html"><a href="dataset.html#case-i-10x-data-stewart-et-al.-nat-cancer-2020"><i class="fa fa-check"></i><b>2.1</b> Case I: 10x data (Stewart et al., Nat Cancer, 2020)</a></li>
<li class="chapter" data-level="2.2" data-path="dataset.html"><a href="dataset.html#case-ii-pdac-a-st-and-indrop-moncada-et-al.-nat-biotechnol-2020"><i class="fa fa-check"></i><b>2.2</b> Case II: PDAC-A ST and inDrop (Moncada et al., Nat Biotechnol, 2020)</a>
<ul>
<li class="chapter" data-level="2.2.1" data-path="dataset.html"><a href="dataset.html#pdac-a-st"><i class="fa fa-check"></i><b>2.2.1</b> PDAC-A ST</a></li>
<li class="chapter" data-level="2.2.2" data-path="dataset.html"><a href="dataset.html#pdac-a-indrop"><i class="fa fa-check"></i><b>2.2.2</b> PDAC-A inDrop</a></li>
</ul></li>
<li class="chapter" data-level="2.3" data-path="dataset.html"><a href="dataset.html#case-iii-input-10x-pbmc-data"><i class="fa fa-check"></i><b>2.3</b> Case III: input 10x PBMC data</a></li>
</ul></li>
<li class="chapter" data-level="3" data-path="dataqc.html"><a href="dataqc.html"><i class="fa fa-check"></i><b>3</b> Data quality control (QC)</a>
<ul>
<li class="chapter" data-level="3.1" data-path="dataqc.html"><a href="dataqc.html#quick-qc"><i class="fa fa-check"></i><b>3.1</b> Quick QC</a></li>
<li class="chapter" data-level="3.2" data-path="dataqc.html"><a href="dataqc.html#sample-qc"><i class="fa fa-check"></i><b>3.2</b> Sample QC</a></li>
<li class="chapter" data-level="3.3" data-path="dataqc.html"><a href="dataqc.html#variable-qc"><i class="fa fa-check"></i><b>3.3</b> Variable QC</a></li>
</ul></li>
<li class="chapter" data-level="4" data-path="normalization.html"><a href="normalization.html"><i class="fa fa-check"></i><b>4</b> Normalizing and centering data</a></li>
<li class="chapter" data-level="5" data-path="integration.html"><a href="integration.html"><i class="fa fa-check"></i><b>5</b> Data integration</a>
<ul>
<li class="chapter" data-level="5.1" data-path="integration.html"><a href="integration.html#integrate-pdac-indrop-and-st-data-by-seurat"><i class="fa fa-check"></i><b>5.1</b> Integrate PDAC inDrop and ST data by Seurat</a></li>
<li class="chapter" data-level="5.2" data-path="integration.html"><a href="integration.html#identifying-differentially-expressed-genes"><i class="fa fa-check"></i><b>5.2</b> Identifying differentially expressed genes</a></li>
<li class="chapter" data-level="5.3" data-path="integration.html"><a href="integration.html#inferring-cell-cycle-phases"><i class="fa fa-check"></i><b>5.3</b> Inferring cell cycle phases</a></li>
<li class="chapter" data-level="5.4" data-path="integration.html"><a href="integration.html#save-results"><i class="fa fa-check"></i><b>5.4</b> Save results</a></li>
</ul></li>
<li class="chapter" data-level="6" data-path="computcorr.html"><a href="computcorr.html"><i class="fa fa-check"></i><b>6</b> Computing correlations among genes</a></li>
<li class="chapter" data-level="7" data-path="markergenes.html"><a href="markergenes.html"><i class="fa fa-check"></i><b>7</b> Checking expression profiles of marker genes</a></li>
<li class="chapter" data-level="8" data-path="collectdb.html"><a href="collectdb.html"><i class="fa fa-check"></i><b>8</b> Collecting databases (optional)</a>
<ul>
<li class="chapter" data-level="8.1" data-path="collectdb.html"><a href="collectdb.html#disease-ontology-db"><i class="fa fa-check"></i><b>8.1</b> Disease Ontology DB</a></li>
<li class="chapter" data-level="8.2" data-path="collectdb.html"><a href="collectdb.html#cell-ontology-db"><i class="fa fa-check"></i><b>8.2</b> Cell Ontology DB</a></li>
<li class="chapter" data-level="8.3" data-path="collectdb.html"><a href="collectdb.html#gene-ontology-db"><i class="fa fa-check"></i><b>8.3</b> Gene Ontology DB</a></li>
<li class="chapter" data-level="8.4" data-path="collectdb.html"><a href="collectdb.html#kegg"><i class="fa fa-check"></i><b>8.4</b> KEGG</a></li>
<li class="chapter" data-level="8.5" data-path="collectdb.html"><a href="collectdb.html#custom-built-db-for-our-analysis-of-pdac-data"><i class="fa fa-check"></i><b>8.5</b> Custom-built DB for our analysis of PDAC data</a></li>
</ul></li>
<li class="chapter" data-level="9" data-path="cell_type.html"><a href="cell_type.html"><i class="fa fa-check"></i><b>9</b> Inferring cell types or diseases (optional)</a>
<ul>
<li class="chapter" data-level="9.1" data-path="cell_type.html"><a href="cell_type.html#add-database-to-asurat-objects"><i class="fa fa-check"></i><b>9.1</b> Add database to ASURAT objects</a></li>
<li class="chapter" data-level="9.2" data-path="cell_type.html"><a href="cell_type.html#quick-qc-1"><i class="fa fa-check"></i><b>9.2</b> Quick QC</a></li>
<li class="chapter" data-level="9.3" data-path="cell_type.html"><a href="cell_type.html#correlation-graph-based-decomposition"><i class="fa fa-check"></i><b>9.3</b> Correlation graph-based decomposition</a></li>
<li class="chapter" data-level="9.4" data-path="cell_type.html"><a href="cell_type.html#creating-systems-of-signs"><i class="fa fa-check"></i><b>9.4</b> Creating systems of signs</a></li>
<li class="chapter" data-level="9.5" data-path="cell_type.html"><a href="cell_type.html#removing-redundant-signs"><i class="fa fa-check"></i><b>9.5</b> Removing redundant signs</a></li>
<li class="chapter" data-level="9.6" data-path="cell_type.html"><a href="cell_type.html#creating-sign-by-sample-matrices"><i class="fa fa-check"></i><b>9.6</b> Creating sign-by-sample matrices</a></li>
<li class="chapter" data-level="9.7" data-path="cell_type.html"><a href="cell_type.html#dimensionality-reduction"><i class="fa fa-check"></i><b>9.7</b> Dimensionality reduction</a>
<ul>
<li class="chapter" data-level="9.7.1" data-path="cell_type.html"><a href="cell_type.html#principal-component-analysis-optional"><i class="fa fa-check"></i><b>9.7.1</b> Principal component analysis (optional)</a></li>
<li class="chapter" data-level="9.7.2" data-path="cell_type.html"><a href="cell_type.html#t-distributed-stochastic-neighbor-embedding-optional"><i class="fa fa-check"></i><b>9.7.2</b> t-distributed stochastic neighbor embedding (optional)</a></li>
<li class="chapter" data-level="9.7.3" data-path="cell_type.html"><a href="cell_type.html#uniform-manifold-approximation-and-projection-optional"><i class="fa fa-check"></i><b>9.7.3</b> Uniform manifold approximation and projection (optional)</a></li>
<li class="chapter" data-level="9.7.4" data-path="cell_type.html"><a href="cell_type.html#diffusion-map-optional"><i class="fa fa-check"></i><b>9.7.4</b> Diffusion map (optional)</a></li>
<li class="chapter" data-level="9.7.5" data-path="cell_type.html"><a href="cell_type.html#visualization"><i class="fa fa-check"></i><b>9.7.5</b> Visualization</a></li>
<li class="chapter" data-level="9.7.6" data-path="cell_type.html"><a href="cell_type.html#save-results."><i class="fa fa-check"></i><b>9.7.6</b> Save results.</a></li>
</ul></li>
<li class="chapter" data-level="9.8" data-path="cell_type.html"><a href="cell_type.html#cell-clustering"><i class="fa fa-check"></i><b>9.8</b> Cell clustering</a>
<ul>
<li class="chapter" data-level="9.8.1" data-path="cell_type.html"><a href="cell_type.html#partitioning-around-medoids-pam-optional"><i class="fa fa-check"></i><b>9.8.1</b> Partitioning around medoids (PAM) (optional)</a></li>
<li class="chapter" data-level="9.8.2" data-path="cell_type.html"><a href="cell_type.html#hierarchical-clustering-and-cutting-tree-optional"><i class="fa fa-check"></i><b>9.8.2</b> Hierarchical clustering and cutting tree (optional)</a></li>
<li class="chapter" data-level="9.8.3" data-path="cell_type.html"><a href="cell_type.html#community-detection-algorithm-with-k-nearest-neighbor-knn-graph-optional"><i class="fa fa-check"></i><b>9.8.3</b> Community detection algorithm with k-nearest neighbor (KNN) graph (optional)</a></li>
<li class="chapter" data-level="9.8.4" data-path="cell_type.html"><a href="cell_type.html#clustering-by-branch-detection-along-pseudotime-optional"><i class="fa fa-check"></i><b>9.8.4</b> Clustering by branch detection along pseudotime (optional)</a></li>
<li class="chapter" data-level="9.8.5" data-path="cell_type.html"><a href="cell_type.html#save-data"><i class="fa fa-check"></i><b>9.8.5</b> Save data</a></li>
</ul></li>
<li class="chapter" data-level="9.9" data-path="cell_type.html"><a href="cell_type.html#visualize-clustering-results"><i class="fa fa-check"></i><b>9.9</b> Visualize clustering results</a></li>
<li class="chapter" data-level="9.10" data-path="cell_type.html"><a href="cell_type.html#find-significant-signs"><i class="fa fa-check"></i><b>9.10</b> Find significant signs</a>
<ul>
<li class="chapter" data-level="9.10.1" data-path="cell_type.html"><a href="cell_type.html#compute-separation-indices"><i class="fa fa-check"></i><b>9.10.1</b> Compute separation indices</a></li>
<li class="chapter" data-level="9.10.2" data-path="cell_type.html"><a href="cell_type.html#investigate-sign-scores"><i class="fa fa-check"></i><b>9.10.2</b> Investigate sign scores</a></li>
</ul></li>
<li class="chapter" data-level="9.11" data-path="cell_type.html"><a href="cell_type.html#investigate-batch-effects"><i class="fa fa-check"></i><b>9.11</b> Investigate batch effects</a></li>
<li class="chapter" data-level="9.12" data-path="cell_type.html"><a href="cell_type.html#spatial-mapping-by-spaniel"><i class="fa fa-check"></i><b>9.12</b> Spatial mapping by Spaniel</a></li>
</ul></li>
<li class="chapter" data-level="10" data-path="biological_function.html"><a href="biological_function.html"><i class="fa fa-check"></i><b>10</b> Inferring biological functions (optional)</a>
<ul>
<li class="chapter" data-level="10.1" data-path="biological_function.html"><a href="biological_function.html#add-database-to-asurat-objects-1"><i class="fa fa-check"></i><b>10.1</b> Add database to ASURAT objects</a></li>
<li class="chapter" data-level="10.2" data-path="biological_function.html"><a href="biological_function.html#quick-qc-2"><i class="fa fa-check"></i><b>10.2</b> Quick QC</a></li>
<li class="chapter" data-level="10.3" data-path="biological_function.html"><a href="biological_function.html#correlation-graph-based-decomposition-1"><i class="fa fa-check"></i><b>10.3</b> Correlation graph-based decomposition</a></li>
<li class="chapter" data-level="10.4" data-path="biological_function.html"><a href="biological_function.html#creating-systems-of-signs-1"><i class="fa fa-check"></i><b>10.4</b> Creating systems of signs</a></li>
<li class="chapter" data-level="10.5" data-path="biological_function.html"><a href="biological_function.html#removing-redundant-signs-1"><i class="fa fa-check"></i><b>10.5</b> Removing redundant signs</a></li>
<li class="chapter" data-level="10.6" data-path="biological_function.html"><a href="biological_function.html#creating-sign-by-sample-matrices-1"><i class="fa fa-check"></i><b>10.6</b> Creating sign-by-sample matrices</a></li>
<li class="chapter" data-level="10.7" data-path="biological_function.html"><a href="biological_function.html#dimensionality-reduction-1"><i class="fa fa-check"></i><b>10.7</b> Dimensionality reduction</a>
<ul>
<li class="chapter" data-level="10.7.1" data-path="biological_function.html"><a href="biological_function.html#principal-component-analysis-optional-1"><i class="fa fa-check"></i><b>10.7.1</b> Principal component analysis (optional)</a></li>
<li class="chapter" data-level="10.7.2" data-path="biological_function.html"><a href="biological_function.html#t-distributed-stochastic-neighbor-embedding-optional-1"><i class="fa fa-check"></i><b>10.7.2</b> t-distributed stochastic neighbor embedding (optional)</a></li>
<li class="chapter" data-level="10.7.3" data-path="biological_function.html"><a href="biological_function.html#uniform-manifold-approximation-and-projection-optional-1"><i class="fa fa-check"></i><b>10.7.3</b> Uniform manifold approximation and projection (optional)</a></li>
<li class="chapter" data-level="10.7.4" data-path="biological_function.html"><a href="biological_function.html#diffusion-map-optional-1"><i class="fa fa-check"></i><b>10.7.4</b> Diffusion map (optional)</a></li>
<li class="chapter" data-level="10.7.5" data-path="biological_function.html"><a href="biological_function.html#visualization-1"><i class="fa fa-check"></i><b>10.7.5</b> Visualization</a></li>
<li class="chapter" data-level="10.7.6" data-path="biological_function.html"><a href="biological_function.html#save-results.-1"><i class="fa fa-check"></i><b>10.7.6</b> Save results.</a></li>
</ul></li>
<li class="chapter" data-level="10.8" data-path="biological_function.html"><a href="biological_function.html#cell-clustering-1"><i class="fa fa-check"></i><b>10.8</b> Cell clustering</a>
<ul>
<li class="chapter" data-level="10.8.1" data-path="biological_function.html"><a href="biological_function.html#partitioning-around-medoids-pam-optional-1"><i class="fa fa-check"></i><b>10.8.1</b> Partitioning around medoids (PAM) (optional)</a></li>
<li class="chapter" data-level="10.8.2" data-path="biological_function.html"><a href="biological_function.html#hierarchical-clustering-and-cutting-tree-optional-1"><i class="fa fa-check"></i><b>10.8.2</b> Hierarchical clustering and cutting tree (optional)</a></li>
<li class="chapter" data-level="10.8.3" data-path="biological_function.html"><a href="biological_function.html#community-detection-algorithm-with-k-nearest-neighbor-knn-graph-optional-1"><i class="fa fa-check"></i><b>10.8.3</b> Community detection algorithm with k-nearest neighbor (KNN) graph (optional)</a></li>
<li class="chapter" data-level="10.8.4" data-path="biological_function.html"><a href="biological_function.html#clustering-by-branch-detection-along-pseudotime-optional-1"><i class="fa fa-check"></i><b>10.8.4</b> Clustering by branch detection along pseudotime (optional)</a></li>
<li class="chapter" data-level="10.8.5" data-path="biological_function.html"><a href="biological_function.html#save-data-1"><i class="fa fa-check"></i><b>10.8.5</b> Save data</a></li>
</ul></li>
<li class="chapter" data-level="10.9" data-path="biological_function.html"><a href="biological_function.html#visualize-clustering-results-1"><i class="fa fa-check"></i><b>10.9</b> Visualize clustering results</a></li>
<li class="chapter" data-level="10.10" data-path="biological_function.html"><a href="biological_function.html#find-significant-signs-1"><i class="fa fa-check"></i><b>10.10</b> Find significant signs</a>
<ul>
<li class="chapter" data-level="10.10.1" data-path="biological_function.html"><a href="biological_function.html#compute-separation-indices-1"><i class="fa fa-check"></i><b>10.10.1</b> Compute separation indices</a></li>
<li class="chapter" data-level="10.10.2" data-path="biological_function.html"><a href="biological_function.html#investigate-sign-scores-1"><i class="fa fa-check"></i><b>10.10.2</b> Investigate sign scores</a></li>
</ul></li>
<li class="chapter" data-level="10.11" data-path="biological_function.html"><a href="biological_function.html#investigate-batch-effects-1"><i class="fa fa-check"></i><b>10.11</b> Investigate batch effects</a></li>
<li class="chapter" data-level="10.12" data-path="biological_function.html"><a href="biological_function.html#spatial-mapping-by-spaniel-1"><i class="fa fa-check"></i><b>10.12</b> Spatial mapping by Spaniel</a></li>
</ul></li>
<li class="chapter" data-level="11" data-path="pathway.html"><a href="pathway.html"><i class="fa fa-check"></i><b>11</b> Inferring pathway activities (optional)</a>
<ul>
<li class="chapter" data-level="11.1" data-path="pathway.html"><a href="pathway.html#add-database-to-asurat-objects-2"><i class="fa fa-check"></i><b>11.1</b> Add database to ASURAT objects</a></li>
<li class="chapter" data-level="11.2" data-path="pathway.html"><a href="pathway.html#quick-qc-3"><i class="fa fa-check"></i><b>11.2</b> Quick QC</a></li>
<li class="chapter" data-level="11.3" data-path="pathway.html"><a href="pathway.html#correlation-graph-based-decomposition-2"><i class="fa fa-check"></i><b>11.3</b> Correlation graph-based decomposition</a></li>
<li class="chapter" data-level="11.4" data-path="pathway.html"><a href="pathway.html#creating-systems-of-signs-2"><i class="fa fa-check"></i><b>11.4</b> Creating systems of signs</a></li>
<li class="chapter" data-level="11.5" data-path="pathway.html"><a href="pathway.html#removing-redundant-signs-2"><i class="fa fa-check"></i><b>11.5</b> Removing redundant signs</a></li>
<li class="chapter" data-level="11.6" data-path="pathway.html"><a href="pathway.html#creating-sign-by-sample-matrices-2"><i class="fa fa-check"></i><b>11.6</b> Creating sign-by-sample matrices</a></li>
<li class="chapter" data-level="11.7" data-path="pathway.html"><a href="pathway.html#dimensionality-reduction-2"><i class="fa fa-check"></i><b>11.7</b> Dimensionality reduction</a>
<ul>
<li class="chapter" data-level="11.7.1" data-path="pathway.html"><a href="pathway.html#principal-component-analysis-optional-2"><i class="fa fa-check"></i><b>11.7.1</b> Principal component analysis (optional)</a></li>
<li class="chapter" data-level="11.7.2" data-path="pathway.html"><a href="pathway.html#t-distributed-stochastic-neighbor-embedding-optional-2"><i class="fa fa-check"></i><b>11.7.2</b> t-distributed stochastic neighbor embedding (optional)</a></li>
<li class="chapter" data-level="11.7.3" data-path="pathway.html"><a href="pathway.html#uniform-manifold-approximation-and-projection-optional-2"><i class="fa fa-check"></i><b>11.7.3</b> Uniform manifold approximation and projection (optional)</a></li>
<li class="chapter" data-level="11.7.4" data-path="pathway.html"><a href="pathway.html#diffusion-map-optional-2"><i class="fa fa-check"></i><b>11.7.4</b> Diffusion map (optional)</a></li>
<li class="chapter" data-level="11.7.5" data-path="pathway.html"><a href="pathway.html#visualization-2"><i class="fa fa-check"></i><b>11.7.5</b> Visualization</a></li>
<li class="chapter" data-level="11.7.6" data-path="pathway.html"><a href="pathway.html#save-results.-2"><i class="fa fa-check"></i><b>11.7.6</b> Save results.</a></li>
</ul></li>
<li class="chapter" data-level="11.8" data-path="pathway.html"><a href="pathway.html#cell-clustering-2"><i class="fa fa-check"></i><b>11.8</b> Cell clustering</a>
<ul>
<li class="chapter" data-level="11.8.1" data-path="pathway.html"><a href="pathway.html#partitioning-around-medoids-pam-optional-2"><i class="fa fa-check"></i><b>11.8.1</b> Partitioning around medoids (PAM) (optional)</a></li>
<li class="chapter" data-level="11.8.2" data-path="pathway.html"><a href="pathway.html#hierarchical-clustering-and-cutting-tree-optional-2"><i class="fa fa-check"></i><b>11.8.2</b> Hierarchical clustering and cutting tree (optional)</a></li>
<li class="chapter" data-level="11.8.3" data-path="pathway.html"><a href="pathway.html#community-detection-algorithm-with-k-nearest-neighbor-knn-graph-optional-2"><i class="fa fa-check"></i><b>11.8.3</b> Community detection algorithm with k-nearest neighbor (KNN) graph (optional)</a></li>
<li class="chapter" data-level="11.8.4" data-path="pathway.html"><a href="pathway.html#clustering-by-branch-detection-along-pseudotime-optional-2"><i class="fa fa-check"></i><b>11.8.4</b> Clustering by branch detection along pseudotime (optional)</a></li>
<li class="chapter" data-level="11.8.5" data-path="pathway.html"><a href="pathway.html#save-data-2"><i class="fa fa-check"></i><b>11.8.5</b> Save data</a></li>
</ul></li>
<li class="chapter" data-level="11.9" data-path="pathway.html"><a href="pathway.html#visualize-clustering-results-2"><i class="fa fa-check"></i><b>11.9</b> Visualize clustering results</a></li>
<li class="chapter" data-level="11.10" data-path="pathway.html"><a href="pathway.html#find-significant-signs-2"><i class="fa fa-check"></i><b>11.10</b> Find significant signs</a>
<ul>
<li class="chapter" data-level="11.10.1" data-path="pathway.html"><a href="pathway.html#compute-separation-indices-2"><i class="fa fa-check"></i><b>11.10.1</b> Compute separation indices</a></li>
<li class="chapter" data-level="11.10.2" data-path="pathway.html"><a href="pathway.html#investigate-sign-scores-2"><i class="fa fa-check"></i><b>11.10.2</b> Investigate sign scores</a></li>
</ul></li>
<li class="chapter" data-level="11.11" data-path="pathway.html"><a href="pathway.html#investigate-batch-effects-2"><i class="fa fa-check"></i><b>11.11</b> Investigate batch effects</a></li>
</ul></li>
<li class="chapter" data-level="12" data-path="multi_layered.html"><a href="multi_layered.html"><i class="fa fa-check"></i><b>12</b> Multi-layered analysis across all the signs</a>
<ul>
<li class="chapter" data-level="12.1" data-path="multi_layered.html"><a href="multi_layered.html#concatenate-asurat-objects"><i class="fa fa-check"></i><b>12.1</b> Concatenate ASURAT objects</a></li>
<li class="chapter" data-level="12.2" data-path="multi_layered.html"><a href="multi_layered.html#find-significant-signs-and-genes-across-different-data-types"><i class="fa fa-check"></i><b>12.2</b> Find significant signs and genes across different data types</a></li>
<li class="chapter" data-level="12.3" data-path="multi_layered.html"><a href="multi_layered.html#spatial-mapping-by-spaniel-2"><i class="fa fa-check"></i><b>12.3</b> Spatial mapping by Spaniel</a></li>
</ul></li>
<li class="chapter" data-level="13" data-path="appendix-a-analyzing-scrna-seq-data-by-existing-method-i.html"><a href="appendix-a-analyzing-scrna-seq-data-by-existing-method-i.html"><i class="fa fa-check"></i><b>13</b> Appendix A: analyzing scRNA-seq data by existing method I</a>
<ul>
<li class="chapter" data-level="13.1" data-path="appendix-a-analyzing-scrna-seq-data-by-existing-method-i.html"><a href="appendix-a-analyzing-scrna-seq-data-by-existing-method-i.html#preprocessing"><i class="fa fa-check"></i><b>13.1</b> Preprocessing</a></li>
<li class="chapter" data-level="13.2" data-path="appendix-a-analyzing-scrna-seq-data-by-existing-method-i.html"><a href="appendix-a-analyzing-scrna-seq-data-by-existing-method-i.html#cluster-cells"><i class="fa fa-check"></i><b>13.2</b> Cluster cells</a></li>
<li class="chapter" data-level="13.3" data-path="appendix-a-analyzing-scrna-seq-data-by-existing-method-i.html"><a href="appendix-a-analyzing-scrna-seq-data-by-existing-method-i.html#visualize-data-in-low-dimensional-space"><i class="fa fa-check"></i><b>13.3</b> Visualize data in low dimensional space</a></li>
<li class="chapter" data-level="13.4" data-path="appendix-a-analyzing-scrna-seq-data-by-existing-method-i.html"><a href="appendix-a-analyzing-scrna-seq-data-by-existing-method-i.html#find-differentially-expressed-genes"><i class="fa fa-check"></i><b>13.4</b> Find differentially expressed genes</a></li>
<li class="chapter" data-level="13.5" data-path="appendix-a-analyzing-scrna-seq-data-by-existing-method-i.html"><a href="appendix-a-analyzing-scrna-seq-data-by-existing-method-i.html#save-results-1"><i class="fa fa-check"></i><b>13.5</b> Save results</a></li>
<li class="chapter" data-level="13.6" data-path="appendix-a-analyzing-scrna-seq-data-by-existing-method-i.html"><a href="appendix-a-analyzing-scrna-seq-data-by-existing-method-i.html#infer-cell-types-for-pbmc-datasets"><i class="fa fa-check"></i><b>13.6</b> Infer cell types for PBMC datasets</a></li>
</ul></li>
<li class="chapter" data-level="14" data-path="appendix-b-analyzing-scrna-seq-data-by-existing-method-ii.html"><a href="appendix-b-analyzing-scrna-seq-data-by-existing-method-ii.html"><i class="fa fa-check"></i><b>14</b> Appendix B: analyzing scRNA-seq data by existing method II</a>
<ul>
<li class="chapter" data-level="14.1" data-path="appendix-b-analyzing-scrna-seq-data-by-existing-method-ii.html"><a href="appendix-b-analyzing-scrna-seq-data-by-existing-method-ii.html#preprocessing-1"><i class="fa fa-check"></i><b>14.1</b> Preprocessing</a></li>
<li class="chapter" data-level="14.2" data-path="appendix-b-analyzing-scrna-seq-data-by-existing-method-ii.html"><a href="appendix-b-analyzing-scrna-seq-data-by-existing-method-ii.html#cluster-cells-1"><i class="fa fa-check"></i><b>14.2</b> Cluster cells</a></li>
<li class="chapter" data-level="14.3" data-path="appendix-b-analyzing-scrna-seq-data-by-existing-method-ii.html"><a href="appendix-b-analyzing-scrna-seq-data-by-existing-method-ii.html#visualize-data-in-low-dimensional-space-1"><i class="fa fa-check"></i><b>14.3</b> Visualize data in low dimensional space</a></li>
<li class="chapter" data-level="14.4" data-path="appendix-b-analyzing-scrna-seq-data-by-existing-method-ii.html"><a href="appendix-b-analyzing-scrna-seq-data-by-existing-method-ii.html#cell-cycle-scoring"><i class="fa fa-check"></i><b>14.4</b> Cell cycle scoring</a></li>
<li class="chapter" data-level="14.5" data-path="appendix-b-analyzing-scrna-seq-data-by-existing-method-ii.html"><a href="appendix-b-analyzing-scrna-seq-data-by-existing-method-ii.html#find-differentially-expressed-genes-1"><i class="fa fa-check"></i><b>14.5</b> Find differentially expressed genes</a></li>
<li class="chapter" data-level="14.6" data-path="appendix-b-analyzing-scrna-seq-data-by-existing-method-ii.html"><a href="appendix-b-analyzing-scrna-seq-data-by-existing-method-ii.html#enrichment-analysis"><i class="fa fa-check"></i><b>14.6</b> Enrichment analysis</a></li>
<li class="chapter" data-level="14.7" data-path="appendix-b-analyzing-scrna-seq-data-by-existing-method-ii.html"><a href="appendix-b-analyzing-scrna-seq-data-by-existing-method-ii.html#save-results-2"><i class="fa fa-check"></i><b>14.7</b> Save results</a></li>
<li class="chapter" data-level="14.8" data-path="appendix-b-analyzing-scrna-seq-data-by-existing-method-ii.html"><a href="appendix-b-analyzing-scrna-seq-data-by-existing-method-ii.html#infer-cell-types-for-pbmc-datasets-1"><i class="fa fa-check"></i><b>14.8</b> Infer cell types for PBMC datasets</a></li>
<li class="chapter" data-level="14.9" data-path="appendix-b-analyzing-scrna-seq-data-by-existing-method-ii.html"><a href="appendix-b-analyzing-scrna-seq-data-by-existing-method-ii.html#infer-cell-types-for-pbmc-datasets-by-seurat-and-sccatch"><i class="fa fa-check"></i><b>14.9</b> Infer cell types for PBMC datasets by Seurat and scCATCH</a></li>
</ul></li>
<li class="chapter" data-level="15" data-path="appendix-c-analyzing-scrna-seq-data-by-existing-method-iii.html"><a href="appendix-c-analyzing-scrna-seq-data-by-existing-method-iii.html"><i class="fa fa-check"></i><b>15</b> Appendix C: analyzing scRNA-seq data by existing method III</a>
<ul>
<li class="chapter" data-level="15.1" data-path="appendix-c-analyzing-scrna-seq-data-by-existing-method-iii.html"><a href="appendix-c-analyzing-scrna-seq-data-by-existing-method-iii.html#preprocessing-2"><i class="fa fa-check"></i><b>15.1</b> Preprocessing</a></li>
<li class="chapter" data-level="15.2" data-path="appendix-c-analyzing-scrna-seq-data-by-existing-method-iii.html"><a href="appendix-c-analyzing-scrna-seq-data-by-existing-method-iii.html#cluster-cells-2"><i class="fa fa-check"></i><b>15.2</b> Cluster cells</a></li>
<li class="chapter" data-level="15.3" data-path="appendix-c-analyzing-scrna-seq-data-by-existing-method-iii.html"><a href="appendix-c-analyzing-scrna-seq-data-by-existing-method-iii.html#visualize-data-in-low-dimensional-space-2"><i class="fa fa-check"></i><b>15.3</b> Visualize data in low dimensional space</a></li>
<li class="chapter" data-level="15.4" data-path="appendix-c-analyzing-scrna-seq-data-by-existing-method-iii.html"><a href="appendix-c-analyzing-scrna-seq-data-by-existing-method-iii.html#find-differentially-expressed-genes-2"><i class="fa fa-check"></i><b>15.4</b> Find differentially expressed genes</a></li>
<li class="chapter" data-level="15.5" data-path="appendix-c-analyzing-scrna-seq-data-by-existing-method-iii.html"><a href="appendix-c-analyzing-scrna-seq-data-by-existing-method-iii.html#save-results-3"><i class="fa fa-check"></i><b>15.5</b> Save results</a></li>
<li class="chapter" data-level="15.6" data-path="appendix-c-analyzing-scrna-seq-data-by-existing-method-iii.html"><a href="appendix-c-analyzing-scrna-seq-data-by-existing-method-iii.html#infer-cell-types-for-pbmc-datasets-2"><i class="fa fa-check"></i><b>15.6</b> Infer cell types for PBMC datasets</a></li>
</ul></li>
<li class="chapter" data-level="16" data-path="appendix-d-analyzing-scrna-seq-data-by-existing-method-iv.html"><a href="appendix-d-analyzing-scrna-seq-data-by-existing-method-iv.html"><i class="fa fa-check"></i><b>16</b> Appendix D: analyzing scRNA-seq data by existing method IV</a>
<ul>
<li class="chapter" data-level="16.1" data-path="appendix-d-analyzing-scrna-seq-data-by-existing-method-iv.html"><a href="appendix-d-analyzing-scrna-seq-data-by-existing-method-iv.html#preprocessing-3"><i class="fa fa-check"></i><b>16.1</b> Preprocessing</a></li>
<li class="chapter" data-level="16.2" data-path="appendix-d-analyzing-scrna-seq-data-by-existing-method-iv.html"><a href="appendix-d-analyzing-scrna-seq-data-by-existing-method-iv.html#cluster-cells-3"><i class="fa fa-check"></i><b>16.2</b> Cluster cells</a></li>
<li class="chapter" data-level="16.3" data-path="appendix-d-analyzing-scrna-seq-data-by-existing-method-iv.html"><a href="appendix-d-analyzing-scrna-seq-data-by-existing-method-iv.html#find-differentially-expressed-genes-3"><i class="fa fa-check"></i><b>16.3</b> Find differentially expressed genes</a></li>
<li class="chapter" data-level="16.4" data-path="appendix-d-analyzing-scrna-seq-data-by-existing-method-iv.html"><a href="appendix-d-analyzing-scrna-seq-data-by-existing-method-iv.html#save-results-4"><i class="fa fa-check"></i><b>16.4</b> Save results</a></li>
<li class="chapter" data-level="16.5" data-path="appendix-d-analyzing-scrna-seq-data-by-existing-method-iv.html"><a href="appendix-d-analyzing-scrna-seq-data-by-existing-method-iv.html#infer-cell-types-for-pbmc-datasets-3"><i class="fa fa-check"></i><b>16.5</b> Infer cell types for PBMC datasets</a></li>
</ul></li>
<li class="chapter" data-level="17" data-path="appendix-e-analyzing-scrna-seq-data-using-asurat.html"><a href="appendix-e-analyzing-scrna-seq-data-using-asurat.html"><i class="fa fa-check"></i><b>17</b> Appendix E: analyzing scRNA-seq data using ASURAT</a>
<ul>
<li class="chapter" data-level="17.1" data-path="appendix-e-analyzing-scrna-seq-data-using-asurat.html"><a href="appendix-e-analyzing-scrna-seq-data-using-asurat.html#infer-cell-types-for-pbmc-datasets-4"><i class="fa fa-check"></i><b>17.1</b> Infer cell types for PBMC datasets</a></li>
</ul></li>
<li class="chapter" data-level="18" data-path="appendix-f-creating-figures.html"><a href="appendix-f-creating-figures.html"><i class="fa fa-check"></i><b>18</b> Appendix F: creating figures</a>
<ul>
<li class="chapter" data-level="18.1" data-path="appendix-f-creating-figures.html"><a href="appendix-f-creating-figures.html#pbmcs-datasets"><i class="fa fa-check"></i><b>18.1</b> PBMCs datasets</a>
<ul>
<li class="chapter" data-level="18.1.1" data-path="appendix-f-creating-figures.html"><a href="appendix-f-creating-figures.html#multi-layered-analysis"><i class="fa fa-check"></i><b>18.1.1</b> Multi-layered analysis</a></li>
<li class="chapter" data-level="18.1.2" data-path="appendix-f-creating-figures.html"><a href="appendix-f-creating-figures.html#infer-population-ratios"><i class="fa fa-check"></i><b>18.1.2</b> Infer population ratios</a></li>
</ul></li>
<li class="chapter" data-level="18.2" data-path="appendix-f-creating-figures.html"><a href="appendix-f-creating-figures.html#lung-cancer-dataset"><i class="fa fa-check"></i><b>18.2</b> Lung cancer dataset</a>
<ul>
<li class="chapter" data-level="18.2.1" data-path="appendix-f-creating-figures.html"><a href="appendix-f-creating-figures.html#heatmaps-of-marker-gene-expressions"><i class="fa fa-check"></i><b>18.2.1</b> Heatmaps of marker gene expressions</a></li>
<li class="chapter" data-level="18.2.2" data-path="appendix-f-creating-figures.html"><a href="appendix-f-creating-figures.html#multi-layered-analysis-1"><i class="fa fa-check"></i><b>18.2.2</b> Multi-layered analysis</a></li>
</ul></li>
<li class="chapter" data-level="18.3" data-path="appendix-f-creating-figures.html"><a href="appendix-f-creating-figures.html#pancreatic-ductal-adenocarcinoma-pdac-datasets"><i class="fa fa-check"></i><b>18.3</b> Pancreatic ductal adenocarcinoma (PDAC) datasets</a>
<ul>
<li class="chapter" data-level="18.3.1" data-path="appendix-f-creating-figures.html"><a href="appendix-f-creating-figures.html#clustering-pdac-spatial-transcriptomes"><i class="fa fa-check"></i><b>18.3.1</b> Clustering PDAC spatial transcriptomes</a></li>
</ul></li>
</ul></li>
<li class="divider"></li>
<li><a href="https://github.com/rstudio/bookdown" target="blank">Published with bookdown</a></li>

</ul>

      </nav>
    </div>

    <div class="book-body">
      <div class="body-inner">
        <div class="book-header" role="navigation">
          <h1>
            <i class="fa fa-circle-o-notch fa-spin"></i><a href="./">Introduction to ASURAT (version 0.0.0.9001)</a>
          </h1>
        </div>

        <div class="page-wrapper" tabindex="-1" role="main">
          <div class="page-inner">

            <section class="normal" id="section-">
<div id="appendix-f-creating-figures" class="section level1" number="18">
<h1><span class="header-section-number">Chapter 18</span> Appendix F: creating figures</h1>
<p>In this section, we create figures for the paper.</p>
<p>The following libraries are used.</p>
<div class="sourceCode" id="cb1"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb1-1"><a href="appendix-f-creating-figures.html#cb1-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(tidyverse)       <span class="co"># For efficient handling of data.frame</span></span>
<span id="cb1-2"><a href="appendix-f-creating-figures.html#cb1-2" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(DT)              <span class="co"># For using datatable</span></span>
<span id="cb1-3"><a href="appendix-f-creating-figures.html#cb1-3" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(scran)           <span class="co"># For using scran</span></span>
<span id="cb1-4"><a href="appendix-f-creating-figures.html#cb1-4" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(Seurat)          <span class="co"># For using Seurat</span></span>
<span id="cb1-5"><a href="appendix-f-creating-figures.html#cb1-5" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(monocle3)        <span class="co"># For using Monocle 3</span></span>
<span id="cb1-6"><a href="appendix-f-creating-figures.html#cb1-6" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(SC3)             <span class="co"># For using SC3</span></span>
<span id="cb1-7"><a href="appendix-f-creating-figures.html#cb1-7" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(ComplexHeatmap)  <span class="co"># For ComplexHeatmap</span></span>
<span id="cb1-8"><a href="appendix-f-creating-figures.html#cb1-8" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(circlize)        <span class="co"># For color mapping function in ComplexHeatmap</span></span>
<span id="cb1-9"><a href="appendix-f-creating-figures.html#cb1-9" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(RColorBrewer)    <span class="co"># For using brewer.pal()</span></span>
<span id="cb1-10"><a href="appendix-f-creating-figures.html#cb1-10" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(scales)          <span class="co"># For using show_col()</span></span>
<span id="cb1-11"><a href="appendix-f-creating-figures.html#cb1-11" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(Spaniel)         <span class="co"># For using Spaniel</span></span></code></pre></div>
<div id="pbmcs-datasets" class="section level2" number="18.1">
<h2><span class="header-section-number">18.1</span> PBMCs datasets</h2>
<div id="multi-layered-analysis" class="section level3" number="18.1.1">
<h3><span class="header-section-number">18.1.1</span> Multi-layered analysis</h3>
<p>Load the data.</p>
<div class="sourceCode" id="cb2"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb2-1"><a href="appendix-f-creating-figures.html#cb2-1" aria-hidden="true" tabindex="-1"></a><span class="fu">source</span>(<span class="st">&quot;R/plot.R&quot;</span>)</span>
<span id="cb2-2"><a href="appendix-f-creating-figures.html#cb2-2" aria-hidden="true" tabindex="-1"></a>pbmc_4000 <span class="ot">&lt;-</span> <span class="fu">readRDS</span>(<span class="at">file =</span> <span class="st">&quot;backup/10_400_pbmc_4000_all.rds&quot;</span>)</span>
<span id="cb2-3"><a href="appendix-f-creating-figures.html#cb2-3" aria-hidden="true" tabindex="-1"></a>pbmc_6000 <span class="ot">&lt;-</span> <span class="fu">readRDS</span>(<span class="at">file =</span> <span class="st">&quot;backup/11_400_pbmc_6000_all.rds&quot;</span>)</span></code></pre></div>
<p>Change the labels of <code>pbmc_4000[["sample"]][["seuratFindClusters_CO_cell"]]</code>
to make it correspond to those of
<code>pbmc_6000[["sample"]][["seuratFindClusters_CO_cell"]]</code>.</p>
<div class="sourceCode" id="cb3"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb3-1"><a href="appendix-f-creating-figures.html#cb3-1" aria-hidden="true" tabindex="-1"></a>df <span class="ot">&lt;-</span> pbmc_4000[[<span class="st">&quot;sample&quot;</span>]]</span>
<span id="cb3-2"><a href="appendix-f-creating-figures.html#cb3-2" aria-hidden="true" tabindex="-1"></a>df<span class="sc">$</span>seuratFindClusters_CO_cell <span class="ot">&lt;-</span> <span class="fu">as.character</span>(df<span class="sc">$</span>seuratFindClusters_CO_cell)</span>
<span id="cb3-3"><a href="appendix-f-creating-figures.html#cb3-3" aria-hidden="true" tabindex="-1"></a>df[df<span class="sc">$</span>seuratFindClusters_CO_cell <span class="sc">==</span> <span class="st">&quot;3&quot;</span>,]<span class="sc">$</span>seuratFindClusters_CO_cell <span class="ot">&lt;-</span> <span class="st">&quot;333&quot;</span></span>
<span id="cb3-4"><a href="appendix-f-creating-figures.html#cb3-4" aria-hidden="true" tabindex="-1"></a>df[df<span class="sc">$</span>seuratFindClusters_CO_cell <span class="sc">==</span> <span class="st">&quot;4&quot;</span>,]<span class="sc">$</span>seuratFindClusters_CO_cell <span class="ot">&lt;-</span> <span class="st">&quot;444&quot;</span></span>
<span id="cb3-5"><a href="appendix-f-creating-figures.html#cb3-5" aria-hidden="true" tabindex="-1"></a>df[df<span class="sc">$</span>seuratFindClusters_CO_cell <span class="sc">==</span> <span class="st">&quot;333&quot;</span>,]<span class="sc">$</span>seuratFindClusters_CO_cell <span class="ot">&lt;-</span> <span class="st">&quot;4&quot;</span></span>
<span id="cb3-6"><a href="appendix-f-creating-figures.html#cb3-6" aria-hidden="true" tabindex="-1"></a>df[df<span class="sc">$</span>seuratFindClusters_CO_cell <span class="sc">==</span> <span class="st">&quot;444&quot;</span>,]<span class="sc">$</span>seuratFindClusters_CO_cell <span class="ot">&lt;-</span> <span class="st">&quot;3&quot;</span></span>
<span id="cb3-7"><a href="appendix-f-creating-figures.html#cb3-7" aria-hidden="true" tabindex="-1"></a>df<span class="sc">$</span>seuratFindClusters_CO_cell <span class="ot">&lt;-</span> <span class="fu">as.factor</span>(df<span class="sc">$</span>seuratFindClusters_CO_cell)</span>
<span id="cb3-8"><a href="appendix-f-creating-figures.html#cb3-8" aria-hidden="true" tabindex="-1"></a>pbmc_4000[[<span class="st">&quot;sample&quot;</span>]] <span class="ot">&lt;-</span> df</span></code></pre></div>
<p>The following function <code>plot_MultiHeatmaps_SignxSamp()</code> vertically concatenates
two or more heat maps across different databases.</p>
<div class="sourceCode" id="cb4"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb4-1"><a href="appendix-f-creating-figures.html#cb4-1" aria-hidden="true" tabindex="-1"></a><span class="co"># ----------------------------------------</span></span>
<span id="cb4-2"><a href="appendix-f-creating-figures.html#cb4-2" aria-hidden="true" tabindex="-1"></a><span class="co"># pbmc_4000</span></span>
<span id="cb4-3"><a href="appendix-f-creating-figures.html#cb4-3" aria-hidden="true" tabindex="-1"></a><span class="co"># ----------------------------------------</span></span>
<span id="cb4-4"><a href="appendix-f-creating-figures.html#cb4-4" aria-hidden="true" tabindex="-1"></a>filename <span class="ot">&lt;-</span> <span class="st">&quot;figures/figure_20_0010.png&quot;</span></span>
<span id="cb4-5"><a href="appendix-f-creating-figures.html#cb4-5" aria-hidden="true" tabindex="-1"></a><span class="fu">png</span>(<span class="at">file =</span> filename, <span class="at">height =</span> <span class="dv">1500</span>, <span class="at">width =</span> <span class="dv">1500</span>, <span class="at">res =</span> <span class="dv">300</span>)</span>
<span id="cb4-6"><a href="appendix-f-creating-figures.html#cb4-6" aria-hidden="true" tabindex="-1"></a><span class="fu">plot_MultiHeatmaps_SignxSamp</span>(</span>
<span id="cb4-7"><a href="appendix-f-creating-figures.html#cb4-7" aria-hidden="true" tabindex="-1"></a>  <span class="at">obj =</span> pbmc_4000,</span>
<span id="cb4-8"><a href="appendix-f-creating-figures.html#cb4-8" aria-hidden="true" tabindex="-1"></a>  <span class="at">data_types =</span> <span class="fu">c</span>(<span class="st">&quot;CO&quot;</span>, <span class="st">&quot;GO&quot;</span>, <span class="st">&quot;KEGG&quot;</span>),</span>
<span id="cb4-9"><a href="appendix-f-creating-figures.html#cb4-9" aria-hidden="true" tabindex="-1"></a>  <span class="at">categories =</span> <span class="fu">c</span>(<span class="st">&quot;cell&quot;</span>, <span class="st">&quot;BP&quot;</span>, <span class="st">&quot;pathway&quot;</span>),</span>
<span id="cb4-10"><a href="appendix-f-creating-figures.html#cb4-10" aria-hidden="true" tabindex="-1"></a>  <span class="at">algo_names =</span> <span class="fu">c</span>(<span class="st">&quot;seuratFindClusters&quot;</span>, <span class="st">&quot;seuratFindClusters&quot;</span>,</span>
<span id="cb4-11"><a href="appendix-f-creating-figures.html#cb4-11" aria-hidden="true" tabindex="-1"></a>                 <span class="st">&quot;seuratFindClusters&quot;</span>),</span>
<span id="cb4-12"><a href="appendix-f-creating-figures.html#cb4-12" aria-hidden="true" tabindex="-1"></a>  <span class="at">show_labels =</span> <span class="fu">c</span>(<span class="cn">TRUE</span>, <span class="cn">TRUE</span>, <span class="cn">TRUE</span>),</span>
<span id="cb4-13"><a href="appendix-f-creating-figures.html#cb4-13" aria-hidden="true" tabindex="-1"></a>  <span class="at">splitting =</span> <span class="cn">TRUE</span>, <span class="at">method =</span> <span class="st">&quot;ward.D2&quot;</span>, <span class="at">show_nReads =</span> <span class="cn">FALSE</span>,</span>
<span id="cb4-14"><a href="appendix-f-creating-figures.html#cb4-14" aria-hidden="true" tabindex="-1"></a>  <span class="at">title =</span> <span class="st">&quot;pbmc_4000&quot;</span>,</span>
<span id="cb4-15"><a href="appendix-f-creating-figures.html#cb4-15" aria-hidden="true" tabindex="-1"></a>  <span class="at">names =</span> <span class="fu">c</span>(<span class="st">&quot;CO:</span><span class="sc">\n</span><span class="st">cell&quot;</span>, <span class="st">&quot;GO:</span><span class="sc">\n</span><span class="st">BP&quot;</span>, <span class="st">&quot;KEGG:</span><span class="sc">\n</span><span class="st">pathway&quot;</span>),</span>
<span id="cb4-16"><a href="appendix-f-creating-figures.html#cb4-16" aria-hidden="true" tabindex="-1"></a>  <span class="at">show_rownames_sign =</span> <span class="cn">FALSE</span>, <span class="at">show_rownames_label =</span> <span class="cn">FALSE</span>,</span>
<span id="cb4-17"><a href="appendix-f-creating-figures.html#cb4-17" aria-hidden="true" tabindex="-1"></a>  <span class="at">show_rownames_nReads =</span> <span class="cn">FALSE</span>, <span class="at">default_colors =</span> <span class="fu">c</span>(<span class="cn">TRUE</span>, <span class="cn">FALSE</span>, <span class="cn">FALSE</span>))</span>
<span id="cb4-18"><a href="appendix-f-creating-figures.html#cb4-18" aria-hidden="true" tabindex="-1"></a><span class="fu">dev.off</span>()</span>
<span id="cb4-19"><a href="appendix-f-creating-figures.html#cb4-19" aria-hidden="true" tabindex="-1"></a><span class="co"># ----------------------------------------</span></span>
<span id="cb4-20"><a href="appendix-f-creating-figures.html#cb4-20" aria-hidden="true" tabindex="-1"></a><span class="co"># pbmc_6000</span></span>
<span id="cb4-21"><a href="appendix-f-creating-figures.html#cb4-21" aria-hidden="true" tabindex="-1"></a><span class="co"># ----------------------------------------</span></span>
<span id="cb4-22"><a href="appendix-f-creating-figures.html#cb4-22" aria-hidden="true" tabindex="-1"></a>filename <span class="ot">&lt;-</span> <span class="st">&quot;figures/figure_20_0011.png&quot;</span></span>
<span id="cb4-23"><a href="appendix-f-creating-figures.html#cb4-23" aria-hidden="true" tabindex="-1"></a><span class="fu">png</span>(<span class="at">file =</span> filename, <span class="at">height =</span> <span class="dv">1500</span>, <span class="at">width =</span> <span class="dv">1500</span>, <span class="at">res =</span> <span class="dv">300</span>)</span>
<span id="cb4-24"><a href="appendix-f-creating-figures.html#cb4-24" aria-hidden="true" tabindex="-1"></a><span class="fu">plot_MultiHeatmaps_SignxSamp</span>(</span>
<span id="cb4-25"><a href="appendix-f-creating-figures.html#cb4-25" aria-hidden="true" tabindex="-1"></a>  <span class="at">obj =</span> pbmc_6000,</span>
<span id="cb4-26"><a href="appendix-f-creating-figures.html#cb4-26" aria-hidden="true" tabindex="-1"></a>  <span class="at">data_types =</span> <span class="fu">c</span>(<span class="st">&quot;CO&quot;</span>, <span class="st">&quot;GO&quot;</span>, <span class="st">&quot;KEGG&quot;</span>),</span>
<span id="cb4-27"><a href="appendix-f-creating-figures.html#cb4-27" aria-hidden="true" tabindex="-1"></a>  <span class="at">categories =</span> <span class="fu">c</span>(<span class="st">&quot;cell&quot;</span>, <span class="st">&quot;BP&quot;</span>, <span class="st">&quot;pathway&quot;</span>),</span>
<span id="cb4-28"><a href="appendix-f-creating-figures.html#cb4-28" aria-hidden="true" tabindex="-1"></a>  <span class="at">algo_names =</span> <span class="fu">c</span>(<span class="st">&quot;seuratFindClusters&quot;</span>, <span class="st">&quot;seuratFindClusters&quot;</span>,</span>
<span id="cb4-29"><a href="appendix-f-creating-figures.html#cb4-29" aria-hidden="true" tabindex="-1"></a>                 <span class="st">&quot;seuratFindClusters&quot;</span>),</span>
<span id="cb4-30"><a href="appendix-f-creating-figures.html#cb4-30" aria-hidden="true" tabindex="-1"></a>  <span class="at">show_labels =</span> <span class="fu">c</span>(<span class="cn">TRUE</span>, <span class="cn">TRUE</span>, <span class="cn">TRUE</span>),</span>
<span id="cb4-31"><a href="appendix-f-creating-figures.html#cb4-31" aria-hidden="true" tabindex="-1"></a>  <span class="at">splitting =</span> <span class="cn">TRUE</span>, <span class="at">method =</span> <span class="st">&quot;ward.D2&quot;</span>, <span class="at">show_nReads =</span> <span class="cn">FALSE</span>,</span>
<span id="cb4-32"><a href="appendix-f-creating-figures.html#cb4-32" aria-hidden="true" tabindex="-1"></a>  <span class="at">title =</span> <span class="st">&quot;pbmc_6000&quot;</span>,</span>
<span id="cb4-33"><a href="appendix-f-creating-figures.html#cb4-33" aria-hidden="true" tabindex="-1"></a>  <span class="at">names =</span> <span class="fu">c</span>(<span class="st">&quot;CO:</span><span class="sc">\n</span><span class="st">cell&quot;</span>, <span class="st">&quot;GO:</span><span class="sc">\n</span><span class="st">BP&quot;</span>, <span class="st">&quot;KEGG:</span><span class="sc">\n</span><span class="st">pathway&quot;</span>),</span>
<span id="cb4-34"><a href="appendix-f-creating-figures.html#cb4-34" aria-hidden="true" tabindex="-1"></a>  <span class="at">show_rownames_sign =</span> <span class="cn">FALSE</span>, <span class="at">show_rownames_label =</span> <span class="cn">FALSE</span>,</span>
<span id="cb4-35"><a href="appendix-f-creating-figures.html#cb4-35" aria-hidden="true" tabindex="-1"></a>  <span class="at">show_rownames_nReads =</span> <span class="cn">FALSE</span>, <span class="at">default_colors =</span> <span class="fu">c</span>(<span class="cn">TRUE</span>, <span class="cn">FALSE</span>, <span class="cn">FALSE</span>))</span>
<span id="cb4-36"><a href="appendix-f-creating-figures.html#cb4-36" aria-hidden="true" tabindex="-1"></a><span class="fu">dev.off</span>()</span></code></pre></div>
<p><img src="figures/figure_20_0010.png" width="400px">
<img src="figures/figure_20_0011.png" width="400px"></p>
<p>Make UMAPs with clustering results.</p>
<div class="sourceCode" id="cb5"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb5-1"><a href="appendix-f-creating-figures.html#cb5-1" aria-hidden="true" tabindex="-1"></a><span class="co"># ----------------------------------------</span></span>
<span id="cb5-2"><a href="appendix-f-creating-figures.html#cb5-2" aria-hidden="true" tabindex="-1"></a><span class="co"># pbmc_4000</span></span>
<span id="cb5-3"><a href="appendix-f-creating-figures.html#cb5-3" aria-hidden="true" tabindex="-1"></a><span class="co"># ----------------------------------------</span></span>
<span id="cb5-4"><a href="appendix-f-creating-figures.html#cb5-4" aria-hidden="true" tabindex="-1"></a>p <span class="ot">&lt;-</span> <span class="fu">plot_umap_sign</span>(<span class="at">obj =</span> pbmc_4000, <span class="at">data_type =</span> <span class="st">&quot;CO&quot;</span>, <span class="at">category =</span> <span class="st">&quot;cell&quot;</span>,</span>
<span id="cb5-5"><a href="appendix-f-creating-figures.html#cb5-5" aria-hidden="true" tabindex="-1"></a>                    <span class="at">algo_name =</span> <span class="st">&quot;seuratFindClusters&quot;</span>,</span>
<span id="cb5-6"><a href="appendix-f-creating-figures.html#cb5-6" aria-hidden="true" tabindex="-1"></a>                    <span class="at">theta =</span> <span class="cn">NULL</span>, <span class="at">phi =</span> <span class="cn">NULL</span>,</span>
<span id="cb5-7"><a href="appendix-f-creating-figures.html#cb5-7" aria-hidden="true" tabindex="-1"></a>                    <span class="at">title =</span> <span class="st">&quot;pbmc_4000 (CO: cell)&quot;</span>, <span class="at">title_size =</span> <span class="dv">16</span>,</span>
<span id="cb5-8"><a href="appendix-f-creating-figures.html#cb5-8" aria-hidden="true" tabindex="-1"></a>                    <span class="at">xlabel =</span> <span class="st">&quot;UMAP_1&quot;</span>, <span class="at">ylabel =</span> <span class="st">&quot;UMAP_2&quot;</span>, <span class="at">zlabel =</span> <span class="cn">NULL</span>,</span>
<span id="cb5-9"><a href="appendix-f-creating-figures.html#cb5-9" aria-hidden="true" tabindex="-1"></a>                    <span class="at">default_color =</span> <span class="cn">TRUE</span>)</span>
<span id="cb5-10"><a href="appendix-f-creating-figures.html#cb5-10" aria-hidden="true" tabindex="-1"></a>filename <span class="ot">&lt;-</span> <span class="st">&quot;figures/figure_20_0020.png&quot;</span></span>
<span id="cb5-11"><a href="appendix-f-creating-figures.html#cb5-11" aria-hidden="true" tabindex="-1"></a><span class="fu">ggsave</span>(<span class="at">file =</span> filename, <span class="at">plot =</span> p, <span class="at">dpi =</span> <span class="dv">300</span>, <span class="at">width =</span> <span class="fl">4.7</span>, <span class="at">height =</span> <span class="dv">4</span>)</span>
<span id="cb5-12"><a href="appendix-f-creating-figures.html#cb5-12" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb5-13"><a href="appendix-f-creating-figures.html#cb5-13" aria-hidden="true" tabindex="-1"></a>p <span class="ot">&lt;-</span> <span class="fu">plot_umap_sign</span>(<span class="at">obj =</span> pbmc_4000, <span class="at">data_type =</span> <span class="st">&quot;GO&quot;</span>, <span class="at">category =</span> <span class="st">&quot;BP&quot;</span>,</span>
<span id="cb5-14"><a href="appendix-f-creating-figures.html#cb5-14" aria-hidden="true" tabindex="-1"></a>                    <span class="at">algo_name =</span> <span class="st">&quot;seuratFindClusters&quot;</span>,</span>
<span id="cb5-15"><a href="appendix-f-creating-figures.html#cb5-15" aria-hidden="true" tabindex="-1"></a>                    <span class="at">theta =</span> <span class="cn">NULL</span>, <span class="at">phi =</span> <span class="cn">NULL</span>,</span>
<span id="cb5-16"><a href="appendix-f-creating-figures.html#cb5-16" aria-hidden="true" tabindex="-1"></a>                    <span class="at">title =</span> <span class="st">&quot;pbmc_4000 (GO: BP)&quot;</span>, <span class="at">title_size =</span> <span class="dv">16</span>,</span>
<span id="cb5-17"><a href="appendix-f-creating-figures.html#cb5-17" aria-hidden="true" tabindex="-1"></a>                    <span class="at">xlabel =</span> <span class="st">&quot;UMAP_1&quot;</span>, <span class="at">ylabel =</span> <span class="st">&quot;UMAP_2&quot;</span>, <span class="at">zlabel =</span> <span class="cn">NULL</span>,</span>
<span id="cb5-18"><a href="appendix-f-creating-figures.html#cb5-18" aria-hidden="true" tabindex="-1"></a>                    <span class="at">default_color =</span> <span class="cn">TRUE</span>)</span>
<span id="cb5-19"><a href="appendix-f-creating-figures.html#cb5-19" aria-hidden="true" tabindex="-1"></a>filename <span class="ot">&lt;-</span> <span class="st">&quot;figures/figure_20_0021.png&quot;</span></span>
<span id="cb5-20"><a href="appendix-f-creating-figures.html#cb5-20" aria-hidden="true" tabindex="-1"></a><span class="fu">ggsave</span>(<span class="at">file =</span> filename, <span class="at">plot =</span> p, <span class="at">dpi =</span> <span class="dv">300</span>, <span class="at">width =</span> <span class="fl">4.7</span>, <span class="at">height =</span> <span class="dv">4</span>)</span>
<span id="cb5-21"><a href="appendix-f-creating-figures.html#cb5-21" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb5-22"><a href="appendix-f-creating-figures.html#cb5-22" aria-hidden="true" tabindex="-1"></a>p <span class="ot">&lt;-</span> <span class="fu">plot_umap_sign</span>(<span class="at">obj =</span> pbmc_4000, <span class="at">data_type =</span> <span class="st">&quot;KEGG&quot;</span>, <span class="at">category =</span> <span class="st">&quot;pathway&quot;</span>,</span>
<span id="cb5-23"><a href="appendix-f-creating-figures.html#cb5-23" aria-hidden="true" tabindex="-1"></a>                    <span class="at">algo_name =</span> <span class="st">&quot;seuratFindClusters&quot;</span>,</span>
<span id="cb5-24"><a href="appendix-f-creating-figures.html#cb5-24" aria-hidden="true" tabindex="-1"></a>                    <span class="at">theta =</span> <span class="cn">NULL</span>, <span class="at">phi =</span> <span class="cn">NULL</span>,</span>
<span id="cb5-25"><a href="appendix-f-creating-figures.html#cb5-25" aria-hidden="true" tabindex="-1"></a>                    <span class="at">title =</span> <span class="st">&quot;pbmc_4000 (KEGG: pathway)&quot;</span>, <span class="at">title_size =</span> <span class="dv">16</span>,</span>
<span id="cb5-26"><a href="appendix-f-creating-figures.html#cb5-26" aria-hidden="true" tabindex="-1"></a>                    <span class="at">xlabel =</span> <span class="st">&quot;UMAP_1&quot;</span>, <span class="at">ylabel =</span> <span class="st">&quot;UMAP_2&quot;</span>, <span class="at">zlabel =</span> <span class="cn">NULL</span>,</span>
<span id="cb5-27"><a href="appendix-f-creating-figures.html#cb5-27" aria-hidden="true" tabindex="-1"></a>                    <span class="at">default_color =</span> <span class="cn">TRUE</span>)</span>
<span id="cb5-28"><a href="appendix-f-creating-figures.html#cb5-28" aria-hidden="true" tabindex="-1"></a>filename <span class="ot">&lt;-</span> <span class="st">&quot;figures/figure_20_0022.png&quot;</span></span>
<span id="cb5-29"><a href="appendix-f-creating-figures.html#cb5-29" aria-hidden="true" tabindex="-1"></a><span class="fu">ggsave</span>(<span class="at">file =</span> filename, <span class="at">plot =</span> p, <span class="at">dpi =</span> <span class="dv">300</span>, <span class="at">width =</span> <span class="fl">4.7</span>, <span class="at">height =</span> <span class="dv">4</span>)</span>
<span id="cb5-30"><a href="appendix-f-creating-figures.html#cb5-30" aria-hidden="true" tabindex="-1"></a><span class="co"># ----------------------------------------</span></span>
<span id="cb5-31"><a href="appendix-f-creating-figures.html#cb5-31" aria-hidden="true" tabindex="-1"></a><span class="co"># pbmc_6000</span></span>
<span id="cb5-32"><a href="appendix-f-creating-figures.html#cb5-32" aria-hidden="true" tabindex="-1"></a><span class="co"># ----------------------------------------</span></span>
<span id="cb5-33"><a href="appendix-f-creating-figures.html#cb5-33" aria-hidden="true" tabindex="-1"></a>p <span class="ot">&lt;-</span> <span class="fu">plot_umap_sign</span>(<span class="at">obj =</span> pbmc_6000, <span class="at">data_type =</span> <span class="st">&quot;CO&quot;</span>, <span class="at">category =</span> <span class="st">&quot;cell&quot;</span>,</span>
<span id="cb5-34"><a href="appendix-f-creating-figures.html#cb5-34" aria-hidden="true" tabindex="-1"></a>                    <span class="at">algo_name =</span> <span class="st">&quot;seuratFindClusters&quot;</span>,</span>
<span id="cb5-35"><a href="appendix-f-creating-figures.html#cb5-35" aria-hidden="true" tabindex="-1"></a>                    <span class="at">theta =</span> <span class="cn">NULL</span>, <span class="at">phi =</span> <span class="cn">NULL</span>,</span>
<span id="cb5-36"><a href="appendix-f-creating-figures.html#cb5-36" aria-hidden="true" tabindex="-1"></a>                    <span class="at">title =</span> <span class="st">&quot;pbmc_6000 (CO: cell)&quot;</span>, <span class="at">title_size =</span> <span class="dv">16</span>,</span>
<span id="cb5-37"><a href="appendix-f-creating-figures.html#cb5-37" aria-hidden="true" tabindex="-1"></a>                    <span class="at">xlabel =</span> <span class="st">&quot;UMAP_1&quot;</span>, <span class="at">ylabel =</span> <span class="st">&quot;UMAP_2&quot;</span>, <span class="at">zlabel =</span> <span class="cn">NULL</span>,</span>
<span id="cb5-38"><a href="appendix-f-creating-figures.html#cb5-38" aria-hidden="true" tabindex="-1"></a>                    <span class="at">default_color =</span> <span class="cn">TRUE</span>)</span>
<span id="cb5-39"><a href="appendix-f-creating-figures.html#cb5-39" aria-hidden="true" tabindex="-1"></a>filename <span class="ot">&lt;-</span> <span class="st">&quot;figures/figure_20_0025.png&quot;</span></span>
<span id="cb5-40"><a href="appendix-f-creating-figures.html#cb5-40" aria-hidden="true" tabindex="-1"></a><span class="fu">ggsave</span>(<span class="at">file =</span> filename, <span class="at">plot =</span> p, <span class="at">dpi =</span> <span class="dv">300</span>, <span class="at">width =</span> <span class="fl">4.8</span>, <span class="at">height =</span> <span class="dv">4</span>)</span>
<span id="cb5-41"><a href="appendix-f-creating-figures.html#cb5-41" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb5-42"><a href="appendix-f-creating-figures.html#cb5-42" aria-hidden="true" tabindex="-1"></a>p <span class="ot">&lt;-</span> <span class="fu">plot_umap_sign</span>(<span class="at">obj =</span> pbmc_6000, <span class="at">data_type =</span> <span class="st">&quot;GO&quot;</span>, <span class="at">category =</span> <span class="st">&quot;BP&quot;</span>,</span>
<span id="cb5-43"><a href="appendix-f-creating-figures.html#cb5-43" aria-hidden="true" tabindex="-1"></a>                    <span class="at">algo_name =</span> <span class="st">&quot;seuratFindClusters&quot;</span>,</span>
<span id="cb5-44"><a href="appendix-f-creating-figures.html#cb5-44" aria-hidden="true" tabindex="-1"></a>                    <span class="at">theta =</span> <span class="cn">NULL</span>, <span class="at">phi =</span> <span class="cn">NULL</span>,</span>
<span id="cb5-45"><a href="appendix-f-creating-figures.html#cb5-45" aria-hidden="true" tabindex="-1"></a>                    <span class="at">title =</span> <span class="st">&quot;pbmc_6000 (GO: BP)&quot;</span>, <span class="at">title_size =</span> <span class="dv">16</span>,</span>
<span id="cb5-46"><a href="appendix-f-creating-figures.html#cb5-46" aria-hidden="true" tabindex="-1"></a>                    <span class="at">xlabel =</span> <span class="st">&quot;UMAP_1&quot;</span>, <span class="at">ylabel =</span> <span class="st">&quot;UMAP_2&quot;</span>, <span class="at">zlabel =</span> <span class="cn">NULL</span>,</span>
<span id="cb5-47"><a href="appendix-f-creating-figures.html#cb5-47" aria-hidden="true" tabindex="-1"></a>                    <span class="at">default_color =</span> <span class="cn">TRUE</span>)</span>
<span id="cb5-48"><a href="appendix-f-creating-figures.html#cb5-48" aria-hidden="true" tabindex="-1"></a>filename <span class="ot">&lt;-</span> <span class="st">&quot;figures/figure_20_0026.png&quot;</span></span>
<span id="cb5-49"><a href="appendix-f-creating-figures.html#cb5-49" aria-hidden="true" tabindex="-1"></a><span class="fu">ggsave</span>(<span class="at">file =</span> filename, <span class="at">plot =</span> p, <span class="at">dpi =</span> <span class="dv">300</span>, <span class="at">width =</span> <span class="fl">4.8</span>, <span class="at">height =</span> <span class="dv">4</span>)</span>
<span id="cb5-50"><a href="appendix-f-creating-figures.html#cb5-50" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb5-51"><a href="appendix-f-creating-figures.html#cb5-51" aria-hidden="true" tabindex="-1"></a>p <span class="ot">&lt;-</span> <span class="fu">plot_umap_sign</span>(<span class="at">obj =</span> pbmc_6000, <span class="at">data_type =</span> <span class="st">&quot;KEGG&quot;</span>, <span class="at">category =</span> <span class="st">&quot;pathway&quot;</span>,</span>
<span id="cb5-52"><a href="appendix-f-creating-figures.html#cb5-52" aria-hidden="true" tabindex="-1"></a>                    <span class="at">algo_name =</span> <span class="st">&quot;seuratFindClusters&quot;</span>,</span>
<span id="cb5-53"><a href="appendix-f-creating-figures.html#cb5-53" aria-hidden="true" tabindex="-1"></a>                    <span class="at">theta =</span> <span class="cn">NULL</span>, <span class="at">phi =</span> <span class="cn">NULL</span>,</span>
<span id="cb5-54"><a href="appendix-f-creating-figures.html#cb5-54" aria-hidden="true" tabindex="-1"></a>                    <span class="at">title =</span> <span class="st">&quot;pbmc_6000 (KEGG: pathway)&quot;</span>, <span class="at">title_size =</span> <span class="dv">16</span>,</span>
<span id="cb5-55"><a href="appendix-f-creating-figures.html#cb5-55" aria-hidden="true" tabindex="-1"></a>                    <span class="at">xlabel =</span> <span class="st">&quot;UMAP_1&quot;</span>, <span class="at">ylabel =</span> <span class="st">&quot;UMAP_2&quot;</span>, <span class="at">zlabel =</span> <span class="cn">NULL</span>,</span>
<span id="cb5-56"><a href="appendix-f-creating-figures.html#cb5-56" aria-hidden="true" tabindex="-1"></a>                    <span class="at">default_color =</span> <span class="cn">TRUE</span>)</span>
<span id="cb5-57"><a href="appendix-f-creating-figures.html#cb5-57" aria-hidden="true" tabindex="-1"></a>filename <span class="ot">&lt;-</span> <span class="st">&quot;figures/figure_20_0027.png&quot;</span></span>
<span id="cb5-58"><a href="appendix-f-creating-figures.html#cb5-58" aria-hidden="true" tabindex="-1"></a><span class="fu">ggsave</span>(<span class="at">file =</span> filename, <span class="at">plot =</span> p, <span class="at">dpi =</span> <span class="dv">300</span>, <span class="at">width =</span> <span class="fl">4.8</span>, <span class="at">height =</span> <span class="dv">4</span>)</span></code></pre></div>
<p><img src="figures/figure_20_0020.png" width="250px">
<img src="figures/figure_20_0021.png" width="250px">
<img src="figures/figure_20_0022.png" width="250px">
<img src="figures/figure_20_0025.png" width="250px">
<img src="figures/figure_20_0026.png" width="250px">
<img src="figures/figure_20_0027.png" width="250px"></p>
<p>Make violin plots.</p>
<div class="sourceCode" id="cb6"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb6-1"><a href="appendix-f-creating-figures.html#cb6-1" aria-hidden="true" tabindex="-1"></a><span class="co"># ----------------------------------------</span></span>
<span id="cb6-2"><a href="appendix-f-creating-figures.html#cb6-2" aria-hidden="true" tabindex="-1"></a><span class="co"># pbmc_4000</span></span>
<span id="cb6-3"><a href="appendix-f-creating-figures.html#cb6-3" aria-hidden="true" tabindex="-1"></a><span class="co"># ----------------------------------------</span></span>
<span id="cb6-4"><a href="appendix-f-creating-figures.html#cb6-4" aria-hidden="true" tabindex="-1"></a><span class="co">#------------------------------</span></span>
<span id="cb6-5"><a href="appendix-f-creating-figures.html#cb6-5" aria-hidden="true" tabindex="-1"></a><span class="co"># CO</span></span>
<span id="cb6-6"><a href="appendix-f-creating-figures.html#cb6-6" aria-hidden="true" tabindex="-1"></a><span class="co">#------------------------------</span></span>
<span id="cb6-7"><a href="appendix-f-creating-figures.html#cb6-7" aria-hidden="true" tabindex="-1"></a>p <span class="ot">&lt;-</span> <span class="fu">plot_violin_signScore</span>(</span>
<span id="cb6-8"><a href="appendix-f-creating-figures.html#cb6-8" aria-hidden="true" tabindex="-1"></a>  <span class="at">obj =</span> pbmc_4000, <span class="at">sign_name =</span> <span class="st">&quot;CL:0000576_S&quot;</span>,</span>
<span id="cb6-9"><a href="appendix-f-creating-figures.html#cb6-9" aria-hidden="true" tabindex="-1"></a>  <span class="at">data_type_for_label =</span> <span class="st">&quot;CO&quot;</span>, <span class="at">category_for_label =</span> <span class="st">&quot;cell&quot;</span>,</span>
<span id="cb6-10"><a href="appendix-f-creating-figures.html#cb6-10" aria-hidden="true" tabindex="-1"></a>  <span class="at">algo_name_for_label =</span> <span class="st">&quot;seuratFindClusters&quot;</span>,</span>
<span id="cb6-11"><a href="appendix-f-creating-figures.html#cb6-11" aria-hidden="true" tabindex="-1"></a>  <span class="at">data_type_for_expr =</span> <span class="st">&quot;CO&quot;</span>, <span class="at">category_for_expr =</span> <span class="st">&quot;cell&quot;</span>,</span>
<span id="cb6-12"><a href="appendix-f-creating-figures.html#cb6-12" aria-hidden="true" tabindex="-1"></a>  <span class="at">title =</span> <span class="st">&quot;CL:0000576_S</span><span class="sc">\n</span><span class="st">Monocyte</span><span class="sc">\n</span><span class="st">FGS = {MEF2C, LYN, ...}&quot;</span>,</span>
<span id="cb6-13"><a href="appendix-f-creating-figures.html#cb6-13" aria-hidden="true" tabindex="-1"></a>  <span class="at">title_size =</span> <span class="dv">16</span>, <span class="at">default_color =</span> <span class="cn">TRUE</span>)</span>
<span id="cb6-14"><a href="appendix-f-creating-figures.html#cb6-14" aria-hidden="true" tabindex="-1"></a>filename <span class="ot">&lt;-</span> <span class="st">&quot;figures/figure_20_0030.png&quot;</span></span>
<span id="cb6-15"><a href="appendix-f-creating-figures.html#cb6-15" aria-hidden="true" tabindex="-1"></a><span class="fu">ggsave</span>(<span class="at">file =</span> filename, <span class="at">plot =</span> p, <span class="at">dpi =</span> <span class="dv">300</span>, <span class="at">width =</span> <span class="dv">4</span>, <span class="at">height =</span> <span class="fl">3.5</span>)</span>
<span id="cb6-16"><a href="appendix-f-creating-figures.html#cb6-16" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb6-17"><a href="appendix-f-creating-figures.html#cb6-17" aria-hidden="true" tabindex="-1"></a>p <span class="ot">&lt;-</span> <span class="fu">plot_violin_signScore</span>(</span>
<span id="cb6-18"><a href="appendix-f-creating-figures.html#cb6-18" aria-hidden="true" tabindex="-1"></a>  <span class="at">obj =</span> pbmc_4000, <span class="at">sign_name =</span> <span class="st">&quot;CL:0000623_V&quot;</span>,</span>
<span id="cb6-19"><a href="appendix-f-creating-figures.html#cb6-19" aria-hidden="true" tabindex="-1"></a>  <span class="at">data_type_for_label =</span> <span class="st">&quot;CO&quot;</span>, <span class="at">category_for_label =</span> <span class="st">&quot;cell&quot;</span>,</span>
<span id="cb6-20"><a href="appendix-f-creating-figures.html#cb6-20" aria-hidden="true" tabindex="-1"></a>  <span class="at">algo_name_for_label =</span> <span class="st">&quot;seuratFindClusters&quot;</span>,</span>
<span id="cb6-21"><a href="appendix-f-creating-figures.html#cb6-21" aria-hidden="true" tabindex="-1"></a>  <span class="at">data_type_for_expr =</span> <span class="st">&quot;CO&quot;</span>, <span class="at">category_for_expr =</span> <span class="st">&quot;cell&quot;</span>,</span>
<span id="cb6-22"><a href="appendix-f-creating-figures.html#cb6-22" aria-hidden="true" tabindex="-1"></a>  <span class="at">title =</span> <span class="st">&quot;CL:0000623_V</span><span class="sc">\n</span><span class="st">Natural killer cell</span><span class="sc">\n</span><span class="st">FGS = {SH2D1A, KLRD1, ...}&quot;</span>,</span>
<span id="cb6-23"><a href="appendix-f-creating-figures.html#cb6-23" aria-hidden="true" tabindex="-1"></a>  <span class="at">title_size =</span> <span class="dv">16</span>, <span class="at">default_color =</span> <span class="cn">TRUE</span>)</span>
<span id="cb6-24"><a href="appendix-f-creating-figures.html#cb6-24" aria-hidden="true" tabindex="-1"></a>filename <span class="ot">&lt;-</span> <span class="st">&quot;figures/figure_20_0031.png&quot;</span></span>
<span id="cb6-25"><a href="appendix-f-creating-figures.html#cb6-25" aria-hidden="true" tabindex="-1"></a><span class="fu">ggsave</span>(<span class="at">file =</span> filename, <span class="at">plot =</span> p, <span class="at">dpi =</span> <span class="dv">300</span>, <span class="at">width =</span> <span class="dv">4</span>, <span class="at">height =</span> <span class="fl">3.5</span>)</span>
<span id="cb6-26"><a href="appendix-f-creating-figures.html#cb6-26" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb6-27"><a href="appendix-f-creating-figures.html#cb6-27" aria-hidden="true" tabindex="-1"></a>p <span class="ot">&lt;-</span> <span class="fu">plot_violin_signScore</span>(</span>
<span id="cb6-28"><a href="appendix-f-creating-figures.html#cb6-28" aria-hidden="true" tabindex="-1"></a>  <span class="at">obj =</span> pbmc_4000, <span class="at">sign_name =</span> <span class="st">&quot;CL:0000789_S&quot;</span>,</span>
<span id="cb6-29"><a href="appendix-f-creating-figures.html#cb6-29" aria-hidden="true" tabindex="-1"></a>  <span class="at">data_type_for_label =</span> <span class="st">&quot;CO&quot;</span>, <span class="at">category_for_label =</span> <span class="st">&quot;cell&quot;</span>,</span>
<span id="cb6-30"><a href="appendix-f-creating-figures.html#cb6-30" aria-hidden="true" tabindex="-1"></a>  <span class="at">algo_name_for_label =</span> <span class="st">&quot;seuratFindClusters&quot;</span>,</span>
<span id="cb6-31"><a href="appendix-f-creating-figures.html#cb6-31" aria-hidden="true" tabindex="-1"></a>  <span class="at">data_type_for_expr =</span> <span class="st">&quot;CO&quot;</span>, <span class="at">category_for_expr =</span> <span class="st">&quot;cell&quot;</span>,</span>
<span id="cb6-32"><a href="appendix-f-creating-figures.html#cb6-32" aria-hidden="true" tabindex="-1"></a>  <span class="at">title =</span> <span class="st">&quot;CL:0000789_S</span><span class="sc">\n</span><span class="st">Alpha-beta T cell</span><span class="sc">\n</span><span class="st">FGS = {CD3E, CD3D, ...}&quot;</span>,</span>
<span id="cb6-33"><a href="appendix-f-creating-figures.html#cb6-33" aria-hidden="true" tabindex="-1"></a>  <span class="at">title_size =</span> <span class="dv">16</span>, <span class="at">default_color =</span> <span class="cn">TRUE</span>)</span>
<span id="cb6-34"><a href="appendix-f-creating-figures.html#cb6-34" aria-hidden="true" tabindex="-1"></a>filename <span class="ot">&lt;-</span> <span class="st">&quot;figures/figure_20_0032.png&quot;</span></span>
<span id="cb6-35"><a href="appendix-f-creating-figures.html#cb6-35" aria-hidden="true" tabindex="-1"></a><span class="fu">ggsave</span>(<span class="at">file =</span> filename, <span class="at">plot =</span> p, <span class="at">dpi =</span> <span class="dv">300</span>, <span class="at">width =</span> <span class="dv">4</span>, <span class="at">height =</span> <span class="fl">3.5</span>)</span>
<span id="cb6-36"><a href="appendix-f-creating-figures.html#cb6-36" aria-hidden="true" tabindex="-1"></a><span class="co">#------------------------------</span></span>
<span id="cb6-37"><a href="appendix-f-creating-figures.html#cb6-37" aria-hidden="true" tabindex="-1"></a><span class="co"># GO</span></span>
<span id="cb6-38"><a href="appendix-f-creating-figures.html#cb6-38" aria-hidden="true" tabindex="-1"></a><span class="co">#------------------------------</span></span>
<span id="cb6-39"><a href="appendix-f-creating-figures.html#cb6-39" aria-hidden="true" tabindex="-1"></a>p <span class="ot">&lt;-</span> <span class="fu">plot_violin_signScore</span>(</span>
<span id="cb6-40"><a href="appendix-f-creating-figures.html#cb6-40" aria-hidden="true" tabindex="-1"></a>  <span class="at">obj =</span> pbmc_4000, <span class="at">sign_name =</span> <span class="st">&quot;GO:0001914_S&quot;</span>,</span>
<span id="cb6-41"><a href="appendix-f-creating-figures.html#cb6-41" aria-hidden="true" tabindex="-1"></a>  <span class="at">data_type_for_label =</span> <span class="st">&quot;GO&quot;</span>, <span class="at">category_for_label =</span> <span class="st">&quot;BP&quot;</span>,</span>
<span id="cb6-42"><a href="appendix-f-creating-figures.html#cb6-42" aria-hidden="true" tabindex="-1"></a>  <span class="at">algo_name_for_label =</span> <span class="st">&quot;seuratFindClusters&quot;</span>,</span>
<span id="cb6-43"><a href="appendix-f-creating-figures.html#cb6-43" aria-hidden="true" tabindex="-1"></a>  <span class="at">data_type_for_expr =</span> <span class="st">&quot;GO&quot;</span>, <span class="at">category_for_expr =</span> <span class="st">&quot;BP&quot;</span>,</span>
<span id="cb6-44"><a href="appendix-f-creating-figures.html#cb6-44" aria-hidden="true" tabindex="-1"></a>  <span class="at">title =</span> <span class="st">&quot;GO:0001914_S</span><span class="sc">\n</span><span class="st">Regulation of T cell..</span><span class="sc">\n</span><span class="st">FGS = {PTPRC, IL7R, ...}&quot;</span>,</span>
<span id="cb6-45"><a href="appendix-f-creating-figures.html#cb6-45" aria-hidden="true" tabindex="-1"></a>  <span class="at">title_size =</span> <span class="dv">16</span>, <span class="at">default_color =</span> <span class="cn">FALSE</span>)</span>
<span id="cb6-46"><a href="appendix-f-creating-figures.html#cb6-46" aria-hidden="true" tabindex="-1"></a>filename <span class="ot">&lt;-</span> <span class="st">&quot;figures/figure_20_0033.png&quot;</span></span>
<span id="cb6-47"><a href="appendix-f-creating-figures.html#cb6-47" aria-hidden="true" tabindex="-1"></a><span class="fu">ggsave</span>(<span class="at">file =</span> filename, <span class="at">plot =</span> p, <span class="at">dpi =</span> <span class="dv">300</span>, <span class="at">width =</span> <span class="dv">4</span>, <span class="at">height =</span> <span class="fl">3.5</span>)</span>
<span id="cb6-48"><a href="appendix-f-creating-figures.html#cb6-48" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb6-49"><a href="appendix-f-creating-figures.html#cb6-49" aria-hidden="true" tabindex="-1"></a>p <span class="ot">&lt;-</span> <span class="fu">plot_violin_signScore</span>(</span>
<span id="cb6-50"><a href="appendix-f-creating-figures.html#cb6-50" aria-hidden="true" tabindex="-1"></a>  <span class="at">obj =</span> pbmc_4000, <span class="at">sign_name =</span> <span class="st">&quot;GO:0050853_V&quot;</span>,</span>
<span id="cb6-51"><a href="appendix-f-creating-figures.html#cb6-51" aria-hidden="true" tabindex="-1"></a>  <span class="at">data_type_for_label =</span> <span class="st">&quot;GO&quot;</span>, <span class="at">category_for_label =</span> <span class="st">&quot;BP&quot;</span>,</span>
<span id="cb6-52"><a href="appendix-f-creating-figures.html#cb6-52" aria-hidden="true" tabindex="-1"></a>  <span class="at">algo_name_for_label =</span> <span class="st">&quot;seuratFindClusters&quot;</span>,</span>
<span id="cb6-53"><a href="appendix-f-creating-figures.html#cb6-53" aria-hidden="true" tabindex="-1"></a>  <span class="at">data_type_for_expr =</span> <span class="st">&quot;GO&quot;</span>, <span class="at">category_for_expr =</span> <span class="st">&quot;BP&quot;</span>,</span>
<span id="cb6-54"><a href="appendix-f-creating-figures.html#cb6-54" aria-hidden="true" tabindex="-1"></a>  <span class="at">title =</span> <span class="st">&quot;GO:0050853_V</span><span class="sc">\n</span><span class="st">B cell differentiation</span><span class="sc">\n</span><span class="st">FGS = {CD19, CD79B, ...}&quot;</span>,</span>
<span id="cb6-55"><a href="appendix-f-creating-figures.html#cb6-55" aria-hidden="true" tabindex="-1"></a>  <span class="at">title_size =</span> <span class="dv">16</span>, <span class="at">default_color =</span> <span class="cn">FALSE</span>)</span>
<span id="cb6-56"><a href="appendix-f-creating-figures.html#cb6-56" aria-hidden="true" tabindex="-1"></a>filename <span class="ot">&lt;-</span> <span class="st">&quot;figures/figure_20_0034.png&quot;</span></span>
<span id="cb6-57"><a href="appendix-f-creating-figures.html#cb6-57" aria-hidden="true" tabindex="-1"></a><span class="fu">ggsave</span>(<span class="at">file =</span> filename, <span class="at">plot =</span> p, <span class="at">dpi =</span> <span class="dv">300</span>, <span class="at">width =</span> <span class="dv">4</span>, <span class="at">height =</span> <span class="fl">3.5</span>)</span>
<span id="cb6-58"><a href="appendix-f-creating-figures.html#cb6-58" aria-hidden="true" tabindex="-1"></a><span class="co">#------------------------------</span></span>
<span id="cb6-59"><a href="appendix-f-creating-figures.html#cb6-59" aria-hidden="true" tabindex="-1"></a><span class="co"># KEGG</span></span>
<span id="cb6-60"><a href="appendix-f-creating-figures.html#cb6-60" aria-hidden="true" tabindex="-1"></a><span class="co">#------------------------------</span></span>
<span id="cb6-61"><a href="appendix-f-creating-figures.html#cb6-61" aria-hidden="true" tabindex="-1"></a>p <span class="ot">&lt;-</span> <span class="fu">plot_violin_signScore</span>(</span>
<span id="cb6-62"><a href="appendix-f-creating-figures.html#cb6-62" aria-hidden="true" tabindex="-1"></a>  <span class="at">obj =</span> pbmc_4000, <span class="at">sign_name =</span> <span class="st">&quot;path:hsa04662_V&quot;</span>,</span>
<span id="cb6-63"><a href="appendix-f-creating-figures.html#cb6-63" aria-hidden="true" tabindex="-1"></a>  <span class="at">data_type_for_label =</span> <span class="st">&quot;KEGG&quot;</span>, <span class="at">category_for_label =</span> <span class="st">&quot;pathway&quot;</span>,</span>
<span id="cb6-64"><a href="appendix-f-creating-figures.html#cb6-64" aria-hidden="true" tabindex="-1"></a>  <span class="at">algo_name_for_label =</span> <span class="st">&quot;seuratFindClusters&quot;</span>,</span>
<span id="cb6-65"><a href="appendix-f-creating-figures.html#cb6-65" aria-hidden="true" tabindex="-1"></a>  <span class="at">data_type_for_expr =</span> <span class="st">&quot;KEGG&quot;</span>, <span class="at">category_for_expr =</span> <span class="st">&quot;pathway&quot;</span>,</span>
<span id="cb6-66"><a href="appendix-f-creating-figures.html#cb6-66" aria-hidden="true" tabindex="-1"></a>  <span class="at">title =</span> <span class="st">&quot;path:hsa04662_V</span><span class="sc">\n</span><span class="st">B cell receptor..</span><span class="sc">\n</span><span class="st">FGS = {CD19, CD72, ..}&quot;</span>,</span>
<span id="cb6-67"><a href="appendix-f-creating-figures.html#cb6-67" aria-hidden="true" tabindex="-1"></a>  <span class="at">title_size =</span> <span class="dv">16</span>, <span class="at">default_color =</span> <span class="cn">FALSE</span>)</span>
<span id="cb6-68"><a href="appendix-f-creating-figures.html#cb6-68" aria-hidden="true" tabindex="-1"></a>filename <span class="ot">&lt;-</span> <span class="st">&quot;figures/figure_20_0035.png&quot;</span></span>
<span id="cb6-69"><a href="appendix-f-creating-figures.html#cb6-69" aria-hidden="true" tabindex="-1"></a><span class="fu">ggsave</span>(<span class="at">file =</span> filename, <span class="at">plot =</span> p, <span class="at">dpi =</span> <span class="dv">300</span>, <span class="at">width =</span> <span class="dv">4</span>, <span class="at">height =</span> <span class="fl">3.5</span>)</span></code></pre></div>
<p><img src="figures/figure_20_0030.png" width="250px">
<img src="figures/figure_20_0031.png" width="250px">
<img src="figures/figure_20_0032.png" width="250px">
<img src="figures/figure_20_0033.png" width="250px">
<img src="figures/figure_20_0034.png" width="250px">
<img src="figures/figure_20_0035.png" width="250px"></p>
<div class="sourceCode" id="cb7"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb7-1"><a href="appendix-f-creating-figures.html#cb7-1" aria-hidden="true" tabindex="-1"></a><span class="co"># ----------------------------------------</span></span>
<span id="cb7-2"><a href="appendix-f-creating-figures.html#cb7-2" aria-hidden="true" tabindex="-1"></a><span class="co"># pbmc_6000</span></span>
<span id="cb7-3"><a href="appendix-f-creating-figures.html#cb7-3" aria-hidden="true" tabindex="-1"></a><span class="co"># ----------------------------------------</span></span>
<span id="cb7-4"><a href="appendix-f-creating-figures.html#cb7-4" aria-hidden="true" tabindex="-1"></a><span class="co">#------------------------------</span></span>
<span id="cb7-5"><a href="appendix-f-creating-figures.html#cb7-5" aria-hidden="true" tabindex="-1"></a><span class="co"># CO</span></span>
<span id="cb7-6"><a href="appendix-f-creating-figures.html#cb7-6" aria-hidden="true" tabindex="-1"></a><span class="co">#------------------------------</span></span>
<span id="cb7-7"><a href="appendix-f-creating-figures.html#cb7-7" aria-hidden="true" tabindex="-1"></a>p <span class="ot">&lt;-</span> <span class="fu">plot_violin_signScore</span>(</span>
<span id="cb7-8"><a href="appendix-f-creating-figures.html#cb7-8" aria-hidden="true" tabindex="-1"></a>  <span class="at">obj =</span> pbmc_6000, <span class="at">sign_name =</span> <span class="st">&quot;CL:0000234_S&quot;</span>,</span>
<span id="cb7-9"><a href="appendix-f-creating-figures.html#cb7-9" aria-hidden="true" tabindex="-1"></a>  <span class="at">data_type_for_label =</span> <span class="st">&quot;CO&quot;</span>, <span class="at">category_for_label =</span> <span class="st">&quot;cell&quot;</span>,</span>
<span id="cb7-10"><a href="appendix-f-creating-figures.html#cb7-10" aria-hidden="true" tabindex="-1"></a>  <span class="at">algo_name_for_label =</span> <span class="st">&quot;seuratFindClusters&quot;</span>,</span>
<span id="cb7-11"><a href="appendix-f-creating-figures.html#cb7-11" aria-hidden="true" tabindex="-1"></a>  <span class="at">data_type_for_expr =</span> <span class="st">&quot;CO&quot;</span>, <span class="at">category_for_expr =</span> <span class="st">&quot;cell&quot;</span>,</span>
<span id="cb7-12"><a href="appendix-f-creating-figures.html#cb7-12" aria-hidden="true" tabindex="-1"></a>  <span class="at">title =</span> <span class="st">&quot;CL:0000234_S</span><span class="sc">\n</span><span class="st">Phagocyte</span><span class="sc">\n</span><span class="st">FGS = {CD14, FGR, ...}&quot;</span>,</span>
<span id="cb7-13"><a href="appendix-f-creating-figures.html#cb7-13" aria-hidden="true" tabindex="-1"></a>  <span class="at">title_size =</span> <span class="dv">16</span>, <span class="at">default_color =</span> <span class="cn">TRUE</span>)</span>
<span id="cb7-14"><a href="appendix-f-creating-figures.html#cb7-14" aria-hidden="true" tabindex="-1"></a>filename <span class="ot">&lt;-</span> <span class="st">&quot;figures/figure_20_0040.png&quot;</span></span>
<span id="cb7-15"><a href="appendix-f-creating-figures.html#cb7-15" aria-hidden="true" tabindex="-1"></a><span class="fu">ggsave</span>(<span class="at">file =</span> filename, <span class="at">plot =</span> p, <span class="at">dpi =</span> <span class="dv">300</span>, <span class="at">width =</span> <span class="dv">4</span>, <span class="at">height =</span> <span class="fl">3.5</span>)</span>
<span id="cb7-16"><a href="appendix-f-creating-figures.html#cb7-16" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb7-17"><a href="appendix-f-creating-figures.html#cb7-17" aria-hidden="true" tabindex="-1"></a>p <span class="ot">&lt;-</span> <span class="fu">plot_violin_signScore</span>(</span>
<span id="cb7-18"><a href="appendix-f-creating-figures.html#cb7-18" aria-hidden="true" tabindex="-1"></a>  <span class="at">obj =</span> pbmc_6000, <span class="at">sign_name =</span> <span class="st">&quot;CL:0000623_S&quot;</span>,</span>
<span id="cb7-19"><a href="appendix-f-creating-figures.html#cb7-19" aria-hidden="true" tabindex="-1"></a>  <span class="at">data_type_for_label =</span> <span class="st">&quot;CO&quot;</span>, <span class="at">category_for_label =</span> <span class="st">&quot;cell&quot;</span>,</span>
<span id="cb7-20"><a href="appendix-f-creating-figures.html#cb7-20" aria-hidden="true" tabindex="-1"></a>  <span class="at">algo_name_for_label =</span> <span class="st">&quot;seuratFindClusters&quot;</span>,</span>
<span id="cb7-21"><a href="appendix-f-creating-figures.html#cb7-21" aria-hidden="true" tabindex="-1"></a>  <span class="at">data_type_for_expr =</span> <span class="st">&quot;CO&quot;</span>, <span class="at">category_for_expr =</span> <span class="st">&quot;cell&quot;</span>,</span>
<span id="cb7-22"><a href="appendix-f-creating-figures.html#cb7-22" aria-hidden="true" tabindex="-1"></a>  <span class="at">title =</span> <span class="st">&quot;CL:0000623_S</span><span class="sc">\n</span><span class="st">Natural killer cell</span><span class="sc">\n</span><span class="st">FGS = {CD160, KLRD1, ...}&quot;</span>,</span>
<span id="cb7-23"><a href="appendix-f-creating-figures.html#cb7-23" aria-hidden="true" tabindex="-1"></a>  <span class="at">title_size =</span> <span class="dv">16</span>, <span class="at">default_color =</span> <span class="cn">TRUE</span>)</span>
<span id="cb7-24"><a href="appendix-f-creating-figures.html#cb7-24" aria-hidden="true" tabindex="-1"></a>filename <span class="ot">&lt;-</span> <span class="st">&quot;figures/figure_20_0041.png&quot;</span></span>
<span id="cb7-25"><a href="appendix-f-creating-figures.html#cb7-25" aria-hidden="true" tabindex="-1"></a><span class="fu">ggsave</span>(<span class="at">file =</span> filename, <span class="at">plot =</span> p, <span class="at">dpi =</span> <span class="dv">300</span>, <span class="at">width =</span> <span class="dv">4</span>, <span class="at">height =</span> <span class="fl">3.5</span>)</span>
<span id="cb7-26"><a href="appendix-f-creating-figures.html#cb7-26" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb7-27"><a href="appendix-f-creating-figures.html#cb7-27" aria-hidden="true" tabindex="-1"></a>p <span class="ot">&lt;-</span> <span class="fu">plot_violin_signScore</span>(</span>
<span id="cb7-28"><a href="appendix-f-creating-figures.html#cb7-28" aria-hidden="true" tabindex="-1"></a>  <span class="at">obj =</span> pbmc_6000, <span class="at">sign_name =</span> <span class="st">&quot;CL:0000782_V&quot;</span>,</span>
<span id="cb7-29"><a href="appendix-f-creating-figures.html#cb7-29" aria-hidden="true" tabindex="-1"></a>  <span class="at">data_type_for_label =</span> <span class="st">&quot;CO&quot;</span>, <span class="at">category_for_label =</span> <span class="st">&quot;cell&quot;</span>,</span>
<span id="cb7-30"><a href="appendix-f-creating-figures.html#cb7-30" aria-hidden="true" tabindex="-1"></a>  <span class="at">algo_name_for_label =</span> <span class="st">&quot;seuratFindClusters&quot;</span>,</span>
<span id="cb7-31"><a href="appendix-f-creating-figures.html#cb7-31" aria-hidden="true" tabindex="-1"></a>  <span class="at">data_type_for_expr =</span> <span class="st">&quot;CO&quot;</span>, <span class="at">category_for_expr =</span> <span class="st">&quot;cell&quot;</span>,</span>
<span id="cb7-32"><a href="appendix-f-creating-figures.html#cb7-32" aria-hidden="true" tabindex="-1"></a>  <span class="at">title =</span> <span class="st">&quot;CL:0000782_V</span><span class="sc">\n</span><span class="st">Myeloid dendritic cell</span><span class="sc">\n</span><span class="st">FGS = {CD2, CCR7, ...}&quot;</span>,</span>
<span id="cb7-33"><a href="appendix-f-creating-figures.html#cb7-33" aria-hidden="true" tabindex="-1"></a>  <span class="at">title_size =</span> <span class="dv">16</span>, <span class="at">default_color =</span> <span class="cn">TRUE</span>)</span>
<span id="cb7-34"><a href="appendix-f-creating-figures.html#cb7-34" aria-hidden="true" tabindex="-1"></a>filename <span class="ot">&lt;-</span> <span class="st">&quot;figures/figure_20_0042.png&quot;</span></span>
<span id="cb7-35"><a href="appendix-f-creating-figures.html#cb7-35" aria-hidden="true" tabindex="-1"></a><span class="fu">ggsave</span>(<span class="at">file =</span> filename, <span class="at">plot =</span> p, <span class="at">dpi =</span> <span class="dv">300</span>, <span class="at">width =</span> <span class="dv">4</span>, <span class="at">height =</span> <span class="fl">3.5</span>)</span>
<span id="cb7-36"><a href="appendix-f-creating-figures.html#cb7-36" aria-hidden="true" tabindex="-1"></a><span class="co">#------------------------------</span></span>
<span id="cb7-37"><a href="appendix-f-creating-figures.html#cb7-37" aria-hidden="true" tabindex="-1"></a><span class="co"># GO</span></span>
<span id="cb7-38"><a href="appendix-f-creating-figures.html#cb7-38" aria-hidden="true" tabindex="-1"></a><span class="co">#------------------------------</span></span>
<span id="cb7-39"><a href="appendix-f-creating-figures.html#cb7-39" aria-hidden="true" tabindex="-1"></a>p <span class="ot">&lt;-</span> <span class="fu">plot_violin_signScore</span>(</span>
<span id="cb7-40"><a href="appendix-f-creating-figures.html#cb7-40" aria-hidden="true" tabindex="-1"></a>  <span class="at">obj =</span> pbmc_6000, <span class="at">sign_name =</span> <span class="st">&quot;GO:0050852_V&quot;</span>,</span>
<span id="cb7-41"><a href="appendix-f-creating-figures.html#cb7-41" aria-hidden="true" tabindex="-1"></a>  <span class="at">data_type_for_label =</span> <span class="st">&quot;GO&quot;</span>, <span class="at">category_for_label =</span> <span class="st">&quot;BP&quot;</span>,</span>
<span id="cb7-42"><a href="appendix-f-creating-figures.html#cb7-42" aria-hidden="true" tabindex="-1"></a>  <span class="at">algo_name_for_label =</span> <span class="st">&quot;seuratFindClusters&quot;</span>,</span>
<span id="cb7-43"><a href="appendix-f-creating-figures.html#cb7-43" aria-hidden="true" tabindex="-1"></a>  <span class="at">data_type_for_expr =</span> <span class="st">&quot;GO&quot;</span>, <span class="at">category_for_expr =</span> <span class="st">&quot;BP&quot;</span>,</span>
<span id="cb7-44"><a href="appendix-f-creating-figures.html#cb7-44" aria-hidden="true" tabindex="-1"></a>  <span class="at">title =</span> <span class="st">&quot;GO:0050852_V</span><span class="sc">\n</span><span class="st">T cell receptor..</span><span class="sc">\n</span><span class="st">FGS = {CD3E, CD247, ...}&quot;</span>,</span>
<span id="cb7-45"><a href="appendix-f-creating-figures.html#cb7-45" aria-hidden="true" tabindex="-1"></a>  <span class="at">title_size =</span> <span class="dv">16</span>, <span class="at">default_color =</span> <span class="cn">FALSE</span>)</span>
<span id="cb7-46"><a href="appendix-f-creating-figures.html#cb7-46" aria-hidden="true" tabindex="-1"></a>filename <span class="ot">&lt;-</span> <span class="st">&quot;figures/figure_20_0043.png&quot;</span></span>
<span id="cb7-47"><a href="appendix-f-creating-figures.html#cb7-47" aria-hidden="true" tabindex="-1"></a><span class="fu">ggsave</span>(<span class="at">file =</span> filename, <span class="at">plot =</span> p, <span class="at">dpi =</span> <span class="dv">300</span>, <span class="at">width =</span> <span class="dv">4</span>, <span class="at">height =</span> <span class="fl">3.5</span>)</span>
<span id="cb7-48"><a href="appendix-f-creating-figures.html#cb7-48" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb7-49"><a href="appendix-f-creating-figures.html#cb7-49" aria-hidden="true" tabindex="-1"></a>p <span class="ot">&lt;-</span> <span class="fu">plot_violin_signScore</span>(</span>
<span id="cb7-50"><a href="appendix-f-creating-figures.html#cb7-50" aria-hidden="true" tabindex="-1"></a>  <span class="at">obj =</span> pbmc_6000, <span class="at">sign_name =</span> <span class="st">&quot;GO:0002577_V&quot;</span>,</span>
<span id="cb7-51"><a href="appendix-f-creating-figures.html#cb7-51" aria-hidden="true" tabindex="-1"></a>  <span class="at">data_type_for_label =</span> <span class="st">&quot;GO&quot;</span>, <span class="at">category_for_label =</span> <span class="st">&quot;BP&quot;</span>,</span>
<span id="cb7-52"><a href="appendix-f-creating-figures.html#cb7-52" aria-hidden="true" tabindex="-1"></a>  <span class="at">algo_name_for_label =</span> <span class="st">&quot;seuratFindClusters&quot;</span>,</span>
<span id="cb7-53"><a href="appendix-f-creating-figures.html#cb7-53" aria-hidden="true" tabindex="-1"></a>  <span class="at">data_type_for_expr =</span> <span class="st">&quot;GO&quot;</span>, <span class="at">category_for_expr =</span> <span class="st">&quot;BP&quot;</span>,</span>
<span id="cb7-54"><a href="appendix-f-creating-figures.html#cb7-54" aria-hidden="true" tabindex="-1"></a>  <span class="at">title =</span> <span class="st">&quot;GO:0002577_V</span><span class="sc">\n</span><span class="st">Regulation of antigen..</span><span class="sc">\n</span><span class="st">FGS = {HLA-DOB, CCR7, ...}&quot;</span>,</span>
<span id="cb7-55"><a href="appendix-f-creating-figures.html#cb7-55" aria-hidden="true" tabindex="-1"></a>  <span class="at">title_size =</span> <span class="dv">16</span>, <span class="at">default_color =</span> <span class="cn">FALSE</span>)</span>
<span id="cb7-56"><a href="appendix-f-creating-figures.html#cb7-56" aria-hidden="true" tabindex="-1"></a>filename <span class="ot">&lt;-</span> <span class="st">&quot;figures/figure_20_0044.png&quot;</span></span>
<span id="cb7-57"><a href="appendix-f-creating-figures.html#cb7-57" aria-hidden="true" tabindex="-1"></a><span class="fu">ggsave</span>(<span class="at">file =</span> filename, <span class="at">plot =</span> p, <span class="at">dpi =</span> <span class="dv">300</span>, <span class="at">width =</span> <span class="dv">4</span>, <span class="at">height =</span> <span class="fl">3.5</span>)</span>
<span id="cb7-58"><a href="appendix-f-creating-figures.html#cb7-58" aria-hidden="true" tabindex="-1"></a><span class="co">#------------------------------</span></span>
<span id="cb7-59"><a href="appendix-f-creating-figures.html#cb7-59" aria-hidden="true" tabindex="-1"></a><span class="co"># KEGG</span></span>
<span id="cb7-60"><a href="appendix-f-creating-figures.html#cb7-60" aria-hidden="true" tabindex="-1"></a><span class="co">#------------------------------</span></span>
<span id="cb7-61"><a href="appendix-f-creating-figures.html#cb7-61" aria-hidden="true" tabindex="-1"></a>p <span class="ot">&lt;-</span> <span class="fu">plot_violin_signScore</span>(</span>
<span id="cb7-62"><a href="appendix-f-creating-figures.html#cb7-62" aria-hidden="true" tabindex="-1"></a>  <span class="at">obj =</span> pbmc_6000, <span class="at">sign_name =</span> <span class="st">&quot;path:hsa04658_V&quot;</span>,</span>
<span id="cb7-63"><a href="appendix-f-creating-figures.html#cb7-63" aria-hidden="true" tabindex="-1"></a>  <span class="at">data_type_for_label =</span> <span class="st">&quot;KEGG&quot;</span>, <span class="at">category_for_label =</span> <span class="st">&quot;pathway&quot;</span>,</span>
<span id="cb7-64"><a href="appendix-f-creating-figures.html#cb7-64" aria-hidden="true" tabindex="-1"></a>  <span class="at">algo_name_for_label =</span> <span class="st">&quot;seuratFindClusters&quot;</span>,</span>
<span id="cb7-65"><a href="appendix-f-creating-figures.html#cb7-65" aria-hidden="true" tabindex="-1"></a>  <span class="at">data_type_for_expr =</span> <span class="st">&quot;KEGG&quot;</span>, <span class="at">category_for_expr =</span> <span class="st">&quot;pathway&quot;</span>,</span>
<span id="cb7-66"><a href="appendix-f-creating-figures.html#cb7-66" aria-hidden="true" tabindex="-1"></a>  <span class="at">title =</span> <span class="st">&quot;path:hsa04658_V</span><span class="sc">\n</span><span class="st">Th1 and Th2 cell diff..</span><span class="sc">\n</span><span class="st">FGS = {CD247, CD3E, ...}&quot;</span>,</span>
<span id="cb7-67"><a href="appendix-f-creating-figures.html#cb7-67" aria-hidden="true" tabindex="-1"></a>  <span class="at">title_size =</span> <span class="dv">16</span>, <span class="at">default_color =</span> <span class="cn">FALSE</span>)</span>
<span id="cb7-68"><a href="appendix-f-creating-figures.html#cb7-68" aria-hidden="true" tabindex="-1"></a>filename <span class="ot">&lt;-</span> <span class="st">&quot;figures/figure_20_0045.png&quot;</span></span>
<span id="cb7-69"><a href="appendix-f-creating-figures.html#cb7-69" aria-hidden="true" tabindex="-1"></a><span class="fu">ggsave</span>(<span class="at">file =</span> filename, <span class="at">plot =</span> p, <span class="at">dpi =</span> <span class="dv">300</span>, <span class="at">width =</span> <span class="dv">4</span>, <span class="at">height =</span> <span class="fl">3.5</span>)</span>
<span id="cb7-70"><a href="appendix-f-creating-figures.html#cb7-70" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb7-71"><a href="appendix-f-creating-figures.html#cb7-71" aria-hidden="true" tabindex="-1"></a>p <span class="ot">&lt;-</span> <span class="fu">plot_violin_signScore</span>(</span>
<span id="cb7-72"><a href="appendix-f-creating-figures.html#cb7-72" aria-hidden="true" tabindex="-1"></a>  <span class="at">obj =</span> pbmc_6000, <span class="at">sign_name =</span> <span class="st">&quot;path:hsa04662_S&quot;</span>,</span>
<span id="cb7-73"><a href="appendix-f-creating-figures.html#cb7-73" aria-hidden="true" tabindex="-1"></a>  <span class="at">data_type_for_label =</span> <span class="st">&quot;KEGG&quot;</span>, <span class="at">category_for_label =</span> <span class="st">&quot;pathway&quot;</span>,</span>
<span id="cb7-74"><a href="appendix-f-creating-figures.html#cb7-74" aria-hidden="true" tabindex="-1"></a>  <span class="at">algo_name_for_label =</span> <span class="st">&quot;seuratFindClusters&quot;</span>,</span>
<span id="cb7-75"><a href="appendix-f-creating-figures.html#cb7-75" aria-hidden="true" tabindex="-1"></a>  <span class="at">data_type_for_expr =</span> <span class="st">&quot;KEGG&quot;</span>, <span class="at">category_for_expr =</span> <span class="st">&quot;pathway&quot;</span>,</span>
<span id="cb7-76"><a href="appendix-f-creating-figures.html#cb7-76" aria-hidden="true" tabindex="-1"></a>  <span class="at">title =</span> <span class="st">&quot;path:hsa04662_S</span><span class="sc">\n</span><span class="st">B cell receptor..</span><span class="sc">\n</span><span class="st">FGS = {CD19, CD72, ...}&quot;</span>,</span>
<span id="cb7-77"><a href="appendix-f-creating-figures.html#cb7-77" aria-hidden="true" tabindex="-1"></a>  <span class="at">title_size =</span> <span class="dv">16</span>, <span class="at">default_color =</span> <span class="cn">FALSE</span>)</span>
<span id="cb7-78"><a href="appendix-f-creating-figures.html#cb7-78" aria-hidden="true" tabindex="-1"></a>filename <span class="ot">&lt;-</span> <span class="st">&quot;figures/figure_20_0046.png&quot;</span></span>
<span id="cb7-79"><a href="appendix-f-creating-figures.html#cb7-79" aria-hidden="true" tabindex="-1"></a><span class="fu">ggsave</span>(<span class="at">file =</span> filename, <span class="at">plot =</span> p, <span class="at">dpi =</span> <span class="dv">300</span>, <span class="at">width =</span> <span class="dv">4</span>, <span class="at">height =</span> <span class="fl">3.5</span>)</span></code></pre></div>
<p><img src="figures/figure_20_0040.png" width="250px">
<img src="figures/figure_20_0041.png" width="250px">
<img src="figures/figure_20_0042.png" width="250px">
<img src="figures/figure_20_0043.png" width="250px">
<img src="figures/figure_20_0044.png" width="250px">
<img src="figures/figure_20_0045.png" width="250px">
<img src="figures/figure_20_0046.png" width="250px"></p>
</div>
<div id="infer-population-ratios" class="section level3" number="18.1.2">
<h3><span class="header-section-number">18.1.2</span> Infer population ratios</h3>
<p>Load the data.</p>
<div class="sourceCode" id="cb8"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb8-1"><a href="appendix-f-creating-figures.html#cb8-1" aria-hidden="true" tabindex="-1"></a><span class="fu">rm</span>(<span class="at">list =</span> <span class="fu">ls</span>())</span>
<span id="cb8-2"><a href="appendix-f-creating-figures.html#cb8-2" aria-hidden="true" tabindex="-1"></a><span class="co"># ----------------------------------------</span></span>
<span id="cb8-3"><a href="appendix-f-creating-figures.html#cb8-3" aria-hidden="true" tabindex="-1"></a><span class="co"># pbmc_4000</span></span>
<span id="cb8-4"><a href="appendix-f-creating-figures.html#cb8-4" aria-hidden="true" tabindex="-1"></a><span class="co"># ----------------------------------------</span></span>
<span id="cb8-5"><a href="appendix-f-creating-figures.html#cb8-5" aria-hidden="true" tabindex="-1"></a>pbmc_4000_scran <span class="ot">&lt;-</span> <span class="fu">readRDS</span>(<span class="at">file =</span> <span class="st">&quot;backup/10_501_pbmc_4000_scran.rds&quot;</span>)</span>
<span id="cb8-6"><a href="appendix-f-creating-figures.html#cb8-6" aria-hidden="true" tabindex="-1"></a>pbmc_4000_seurat <span class="ot">&lt;-</span> <span class="fu">readRDS</span>(<span class="at">file =</span> <span class="st">&quot;backup/10_601_pbmc_4000_seurat.rds&quot;</span>)</span>
<span id="cb8-7"><a href="appendix-f-creating-figures.html#cb8-7" aria-hidden="true" tabindex="-1"></a>pbmc_4000_sccatch <span class="ot">&lt;-</span> <span class="fu">readRDS</span>(<span class="at">file =</span><span class="st">&quot;backup/10_602_pbmc_4000_seurat_sccatch.rds&quot;</span>)</span>
<span id="cb8-8"><a href="appendix-f-creating-figures.html#cb8-8" aria-hidden="true" tabindex="-1"></a>pbmc_4000_monocle <span class="ot">&lt;-</span> <span class="fu">readRDS</span>(<span class="at">file =</span> <span class="st">&quot;backup/10_701_pbmc_4000_monocle3.rds&quot;</span>)</span>
<span id="cb8-9"><a href="appendix-f-creating-figures.html#cb8-9" aria-hidden="true" tabindex="-1"></a>pbmc_4000_sc3 <span class="ot">&lt;-</span> <span class="fu">readRDS</span>(<span class="at">file =</span> <span class="st">&quot;backup/10_801_pbmc_4000_sc3.rds&quot;</span>)</span>
<span id="cb8-10"><a href="appendix-f-creating-figures.html#cb8-10" aria-hidden="true" tabindex="-1"></a>pbmc_4000_asurat <span class="ot">&lt;-</span> <span class="fu">readRDS</span>(<span class="at">file =</span> <span class="st">&quot;backup/10_901_pbmc_4000_asurat.rds&quot;</span>)</span>
<span id="cb8-11"><a href="appendix-f-creating-figures.html#cb8-11" aria-hidden="true" tabindex="-1"></a><span class="co"># ----------------------------------------</span></span>
<span id="cb8-12"><a href="appendix-f-creating-figures.html#cb8-12" aria-hidden="true" tabindex="-1"></a><span class="co"># pbmc_6000</span></span>
<span id="cb8-13"><a href="appendix-f-creating-figures.html#cb8-13" aria-hidden="true" tabindex="-1"></a><span class="co"># ----------------------------------------</span></span>
<span id="cb8-14"><a href="appendix-f-creating-figures.html#cb8-14" aria-hidden="true" tabindex="-1"></a>pbmc_6000_scran <span class="ot">&lt;-</span> <span class="fu">readRDS</span>(<span class="at">file =</span> <span class="st">&quot;backup/11_501_pbmc_6000_scran.rds&quot;</span>)</span>
<span id="cb8-15"><a href="appendix-f-creating-figures.html#cb8-15" aria-hidden="true" tabindex="-1"></a>pbmc_6000_seurat <span class="ot">&lt;-</span> <span class="fu">readRDS</span>(<span class="at">file =</span> <span class="st">&quot;backup/11_601_pbmc_6000_seurat.rds&quot;</span>)</span>
<span id="cb8-16"><a href="appendix-f-creating-figures.html#cb8-16" aria-hidden="true" tabindex="-1"></a>pbmc_6000_sccatch <span class="ot">&lt;-</span> <span class="fu">readRDS</span>(<span class="at">file =</span><span class="st">&quot;backup/11_602_pbmc_6000_seurat_sccatch.rds&quot;</span>)</span>
<span id="cb8-17"><a href="appendix-f-creating-figures.html#cb8-17" aria-hidden="true" tabindex="-1"></a>pbmc_6000_monocle <span class="ot">&lt;-</span> <span class="fu">readRDS</span>(<span class="at">file =</span> <span class="st">&quot;backup/11_701_pbmc_6000_monocle3.rds&quot;</span>)</span>
<span id="cb8-18"><a href="appendix-f-creating-figures.html#cb8-18" aria-hidden="true" tabindex="-1"></a>pbmc_6000_sc3 <span class="ot">&lt;-</span> <span class="fu">readRDS</span>(<span class="at">file =</span> <span class="st">&quot;backup/11_801_pbmc_6000_sc3.rds&quot;</span>)</span>
<span id="cb8-19"><a href="appendix-f-creating-figures.html#cb8-19" aria-hidden="true" tabindex="-1"></a>pbmc_6000_asurat <span class="ot">&lt;-</span> <span class="fu">readRDS</span>(<span class="at">file =</span> <span class="st">&quot;backup/11_901_pbmc_6000_asurat.rds&quot;</span>)</span></code></pre></div>
<p>The hypothetical results can be obtained from Cao et al., Front. Genet. (2020).</p>
<div class="sourceCode" id="cb9"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb9-1"><a href="appendix-f-creating-figures.html#cb9-1" aria-hidden="true" tabindex="-1"></a>pbmc_4000_scsa <span class="ot">&lt;-</span> <span class="fu">data.frame</span>(</span>
<span id="cb9-2"><a href="appendix-f-creating-figures.html#cb9-2" aria-hidden="true" tabindex="-1"></a>  <span class="at">cell_type =</span> <span class="fu">c</span>(<span class="st">&quot;T cell&quot;</span>, <span class="st">&quot;Monocyte&quot;</span>, <span class="st">&quot;B cell&quot;</span>, <span class="st">&quot;NK or NKT cell&quot;</span>,</span>
<span id="cb9-3"><a href="appendix-f-creating-figures.html#cb9-3" aria-hidden="true" tabindex="-1"></a>                <span class="st">&quot;Megakaryocyte&quot;</span>),</span>
<span id="cb9-4"><a href="appendix-f-creating-figures.html#cb9-4" aria-hidden="true" tabindex="-1"></a>  <span class="at">n =</span> <span class="fu">c</span>(<span class="fl">0.4226</span>, <span class="fl">0.2707</span>, <span class="fl">0.1509</span>, <span class="fl">0.1546</span>, <span class="fl">0.0012</span>)</span>
<span id="cb9-5"><a href="appendix-f-creating-figures.html#cb9-5" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb9-6"><a href="appendix-f-creating-figures.html#cb9-6" aria-hidden="true" tabindex="-1"></a>pbmc_6000_scsa <span class="ot">&lt;-</span> <span class="fu">data.frame</span>(</span>
<span id="cb9-7"><a href="appendix-f-creating-figures.html#cb9-7" aria-hidden="true" tabindex="-1"></a>  <span class="at">cell_type =</span> <span class="fu">c</span>(<span class="st">&quot;T cell&quot;</span>, <span class="st">&quot;Monocyte&quot;</span>, <span class="st">&quot;B cell&quot;</span>, <span class="st">&quot;NK or NKT cell&quot;</span>,</span>
<span id="cb9-8"><a href="appendix-f-creating-figures.html#cb9-8" aria-hidden="true" tabindex="-1"></a>                <span class="st">&quot;Megakaryocyte&quot;</span>),</span>
<span id="cb9-9"><a href="appendix-f-creating-figures.html#cb9-9" aria-hidden="true" tabindex="-1"></a>  <span class="at">n =</span> <span class="fu">c</span>(<span class="fl">0.4920</span>, <span class="fl">0.2652</span>, <span class="fl">0.1292</span>, <span class="fl">0.1102</span>, <span class="fl">0.0035</span>)</span>
<span id="cb9-10"><a href="appendix-f-creating-figures.html#cb9-10" aria-hidden="true" tabindex="-1"></a>)</span></code></pre></div>
<p>Change the numbers of cells into ratios.</p>
<div class="sourceCode" id="cb10"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb10-1"><a href="appendix-f-creating-figures.html#cb10-1" aria-hidden="true" tabindex="-1"></a>take_ratio <span class="ot">&lt;-</span> <span class="cf">function</span>(df){</span>
<span id="cb10-2"><a href="appendix-f-creating-figures.html#cb10-2" aria-hidden="true" tabindex="-1"></a>  df<span class="sc">$</span>n <span class="ot">&lt;-</span> <span class="fu">as.numeric</span>(df<span class="sc">$</span>n) <span class="sc">/</span> <span class="fu">sum</span>(<span class="fu">as.numeric</span>(df<span class="sc">$</span>n))</span>
<span id="cb10-3"><a href="appendix-f-creating-figures.html#cb10-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">return</span>(df)</span>
<span id="cb10-4"><a href="appendix-f-creating-figures.html#cb10-4" aria-hidden="true" tabindex="-1"></a>}</span>
<span id="cb10-5"><a href="appendix-f-creating-figures.html#cb10-5" aria-hidden="true" tabindex="-1"></a><span class="co"># ----------------------------------------</span></span>
<span id="cb10-6"><a href="appendix-f-creating-figures.html#cb10-6" aria-hidden="true" tabindex="-1"></a><span class="co"># pbmc_4000</span></span>
<span id="cb10-7"><a href="appendix-f-creating-figures.html#cb10-7" aria-hidden="true" tabindex="-1"></a><span class="co"># ----------------------------------------</span></span>
<span id="cb10-8"><a href="appendix-f-creating-figures.html#cb10-8" aria-hidden="true" tabindex="-1"></a>pbmc_4000_scran <span class="ot">&lt;-</span> <span class="fu">take_ratio</span>(pbmc_4000_scran[[<span class="st">&quot;population&quot;</span>]])</span>
<span id="cb10-9"><a href="appendix-f-creating-figures.html#cb10-9" aria-hidden="true" tabindex="-1"></a>pbmc_4000_seurat <span class="ot">&lt;-</span> <span class="fu">take_ratio</span>(pbmc_4000_seurat<span class="sc">@</span>misc[[<span class="st">&quot;population&quot;</span>]])</span>
<span id="cb10-10"><a href="appendix-f-creating-figures.html#cb10-10" aria-hidden="true" tabindex="-1"></a>pbmc_4000_sccatch <span class="ot">&lt;-</span> <span class="fu">take_ratio</span>(pbmc_4000_sccatch<span class="sc">@</span>misc[[<span class="st">&quot;population&quot;</span>]])</span>
<span id="cb10-11"><a href="appendix-f-creating-figures.html#cb10-11" aria-hidden="true" tabindex="-1"></a>pbmc_4000_monocle <span class="ot">&lt;-</span> <span class="fu">take_ratio</span>(pbmc_4000_monocle[[<span class="st">&quot;population&quot;</span>]])</span>
<span id="cb10-12"><a href="appendix-f-creating-figures.html#cb10-12" aria-hidden="true" tabindex="-1"></a>pbmc_4000_sc3 <span class="ot">&lt;-</span> <span class="fu">take_ratio</span>(pbmc_4000_sc3[[<span class="st">&quot;population&quot;</span>]])</span>
<span id="cb10-13"><a href="appendix-f-creating-figures.html#cb10-13" aria-hidden="true" tabindex="-1"></a>pbmc_4000_asurat <span class="ot">&lt;-</span> <span class="fu">take_ratio</span>(pbmc_4000_asurat[[<span class="st">&quot;population&quot;</span>]])</span>
<span id="cb10-14"><a href="appendix-f-creating-figures.html#cb10-14" aria-hidden="true" tabindex="-1"></a><span class="co"># ----------------------------------------</span></span>
<span id="cb10-15"><a href="appendix-f-creating-figures.html#cb10-15" aria-hidden="true" tabindex="-1"></a><span class="co"># pbmc_6000</span></span>
<span id="cb10-16"><a href="appendix-f-creating-figures.html#cb10-16" aria-hidden="true" tabindex="-1"></a><span class="co"># ----------------------------------------</span></span>
<span id="cb10-17"><a href="appendix-f-creating-figures.html#cb10-17" aria-hidden="true" tabindex="-1"></a>pbmc_6000_scran <span class="ot">&lt;-</span> <span class="fu">take_ratio</span>(pbmc_6000_scran[[<span class="st">&quot;population&quot;</span>]])</span>
<span id="cb10-18"><a href="appendix-f-creating-figures.html#cb10-18" aria-hidden="true" tabindex="-1"></a>pbmc_6000_seurat <span class="ot">&lt;-</span> <span class="fu">take_ratio</span>(pbmc_6000_seurat<span class="sc">@</span>misc[[<span class="st">&quot;population&quot;</span>]])</span>
<span id="cb10-19"><a href="appendix-f-creating-figures.html#cb10-19" aria-hidden="true" tabindex="-1"></a>pbmc_6000_sccatch <span class="ot">&lt;-</span> <span class="fu">take_ratio</span>(pbmc_6000_sccatch<span class="sc">@</span>misc[[<span class="st">&quot;population&quot;</span>]])</span>
<span id="cb10-20"><a href="appendix-f-creating-figures.html#cb10-20" aria-hidden="true" tabindex="-1"></a>pbmc_6000_monocle <span class="ot">&lt;-</span> <span class="fu">take_ratio</span>(pbmc_6000_monocle[[<span class="st">&quot;population&quot;</span>]])</span>
<span id="cb10-21"><a href="appendix-f-creating-figures.html#cb10-21" aria-hidden="true" tabindex="-1"></a>pbmc_6000_sc3 <span class="ot">&lt;-</span> <span class="fu">take_ratio</span>(pbmc_6000_sc3[[<span class="st">&quot;population&quot;</span>]])</span>
<span id="cb10-22"><a href="appendix-f-creating-figures.html#cb10-22" aria-hidden="true" tabindex="-1"></a>pbmc_6000_asurat <span class="ot">&lt;-</span> <span class="fu">take_ratio</span>(pbmc_6000_asurat[[<span class="st">&quot;population&quot;</span>]])</span></code></pre></div>
<p>Add the names of algorithms.</p>
<div class="sourceCode" id="cb11"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb11-1"><a href="appendix-f-creating-figures.html#cb11-1" aria-hidden="true" tabindex="-1"></a><span class="co"># ----------------------------------------</span></span>
<span id="cb11-2"><a href="appendix-f-creating-figures.html#cb11-2" aria-hidden="true" tabindex="-1"></a><span class="co"># pbmc_4000</span></span>
<span id="cb11-3"><a href="appendix-f-creating-figures.html#cb11-3" aria-hidden="true" tabindex="-1"></a><span class="co"># ----------------------------------------</span></span>
<span id="cb11-4"><a href="appendix-f-creating-figures.html#cb11-4" aria-hidden="true" tabindex="-1"></a>pbmc_4000_scsa<span class="sc">$</span>algorithm <span class="ot">&lt;-</span> <span class="st">&quot;SCSA&quot;</span></span>
<span id="cb11-5"><a href="appendix-f-creating-figures.html#cb11-5" aria-hidden="true" tabindex="-1"></a>pbmc_4000_scran<span class="sc">$</span>algorithm <span class="ot">&lt;-</span> <span class="st">&quot;scran&quot;</span></span>
<span id="cb11-6"><a href="appendix-f-creating-figures.html#cb11-6" aria-hidden="true" tabindex="-1"></a>pbmc_4000_seurat<span class="sc">$</span>algorithm <span class="ot">&lt;-</span> <span class="st">&quot;Seurat&quot;</span></span>
<span id="cb11-7"><a href="appendix-f-creating-figures.html#cb11-7" aria-hidden="true" tabindex="-1"></a>pbmc_4000_sccatch<span class="sc">$</span>algorithm <span class="ot">&lt;-</span> <span class="st">&quot;Seurat with scCATCH&quot;</span></span>
<span id="cb11-8"><a href="appendix-f-creating-figures.html#cb11-8" aria-hidden="true" tabindex="-1"></a>pbmc_4000_monocle<span class="sc">$</span>algorithm <span class="ot">&lt;-</span> <span class="st">&quot;Monocle 3&quot;</span></span>
<span id="cb11-9"><a href="appendix-f-creating-figures.html#cb11-9" aria-hidden="true" tabindex="-1"></a>pbmc_4000_sc3<span class="sc">$</span>algorithm <span class="ot">&lt;-</span> <span class="st">&quot;SC3&quot;</span></span>
<span id="cb11-10"><a href="appendix-f-creating-figures.html#cb11-10" aria-hidden="true" tabindex="-1"></a>pbmc_4000_asurat<span class="sc">$</span>algorithm <span class="ot">&lt;-</span> <span class="st">&quot;ASURAT&quot;</span></span>
<span id="cb11-11"><a href="appendix-f-creating-figures.html#cb11-11" aria-hidden="true" tabindex="-1"></a><span class="co"># ----------------------------------------</span></span>
<span id="cb11-12"><a href="appendix-f-creating-figures.html#cb11-12" aria-hidden="true" tabindex="-1"></a><span class="co"># pbmc_6000</span></span>
<span id="cb11-13"><a href="appendix-f-creating-figures.html#cb11-13" aria-hidden="true" tabindex="-1"></a><span class="co"># ----------------------------------------</span></span>
<span id="cb11-14"><a href="appendix-f-creating-figures.html#cb11-14" aria-hidden="true" tabindex="-1"></a>pbmc_6000_scsa<span class="sc">$</span>algorithm <span class="ot">&lt;-</span> <span class="st">&quot;SCSA&quot;</span></span>
<span id="cb11-15"><a href="appendix-f-creating-figures.html#cb11-15" aria-hidden="true" tabindex="-1"></a>pbmc_6000_scran<span class="sc">$</span>algorithm <span class="ot">&lt;-</span> <span class="st">&quot;scran&quot;</span></span>
<span id="cb11-16"><a href="appendix-f-creating-figures.html#cb11-16" aria-hidden="true" tabindex="-1"></a>pbmc_6000_seurat<span class="sc">$</span>algorithm <span class="ot">&lt;-</span> <span class="st">&quot;Seurat&quot;</span></span>
<span id="cb11-17"><a href="appendix-f-creating-figures.html#cb11-17" aria-hidden="true" tabindex="-1"></a>pbmc_6000_sccatch<span class="sc">$</span>algorithm <span class="ot">&lt;-</span> <span class="st">&quot;Seurat with scCATCH&quot;</span></span>
<span id="cb11-18"><a href="appendix-f-creating-figures.html#cb11-18" aria-hidden="true" tabindex="-1"></a>pbmc_6000_monocle<span class="sc">$</span>algorithm <span class="ot">&lt;-</span> <span class="st">&quot;Monocle 3&quot;</span></span>
<span id="cb11-19"><a href="appendix-f-creating-figures.html#cb11-19" aria-hidden="true" tabindex="-1"></a>pbmc_6000_sc3<span class="sc">$</span>algorithm <span class="ot">&lt;-</span> <span class="st">&quot;SC3&quot;</span></span>
<span id="cb11-20"><a href="appendix-f-creating-figures.html#cb11-20" aria-hidden="true" tabindex="-1"></a>pbmc_6000_asurat<span class="sc">$</span>algorithm <span class="ot">&lt;-</span> <span class="st">&quot;ASURAT&quot;</span></span></code></pre></div>
<p>Concatenate all of the data.</p>
<div class="sourceCode" id="cb12"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb12-1"><a href="appendix-f-creating-figures.html#cb12-1" aria-hidden="true" tabindex="-1"></a><span class="co"># ----------------------------------------</span></span>
<span id="cb12-2"><a href="appendix-f-creating-figures.html#cb12-2" aria-hidden="true" tabindex="-1"></a><span class="co"># pbmc_4000</span></span>
<span id="cb12-3"><a href="appendix-f-creating-figures.html#cb12-3" aria-hidden="true" tabindex="-1"></a><span class="co"># ----------------------------------------</span></span>
<span id="cb12-4"><a href="appendix-f-creating-figures.html#cb12-4" aria-hidden="true" tabindex="-1"></a>pbmc_4000_df <span class="ot">&lt;-</span> <span class="fu">rbind</span>(pbmc_4000_scsa, pbmc_4000_scran, pbmc_4000_seurat,</span>
<span id="cb12-5"><a href="appendix-f-creating-figures.html#cb12-5" aria-hidden="true" tabindex="-1"></a>                      pbmc_4000_sccatch, pbmc_4000_monocle, pbmc_4000_sc3,</span>
<span id="cb12-6"><a href="appendix-f-creating-figures.html#cb12-6" aria-hidden="true" tabindex="-1"></a>                      pbmc_4000_asurat)</span>
<span id="cb12-7"><a href="appendix-f-creating-figures.html#cb12-7" aria-hidden="true" tabindex="-1"></a><span class="co"># ----------------------------------------</span></span>
<span id="cb12-8"><a href="appendix-f-creating-figures.html#cb12-8" aria-hidden="true" tabindex="-1"></a><span class="co"># pbmc_6000</span></span>
<span id="cb12-9"><a href="appendix-f-creating-figures.html#cb12-9" aria-hidden="true" tabindex="-1"></a><span class="co"># ----------------------------------------</span></span>
<span id="cb12-10"><a href="appendix-f-creating-figures.html#cb12-10" aria-hidden="true" tabindex="-1"></a>pbmc_6000_df <span class="ot">&lt;-</span> <span class="fu">rbind</span>(pbmc_6000_scsa, pbmc_6000_scran, pbmc_6000_seurat,</span>
<span id="cb12-11"><a href="appendix-f-creating-figures.html#cb12-11" aria-hidden="true" tabindex="-1"></a>                      pbmc_6000_sccatch, pbmc_6000_monocle, pbmc_6000_sc3,</span>
<span id="cb12-12"><a href="appendix-f-creating-figures.html#cb12-12" aria-hidden="true" tabindex="-1"></a>                      pbmc_6000_asurat)</span></code></pre></div>
<p>Show the population sizes.</p>
<div class="sourceCode" id="cb13"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb13-1"><a href="appendix-f-creating-figures.html#cb13-1" aria-hidden="true" tabindex="-1"></a>ymax <span class="ot">&lt;-</span> <span class="fl">0.65</span></span>
<span id="cb13-2"><a href="appendix-f-creating-figures.html#cb13-2" aria-hidden="true" tabindex="-1"></a>xlabel <span class="ot">&lt;-</span> <span class="st">&quot;&quot;</span></span>
<span id="cb13-3"><a href="appendix-f-creating-figures.html#cb13-3" aria-hidden="true" tabindex="-1"></a>ylabel <span class="ot">&lt;-</span> <span class="st">&quot;Ratio&quot;</span></span>
<span id="cb13-4"><a href="appendix-f-creating-figures.html#cb13-4" aria-hidden="true" tabindex="-1"></a>my_colors <span class="ot">&lt;-</span> <span class="fu">brewer.pal</span>(<span class="dv">6</span>, <span class="at">name =</span> <span class="st">&quot;Set1&quot;</span>)</span>
<span id="cb13-5"><a href="appendix-f-creating-figures.html#cb13-5" aria-hidden="true" tabindex="-1"></a>size <span class="ot">&lt;-</span> <span class="fl">4.5</span></span>
<span id="cb13-6"><a href="appendix-f-creating-figures.html#cb13-6" aria-hidden="true" tabindex="-1"></a>alpha <span class="ot">&lt;-</span> <span class="fl">0.85</span></span>
<span id="cb13-7"><a href="appendix-f-creating-figures.html#cb13-7" aria-hidden="true" tabindex="-1"></a>jitter_size <span class="ot">&lt;-</span> <span class="fl">0.2</span></span>
<span id="cb13-8"><a href="appendix-f-creating-figures.html#cb13-8" aria-hidden="true" tabindex="-1"></a><span class="fu">set.seed</span>(<span class="dv">16</span>)</span>
<span id="cb13-9"><a href="appendix-f-creating-figures.html#cb13-9" aria-hidden="true" tabindex="-1"></a>p <span class="ot">&lt;-</span> <span class="fu">ggplot</span>() <span class="sc">+</span></span>
<span id="cb13-10"><a href="appendix-f-creating-figures.html#cb13-10" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_boxplot</span>(<span class="fu">aes</span>(<span class="at">x =</span> pbmc_4000_df<span class="sc">$</span>cell_type, <span class="at">y =</span> pbmc_4000_df<span class="sc">$</span>n),</span>
<span id="cb13-11"><a href="appendix-f-creating-figures.html#cb13-11" aria-hidden="true" tabindex="-1"></a>               <span class="at">color =</span> <span class="st">&quot;white&quot;</span>, <span class="at">alpha =</span> <span class="dv">0</span>, <span class="at">size =</span> <span class="dv">0</span>, <span class="at">coef =</span> <span class="fl">1e10</span>) <span class="sc">+</span></span>
<span id="cb13-12"><a href="appendix-f-creating-figures.html#cb13-12" aria-hidden="true" tabindex="-1"></a>  <span class="co"># ----------------------------------------</span></span>
<span id="cb13-13"><a href="appendix-f-creating-figures.html#cb13-13" aria-hidden="true" tabindex="-1"></a>  <span class="co"># Cao et al., 2020.</span></span>
<span id="cb13-14"><a href="appendix-f-creating-figures.html#cb13-14" aria-hidden="true" tabindex="-1"></a>  <span class="co"># ----------------------------------------</span></span>
<span id="cb13-15"><a href="appendix-f-creating-figures.html#cb13-15" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_jitter</span>(<span class="fu">aes</span>(</span>
<span id="cb13-16"><a href="appendix-f-creating-figures.html#cb13-16" aria-hidden="true" tabindex="-1"></a>    <span class="at">x =</span> pbmc_4000_df[<span class="fu">which</span>(pbmc_4000_df<span class="sc">$</span>algorithm <span class="sc">==</span> <span class="st">&quot;SCSA&quot;</span>),]<span class="sc">$</span>cell_type,</span>
<span id="cb13-17"><a href="appendix-f-creating-figures.html#cb13-17" aria-hidden="true" tabindex="-1"></a>    <span class="at">y =</span> pbmc_4000_df[<span class="fu">which</span>(pbmc_4000_df<span class="sc">$</span>algorithm <span class="sc">==</span> <span class="st">&quot;SCSA&quot;</span>),]<span class="sc">$</span>n),</span>
<span id="cb13-18"><a href="appendix-f-creating-figures.html#cb13-18" aria-hidden="true" tabindex="-1"></a>    <span class="at">color =</span> <span class="st">&quot;black&quot;</span>, <span class="at">size =</span> size, <span class="at">shape =</span> <span class="dv">16</span>, <span class="at">alpha =</span> <span class="dv">1</span>,</span>
<span id="cb13-19"><a href="appendix-f-creating-figures.html#cb13-19" aria-hidden="true" tabindex="-1"></a>    <span class="at">position =</span> <span class="fu">position_jitter</span>(<span class="at">w =</span> jitter_size, <span class="at">h =</span> <span class="dv">0</span>)) <span class="sc">+</span></span>
<span id="cb13-20"><a href="appendix-f-creating-figures.html#cb13-20" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_jitter</span>(<span class="fu">aes</span>(</span>
<span id="cb13-21"><a href="appendix-f-creating-figures.html#cb13-21" aria-hidden="true" tabindex="-1"></a>    <span class="at">x =</span> pbmc_6000_df[<span class="fu">which</span>(pbmc_6000_df<span class="sc">$</span>algorithm <span class="sc">==</span> <span class="st">&quot;SCSA&quot;</span>),]<span class="sc">$</span>cell_type,</span>
<span id="cb13-22"><a href="appendix-f-creating-figures.html#cb13-22" aria-hidden="true" tabindex="-1"></a>    <span class="at">y =</span> pbmc_6000_df[<span class="fu">which</span>(pbmc_6000_df<span class="sc">$</span>algorithm <span class="sc">==</span> <span class="st">&quot;SCSA&quot;</span>),]<span class="sc">$</span>n),</span>
<span id="cb13-23"><a href="appendix-f-creating-figures.html#cb13-23" aria-hidden="true" tabindex="-1"></a>    <span class="at">color =</span> <span class="st">&quot;black&quot;</span>, <span class="at">size =</span> size, <span class="at">shape =</span> <span class="dv">16</span>, <span class="at">alpha =</span> <span class="dv">1</span>,</span>
<span id="cb13-24"><a href="appendix-f-creating-figures.html#cb13-24" aria-hidden="true" tabindex="-1"></a>    <span class="at">position =</span> <span class="fu">position_jitter</span>(<span class="at">w =</span> jitter_size, <span class="at">h =</span> <span class="dv">0</span>)) <span class="sc">+</span></span>
<span id="cb13-25"><a href="appendix-f-creating-figures.html#cb13-25" aria-hidden="true" tabindex="-1"></a>  <span class="co"># ----------------------------------------</span></span>
<span id="cb13-26"><a href="appendix-f-creating-figures.html#cb13-26" aria-hidden="true" tabindex="-1"></a>  <span class="co"># pbmc_4000</span></span>
<span id="cb13-27"><a href="appendix-f-creating-figures.html#cb13-27" aria-hidden="true" tabindex="-1"></a>  <span class="co"># ----------------------------------------</span></span>
<span id="cb13-28"><a href="appendix-f-creating-figures.html#cb13-28" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_jitter</span>(<span class="fu">aes</span>(</span>
<span id="cb13-29"><a href="appendix-f-creating-figures.html#cb13-29" aria-hidden="true" tabindex="-1"></a>    <span class="at">x =</span> pbmc_4000_df[<span class="fu">which</span>(pbmc_4000_df<span class="sc">$</span>algorithm <span class="sc">==</span> <span class="st">&quot;scran&quot;</span>),]<span class="sc">$</span>cell_type,</span>
<span id="cb13-30"><a href="appendix-f-creating-figures.html#cb13-30" aria-hidden="true" tabindex="-1"></a>    <span class="at">y =</span> pbmc_4000_df[<span class="fu">which</span>(pbmc_4000_df<span class="sc">$</span>algorithm <span class="sc">==</span> <span class="st">&quot;scran&quot;</span>),]<span class="sc">$</span>n),</span>
<span id="cb13-31"><a href="appendix-f-creating-figures.html#cb13-31" aria-hidden="true" tabindex="-1"></a>    <span class="at">color =</span> my_colors[<span class="dv">1</span>], <span class="at">size =</span> size, <span class="at">shape =</span> <span class="dv">16</span>, <span class="at">alpha =</span> alpha,</span>
<span id="cb13-32"><a href="appendix-f-creating-figures.html#cb13-32" aria-hidden="true" tabindex="-1"></a>    <span class="at">position =</span> <span class="fu">position_jitter</span>(<span class="at">w =</span> jitter_size, <span class="at">h =</span> <span class="dv">0</span>)) <span class="sc">+</span></span>
<span id="cb13-33"><a href="appendix-f-creating-figures.html#cb13-33" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_jitter</span>(<span class="fu">aes</span>(</span>
<span id="cb13-34"><a href="appendix-f-creating-figures.html#cb13-34" aria-hidden="true" tabindex="-1"></a>    <span class="at">x =</span> pbmc_4000_df[<span class="fu">which</span>(pbmc_4000_df<span class="sc">$</span>algorithm <span class="sc">==</span> <span class="st">&quot;Seurat&quot;</span>),]<span class="sc">$</span>cell_type,</span>
<span id="cb13-35"><a href="appendix-f-creating-figures.html#cb13-35" aria-hidden="true" tabindex="-1"></a>    <span class="at">y =</span> pbmc_4000_df[<span class="fu">which</span>(pbmc_4000_df<span class="sc">$</span>algorithm <span class="sc">==</span> <span class="st">&quot;Seurat&quot;</span>),]<span class="sc">$</span>n),</span>
<span id="cb13-36"><a href="appendix-f-creating-figures.html#cb13-36" aria-hidden="true" tabindex="-1"></a>    <span class="at">color =</span> my_colors[<span class="dv">2</span>], <span class="at">size =</span> size, <span class="at">shape =</span> <span class="dv">16</span>, <span class="at">alpha =</span> alpha,</span>
<span id="cb13-37"><a href="appendix-f-creating-figures.html#cb13-37" aria-hidden="true" tabindex="-1"></a>    <span class="at">position =</span> <span class="fu">position_jitter</span>(<span class="at">w =</span> jitter_size, <span class="at">h =</span> <span class="dv">0</span>)) <span class="sc">+</span></span>
<span id="cb13-38"><a href="appendix-f-creating-figures.html#cb13-38" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_jitter</span>(<span class="fu">aes</span>(</span>
<span id="cb13-39"><a href="appendix-f-creating-figures.html#cb13-39" aria-hidden="true" tabindex="-1"></a>    <span class="at">x =</span> pbmc_4000_df[<span class="fu">which</span>(pbmc_4000_df<span class="sc">$</span>algorithm <span class="sc">==</span> <span class="st">&quot;Seurat with scCATCH&quot;</span>),]<span class="sc">$</span>cell_type,</span>
<span id="cb13-40"><a href="appendix-f-creating-figures.html#cb13-40" aria-hidden="true" tabindex="-1"></a>    <span class="at">y =</span> pbmc_4000_df[<span class="fu">which</span>(pbmc_4000_df<span class="sc">$</span>algorithm <span class="sc">==</span> <span class="st">&quot;Seurat with scCATCH&quot;</span>),]<span class="sc">$</span>n),</span>
<span id="cb13-41"><a href="appendix-f-creating-figures.html#cb13-41" aria-hidden="true" tabindex="-1"></a>    <span class="at">color =</span> my_colors[<span class="dv">3</span>], <span class="at">size =</span> size, <span class="at">shape =</span> <span class="dv">16</span>, <span class="at">alpha =</span> alpha,</span>
<span id="cb13-42"><a href="appendix-f-creating-figures.html#cb13-42" aria-hidden="true" tabindex="-1"></a>    <span class="at">position =</span> <span class="fu">position_jitter</span>(<span class="at">w =</span> jitter_size, <span class="at">h =</span> <span class="dv">0</span>)) <span class="sc">+</span></span>
<span id="cb13-43"><a href="appendix-f-creating-figures.html#cb13-43" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_jitter</span>(<span class="fu">aes</span>(</span>
<span id="cb13-44"><a href="appendix-f-creating-figures.html#cb13-44" aria-hidden="true" tabindex="-1"></a>    <span class="at">x =</span> pbmc_4000_df[<span class="fu">which</span>(pbmc_4000_df<span class="sc">$</span>algorithm <span class="sc">==</span> <span class="st">&quot;Monocle 3&quot;</span>),]<span class="sc">$</span>cell_type,</span>
<span id="cb13-45"><a href="appendix-f-creating-figures.html#cb13-45" aria-hidden="true" tabindex="-1"></a>    <span class="at">y =</span> pbmc_4000_df[<span class="fu">which</span>(pbmc_4000_df<span class="sc">$</span>algorithm <span class="sc">==</span> <span class="st">&quot;Monocle 3&quot;</span>),]<span class="sc">$</span>n),</span>
<span id="cb13-46"><a href="appendix-f-creating-figures.html#cb13-46" aria-hidden="true" tabindex="-1"></a>    <span class="at">color =</span> my_colors[<span class="dv">4</span>], <span class="at">size =</span> size, <span class="at">shape =</span> <span class="dv">16</span>, <span class="at">alpha =</span> alpha,</span>
<span id="cb13-47"><a href="appendix-f-creating-figures.html#cb13-47" aria-hidden="true" tabindex="-1"></a>    <span class="at">position =</span> <span class="fu">position_jitter</span>(<span class="at">w =</span> jitter_size, <span class="at">h =</span> <span class="dv">0</span>)) <span class="sc">+</span></span>
<span id="cb13-48"><a href="appendix-f-creating-figures.html#cb13-48" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_jitter</span>(<span class="fu">aes</span>(</span>
<span id="cb13-49"><a href="appendix-f-creating-figures.html#cb13-49" aria-hidden="true" tabindex="-1"></a>    <span class="at">x =</span> pbmc_4000_df[<span class="fu">which</span>(pbmc_4000_df<span class="sc">$</span>algorithm <span class="sc">==</span> <span class="st">&quot;SC3&quot;</span>),]<span class="sc">$</span>cell_type,</span>
<span id="cb13-50"><a href="appendix-f-creating-figures.html#cb13-50" aria-hidden="true" tabindex="-1"></a>    <span class="at">y =</span> pbmc_4000_df[<span class="fu">which</span>(pbmc_4000_df<span class="sc">$</span>algorithm <span class="sc">==</span> <span class="st">&quot;SC3&quot;</span>),]<span class="sc">$</span>n),</span>
<span id="cb13-51"><a href="appendix-f-creating-figures.html#cb13-51" aria-hidden="true" tabindex="-1"></a>    <span class="at">color =</span> my_colors[<span class="dv">5</span>], <span class="at">size =</span> size, <span class="at">shape =</span> <span class="dv">16</span>, <span class="at">alpha =</span> alpha,</span>
<span id="cb13-52"><a href="appendix-f-creating-figures.html#cb13-52" aria-hidden="true" tabindex="-1"></a>    <span class="at">position =</span> <span class="fu">position_jitter</span>(<span class="at">w =</span> jitter_size, <span class="at">h =</span> <span class="dv">0</span>)) <span class="sc">+</span></span>
<span id="cb13-53"><a href="appendix-f-creating-figures.html#cb13-53" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_jitter</span>(<span class="fu">aes</span>(</span>
<span id="cb13-54"><a href="appendix-f-creating-figures.html#cb13-54" aria-hidden="true" tabindex="-1"></a>    <span class="at">x =</span> pbmc_4000_df[<span class="fu">which</span>(pbmc_4000_df<span class="sc">$</span>algorithm <span class="sc">==</span> <span class="st">&quot;ASURAT&quot;</span>),]<span class="sc">$</span>cell_type,</span>
<span id="cb13-55"><a href="appendix-f-creating-figures.html#cb13-55" aria-hidden="true" tabindex="-1"></a>    <span class="at">y =</span> pbmc_4000_df[<span class="fu">which</span>(pbmc_4000_df<span class="sc">$</span>algorithm <span class="sc">==</span> <span class="st">&quot;ASURAT&quot;</span>),]<span class="sc">$</span>n),</span>
<span id="cb13-56"><a href="appendix-f-creating-figures.html#cb13-56" aria-hidden="true" tabindex="-1"></a>    <span class="at">color =</span> my_colors[<span class="dv">6</span>], <span class="at">size =</span> size, <span class="at">shape =</span> <span class="dv">16</span>, <span class="at">alpha =</span> alpha,</span>
<span id="cb13-57"><a href="appendix-f-creating-figures.html#cb13-57" aria-hidden="true" tabindex="-1"></a>    <span class="at">position =</span> <span class="fu">position_jitter</span>(<span class="at">w =</span> jitter_size, <span class="at">h =</span> <span class="dv">0</span>)) <span class="sc">+</span></span>
<span id="cb13-58"><a href="appendix-f-creating-figures.html#cb13-58" aria-hidden="true" tabindex="-1"></a>  <span class="co"># ----------------------------------------</span></span>
<span id="cb13-59"><a href="appendix-f-creating-figures.html#cb13-59" aria-hidden="true" tabindex="-1"></a>  <span class="co"># pbmc_6000</span></span>
<span id="cb13-60"><a href="appendix-f-creating-figures.html#cb13-60" aria-hidden="true" tabindex="-1"></a>  <span class="co"># ----------------------------------------</span></span>
<span id="cb13-61"><a href="appendix-f-creating-figures.html#cb13-61" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_jitter</span>(<span class="fu">aes</span>(</span>
<span id="cb13-62"><a href="appendix-f-creating-figures.html#cb13-62" aria-hidden="true" tabindex="-1"></a>    <span class="at">x =</span> pbmc_6000_df[<span class="fu">which</span>(pbmc_6000_df<span class="sc">$</span>algorithm <span class="sc">==</span> <span class="st">&quot;scran&quot;</span>),]<span class="sc">$</span>cell_type,</span>
<span id="cb13-63"><a href="appendix-f-creating-figures.html#cb13-63" aria-hidden="true" tabindex="-1"></a>    <span class="at">y =</span> pbmc_6000_df[<span class="fu">which</span>(pbmc_6000_df<span class="sc">$</span>algorithm <span class="sc">==</span> <span class="st">&quot;scran&quot;</span>),]<span class="sc">$</span>n),</span>
<span id="cb13-64"><a href="appendix-f-creating-figures.html#cb13-64" aria-hidden="true" tabindex="-1"></a>    <span class="at">color =</span> my_colors[<span class="dv">1</span>], <span class="at">size =</span> size, <span class="at">shape =</span> <span class="dv">16</span>, <span class="at">alpha =</span> alpha,</span>
<span id="cb13-65"><a href="appendix-f-creating-figures.html#cb13-65" aria-hidden="true" tabindex="-1"></a>    <span class="at">position =</span> <span class="fu">position_jitter</span>(<span class="at">w =</span> jitter_size, <span class="at">h =</span> <span class="dv">0</span>)) <span class="sc">+</span></span>
<span id="cb13-66"><a href="appendix-f-creating-figures.html#cb13-66" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_jitter</span>(<span class="fu">aes</span>(</span>
<span id="cb13-67"><a href="appendix-f-creating-figures.html#cb13-67" aria-hidden="true" tabindex="-1"></a>    <span class="at">x =</span> pbmc_6000_df[<span class="fu">which</span>(pbmc_6000_df<span class="sc">$</span>algorithm <span class="sc">==</span> <span class="st">&quot;Seurat&quot;</span>),]<span class="sc">$</span>cell_type,</span>
<span id="cb13-68"><a href="appendix-f-creating-figures.html#cb13-68" aria-hidden="true" tabindex="-1"></a>    <span class="at">y =</span> pbmc_6000_df[<span class="fu">which</span>(pbmc_6000_df<span class="sc">$</span>algorithm <span class="sc">==</span> <span class="st">&quot;Seurat&quot;</span>),]<span class="sc">$</span>n),</span>
<span id="cb13-69"><a href="appendix-f-creating-figures.html#cb13-69" aria-hidden="true" tabindex="-1"></a>    <span class="at">color =</span> my_colors[<span class="dv">2</span>], <span class="at">size =</span> size, <span class="at">shape =</span> <span class="dv">16</span>, <span class="at">alpha =</span> alpha,</span>
<span id="cb13-70"><a href="appendix-f-creating-figures.html#cb13-70" aria-hidden="true" tabindex="-1"></a>    <span class="at">position =</span> <span class="fu">position_jitter</span>(<span class="at">w =</span> jitter_size, <span class="at">h =</span> <span class="dv">0</span>)) <span class="sc">+</span></span>
<span id="cb13-71"><a href="appendix-f-creating-figures.html#cb13-71" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_jitter</span>(<span class="fu">aes</span>(</span>
<span id="cb13-72"><a href="appendix-f-creating-figures.html#cb13-72" aria-hidden="true" tabindex="-1"></a>    <span class="at">x =</span> pbmc_6000_df[<span class="fu">which</span>(pbmc_6000_df<span class="sc">$</span>algorithm <span class="sc">==</span> <span class="st">&quot;Seurat with scCATCH&quot;</span>),]<span class="sc">$</span>cell_type,</span>
<span id="cb13-73"><a href="appendix-f-creating-figures.html#cb13-73" aria-hidden="true" tabindex="-1"></a>    <span class="at">y =</span> pbmc_6000_df[<span class="fu">which</span>(pbmc_6000_df<span class="sc">$</span>algorithm <span class="sc">==</span> <span class="st">&quot;Seurat with scCATCH&quot;</span>),]<span class="sc">$</span>n),</span>
<span id="cb13-74"><a href="appendix-f-creating-figures.html#cb13-74" aria-hidden="true" tabindex="-1"></a>    <span class="at">color =</span> my_colors[<span class="dv">3</span>], <span class="at">size =</span> size, <span class="at">shape =</span> <span class="dv">16</span>, <span class="at">alpha =</span> alpha,</span>
<span id="cb13-75"><a href="appendix-f-creating-figures.html#cb13-75" aria-hidden="true" tabindex="-1"></a>    <span class="at">position =</span> <span class="fu">position_jitter</span>(<span class="at">w =</span> jitter_size, <span class="at">h =</span> <span class="dv">0</span>)) <span class="sc">+</span></span>
<span id="cb13-76"><a href="appendix-f-creating-figures.html#cb13-76" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_jitter</span>(<span class="fu">aes</span>(</span>
<span id="cb13-77"><a href="appendix-f-creating-figures.html#cb13-77" aria-hidden="true" tabindex="-1"></a>    <span class="at">x =</span> pbmc_6000_df[<span class="fu">which</span>(pbmc_6000_df<span class="sc">$</span>algorithm <span class="sc">==</span> <span class="st">&quot;Monocle 3&quot;</span>),]<span class="sc">$</span>cell_type,</span>
<span id="cb13-78"><a href="appendix-f-creating-figures.html#cb13-78" aria-hidden="true" tabindex="-1"></a>    <span class="at">y =</span> pbmc_6000_df[<span class="fu">which</span>(pbmc_6000_df<span class="sc">$</span>algorithm <span class="sc">==</span> <span class="st">&quot;Monocle 3&quot;</span>),]<span class="sc">$</span>n),</span>
<span id="cb13-79"><a href="appendix-f-creating-figures.html#cb13-79" aria-hidden="true" tabindex="-1"></a>    <span class="at">color =</span> my_colors[<span class="dv">4</span>], <span class="at">size =</span> size, <span class="at">shape =</span> <span class="dv">16</span>, <span class="at">alpha =</span> alpha,</span>
<span id="cb13-80"><a href="appendix-f-creating-figures.html#cb13-80" aria-hidden="true" tabindex="-1"></a>    <span class="at">position =</span> <span class="fu">position_jitter</span>(<span class="at">w =</span> jitter_size, <span class="at">h =</span> <span class="dv">0</span>)) <span class="sc">+</span></span>
<span id="cb13-81"><a href="appendix-f-creating-figures.html#cb13-81" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_jitter</span>(<span class="fu">aes</span>(</span>
<span id="cb13-82"><a href="appendix-f-creating-figures.html#cb13-82" aria-hidden="true" tabindex="-1"></a>    <span class="at">x =</span> pbmc_6000_df[<span class="fu">which</span>(pbmc_6000_df<span class="sc">$</span>algorithm <span class="sc">==</span> <span class="st">&quot;SC3&quot;</span>),]<span class="sc">$</span>cell_type,</span>
<span id="cb13-83"><a href="appendix-f-creating-figures.html#cb13-83" aria-hidden="true" tabindex="-1"></a>    <span class="at">y =</span> pbmc_6000_df[<span class="fu">which</span>(pbmc_6000_df<span class="sc">$</span>algorithm <span class="sc">==</span> <span class="st">&quot;SC3&quot;</span>),]<span class="sc">$</span>n),</span>
<span id="cb13-84"><a href="appendix-f-creating-figures.html#cb13-84" aria-hidden="true" tabindex="-1"></a>    <span class="at">color =</span> my_colors[<span class="dv">5</span>], <span class="at">size =</span> size, <span class="at">shape =</span> <span class="dv">16</span>, <span class="at">alpha =</span> alpha,</span>
<span id="cb13-85"><a href="appendix-f-creating-figures.html#cb13-85" aria-hidden="true" tabindex="-1"></a>    <span class="at">position =</span> <span class="fu">position_jitter</span>(<span class="at">w =</span> jitter_size, <span class="at">h =</span> <span class="dv">0</span>)) <span class="sc">+</span></span>
<span id="cb13-86"><a href="appendix-f-creating-figures.html#cb13-86" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_jitter</span>(<span class="fu">aes</span>(</span>
<span id="cb13-87"><a href="appendix-f-creating-figures.html#cb13-87" aria-hidden="true" tabindex="-1"></a>    <span class="at">x =</span> pbmc_6000_df[<span class="fu">which</span>(pbmc_6000_df<span class="sc">$</span>algorithm <span class="sc">==</span> <span class="st">&quot;ASURAT&quot;</span>),]<span class="sc">$</span>cell_type,</span>
<span id="cb13-88"><a href="appendix-f-creating-figures.html#cb13-88" aria-hidden="true" tabindex="-1"></a>    <span class="at">y =</span> pbmc_6000_df[<span class="fu">which</span>(pbmc_6000_df<span class="sc">$</span>algorithm <span class="sc">==</span> <span class="st">&quot;ASURAT&quot;</span>),]<span class="sc">$</span>n),</span>
<span id="cb13-89"><a href="appendix-f-creating-figures.html#cb13-89" aria-hidden="true" tabindex="-1"></a>    <span class="at">color =</span> my_colors[<span class="dv">6</span>], <span class="at">size =</span> size, <span class="at">shape =</span> <span class="dv">16</span>, <span class="at">alpha =</span> alpha,</span>
<span id="cb13-90"><a href="appendix-f-creating-figures.html#cb13-90" aria-hidden="true" tabindex="-1"></a>    <span class="at">position =</span> <span class="fu">position_jitter</span>(<span class="at">w =</span> jitter_size, <span class="at">h =</span> <span class="dv">0</span>)) <span class="sc">+</span></span>
<span id="cb13-91"><a href="appendix-f-creating-figures.html#cb13-91" aria-hidden="true" tabindex="-1"></a>  <span class="co"># ----------------------------------------</span></span>
<span id="cb13-92"><a href="appendix-f-creating-figures.html#cb13-92" aria-hidden="true" tabindex="-1"></a>  <span class="co"># option</span></span>
<span id="cb13-93"><a href="appendix-f-creating-figures.html#cb13-93" aria-hidden="true" tabindex="-1"></a>  <span class="co"># ----------------------------------------</span></span>
<span id="cb13-94"><a href="appendix-f-creating-figures.html#cb13-94" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ylim</span>(<span class="dv">0</span>, ymax) <span class="sc">+</span> <span class="fu">scale_x_discrete</span>(<span class="at">guide =</span> <span class="fu">guide_axis</span>(<span class="at">angle =</span> <span class="dv">45</span>)) <span class="sc">+</span></span>
<span id="cb13-95"><a href="appendix-f-creating-figures.html#cb13-95" aria-hidden="true" tabindex="-1"></a>  <span class="fu">labs</span>(<span class="at">title =</span> <span class="st">&quot;&quot;</span>, <span class="at">x =</span> xlabel, <span class="at">y =</span> ylabel, <span class="at">colour =</span> <span class="st">&quot;&quot;</span>) <span class="sc">+</span></span>
<span id="cb13-96"><a href="appendix-f-creating-figures.html#cb13-96" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme_minimal</span>(<span class="at">base_size =</span> <span class="dv">16</span>, <span class="at">base_family =</span> <span class="st">&quot;Helvetica&quot;</span>) <span class="sc">+</span></span>
<span id="cb13-97"><a href="appendix-f-creating-figures.html#cb13-97" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme</span>(<span class="at">plot.title =</span> <span class="fu">element_text</span>(<span class="at">hjust =</span> <span class="fl">0.5</span>),</span>
<span id="cb13-98"><a href="appendix-f-creating-figures.html#cb13-98" aria-hidden="true" tabindex="-1"></a>        <span class="at">panel.grid =</span> <span class="fu">element_line</span>(<span class="at">color =</span> <span class="st">&quot;gray80&quot;</span>),</span>
<span id="cb13-99"><a href="appendix-f-creating-figures.html#cb13-99" aria-hidden="true" tabindex="-1"></a>        <span class="at">plot.margin =</span> <span class="fu">unit</span>(<span class="fu">c</span>(<span class="fl">1.5</span>,<span class="fl">1.5</span>,<span class="fl">1.5</span>,<span class="fl">1.5</span>), <span class="st">&quot;lines&quot;</span>),</span>
<span id="cb13-100"><a href="appendix-f-creating-figures.html#cb13-100" aria-hidden="true" tabindex="-1"></a>        <span class="at">legend.position =</span> <span class="st">&quot;right&quot;</span>)</span>
<span id="cb13-101"><a href="appendix-f-creating-figures.html#cb13-101" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb13-102"><a href="appendix-f-creating-figures.html#cb13-102" aria-hidden="true" tabindex="-1"></a>filename <span class="ot">&lt;-</span> <span class="st">&quot;figures/figure_20_0100.png&quot;</span></span>
<span id="cb13-103"><a href="appendix-f-creating-figures.html#cb13-103" aria-hidden="true" tabindex="-1"></a><span class="fu">ggsave</span>(<span class="at">file =</span> filename, <span class="at">plot =</span> p, <span class="at">dpi =</span> <span class="dv">300</span>, <span class="at">width =</span> <span class="fl">4.5</span>, <span class="at">height =</span> <span class="dv">5</span>)</span></code></pre></div>
<p>Below are the labels of methods.</p>
<ul>
<li>black: SCSA</li>
<li>red: scran</li>
<li>cyan: Seurat</li>
<li>green: Seurat with scCATCH</li>
<li>purple: Monocle 3</li>
<li>orange: SC3</li>
<li>yellow: ASURAT</li>
</ul>
<p><img src="figures/figure_20_0100.png" width="400px"></p>
</div>
</div>
<div id="lung-cancer-dataset" class="section level2" number="18.2">
<h2><span class="header-section-number">18.2</span> Lung cancer dataset</h2>
<div id="heatmaps-of-marker-gene-expressions" class="section level3" number="18.2.1">
<h3><span class="header-section-number">18.2.1</span> Heatmaps of marker gene expressions</h3>
<p>Load the data.</p>
<div class="sourceCode" id="cb14"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb14-1"><a href="appendix-f-creating-figures.html#cb14-1" aria-hidden="true" tabindex="-1"></a><span class="fu">rm</span>(<span class="at">list =</span> <span class="fu">ls</span>())</span>
<span id="cb14-2"><a href="appendix-f-creating-figures.html#cb14-2" aria-hidden="true" tabindex="-1"></a><span class="fu">source</span>(<span class="st">&quot;R/plot.R&quot;</span>)</span>
<span id="cb14-3"><a href="appendix-f-creating-figures.html#cb14-3" aria-hidden="true" tabindex="-1"></a>sc68_cisp <span class="ot">&lt;-</span> <span class="fu">readRDS</span>(<span class="at">file =</span> <span class="st">&quot;backup/02_005_sc68_cisp_normalized.rds&quot;</span>)</span>
<span id="cb14-4"><a href="appendix-f-creating-figures.html#cb14-4" aria-hidden="true" tabindex="-1"></a>sc68_cisp_c <span class="ot">&lt;-</span> <span class="fu">readRDS</span>(<span class="at">file =</span> <span class="st">&quot;backup/02_001_sc68_cisp_raw.rds&quot;</span>)</span></code></pre></div>
<p>Make a log-normalized read count table for ASCL1, NEUROD1, YAP1, and POU2F3.
Load the data.</p>
<div class="sourceCode" id="cb15"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb15-1"><a href="appendix-f-creating-figures.html#cb15-1" aria-hidden="true" tabindex="-1"></a>cells <span class="ot">&lt;-</span> sc68_cisp[[<span class="st">&quot;sample&quot;</span>]][[<span class="st">&quot;barcode&quot;</span>]]</span>
<span id="cb15-2"><a href="appendix-f-creating-figures.html#cb15-2" aria-hidden="true" tabindex="-1"></a>genes <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">&quot;ASCL1&quot;</span>, <span class="st">&quot;NEUROD1&quot;</span>, <span class="st">&quot;YAP1&quot;</span>, <span class="st">&quot;POU2F3&quot;</span>)</span>
<span id="cb15-3"><a href="appendix-f-creating-figures.html#cb15-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb15-4"><a href="appendix-f-creating-figures.html#cb15-4" aria-hidden="true" tabindex="-1"></a>mat <span class="ot">&lt;-</span> sc68_cisp_c[[<span class="st">&quot;data&quot;</span>]][[<span class="st">&quot;raw&quot;</span>]]</span>
<span id="cb15-5"><a href="appendix-f-creating-figures.html#cb15-5" aria-hidden="true" tabindex="-1"></a>mat <span class="ot">&lt;-</span> mat[<span class="fu">which</span>(<span class="fu">rownames</span>(mat) <span class="sc">%in%</span> genes),]</span>
<span id="cb15-6"><a href="appendix-f-creating-figures.html#cb15-6" aria-hidden="true" tabindex="-1"></a>mat <span class="ot">&lt;-</span> mat[, <span class="fu">which</span>(<span class="fu">colnames</span>(mat) <span class="sc">%in%</span> cells)]</span>
<span id="cb15-7"><a href="appendix-f-creating-figures.html#cb15-7" aria-hidden="true" tabindex="-1"></a>tmp <span class="ot">&lt;-</span> <span class="fu">log</span>(mat <span class="sc">+</span> <span class="dv">1</span>)</span>
<span id="cb15-8"><a href="appendix-f-creating-figures.html#cb15-8" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb15-9"><a href="appendix-f-creating-figures.html#cb15-9" aria-hidden="true" tabindex="-1"></a>sc68_cisp[[<span class="st">&quot;data&quot;</span>]][[<span class="st">&quot;normalized&quot;</span>]] <span class="ot">&lt;-</span> tmp</span></code></pre></div>
<p>The following function <code>plot_Heatmap_GenexSamp()</code> shows normalized read counts
of a given gene set as a heatmap, using the results of
<code>obj[["data"]][["normalized"]]</code>.</p>
<div class="sourceCode" id="cb16"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb16-1"><a href="appendix-f-creating-figures.html#cb16-1" aria-hidden="true" tabindex="-1"></a><span class="co"># ----------------------------------------</span></span>
<span id="cb16-2"><a href="appendix-f-creating-figures.html#cb16-2" aria-hidden="true" tabindex="-1"></a><span class="co"># sc68_cisp</span></span>
<span id="cb16-3"><a href="appendix-f-creating-figures.html#cb16-3" aria-hidden="true" tabindex="-1"></a><span class="co"># ----------------------------------------</span></span>
<span id="cb16-4"><a href="appendix-f-creating-figures.html#cb16-4" aria-hidden="true" tabindex="-1"></a>filename <span class="ot">&lt;-</span> <span class="st">&quot;figures/figure_20_0200.png&quot;</span></span>
<span id="cb16-5"><a href="appendix-f-creating-figures.html#cb16-5" aria-hidden="true" tabindex="-1"></a><span class="fu">png</span>(<span class="at">file =</span> filename, <span class="at">height =</span> <span class="dv">900</span>, <span class="at">width =</span> <span class="dv">2000</span>, <span class="at">res =</span> <span class="dv">300</span>)</span>
<span id="cb16-6"><a href="appendix-f-creating-figures.html#cb16-6" aria-hidden="true" tabindex="-1"></a><span class="fu">plot_Heatmap_GenexSamp</span>(<span class="at">obj =</span> sc68_cisp, <span class="at">genes =</span> genes,</span>
<span id="cb16-7"><a href="appendix-f-creating-figures.html#cb16-7" aria-hidden="true" tabindex="-1"></a>                       <span class="at">method =</span> <span class="st">&quot;ward.D2&quot;</span>, <span class="at">show_nReads =</span> <span class="cn">TRUE</span>,</span>
<span id="cb16-8"><a href="appendix-f-creating-figures.html#cb16-8" aria-hidden="true" tabindex="-1"></a>                       <span class="at">title =</span> <span class="st">&quot;Lung cancer scRNA-seq&quot;</span>, <span class="at">name =</span> <span class="st">&quot;Expression&quot;</span>)</span>
<span id="cb16-9"><a href="appendix-f-creating-figures.html#cb16-9" aria-hidden="true" tabindex="-1"></a><span class="fu">dev.off</span>()</span></code></pre></div>
</div>
<div id="multi-layered-analysis-1" class="section level3" number="18.2.2">
<h3><span class="header-section-number">18.2.2</span> Multi-layered analysis</h3>
<p>Load the data.</p>
<div class="sourceCode" id="cb17"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb17-1"><a href="appendix-f-creating-figures.html#cb17-1" aria-hidden="true" tabindex="-1"></a><span class="fu">rm</span>(<span class="at">list =</span> <span class="fu">ls</span>())</span>
<span id="cb17-2"><a href="appendix-f-creating-figures.html#cb17-2" aria-hidden="true" tabindex="-1"></a><span class="fu">source</span>(<span class="st">&quot;R/function_seurat.R&quot;</span>)</span>
<span id="cb17-3"><a href="appendix-f-creating-figures.html#cb17-3" aria-hidden="true" tabindex="-1"></a><span class="fu">source</span>(<span class="st">&quot;R/function_asurat.R&quot;</span>)</span>
<span id="cb17-4"><a href="appendix-f-creating-figures.html#cb17-4" aria-hidden="true" tabindex="-1"></a><span class="fu">source</span>(<span class="st">&quot;R/plot.R&quot;</span>)</span>
<span id="cb17-5"><a href="appendix-f-creating-figures.html#cb17-5" aria-hidden="true" tabindex="-1"></a>sc68_cisp_asurat <span class="ot">&lt;-</span> <span class="fu">readRDS</span>(<span class="at">file =</span> <span class="st">&quot;backup/02_400_sc68_cisp_all.rds&quot;</span>)</span></code></pre></div>
<p>Check the size of the gene-by-sample matrices.</p>
<div class="sourceCode" id="cb18"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb18-1"><a href="appendix-f-creating-figures.html#cb18-1" aria-hidden="true" tabindex="-1"></a><span class="fu">dim</span>(sc68_cisp_asurat[[<span class="st">&quot;data&quot;</span>]][[<span class="st">&quot;bayNorm&quot;</span>]][[<span class="st">&quot;Bay_out&quot;</span>]])</span></code></pre></div>
<pre><code>[1,] 6347 2285</code></pre>
<p>Create Seruat objects.</p>
<div class="sourceCode" id="cb20"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb20-1"><a href="appendix-f-creating-figures.html#cb20-1" aria-hidden="true" tabindex="-1"></a>sc68_cisp <span class="ot">&lt;-</span> <span class="fu">CreateSeuratObject</span>(</span>
<span id="cb20-2"><a href="appendix-f-creating-figures.html#cb20-2" aria-hidden="true" tabindex="-1"></a>  <span class="at">counts =</span> <span class="fu">as.matrix</span>(sc68_cisp_asurat[[<span class="st">&quot;data&quot;</span>]][[<span class="st">&quot;bayNorm&quot;</span>]][[<span class="st">&quot;Bay_out&quot;</span>]]),</span>
<span id="cb20-3"><a href="appendix-f-creating-figures.html#cb20-3" aria-hidden="true" tabindex="-1"></a>  <span class="at">project =</span> <span class="st">&quot;sc68_cisp&quot;</span>)</span></code></pre></div>
<p>Run Seurat’s standard procedures but the clustering result will be replaced
with that of ASURAT.
Hence, parameters <code>nfeatures</code>, <code>sc68_cisp_pc</code>, and <code>resolution</code> are roughly set.</p>
<div class="sourceCode" id="cb21"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb21-1"><a href="appendix-f-creating-figures.html#cb21-1" aria-hidden="true" tabindex="-1"></a>nfeatures <span class="ot">&lt;-</span> <span class="fu">round</span>(<span class="fl">0.2</span> <span class="sc">*</span> sc68_cisp<span class="sc">@</span>assays[[<span class="st">&quot;RNA&quot;</span>]]<span class="sc">@</span>counts<span class="sc">@</span>Dim[<span class="dv">2</span>])</span>
<span id="cb21-2"><a href="appendix-f-creating-figures.html#cb21-2" aria-hidden="true" tabindex="-1"></a>sc68_cisp <span class="ot">&lt;-</span> <span class="fu">process_001_seurat</span>(<span class="at">obj =</span> sc68_cisp, <span class="at">nfeatures =</span> nfeatures)</span>
<span id="cb21-3"><a href="appendix-f-creating-figures.html#cb21-3" aria-hidden="true" tabindex="-1"></a>sc68_cisp_pc <span class="ot">&lt;-</span> <span class="fu">which</span>(<span class="fu">cumsum</span>(sc68_cisp<span class="sc">@</span>reductions[[<span class="st">&quot;pca&quot;</span>]]<span class="sc">@</span>stdev) <span class="sc">/</span></span>
<span id="cb21-4"><a href="appendix-f-creating-figures.html#cb21-4" aria-hidden="true" tabindex="-1"></a>                      <span class="fu">sum</span>(sc68_cisp<span class="sc">@</span>reductions[[<span class="st">&quot;pca&quot;</span>]]<span class="sc">@</span>stdev) <span class="sc">&gt;</span> <span class="fl">0.9</span>)[<span class="dv">1</span>]</span>
<span id="cb21-5"><a href="appendix-f-creating-figures.html#cb21-5" aria-hidden="true" tabindex="-1"></a>sc68_cisp <span class="ot">&lt;-</span> <span class="fu">process_002_seurat</span>(<span class="at">obj =</span> sc68_cisp, <span class="at">pc =</span> sc68_cisp_pc,</span>
<span id="cb21-6"><a href="appendix-f-creating-figures.html#cb21-6" aria-hidden="true" tabindex="-1"></a>                                <span class="at">resolution =</span> <span class="fl">0.1</span>)</span></code></pre></div>
<p>Assign a cell cycle phase to each cell.</p>
<div class="sourceCode" id="cb22"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb22-1"><a href="appendix-f-creating-figures.html#cb22-1" aria-hidden="true" tabindex="-1"></a>s.genes <span class="ot">&lt;-</span> cc.genes<span class="sc">$</span>s.genes</span>
<span id="cb22-2"><a href="appendix-f-creating-figures.html#cb22-2" aria-hidden="true" tabindex="-1"></a>g2m.genes <span class="ot">&lt;-</span> cc.genes<span class="sc">$</span>g2m.genes</span>
<span id="cb22-3"><a href="appendix-f-creating-figures.html#cb22-3" aria-hidden="true" tabindex="-1"></a>sc68_cisp <span class="ot">&lt;-</span> <span class="fu">CellCycleScoring</span>(sc68_cisp, <span class="at">s.features =</span> s.genes,</span>
<span id="cb22-4"><a href="appendix-f-creating-figures.html#cb22-4" aria-hidden="true" tabindex="-1"></a>                              <span class="at">g2m.features =</span> g2m.genes, <span class="at">set.ident =</span> <span class="cn">FALSE</span>)</span>
<span id="cb22-5"><a href="appendix-f-creating-figures.html#cb22-5" aria-hidden="true" tabindex="-1"></a>tmp <span class="ot">&lt;-</span> <span class="fu">factor</span>(sc68_cisp<span class="sc">@</span>meta.data[[<span class="st">&quot;Phase&quot;</span>]],</span>
<span id="cb22-6"><a href="appendix-f-creating-figures.html#cb22-6" aria-hidden="true" tabindex="-1"></a>              <span class="at">levels =</span> <span class="fu">c</span>(<span class="st">&quot;G1&quot;</span>, <span class="st">&quot;S&quot;</span>, <span class="st">&quot;G2M&quot;</span>, <span class="st">&quot;Undecided&quot;</span>))</span>
<span id="cb22-7"><a href="appendix-f-creating-figures.html#cb22-7" aria-hidden="true" tabindex="-1"></a>sc68_cisp<span class="sc">@</span>meta.data[[<span class="st">&quot;Phase&quot;</span>]] <span class="ot">&lt;-</span> tmp</span>
<span id="cb22-8"><a href="appendix-f-creating-figures.html#cb22-8" aria-hidden="true" tabindex="-1"></a>sc68_cisp_asurat[[<span class="st">&quot;sample&quot;</span>]][[<span class="st">&quot;phase&quot;</span>]] <span class="ot">&lt;-</span> sc68_cisp<span class="sc">@</span>meta.data[[<span class="st">&quot;Phase&quot;</span>]]</span></code></pre></div>
<p>Add the cell labels to <code>obj@meta.data[["asurat_clusters"]]</code>, which were inferred
by ASURAT’s computations (see <a href="cell_type.html#cell_type">previous section</a> for the details).</p>
<div class="sourceCode" id="cb23"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb23-1"><a href="appendix-f-creating-figures.html#cb23-1" aria-hidden="true" tabindex="-1"></a>tmp <span class="ot">&lt;-</span> <span class="fu">as.factor</span>(sc68_cisp_asurat[[<span class="st">&quot;sample&quot;</span>]][[<span class="st">&quot;merlot_DO_disease&quot;</span>]])</span>
<span id="cb23-2"><a href="appendix-f-creating-figures.html#cb23-2" aria-hidden="true" tabindex="-1"></a>sc68_cisp<span class="sc">@</span>meta.data[[<span class="st">&quot;asurat_clusters&quot;</span>]] <span class="ot">&lt;-</span> tmp</span>
<span id="cb23-3"><a href="appendix-f-creating-figures.html#cb23-3" aria-hidden="true" tabindex="-1"></a>sc68_cisp <span class="ot">&lt;-</span> <span class="fu">SetIdent</span>(sc68_cisp, <span class="at">value =</span> <span class="st">&quot;asurat_clusters&quot;</span>)</span></code></pre></div>
<p>Find differentially expressed genes using Seurat’s function and stores the
results into <code>obj@misc[["markers"]]</code>.</p>
<div class="sourceCode" id="cb24"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb24-1"><a href="appendix-f-creating-figures.html#cb24-1" aria-hidden="true" tabindex="-1"></a>sc68_cisp<span class="sc">@</span>misc[[<span class="st">&quot;markers&quot;</span>]] <span class="ot">&lt;-</span> <span class="fu">FindAllMarkers</span>(</span>
<span id="cb24-2"><a href="appendix-f-creating-figures.html#cb24-2" aria-hidden="true" tabindex="-1"></a>  sc68_cisp, <span class="at">only.pos =</span> T, <span class="at">min.pct =</span> <span class="fl">0.25</span>, <span class="at">logfc.threshold =</span> <span class="fl">0.25</span>)</span></code></pre></div>
<p>Users can check the marker genes of each cluster by the following command:</p>
<div class="sourceCode" id="cb25"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb25-1"><a href="appendix-f-creating-figures.html#cb25-1" aria-hidden="true" tabindex="-1"></a><span class="fu">datatable</span>(sc68_cisp<span class="sc">@</span>misc[[<span class="st">&quot;markers&quot;</span>]], <span class="at">rownames =</span> <span class="cn">FALSE</span>)</span></code></pre></div>
<ul>
<li><code>sc68_cisp</code></li>
</ul>
<iframe src="figures/sc68_cisp_marker_asurat.html" style="width:100%; height:400px;">
</iframe>
<p>Select top <code>topn</code> genes, in terms of p-value, and show the expression levels
by complex heat map.</p>
<div class="sourceCode" id="cb26"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb26-1"><a href="appendix-f-creating-figures.html#cb26-1" aria-hidden="true" tabindex="-1"></a>topn <span class="ot">&lt;-</span> sc68_cisp<span class="sc">@</span>misc[[<span class="st">&quot;markers&quot;</span>]] <span class="sc">%&gt;%</span> <span class="fu">group_by</span>(cluster)</span>
<span id="cb26-2"><a href="appendix-f-creating-figures.html#cb26-2" aria-hidden="true" tabindex="-1"></a>topn_new <span class="ot">&lt;-</span> topn <span class="sc">%&gt;%</span> <span class="fu">top_n</span>(<span class="at">n=</span><span class="sc">-</span><span class="dv">10</span>, p_val_adj) <span class="co"># Selecting by p_val_adj</span></span>
<span id="cb26-3"><a href="appendix-f-creating-figures.html#cb26-3" aria-hidden="true" tabindex="-1"></a>topn_genes <span class="ot">&lt;-</span> <span class="fu">data.frame</span>(<span class="at">gene =</span> topn_new<span class="sc">$</span>gene, <span class="at">label =</span> topn_new<span class="sc">$</span>cluster)</span>
<span id="cb26-4"><a href="appendix-f-creating-figures.html#cb26-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb26-5"><a href="appendix-f-creating-figures.html#cb26-5" aria-hidden="true" tabindex="-1"></a>filename <span class="ot">&lt;-</span> <span class="st">&quot;figures/figure_20_0210.png&quot;</span></span>
<span id="cb26-6"><a href="appendix-f-creating-figures.html#cb26-6" aria-hidden="true" tabindex="-1"></a><span class="fu">png</span>(<span class="at">file =</span> filename, <span class="at">height =</span> <span class="dv">2000</span>, <span class="at">width =</span> <span class="dv">1500</span>, <span class="at">res =</span> <span class="dv">300</span>)</span>
<span id="cb26-7"><a href="appendix-f-creating-figures.html#cb26-7" aria-hidden="true" tabindex="-1"></a><span class="fu">plot_Heatmap_seurat</span>(<span class="at">obj =</span> sc68_cisp, <span class="at">topn_genes =</span> topn_genes,</span>
<span id="cb26-8"><a href="appendix-f-creating-figures.html#cb26-8" aria-hidden="true" tabindex="-1"></a>                    <span class="at">method =</span> <span class="st">&quot;ward.D2&quot;</span>, <span class="at">show_nReads =</span> <span class="cn">TRUE</span>,</span>
<span id="cb26-9"><a href="appendix-f-creating-figures.html#cb26-9" aria-hidden="true" tabindex="-1"></a>                    <span class="at">title =</span> <span class="st">&quot;sc68_cisp (ASURAT)&quot;</span>, <span class="at">name =</span> <span class="st">&quot;Expr level&quot;</span>,</span>
<span id="cb26-10"><a href="appendix-f-creating-figures.html#cb26-10" aria-hidden="true" tabindex="-1"></a>                    <span class="at">show_rownames_nReads =</span> <span class="cn">FALSE</span>, <span class="at">default_color =</span> <span class="cn">TRUE</span>)</span>
<span id="cb26-11"><a href="appendix-f-creating-figures.html#cb26-11" aria-hidden="true" tabindex="-1"></a><span class="fu">dev.off</span>()</span></code></pre></div>
<p><img src="figures/figure_20_0210.png" width="400px"></p>
<p>The following function <code>plot_MultiHeatmaps_SignxSamp_GenexSamp()</code> vertically
concatenates two or more heat maps for top n signs, with respect to sign score,
and top n genes with respect to p-value.
Users can remove signs by setting <code>keywords</code> (list of keywords separated by <code>|</code>)
which obviously seem to be unimportant or untranslatable.</p>
<div class="sourceCode" id="cb27"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb27-1"><a href="appendix-f-creating-figures.html#cb27-1" aria-hidden="true" tabindex="-1"></a>topn <span class="ot">&lt;-</span> sc68_cisp<span class="sc">@</span>misc[[<span class="st">&quot;markers&quot;</span>]] <span class="sc">%&gt;%</span> <span class="fu">group_by</span>(cluster)</span>
<span id="cb27-2"><a href="appendix-f-creating-figures.html#cb27-2" aria-hidden="true" tabindex="-1"></a>topn_new <span class="ot">&lt;-</span> topn <span class="sc">%&gt;%</span> <span class="fu">top_n</span>(<span class="at">n=</span><span class="sc">-</span><span class="dv">8</span>, p_val_adj) <span class="co"># Selecting by p_val_adj</span></span>
<span id="cb27-3"><a href="appendix-f-creating-figures.html#cb27-3" aria-hidden="true" tabindex="-1"></a>topn_genes <span class="ot">&lt;-</span> <span class="fu">data.frame</span>(<span class="at">gene =</span> topn_new<span class="sc">$</span>gene, <span class="at">label =</span> topn_new<span class="sc">$</span>cluster)</span>
<span id="cb27-4"><a href="appendix-f-creating-figures.html#cb27-4" aria-hidden="true" tabindex="-1"></a>topn_genes<span class="sc">$</span>label <span class="ot">&lt;-</span> <span class="fu">paste</span>(<span class="st">&quot;DEG (group &quot;</span>, topn_genes<span class="sc">$</span>label, <span class="st">&quot;)&quot;</span>, <span class="at">sep =</span> <span class="st">&quot;&quot;</span>)</span>
<span id="cb27-5"><a href="appendix-f-creating-figures.html#cb27-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb27-6"><a href="appendix-f-creating-figures.html#cb27-6" aria-hidden="true" tabindex="-1"></a>tmp <span class="ot">&lt;-</span> sc68_cisp_asurat[[<span class="st">&quot;marker&quot;</span>]][[<span class="st">&quot;DO&quot;</span>]][[<span class="st">&quot;disease&quot;</span>]][[<span class="st">&quot;all&quot;</span>]]</span>
<span id="cb27-7"><a href="appendix-f-creating-figures.html#cb27-7" aria-hidden="true" tabindex="-1"></a>topn <span class="ot">&lt;-</span> tmp <span class="sc">%&gt;%</span> <span class="fu">group_by</span>(Label)</span>
<span id="cb27-8"><a href="appendix-f-creating-figures.html#cb27-8" aria-hidden="true" tabindex="-1"></a>keywords <span class="ot">&lt;-</span> <span class="st">&quot;embryonal cancer|aplastic anemia|Diamond-Blackfan anemia&quot;</span></span>
<span id="cb27-9"><a href="appendix-f-creating-figures.html#cb27-9" aria-hidden="true" tabindex="-1"></a>keywords <span class="ot">&lt;-</span> <span class="fu">paste</span>(keywords, <span class="st">&quot;|meningioma|cervi|spondylitis&quot;</span>)</span>
<span id="cb27-10"><a href="appendix-f-creating-figures.html#cb27-10" aria-hidden="true" tabindex="-1"></a>topn <span class="ot">&lt;-</span> topn[<span class="sc">!</span>((<span class="fu">grepl</span>(keywords, topn<span class="sc">$</span>Sign)) <span class="sc">|</span></span>
<span id="cb27-11"><a href="appendix-f-creating-figures.html#cb27-11" aria-hidden="true" tabindex="-1"></a>                 (<span class="fu">grepl</span>(keywords, topn<span class="sc">$</span>ParentDescription))),]</span>
<span id="cb27-12"><a href="appendix-f-creating-figures.html#cb27-12" aria-hidden="true" tabindex="-1"></a>topn_new <span class="ot">&lt;-</span> topn <span class="sc">%&gt;%</span> <span class="fu">top_n</span>(<span class="dv">3</span>, sep_I)</span>
<span id="cb27-13"><a href="appendix-f-creating-figures.html#cb27-13" aria-hidden="true" tabindex="-1"></a>topn_signs_DO <span class="ot">&lt;-</span> <span class="fu">data.frame</span>(</span>
<span id="cb27-14"><a href="appendix-f-creating-figures.html#cb27-14" aria-hidden="true" tabindex="-1"></a>  <span class="at">sign =</span> topn_new<span class="sc">$</span>Sign, <span class="at">description =</span> topn_new<span class="sc">$</span>ParentDescription,</span>
<span id="cb27-15"><a href="appendix-f-creating-figures.html#cb27-15" aria-hidden="true" tabindex="-1"></a>  <span class="at">label =</span> topn_new<span class="sc">$</span>Label)</span>
<span id="cb27-16"><a href="appendix-f-creating-figures.html#cb27-16" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb27-17"><a href="appendix-f-creating-figures.html#cb27-17" aria-hidden="true" tabindex="-1"></a>keywords <span class="ot">&lt;-</span> <span class="st">&quot;viral|learning&quot;</span></span>
<span id="cb27-18"><a href="appendix-f-creating-figures.html#cb27-18" aria-hidden="true" tabindex="-1"></a>tmp <span class="ot">&lt;-</span> sc68_cisp_asurat[[<span class="st">&quot;marker&quot;</span>]][[<span class="st">&quot;GO&quot;</span>]][[<span class="st">&quot;BP&quot;</span>]][[<span class="st">&quot;all&quot;</span>]]</span>
<span id="cb27-19"><a href="appendix-f-creating-figures.html#cb27-19" aria-hidden="true" tabindex="-1"></a>topn <span class="ot">&lt;-</span> tmp <span class="sc">%&gt;%</span> <span class="fu">group_by</span>(Label)</span>
<span id="cb27-20"><a href="appendix-f-creating-figures.html#cb27-20" aria-hidden="true" tabindex="-1"></a>topn <span class="ot">&lt;-</span> tmp <span class="sc">%&gt;%</span> <span class="fu">group_by</span>(Label)</span>
<span id="cb27-21"><a href="appendix-f-creating-figures.html#cb27-21" aria-hidden="true" tabindex="-1"></a>topn <span class="ot">&lt;-</span> topn[<span class="sc">!</span>((<span class="fu">grepl</span>(keywords, topn<span class="sc">$</span>Sign)) <span class="sc">|</span></span>
<span id="cb27-22"><a href="appendix-f-creating-figures.html#cb27-22" aria-hidden="true" tabindex="-1"></a>                 (<span class="fu">grepl</span>(keywords, topn<span class="sc">$</span>ParentDescription))),]</span>
<span id="cb27-23"><a href="appendix-f-creating-figures.html#cb27-23" aria-hidden="true" tabindex="-1"></a>topn_new <span class="ot">&lt;-</span> topn <span class="sc">%&gt;%</span> <span class="fu">top_n</span>(<span class="dv">2</span>, sep_I)</span>
<span id="cb27-24"><a href="appendix-f-creating-figures.html#cb27-24" aria-hidden="true" tabindex="-1"></a>topn_signs_GO <span class="ot">&lt;-</span> <span class="fu">data.frame</span>(</span>
<span id="cb27-25"><a href="appendix-f-creating-figures.html#cb27-25" aria-hidden="true" tabindex="-1"></a>  <span class="at">sign =</span> topn_new<span class="sc">$</span>Sign, <span class="at">description =</span> topn_new<span class="sc">$</span>ParentDescription,</span>
<span id="cb27-26"><a href="appendix-f-creating-figures.html#cb27-26" aria-hidden="true" tabindex="-1"></a>  <span class="at">label =</span> topn_new<span class="sc">$</span>Label)</span>
<span id="cb27-27"><a href="appendix-f-creating-figures.html#cb27-27" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb27-28"><a href="appendix-f-creating-figures.html#cb27-28" aria-hidden="true" tabindex="-1"></a>keywords <span class="ot">&lt;-</span> <span class="fu">paste</span>(<span class="st">&quot;Glioma|Ribosome|Oocyte|oocyte|path:hsa04650_S|&quot;</span>,</span>
<span id="cb27-29"><a href="appendix-f-creating-figures.html#cb27-29" aria-hidden="true" tabindex="-1"></a>                  <span class="st">&quot;path:hsa04380_S|path:hsa05162_S|path:hsa05416_S|&quot;</span>,</span>
<span id="cb27-30"><a href="appendix-f-creating-figures.html#cb27-30" aria-hidden="true" tabindex="-1"></a>                  <span class="st">&quot;path:hsa05140_S|path:hsa04662_S&quot;</span>, <span class="at">sep =</span> <span class="st">&quot;&quot;</span>)</span>
<span id="cb27-31"><a href="appendix-f-creating-figures.html#cb27-31" aria-hidden="true" tabindex="-1"></a>tmp <span class="ot">&lt;-</span> sc68_cisp_asurat[[<span class="st">&quot;marker&quot;</span>]][[<span class="st">&quot;KEGG&quot;</span>]][[<span class="st">&quot;pathway&quot;</span>]][[<span class="st">&quot;all&quot;</span>]]</span>
<span id="cb27-32"><a href="appendix-f-creating-figures.html#cb27-32" aria-hidden="true" tabindex="-1"></a>topn <span class="ot">&lt;-</span> tmp <span class="sc">%&gt;%</span> <span class="fu">group_by</span>(Label)</span>
<span id="cb27-33"><a href="appendix-f-creating-figures.html#cb27-33" aria-hidden="true" tabindex="-1"></a>topn <span class="ot">&lt;-</span> tmp <span class="sc">%&gt;%</span> <span class="fu">group_by</span>(Label)</span>
<span id="cb27-34"><a href="appendix-f-creating-figures.html#cb27-34" aria-hidden="true" tabindex="-1"></a>topn <span class="ot">&lt;-</span> topn[<span class="sc">!</span>((<span class="fu">grepl</span>(keywords, topn<span class="sc">$</span>Sign)) <span class="sc">|</span></span>
<span id="cb27-35"><a href="appendix-f-creating-figures.html#cb27-35" aria-hidden="true" tabindex="-1"></a>                 (<span class="fu">grepl</span>(keywords, topn<span class="sc">$</span>ParentDescription))),]</span>
<span id="cb27-36"><a href="appendix-f-creating-figures.html#cb27-36" aria-hidden="true" tabindex="-1"></a>topn_new <span class="ot">&lt;-</span> topn <span class="sc">%&gt;%</span> <span class="fu">top_n</span>(<span class="dv">5</span>, sep_I)</span>
<span id="cb27-37"><a href="appendix-f-creating-figures.html#cb27-37" aria-hidden="true" tabindex="-1"></a>topn_signs_KEGG <span class="ot">&lt;-</span> <span class="fu">data.frame</span>(</span>
<span id="cb27-38"><a href="appendix-f-creating-figures.html#cb27-38" aria-hidden="true" tabindex="-1"></a>  <span class="at">sign =</span> topn_new<span class="sc">$</span>Sign, <span class="at">description =</span> topn_new<span class="sc">$</span>ParentDescription,</span>
<span id="cb27-39"><a href="appendix-f-creating-figures.html#cb27-39" aria-hidden="true" tabindex="-1"></a>  <span class="at">label =</span> topn_new<span class="sc">$</span>Label)</span>
<span id="cb27-40"><a href="appendix-f-creating-figures.html#cb27-40" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb27-41"><a href="appendix-f-creating-figures.html#cb27-41" aria-hidden="true" tabindex="-1"></a>topn_signs <span class="ot">&lt;-</span> <span class="fu">list</span>(topn_signs_DO, topn_signs_GO, topn_signs_KEGG)</span>
<span id="cb27-42"><a href="appendix-f-creating-figures.html#cb27-42" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb27-43"><a href="appendix-f-creating-figures.html#cb27-43" aria-hidden="true" tabindex="-1"></a>filename <span class="ot">&lt;-</span> <span class="st">&quot;figures/figure_20_0211.png&quot;</span></span>
<span id="cb27-44"><a href="appendix-f-creating-figures.html#cb27-44" aria-hidden="true" tabindex="-1"></a><span class="fu">png</span>(<span class="at">file =</span> filename, <span class="at">height =</span> <span class="dv">3200</span>, <span class="at">width =</span> <span class="dv">2000</span>, <span class="at">res =</span> <span class="dv">300</span>)</span>
<span id="cb27-45"><a href="appendix-f-creating-figures.html#cb27-45" aria-hidden="true" tabindex="-1"></a><span class="fu">plot_MultiHeatmaps_SignxSamp_GenexSamp</span>(</span>
<span id="cb27-46"><a href="appendix-f-creating-figures.html#cb27-46" aria-hidden="true" tabindex="-1"></a>  <span class="at">obj =</span> sc68_cisp_asurat, <span class="at">topn_genes =</span> topn_genes, <span class="at">topn_signs =</span> topn_signs,</span>
<span id="cb27-47"><a href="appendix-f-creating-figures.html#cb27-47" aria-hidden="true" tabindex="-1"></a>  <span class="at">data_types =</span> <span class="fu">c</span>(<span class="st">&quot;DO&quot;</span>, <span class="st">&quot;GO&quot;</span>, <span class="st">&quot;KEGG&quot;</span>),</span>
<span id="cb27-48"><a href="appendix-f-creating-figures.html#cb27-48" aria-hidden="true" tabindex="-1"></a>  <span class="at">categories =</span> <span class="fu">c</span>(<span class="st">&quot;disease&quot;</span>, <span class="st">&quot;BP&quot;</span>, <span class="st">&quot;pathway&quot;</span>),</span>
<span id="cb27-49"><a href="appendix-f-creating-figures.html#cb27-49" aria-hidden="true" tabindex="-1"></a>  <span class="at">algo_names =</span> <span class="fu">c</span>(<span class="st">&quot;merlot&quot;</span>, <span class="cn">NULL</span>, <span class="cn">NULL</span>),</span>
<span id="cb27-50"><a href="appendix-f-creating-figures.html#cb27-50" aria-hidden="true" tabindex="-1"></a>  <span class="at">show_labels =</span> <span class="fu">c</span>(<span class="cn">TRUE</span>, <span class="cn">FALSE</span>, <span class="cn">FALSE</span>),</span>
<span id="cb27-51"><a href="appendix-f-creating-figures.html#cb27-51" aria-hidden="true" tabindex="-1"></a>  <span class="at">method =</span> <span class="st">&quot;ward.D2&quot;</span>, <span class="at">show_nReads =</span> <span class="cn">FALSE</span>,</span>
<span id="cb27-52"><a href="appendix-f-creating-figures.html#cb27-52" aria-hidden="true" tabindex="-1"></a>  <span class="at">title =</span> <span class="st">&quot;sc68_cisp (ASURAT)&quot;</span>,</span>
<span id="cb27-53"><a href="appendix-f-creating-figures.html#cb27-53" aria-hidden="true" tabindex="-1"></a>  <span class="at">names_signs =</span> <span class="fu">c</span>(<span class="st">&quot;DO:</span><span class="sc">\n</span><span class="st">disease&quot;</span>, <span class="st">&quot;GO:</span><span class="sc">\n</span><span class="st">BP&quot;</span>, <span class="st">&quot;KEGG:</span><span class="sc">\n</span><span class="st">pathway&quot;</span>),</span>
<span id="cb27-54"><a href="appendix-f-creating-figures.html#cb27-54" aria-hidden="true" tabindex="-1"></a>  <span class="at">name_genes =</span> <span class="st">&quot;Expression&quot;</span>,</span>
<span id="cb27-55"><a href="appendix-f-creating-figures.html#cb27-55" aria-hidden="true" tabindex="-1"></a>  <span class="at">show_rownames_sign =</span> <span class="cn">TRUE</span>, <span class="at">show_rownames_label =</span> <span class="cn">FALSE</span>,</span>
<span id="cb27-56"><a href="appendix-f-creating-figures.html#cb27-56" aria-hidden="true" tabindex="-1"></a>  <span class="at">show_rownames_nReads =</span> <span class="cn">FALSE</span>, <span class="at">default_colors =</span> <span class="fu">c</span>(<span class="cn">TRUE</span>, <span class="cn">FALSE</span>, <span class="cn">FALSE</span>),</span>
<span id="cb27-57"><a href="appendix-f-creating-figures.html#cb27-57" aria-hidden="true" tabindex="-1"></a>  <span class="at">show_cellcycle =</span> <span class="cn">TRUE</span></span>
<span id="cb27-58"><a href="appendix-f-creating-figures.html#cb27-58" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb27-59"><a href="appendix-f-creating-figures.html#cb27-59" aria-hidden="true" tabindex="-1"></a><span class="fu">dev.off</span>()</span></code></pre></div>
<p><img src="figures/figure_20_0211.png" width="400px"></p>
<p>Save the objects.</p>
<div class="sourceCode" id="cb28"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb28-1"><a href="appendix-f-creating-figures.html#cb28-1" aria-hidden="true" tabindex="-1"></a><span class="fu">saveRDS</span>(sc68_cisp, <span class="at">file =</span> <span class="st">&quot;backup/20_001_sc68_cisp.rds&quot;</span>)</span>
<span id="cb28-2"><a href="appendix-f-creating-figures.html#cb28-2" aria-hidden="true" tabindex="-1"></a><span class="fu">saveRDS</span>(sc68_cisp_asurat, <span class="at">file =</span> <span class="st">&quot;backup/20_002_sc68_cisp_asurat.rds&quot;</span>)</span></code></pre></div>
<p>Load the data.</p>
<div class="sourceCode" id="cb29"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb29-1"><a href="appendix-f-creating-figures.html#cb29-1" aria-hidden="true" tabindex="-1"></a>sc68_cisp <span class="ot">&lt;-</span> <span class="fu">readRDS</span>(<span class="at">file =</span> <span class="st">&quot;backup/20_001_sc68_cisp.rds&quot;</span>)</span>
<span id="cb29-2"><a href="appendix-f-creating-figures.html#cb29-2" aria-hidden="true" tabindex="-1"></a>sc68_cisp_asurat <span class="ot">&lt;-</span> <span class="fu">readRDS</span>(<span class="at">file =</span> <span class="st">&quot;backup/20_002_sc68_cisp_asurat.rds&quot;</span>)</span></code></pre></div>
</div>
</div>
<div id="pancreatic-ductal-adenocarcinoma-pdac-datasets" class="section level2" number="18.3">
<h2><span class="header-section-number">18.3</span> Pancreatic ductal adenocarcinoma (PDAC) datasets</h2>
<div id="clustering-pdac-spatial-transcriptomes" class="section level3" number="18.3.1">
<h3><span class="header-section-number">18.3.1</span> Clustering PDAC spatial transcriptomes</h3>
<p>Load the data.</p>
<div class="sourceCode" id="cb30"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb30-1"><a href="appendix-f-creating-figures.html#cb30-1" aria-hidden="true" tabindex="-1"></a>pdac_aint <span class="ot">&lt;-</span> <span class="fu">readRDS</span>(<span class="at">file =</span> <span class="st">&quot;backup/07_400_pdac_aint_all.rds&quot;</span>)</span>
<span id="cb30-2"><a href="appendix-f-creating-figures.html#cb30-2" aria-hidden="true" tabindex="-1"></a>pdac_ast1 <span class="ot">&lt;-</span> <span class="fu">readRDS</span>(<span class="at">file =</span> <span class="st">&quot;backup/05_005_pdac_ast1_normalized.rds&quot;</span>)</span>
<span id="cb30-3"><a href="appendix-f-creating-figures.html#cb30-3" aria-hidden="true" tabindex="-1"></a>fn <span class="ot">&lt;-</span> <span class="st">&quot;data/2020_001_Moncada/pdac_ast1/tissue_positions_list_spatial_object.tsv&quot;</span></span></code></pre></div>
<p>Create a Seurat object keeping only ST spots in integrated data <code>pdac_aint</code>.</p>
<div class="sourceCode" id="cb31"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb31-1"><a href="appendix-f-creating-figures.html#cb31-1" aria-hidden="true" tabindex="-1"></a><span class="co"># ----------------------------------------</span></span>
<span id="cb31-2"><a href="appendix-f-creating-figures.html#cb31-2" aria-hidden="true" tabindex="-1"></a><span class="co"># pdac_aint</span></span>
<span id="cb31-3"><a href="appendix-f-creating-figures.html#cb31-3" aria-hidden="true" tabindex="-1"></a><span class="co"># ----------------------------------------</span></span>
<span id="cb31-4"><a href="appendix-f-creating-figures.html#cb31-4" aria-hidden="true" tabindex="-1"></a>tmp <span class="ot">&lt;-</span> pdac_aint[[<span class="st">&quot;sample&quot;</span>]]</span>
<span id="cb31-5"><a href="appendix-f-creating-figures.html#cb31-5" aria-hidden="true" tabindex="-1"></a>tmp <span class="ot">&lt;-</span> tmp[<span class="fu">which</span>(tmp<span class="sc">$</span>orig_ident <span class="sc">==</span> <span class="st">&quot;ST&quot;</span>),]</span>
<span id="cb31-6"><a href="appendix-f-creating-figures.html#cb31-6" aria-hidden="true" tabindex="-1"></a>pdac_aint[[<span class="st">&quot;sample&quot;</span>]] <span class="ot">&lt;-</span> tmp</span>
<span id="cb31-7"><a href="appendix-f-creating-figures.html#cb31-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb31-8"><a href="appendix-f-creating-figures.html#cb31-8" aria-hidden="true" tabindex="-1"></a>mat <span class="ot">&lt;-</span> pdac_aint[[<span class="st">&quot;sign&quot;</span>]][[<span class="st">&quot;CBxSample&quot;</span>]][[<span class="st">&quot;cell&quot;</span>]]</span>
<span id="cb31-9"><a href="appendix-f-creating-figures.html#cb31-9" aria-hidden="true" tabindex="-1"></a>tmp <span class="ot">&lt;-</span> pdac_aint[[<span class="st">&quot;sample&quot;</span>]][[<span class="st">&quot;barcode&quot;</span>]]</span>
<span id="cb31-10"><a href="appendix-f-creating-figures.html#cb31-10" aria-hidden="true" tabindex="-1"></a>mat <span class="ot">&lt;-</span> mat[, <span class="fu">which</span>(<span class="fu">colnames</span>(mat) <span class="sc">%in%</span> tmp)]</span>
<span id="cb31-11"><a href="appendix-f-creating-figures.html#cb31-11" aria-hidden="true" tabindex="-1"></a><span class="fu">colnames</span>(mat) <span class="ot">&lt;-</span> pdac_ast1[[<span class="st">&quot;sample&quot;</span>]][[<span class="st">&quot;spot&quot;</span>]]  <span class="co"># This is required by Spaniel</span></span>
<span id="cb31-12"><a href="appendix-f-creating-figures.html#cb31-12" aria-hidden="true" tabindex="-1"></a><span class="co"># ----------------------------------------</span></span>
<span id="cb31-13"><a href="appendix-f-creating-figures.html#cb31-13" aria-hidden="true" tabindex="-1"></a><span class="co"># se</span></span>
<span id="cb31-14"><a href="appendix-f-creating-figures.html#cb31-14" aria-hidden="true" tabindex="-1"></a><span class="co"># ----------------------------------------</span></span>
<span id="cb31-15"><a href="appendix-f-creating-figures.html#cb31-15" aria-hidden="true" tabindex="-1"></a>se <span class="ot">&lt;-</span> Spaniel<span class="sc">::</span><span class="fu">createSeurat</span>(<span class="at">counts =</span> mat, <span class="at">barcodeFile =</span> fn,</span>
<span id="cb31-16"><a href="appendix-f-creating-figures.html#cb31-16" aria-hidden="true" tabindex="-1"></a>                            <span class="at">projectName =</span> <span class="st">&quot;PDAC-A&quot;</span>, <span class="at">sectionNumber =</span> <span class="st">&quot;1&quot;</span>)</span>
<span id="cb31-17"><a href="appendix-f-creating-figures.html#cb31-17" aria-hidden="true" tabindex="-1"></a>se<span class="sc">@</span>images <span class="ot">&lt;-</span> pdac_aint[[<span class="st">&quot;misc&quot;</span>]][[<span class="st">&quot;images&quot;</span>]]</span></code></pre></div>
<p>Pre-processing.</p>
<div class="sourceCode" id="cb32"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb32-1"><a href="appendix-f-creating-figures.html#cb32-1" aria-hidden="true" tabindex="-1"></a>tmp <span class="ot">&lt;-</span> pdac_aint[[<span class="st">&quot;sample&quot;</span>]][[<span class="st">&quot;seuratFindClusters_CB_cell&quot;</span>]]</span>
<span id="cb32-2"><a href="appendix-f-creating-figures.html#cb32-2" aria-hidden="true" tabindex="-1"></a>se<span class="sc">@</span>meta.data[[<span class="st">&quot;cluster&quot;</span>]] <span class="ot">&lt;-</span> tmp</span>
<span id="cb32-3"><a href="appendix-f-creating-figures.html#cb32-3" aria-hidden="true" tabindex="-1"></a>tmp <span class="ot">&lt;-</span> pdac_aint[[<span class="st">&quot;sample&quot;</span>]]</span>
<span id="cb32-4"><a href="appendix-f-creating-figures.html#cb32-4" aria-hidden="true" tabindex="-1"></a>tmp <span class="ot">&lt;-</span> tmp[<span class="fu">which</span>(tmp<span class="sc">$</span>orig_ident <span class="sc">==</span> <span class="st">&quot;ST&quot;</span>),]</span>
<span id="cb32-5"><a href="appendix-f-creating-figures.html#cb32-5" aria-hidden="true" tabindex="-1"></a>se<span class="sc">@</span>meta.data[[<span class="st">&quot;phase&quot;</span>]] <span class="ot">&lt;-</span> tmp<span class="sc">$</span>phase</span>
<span id="cb32-6"><a href="appendix-f-creating-figures.html#cb32-6" aria-hidden="true" tabindex="-1"></a><span class="fu">SetIdent</span>(se, <span class="at">value =</span> <span class="st">&quot;cluster&quot;</span>)</span>
<span id="cb32-7"><a href="appendix-f-creating-figures.html#cb32-7" aria-hidden="true" tabindex="-1"></a>se <span class="ot">&lt;-</span> <span class="fu">ScaleData</span>(se)</span></code></pre></div>
<p>Visualize putative clusters on histological images.
Note that <code>spanielPlot_2()</code> is custom-build.</p>
<div class="sourceCode" id="cb33"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb33-1"><a href="appendix-f-creating-figures.html#cb33-1" aria-hidden="true" tabindex="-1"></a>p <span class="ot">&lt;-</span> <span class="fu">spanielPlot</span>(<span class="at">object =</span> se, <span class="at">grob =</span> se<span class="sc">@</span>images[[<span class="dv">1</span>]], <span class="at">plotType =</span> <span class="st">&quot;Cluster&quot;</span>,</span>
<span id="cb33-2"><a href="appendix-f-creating-figures.html#cb33-2" aria-hidden="true" tabindex="-1"></a>                 <span class="at">clusterRes =</span> <span class="st">&quot;cluster&quot;</span>, <span class="at">ptSize =</span> <span class="fl">2.5</span>,</span>
<span id="cb33-3"><a href="appendix-f-creating-figures.html#cb33-3" aria-hidden="true" tabindex="-1"></a>                 <span class="at">customTitle =</span> <span class="st">&quot;Cluster (CB: cell)&quot;</span>)</span>
<span id="cb33-4"><a href="appendix-f-creating-figures.html#cb33-4" aria-hidden="true" tabindex="-1"></a>filename <span class="ot">&lt;-</span> <span class="st">&quot;figures/figure_20_0300.png&quot;</span></span>
<span id="cb33-5"><a href="appendix-f-creating-figures.html#cb33-5" aria-hidden="true" tabindex="-1"></a><span class="fu">ggsave</span>(<span class="at">file =</span> filename, <span class="at">plot =</span> p, <span class="at">dpi =</span> <span class="dv">300</span>, <span class="at">width =</span> <span class="fl">5.1</span>, <span class="at">height =</span> <span class="fl">4.2</span>)</span></code></pre></div>
<p><img src="figures/figure_20_0300.png" width="400px"></p>
<p>Visualize cell cycle phases on histological images.
Note that <code>spanielPlot_3()</code> is custom-build, which can only be used for
plotting cell cycle phases.</p>
<div class="sourceCode" id="cb34"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb34-1"><a href="appendix-f-creating-figures.html#cb34-1" aria-hidden="true" tabindex="-1"></a>p <span class="ot">&lt;-</span> <span class="fu">spanielPlot</span>(<span class="at">object =</span> se, <span class="at">grob =</span> se<span class="sc">@</span>images[[<span class="dv">1</span>]], <span class="at">plotType =</span> <span class="st">&quot;Cluster&quot;</span>,</span>
<span id="cb34-2"><a href="appendix-f-creating-figures.html#cb34-2" aria-hidden="true" tabindex="-1"></a>                 <span class="at">clusterRes =</span> <span class="st">&quot;phase&quot;</span>, <span class="at">ptSize =</span> <span class="fl">2.5</span>,</span>
<span id="cb34-3"><a href="appendix-f-creating-figures.html#cb34-3" aria-hidden="true" tabindex="-1"></a>                 <span class="at">customTitle =</span> <span class="st">&quot;Cell cycle phase&quot;</span>)</span>
<span id="cb34-4"><a href="appendix-f-creating-figures.html#cb34-4" aria-hidden="true" tabindex="-1"></a>filename <span class="ot">&lt;-</span> <span class="st">&quot;figures/figure_20_0301.png&quot;</span></span>
<span id="cb34-5"><a href="appendix-f-creating-figures.html#cb34-5" aria-hidden="true" tabindex="-1"></a><span class="fu">ggsave</span>(<span class="at">file =</span> filename, <span class="at">plot =</span> p, <span class="at">dpi =</span> <span class="dv">300</span>, <span class="at">width =</span> <span class="fl">5.1</span>, <span class="at">height =</span> <span class="fl">4.2</span>)</span></code></pre></div>
<p><img src="figures/figure_20_0301.png" width="400px"></p>
<p>Visualize sign scores on histological images.
Note that <code>spanielPlot_2()</code> is custom-build.
Note here that in <code>se</code> hyphens “-” were replaced with underscores "_".</p>
<div class="sourceCode" id="cb35"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb35-1"><a href="appendix-f-creating-figures.html#cb35-1" aria-hidden="true" tabindex="-1"></a>p <span class="ot">&lt;-</span> <span class="fu">spanielPlot</span>(<span class="at">object =</span> se, <span class="at">grob =</span> se<span class="sc">@</span>images[[<span class="dv">1</span>]], <span class="at">plotType =</span> <span class="st">&quot;Gene&quot;</span>,</span>
<span id="cb35-2"><a href="appendix-f-creating-figures.html#cb35-2" aria-hidden="true" tabindex="-1"></a>                 <span class="at">gene =</span> <span class="st">&quot;DOID:3498-S&quot;</span>, <span class="at">ptSizeMax =</span> <span class="fl">2.5</span>,  <span class="at">ptSizeMin =</span> <span class="dv">0</span>,</span>
<span id="cb35-3"><a href="appendix-f-creating-figures.html#cb35-3" aria-hidden="true" tabindex="-1"></a>                 <span class="at">customTitle =</span> <span class="st">&quot;DOID:3498_S&quot;</span>)</span>
<span id="cb35-4"><a href="appendix-f-creating-figures.html#cb35-4" aria-hidden="true" tabindex="-1"></a>filename <span class="ot">&lt;-</span> <span class="st">&quot;figures/figure_20_0310.png&quot;</span></span>
<span id="cb35-5"><a href="appendix-f-creating-figures.html#cb35-5" aria-hidden="true" tabindex="-1"></a><span class="fu">ggsave</span>(<span class="at">file =</span> filename, <span class="at">plot =</span> p, <span class="at">dpi =</span> <span class="dv">300</span>, <span class="at">width =</span> <span class="fl">4.7</span>, <span class="at">height =</span> <span class="fl">4.1</span>)</span></code></pre></div>
<p><img src="figures/figure_20_0310.png" width="400px">
<img src="figures/figure_20_0311.png" width="400px"></p>
<p>Change datatype.</p>
<div class="sourceCode" id="cb36"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb36-1"><a href="appendix-f-creating-figures.html#cb36-1" aria-hidden="true" tabindex="-1"></a>mat <span class="ot">&lt;-</span> pdac_aint[[<span class="st">&quot;sign&quot;</span>]][[<span class="st">&quot;GOxSample&quot;</span>]][[<span class="st">&quot;BP&quot;</span>]]</span>
<span id="cb36-2"><a href="appendix-f-creating-figures.html#cb36-2" aria-hidden="true" tabindex="-1"></a>tmp <span class="ot">&lt;-</span> pdac_aint[[<span class="st">&quot;sample&quot;</span>]][[<span class="st">&quot;barcode&quot;</span>]]</span>
<span id="cb36-3"><a href="appendix-f-creating-figures.html#cb36-3" aria-hidden="true" tabindex="-1"></a>mat <span class="ot">&lt;-</span> mat[, <span class="fu">which</span>(<span class="fu">colnames</span>(mat) <span class="sc">%in%</span> tmp)]</span>
<span id="cb36-4"><a href="appendix-f-creating-figures.html#cb36-4" aria-hidden="true" tabindex="-1"></a><span class="fu">colnames</span>(mat) <span class="ot">&lt;-</span> pdac_ast1[[<span class="st">&quot;sample&quot;</span>]][[<span class="st">&quot;spot&quot;</span>]]  <span class="co"># This is required by Spaniel</span></span>
<span id="cb36-5"><a href="appendix-f-creating-figures.html#cb36-5" aria-hidden="true" tabindex="-1"></a>se <span class="ot">&lt;-</span> Spaniel<span class="sc">::</span><span class="fu">createSeurat</span>(<span class="at">counts =</span> mat, <span class="at">barcodeFile =</span> fn,</span>
<span id="cb36-6"><a href="appendix-f-creating-figures.html#cb36-6" aria-hidden="true" tabindex="-1"></a>                            <span class="at">projectName =</span> <span class="st">&quot;PDAC-A&quot;</span>, <span class="at">sectionNumber =</span> <span class="st">&quot;1&quot;</span>)</span>
<span id="cb36-7"><a href="appendix-f-creating-figures.html#cb36-7" aria-hidden="true" tabindex="-1"></a>se<span class="sc">@</span>images <span class="ot">&lt;-</span> pdac_aint[[<span class="st">&quot;misc&quot;</span>]][[<span class="st">&quot;images&quot;</span>]]</span>
<span id="cb36-8"><a href="appendix-f-creating-figures.html#cb36-8" aria-hidden="true" tabindex="-1"></a>se <span class="ot">&lt;-</span> <span class="fu">ScaleData</span>(se)</span></code></pre></div>
<p>Visualize sign scores on histological images.
Note that <code>spanielPlot_2()</code> is custom-build.
Note here that in <code>se</code> hyphens “-” were replaced with underscores "_".</p>
<div class="sourceCode" id="cb37"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb37-1"><a href="appendix-f-creating-figures.html#cb37-1" aria-hidden="true" tabindex="-1"></a>p <span class="ot">&lt;-</span> <span class="fu">spanielPlot</span>(<span class="at">object =</span> se, <span class="at">grob =</span> se<span class="sc">@</span>images[[<span class="dv">1</span>]], <span class="at">plotType =</span> <span class="st">&quot;Gene&quot;</span>,</span>
<span id="cb37-2"><a href="appendix-f-creating-figures.html#cb37-2" aria-hidden="true" tabindex="-1"></a>                 <span class="at">gene =</span> <span class="st">&quot;GO:0030857-S&quot;</span>, <span class="at">ptSizeMax =</span> <span class="fl">2.5</span>,  <span class="at">ptSizeMin =</span> <span class="dv">0</span>,</span>
<span id="cb37-3"><a href="appendix-f-creating-figures.html#cb37-3" aria-hidden="true" tabindex="-1"></a>                 <span class="at">customTitle =</span> <span class="st">&quot;GO:0030857_S&quot;</span>)</span>
<span id="cb37-4"><a href="appendix-f-creating-figures.html#cb37-4" aria-hidden="true" tabindex="-1"></a>filename <span class="ot">&lt;-</span> <span class="st">&quot;figures/figure_20_0312.png&quot;</span></span>
<span id="cb37-5"><a href="appendix-f-creating-figures.html#cb37-5" aria-hidden="true" tabindex="-1"></a><span class="fu">ggsave</span>(<span class="at">file =</span> filename, <span class="at">plot =</span> p, <span class="at">dpi =</span> <span class="dv">300</span>, <span class="at">width =</span> <span class="fl">4.7</span>, <span class="at">height =</span> <span class="fl">4.1</span>)</span></code></pre></div>
<p><img src="figures/figure_20_0312.png" width="400px"></p>
<p>Change datatype.</p>
<div class="sourceCode" id="cb38"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb38-1"><a href="appendix-f-creating-figures.html#cb38-1" aria-hidden="true" tabindex="-1"></a>mat <span class="ot">&lt;-</span> pdac_aint[[<span class="st">&quot;sign&quot;</span>]][[<span class="st">&quot;KEGGxSample&quot;</span>]][[<span class="st">&quot;pathway&quot;</span>]]</span>
<span id="cb38-2"><a href="appendix-f-creating-figures.html#cb38-2" aria-hidden="true" tabindex="-1"></a>tmp <span class="ot">&lt;-</span> pdac_aint[[<span class="st">&quot;sample&quot;</span>]][[<span class="st">&quot;barcode&quot;</span>]]</span>
<span id="cb38-3"><a href="appendix-f-creating-figures.html#cb38-3" aria-hidden="true" tabindex="-1"></a>mat <span class="ot">&lt;-</span> mat[, <span class="fu">which</span>(<span class="fu">colnames</span>(mat) <span class="sc">%in%</span> tmp)]</span>
<span id="cb38-4"><a href="appendix-f-creating-figures.html#cb38-4" aria-hidden="true" tabindex="-1"></a><span class="fu">colnames</span>(mat) <span class="ot">&lt;-</span> pdac_ast1[[<span class="st">&quot;sample&quot;</span>]][[<span class="st">&quot;spot&quot;</span>]]  <span class="co"># This is required by Spaniel</span></span>
<span id="cb38-5"><a href="appendix-f-creating-figures.html#cb38-5" aria-hidden="true" tabindex="-1"></a>se <span class="ot">&lt;-</span> Spaniel<span class="sc">::</span><span class="fu">createSeurat</span>(<span class="at">counts =</span> mat, <span class="at">barcodeFile =</span> fn,</span>
<span id="cb38-6"><a href="appendix-f-creating-figures.html#cb38-6" aria-hidden="true" tabindex="-1"></a>                            <span class="at">projectName =</span> <span class="st">&quot;PDAC-A&quot;</span>, <span class="at">sectionNumber =</span> <span class="st">&quot;1&quot;</span>)</span>
<span id="cb38-7"><a href="appendix-f-creating-figures.html#cb38-7" aria-hidden="true" tabindex="-1"></a>se<span class="sc">@</span>images <span class="ot">&lt;-</span> pdac_aint[[<span class="st">&quot;misc&quot;</span>]][[<span class="st">&quot;images&quot;</span>]]</span>
<span id="cb38-8"><a href="appendix-f-creating-figures.html#cb38-8" aria-hidden="true" tabindex="-1"></a>se <span class="ot">&lt;-</span> <span class="fu">ScaleData</span>(se)</span></code></pre></div>
<p>Visualize sign scores on histological images.
Note that <code>spanielPlot_2()</code> is custom-build.
Note here that in <code>se</code> hyphens “-” were replaced with underscores "_".</p>
<div class="sourceCode" id="cb39"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb39-1"><a href="appendix-f-creating-figures.html#cb39-1" aria-hidden="true" tabindex="-1"></a>p <span class="ot">&lt;-</span> <span class="fu">spanielPlot</span>(<span class="at">object =</span> se, <span class="at">grob =</span> se<span class="sc">@</span>images[[<span class="dv">1</span>]], <span class="at">plotType =</span> <span class="st">&quot;Gene&quot;</span>,</span>
<span id="cb39-2"><a href="appendix-f-creating-figures.html#cb39-2" aria-hidden="true" tabindex="-1"></a>                 <span class="at">gene =</span> <span class="st">&quot;path:hsa04972-S&quot;</span>, <span class="at">ptSizeMax =</span> <span class="fl">2.5</span>,  <span class="at">ptSizeMin =</span> <span class="dv">0</span>,</span>
<span id="cb39-3"><a href="appendix-f-creating-figures.html#cb39-3" aria-hidden="true" tabindex="-1"></a>                 <span class="at">customTitle =</span> <span class="st">&quot;KEGG:hsa04972_S&quot;</span>)</span>
<span id="cb39-4"><a href="appendix-f-creating-figures.html#cb39-4" aria-hidden="true" tabindex="-1"></a>filename <span class="ot">&lt;-</span> <span class="st">&quot;figures/figure_20_0313.png&quot;</span></span>
<span id="cb39-5"><a href="appendix-f-creating-figures.html#cb39-5" aria-hidden="true" tabindex="-1"></a><span class="fu">ggsave</span>(<span class="at">file =</span> filename, <span class="at">plot =</span> p, <span class="at">dpi =</span> <span class="dv">300</span>, <span class="at">width =</span> <span class="fl">4.63</span>, <span class="at">height =</span> <span class="fl">4.1</span>)</span></code></pre></div>
<p><img src="figures/figure_20_0313.png" width="400px">
<img src="figures/figure_20_0314.png" width="400px"></p>

</div>
</div>
</div>
            </section>

          </div>
        </div>
      </div>
<a href="appendix-e-analyzing-scrna-seq-data-using-asurat.html" class="navigation navigation-prev navigation-unique" aria-label="Previous page"><i class="fa fa-angle-left"></i></a>

    </div>
  </div>
<script src="libs/gitbook-2.6.7/js/app.min.js"></script>
<script src="libs/gitbook-2.6.7/js/lunr.js"></script>
<script src="libs/gitbook-2.6.7/js/clipboard.min.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-search.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-sharing.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-fontsettings.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-bookdown.js"></script>
<script src="libs/gitbook-2.6.7/js/jquery.highlight.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-clipboard.js"></script>
<script>
gitbook.require(["gitbook"], function(gitbook) {
gitbook.start({
"sharing": {
"github": false,
"facebook": true,
"twitter": true,
"linkedin": false,
"weibo": false,
"instapaper": false,
"vk": false,
"whatsapp": false,
"all": ["facebook", "twitter", "linkedin", "weibo", "instapaper"]
},
"fontsettings": {
"theme": "white",
"family": "sans",
"size": 2
},
"edit": {
"link": null,
"text": null
},
"history": {
"link": null,
"text": null
},
"view": {
"link": null,
"text": null
},
"download": ["Introduction to ASURAT (ver. 0.0.0.9000).pdf", "Introduction to ASURAT (ver. 0.0.0.9000).epub"],
"search": {
"engine": "lunr",
"options": null
},
"toc": {
"collapse": "subsection"
}
});
});
</script>

</body>

</html>
