<!DOCTYPE html>
<html lang="" xml:lang="">
<head>

  <meta charset="utf-8" />
  <meta http-equiv="X-UA-Compatible" content="IE=edge" />
  <title>Chapter 7 Collecting databases (optional) | Introduction to ASURAT (version 0.0.0.9000)</title>
  <meta name="description" content="This is a developmental version of ASURAT." />
  <meta name="generator" content="bookdown 0.21 and GitBook 2.6.7" />

  <meta property="og:title" content="Chapter 7 Collecting databases (optional) | Introduction to ASURAT (version 0.0.0.9000)" />
  <meta property="og:type" content="book" />
  
  
  <meta property="og:description" content="This is a developmental version of ASURAT." />
  

  <meta name="twitter:card" content="summary" />
  <meta name="twitter:title" content="Chapter 7 Collecting databases (optional) | Introduction to ASURAT (version 0.0.0.9000)" />
  
  <meta name="twitter:description" content="This is a developmental version of ASURAT." />
  

<meta name="author" content="Keita Iida" />


<meta name="date" content="2021-06-09" />

  <meta name="viewport" content="width=device-width, initial-scale=1" />
  <meta name="apple-mobile-web-app-capable" content="yes" />
  <meta name="apple-mobile-web-app-status-bar-style" content="black" />
  
  
<link rel="prev" href="markergenes.html"/>
<link rel="next" href="asurat-using-disease-ontology-database.html"/>
<script src="libs/jquery-2.2.3/jquery.min.js"></script>
<link href="libs/gitbook-2.6.7/css/style.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-table.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-bookdown.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-highlight.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-search.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-fontsettings.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-clipboard.css" rel="stylesheet" />









<script src="libs/accessible-code-block-0.0.1/empty-anchor.js"></script>


<style type="text/css">
code.sourceCode > span { display: inline-block; line-height: 1.25; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode { white-space: pre; position: relative; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
code.sourceCode { white-space: pre-wrap; }
code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } /* Alert */
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code span.at { color: #7d9029; } /* Attribute */
code span.bn { color: #40a070; } /* BaseN */
code span.bu { } /* BuiltIn */
code span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code span.ch { color: #4070a0; } /* Char */
code span.cn { color: #880000; } /* Constant */
code span.co { color: #60a0b0; font-style: italic; } /* Comment */
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code span.do { color: #ba2121; font-style: italic; } /* Documentation */
code span.dt { color: #902000; } /* DataType */
code span.dv { color: #40a070; } /* DecVal */
code span.er { color: #ff0000; font-weight: bold; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #40a070; } /* Float */
code span.fu { color: #06287e; } /* Function */
code span.im { } /* Import */
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
code span.kw { color: #007020; font-weight: bold; } /* Keyword */
code span.op { color: #666666; } /* Operator */
code span.ot { color: #007020; } /* Other */
code span.pp { color: #bc7a00; } /* Preprocessor */
code span.sc { color: #4070a0; } /* SpecialChar */
code span.ss { color: #bb6688; } /* SpecialString */
code span.st { color: #4070a0; } /* String */
code span.va { color: #19177c; } /* Variable */
code span.vs { color: #4070a0; } /* VerbatimString */
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */
</style>

<link rel="stylesheet" href="style.css" type="text/css" />
</head>

<body>



  <div class="book without-animation with-summary font-size-2 font-family-1" data-basepath=".">

    <div class="book-summary">
      <nav role="navigation">

<ul class="summary">
<li><a href="./">A Minimal Book Example</a></li>

<li class="divider"></li>
<li class="chapter" data-level="1" data-path="index.html"><a href="index.html"><i class="fa fa-check"></i><b>1</b> Overview</a><ul>
<li class="chapter" data-level="1.1" data-path="index.html"><a href="index.html#license-and-description"><i class="fa fa-check"></i><b>1.1</b> License and description</a></li>
<li class="chapter" data-level="1.2" data-path="index.html"><a href="index.html#computational-environment"><i class="fa fa-check"></i><b>1.2</b> Computational environment</a></li>
<li class="chapter" data-level="1.3" data-path="index.html"><a href="index.html#installation"><i class="fa fa-check"></i><b>1.3</b> Installation</a></li>
</ul></li>
<li class="chapter" data-level="2" data-path="dataset.html"><a href="dataset.html"><i class="fa fa-check"></i><b>2</b> Preparing data sets</a><ul>
<li class="chapter" data-level="2.1" data-path="dataset.html"><a href="dataset.html#case-i-input-gene-expression-matrix-with-row-and-column-names"><i class="fa fa-check"></i><b>2.1</b> Case I: input gene expression matrix with row and column names</a></li>
<li class="chapter" data-level="2.2" data-path="dataset.html"><a href="dataset.html#case-ii-input-10x-data"><i class="fa fa-check"></i><b>2.2</b> Case II: input 10x data</a></li>
<li class="chapter" data-level="2.3" data-path="dataset.html"><a href="dataset.html#case-iii-input-10x-pbmc-data"><i class="fa fa-check"></i><b>2.3</b> Case III: input 10x PBMC data</a></li>
</ul></li>
<li class="chapter" data-level="3" data-path="dataqc.html"><a href="dataqc.html"><i class="fa fa-check"></i><b>3</b> Data quality control (QC)</a><ul>
<li class="chapter" data-level="3.1" data-path="dataqc.html"><a href="dataqc.html#quick-qc"><i class="fa fa-check"></i><b>3.1</b> Quick QC</a></li>
<li class="chapter" data-level="3.2" data-path="dataqc.html"><a href="dataqc.html#sample-qc"><i class="fa fa-check"></i><b>3.2</b> Sample QC</a></li>
<li class="chapter" data-level="3.3" data-path="dataqc.html"><a href="dataqc.html#variable-qc"><i class="fa fa-check"></i><b>3.3</b> Variable QC</a></li>
</ul></li>
<li class="chapter" data-level="4" data-path="normalization.html"><a href="normalization.html"><i class="fa fa-check"></i><b>4</b> Normalizing and centering data</a></li>
<li class="chapter" data-level="5" data-path="computcorr.html"><a href="computcorr.html"><i class="fa fa-check"></i><b>5</b> Computing correlations among genes</a></li>
<li class="chapter" data-level="6" data-path="markergenes.html"><a href="markergenes.html"><i class="fa fa-check"></i><b>6</b> Checking expression profiles of marker genes</a></li>
<li class="chapter" data-level="7" data-path="collectdb.html"><a href="collectdb.html"><i class="fa fa-check"></i><b>7</b> Collecting databases (optional)</a><ul>
<li class="chapter" data-level="7.1" data-path="collectdb.html"><a href="collectdb.html#disease-ontology-db"><i class="fa fa-check"></i><b>7.1</b> Disease Ontology DB</a></li>
<li class="chapter" data-level="7.2" data-path="collectdb.html"><a href="collectdb.html#cell-ontology-db"><i class="fa fa-check"></i><b>7.2</b> Cell Ontology DB</a></li>
<li class="chapter" data-level="7.3" data-path="collectdb.html"><a href="collectdb.html#gene-ontology-db"><i class="fa fa-check"></i><b>7.3</b> Gene Ontology DB</a></li>
<li class="chapter" data-level="7.4" data-path="collectdb.html"><a href="collectdb.html#kegg"><i class="fa fa-check"></i><b>7.4</b> KEGG</a></li>
<li class="chapter" data-level="7.5" data-path="collectdb.html"><a href="collectdb.html#reactome"><i class="fa fa-check"></i><b>7.5</b> Reactome</a></li>
</ul></li>
<li class="chapter" data-level="8" data-path="asurat-using-disease-ontology-database.html"><a href="asurat-using-disease-ontology-database.html"><i class="fa fa-check"></i><b>8</b> ASURAT using Disease Ontology database</a><ul>
<li class="chapter" data-level="8.1" data-path="asurat-using-disease-ontology-database.html"><a href="asurat-using-disease-ontology-database.html#creating-sign-by-sample-matrices"><i class="fa fa-check"></i><b>8.1</b> Creating sign-by-sample matrices</a></li>
<li class="chapter" data-level="8.2" data-path="asurat-using-disease-ontology-database.html"><a href="asurat-using-disease-ontology-database.html#dimensionality-reduction"><i class="fa fa-check"></i><b>8.2</b> Dimensionality reduction</a></li>
<li class="chapter" data-level="8.3" data-path="asurat-using-disease-ontology-database.html"><a href="asurat-using-disease-ontology-database.html#cell-clustering"><i class="fa fa-check"></i><b>8.3</b> Cell clustering</a></li>
<li class="chapter" data-level="8.4" data-path="asurat-using-disease-ontology-database.html"><a href="asurat-using-disease-ontology-database.html#visualization"><i class="fa fa-check"></i><b>8.4</b> Visualization</a></li>
<li class="chapter" data-level="8.5" data-path="asurat-using-disease-ontology-database.html"><a href="asurat-using-disease-ontology-database.html#significant-signs"><i class="fa fa-check"></i><b>8.5</b> Significant signs</a></li>
</ul></li>
<li class="chapter" data-level="9" data-path="asurat-using-cell-ontology-database-optional.html"><a href="asurat-using-cell-ontology-database-optional.html"><i class="fa fa-check"></i><b>9</b> ASURAT using Cell Ontology database (optional)</a><ul>
<li class="chapter" data-level="9.1" data-path="asurat-using-cell-ontology-database-optional.html"><a href="asurat-using-cell-ontology-database-optional.html#creating-sign-by-sample-matrices-1"><i class="fa fa-check"></i><b>9.1</b> Creating sign-by-sample matrices</a></li>
<li class="chapter" data-level="9.2" data-path="asurat-using-cell-ontology-database-optional.html"><a href="asurat-using-cell-ontology-database-optional.html#dimensionality-reduction-1"><i class="fa fa-check"></i><b>9.2</b> Dimensionality reduction</a></li>
<li class="chapter" data-level="9.3" data-path="asurat-using-cell-ontology-database-optional.html"><a href="asurat-using-cell-ontology-database-optional.html#cell-clustering-1"><i class="fa fa-check"></i><b>9.3</b> Cell clustering</a></li>
<li class="chapter" data-level="9.4" data-path="asurat-using-cell-ontology-database-optional.html"><a href="asurat-using-cell-ontology-database-optional.html#visualization-1"><i class="fa fa-check"></i><b>9.4</b> Visualization</a></li>
<li class="chapter" data-level="9.5" data-path="asurat-using-cell-ontology-database-optional.html"><a href="asurat-using-cell-ontology-database-optional.html#significant-signs-1"><i class="fa fa-check"></i><b>9.5</b> Significant signs</a></li>
</ul></li>
<li class="chapter" data-level="10" data-path="asurat-using-gene-ontology-database-optional.html"><a href="asurat-using-gene-ontology-database-optional.html"><i class="fa fa-check"></i><b>10</b> ASURAT using Gene Ontology database (optional)</a><ul>
<li class="chapter" data-level="10.1" data-path="asurat-using-gene-ontology-database-optional.html"><a href="asurat-using-gene-ontology-database-optional.html#creating-sign-by-sample-matrices-2"><i class="fa fa-check"></i><b>10.1</b> Creating sign-by-sample matrices</a></li>
<li class="chapter" data-level="10.2" data-path="asurat-using-gene-ontology-database-optional.html"><a href="asurat-using-gene-ontology-database-optional.html#dimensionality-reduction-2"><i class="fa fa-check"></i><b>10.2</b> Dimensionality reduction</a></li>
<li class="chapter" data-level="10.3" data-path="asurat-using-gene-ontology-database-optional.html"><a href="asurat-using-gene-ontology-database-optional.html#cell-clustering-2"><i class="fa fa-check"></i><b>10.3</b> Cell clustering</a></li>
<li class="chapter" data-level="10.4" data-path="asurat-using-gene-ontology-database-optional.html"><a href="asurat-using-gene-ontology-database-optional.html#visualization-2"><i class="fa fa-check"></i><b>10.4</b> Visualization</a></li>
<li class="chapter" data-level="10.5" data-path="asurat-using-gene-ontology-database-optional.html"><a href="asurat-using-gene-ontology-database-optional.html#significant-signs-2"><i class="fa fa-check"></i><b>10.5</b> Significant signs</a></li>
</ul></li>
<li class="chapter" data-level="11" data-path="asurat-using-kegg-optional.html"><a href="asurat-using-kegg-optional.html"><i class="fa fa-check"></i><b>11</b> ASURAT using KEGG (optional)</a><ul>
<li class="chapter" data-level="11.1" data-path="asurat-using-kegg-optional.html"><a href="asurat-using-kegg-optional.html#creating-sign-by-sample-matrices-3"><i class="fa fa-check"></i><b>11.1</b> Creating sign-by-sample matrices</a></li>
<li class="chapter" data-level="11.2" data-path="asurat-using-kegg-optional.html"><a href="asurat-using-kegg-optional.html#dimensionality-reduction-3"><i class="fa fa-check"></i><b>11.2</b> Dimensionality reduction</a></li>
<li class="chapter" data-level="11.3" data-path="asurat-using-kegg-optional.html"><a href="asurat-using-kegg-optional.html#cell-clustering-3"><i class="fa fa-check"></i><b>11.3</b> Cell clustering</a></li>
<li class="chapter" data-level="11.4" data-path="asurat-using-kegg-optional.html"><a href="asurat-using-kegg-optional.html#visualizing"><i class="fa fa-check"></i><b>11.4</b> Visualizing</a></li>
<li class="chapter" data-level="11.5" data-path="asurat-using-kegg-optional.html"><a href="asurat-using-kegg-optional.html#significant-signs-3"><i class="fa fa-check"></i><b>11.5</b> Significant signs</a></li>
</ul></li>
<li class="chapter" data-level="12" data-path="asurat-using-reactome-optional.html"><a href="asurat-using-reactome-optional.html"><i class="fa fa-check"></i><b>12</b> ASURAT using Reactome (optional)</a><ul>
<li class="chapter" data-level="12.1" data-path="asurat-using-reactome-optional.html"><a href="asurat-using-reactome-optional.html#creating-sign-by-sample-matrices-4"><i class="fa fa-check"></i><b>12.1</b> Creating sign-by-sample matrices</a></li>
<li class="chapter" data-level="12.2" data-path="asurat-using-reactome-optional.html"><a href="asurat-using-reactome-optional.html#dimensionality-reduction-4"><i class="fa fa-check"></i><b>12.2</b> Dimensionality reduction</a></li>
<li class="chapter" data-level="12.3" data-path="asurat-using-reactome-optional.html"><a href="asurat-using-reactome-optional.html#sample-clustering"><i class="fa fa-check"></i><b>12.3</b> Sample clustering</a></li>
</ul></li>
<li class="chapter" data-level="13" data-path="multiple-sign-analysis-by-concatenating-do-co-go-kegg-and-reactome.html"><a href="multiple-sign-analysis-by-concatenating-do-co-go-kegg-and-reactome.html"><i class="fa fa-check"></i><b>13</b> Multiple sign analysis by concatenating DO, CO, GO, KEGG, and Reactome</a></li>
<li class="chapter" data-level="14" data-path="appendix-a-comparing-performances-of-asurat-and-existing-methods.html"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html"><i class="fa fa-check"></i><b>14</b> Appendix A: comparing performances of ASURAT and existing methods</a><ul>
<li class="chapter" data-level="14.1" data-path="appendix-a-comparing-performances-of-asurat-and-existing-methods.html"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#installation-1"><i class="fa fa-check"></i><b>14.1</b> Installation</a></li>
<li class="chapter" data-level="14.2" data-path="appendix-a-comparing-performances-of-asurat-and-existing-methods.html"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#removing-the-outlier-cells-detected-by-asurat"><i class="fa fa-check"></i><b>14.2</b> Removing the outlier cells detected by ASURAT</a></li>
<li class="chapter" data-level="14.3" data-path="appendix-a-comparing-performances-of-asurat-and-existing-methods.html"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#assessing-clustering-performances"><i class="fa fa-check"></i><b>14.3</b> Assessing clustering performances</a><ul>
<li class="chapter" data-level="14.3.1" data-path="appendix-a-comparing-performances-of-asurat-and-existing-methods.html"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#seurat"><i class="fa fa-check"></i><b>14.3.1</b> Seurat</a></li>
<li class="chapter" data-level="14.3.2" data-path="appendix-a-comparing-performances-of-asurat-and-existing-methods.html"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#monocle-3"><i class="fa fa-check"></i><b>14.3.2</b> Monocle 3</a></li>
<li class="chapter" data-level="14.3.3" data-path="appendix-a-comparing-performances-of-asurat-and-existing-methods.html"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#sc3"><i class="fa fa-check"></i><b>14.3.3</b> SC3</a></li>
<li class="chapter" data-level="14.3.4" data-path="appendix-a-comparing-performances-of-asurat-and-existing-methods.html"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#principal-component-analysis-pca"><i class="fa fa-check"></i><b>14.3.4</b> Principal component analysis (PCA)</a></li>
<li class="chapter" data-level="14.3.5" data-path="appendix-a-comparing-performances-of-asurat-and-existing-methods.html"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#asurat"><i class="fa fa-check"></i><b>14.3.5</b> ASURAT</a></li>
<li class="chapter" data-level="14.3.6" data-path="appendix-a-comparing-performances-of-asurat-and-existing-methods.html"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#comparing-clustering-validity-indices"><i class="fa fa-check"></i><b>14.3.6</b> Comparing clustering validity indices</a></li>
</ul></li>
<li class="chapter" data-level="14.4" data-path="appendix-a-comparing-performances-of-asurat-and-existing-methods.html"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cell-type-inference-in-pbmc_4000-and-pbmc_6000"><i class="fa fa-check"></i><b>14.4</b> Cell-type inference in <code>pbmc_4000</code> and <code>pbmc_6000</code></a><ul>
<li class="chapter" data-level="14.4.1" data-path="appendix-a-comparing-performances-of-asurat-and-existing-methods.html"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#seurat-1"><i class="fa fa-check"></i><b>14.4.1</b> Seurat</a></li>
<li class="chapter" data-level="14.4.2" data-path="appendix-a-comparing-performances-of-asurat-and-existing-methods.html"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#monocle-3-1"><i class="fa fa-check"></i><b>14.4.2</b> Monocle 3</a></li>
<li class="chapter" data-level="14.4.3" data-path="appendix-a-comparing-performances-of-asurat-and-existing-methods.html"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#sc3-1"><i class="fa fa-check"></i><b>14.4.3</b> SC3</a></li>
<li class="chapter" data-level="14.4.4" data-path="appendix-a-comparing-performances-of-asurat-and-existing-methods.html"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#asurat-1"><i class="fa fa-check"></i><b>14.4.4</b> ASURAT</a></li>
</ul></li>
<li class="chapter" data-level="14.5" data-path="appendix-a-comparing-performances-of-asurat-and-existing-methods.html"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#identifying-cell-types-in-pbmc_4000-and-pbmc_6000"><i class="fa fa-check"></i><b>14.5</b> Identifying cell types in <code>pbmc_4000</code> and <code>pbmc_6000</code></a></li>
</ul></li>
<li class="chapter" data-level="15" data-path="appendix-b-automatically-tuning-asurats-parameters.html"><a href="appendix-b-automatically-tuning-asurats-parameters.html"><i class="fa fa-check"></i><b>15</b> Appendix B: automatically tuning ASURAT’s parameters</a></li>
<li class="divider"></li>
<li><a href="https://github.com/rstudio/bookdown" target="blank">Published with bookdown</a></li>

</ul>

      </nav>
    </div>

    <div class="book-body">
      <div class="body-inner">
        <div class="book-header" role="navigation">
          <h1>
            <i class="fa fa-circle-o-notch fa-spin"></i><a href="./">Introduction to ASURAT (version 0.0.0.9000)</a>
          </h1>
        </div>

        <div class="page-wrapper" tabindex="-1" role="main">
          <div class="page-inner">

            <section class="normal" id="section-">
<div id="collectdb" class="section level1">
<h1><span class="header-section-number">Chapter 7</span> Collecting databases (optional)</h1>
<p>In this section, the following databases (DBs) are downloaded and properly
formatted.
Since all of the results in this section have already been prepared for human
and mouse datasets in December 2020 (see <code>data/2020_001_databases/</code>),
users may be able to skip this section.</p>
<ol style="list-style-type: decimal">
<li>DB for Disease Ontology (DO) for human disease</li>
<li>DB for Cell Ontology (CO)</li>
<li>DB for Gene Ontology (GO)</li>
<li>DB for Kyoto Encyclopedia of Genes and Genomes (KEGG)</li>
<li>DB for Reactome</li>
</ol>
<div id="disease-ontology-db" class="section level2">
<h2><span class="header-section-number">7.1</span> Disease Ontology DB</h2>
<div class="sourceCode" id="cb1"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb1-1"><a href="collectdb.html#cb1-1"></a><span class="kw">source</span>(<span class="st">&quot;R/function_DiseaseOntology.R&quot;</span>)</span></code></pre></div>
<p>The following functions <code>collect_DO()</code>, <code>format_DO()</code>, and <code>compute_IC_DO()</code>
(i) downloads a database for Disease Ontology,
(ii) formats the results, and
(iii) prepares a list of information content (IC) (Mistry and Pavlidis,
BMC Bioinformatics, 2008), respectively.
IC value represents the depth of the biological term in the ontology,
which will be later used for computing similarities between terms.</p>
<div class="sourceCode" id="cb2"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb2-1"><a href="collectdb.html#cb2-1"></a><span class="co">#</span></span>
<span id="cb2-2"><a href="collectdb.html#cb2-2"></a><span class="co"># Note again that the following result has already been prepared for human and</span></span>
<span id="cb2-3"><a href="collectdb.html#cb2-3"></a><span class="co"># mouse datasets in December 2020 (see `data/2020_001_databases/`).</span></span>
<span id="cb2-4"><a href="collectdb.html#cb2-4"></a><span class="co">#</span></span>
<span id="cb2-5"><a href="collectdb.html#cb2-5"></a><span class="kw">library</span>(DO.db)</span>
<span id="cb2-6"><a href="collectdb.html#cb2-6"></a>dict_DO &lt;-<span class="st"> </span><span class="kw">collect_DO</span>()</span>
<span id="cb2-7"><a href="collectdb.html#cb2-7"></a>tidy_DO &lt;-<span class="st"> </span><span class="kw">format_DO</span>(<span class="dt">dict =</span> dict_DO)</span>
<span id="cb2-8"><a href="collectdb.html#cb2-8"></a>IC_DO &lt;-<span class="st"> </span><span class="kw">compute_IC_DO</span>(<span class="dt">tidy =</span> tidy_DO)</span>
<span id="cb2-9"><a href="collectdb.html#cb2-9"></a><span class="co">#</span></span>
<span id="cb2-10"><a href="collectdb.html#cb2-10"></a><span class="co"># Save the result.</span></span>
<span id="cb2-11"><a href="collectdb.html#cb2-11"></a><span class="co">#</span></span>
<span id="cb2-12"><a href="collectdb.html#cb2-12"></a><span class="kw">saveRDS</span>(dict_DO, <span class="dt">file =</span> <span class="st">&quot;data/2020_001_databases/YYYYMMDD_dict_DO_human.rds&quot;</span>)</span>
<span id="cb2-13"><a href="collectdb.html#cb2-13"></a><span class="kw">saveRDS</span>(tidy_DO, <span class="dt">file =</span> <span class="st">&quot;data/2020_001_databases/YYYYMMDD_tidy_DO_human.rds&quot;</span>)</span>
<span id="cb2-14"><a href="collectdb.html#cb2-14"></a><span class="kw">saveRDS</span>(IC_DO, <span class="dt">file =</span> <span class="st">&quot;data/2020_001_databases/YYYYMMDD_IC_DO_human.rds&quot;</span>)</span></code></pre></div>
</div>
<div id="cell-ontology-db" class="section level2">
<h2><span class="header-section-number">7.2</span> Cell Ontology DB</h2>
<div class="sourceCode" id="cb3"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb3-1"><a href="collectdb.html#cb3-1"></a><span class="kw">source</span>(<span class="st">&quot;R/function_CellOntology.R&quot;</span>)</span></code></pre></div>
<p>The following functions <code>collect_CO()</code>, <code>format_CO()</code>, <code>make_treeTable_CO()</code>,
and <code>compute_IC_CO()</code>
(i) downloads a database for Disease Ontology,
(ii) formats the results, and
(iii) prepares a list of information content (IC) (Mistry and Pavlidis,
BMC Bioinformatics, 2008), respectively.</p>
<p><strong>Tips:</strong>
As of December 2020, Cell Ontology database might not be complete enough
for some biological contexts.
For example, the well-known marker genes for pancreatic beta cell,
Ins1 and Ins2, are not registered in the gene slot of <code>type B pancreatic cell</code>
with CO ID <code>CL:0000169</code>.
Be careful of using IC values of CO, which may not be always useful for
computing similarities between biological terms.
Nevertheless, we computed the IC values as in the future CO database might be
better annotated.</p>
<div class="sourceCode" id="cb4"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb4-1"><a href="collectdb.html#cb4-1"></a><span class="co">#</span></span>
<span id="cb4-2"><a href="collectdb.html#cb4-2"></a><span class="co"># Note again that the following result has already been prepared for human and</span></span>
<span id="cb4-3"><a href="collectdb.html#cb4-3"></a><span class="co"># mouse datasets in December 2020 (see `data/2020_001_databases/`).</span></span>
<span id="cb4-4"><a href="collectdb.html#cb4-4"></a><span class="co">#</span></span>
<span id="cb4-5"><a href="collectdb.html#cb4-5"></a>dict_CO &lt;-<span class="st"> </span><span class="kw">collect_CO</span>(<span class="dt">orgdb =</span> org.Hs.eg.db)</span>
<span id="cb4-6"><a href="collectdb.html#cb4-6"></a>tidy_CO &lt;-<span class="st"> </span><span class="kw">format_CO</span>(<span class="dt">dict =</span> dict_CO)</span>
<span id="cb4-7"><a href="collectdb.html#cb4-7"></a>treeTable_CO &lt;-<span class="st"> </span><span class="kw">make_treeTable_CO</span>(<span class="dt">tidy =</span> tidy_CO)</span>
<span id="cb4-8"><a href="collectdb.html#cb4-8"></a>IC_CO &lt;-<span class="st"> </span><span class="kw">compute_IC_CO</span>(<span class="dt">tidy =</span> tidy_CO, <span class="dt">treeTable =</span> treeTable_CO)</span>
<span id="cb4-9"><a href="collectdb.html#cb4-9"></a><span class="co">#</span></span>
<span id="cb4-10"><a href="collectdb.html#cb4-10"></a><span class="co"># Save the result.</span></span>
<span id="cb4-11"><a href="collectdb.html#cb4-11"></a><span class="co">#</span></span>
<span id="cb4-12"><a href="collectdb.html#cb4-12"></a><span class="kw">saveRDS</span>(dict_CO, <span class="dt">file =</span> <span class="st">&quot;data/2020_001_databases//YYYYMMDD_dict_CO_human.rds&quot;</span>)</span>
<span id="cb4-13"><a href="collectdb.html#cb4-13"></a><span class="kw">saveRDS</span>(tidy_CO, <span class="dt">file =</span> <span class="st">&quot;data/2020_001_databases//YYYYMMDD_tidy_CO_human.rds&quot;</span>)</span>
<span id="cb4-14"><a href="collectdb.html#cb4-14"></a><span class="kw">saveRDS</span>(treeTable_CO,</span>
<span id="cb4-15"><a href="collectdb.html#cb4-15"></a>        <span class="dt">file =</span> <span class="st">&quot;data/2020_001_databases//YYYYMMDD_treeTable_CO_human.rds&quot;</span>)</span>
<span id="cb4-16"><a href="collectdb.html#cb4-16"></a><span class="kw">saveRDS</span>(IC_CO, <span class="dt">file =</span> <span class="st">&quot;data/2020_001_databases//YYYYMMDD_IC_CO_human.rds&quot;</span>)</span></code></pre></div>
</div>
<div id="gene-ontology-db" class="section level2">
<h2><span class="header-section-number">7.3</span> Gene Ontology DB</h2>
<div class="sourceCode" id="cb5"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb5-1"><a href="collectdb.html#cb5-1"></a><span class="kw">source</span>(<span class="st">&quot;R/function_GeneOntology.R&quot;</span>)</span></code></pre></div>
<p>The following functions <code>collect_GO()</code>, <code>format_GO()</code>, and <code>compute_IC_GO()</code>
(i) downloads a database for Gene Ontology,
(ii) formats the results, and
(iii) prepares a list of information content (IC), respectively.
The arguments of <code>collect_GO()</code> are <code>orgdb</code>
(annotation data package such as <code>org.Hs.eg.db</code> and <code>org.Mm.eg.db</code>) and
<code>all_geneIDs</code> (all Entrez gene IDs of a specific organism).</p>
<div class="sourceCode" id="cb6"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb6-1"><a href="collectdb.html#cb6-1"></a><span class="co">#</span></span>
<span id="cb6-2"><a href="collectdb.html#cb6-2"></a><span class="co"># Note again that the following result has already been prepared for human and</span></span>
<span id="cb6-3"><a href="collectdb.html#cb6-3"></a><span class="co"># mouse datasets in December 2020 (see `data/2020_001_databases/`).</span></span>
<span id="cb6-4"><a href="collectdb.html#cb6-4"></a><span class="co">#</span></span>
<span id="cb6-5"><a href="collectdb.html#cb6-5"></a><span class="kw">library</span>(GO.db)</span>
<span id="cb6-6"><a href="collectdb.html#cb6-6"></a>all_geneIDs &lt;-<span class="st"> </span><span class="kw">as.list</span>(org.Hs.egGO)</span>
<span id="cb6-7"><a href="collectdb.html#cb6-7"></a>all_geneIDs &lt;-<span class="st"> </span><span class="kw">names</span>(all_geneIDs[<span class="op">!</span><span class="kw">is.na</span>(all_geneIDs)])</span>
<span id="cb6-8"><a href="collectdb.html#cb6-8"></a>dict_GO &lt;-<span class="st"> </span><span class="kw">collect_GO</span>(<span class="dt">orgdb =</span> org.Hs.eg.db, <span class="dt">all_geneIDs =</span> all_geneIDs)</span>
<span id="cb6-9"><a href="collectdb.html#cb6-9"></a>tidy_GO &lt;-<span class="st"> </span><span class="kw">format_GO</span>(<span class="dt">dict =</span> dict_GO)</span>
<span id="cb6-10"><a href="collectdb.html#cb6-10"></a>IC_GO &lt;-<span class="st"> </span><span class="kw">compute_IC_GO</span>(<span class="dt">tidy =</span> tidy_GO)</span>
<span id="cb6-11"><a href="collectdb.html#cb6-11"></a><span class="co">#</span></span>
<span id="cb6-12"><a href="collectdb.html#cb6-12"></a><span class="co"># Save the result.</span></span>
<span id="cb6-13"><a href="collectdb.html#cb6-13"></a><span class="co">#</span></span>
<span id="cb6-14"><a href="collectdb.html#cb6-14"></a><span class="kw">saveRDS</span>(dict_DO, <span class="dt">file =</span> <span class="st">&quot;data/2020_001_databases/YYYYMMDD_dict_GO_human.rds&quot;</span>)</span>
<span id="cb6-15"><a href="collectdb.html#cb6-15"></a><span class="kw">saveRDS</span>(tidy_DO, <span class="dt">file =</span> <span class="st">&quot;data/2020_001_databases/YYYYMMDD_tidy_GO_human.rds&quot;</span>)</span>
<span id="cb6-16"><a href="collectdb.html#cb6-16"></a><span class="kw">saveRDS</span>(IC_DO, <span class="dt">file =</span> <span class="st">&quot;data/2020_001_databases/YYYYMMDD_IC_GO_human.rds&quot;</span>)</span></code></pre></div>
</div>
<div id="kegg" class="section level2">
<h2><span class="header-section-number">7.4</span> KEGG</h2>
<div class="sourceCode" id="cb7"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb7-1"><a href="collectdb.html#cb7-1"></a><span class="kw">source</span>(<span class="st">&quot;R/function_KEGG.R&quot;</span>)</span></code></pre></div>
<p>The following functions <code>collect_KEGG()</code> and <code>format_KEGG()</code>
(i) downloads a database for KEGG and (ii) formats the results, respectively.
In the current version, KEGG PATHWAY (<code>"pathway"</code>) and MODULE (<code>"module"</code>)
can be input into <code>collect_KEGG()</code> (cf. <code>KEGGREST</code> function <code>listDatabases()</code>).
The arguments of <code>collect_KEGG()</code> are <code>organism</code> (such as Homo sapiens <code>"hsa"</code>,
obeying the naming rule of <a href="http://rest.kegg.jp/list/organism">KEGG</a>) and
<code>category_names</code> (such as <code>"pathway"</code> and <code>"module"</code>).
This function requires an internet connection.</p>
<div class="sourceCode" id="cb8"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb8-1"><a href="collectdb.html#cb8-1"></a><span class="co">#</span></span>
<span id="cb8-2"><a href="collectdb.html#cb8-2"></a><span class="co"># Note again that the following result has already been prepared for human and</span></span>
<span id="cb8-3"><a href="collectdb.html#cb8-3"></a><span class="co"># mouse datasets in December 2020 (see `data/2020_001_databases/`).</span></span>
<span id="cb8-4"><a href="collectdb.html#cb8-4"></a><span class="co">#</span></span>
<span id="cb8-5"><a href="collectdb.html#cb8-5"></a><span class="co"># Note that `collect_KEGG` uses `KEGGREST` function `keggGet` with internet</span></span>
<span id="cb8-6"><a href="collectdb.html#cb8-6"></a><span class="co"># connection, which may produce both successful and unsuccessful results.</span></span>
<span id="cb8-7"><a href="collectdb.html#cb8-7"></a><span class="co">#</span></span>
<span id="cb8-8"><a href="collectdb.html#cb8-8"></a>dict_KEGG_raw &lt;-<span class="st"> </span><span class="kw">collect_KEGG</span>(<span class="dt">organism =</span> <span class="st">&quot;hsa&quot;</span>, <span class="dt">category_names =</span> <span class="kw">c</span>(<span class="st">&quot;pathway&quot;</span>))</span>
<span id="cb8-9"><a href="collectdb.html#cb8-9"></a>dict_KEGG &lt;-<span class="st"> </span><span class="kw">list</span>(<span class="dt">pathway =</span> dict_KEGG_raw[[<span class="st">&quot;pathway&quot;</span>]][[<span class="st">&quot;success&quot;</span>]])</span>
<span id="cb8-10"><a href="collectdb.html#cb8-10"></a>tidy_KEGG &lt;-<span class="st"> </span><span class="kw">format_KEGG</span>(<span class="dt">dict =</span> dict_KEGG)</span>
<span id="cb8-11"><a href="collectdb.html#cb8-11"></a><span class="co">#</span></span>
<span id="cb8-12"><a href="collectdb.html#cb8-12"></a><span class="co"># Save the result.</span></span>
<span id="cb8-13"><a href="collectdb.html#cb8-13"></a><span class="co">#</span></span>
<span id="cb8-14"><a href="collectdb.html#cb8-14"></a><span class="kw">saveRDS</span>(dict_KEGG_raw,</span>
<span id="cb8-15"><a href="collectdb.html#cb8-15"></a>        <span class="dt">file =</span> <span class="st">&quot;data/2020_001_databases/YYYYMMDD_dict_KEGG_raw_human.rds&quot;</span>)</span>
<span id="cb8-16"><a href="collectdb.html#cb8-16"></a><span class="kw">saveRDS</span>(dict_KEGG,</span>
<span id="cb8-17"><a href="collectdb.html#cb8-17"></a>        <span class="dt">file =</span> <span class="st">&quot;data/2020_001_databases/YYYYMMDD_dict_KEGG_human.rds&quot;</span>)</span>
<span id="cb8-18"><a href="collectdb.html#cb8-18"></a><span class="kw">saveRDS</span>(tidy_KEGG,</span>
<span id="cb8-19"><a href="collectdb.html#cb8-19"></a>        <span class="dt">file =</span> <span class="st">&quot;data/2020_001_databases/YYYYMMDD_tidy_KEGG_human.rds&quot;</span>)</span></code></pre></div>
</div>
<div id="reactome" class="section level2">
<h2><span class="header-section-number">7.5</span> Reactome</h2>
<div class="sourceCode" id="cb9"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb9-1"><a href="collectdb.html#cb9-1"></a><span class="kw">source</span>(<span class="st">&quot;R/function_Reactome.R&quot;</span>)</span></code></pre></div>
<p>The following functions <code>collect_Reactome()</code> and <code>format_Reactome()</code>
(i) downloads a database for Reactome and
(ii) formats the results, respectively.
In the current version, Reactome pathway (<code>"PATHID"</code>) can be input into
<code>collect_Reactome()</code> (cf. <code>reactome.db</code> functions <code>ls("package:reactome.db")</code>
and <code>keytypes(reactome.db)</code>).
The arguments are <code>organism</code> (such as Homo sapiens <code>Homo sapiens</code>, obeying the
naming rule of <code>reactome.db</code>) and <code>category_names</code> (such as <code>PATHID</code>).</p>
<div class="sourceCode" id="cb10"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb10-1"><a href="collectdb.html#cb10-1"></a><span class="co">#</span></span>
<span id="cb10-2"><a href="collectdb.html#cb10-2"></a><span class="co"># Note again that the following result has already been prepared for human and</span></span>
<span id="cb10-3"><a href="collectdb.html#cb10-3"></a><span class="co"># mouse datasets in December 2020 (see `data/2020_001_databases/`).</span></span>
<span id="cb10-4"><a href="collectdb.html#cb10-4"></a><span class="co">#</span></span>
<span id="cb10-5"><a href="collectdb.html#cb10-5"></a><span class="co"># Note that `collect_Reactome` may produce both successful and unsuccessful</span></span>
<span id="cb10-6"><a href="collectdb.html#cb10-6"></a><span class="co"># results.</span></span>
<span id="cb10-7"><a href="collectdb.html#cb10-7"></a><span class="co">#</span></span>
<span id="cb10-8"><a href="collectdb.html#cb10-8"></a>dict_Reactome_raw &lt;-<span class="st"> </span><span class="kw">collect_Reactome</span>(<span class="dt">organism =</span> <span class="st">&quot;Homo sapiens&quot;</span>,</span>
<span id="cb10-9"><a href="collectdb.html#cb10-9"></a>                                      <span class="dt">category_names =</span> <span class="kw">c</span>(<span class="st">&quot;PATHID&quot;</span>))</span>
<span id="cb10-10"><a href="collectdb.html#cb10-10"></a>dict_Reactome &lt;-<span class="st"> </span><span class="kw">list</span>(<span class="dt">pathid =</span> dict_Reactome_raw[[<span class="st">&quot;PATHID&quot;</span>]][[<span class="st">&quot;success&quot;</span>]])</span>
<span id="cb10-11"><a href="collectdb.html#cb10-11"></a>tidy_Reactome &lt;-<span class="st"> </span><span class="kw">format_Reactome</span>(<span class="dt">dict =</span> dict_Reactome)</span>
<span id="cb10-12"><a href="collectdb.html#cb10-12"></a><span class="co">#</span></span>
<span id="cb10-13"><a href="collectdb.html#cb10-13"></a><span class="co"># Save the result.</span></span>
<span id="cb10-14"><a href="collectdb.html#cb10-14"></a><span class="co">#</span></span>
<span id="cb10-15"><a href="collectdb.html#cb10-15"></a><span class="kw">saveRDS</span>(dict_Reactome_raw,</span>
<span id="cb10-16"><a href="collectdb.html#cb10-16"></a>        <span class="dt">file =</span> <span class="st">&quot;data/2020_001_databases/YYYYMMDD_dict_Reactome_raw_human.rds&quot;</span>)</span>
<span id="cb10-17"><a href="collectdb.html#cb10-17"></a><span class="kw">saveRDS</span>(dict_Reactome,</span>
<span id="cb10-18"><a href="collectdb.html#cb10-18"></a>        <span class="dt">file =</span> <span class="st">&quot;data/2020_001_databases/YYYYMMDD_dict_Reactome_human.rds&quot;</span>)</span>
<span id="cb10-19"><a href="collectdb.html#cb10-19"></a><span class="kw">saveRDS</span>(tidy_Reactome,</span>
<span id="cb10-20"><a href="collectdb.html#cb10-20"></a>        <span class="dt">file =</span> <span class="st">&quot;data/2020_001_databases/YYYYMMDD_tidy_Reactome_human.rds&quot;</span>)</span></code></pre></div>

</div>
</div>
            </section>

          </div>
        </div>
      </div>
<a href="markergenes.html" class="navigation navigation-prev " aria-label="Previous page"><i class="fa fa-angle-left"></i></a>
<a href="asurat-using-disease-ontology-database.html" class="navigation navigation-next " aria-label="Next page"><i class="fa fa-angle-right"></i></a>
    </div>
  </div>
<script src="libs/gitbook-2.6.7/js/app.min.js"></script>
<script src="libs/gitbook-2.6.7/js/lunr.js"></script>
<script src="libs/gitbook-2.6.7/js/clipboard.min.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-search.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-sharing.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-fontsettings.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-bookdown.js"></script>
<script src="libs/gitbook-2.6.7/js/jquery.highlight.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-clipboard.js"></script>
<script>
gitbook.require(["gitbook"], function(gitbook) {
gitbook.start({
"sharing": {
"github": false,
"facebook": true,
"twitter": true,
"linkedin": false,
"weibo": false,
"instapaper": false,
"vk": false,
"all": ["facebook", "twitter", "linkedin", "weibo", "instapaper"]
},
"fontsettings": {
"theme": "white",
"family": "sans",
"size": 2
},
"edit": {
"link": null,
"text": null
},
"history": {
"link": null,
"text": null
},
"view": {
"link": null,
"text": null
},
"download": ["Introduction to ASURAT (ver. 0.0.0.9000).pdf", "Introduction to ASURAT (ver. 0.0.0.9000).epub"],
"toc": {
"collapse": "subsection"
}
});
});
</script>

</body>

</html>
