<!DOCTYPE html>
<html lang="" xml:lang="">
<head>

  <meta charset="utf-8" />
  <meta http-equiv="X-UA-Compatible" content="IE=edge" />
  <title>Chapter 3 Data quality control (QC) | Introduction to ASURAT (version 0.0.0.9000)</title>
  <meta name="description" content="This is a developmental version of ASURAT." />
  <meta name="generator" content="bookdown 0.21 and GitBook 2.6.7" />

  <meta property="og:title" content="Chapter 3 Data quality control (QC) | Introduction to ASURAT (version 0.0.0.9000)" />
  <meta property="og:type" content="book" />
  
  
  <meta property="og:description" content="This is a developmental version of ASURAT." />
  

  <meta name="twitter:card" content="summary" />
  <meta name="twitter:title" content="Chapter 3 Data quality control (QC) | Introduction to ASURAT (version 0.0.0.9000)" />
  
  <meta name="twitter:description" content="This is a developmental version of ASURAT." />
  

<meta name="author" content="Keita Iida" />


<meta name="date" content="2021-06-16" />

  <meta name="viewport" content="width=device-width, initial-scale=1" />
  <meta name="apple-mobile-web-app-capable" content="yes" />
  <meta name="apple-mobile-web-app-status-bar-style" content="black" />
  
  
<link rel="prev" href="dataset.html"/>
<link rel="next" href="normalization.html"/>
<script src="libs/jquery-2.2.3/jquery.min.js"></script>
<link href="libs/gitbook-2.6.7/css/style.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-table.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-bookdown.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-highlight.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-search.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-fontsettings.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-clipboard.css" rel="stylesheet" />









<script src="libs/accessible-code-block-0.0.1/empty-anchor.js"></script>


<style type="text/css">
code.sourceCode > span { display: inline-block; line-height: 1.25; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode { white-space: pre; position: relative; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
code.sourceCode { white-space: pre-wrap; }
code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } /* Alert */
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code span.at { color: #7d9029; } /* Attribute */
code span.bn { color: #40a070; } /* BaseN */
code span.bu { } /* BuiltIn */
code span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code span.ch { color: #4070a0; } /* Char */
code span.cn { color: #880000; } /* Constant */
code span.co { color: #60a0b0; font-style: italic; } /* Comment */
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code span.do { color: #ba2121; font-style: italic; } /* Documentation */
code span.dt { color: #902000; } /* DataType */
code span.dv { color: #40a070; } /* DecVal */
code span.er { color: #ff0000; font-weight: bold; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #40a070; } /* Float */
code span.fu { color: #06287e; } /* Function */
code span.im { } /* Import */
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
code span.kw { color: #007020; font-weight: bold; } /* Keyword */
code span.op { color: #666666; } /* Operator */
code span.ot { color: #007020; } /* Other */
code span.pp { color: #bc7a00; } /* Preprocessor */
code span.sc { color: #4070a0; } /* SpecialChar */
code span.ss { color: #bb6688; } /* SpecialString */
code span.st { color: #4070a0; } /* String */
code span.va { color: #19177c; } /* Variable */
code span.vs { color: #4070a0; } /* VerbatimString */
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */
</style>

<link rel="stylesheet" href="style.css" type="text/css" />
</head>

<body>



  <div class="book without-animation with-summary font-size-2 font-family-1" data-basepath=".">

    <div class="book-summary">
      <nav role="navigation">

<ul class="summary">
<li><a href="./">A Minimal Book Example</a></li>

<li class="divider"></li>
<li class="chapter" data-level="1" data-path="index.html"><a href="index.html"><i class="fa fa-check"></i><b>1</b> Overview</a><ul>
<li class="chapter" data-level="1.1" data-path="index.html"><a href="index.html#license-and-description"><i class="fa fa-check"></i><b>1.1</b> License and description</a></li>
<li class="chapter" data-level="1.2" data-path="index.html"><a href="index.html#computational-environment"><i class="fa fa-check"></i><b>1.2</b> Computational environment</a></li>
<li class="chapter" data-level="1.3" data-path="index.html"><a href="index.html#installation"><i class="fa fa-check"></i><b>1.3</b> Installation</a></li>
</ul></li>
<li class="chapter" data-level="2" data-path="dataset.html"><a href="dataset.html"><i class="fa fa-check"></i><b>2</b> Preparing data sets</a><ul>
<li class="chapter" data-level="2.1" data-path="dataset.html"><a href="dataset.html#case-i-input-gene-expression-matrix-with-row-and-column-names"><i class="fa fa-check"></i><b>2.1</b> Case I: input gene expression matrix with row and column names</a></li>
<li class="chapter" data-level="2.2" data-path="dataset.html"><a href="dataset.html#case-ii-input-10x-data"><i class="fa fa-check"></i><b>2.2</b> Case II: input 10x data</a></li>
<li class="chapter" data-level="2.3" data-path="dataset.html"><a href="dataset.html#case-iii-input-10x-pbmc-data"><i class="fa fa-check"></i><b>2.3</b> Case III: input 10x PBMC data</a></li>
</ul></li>
<li class="chapter" data-level="3" data-path="dataqc.html"><a href="dataqc.html"><i class="fa fa-check"></i><b>3</b> Data quality control (QC)</a><ul>
<li class="chapter" data-level="3.1" data-path="dataqc.html"><a href="dataqc.html#quick-qc"><i class="fa fa-check"></i><b>3.1</b> Quick QC</a></li>
<li class="chapter" data-level="3.2" data-path="dataqc.html"><a href="dataqc.html#sample-qc"><i class="fa fa-check"></i><b>3.2</b> Sample QC</a></li>
<li class="chapter" data-level="3.3" data-path="dataqc.html"><a href="dataqc.html#variable-qc"><i class="fa fa-check"></i><b>3.3</b> Variable QC</a></li>
</ul></li>
<li class="chapter" data-level="4" data-path="normalization.html"><a href="normalization.html"><i class="fa fa-check"></i><b>4</b> Normalizing and centering data</a></li>
<li class="chapter" data-level="5" data-path="computcorr.html"><a href="computcorr.html"><i class="fa fa-check"></i><b>5</b> Computing correlations among genes</a></li>
<li class="chapter" data-level="6" data-path="markergenes.html"><a href="markergenes.html"><i class="fa fa-check"></i><b>6</b> Checking expression profiles of marker genes</a></li>
<li class="chapter" data-level="7" data-path="collectdb.html"><a href="collectdb.html"><i class="fa fa-check"></i><b>7</b> Collecting databases (optional)</a><ul>
<li class="chapter" data-level="7.1" data-path="collectdb.html"><a href="collectdb.html#disease-ontology-db"><i class="fa fa-check"></i><b>7.1</b> Disease Ontology DB</a></li>
<li class="chapter" data-level="7.2" data-path="collectdb.html"><a href="collectdb.html#cell-ontology-db"><i class="fa fa-check"></i><b>7.2</b> Cell Ontology DB</a></li>
<li class="chapter" data-level="7.3" data-path="collectdb.html"><a href="collectdb.html#gene-ontology-db"><i class="fa fa-check"></i><b>7.3</b> Gene Ontology DB</a></li>
<li class="chapter" data-level="7.4" data-path="collectdb.html"><a href="collectdb.html#kegg"><i class="fa fa-check"></i><b>7.4</b> KEGG</a></li>
<li class="chapter" data-level="7.5" data-path="collectdb.html"><a href="collectdb.html#reactome"><i class="fa fa-check"></i><b>7.5</b> Reactome</a></li>
</ul></li>
<li class="chapter" data-level="8" data-path="asurat-using-disease-ontology-database.html"><a href="asurat-using-disease-ontology-database.html"><i class="fa fa-check"></i><b>8</b> ASURAT using Disease Ontology database</a><ul>
<li class="chapter" data-level="8.1" data-path="asurat-using-disease-ontology-database.html"><a href="asurat-using-disease-ontology-database.html#creating-sign-by-sample-matrices"><i class="fa fa-check"></i><b>8.1</b> Creating sign-by-sample matrices</a></li>
<li class="chapter" data-level="8.2" data-path="asurat-using-disease-ontology-database.html"><a href="asurat-using-disease-ontology-database.html#dimensionality-reduction"><i class="fa fa-check"></i><b>8.2</b> Dimensionality reduction</a></li>
<li class="chapter" data-level="8.3" data-path="asurat-using-disease-ontology-database.html"><a href="asurat-using-disease-ontology-database.html#cell-clustering"><i class="fa fa-check"></i><b>8.3</b> Cell clustering</a></li>
<li class="chapter" data-level="8.4" data-path="asurat-using-disease-ontology-database.html"><a href="asurat-using-disease-ontology-database.html#visualization"><i class="fa fa-check"></i><b>8.4</b> Visualization</a></li>
<li class="chapter" data-level="8.5" data-path="asurat-using-disease-ontology-database.html"><a href="asurat-using-disease-ontology-database.html#significant-signs"><i class="fa fa-check"></i><b>8.5</b> Significant signs</a></li>
</ul></li>
<li class="chapter" data-level="9" data-path="asurat-using-cell-ontology-database-optional.html"><a href="asurat-using-cell-ontology-database-optional.html"><i class="fa fa-check"></i><b>9</b> ASURAT using Cell Ontology database (optional)</a><ul>
<li class="chapter" data-level="9.1" data-path="asurat-using-cell-ontology-database-optional.html"><a href="asurat-using-cell-ontology-database-optional.html#creating-sign-by-sample-matrices-1"><i class="fa fa-check"></i><b>9.1</b> Creating sign-by-sample matrices</a></li>
<li class="chapter" data-level="9.2" data-path="asurat-using-cell-ontology-database-optional.html"><a href="asurat-using-cell-ontology-database-optional.html#dimensionality-reduction-1"><i class="fa fa-check"></i><b>9.2</b> Dimensionality reduction</a></li>
<li class="chapter" data-level="9.3" data-path="asurat-using-cell-ontology-database-optional.html"><a href="asurat-using-cell-ontology-database-optional.html#cell-clustering-1"><i class="fa fa-check"></i><b>9.3</b> Cell clustering</a></li>
<li class="chapter" data-level="9.4" data-path="asurat-using-cell-ontology-database-optional.html"><a href="asurat-using-cell-ontology-database-optional.html#visualization-1"><i class="fa fa-check"></i><b>9.4</b> Visualization</a></li>
<li class="chapter" data-level="9.5" data-path="asurat-using-cell-ontology-database-optional.html"><a href="asurat-using-cell-ontology-database-optional.html#significant-signs-1"><i class="fa fa-check"></i><b>9.5</b> Significant signs</a></li>
</ul></li>
<li class="chapter" data-level="10" data-path="asurat-using-gene-ontology-database-optional.html"><a href="asurat-using-gene-ontology-database-optional.html"><i class="fa fa-check"></i><b>10</b> ASURAT using Gene Ontology database (optional)</a><ul>
<li class="chapter" data-level="10.1" data-path="asurat-using-gene-ontology-database-optional.html"><a href="asurat-using-gene-ontology-database-optional.html#creating-sign-by-sample-matrices-2"><i class="fa fa-check"></i><b>10.1</b> Creating sign-by-sample matrices</a></li>
<li class="chapter" data-level="10.2" data-path="asurat-using-gene-ontology-database-optional.html"><a href="asurat-using-gene-ontology-database-optional.html#dimensionality-reduction-2"><i class="fa fa-check"></i><b>10.2</b> Dimensionality reduction</a></li>
<li class="chapter" data-level="10.3" data-path="asurat-using-gene-ontology-database-optional.html"><a href="asurat-using-gene-ontology-database-optional.html#cell-clustering-2"><i class="fa fa-check"></i><b>10.3</b> Cell clustering</a></li>
<li class="chapter" data-level="10.4" data-path="asurat-using-gene-ontology-database-optional.html"><a href="asurat-using-gene-ontology-database-optional.html#visualization-2"><i class="fa fa-check"></i><b>10.4</b> Visualization</a></li>
<li class="chapter" data-level="10.5" data-path="asurat-using-gene-ontology-database-optional.html"><a href="asurat-using-gene-ontology-database-optional.html#significant-signs-2"><i class="fa fa-check"></i><b>10.5</b> Significant signs</a></li>
</ul></li>
<li class="chapter" data-level="11" data-path="asurat-using-kegg-optional.html"><a href="asurat-using-kegg-optional.html"><i class="fa fa-check"></i><b>11</b> ASURAT using KEGG (optional)</a><ul>
<li class="chapter" data-level="11.1" data-path="asurat-using-kegg-optional.html"><a href="asurat-using-kegg-optional.html#creating-sign-by-sample-matrices-3"><i class="fa fa-check"></i><b>11.1</b> Creating sign-by-sample matrices</a></li>
<li class="chapter" data-level="11.2" data-path="asurat-using-kegg-optional.html"><a href="asurat-using-kegg-optional.html#dimensionality-reduction-3"><i class="fa fa-check"></i><b>11.2</b> Dimensionality reduction</a></li>
<li class="chapter" data-level="11.3" data-path="asurat-using-kegg-optional.html"><a href="asurat-using-kegg-optional.html#cell-clustering-3"><i class="fa fa-check"></i><b>11.3</b> Cell clustering</a></li>
<li class="chapter" data-level="11.4" data-path="asurat-using-kegg-optional.html"><a href="asurat-using-kegg-optional.html#visualizing"><i class="fa fa-check"></i><b>11.4</b> Visualizing</a></li>
<li class="chapter" data-level="11.5" data-path="asurat-using-kegg-optional.html"><a href="asurat-using-kegg-optional.html#significant-signs-3"><i class="fa fa-check"></i><b>11.5</b> Significant signs</a></li>
</ul></li>
<li class="chapter" data-level="12" data-path="asurat-using-reactome-optional.html"><a href="asurat-using-reactome-optional.html"><i class="fa fa-check"></i><b>12</b> ASURAT using Reactome (optional)</a><ul>
<li class="chapter" data-level="12.1" data-path="asurat-using-reactome-optional.html"><a href="asurat-using-reactome-optional.html#creating-sign-by-sample-matrices-4"><i class="fa fa-check"></i><b>12.1</b> Creating sign-by-sample matrices</a></li>
<li class="chapter" data-level="12.2" data-path="asurat-using-reactome-optional.html"><a href="asurat-using-reactome-optional.html#dimensionality-reduction-4"><i class="fa fa-check"></i><b>12.2</b> Dimensionality reduction</a></li>
<li class="chapter" data-level="12.3" data-path="asurat-using-reactome-optional.html"><a href="asurat-using-reactome-optional.html#sample-clustering"><i class="fa fa-check"></i><b>12.3</b> Sample clustering</a></li>
</ul></li>
<li class="chapter" data-level="13" data-path="multiple-sign-analysis-by-concatenating-do-co-go-kegg-and-reactome.html"><a href="multiple-sign-analysis-by-concatenating-do-co-go-kegg-and-reactome.html"><i class="fa fa-check"></i><b>13</b> Multiple sign analysis by concatenating DO, CO, GO, KEGG, and Reactome</a></li>
<li class="chapter" data-level="14" data-path="appendix-a-comparing-performances-of-asurat-and-existing-methods.html"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html"><i class="fa fa-check"></i><b>14</b> Appendix A: comparing performances of ASURAT and existing methods</a><ul>
<li class="chapter" data-level="14.1" data-path="appendix-a-comparing-performances-of-asurat-and-existing-methods.html"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#installation-1"><i class="fa fa-check"></i><b>14.1</b> Installation</a></li>
<li class="chapter" data-level="14.2" data-path="appendix-a-comparing-performances-of-asurat-and-existing-methods.html"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#removing-the-outlier-cells-detected-by-asurat"><i class="fa fa-check"></i><b>14.2</b> Removing the outlier cells detected by ASURAT</a></li>
<li class="chapter" data-level="14.3" data-path="appendix-a-comparing-performances-of-asurat-and-existing-methods.html"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#assessing-clustering-performances"><i class="fa fa-check"></i><b>14.3</b> Assessing clustering performances</a><ul>
<li class="chapter" data-level="14.3.1" data-path="appendix-a-comparing-performances-of-asurat-and-existing-methods.html"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#seurat"><i class="fa fa-check"></i><b>14.3.1</b> Seurat</a></li>
<li class="chapter" data-level="14.3.2" data-path="appendix-a-comparing-performances-of-asurat-and-existing-methods.html"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#monocle-3"><i class="fa fa-check"></i><b>14.3.2</b> Monocle 3</a></li>
<li class="chapter" data-level="14.3.3" data-path="appendix-a-comparing-performances-of-asurat-and-existing-methods.html"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#sc3"><i class="fa fa-check"></i><b>14.3.3</b> SC3</a></li>
<li class="chapter" data-level="14.3.4" data-path="appendix-a-comparing-performances-of-asurat-and-existing-methods.html"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#principal-component-analysis-pca"><i class="fa fa-check"></i><b>14.3.4</b> Principal component analysis (PCA)</a></li>
<li class="chapter" data-level="14.3.5" data-path="appendix-a-comparing-performances-of-asurat-and-existing-methods.html"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#asurat"><i class="fa fa-check"></i><b>14.3.5</b> ASURAT</a></li>
<li class="chapter" data-level="14.3.6" data-path="appendix-a-comparing-performances-of-asurat-and-existing-methods.html"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#comparing-clustering-validity-indices"><i class="fa fa-check"></i><b>14.3.6</b> Comparing clustering validity indices</a></li>
</ul></li>
<li class="chapter" data-level="14.4" data-path="appendix-a-comparing-performances-of-asurat-and-existing-methods.html"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#cell-type-inference-in-pbmc_4000-and-pbmc_6000"><i class="fa fa-check"></i><b>14.4</b> Cell-type inference in <code>pbmc_4000</code> and <code>pbmc_6000</code></a><ul>
<li class="chapter" data-level="14.4.1" data-path="appendix-a-comparing-performances-of-asurat-and-existing-methods.html"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#seurat-1"><i class="fa fa-check"></i><b>14.4.1</b> Seurat</a></li>
<li class="chapter" data-level="14.4.2" data-path="appendix-a-comparing-performances-of-asurat-and-existing-methods.html"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#monocle-3-1"><i class="fa fa-check"></i><b>14.4.2</b> Monocle 3</a></li>
<li class="chapter" data-level="14.4.3" data-path="appendix-a-comparing-performances-of-asurat-and-existing-methods.html"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#sc3-1"><i class="fa fa-check"></i><b>14.4.3</b> SC3</a></li>
<li class="chapter" data-level="14.4.4" data-path="appendix-a-comparing-performances-of-asurat-and-existing-methods.html"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#asurat-1"><i class="fa fa-check"></i><b>14.4.4</b> ASURAT</a></li>
</ul></li>
<li class="chapter" data-level="14.5" data-path="appendix-a-comparing-performances-of-asurat-and-existing-methods.html"><a href="appendix-a-comparing-performances-of-asurat-and-existing-methods.html#identifying-cell-types-in-pbmc_4000-and-pbmc_6000"><i class="fa fa-check"></i><b>14.5</b> Identifying cell types in <code>pbmc_4000</code> and <code>pbmc_6000</code></a></li>
</ul></li>
<li class="chapter" data-level="15" data-path="appendix-b-automatically-tuning-asurats-parameters.html"><a href="appendix-b-automatically-tuning-asurats-parameters.html"><i class="fa fa-check"></i><b>15</b> Appendix B: automatically tuning ASURAT’s parameters</a></li>
<li class="divider"></li>
<li><a href="https://github.com/rstudio/bookdown" target="blank">Published with bookdown</a></li>

</ul>

      </nav>
    </div>

    <div class="book-body">
      <div class="body-inner">
        <div class="book-header" role="navigation">
          <h1>
            <i class="fa fa-circle-o-notch fa-spin"></i><a href="./">Introduction to ASURAT (version 0.0.0.9000)</a>
          </h1>
        </div>

        <div class="page-wrapper" tabindex="-1" role="main">
          <div class="page-inner">

            <section class="normal" id="section-">
<div id="dataqc" class="section level1">
<h1><span class="header-section-number">Chapter 3</span> Data quality control (QC)</h1>
<p>In this section, we remove the genes and cells with low quality,
by processing the following three steps:
(1) quick gene filtering based on expression profile across cells,
(2) cell filtering based on QC metrics, and
(3) second gene filtering based on mean expression levels across cells.</p>
<div id="quick-qc" class="section level2">
<h2><span class="header-section-number">3.1</span> Quick QC</h2>
<p>Load the data.</p>
<div class="sourceCode" id="cb1"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb1-1"><a href="dataqc.html#cb1-1"></a><span class="kw">source</span>(<span class="st">&quot;R/function_general.R&quot;</span>)</span>
<span id="cb1-2"><a href="dataqc.html#cb1-2"></a>day1_norm &lt;-<span class="st"> </span><span class="kw">readRDS</span>(<span class="dt">file =</span> <span class="st">&quot;backup/01_001_day1_norm_raw.rds&quot;</span>)</span>
<span id="cb1-3"><a href="dataqc.html#cb1-3"></a>day7_hypo &lt;-<span class="st"> </span><span class="kw">readRDS</span>(<span class="dt">file =</span> <span class="st">&quot;backup/02_001_day7_hypo_raw.rds&quot;</span>)</span>
<span id="cb1-4"><a href="dataqc.html#cb1-4"></a>sc68_vehi &lt;-<span class="st"> </span><span class="kw">readRDS</span>(<span class="dt">file =</span> <span class="st">&quot;backup/03_001_sc68_vehi_raw.rds&quot;</span>)</span>
<span id="cb1-5"><a href="dataqc.html#cb1-5"></a>sc68_cisp &lt;-<span class="st"> </span><span class="kw">readRDS</span>(<span class="dt">file =</span> <span class="st">&quot;backup/04_001_sc68_cisp_raw.rds&quot;</span>)</span>
<span id="cb1-6"><a href="dataqc.html#cb1-6"></a>pbmc_<span class="dv">4000</span> &lt;-<span class="st"> </span><span class="kw">readRDS</span>(<span class="dt">file =</span> <span class="st">&quot;backup/05_001_pbmc_4000_raw.rds&quot;</span>)</span>
<span id="cb1-7"><a href="dataqc.html#cb1-7"></a>pbmc_<span class="dv">6000</span> &lt;-<span class="st"> </span><span class="kw">readRDS</span>(<span class="dt">file =</span> <span class="st">&quot;backup/06_001_pbmc_6000_raw.rds&quot;</span>)</span></code></pre></div>
<p>Check the size of the gene-by-sample matrices.</p>
<div class="sourceCode" id="cb2"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb2-1"><a href="dataqc.html#cb2-1"></a><span class="kw">rbind</span>(<span class="kw">dim</span>(day1_norm[[<span class="st">&quot;data&quot;</span>]][[<span class="st">&quot;raw&quot;</span>]]), <span class="kw">dim</span>(day7_hypo[[<span class="st">&quot;data&quot;</span>]][[<span class="st">&quot;raw&quot;</span>]]),</span>
<span id="cb2-2"><a href="dataqc.html#cb2-2"></a>      <span class="kw">dim</span>(sc68_vehi[[<span class="st">&quot;data&quot;</span>]][[<span class="st">&quot;raw&quot;</span>]]), <span class="kw">dim</span>(sc68_cisp[[<span class="st">&quot;data&quot;</span>]][[<span class="st">&quot;raw&quot;</span>]]),</span>
<span id="cb2-3"><a href="dataqc.html#cb2-3"></a>      <span class="kw">dim</span>(pbmc_<span class="dv">4000</span>[[<span class="st">&quot;data&quot;</span>]][[<span class="st">&quot;raw&quot;</span>]]), <span class="kw">dim</span>(pbmc_<span class="dv">6000</span>[[<span class="st">&quot;data&quot;</span>]][[<span class="st">&quot;raw&quot;</span>]]))</span></code></pre></div>
<pre><code>[1,] 21383 11417
[2,] 20073  6117
[3,] 33538  5905
[4,] 33538  3433
[5,] 33694  4340
[6,] 32738  5419</code></pre>
<p>The following function <code>do_quickQC()</code> removes the genes for which the number
of non-zero expressing cells is less than <code>min_nsamples</code>.
In addition, this function counts the number of reads, the number of non-zero
expressing genes, and the percent of reads that map to the mitochondrial genes,
e.g., <code>MT-Co1</code>, <code>mt-Co1</code>, etc.</p>
<div class="sourceCode" id="cb4"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb4-1"><a href="dataqc.html#cb4-1"></a>day1_norm &lt;-<span class="st"> </span><span class="kw">do_quickQC</span>(<span class="dt">obj =</span> day1_norm, <span class="dt">min_nsamples =</span> <span class="dv">10</span>,</span>
<span id="cb4-2"><a href="dataqc.html#cb4-2"></a>                        <span class="dt">mitochondria_symbol =</span> <span class="st">&quot;^MT-&quot;</span>)</span>
<span id="cb4-3"><a href="dataqc.html#cb4-3"></a>day7_hypo &lt;-<span class="st"> </span><span class="kw">do_quickQC</span>(<span class="dt">obj =</span> day7_hypo, <span class="dt">min_nsamples =</span> <span class="dv">10</span>,</span>
<span id="cb4-4"><a href="dataqc.html#cb4-4"></a>                        <span class="dt">mitochondria_symbol =</span> <span class="st">&quot;^MT-&quot;</span>)</span>
<span id="cb4-5"><a href="dataqc.html#cb4-5"></a>sc68_vehi &lt;-<span class="st"> </span><span class="kw">do_quickQC</span>(<span class="dt">obj =</span> sc68_vehi, <span class="dt">min_nsamples =</span> <span class="dv">10</span>,</span>
<span id="cb4-6"><a href="dataqc.html#cb4-6"></a>                        <span class="dt">mitochondria_symbol =</span> <span class="st">&quot;^MT-&quot;</span>)</span>
<span id="cb4-7"><a href="dataqc.html#cb4-7"></a>sc68_cisp &lt;-<span class="st"> </span><span class="kw">do_quickQC</span>(<span class="dt">obj =</span> sc68_cisp, <span class="dt">min_nsamples =</span> <span class="dv">10</span>,</span>
<span id="cb4-8"><a href="dataqc.html#cb4-8"></a>                        <span class="dt">mitochondria_symbol =</span> <span class="st">&quot;^MT-&quot;</span>)</span>
<span id="cb4-9"><a href="dataqc.html#cb4-9"></a>pbmc_<span class="dv">4000</span> &lt;-<span class="st"> </span><span class="kw">do_quickQC</span>(<span class="dt">obj =</span> pbmc_<span class="dv">4000</span>, <span class="dt">min_nsamples =</span> <span class="dv">10</span>,</span>
<span id="cb4-10"><a href="dataqc.html#cb4-10"></a>                        <span class="dt">mitochondria_symbol =</span> <span class="st">&quot;^MT-&quot;</span>)</span>
<span id="cb4-11"><a href="dataqc.html#cb4-11"></a>pbmc_<span class="dv">6000</span> &lt;-<span class="st"> </span><span class="kw">do_quickQC</span>(<span class="dt">obj =</span> pbmc_<span class="dv">6000</span>, <span class="dt">min_nsamples =</span> <span class="dv">10</span>,</span>
<span id="cb4-12"><a href="dataqc.html#cb4-12"></a>                        <span class="dt">mitochondria_symbol =</span> <span class="st">&quot;^MT-&quot;</span>)</span></code></pre></div>
<p>The results are stored in
<code>obj[["sample"]][["nReads"]]</code>, <code>obj[["sample"]][["nGenes"]]</code>,
<code>obj[["sample"]][["percent_MT"]]</code>, and <code>obj[["data"]][["raw"]]</code>.</p>
<p>Users can always check the parameter values in <code>obj[["history"]]</code>.
For example, <code>obj[["history"]][["do_quickQC"]][["min_nsamples"]]</code>.</p>
<p>Check the size of the gene-by-sample matrix.</p>
<div class="sourceCode" id="cb5"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb5-1"><a href="dataqc.html#cb5-1"></a><span class="kw">rbind</span>(<span class="kw">dim</span>(day1_norm[[<span class="st">&quot;data&quot;</span>]][[<span class="st">&quot;raw&quot;</span>]]), <span class="kw">dim</span>(day7_hypo[[<span class="st">&quot;data&quot;</span>]][[<span class="st">&quot;raw&quot;</span>]]),</span>
<span id="cb5-2"><a href="dataqc.html#cb5-2"></a>      <span class="kw">dim</span>(sc68_vehi[[<span class="st">&quot;data&quot;</span>]][[<span class="st">&quot;raw&quot;</span>]]), <span class="kw">dim</span>(sc68_cisp[[<span class="st">&quot;data&quot;</span>]][[<span class="st">&quot;raw&quot;</span>]]),</span>
<span id="cb5-3"><a href="dataqc.html#cb5-3"></a>      <span class="kw">dim</span>(pbmc_<span class="dv">4000</span>[[<span class="st">&quot;data&quot;</span>]][[<span class="st">&quot;raw&quot;</span>]]), <span class="kw">dim</span>(pbmc_<span class="dv">6000</span>[[<span class="st">&quot;data&quot;</span>]][[<span class="st">&quot;raw&quot;</span>]]))</span></code></pre></div>
<pre><code>[1,] 16820 11417
[2,] 15795  6117
[3,] 16784  5905
[4,] 15605  3433
[5,] 13829  4340
[6,] 11946  5419</code></pre>
<p>Save the objects.</p>
<div class="sourceCode" id="cb7"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb7-1"><a href="dataqc.html#cb7-1"></a><span class="kw">saveRDS</span>(day1_norm, <span class="dt">file =</span> <span class="st">&quot;backup/01_002_day1_norm_quickQC.rds&quot;</span>)</span>
<span id="cb7-2"><a href="dataqc.html#cb7-2"></a><span class="kw">saveRDS</span>(day7_hypo, <span class="dt">file =</span> <span class="st">&quot;backup/02_002_day7_hypo_quickQC.rds&quot;</span>)</span>
<span id="cb7-3"><a href="dataqc.html#cb7-3"></a><span class="kw">saveRDS</span>(sc68_vehi, <span class="dt">file =</span> <span class="st">&quot;backup/03_002_sc68_vehi_quickQC.rds&quot;</span>)</span>
<span id="cb7-4"><a href="dataqc.html#cb7-4"></a><span class="kw">saveRDS</span>(sc68_cisp, <span class="dt">file =</span> <span class="st">&quot;backup/04_002_sc68_cisp_quickQC.rds&quot;</span>)</span>
<span id="cb7-5"><a href="dataqc.html#cb7-5"></a><span class="kw">saveRDS</span>(pbmc_<span class="dv">4000</span>, <span class="dt">file =</span> <span class="st">&quot;backup/05_002_pbmc_4000_quickQC.rds&quot;</span>)</span>
<span id="cb7-6"><a href="dataqc.html#cb7-6"></a><span class="kw">saveRDS</span>(pbmc_<span class="dv">6000</span>, <span class="dt">file =</span> <span class="st">&quot;backup/06_002_pbmc_6000_quickQC.rds&quot;</span>)</span></code></pre></div>
</div>
<div id="sample-qc" class="section level2">
<h2><span class="header-section-number">3.2</span> Sample QC</h2>
<p>Load the data.</p>
<div class="sourceCode" id="cb8"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb8-1"><a href="dataqc.html#cb8-1"></a><span class="kw">source</span>(<span class="st">&quot;R/function_general.R&quot;</span>)</span>
<span id="cb8-2"><a href="dataqc.html#cb8-2"></a><span class="kw">source</span>(<span class="st">&quot;R/plot.R&quot;</span>)</span>
<span id="cb8-3"><a href="dataqc.html#cb8-3"></a>day1_norm &lt;-<span class="st"> </span><span class="kw">readRDS</span>(<span class="dt">file =</span> <span class="st">&quot;backup/01_002_day1_norm_quickQC.rds&quot;</span>)</span>
<span id="cb8-4"><a href="dataqc.html#cb8-4"></a>day7_hypo &lt;-<span class="st"> </span><span class="kw">readRDS</span>(<span class="dt">file =</span> <span class="st">&quot;backup/02_002_day7_hypo_quickQC.rds&quot;</span>)</span>
<span id="cb8-5"><a href="dataqc.html#cb8-5"></a>sc68_vehi &lt;-<span class="st"> </span><span class="kw">readRDS</span>(<span class="dt">file =</span> <span class="st">&quot;backup/03_002_sc68_vehi_quickQC.rds&quot;</span>)</span>
<span id="cb8-6"><a href="dataqc.html#cb8-6"></a>sc68_cisp &lt;-<span class="st"> </span><span class="kw">readRDS</span>(<span class="dt">file =</span> <span class="st">&quot;backup/04_002_sc68_cisp_quickQC.rds&quot;</span>)</span>
<span id="cb8-7"><a href="dataqc.html#cb8-7"></a>pbmc_<span class="dv">4000</span> &lt;-<span class="st"> </span><span class="kw">readRDS</span>(<span class="dt">file =</span> <span class="st">&quot;backup/05_002_pbmc_4000_quickQC.rds&quot;</span>)</span>
<span id="cb8-8"><a href="dataqc.html#cb8-8"></a>pbmc_<span class="dv">6000</span> &lt;-<span class="st"> </span><span class="kw">readRDS</span>(<span class="dt">file =</span> <span class="st">&quot;backup/06_002_pbmc_6000_quickQC.rds&quot;</span>)</span></code></pre></div>
<p>Check the size of the gene-by-sample matrices.</p>
<div class="sourceCode" id="cb9"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb9-1"><a href="dataqc.html#cb9-1"></a><span class="kw">rbind</span>(<span class="kw">dim</span>(day1_norm[[<span class="st">&quot;data&quot;</span>]][[<span class="st">&quot;raw&quot;</span>]]), <span class="kw">dim</span>(day7_hypo[[<span class="st">&quot;data&quot;</span>]][[<span class="st">&quot;raw&quot;</span>]]),</span>
<span id="cb9-2"><a href="dataqc.html#cb9-2"></a>      <span class="kw">dim</span>(sc68_vehi[[<span class="st">&quot;data&quot;</span>]][[<span class="st">&quot;raw&quot;</span>]]), <span class="kw">dim</span>(sc68_cisp[[<span class="st">&quot;data&quot;</span>]][[<span class="st">&quot;raw&quot;</span>]]),</span>
<span id="cb9-3"><a href="dataqc.html#cb9-3"></a>      <span class="kw">dim</span>(pbmc_<span class="dv">4000</span>[[<span class="st">&quot;data&quot;</span>]][[<span class="st">&quot;raw&quot;</span>]]), <span class="kw">dim</span>(pbmc_<span class="dv">6000</span>[[<span class="st">&quot;data&quot;</span>]][[<span class="st">&quot;raw&quot;</span>]]))</span></code></pre></div>
<pre><code>[1,] 16820 11417
[2,] 15795  6117
[3,] 16784  5905
[4,] 15605  3433
[5,] 13829  4340
[6,] 11946  5419</code></pre>
<p>The following function <code>plot_metaData()</code> outputs a scatter plot between the
input meta data for each cell
The arguments are <code>df</code> (two-dimensional data frame),
<code>title</code> (title of the plot), <code>xlabel</code> (x-axis label), and
<code>ylabel</code> (y-axis label).</p>
<div class="sourceCode" id="cb11"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb11-1"><a href="dataqc.html#cb11-1"></a>my_func &lt;-<span class="st"> </span><span class="cf">function</span>(obj, metaname1, metaname2){</span>
<span id="cb11-2"><a href="dataqc.html#cb11-2"></a>  df &lt;-<span class="st"> </span><span class="kw">data.frame</span>(<span class="dt">x =</span> obj[[<span class="st">&quot;sample&quot;</span>]][[metaname1]],</span>
<span id="cb11-3"><a href="dataqc.html#cb11-3"></a>                   <span class="dt">y =</span> obj[[<span class="st">&quot;sample&quot;</span>]][[metaname2]])</span>
<span id="cb11-4"><a href="dataqc.html#cb11-4"></a>  p &lt;-<span class="st"> </span><span class="kw">plot_metaData</span>(<span class="dt">df =</span> df,</span>
<span id="cb11-5"><a href="dataqc.html#cb11-5"></a>                    <span class="dt">title =</span> obj[[<span class="st">&quot;history&quot;</span>]][[<span class="st">&quot;make_asurat_obj&quot;</span>]][[<span class="st">&quot;obj_name&quot;</span>]],</span>
<span id="cb11-6"><a href="dataqc.html#cb11-6"></a>                    <span class="dt">title_size =</span> <span class="dv">16</span>, <span class="dt">xlabel =</span> metaname1, <span class="dt">ylabel =</span> metaname2)</span>
<span id="cb11-7"><a href="dataqc.html#cb11-7"></a>  <span class="kw">return</span>(p)</span>
<span id="cb11-8"><a href="dataqc.html#cb11-8"></a>}</span>
<span id="cb11-9"><a href="dataqc.html#cb11-9"></a><span class="co"># ----------------------------------------</span></span>
<span id="cb11-10"><a href="dataqc.html#cb11-10"></a><span class="co"># For `day1_norm`</span></span>
<span id="cb11-11"><a href="dataqc.html#cb11-11"></a><span class="co"># ----------------------------------------</span></span>
<span id="cb11-12"><a href="dataqc.html#cb11-12"></a>p &lt;-<span class="st"> </span><span class="kw">my_func</span>(<span class="dt">obj =</span> day1_norm, <span class="dt">metaname1 =</span> <span class="st">&quot;nReads&quot;</span>, <span class="dt">metaname2 =</span> <span class="st">&quot;nGenes&quot;</span>)</span>
<span id="cb11-13"><a href="dataqc.html#cb11-13"></a>filename &lt;-<span class="st"> &quot;figures/figure_01_0010.png&quot;</span></span>
<span id="cb11-14"><a href="dataqc.html#cb11-14"></a><span class="kw">ggsave</span>(<span class="dt">file =</span> filename, <span class="dt">plot =</span> p, <span class="dt">dpi =</span> <span class="dv">300</span>, <span class="dt">width =</span> <span class="dv">4</span>, <span class="dt">height =</span> <span class="dv">4</span>)</span>
<span id="cb11-15"><a href="dataqc.html#cb11-15"></a></span>
<span id="cb11-16"><a href="dataqc.html#cb11-16"></a>p &lt;-<span class="st"> </span><span class="kw">my_func</span>(<span class="dt">obj =</span> day1_norm, <span class="dt">metaname1 =</span> <span class="st">&quot;nGenes&quot;</span>, <span class="dt">metaname2 =</span> <span class="st">&quot;percent_MT&quot;</span>)</span>
<span id="cb11-17"><a href="dataqc.html#cb11-17"></a>filename &lt;-<span class="st"> &quot;figures/figure_01_0011.png&quot;</span></span>
<span id="cb11-18"><a href="dataqc.html#cb11-18"></a><span class="kw">ggsave</span>(<span class="dt">file =</span> filename, <span class="dt">plot =</span> p, <span class="dt">dpi =</span> <span class="dv">300</span>, <span class="dt">width =</span> <span class="dv">4</span>, <span class="dt">height =</span> <span class="dv">4</span>)</span></code></pre></div>
<p><img src="figures/figure_01_0010.png" width="150px">
<img src="figures/figure_01_0011.png" width="150px">
<img src="figures/figure_02_0010.png" width="150px">
<img src="figures/figure_02_0011.png" width="150px"></p>
<p><img src="figures/figure_03_0010.png" width="150px">
<img src="figures/figure_03_0011.png" width="150px">
<img src="figures/figure_04_0010.png" width="150px">
<img src="figures/figure_04_0011.png" width="150px"></p>
<p><img src="figures/figure_05_0010.png" width="150px">
<img src="figures/figure_05_0011.png" width="150px">
<img src="figures/figure_06_0010.png" width="150px">
<img src="figures/figure_06_0011.png" width="150px"></p>
<p>The following function <code>plot_metaData_wAnnotation()</code> outputs a scatter plot for
the meta data input, which is ranked according to their values.
The arguments are <code>vec</code> (one-dimensional vector of single-cell QC metrics,
e.g., one of <code>obj[["sample"]][["nReads"]]</code>, <code>obj[["sample"]][["nGenes"]]</code>),
<code>target_xRange</code> (x-axis interval in which one would like to find an
inflection point), <code>n</code> (number of dividing points for <code>target_xRange</code>),
<code>title</code>, <code>xlabel</code>, and <code>ylabel</code>.
Note that the inflection points are used for the cell filtering below.
Note that the calculated inflection points are approximated to the nearest
integers.</p>
<div class="sourceCode" id="cb12"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb12-1"><a href="dataqc.html#cb12-1"></a>my_func &lt;-<span class="st"> </span><span class="cf">function</span>(obj, metaname, target_xRange){</span>
<span id="cb12-2"><a href="dataqc.html#cb12-2"></a>  vec &lt;-<span class="st"> </span>obj[[<span class="st">&quot;sample&quot;</span>]][[metaname]]</span>
<span id="cb12-3"><a href="dataqc.html#cb12-3"></a>  p &lt;-<span class="st"> </span><span class="kw">plot_metaData_wAnnotation</span>(</span>
<span id="cb12-4"><a href="dataqc.html#cb12-4"></a>    <span class="dt">vec =</span> obj[[<span class="st">&quot;sample&quot;</span>]][[metaname]], <span class="dt">target_xRange =</span> target_xRange,</span>
<span id="cb12-5"><a href="dataqc.html#cb12-5"></a>    <span class="dt">n =</span> <span class="kw">round</span>(<span class="fl">0.1</span> <span class="op">*</span><span class="st"> </span>(target_xRange[<span class="dv">2</span>] <span class="op">-</span><span class="st"> </span>target_xRange[<span class="dv">1</span>])),</span>
<span id="cb12-6"><a href="dataqc.html#cb12-6"></a>    <span class="dt">title =</span> obj[[<span class="st">&quot;history&quot;</span>]][[<span class="st">&quot;make_asurat_obj&quot;</span>]][[<span class="st">&quot;obj_name&quot;</span>]],</span>
<span id="cb12-7"><a href="dataqc.html#cb12-7"></a>    <span class="dt">title_size =</span> <span class="dv">16</span>, <span class="dt">xlabel =</span> <span class="st">&quot;Ranked samples&quot;</span>, <span class="dt">ylabel =</span> metaname</span>
<span id="cb12-8"><a href="dataqc.html#cb12-8"></a>  )</span>
<span id="cb12-9"><a href="dataqc.html#cb12-9"></a>  <span class="kw">return</span>(p)</span>
<span id="cb12-10"><a href="dataqc.html#cb12-10"></a>}</span>
<span id="cb12-11"><a href="dataqc.html#cb12-11"></a><span class="co"># ----------------------------------------</span></span>
<span id="cb12-12"><a href="dataqc.html#cb12-12"></a><span class="co"># day1_norm</span></span>
<span id="cb12-13"><a href="dataqc.html#cb12-13"></a><span class="co"># ----------------------------------------</span></span>
<span id="cb12-14"><a href="dataqc.html#cb12-14"></a>target_xRange &lt;-<span class="st"> </span><span class="kw">c</span>(<span class="dv">100</span>, <span class="dv">10000</span>)</span>
<span id="cb12-15"><a href="dataqc.html#cb12-15"></a>p &lt;-<span class="st"> </span><span class="kw">my_func</span>(day1_norm, <span class="dt">metaname =</span> <span class="st">&quot;nReads&quot;</span>, <span class="dt">target_xRange =</span> target_xRange)</span>
<span id="cb12-16"><a href="dataqc.html#cb12-16"></a>filename &lt;-<span class="st"> &quot;figures/figure_01_0020.png&quot;</span></span>
<span id="cb12-17"><a href="dataqc.html#cb12-17"></a><span class="kw">ggsave</span>(<span class="dt">file =</span> filename, <span class="dt">plot =</span> p, <span class="dt">dpi =</span> <span class="dv">300</span>, <span class="dt">width =</span> <span class="dv">4</span>, <span class="dt">height =</span> <span class="fl">3.7</span>)</span>
<span id="cb12-18"><a href="dataqc.html#cb12-18"></a></span>
<span id="cb12-19"><a href="dataqc.html#cb12-19"></a>target_xRange &lt;-<span class="st"> </span><span class="kw">c</span>(<span class="dv">100</span>, <span class="dv">10000</span>)</span>
<span id="cb12-20"><a href="dataqc.html#cb12-20"></a>p &lt;-<span class="st"> </span><span class="kw">my_func</span>(day1_norm, <span class="dt">metaname =</span> <span class="st">&quot;nGenes&quot;</span>, <span class="dt">target_xRange =</span> target_xRange)</span>
<span id="cb12-21"><a href="dataqc.html#cb12-21"></a>filename &lt;-<span class="st"> &quot;figures/figure_01_0021.png&quot;</span></span>
<span id="cb12-22"><a href="dataqc.html#cb12-22"></a><span class="kw">ggsave</span>(<span class="dt">file =</span> filename, <span class="dt">plot =</span> p, <span class="dt">dpi =</span> <span class="dv">300</span>, <span class="dt">width =</span> <span class="dv">4</span>, <span class="dt">height =</span> <span class="fl">3.7</span>)</span></code></pre></div>
<p><img src="figures/figure_01_0020.png" width="150px">
<img src="figures/figure_01_0021.png" width="150px">
<img src="figures/figure_02_0020.png" width="150px">
<img src="figures/figure_02_0021.png" width="150px"></p>
<p><img src="figures/figure_03_0020.png" width="150px">
<img src="figures/figure_03_0021.png" width="150px">
<img src="figures/figure_04_0020.png" width="150px">
<img src="figures/figure_04_0021.png" width="150px"></p>
<p><img src="figures/figure_05_0020.png" width="150px">
<img src="figures/figure_05_0021.png" width="150px">
<img src="figures/figure_06_0020.png" width="150px">
<img src="figures/figure_06_0021.png" width="150px"></p>
<p>The following function <code>trim_samples()</code> filters out the low quality
cells by setting minimum and maximum values of the calculated QC metrics.</p>
<div class="sourceCode" id="cb13"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb13-1"><a href="dataqc.html#cb13-1"></a>day1_norm_red &lt;-<span class="st"> </span><span class="kw">trim_samples</span>(<span class="dt">obj =</span> day1_norm,</span>
<span id="cb13-2"><a href="dataqc.html#cb13-2"></a>                              <span class="dt">min_nReads =</span> <span class="dv">2606</span>, <span class="dt">max_nReads =</span> <span class="dv">30000</span>,</span>
<span id="cb13-3"><a href="dataqc.html#cb13-3"></a>                              <span class="dt">min_nGenes =</span> <span class="dv">993</span>, <span class="dt">max_nGenes =</span> <span class="fl">1e+10</span>,</span>
<span id="cb13-4"><a href="dataqc.html#cb13-4"></a>                              <span class="dt">min_percent_MT =</span> <span class="dv">0</span>, <span class="dt">max_percent_MT =</span> <span class="dv">14</span>)</span>
<span id="cb13-5"><a href="dataqc.html#cb13-5"></a>day7_hypo_red &lt;-<span class="st"> </span><span class="kw">trim_samples</span>(<span class="dt">obj =</span> day7_hypo,</span>
<span id="cb13-6"><a href="dataqc.html#cb13-6"></a>                              <span class="dt">min_nReads =</span> <span class="dv">2748</span>, <span class="dt">max_nReads =</span> <span class="dv">20000</span>,</span>
<span id="cb13-7"><a href="dataqc.html#cb13-7"></a>                              <span class="dt">min_nGenes =</span> <span class="dv">1205</span>, <span class="dt">max_nGenes =</span> <span class="fl">1e+10</span>,</span>
<span id="cb13-8"><a href="dataqc.html#cb13-8"></a>                              <span class="dt">min_percent_MT =</span> <span class="dv">0</span>, <span class="dt">max_percent_MT =</span> <span class="dv">12</span>)</span>
<span id="cb13-9"><a href="dataqc.html#cb13-9"></a>sc68_vehi_red &lt;-<span class="st"> </span><span class="kw">trim_samples</span>(<span class="dt">obj =</span> sc68_vehi,</span>
<span id="cb13-10"><a href="dataqc.html#cb13-10"></a>                              <span class="dt">min_nReads =</span> <span class="dv">4828</span>, <span class="dt">max_nReads =</span> <span class="dv">40000</span>,</span>
<span id="cb13-11"><a href="dataqc.html#cb13-11"></a>                              <span class="dt">min_nGenes =</span> <span class="dv">1363</span>, <span class="dt">max_nGenes =</span> <span class="fl">1e+10</span>,</span>
<span id="cb13-12"><a href="dataqc.html#cb13-12"></a>                              <span class="dt">min_percent_MT =</span> <span class="dv">0</span>, <span class="dt">max_percent_MT =</span> <span class="dv">13</span>)</span>
<span id="cb13-13"><a href="dataqc.html#cb13-13"></a>sc68_cisp_red &lt;-<span class="st"> </span><span class="kw">trim_samples</span>(<span class="dt">obj =</span> sc68_cisp,</span>
<span id="cb13-14"><a href="dataqc.html#cb13-14"></a>                              <span class="dt">min_nReads =</span> <span class="dv">1424</span>, <span class="dt">max_nReads =</span> <span class="dv">40000</span>,</span>
<span id="cb13-15"><a href="dataqc.html#cb13-15"></a>                              <span class="dt">min_nGenes =</span> <span class="dv">1138</span>, <span class="dt">max_nGenes =</span> <span class="fl">1e+10</span>,</span>
<span id="cb13-16"><a href="dataqc.html#cb13-16"></a>                              <span class="dt">min_percent_MT =</span> <span class="dv">0</span>, <span class="dt">max_percent_MT =</span> <span class="dv">15</span>)</span>
<span id="cb13-17"><a href="dataqc.html#cb13-17"></a>pbmc_<span class="dv">4000</span>_red &lt;-<span class="st"> </span><span class="kw">trim_samples</span>(<span class="dt">obj =</span> pbmc_<span class="dv">4000</span>,</span>
<span id="cb13-18"><a href="dataqc.html#cb13-18"></a>                              <span class="dt">min_nReads =</span> <span class="dv">2239</span>, <span class="dt">max_nReads =</span> <span class="dv">20000</span>,</span>
<span id="cb13-19"><a href="dataqc.html#cb13-19"></a>                              <span class="dt">min_nGenes =</span> <span class="dv">884</span>, <span class="dt">max_nGenes =</span> <span class="fl">1e+10</span>,</span>
<span id="cb13-20"><a href="dataqc.html#cb13-20"></a>                              <span class="dt">min_percent_MT =</span> <span class="dv">0</span>, <span class="dt">max_percent_MT =</span> <span class="dv">8</span>)</span>
<span id="cb13-21"><a href="dataqc.html#cb13-21"></a>pbmc_<span class="dv">6000</span>_red &lt;-<span class="st"> </span><span class="kw">trim_samples</span>(<span class="dt">obj =</span> pbmc_<span class="dv">6000</span>,</span>
<span id="cb13-22"><a href="dataqc.html#cb13-22"></a>                              <span class="dt">min_nReads =</span> <span class="dv">1016</span>, <span class="dt">max_nReads =</span> <span class="dv">7000</span>,</span>
<span id="cb13-23"><a href="dataqc.html#cb13-23"></a>                              <span class="dt">min_nGenes =</span> <span class="dv">457</span>, <span class="dt">max_nGenes =</span> <span class="fl">1e+10</span>,</span>
<span id="cb13-24"><a href="dataqc.html#cb13-24"></a>                              <span class="dt">min_percent_MT =</span> <span class="dv">0</span>, <span class="dt">max_percent_MT =</span> <span class="dv">8</span>)</span></code></pre></div>
<p>Save the objects.</p>
<div class="sourceCode" id="cb14"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb14-1"><a href="dataqc.html#cb14-1"></a><span class="kw">saveRDS</span>(day1_norm_red, <span class="dt">file =</span> <span class="st">&quot;backup/01_003_day1_norm_trim_samples.rds&quot;</span>)</span>
<span id="cb14-2"><a href="dataqc.html#cb14-2"></a><span class="kw">saveRDS</span>(day7_hypo_red, <span class="dt">file =</span> <span class="st">&quot;backup/02_003_day7_hypo_trim_samples.rds&quot;</span>)</span>
<span id="cb14-3"><a href="dataqc.html#cb14-3"></a><span class="kw">saveRDS</span>(sc68_vehi_red, <span class="dt">file =</span> <span class="st">&quot;backup/03_003_sc68_vehi_trim_samples.rds&quot;</span>)</span>
<span id="cb14-4"><a href="dataqc.html#cb14-4"></a><span class="kw">saveRDS</span>(sc68_cisp_red, <span class="dt">file =</span> <span class="st">&quot;backup/04_003_sc68_cisp_trim_samples.rds&quot;</span>)</span>
<span id="cb14-5"><a href="dataqc.html#cb14-5"></a><span class="kw">saveRDS</span>(pbmc_<span class="dv">4000</span>_red, <span class="dt">file =</span> <span class="st">&quot;backup/05_003_pbmc_4000_trim_samples.rds&quot;</span>)</span>
<span id="cb14-6"><a href="dataqc.html#cb14-6"></a><span class="kw">saveRDS</span>(pbmc_<span class="dv">6000</span>_red, <span class="dt">file =</span> <span class="st">&quot;backup/06_003_pbmc_6000_trim_samples.rds&quot;</span>)</span></code></pre></div>
</div>
<div id="variable-qc" class="section level2">
<h2><span class="header-section-number">3.3</span> Variable QC</h2>
<p>Load the data.</p>
<div class="sourceCode" id="cb15"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb15-1"><a href="dataqc.html#cb15-1"></a><span class="kw">source</span>(<span class="st">&quot;R/function_general.R&quot;</span>)</span>
<span id="cb15-2"><a href="dataqc.html#cb15-2"></a><span class="kw">source</span>(<span class="st">&quot;R/plot.R&quot;</span>)</span>
<span id="cb15-3"><a href="dataqc.html#cb15-3"></a>day1_norm &lt;-<span class="st"> </span><span class="kw">readRDS</span>(<span class="dt">file =</span> <span class="st">&quot;backup/01_003_day1_norm_trim_samples.rds&quot;</span>)</span>
<span id="cb15-4"><a href="dataqc.html#cb15-4"></a>day7_hypo &lt;-<span class="st"> </span><span class="kw">readRDS</span>(<span class="dt">file =</span> <span class="st">&quot;backup/02_003_day7_hypo_trim_samples.rds&quot;</span>)</span>
<span id="cb15-5"><a href="dataqc.html#cb15-5"></a>sc68_vehi &lt;-<span class="st"> </span><span class="kw">readRDS</span>(<span class="dt">file =</span> <span class="st">&quot;backup/03_003_sc68_vehi_trim_samples.rds&quot;</span>)</span>
<span id="cb15-6"><a href="dataqc.html#cb15-6"></a>sc68_cisp &lt;-<span class="st"> </span><span class="kw">readRDS</span>(<span class="dt">file =</span> <span class="st">&quot;backup/04_003_sc68_cisp_trim_samples.rds&quot;</span>)</span>
<span id="cb15-7"><a href="dataqc.html#cb15-7"></a>pbmc_<span class="dv">4000</span> &lt;-<span class="st"> </span><span class="kw">readRDS</span>(<span class="dt">file =</span> <span class="st">&quot;backup/05_003_pbmc_4000_trim_samples.rds&quot;</span>)</span>
<span id="cb15-8"><a href="dataqc.html#cb15-8"></a>pbmc_<span class="dv">6000</span> &lt;-<span class="st"> </span><span class="kw">readRDS</span>(<span class="dt">file =</span> <span class="st">&quot;backup/06_003_pbmc_6000_trim_samples.rds&quot;</span>)</span></code></pre></div>
<p>Check the size of the gene-by-sample matrices.</p>
<div class="sourceCode" id="cb16"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb16-1"><a href="dataqc.html#cb16-1"></a><span class="kw">rbind</span>(<span class="kw">dim</span>(day1_norm[[<span class="st">&quot;data&quot;</span>]][[<span class="st">&quot;raw&quot;</span>]]), <span class="kw">dim</span>(day7_hypo[[<span class="st">&quot;data&quot;</span>]][[<span class="st">&quot;raw&quot;</span>]]),</span>
<span id="cb16-2"><a href="dataqc.html#cb16-2"></a>      <span class="kw">dim</span>(sc68_vehi[[<span class="st">&quot;data&quot;</span>]][[<span class="st">&quot;raw&quot;</span>]]), <span class="kw">dim</span>(sc68_cisp[[<span class="st">&quot;data&quot;</span>]][[<span class="st">&quot;raw&quot;</span>]]),</span>
<span id="cb16-3"><a href="dataqc.html#cb16-3"></a>      <span class="kw">dim</span>(pbmc_<span class="dv">4000</span>[[<span class="st">&quot;data&quot;</span>]][[<span class="st">&quot;raw&quot;</span>]]), <span class="kw">dim</span>(pbmc_<span class="dv">6000</span>[[<span class="st">&quot;data&quot;</span>]][[<span class="st">&quot;raw&quot;</span>]]))</span></code></pre></div>
<pre><code>[1,] 16820 3663
[2,] 15795 1947
[3,] 16784 3923
[4,] 15605 2285
[5,] 13829 3815
[6,] 11946 4878</code></pre>
<p>The following function <code>plot_meanReads_wAnnotation()</code> outputs a scatter plot for
mean of number of reads for each gene, which is ranked according to their
values.
The arguments are <code>obj</code>, <code>threshold</code> (set by trial and error),
<code>digits</code> (number of digits after the decimal point for the annotated value),
<code>title</code>, <code>xlabel</code>, and <code>ylabel</code>.</p>
<div class="sourceCode" id="cb18"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb18-1"><a href="dataqc.html#cb18-1"></a>my_func &lt;-<span class="st"> </span><span class="cf">function</span>(obj, threshold){</span>
<span id="cb18-2"><a href="dataqc.html#cb18-2"></a>  p &lt;-<span class="st"> </span><span class="kw">plot_meanReads_wAnnotation</span>(</span>
<span id="cb18-3"><a href="dataqc.html#cb18-3"></a>    <span class="dt">obj =</span> obj, <span class="dt">threshold =</span> threshold, <span class="dt">digits =</span> <span class="dv">2</span>,</span>
<span id="cb18-4"><a href="dataqc.html#cb18-4"></a>    <span class="dt">title =</span> obj[[<span class="st">&quot;history&quot;</span>]][[<span class="st">&quot;make_asurat_obj&quot;</span>]][[<span class="st">&quot;obj_name&quot;</span>]],</span>
<span id="cb18-5"><a href="dataqc.html#cb18-5"></a>    <span class="dt">title_size =</span> <span class="dv">16</span>, <span class="dt">xlabel =</span> <span class="st">&quot;Ranked genes&quot;</span>, <span class="dt">ylabel =</span> <span class="st">&quot;Mean of raw read counts&quot;</span></span>
<span id="cb18-6"><a href="dataqc.html#cb18-6"></a>  )</span>
<span id="cb18-7"><a href="dataqc.html#cb18-7"></a>  <span class="kw">return</span>(p)</span>
<span id="cb18-8"><a href="dataqc.html#cb18-8"></a>}</span>
<span id="cb18-9"><a href="dataqc.html#cb18-9"></a><span class="co"># ----------------------------------------</span></span>
<span id="cb18-10"><a href="dataqc.html#cb18-10"></a><span class="co"># day1_norm</span></span>
<span id="cb18-11"><a href="dataqc.html#cb18-11"></a><span class="co"># ----------------------------------------</span></span>
<span id="cb18-12"><a href="dataqc.html#cb18-12"></a>p &lt;-<span class="st"> </span><span class="kw">my_func</span>(<span class="dt">obj =</span> day1_norm, <span class="dt">threshold =</span> <span class="fl">0.2</span>)</span>
<span id="cb18-13"><a href="dataqc.html#cb18-13"></a>filename &lt;-<span class="st"> &quot;figures/figure_01_0030.png&quot;</span></span>
<span id="cb18-14"><a href="dataqc.html#cb18-14"></a><span class="kw">ggsave</span>(<span class="dt">file =</span> filename, <span class="dt">plot =</span> p, <span class="dt">dpi =</span> <span class="dv">300</span>, <span class="dt">width =</span> <span class="dv">4</span>, <span class="dt">height =</span> <span class="fl">3.7</span>)</span></code></pre></div>
<p><img src="figures/figure_01_0030.png" width="150px">
<img src="figures/figure_02_0030.png" width="150px">
<img src="figures/figure_03_0030.png" width="150px">
<img src="figures/figure_04_0030.png" width="150px"></p>
<p><img src="figures/figure_05_0030.png" width="150px">
<img src="figures/figure_06_0030.png" width="150px"></p>
<p>The following function <code>trim_variables()</code> filters out the low quality
genes by setting a minimum value as <code>min_meanReads</code>.</p>
<div class="sourceCode" id="cb19"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb19-1"><a href="dataqc.html#cb19-1"></a>day1_norm_red &lt;-<span class="st"> </span><span class="kw">trim_variables</span>(<span class="dt">obj =</span> day1_norm, <span class="dt">min_meanReads =</span> <span class="fl">0.2</span>)</span>
<span id="cb19-2"><a href="dataqc.html#cb19-2"></a>day7_hypo_red &lt;-<span class="st"> </span><span class="kw">trim_variables</span>(<span class="dt">obj =</span> day7_hypo, <span class="dt">min_meanReads =</span> <span class="fl">0.2</span>)</span>
<span id="cb19-3"><a href="dataqc.html#cb19-3"></a>sc68_vehi_red &lt;-<span class="st"> </span><span class="kw">trim_variables</span>(<span class="dt">obj =</span> sc68_vehi, <span class="dt">min_meanReads =</span> <span class="fl">0.2</span>)</span>
<span id="cb19-4"><a href="dataqc.html#cb19-4"></a>sc68_cisp_red &lt;-<span class="st"> </span><span class="kw">trim_variables</span>(<span class="dt">obj =</span> sc68_cisp, <span class="dt">min_meanReads =</span> <span class="fl">0.2</span>)</span>
<span id="cb19-5"><a href="dataqc.html#cb19-5"></a>pbmc_<span class="dv">4000</span>_red &lt;-<span class="st"> </span><span class="kw">trim_variables</span>(<span class="dt">obj =</span> pbmc_<span class="dv">4000</span>, <span class="dt">min_meanReads =</span> <span class="fl">0.05</span>)</span>
<span id="cb19-6"><a href="dataqc.html#cb19-6"></a>pbmc_<span class="dv">6000</span>_red &lt;-<span class="st"> </span><span class="kw">trim_variables</span>(<span class="dt">obj =</span> pbmc_<span class="dv">6000</span>, <span class="dt">min_meanReads =</span> <span class="fl">0.03</span>)</span></code></pre></div>
<p>Check the scatter plots between QC metrics for each cell</p>
<div class="sourceCode" id="cb20"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb20-1"><a href="dataqc.html#cb20-1"></a>my_func &lt;-<span class="st"> </span><span class="cf">function</span>(obj, metaname1, metaname2){</span>
<span id="cb20-2"><a href="dataqc.html#cb20-2"></a>  df &lt;-<span class="st"> </span><span class="kw">data.frame</span>(<span class="dt">x =</span> obj[[<span class="st">&quot;sample&quot;</span>]][[metaname1]],</span>
<span id="cb20-3"><a href="dataqc.html#cb20-3"></a>                   <span class="dt">y =</span> obj[[<span class="st">&quot;sample&quot;</span>]][[metaname2]])</span>
<span id="cb20-4"><a href="dataqc.html#cb20-4"></a>  p &lt;-<span class="st"> </span><span class="kw">plot_metaData</span>(<span class="dt">df =</span> df,</span>
<span id="cb20-5"><a href="dataqc.html#cb20-5"></a>                    <span class="dt">title =</span> obj[[<span class="st">&quot;history&quot;</span>]][[<span class="st">&quot;make_asurat_obj&quot;</span>]][[<span class="st">&quot;obj_name&quot;</span>]],</span>
<span id="cb20-6"><a href="dataqc.html#cb20-6"></a>                    <span class="dt">title_size =</span> <span class="dv">16</span>, <span class="dt">xlabel =</span> metaname1, <span class="dt">ylabel =</span> metaname2)</span>
<span id="cb20-7"><a href="dataqc.html#cb20-7"></a>  <span class="kw">return</span>(p)</span>
<span id="cb20-8"><a href="dataqc.html#cb20-8"></a>}</span>
<span id="cb20-9"><a href="dataqc.html#cb20-9"></a><span class="co"># ----------------------------------------</span></span>
<span id="cb20-10"><a href="dataqc.html#cb20-10"></a><span class="co"># For `day1_norm`</span></span>
<span id="cb20-11"><a href="dataqc.html#cb20-11"></a><span class="co"># ----------------------------------------</span></span>
<span id="cb20-12"><a href="dataqc.html#cb20-12"></a>p &lt;-<span class="st"> </span><span class="kw">my_func</span>(day1_norm_red, <span class="dt">metaname1 =</span> <span class="st">&quot;nReads&quot;</span>, <span class="dt">metaname2 =</span> <span class="st">&quot;nGenes&quot;</span>)</span>
<span id="cb20-13"><a href="dataqc.html#cb20-13"></a>filename &lt;-<span class="st"> &quot;figures/figure_01_0035.png&quot;</span></span>
<span id="cb20-14"><a href="dataqc.html#cb20-14"></a><span class="kw">ggsave</span>(<span class="dt">file =</span> filename, <span class="dt">plot =</span> p, <span class="dt">dpi =</span> <span class="dv">300</span>, <span class="dt">width =</span> <span class="dv">4</span>, <span class="dt">height =</span> <span class="dv">4</span>)</span>
<span id="cb20-15"><a href="dataqc.html#cb20-15"></a></span>
<span id="cb20-16"><a href="dataqc.html#cb20-16"></a>p &lt;-<span class="st"> </span><span class="kw">my_func</span>(day1_norm_red, <span class="dt">metaname1 =</span> <span class="st">&quot;nGenes&quot;</span>, <span class="dt">metaname2 =</span> <span class="st">&quot;percent_MT&quot;</span>)</span>
<span id="cb20-17"><a href="dataqc.html#cb20-17"></a>filename &lt;-<span class="st"> &quot;figures/figure_01_0036.png&quot;</span></span>
<span id="cb20-18"><a href="dataqc.html#cb20-18"></a><span class="kw">ggsave</span>(<span class="dt">file =</span> filename, <span class="dt">plot =</span> p, <span class="dt">dpi =</span> <span class="dv">300</span>, <span class="dt">width =</span> <span class="dv">4</span>, <span class="dt">height =</span> <span class="dv">4</span>)</span></code></pre></div>
<p><img src="figures/figure_01_0035.png" width="150px">
<img src="figures/figure_01_0036.png" width="150px">
<img src="figures/figure_02_0035.png" width="150px">
<img src="figures/figure_02_0036.png" width="150px"></p>
<p><img src="figures/figure_03_0035.png" width="150px">
<img src="figures/figure_03_0036.png" width="150px">
<img src="figures/figure_04_0035.png" width="150px">
<img src="figures/figure_04_0036.png" width="150px"></p>
<p><img src="figures/figure_05_0035.png" width="150px">
<img src="figures/figure_05_0036.png" width="150px">
<img src="figures/figure_06_0035.png" width="150px">
<img src="figures/figure_06_0036.png" width="150px"></p>
<p>Check the size of the gene-by-sample matrices.</p>
<div class="sourceCode" id="cb21"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb21-1"><a href="dataqc.html#cb21-1"></a><span class="kw">rbind</span>(<span class="kw">dim</span>(day1_norm_red[[<span class="st">&quot;data&quot;</span>]][[<span class="st">&quot;raw&quot;</span>]]),</span>
<span id="cb21-2"><a href="dataqc.html#cb21-2"></a>      <span class="kw">dim</span>(day7_hypo_red[[<span class="st">&quot;data&quot;</span>]][[<span class="st">&quot;raw&quot;</span>]]),</span>
<span id="cb21-3"><a href="dataqc.html#cb21-3"></a>      <span class="kw">dim</span>(sc68_vehi_red[[<span class="st">&quot;data&quot;</span>]][[<span class="st">&quot;raw&quot;</span>]]),</span>
<span id="cb21-4"><a href="dataqc.html#cb21-4"></a>      <span class="kw">dim</span>(sc68_cisp_red[[<span class="st">&quot;data&quot;</span>]][[<span class="st">&quot;raw&quot;</span>]]),</span>
<span id="cb21-5"><a href="dataqc.html#cb21-5"></a>      <span class="kw">dim</span>(pbmc_<span class="dv">4000</span>_red[[<span class="st">&quot;data&quot;</span>]][[<span class="st">&quot;raw&quot;</span>]]),</span>
<span id="cb21-6"><a href="dataqc.html#cb21-6"></a>      <span class="kw">dim</span>(pbmc_<span class="dv">6000</span>_red[[<span class="st">&quot;data&quot;</span>]][[<span class="st">&quot;raw&quot;</span>]]))</span></code></pre></div>
<pre><code>[1,] 5272 3663
[2,] 6213 1947
[3,] 6581 3923
[4,] 6347 2285
[5,] 6658 3815
[6,] 5169 4878</code></pre>
<p>Save the objects.</p>
<div class="sourceCode" id="cb23"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb23-1"><a href="dataqc.html#cb23-1"></a><span class="kw">saveRDS</span>(day1_norm_red, <span class="dt">file =</span> <span class="st">&quot;backup/01_004_day1_norm_trim_variables.rds&quot;</span>)</span>
<span id="cb23-2"><a href="dataqc.html#cb23-2"></a><span class="kw">saveRDS</span>(day7_hypo_red, <span class="dt">file =</span> <span class="st">&quot;backup/02_004_day7_hypo_trim_variables.rds&quot;</span>)</span>
<span id="cb23-3"><a href="dataqc.html#cb23-3"></a><span class="kw">saveRDS</span>(sc68_vehi_red, <span class="dt">file =</span> <span class="st">&quot;backup/03_004_sc68_vehi_trim_variables.rds&quot;</span>)</span>
<span id="cb23-4"><a href="dataqc.html#cb23-4"></a><span class="kw">saveRDS</span>(sc68_cisp_red, <span class="dt">file =</span> <span class="st">&quot;backup/04_004_sc68_cisp_trim_variables.rds&quot;</span>)</span>
<span id="cb23-5"><a href="dataqc.html#cb23-5"></a><span class="kw">saveRDS</span>(pbmc_<span class="dv">4000</span>_red, <span class="dt">file =</span> <span class="st">&quot;backup/05_004_pbmc_4000_trim_variables.rds&quot;</span>)</span>
<span id="cb23-6"><a href="dataqc.html#cb23-6"></a><span class="kw">saveRDS</span>(pbmc_<span class="dv">6000</span>_red, <span class="dt">file =</span> <span class="st">&quot;backup/06_004_pbmc_6000_trim_variables.rds&quot;</span>)</span></code></pre></div>

</div>
</div>
            </section>

          </div>
        </div>
      </div>
<a href="dataset.html" class="navigation navigation-prev " aria-label="Previous page"><i class="fa fa-angle-left"></i></a>
<a href="normalization.html" class="navigation navigation-next " aria-label="Next page"><i class="fa fa-angle-right"></i></a>
    </div>
  </div>
<script src="libs/gitbook-2.6.7/js/app.min.js"></script>
<script src="libs/gitbook-2.6.7/js/lunr.js"></script>
<script src="libs/gitbook-2.6.7/js/clipboard.min.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-search.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-sharing.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-fontsettings.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-bookdown.js"></script>
<script src="libs/gitbook-2.6.7/js/jquery.highlight.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-clipboard.js"></script>
<script>
gitbook.require(["gitbook"], function(gitbook) {
gitbook.start({
"sharing": {
"github": false,
"facebook": true,
"twitter": true,
"linkedin": false,
"weibo": false,
"instapaper": false,
"vk": false,
"all": ["facebook", "twitter", "linkedin", "weibo", "instapaper"]
},
"fontsettings": {
"theme": "white",
"family": "sans",
"size": 2
},
"edit": {
"link": null,
"text": null
},
"history": {
"link": null,
"text": null
},
"view": {
"link": null,
"text": null
},
"download": ["Introduction to ASURAT (ver. 0.0.0.9000).pdf", "Introduction to ASURAT (ver. 0.0.0.9000).epub"],
"toc": {
"collapse": "subsection"
}
});
});
</script>

</body>

</html>
