<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>ASURAT • ASURAT</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="ASURAT">
<meta property="og:description" content="ASURAT">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    

    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">ASURAT</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">0.99.1</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../articles/ASURAT.html">Get started</a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right"></ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>ASURAT</h1>
                        <h4 data-toc-skip class="author">Keita Iida</h4>
            
            <h4 data-toc-skip class="date">2022-03-02</h4>
      
      
      <div class="hidden name"><code>ASURAT.Rmd</code></div>

    </div>

    
    
<div class="section level2">
<h2 id="installations">Installations<a class="anchor" aria-label="anchor" href="#installations"></a>
</h2>
<p>Attach necessary libraries:</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">ASURAT</span><span class="op">)</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">SingleCellExperiment</span><span class="op">)</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://bioconductor.org/packages/SummarizedExperiment" class="external-link">SummarizedExperiment</a></span><span class="op">)</span></code></pre></div>
<p><br></p>
</div>
<div class="section level2">
<h2 id="goal">Goal<a class="anchor" aria-label="anchor" href="#goal"></a>
</h2>
<p>A goal of ASURAT is to cluster and characterize individual samples (cells) in terms of cell type (or disease), biological function, signaling pathway activity, and so on (see <a href="#multilayered">here</a>).</p>
<p><br></p>
</div>
<div class="section level2">
<h2 id="quick-start-by-singlecellexperiment-objects">Quick start by SingleCellExperiment objects<a class="anchor" aria-label="anchor" href="#quick-start-by-singlecellexperiment-objects"></a>
</h2>
<p>Having a SingleCellExperiment object (e.g., <code>sce</code>), one can use ASURAT by confirming the following requirements:</p>
<ul>
<li>
<code>assays(sce)</code> contains gene expression data with row and column names as variable (gene) and sample (cell), respectively,</li>
</ul>
<p>If <code>sce</code> contains normalized expression data (e.g., <code>assay(sce, "logcounts")</code>), set <code>assay(sce, "centered")</code> by subtracting the data with the mean expression levels across samples (cells).</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">mat</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/matrix.html" class="external-link">as.matrix</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">assay</a></span><span class="op">(</span><span class="va">sce</span>, <span class="st">"logcounts"</span><span class="op">)</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">assay</a></span><span class="op">(</span><span class="va">sce</span>, <span class="st">"centered"</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/sweep.html" class="external-link">sweep</a></span><span class="op">(</span><span class="va">mat</span>, <span class="fl">1</span>, <span class="fu"><a href="https://rdrr.io/r/base/apply.html" class="external-link">apply</a></span><span class="op">(</span><span class="va">mat</span>, <span class="fl">1</span>, <span class="va">mean</span><span class="op">)</span>, FUN <span class="op">=</span> <span class="st">"-"</span><span class="op">)</span></code></pre></div>
<p>One may use a Seurat function <code><a href="https://satijalab.org/seurat/reference/as.SingleCellExperiment.html" class="external-link">Seurat::as.SingleCellExperiment()</a></code> for converting Seurat objects into SingleCellExperiment ones.</p>
<p>Now, ready for the next step <a href="#sign">here</a>.</p>
<p><br></p>
</div>
<div class="section level2">
<h2 id="preprocessing">Preprocessing<a class="anchor" aria-label="anchor" href="#preprocessing"></a>
</h2>
<div class="section level3">
<h3 id="prepare-singlecellexperiment-objects">Prepare SingleCellExperiment objects<a class="anchor" aria-label="anchor" href="#prepare-singlecellexperiment-objects"></a>
</h3>
<p>Load single-cell RNA sequencing (scRNA-seq) data.</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">sce</span> <span class="op">&lt;-</span> <span class="fu">TENxPBMCData</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/TENxPBMCData/man/TENxPBMCData.html" class="external-link">TENxPBMCData</a></span><span class="op">(</span>dataset <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"pbmc4k"</span><span class="op">)</span><span class="op">)</span>
<span class="va">pbmc_counts</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/matrix.html" class="external-link">as.matrix</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">assay</a></span><span class="op">(</span><span class="va">sce</span>, <span class="st">"counts"</span><span class="op">)</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">pbmc_counts</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">rowData</a></span><span class="op">(</span><span class="va">sce</span><span class="op">)</span><span class="op">$</span><span class="va">Symbol_TENx</span>
<span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">pbmc_counts</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">colData</a></span><span class="op">(</span><span class="va">sce</span><span class="op">)</span><span class="op">$</span><span class="va">Barcode</span></code></pre></div>
<p>Here, <code>pbmc_counts</code> is a read count table of peripheral blood mononuclear cells (PBMCs).</p>
<p>Below is a head of <code>pbmc_counts</code>:</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">pbmc_counts</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">5</span>, <span class="fl">1</span><span class="op">:</span><span class="fl">3</span><span class="op">]</span>
<span class="co">#&gt;              AAACCTGAGAAGGCCT-1 AAACCTGAGACAGACC-1 AAACCTGAGATAGTCA-1</span>
<span class="co">#&gt; RP11-34P13.3                  0                  0                  0</span>
<span class="co">#&gt; FAM138A                       0                  0                  0</span>
<span class="co">#&gt; OR4F5                         0                  0                  0</span>
<span class="co">#&gt; RP11-34P13.7                  0                  0                  0</span>
<span class="co">#&gt; RP11-34P13.8                  0                  0                  0</span></code></pre></div>
<p>Create SingleCellExperiment objects by inputting gene expression data.</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">pbmc</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/SingleCellExperiment/man/SingleCellExperiment.html" class="external-link">SingleCellExperiment</a></span><span class="op">(</span>assays <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>counts <span class="op">=</span> <span class="va">pbmc_counts</span><span class="op">)</span>,
                             rowData <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html" class="external-link">data.frame</a></span><span class="op">(</span>gene <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">pbmc_counts</span><span class="op">)</span><span class="op">)</span>,
                             colData <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html" class="external-link">data.frame</a></span><span class="op">(</span>cell <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">pbmc_counts</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></code></pre></div>
<p>Check data sizes.</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/base/dim.html" class="external-link">dim</a></span><span class="op">(</span><span class="va">pbmc</span><span class="op">)</span>
<span class="co">#&gt; [1] 33694  4340</span></code></pre></div>
<p><br></p>
</div>
<div class="section level3">
<h3 id="control-data-quality">Control data quality<a class="anchor" aria-label="anchor" href="#control-data-quality"></a>
</h3>
<p>Remove variables (genes) and samples (cells) with low quality, by processing the following three steps:</p>
<ol style="list-style-type: decimal">
<li>remove variables based on expression profiles across samples,</li>
<li>remove samples based on the numbers of reads and nonzero expressed variables,</li>
<li>remove variables based on the mean read counts across samples.</li>
</ol>
<p>First of all, add metadata for both variables and samples using ASURAT function <code><a href="../reference/add_metadata.html">add_metadata()</a></code>.</p>
<p>The arguments are</p>
<ul>
<li>
<code>sce</code>: SingleCellExperiment object, and</li>
<li>
<code>mitochondria_symbol</code>: a string representing for mitochondrial genes.</li>
</ul>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">pbmc</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/add_metadata.html">add_metadata</a></span><span class="op">(</span>sce <span class="op">=</span> <span class="va">pbmc</span>, mitochondria_symbol <span class="op">=</span> <span class="st">"^MT-"</span><span class="op">)</span></code></pre></div>
<p>One can check the results in <code>rowData(sce)</code> and <code>colData(sce)</code> slots.</p>
<p><br></p>
<div class="section level4">
<h4 id="remove-variables-based-on-expression-profiles">Remove variables based on expression profiles<a class="anchor" aria-label="anchor" href="#remove-variables-based-on-expression-profiles"></a>
</h4>
<p>ASURAT function <code><a href="../reference/remove_variables.html">remove_variables()</a></code> removes variable (gene) data such that the numbers of non-zero expressing samples (cells) are less than <code>min_nsamples</code>.</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">pbmc</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/remove_variables.html">remove_variables</a></span><span class="op">(</span>sce <span class="op">=</span> <span class="va">pbmc</span>, min_nsamples <span class="op">=</span> <span class="fl">10</span><span class="op">)</span></code></pre></div>
<p><br></p>
</div>
<div class="section level4">
<h4 id="remove-samples-based-on-expression-profiles">Remove samples based on expression profiles<a class="anchor" aria-label="anchor" href="#remove-samples-based-on-expression-profiles"></a>
</h4>
<p>Qualities of sample (cell) data are confirmed based on proper visualization of <code>colData(sce)</code>. ASURAT function <code><a href="../reference/plot_dataframe2D.html">plot_dataframe2D()</a></code> shows scatter plots of two-dimensional data (see <a href="#visualization_lowdim">here</a> for details).</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">dataframe2D</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html" class="external-link">data.frame</a></span><span class="op">(</span>x <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">colData</a></span><span class="op">(</span><span class="va">pbmc</span><span class="op">)</span><span class="op">$</span><span class="va">nReads</span>, y <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">colData</a></span><span class="op">(</span><span class="va">pbmc</span><span class="op">)</span><span class="op">$</span><span class="va">nGenes</span><span class="op">)</span>
<span class="fu"><a href="../reference/plot_dataframe2D.html">plot_dataframe2D</a></span><span class="op">(</span>dataframe2D <span class="op">=</span> <span class="va">dataframe2D</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu">ggplot2</span><span class="fu">::</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">labs</a></span><span class="op">(</span>x <span class="op">=</span> <span class="st">"Number of reads"</span>, y <span class="op">=</span> <span class="st">"Number of genes"</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu">ggplot2</span><span class="fu">::</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html" class="external-link">theme_classic</a></span><span class="op">(</span>base_size <span class="op">=</span> <span class="fl">20</span><span class="op">)</span></code></pre></div>
<p><img src="ASURAT_files/figure-html/unnamed-chunk-10-1.png" width="700"></p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">dataframe2D</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html" class="external-link">data.frame</a></span><span class="op">(</span>x <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">colData</a></span><span class="op">(</span><span class="va">pbmc</span><span class="op">)</span><span class="op">$</span><span class="va">nReads</span>, y <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">colData</a></span><span class="op">(</span><span class="va">pbmc</span><span class="op">)</span><span class="op">$</span><span class="va">percMT</span><span class="op">)</span>
<span class="fu"><a href="../reference/plot_dataframe2D.html">plot_dataframe2D</a></span><span class="op">(</span>dataframe2D <span class="op">=</span> <span class="va">dataframe2D</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu">ggplot2</span><span class="fu">::</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">labs</a></span><span class="op">(</span>x <span class="op">=</span> <span class="st">"Number of reads"</span>, y <span class="op">=</span> <span class="st">"Perc of MT reads"</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu">ggplot2</span><span class="fu">::</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html" class="external-link">theme_classic</a></span><span class="op">(</span>base_size <span class="op">=</span> <span class="fl">20</span><span class="op">)</span></code></pre></div>
<p><img src="ASURAT_files/figure-html/unnamed-chunk-11-1.png" width="700"></p>
<p>ASURAT function <code><a href="../reference/remove_samples.html">remove_samples()</a></code> removes sample (cell) data by setting cutoff values for the metadata.</p>
<p>The arguments are</p>
<ul>
<li>
<code>sce</code>: SingleCellExperiment object,</li>
<li>
<code>min_nReads</code> and <code>max_nReads</code>: minimum and maximum number of reads,</li>
<li>
<code>min_nGenes</code> and <code>max_nGenes</code>: minimum and maximum number of non-zero expressed genes, and</li>
<li>
<code>min_percMT</code> and <code>max_percMT</code>: minimum and maximum percent of reads that map to mitochondrial genes, respectively. If there is no mitochondrial genes, set them as <code>NULL</code>.</li>
</ul>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">pbmc</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/remove_samples.html">remove_samples</a></span><span class="op">(</span>sce <span class="op">=</span> <span class="va">pbmc</span>, min_nReads <span class="op">=</span> <span class="fl">2000</span>, max_nReads <span class="op">=</span> <span class="fl">20000</span>,
                       min_nGenes <span class="op">=</span> <span class="fl">500</span>, max_nGenes <span class="op">=</span> <span class="fl">5000</span>,
                       min_percMT <span class="op">=</span> <span class="fl">0</span>, max_percMT <span class="op">=</span> <span class="fl">10</span><span class="op">)</span></code></pre></div>
<p><br></p>
</div>
<div class="section level4">
<h4 id="remove-variables-based-on-the-mean-read-counts">Remove variables based on the mean read counts<a class="anchor" aria-label="anchor" href="#remove-variables-based-on-the-mean-read-counts"></a>
</h4>
<p>Qualities of variable (gene) data are confirmed based on proper visualization of <code>rowData(sce)</code>. ASURAT function <code><a href="../reference/plot_dataframe2D.html">plot_dataframe2D()</a></code> shows scatter plots of two-dimensional data (see <a href="#visualization_lowdim">here</a> for details).</p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">dataframe2D</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html" class="external-link">data.frame</a></span><span class="op">(</span>x <span class="op">=</span> <span class="fl">1</span><span class="op">:</span><span class="fu"><a href="https://rdrr.io/r/base/nrow.html" class="external-link">nrow</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">rowData</a></span><span class="op">(</span><span class="va">pbmc</span><span class="op">)</span><span class="op">)</span>,
                          y <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/sort.html" class="external-link">sort</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">rowData</a></span><span class="op">(</span><span class="va">pbmc</span><span class="op">)</span><span class="op">$</span><span class="va">nSamples</span>, decreasing <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span><span class="op">)</span>
<span class="fu"><a href="../reference/plot_dataframe2D.html">plot_dataframe2D</a></span><span class="op">(</span>dataframe2D <span class="op">=</span> <span class="va">dataframe2D</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu">ggplot2</span><span class="fu">::</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">labs</a></span><span class="op">(</span>x <span class="op">=</span> <span class="st">"Rank of genes"</span>, y <span class="op">=</span> <span class="st">"Mean read counts"</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu">ggplot2</span><span class="fu">::</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html" class="external-link">theme_classic</a></span><span class="op">(</span>base_size <span class="op">=</span> <span class="fl">20</span><span class="op">)</span></code></pre></div>
<p><img src="ASURAT_files/figure-html/unnamed-chunk-13-1.png" width="700"></p>
<p>ASURAT function <code><a href="../reference/remove_variables_second.html">remove_variables_second()</a></code> removes variable (gene) data such that the mean read counts across samples are less than <code>min_meannReads</code>.</p>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">pbmc</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/remove_variables_second.html">remove_variables_second</a></span><span class="op">(</span>sce <span class="op">=</span> <span class="va">pbmc</span>, min_meannReads <span class="op">=</span> <span class="fl">0.05</span><span class="op">)</span></code></pre></div>
<p><br></p>
</div>
</div>
<div class="section level3">
<h3 id="normalize-data">Normalize data<a class="anchor" aria-label="anchor" href="#normalize-data"></a>
</h3>
<p>Normalize data using bayNorm functions (Tang et al., Bioinformatics, 2020) and log transformation with a pseudo count. Then, we center the log-normalized data by subtracting with the mean expression levels across samples (cells). The resulting normalized-and-centered data are used for downstream analyses.</p>
<p>Perform <code>bayNorm()</code> for attenuating technical biases with respect to zero inflation and variation of capture efficiencies between samples (cells).</p>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">bayout</span> <span class="op">&lt;-</span> <span class="fu">bayNorm</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/bayNorm/man/bayNorm.html" class="external-link">bayNorm</a></span><span class="op">(</span>Data <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/SingleCellExperiment/man/assays.html" class="external-link">counts</a></span><span class="op">(</span><span class="va">pbmc</span><span class="op">)</span>, mode_version <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">assay</a></span><span class="op">(</span><span class="va">pbmc</span>, <span class="st">"normalized"</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="va">bayout</span><span class="op">$</span><span class="va">Bay_out</span></code></pre></div>
<p>Perform log-normalization with a pseudo count.</p>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">assay</a></span><span class="op">(</span><span class="va">pbmc</span>, <span class="st">"logcounts"</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/Log.html" class="external-link">log</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">assay</a></span><span class="op">(</span><span class="va">pbmc</span>, <span class="st">"normalized"</span><span class="op">)</span> <span class="op">+</span> <span class="fl">1</span><span class="op">)</span></code></pre></div>
<p>Center row data.</p>
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">mat</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">assay</a></span><span class="op">(</span><span class="va">pbmc</span>, <span class="st">"logcounts"</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">assay</a></span><span class="op">(</span><span class="va">pbmc</span>, <span class="st">"centered"</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/sweep.html" class="external-link">sweep</a></span><span class="op">(</span><span class="va">mat</span>, <span class="fl">1</span>, <span class="fu"><a href="https://rdrr.io/r/base/apply.html" class="external-link">apply</a></span><span class="op">(</span><span class="va">mat</span>, <span class="fl">1</span>, <span class="va">mean</span><span class="op">)</span>, FUN <span class="op">=</span> <span class="st">"-"</span><span class="op">)</span></code></pre></div>
<p><br></p>
</div>
</div>
<div class="section level2">
<h2 id="sign">Multifaceted sign analysis<a class="anchor" aria-label="anchor" href="#sign"></a>
</h2>
<p>Infer cell or disease types, biological functions, and signaling pathway activity at the single-cell level by inputting related databases.</p>
<p>ASURAT transforms centered read count tables to functional feature matrices, termed sign-by-sample matrices (SSMs). Using SSMs, perform unsupervised clustering of samples (cells).</p>
<p><br></p>
<div class="section level3">
<h3 id="load-databases">Load databases<a class="anchor" aria-label="anchor" href="#load-databases"></a>
</h3>
<p>Load databases.</p>
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">urlpath</span> <span class="op">&lt;-</span> <span class="st">"https://github.com/keita-iida/ASURATDB/blob/main/genes2bioterm/"</span>
<span class="fu"><a href="https://rdrr.io/r/base/load.html" class="external-link">load</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/connections.html" class="external-link">url</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="va">urlpath</span>, <span class="st">"20220108_human_COMSig.rda?raw=TRUE"</span><span class="op">)</span><span class="op">)</span><span class="op">)</span> <span class="co"># CO &amp; MSigDB</span>
<span class="fu"><a href="https://rdrr.io/r/base/load.html" class="external-link">load</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/connections.html" class="external-link">url</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="va">urlpath</span>, <span class="st">"20201213_human_GO_red.rda?raw=TRUE"</span><span class="op">)</span><span class="op">)</span><span class="op">)</span> <span class="co"># GO</span>
<span class="fu"><a href="https://rdrr.io/r/base/load.html" class="external-link">load</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/connections.html" class="external-link">url</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="va">urlpath</span>, <span class="st">"20201213_human_KEGG.rda?raw=TRUE"</span><span class="op">)</span><span class="op">)</span><span class="op">)</span>   <span class="co"># KEGG</span></code></pre></div>
<p>The reformatted knowledge-based data were available from the following repositories:</p>
<ul>
<li><a href="https://figshare.com/s/0599d2de970c2deb675c" class="external-link">DOI:10.6084/m9.figshare.19102598</a></li>
<li><a href="https://github.com/keita-iida/ASURATDB" class="external-link">Github ASURATDB</a></li>
</ul>
<p>Prepare correlation matrices of gene expressions.</p>
<div class="sourceCode" id="cb18"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">pbmc_cormat</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/stats/cor.html" class="external-link">cor</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/t.html" class="external-link">t</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">assay</a></span><span class="op">(</span><span class="va">pbmc</span>, <span class="st">"centered"</span><span class="op">)</span><span class="op">)</span>, method <span class="op">=</span> <span class="st">"spearman"</span><span class="op">)</span></code></pre></div>
<p>Set gene expression data into <code>altExp(sce)</code>.</p>
<p><strong>Tips:</strong> Take care not to use a slot name “log-normalized” for <code>altExp(sce)</code>, which may produce an error when using a Seurat (version 4.0.5) function <code>as.Seurat()</code> in the downstream analysis.</p>
<div class="sourceCode" id="cb19"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">sname</span> <span class="op">&lt;-</span> <span class="st">"logcounts"</span>
<span class="fu"><a href="https://rdrr.io/pkg/SingleCellExperiment/man/altExps.html" class="external-link">altExp</a></span><span class="op">(</span><span class="va">pbmc</span>, <span class="va">sname</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/RangedSummarizedExperiment-class.html" class="external-link">SummarizedExperiment</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>counts <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">assay</a></span><span class="op">(</span><span class="va">pbmc</span>, <span class="va">sname</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></code></pre></div>
<p>Add ENTREZ Gene IDs to <code>rowData(sce)</code>.</p>
<div class="sourceCode" id="cb20"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">dictionary</span> <span class="op">&lt;-</span> <span class="fu">AnnotationDbi</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/AnnotationDbi/man/AnnotationDb-class.html" class="external-link">select</a></span><span class="op">(</span><span class="fu">org.Hs.eg.db</span><span class="fu">::</span><span class="va"><a href="https://rdrr.io/pkg/org.Hs.eg.db/man/org.Hs.egBASE.html" class="external-link">org.Hs.eg.db</a></span>,
                                    key <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">pbmc</span><span class="op">)</span>,
                                    columns <span class="op">=</span> <span class="st">"ENTREZID"</span>, keytype <span class="op">=</span> <span class="st">"SYMBOL"</span><span class="op">)</span>
<span class="va">dictionary</span> <span class="op">&lt;-</span> <span class="va">dictionary</span><span class="op">[</span><span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/duplicated.html" class="external-link">duplicated</a></span><span class="op">(</span><span class="va">dictionary</span><span class="op">$</span><span class="va">SYMBOL</span><span class="op">)</span>, <span class="op">]</span>
<span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">rowData</a></span><span class="op">(</span><span class="va">pbmc</span><span class="op">)</span><span class="op">$</span><span class="va">geneID</span> <span class="op">&lt;-</span> <span class="va">dictionary</span><span class="op">$</span><span class="va">ENTREZID</span></code></pre></div>
<p>Add formatted databases to <code>metadata(sce)$sign</code>.</p>
<div class="sourceCode" id="cb21"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">pbmcs</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>CM <span class="op">=</span> <span class="va">pbmc</span>, GO <span class="op">=</span> <span class="va">pbmc</span>, KG <span class="op">=</span> <span class="va">pbmc</span><span class="op">)</span>
<span class="fu">metadata</span><span class="op">(</span><span class="va">pbmcs</span><span class="op">$</span><span class="va">CM</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>sign <span class="op">=</span> <span class="va">human_COMSig</span><span class="op">[[</span><span class="st">"cell"</span><span class="op">]</span><span class="op">]</span><span class="op">)</span>
<span class="fu">metadata</span><span class="op">(</span><span class="va">pbmcs</span><span class="op">$</span><span class="va">GO</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>sign <span class="op">=</span> <span class="va">human_GO</span><span class="op">[[</span><span class="st">"BP"</span><span class="op">]</span><span class="op">]</span><span class="op">)</span>
<span class="fu">metadata</span><span class="op">(</span><span class="va">pbmcs</span><span class="op">$</span><span class="va">KG</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>sign <span class="op">=</span> <span class="va">human_KEGG</span><span class="op">[[</span><span class="st">"pathway"</span><span class="op">]</span><span class="op">]</span><span class="op">)</span></code></pre></div>
<p><br></p>
</div>
<div class="section level3">
<h3 id="create-signs">Create signs<a class="anchor" aria-label="anchor" href="#create-signs"></a>
</h3>
<p>ASURAT function <code><a href="../reference/remove_signs.html">remove_signs()</a></code> redefines functional gene sets for the input database by removing genes, which are not included in <code>rownames(sce)</code>, and further removes biological terms including too few or too many genes.</p>
<p>The arguments are</p>
<ul>
<li>
<code>sce</code>: SingleCellExperiment object,</li>
<li>
<code>min_ngenes</code>: minimal number of genes&gt; 1 (the default value is 2), and</li>
<li>
<code>max_ngenes</code>: maximal number of genes&gt; 1 (the default value is 1000).</li>
</ul>
<div class="sourceCode" id="cb22"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">pbmcs</span><span class="op">$</span><span class="va">CM</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/remove_signs.html">remove_signs</a></span><span class="op">(</span>sce <span class="op">=</span> <span class="va">pbmcs</span><span class="op">$</span><span class="va">CM</span>, min_ngenes <span class="op">=</span> <span class="fl">2</span>, max_ngenes <span class="op">=</span> <span class="fl">1000</span><span class="op">)</span>
<span class="va">pbmcs</span><span class="op">$</span><span class="va">GO</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/remove_signs.html">remove_signs</a></span><span class="op">(</span>sce <span class="op">=</span> <span class="va">pbmcs</span><span class="op">$</span><span class="va">GO</span>, min_ngenes <span class="op">=</span> <span class="fl">2</span>, max_ngenes <span class="op">=</span> <span class="fl">500</span><span class="op">)</span>
<span class="va">pbmcs</span><span class="op">$</span><span class="va">KG</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/remove_signs.html">remove_signs</a></span><span class="op">(</span>sce <span class="op">=</span> <span class="va">pbmcs</span><span class="op">$</span><span class="va">KG</span>, min_ngenes <span class="op">=</span> <span class="fl">2</span>, max_ngenes <span class="op">=</span> <span class="fl">1000</span><span class="op">)</span></code></pre></div>
<p>The results are stored in <code>metadata(sce)$sign</code>.</p>
<p>ASURAT function <code>cluster_genes()</code> clusters functional gene sets using a correlation graph-based decomposition method, which produces strongly, variably, and weakly correlated gene sets (SCG, VCG, and WCG, respectively).</p>
<p>The arguments are</p>
<ul>
<li>
<code>sce</code>: SingleCellExperiment object,</li>
<li>
<code>cormat</code>: correlation matrix of gene expressions,</li>
<li>
<code>th_posi</code> and <code>th_nega</code>: threshold values of positive and negative correlation coefficients, respectively.</li>
</ul>
<p><strong>Tips:</strong> Empirically, typical values of <code>th_posi</code> and <code>th_nega</code> are <span class="math inline">\(0.15 \le {\rm th{\_}posi} \le 0.4\)</span> and <span class="math inline">\(-0.4 \le {\rm th{\_}nega} \le -0.15\)</span>, but one cannot avoid trial and error for setting these values. An exhaustive parameter searching is time-consuming but helpful for obtaining interpretable results.</p>
<div class="sourceCode" id="cb23"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/base/Random.html" class="external-link">set.seed</a></span><span class="op">(</span><span class="fl">1</span><span class="op">)</span>
<span class="va">pbmcs</span><span class="op">$</span><span class="va">CM</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/cluster_genesets.html">cluster_genesets</a></span><span class="op">(</span>sce <span class="op">=</span> <span class="va">pbmcs</span><span class="op">$</span><span class="va">CM</span>, cormat <span class="op">=</span> <span class="va">pbmc_cormat</span>,
                             th_posi <span class="op">=</span> <span class="fl">0.20</span>, th_nega <span class="op">=</span> <span class="op">-</span><span class="fl">0.20</span><span class="op">)</span>

<span class="fu"><a href="https://rdrr.io/r/base/Random.html" class="external-link">set.seed</a></span><span class="op">(</span><span class="fl">1</span><span class="op">)</span>
<span class="va">pbmcs</span><span class="op">$</span><span class="va">GO</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/cluster_genesets.html">cluster_genesets</a></span><span class="op">(</span>sce <span class="op">=</span> <span class="va">pbmcs</span><span class="op">$</span><span class="va">GO</span>, cormat <span class="op">=</span> <span class="va">pbmc_cormat</span>,
                             th_posi <span class="op">=</span> <span class="fl">0.20</span>, th_nega <span class="op">=</span> <span class="op">-</span><span class="fl">0.20</span><span class="op">)</span>

<span class="fu"><a href="https://rdrr.io/r/base/Random.html" class="external-link">set.seed</a></span><span class="op">(</span><span class="fl">1</span><span class="op">)</span>
<span class="va">pbmcs</span><span class="op">$</span><span class="va">KG</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/cluster_genesets.html">cluster_genesets</a></span><span class="op">(</span>sce <span class="op">=</span> <span class="va">pbmcs</span><span class="op">$</span><span class="va">KG</span>, cormat <span class="op">=</span> <span class="va">pbmc_cormat</span>,
                             th_posi <span class="op">=</span> <span class="fl">0.20</span>, th_nega <span class="op">=</span> <span class="op">-</span><span class="fl">0.20</span><span class="op">)</span></code></pre></div>
<p>The results are stored in <code>metadata(sce)$sign</code>.</p>
<p>ASURAT function <code><a href="../reference/create_signs.html">create_signs()</a></code> creates signs by the following criteria:</p>
<ol style="list-style-type: decimal">
<li>the number of genes in SCG&gt;= <code>min_cnt_strg</code> (the default value is 2) and</li>
<li>the number of genes in VCG&gt;= <code>min_cnt_vari</code> (the default value is 2),</li>
</ol>
<p>which are independently applied to SCGs and VCGs, respectively.</p>
<p><strong>Tips:</strong> Empirically, typical values of <code>min_cnt_strg</code> and <code>min_cnt_vari</code> are <span class="math inline">\(2 \le {\rm min\_cnt\_strg} = {\rm min\_cnt\_vari} \le 3\)</span>, but one cannot avoid trial and error for setting these values. An exhaustive parameter searching is time-consuming but helpful for obtaining interpretable results.</p>
<div class="sourceCode" id="cb24"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">pbmcs</span><span class="op">$</span><span class="va">CM</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/create_signs.html">create_signs</a></span><span class="op">(</span>sce <span class="op">=</span> <span class="va">pbmcs</span><span class="op">$</span><span class="va">CM</span>, min_cnt_strg <span class="op">=</span> <span class="fl">2</span>, min_cnt_vari <span class="op">=</span> <span class="fl">2</span><span class="op">)</span>
<span class="va">pbmcs</span><span class="op">$</span><span class="va">GO</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/create_signs.html">create_signs</a></span><span class="op">(</span>sce <span class="op">=</span> <span class="va">pbmcs</span><span class="op">$</span><span class="va">GO</span>, min_cnt_strg <span class="op">=</span> <span class="fl">3</span>, min_cnt_vari <span class="op">=</span> <span class="fl">3</span><span class="op">)</span>
<span class="va">pbmcs</span><span class="op">$</span><span class="va">KG</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/create_signs.html">create_signs</a></span><span class="op">(</span>sce <span class="op">=</span> <span class="va">pbmcs</span><span class="op">$</span><span class="va">KG</span>, min_cnt_strg <span class="op">=</span> <span class="fl">3</span>, min_cnt_vari <span class="op">=</span> <span class="fl">3</span><span class="op">)</span></code></pre></div>
<p>The results are stored in <code>metadata(sce)$sign_all</code>, where “CorrType” indicates SCG or VCG, “Corr” means the average correlation coefficients of SCG or VCG, “CorrWeak” means the average correlation coefficients of WCG, “CorrGene” means SCG or VCG, and “WeakCorrGene” means WCG. The orders of gene symbols and ENTREZ IDs, separated by “/”, are consistent.</p>
<p><strong>Tips:</strong> If one would like to recreate signs, reset the list of objects by, e.g., (<code>pbmcs &lt;- list(CM = pbmc, GO = pbmc, KG = pbmc)</code>), and go back to <code><a href="../reference/remove_signs.html">remove_signs()</a></code>.</p>
<p><br></p>
</div>
<div class="section level3">
<h3 id="select-useful-signs">Select useful signs<a class="anchor" aria-label="anchor" href="#select-useful-signs"></a>
</h3>
<p>If signs have semantic similarity information, one can use ASURAT function <code><a href="../reference/remove_signs_redundant.html">remove_signs_redundant()</a></code> for removing redundant sings using the semantic similarity matrices.</p>
<p>The arguments are</p>
<ul>
<li>
<code>sce</code>: SingleCellExperiment object,</li>
<li>
<code>similarity_matrix</code>: a semantic similarity matrix,</li>
<li>
<code>threshold</code>: a threshold value of semantic similarity, used for regarding biological terms as similar ones, and</li>
<li>
<code>keep_rareID</code>: if <code>TRUE</code>, biological terms with the larger ICs are kept.</li>
</ul>
<p><strong>Tips:</strong> The optimal value of <code>threshold</code> depends on the ontology structure as well as the method for computing semantic similarity matrix.</p>
<div class="sourceCode" id="cb25"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">pbmcs</span><span class="op">$</span><span class="va">GO</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/remove_signs_redundant.html">remove_signs_redundant</a></span><span class="op">(</span>sce <span class="op">=</span> <span class="va">pbmcs</span><span class="op">$</span><span class="va">GO</span>,
                                   similarity_matrix <span class="op">=</span> <span class="va">human_GO</span><span class="op">$</span><span class="va">similarity_matrix</span><span class="op">$</span><span class="va">BP</span>,
                                   threshold <span class="op">=</span> <span class="fl">0.90</span>, keep_rareID <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></code></pre></div>
<p>The results are stored in <code>metadata(sce)$sign_SCG</code>, <code>metadata(sce)$sign_VCG</code>, <code>metadata(sce)$sign_all</code>, and if there exist, <code>metadata(sce)$sign_SCG_redundant</code> and <code>metadata(sce)$sign_VCG_redundant</code>.</p>
<p>ASURAT function <code><a href="../reference/remove_signs_manually.html">remove_signs_manually()</a></code> removes signs by specifying IDs (e.g., <code>GOID:XXX</code>) or descriptions (e.g., <code>metabolic</code>) using <code><a href="https://rdrr.io/r/base/grep.html" class="external-link">grepl()</a></code>. The arguments are <code>sce</code> and <code>keywords</code> (keywords separated by <code>|</code>).</p>
<div class="sourceCode" id="cb26"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">keywords</span> <span class="op">&lt;-</span> <span class="st">"Covid|COVID|foofoo|hogehoge"</span>
<span class="va">pbmcs</span><span class="op">$</span><span class="va">KG</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/remove_signs_manually.html">remove_signs_manually</a></span><span class="op">(</span>sce <span class="op">=</span> <span class="va">pbmcs</span><span class="op">$</span><span class="va">KG</span>, keywords <span class="op">=</span> <span class="va">keywords</span><span class="op">)</span></code></pre></div>
<p>The results are stored in <code>metadata(sce)$sign_SCG</code>, <code>metadata(sce)$sign_VCG</code>, and <code>metadata(sce)$sign_all</code>.</p>
<p>There is another ASURAT function <code><a href="../reference/select_signs_manually.html">select_signs_manually()</a></code>, a counter part of <code><a href="../reference/remove_signs_manually.html">remove_signs_manually()</a></code>, which removes signs that do not include <code>keywords</code> (keywords separated by <code>|</code>).</p>
<div class="sourceCode" id="cb27"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">keywords</span> <span class="op">&lt;-</span> <span class="st">"cell|cyte"</span>
<span class="va">test</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/select_signs_manually.html">select_signs_manually</a></span><span class="op">(</span>sce <span class="op">=</span> <span class="va">pbmcs</span><span class="op">$</span><span class="va">CM</span>, keywords <span class="op">=</span> <span class="va">keywords</span><span class="op">)</span></code></pre></div>
<p>The results are stored in <code>metadata(sce)$sign_SCG</code>, <code>metadata(sce)$sign_VCG</code>, and <code>metadata(sce)$sign_all</code>.</p>
<p><br></p>
</div>
<div class="section level3">
<h3 id="create-sign-by-sample-matrices">Create sign-by-sample matrices<a class="anchor" aria-label="anchor" href="#create-sign-by-sample-matrices"></a>
</h3>
<p>ASURAT function <code>create_sce_signmatrix()</code> creates a new SingleCellExperiment object <code>new_sce</code>, consisting of the following information:</p>
<ul>
<li>
<code>assayNames(new_sce)</code>: counts (SSM whose entries are termed sign scores),</li>
<li>
<code>names(colData(new_sce))</code>: nReads, nGenes, percMT,</li>
<li>
<code>names(rowData(new_sce))</code>: ParentSignID, Description, CorrGene, etc.,</li>
<li>
<code>names(metadata(new_sce))</code>: sign_SCG, sign_VCG, etc.,</li>
<li>
<code>altExpNames(new_sce)</code>: something if there is data in <code>altExp(sce)</code>.</li>
</ul>
<p>The arguments are</p>
<ul>
<li>
<code>sce</code>: SingleCellExperiment object,</li>
<li>
<code>weight_strg</code>: weight parameter for SCG (the default value is 0.5), and</li>
<li>
<code>weight_vari</code>: weight parameter for VCG (the default is 0.5).</li>
</ul>
<div class="sourceCode" id="cb28"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">pbmcs</span><span class="op">$</span><span class="va">CM</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/makeSignMatrix.html">makeSignMatrix</a></span><span class="op">(</span>sce <span class="op">=</span> <span class="va">pbmcs</span><span class="op">$</span><span class="va">CM</span>, weight_strg <span class="op">=</span> <span class="fl">0.5</span>, weight_vari <span class="op">=</span> <span class="fl">0.5</span><span class="op">)</span>
<span class="va">pbmcs</span><span class="op">$</span><span class="va">GO</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/makeSignMatrix.html">makeSignMatrix</a></span><span class="op">(</span>sce <span class="op">=</span> <span class="va">pbmcs</span><span class="op">$</span><span class="va">GO</span>, weight_strg <span class="op">=</span> <span class="fl">0.5</span>, weight_vari <span class="op">=</span> <span class="fl">0.5</span><span class="op">)</span>
<span class="va">pbmcs</span><span class="op">$</span><span class="va">KG</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/makeSignMatrix.html">makeSignMatrix</a></span><span class="op">(</span>sce <span class="op">=</span> <span class="va">pbmcs</span><span class="op">$</span><span class="va">KG</span>, weight_strg <span class="op">=</span> <span class="fl">0.5</span>, weight_vari <span class="op">=</span> <span class="fl">0.5</span><span class="op">)</span></code></pre></div>
<p>Below are head and tail of <code>assay(sce, "counts")</code>:</p>
<div class="sourceCode" id="cb29"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/base/cbind.html" class="external-link">rbind</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">assay</a></span><span class="op">(</span><span class="va">pbmcs</span><span class="op">$</span><span class="va">CM</span>, <span class="st">"counts"</span><span class="op">)</span><span class="op">[</span>, <span class="fl">1</span><span class="op">:</span><span class="fl">3</span><span class="op">]</span>, n <span class="op">=</span> <span class="fl">4</span><span class="op">)</span>,
      <span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">tail</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">assay</a></span><span class="op">(</span><span class="va">pbmcs</span><span class="op">$</span><span class="va">CM</span>, <span class="st">"counts"</span><span class="op">)</span><span class="op">[</span>, <span class="fl">1</span><span class="op">:</span><span class="fl">3</span><span class="op">]</span>, n <span class="op">=</span> <span class="fl">4</span><span class="op">)</span><span class="op">)</span>
<span class="co">#&gt;              AAACCTGAGACAGACC-1 AAACCTGAGCGCCTCA-1 AAACCTGAGGCATGGT-1</span>
<span class="co">#&gt; CL:0000097-S         0.14532707       -0.053723407       -0.063695116</span>
<span class="co">#&gt; CL:0000100-S         0.09553473       -0.028719989        0.083311009</span>
<span class="co">#&gt; CL:0000121-S         0.33933952       -0.083850197       -0.138209457</span>
<span class="co">#&gt; CL:0000126-S         0.11126650       -0.115367841       -0.129355630</span>
<span class="co">#&gt; MSigID:278-V        -0.01188412       -0.053548307       -0.071345499</span>
<span class="co">#&gt; MSigID:289-V        -0.10786674       -0.002297659       -0.088932797</span>
<span class="co">#&gt; MSigID:290-V         0.09279058       -0.060334748       -0.044793026</span>
<span class="co">#&gt; MSigID:295-V        -0.01473417       -0.056002030       -0.004625395</span></code></pre></div>
<p><br></p>
</div>
<div class="section level3">
<h3 id="dim_reduction">Reduce dimensions of sign-by-sample matrices<a class="anchor" aria-label="anchor" href="#dim_reduction"></a>
</h3>
<p>Perform principal component analysis and t-distributed stochastic neighbor embedding.</p>
<div class="sourceCode" id="cb30"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/base/Random.html" class="external-link">set.seed</a></span><span class="op">(</span><span class="fl">1</span><span class="op">)</span>
<span class="va">res</span> <span class="op">&lt;-</span> <span class="fu">Rtsne</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/Rtsne/man/Rtsne.html" class="external-link">Rtsne</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/t.html" class="external-link">t</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">assay</a></span><span class="op">(</span><span class="va">pbmcs</span><span class="op">$</span><span class="va">CM</span>, <span class="st">"counts"</span><span class="op">)</span><span class="op">)</span>, dim <span class="op">=</span> <span class="fl">2</span>, pca <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/pkg/SingleCellExperiment/man/reducedDims.html" class="external-link">reducedDim</a></span><span class="op">(</span><span class="va">pbmcs</span><span class="op">$</span><span class="va">CM</span>, <span class="st">"TSNE"</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="va">res</span><span class="op">[[</span><span class="st">"Y"</span><span class="op">]</span><span class="op">]</span>

<span class="fu"><a href="https://rdrr.io/r/base/Random.html" class="external-link">set.seed</a></span><span class="op">(</span><span class="fl">1</span><span class="op">)</span>
<span class="va">res</span> <span class="op">&lt;-</span> <span class="fu">Rtsne</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/Rtsne/man/Rtsne.html" class="external-link">Rtsne</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/t.html" class="external-link">t</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">assay</a></span><span class="op">(</span><span class="va">pbmcs</span><span class="op">$</span><span class="va">GO</span>, <span class="st">"counts"</span><span class="op">)</span><span class="op">)</span>, dim <span class="op">=</span> <span class="fl">2</span>, pca <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/pkg/SingleCellExperiment/man/reducedDims.html" class="external-link">reducedDim</a></span><span class="op">(</span><span class="va">pbmcs</span><span class="op">$</span><span class="va">GO</span>, <span class="st">"TSNE"</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="va">res</span><span class="op">[[</span><span class="st">"Y"</span><span class="op">]</span><span class="op">]</span>

<span class="fu"><a href="https://rdrr.io/r/base/Random.html" class="external-link">set.seed</a></span><span class="op">(</span><span class="fl">1</span><span class="op">)</span>
<span class="va">res</span> <span class="op">&lt;-</span> <span class="fu">Rtsne</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/Rtsne/man/Rtsne.html" class="external-link">Rtsne</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/t.html" class="external-link">t</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">assay</a></span><span class="op">(</span><span class="va">pbmcs</span><span class="op">$</span><span class="va">KG</span>, <span class="st">"counts"</span><span class="op">)</span><span class="op">)</span>, dim <span class="op">=</span> <span class="fl">3</span>, pca <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/pkg/SingleCellExperiment/man/reducedDims.html" class="external-link">reducedDim</a></span><span class="op">(</span><span class="va">pbmcs</span><span class="op">$</span><span class="va">KG</span>, <span class="st">"TSNE"</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="va">res</span><span class="op">[[</span><span class="st">"Y"</span><span class="op">]</span><span class="op">]</span></code></pre></div>
<p>The results can be visualized by ASURAT functions <code><a href="../reference/plot_dataframe2D.html">plot_dataframe2D()</a></code> or <code><a href="../reference/plot_dataframe3D.html">plot_dataframe3D()</a></code> (see <a href="#visualization_lowdim">here</a> for details).</p>
<div class="sourceCode" id="cb31"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">dataframe2D</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/as.data.frame.html" class="external-link">as.data.frame</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/SingleCellExperiment/man/reducedDims.html" class="external-link">reducedDim</a></span><span class="op">(</span><span class="va">pbmcs</span><span class="op">$</span><span class="va">CM</span>, <span class="st">"TSNE"</span><span class="op">)</span><span class="op">)</span>
<span class="fu"><a href="../reference/plot_dataframe2D.html">plot_dataframe2D</a></span><span class="op">(</span>dataframe2D <span class="op">=</span> <span class="va">dataframe2D</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu">ggplot2</span><span class="fu">::</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">labs</a></span><span class="op">(</span>title <span class="op">=</span> <span class="st">"PBMC (CO &amp; MSigDB)"</span>, x <span class="op">=</span> <span class="st">"tSNE_1"</span>, y <span class="op">=</span> <span class="st">"tSNE_2"</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu">ggplot2</span><span class="fu">::</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html" class="external-link">theme_classic</a></span><span class="op">(</span>base_size <span class="op">=</span> <span class="fl">15</span><span class="op">)</span></code></pre></div>
<p><img src="ASURAT_files/figure-html/unnamed-chunk-32-1.png" width="700"></p>
<div class="sourceCode" id="cb32"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">dataframe2D</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/as.data.frame.html" class="external-link">as.data.frame</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/SingleCellExperiment/man/reducedDims.html" class="external-link">reducedDim</a></span><span class="op">(</span><span class="va">pbmcs</span><span class="op">$</span><span class="va">GO</span>, <span class="st">"TSNE"</span><span class="op">)</span><span class="op">)</span>
<span class="fu"><a href="../reference/plot_dataframe2D.html">plot_dataframe2D</a></span><span class="op">(</span>dataframe2D <span class="op">=</span> <span class="va">dataframe2D</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu">ggplot2</span><span class="fu">::</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">labs</a></span><span class="op">(</span>title <span class="op">=</span> <span class="st">"PBMC (GO)"</span>, x <span class="op">=</span> <span class="st">"tSNE_1"</span>, y <span class="op">=</span> <span class="st">"tSNE_2"</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu">ggplot2</span><span class="fu">::</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html" class="external-link">theme_classic</a></span><span class="op">(</span>base_size <span class="op">=</span> <span class="fl">15</span><span class="op">)</span></code></pre></div>
<p><img src="ASURAT_files/figure-html/unnamed-chunk-33-1.png" width="700"></p>
<div class="sourceCode" id="cb33"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">dataframe3D</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/as.data.frame.html" class="external-link">as.data.frame</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/SingleCellExperiment/man/reducedDims.html" class="external-link">reducedDim</a></span><span class="op">(</span><span class="va">pbmcs</span><span class="op">$</span><span class="va">KG</span>, <span class="st">"TSNE"</span><span class="op">)</span><span class="op">[</span>, <span class="fu"><a href="https://rdrr.io/r/base/seq.html" class="external-link">seq_len</a></span><span class="op">(</span><span class="fl">3</span><span class="op">)</span><span class="op">]</span><span class="op">)</span>
<span class="fu"><a href="../reference/plot_dataframe3D.html">plot_dataframe3D</a></span><span class="op">(</span>dataframe3D <span class="op">=</span> <span class="va">dataframe3D</span>,
                 theta <span class="op">=</span> <span class="fl">40</span>, phi <span class="op">=</span> <span class="fl">25</span>, title <span class="op">=</span> <span class="st">"PBMC (KEGG)"</span>,
                 xlabel <span class="op">=</span> <span class="st">"tSNE_1"</span>, ylabel <span class="op">=</span> <span class="st">"tSNE_2"</span>, zlabel <span class="op">=</span> <span class="st">"tSNE_3"</span><span class="op">)</span></code></pre></div>
<p><img src="ASURAT_files/figure-html/unnamed-chunk-34-1.png" width="700"></p>
<p><br></p>
</div>
<div class="section level3">
<h3 id="cluster-cells">Cluster cells<a class="anchor" aria-label="anchor" href="#cluster-cells"></a>
</h3>
<div class="section level4">
<h4 id="use-seurat-functions">Use Seurat functions<a class="anchor" aria-label="anchor" href="#use-seurat-functions"></a>
</h4>
<p>To date (December, 2021), one of the most useful clustering methods in scRNA-seq data analysis is a combination of a community detection algorithm and graph-based unsupervised clustering, developed in Seurat package.</p>
<p>In this tutorial, our strategy is as follows:</p>
<ol style="list-style-type: decimal">
<li>convert SingleCellExperiment objects into Seurat objects (note that <code><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">rowData()</a></code> and <code><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">colData()</a></code> must have data),</li>
<li>perform <code>ScaleData()</code>, <code>RunPCA()</code>, <code>FindNeighbors()</code>, and <code>FindClusters()</code>,</li>
<li>convert Seurat objects into temporal SingleCellExperiment objects <code>temp</code>,</li>
<li>add <code>colData(temp)$seurat_clusters</code> into <code>colData(sce)$seurat_clusters</code>.</li>
</ol>
<div class="sourceCode" id="cb34"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">resolutions</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">0.15</span>, <span class="fl">0.10</span>, <span class="fl">0.20</span><span class="op">)</span>
<span class="va">dims</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/seq.html" class="external-link">seq_len</a></span><span class="op">(</span><span class="fl">30</span><span class="op">)</span>, <span class="fu"><a href="https://rdrr.io/r/base/seq.html" class="external-link">seq_len</a></span><span class="op">(</span><span class="fl">30</span><span class="op">)</span>, <span class="fu"><a href="https://rdrr.io/r/base/seq.html" class="external-link">seq_len</a></span><span class="op">(</span><span class="fl">10</span><span class="op">)</span><span class="op">)</span>
<span class="kw">for</span><span class="op">(</span><span class="va">i</span> <span class="kw">in</span> <span class="fu"><a href="https://rdrr.io/r/base/seq.html" class="external-link">seq_len</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/length.html" class="external-link">length</a></span><span class="op">(</span><span class="va">pbmcs</span><span class="op">)</span><span class="op">)</span><span class="op">)</span><span class="op">{</span>
  <span class="va">pbmc_surt</span> <span class="op">&lt;-</span> <span class="fu">Seurat</span><span class="fu">::</span><span class="fu"><a href="https://satijalab.org/seurat/reference/as.Seurat.html" class="external-link">as.Seurat</a></span><span class="op">(</span><span class="va">pbmcs</span><span class="op">[[</span><span class="va">i</span><span class="op">]</span><span class="op">]</span>, counts <span class="op">=</span> <span class="st">"counts"</span>, data <span class="op">=</span> <span class="st">"counts"</span><span class="op">)</span>
  <span class="va">pbmc_surt</span><span class="op">[[</span><span class="st">"SSM"</span><span class="op">]</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="fu">Seurat</span><span class="fu">::</span><span class="fu"><a href="https://satijalab.org/seurat/reference/CreateAssayObject.html" class="external-link">CreateAssayObject</a></span><span class="op">(</span>
    counts <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/matrix.html" class="external-link">as.matrix</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">assay</a></span><span class="op">(</span><span class="va">pbmcs</span><span class="op">[[</span><span class="va">i</span><span class="op">]</span><span class="op">]</span>, <span class="st">"counts"</span><span class="op">)</span><span class="op">)</span><span class="op">)</span>
  <span class="fu">Seurat</span><span class="fu">::</span><span class="fu"><a href="https://satijalab.org/seurat/reference/DefaultAssay.html" class="external-link">DefaultAssay</a></span><span class="op">(</span><span class="va">pbmc_surt</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="st">"SSM"</span>
  <span class="va">pbmc_surt</span> <span class="op">&lt;-</span> <span class="fu">Seurat</span><span class="fu">::</span><span class="fu"><a href="https://satijalab.org/seurat/reference/ScaleData.html" class="external-link">ScaleData</a></span><span class="op">(</span><span class="va">pbmc_surt</span>, features <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">pbmc_surt</span><span class="op">)</span><span class="op">)</span>
  <span class="va">pbmc_surt</span> <span class="op">&lt;-</span> <span class="fu">Seurat</span><span class="fu">::</span><span class="fu"><a href="https://satijalab.org/seurat/reference/RunPCA.html" class="external-link">RunPCA</a></span><span class="op">(</span><span class="va">pbmc_surt</span>, features <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">pbmc_surt</span><span class="op">)</span><span class="op">)</span>
  <span class="va">pbmc_surt</span> <span class="op">&lt;-</span> <span class="fu">Seurat</span><span class="fu">::</span><span class="fu"><a href="https://satijalab.org/seurat/reference/FindNeighbors.html" class="external-link">FindNeighbors</a></span><span class="op">(</span><span class="va">pbmc_surt</span>, reduction <span class="op">=</span> <span class="st">"pca"</span>, dims <span class="op">=</span> <span class="va">dims</span><span class="op">[[</span><span class="va">i</span><span class="op">]</span><span class="op">]</span><span class="op">)</span>
  <span class="va">pbmc_surt</span> <span class="op">&lt;-</span> <span class="fu">Seurat</span><span class="fu">::</span><span class="fu"><a href="https://satijalab.org/seurat/reference/FindClusters.html" class="external-link">FindClusters</a></span><span class="op">(</span><span class="va">pbmc_surt</span>, resolution <span class="op">=</span> <span class="va">resolutions</span><span class="op">[</span><span class="va">i</span><span class="op">]</span><span class="op">)</span>
  <span class="va">pbmc_temp</span> <span class="op">&lt;-</span> <span class="fu">Seurat</span><span class="fu">::</span><span class="fu"><a href="https://satijalab.org/seurat/reference/as.SingleCellExperiment.html" class="external-link">as.SingleCellExperiment</a></span><span class="op">(</span><span class="va">pbmc_surt</span><span class="op">)</span>
  <span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">colData</a></span><span class="op">(</span><span class="va">pbmcs</span><span class="op">[[</span><span class="va">i</span><span class="op">]</span><span class="op">]</span><span class="op">)</span><span class="op">$</span><span class="va">seurat_clusters</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">colData</a></span><span class="op">(</span><span class="va">pbmc_temp</span><span class="op">)</span><span class="op">$</span><span class="va">seurat_clusters</span>
<span class="op">}</span></code></pre></div>
<p>The results can be visualized by ASURAT functions <code><a href="../reference/plot_dataframe2D.html">plot_dataframe2D()</a></code> or <code><a href="../reference/plot_dataframe3D.html">plot_dataframe3D()</a></code> (see <a href="#visualization_lowdim">here</a> for details).</p>
<div class="sourceCode" id="cb35"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">labels</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">colData</a></span><span class="op">(</span><span class="va">pbmcs</span><span class="op">$</span><span class="va">CM</span><span class="op">)</span><span class="op">$</span><span class="va">seurat_clusters</span>
<span class="va">dataframe2D</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/as.data.frame.html" class="external-link">as.data.frame</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/SingleCellExperiment/man/reducedDims.html" class="external-link">reducedDim</a></span><span class="op">(</span><span class="va">pbmcs</span><span class="op">$</span><span class="va">CM</span>, <span class="st">"TSNE"</span><span class="op">)</span><span class="op">)</span>
<span class="fu"><a href="../reference/plot_dataframe2D.html">plot_dataframe2D</a></span><span class="op">(</span>dataframe2D <span class="op">=</span> <span class="va">dataframe2D</span>, labels <span class="op">=</span> <span class="va">labels</span>, colors <span class="op">=</span> <span class="cn">NULL</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu">ggplot2</span><span class="fu">::</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">labs</a></span><span class="op">(</span>title <span class="op">=</span> <span class="st">"PBMC (CO &amp; MSigDB)"</span>, x <span class="op">=</span> <span class="st">"tSNE_1"</span>, y <span class="op">=</span> <span class="st">"tSNE_2"</span>, color <span class="op">=</span> <span class="st">""</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu">ggplot2</span><span class="fu">::</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html" class="external-link">theme_classic</a></span><span class="op">(</span>base_size <span class="op">=</span> <span class="fl">15</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu">ggplot2</span><span class="fu">::</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/guides.html" class="external-link">guides</a></span><span class="op">(</span>colour <span class="op">=</span> <span class="fu">ggplot2</span><span class="fu">::</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/guide_legend.html" class="external-link">guide_legend</a></span><span class="op">(</span>override.aes <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>size <span class="op">=</span> <span class="fl">4</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></code></pre></div>
<p><img src="ASURAT_files/figure-html/unnamed-chunk-36-1.png" width="700"></p>
<div class="sourceCode" id="cb36"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">labels</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">colData</a></span><span class="op">(</span><span class="va">pbmcs</span><span class="op">$</span><span class="va">GO</span><span class="op">)</span><span class="op">$</span><span class="va">seurat_clusters</span>
<span class="va">dataframe2D</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/as.data.frame.html" class="external-link">as.data.frame</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/SingleCellExperiment/man/reducedDims.html" class="external-link">reducedDim</a></span><span class="op">(</span><span class="va">pbmcs</span><span class="op">$</span><span class="va">GO</span>, <span class="st">"TSNE"</span><span class="op">)</span><span class="op">)</span>
<span class="fu"><a href="../reference/plot_dataframe2D.html">plot_dataframe2D</a></span><span class="op">(</span>dataframe2D <span class="op">=</span> <span class="va">dataframe2D</span>, labels <span class="op">=</span> <span class="va">labels</span>, colors <span class="op">=</span> <span class="cn">NULL</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu">ggplot2</span><span class="fu">::</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">labs</a></span><span class="op">(</span>title <span class="op">=</span> <span class="st">"PBMC (GO)"</span>, x <span class="op">=</span> <span class="st">"tSNE_1"</span>, y <span class="op">=</span> <span class="st">"tSNE_2"</span>, color <span class="op">=</span> <span class="st">""</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu">ggplot2</span><span class="fu">::</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html" class="external-link">theme_classic</a></span><span class="op">(</span>base_size <span class="op">=</span> <span class="fl">15</span><span class="op">)</span></code></pre></div>
<p><img src="ASURAT_files/figure-html/unnamed-chunk-37-1.png" width="700"></p>
<div class="sourceCode" id="cb37"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">labels</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">colData</a></span><span class="op">(</span><span class="va">pbmcs</span><span class="op">$</span><span class="va">KG</span><span class="op">)</span><span class="op">$</span><span class="va">seurat_clusters</span>
<span class="va">dataframe3D</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/as.data.frame.html" class="external-link">as.data.frame</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/SingleCellExperiment/man/reducedDims.html" class="external-link">reducedDim</a></span><span class="op">(</span><span class="va">pbmcs</span><span class="op">$</span><span class="va">KG</span>, <span class="st">"TSNE"</span><span class="op">)</span><span class="op">[</span>, <span class="fu"><a href="https://rdrr.io/r/base/seq.html" class="external-link">seq_len</a></span><span class="op">(</span><span class="fl">3</span><span class="op">)</span><span class="op">]</span><span class="op">)</span>
<span class="fu"><a href="../reference/plot_dataframe3D.html">plot_dataframe3D</a></span><span class="op">(</span>dataframe3D <span class="op">=</span> <span class="va">dataframe3D</span>, labels <span class="op">=</span> <span class="va">labels</span>, colors <span class="op">=</span> <span class="cn">NULL</span>,
                 theta <span class="op">=</span> <span class="fl">40</span>, phi <span class="op">=</span> <span class="fl">25</span>, title <span class="op">=</span> <span class="st">"PBMC (KEGG)"</span>,
                 xlabel <span class="op">=</span> <span class="st">"tSNE_1"</span>, ylabel <span class="op">=</span> <span class="st">"tSNE_2"</span>, zlabel <span class="op">=</span> <span class="st">"tSNE_3"</span><span class="op">)</span></code></pre></div>
<p><img src="ASURAT_files/figure-html/unnamed-chunk-38-1.png" width="700"></p>
<p><br></p>
</div>
<div class="section level4">
<h4 id="cell-cycle-inference-using-seurat-functions">Cell cycle inference using Seurat functions<a class="anchor" aria-label="anchor" href="#cell-cycle-inference-using-seurat-functions"></a>
</h4>
<p>If there is gene expression data in <code>altExp(sce)</code>, one can infer cell cycle phases by using Seurat functions in the similar manner as above.</p>
<div class="sourceCode" id="cb38"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">pbmc_surt</span> <span class="op">&lt;-</span> <span class="fu">Seurat</span><span class="fu">::</span><span class="fu"><a href="https://satijalab.org/seurat/reference/as.Seurat.html" class="external-link">as.Seurat</a></span><span class="op">(</span><span class="va">pbmcs</span><span class="op">$</span><span class="va">CM</span>, counts <span class="op">=</span> <span class="st">"counts"</span>, data <span class="op">=</span> <span class="st">"counts"</span><span class="op">)</span>
<span class="va">pbmc_surt</span><span class="op">[[</span><span class="st">"GEM"</span><span class="op">]</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="fu">Seurat</span><span class="fu">::</span><span class="fu"><a href="https://satijalab.org/seurat/reference/CreateAssayObject.html" class="external-link">CreateAssayObject</a></span><span class="op">(</span>
  counts <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/matrix.html" class="external-link">as.matrix</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">assay</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/SingleCellExperiment/man/altExps.html" class="external-link">altExp</a></span><span class="op">(</span><span class="va">pbmcs</span><span class="op">$</span><span class="va">CM</span><span class="op">)</span>, <span class="st">"counts"</span><span class="op">)</span><span class="op">)</span><span class="op">)</span>
<span class="fu">Seurat</span><span class="fu">::</span><span class="fu"><a href="https://satijalab.org/seurat/reference/DefaultAssay.html" class="external-link">DefaultAssay</a></span><span class="op">(</span><span class="va">pbmc_surt</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="st">"GEM"</span>
<span class="va">pbmc_surt</span> <span class="op">&lt;-</span> <span class="fu">Seurat</span><span class="fu">::</span><span class="fu"><a href="https://satijalab.org/seurat/reference/ScaleData.html" class="external-link">ScaleData</a></span><span class="op">(</span><span class="va">pbmc_surt</span>, features <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">pbmc_surt</span><span class="op">)</span><span class="op">)</span>
<span class="va">pbmc_surt</span> <span class="op">&lt;-</span> <span class="fu">Seurat</span><span class="fu">::</span><span class="fu"><a href="https://satijalab.org/seurat/reference/RunPCA.html" class="external-link">RunPCA</a></span><span class="op">(</span><span class="va">pbmc_surt</span>, features <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">pbmc_surt</span><span class="op">)</span><span class="op">)</span>
<span class="va">pbmc_surt</span> <span class="op">&lt;-</span> <span class="fu">Seurat</span><span class="fu">::</span><span class="fu"><a href="https://satijalab.org/seurat/reference/CellCycleScoring.html" class="external-link">CellCycleScoring</a></span><span class="op">(</span><span class="va">pbmc_surt</span>,
                                      s.features <span class="op">=</span> <span class="fu">Seurat</span><span class="fu">::</span><span class="va"><a href="https://satijalab.org/seurat/reference/cc.genes.html" class="external-link">cc.genes</a></span><span class="op">$</span><span class="va">s.genes</span>,
                                      g2m.features <span class="op">=</span> <span class="fu">Seurat</span><span class="fu">::</span><span class="va"><a href="https://satijalab.org/seurat/reference/cc.genes.html" class="external-link">cc.genes</a></span><span class="op">$</span><span class="va">g2m.genes</span><span class="op">)</span>
<span class="va">pbmc_temp</span> <span class="op">&lt;-</span> <span class="fu">Seurat</span><span class="fu">::</span><span class="fu"><a href="https://satijalab.org/seurat/reference/as.SingleCellExperiment.html" class="external-link">as.SingleCellExperiment</a></span><span class="op">(</span><span class="va">pbmc_surt</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">colData</a></span><span class="op">(</span><span class="va">pbmcs</span><span class="op">$</span><span class="va">CM</span><span class="op">)</span><span class="op">$</span><span class="va">Phase</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">colData</a></span><span class="op">(</span><span class="va">pbmc_temp</span><span class="op">)</span><span class="op">$</span><span class="va">Phase</span></code></pre></div>
<p><br></p>
</div>
</div>
<div class="section level3">
<h3 id="separation_index">Investigate significant signs<a class="anchor" aria-label="anchor" href="#separation_index"></a>
</h3>
<p>Significant signs are analogous to differentially expressed genes but bear biological meanings. Note that naïve usages of statistical tests should be avoided because the row vectors of SSMs are centered.</p>
<p>Instead, ASURAT function <code><a href="../reference/compute_sepI_all.html">compute_sepI_all()</a></code> computes separation indices for each cluster against the others. Briefly, a separation index “sepI”, ranging from -1 to 1, is a nonparametric measure of significance of a given sign score for a given subpopulation. The larger (resp. smaller) sepI is, the more reliable the sign is as a positive (resp. negative) marker for the cluster.</p>
<p>The arguments are</p>
<ul>
<li>
<code>sce</code>: SingleCellExperiment object,</li>
<li>
<code>labels</code>: a vector of labels of all the samples, and</li>
<li>
<code>nrand_samples</code>: an integer for the number of samples used for random sampling, which samples at least one sample per cluster.</li>
</ul>
<div class="sourceCode" id="cb39"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="kw">for</span><span class="op">(</span><span class="va">i</span> <span class="kw">in</span> <span class="fu"><a href="https://rdrr.io/r/base/seq.html" class="external-link">seq_len</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/length.html" class="external-link">length</a></span><span class="op">(</span><span class="va">pbmcs</span><span class="op">)</span><span class="op">)</span><span class="op">)</span><span class="op">{</span>
  <span class="fu"><a href="https://rdrr.io/r/base/Random.html" class="external-link">set.seed</a></span><span class="op">(</span><span class="fl">1</span><span class="op">)</span>
  <span class="va">labels</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">colData</a></span><span class="op">(</span><span class="va">pbmcs</span><span class="op">[[</span><span class="va">i</span><span class="op">]</span><span class="op">]</span><span class="op">)</span><span class="op">$</span><span class="va">seurat_clusters</span>
  <span class="va">pbmcs</span><span class="op">[[</span><span class="va">i</span><span class="op">]</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/compute_sepI_all.html">compute_sepI_all</a></span><span class="op">(</span>sce <span class="op">=</span> <span class="va">pbmcs</span><span class="op">[[</span><span class="va">i</span><span class="op">]</span><span class="op">]</span>, labels <span class="op">=</span> <span class="va">labels</span>,
                                 nrand_samples <span class="op">=</span> <span class="fl">100</span><span class="op">)</span>
<span class="op">}</span></code></pre></div>
<p>The results are stored in <code>metadata(sce)$marker_signs</code>.</p>
<p>When computing separation indices between given clusters, e.g., cluster 1 versus clusters 2 and 3, use an ASURAT function <code><a href="../reference/compute_sepI_clusters.html">compute_sepI_clusters()</a></code>. See <code><a href="../reference/compute_sepI_clusters.html">?compute_sepI_clusters</a></code> for details.</p>
<p>ASURAT function <code><a href="../reference/plot_violin.html">plot_violin()</a></code> shows violin plots showing sign score distributions (see <a href="#violin">here</a> for details).</p>
<div class="sourceCode" id="cb40"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">vname</span> <span class="op">&lt;-</span> <span class="st">"MSigID:92-S"</span>
<span class="va">pbmc_sub</span> <span class="op">&lt;-</span> <span class="va">pbmcs</span><span class="op">$</span><span class="va">CM</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">pbmcs</span><span class="op">$</span><span class="va">CM</span><span class="op">)</span> <span class="op"><a href="https://rdrr.io/r/base/match.html" class="external-link">%in%</a></span> <span class="va">vname</span>, <span class="op">]</span>
<span class="va">labels</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">colData</a></span><span class="op">(</span><span class="va">pbmc_sub</span><span class="op">)</span><span class="op">$</span><span class="va">seurat_clusters</span>
<span class="va">dataframe1D</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/as.data.frame.html" class="external-link">as.data.frame</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/t.html" class="external-link">t</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">assay</a></span><span class="op">(</span><span class="va">pbmc_sub</span>, <span class="st">"counts"</span><span class="op">)</span><span class="op">)</span><span class="op">)</span>
<span class="fu"><a href="../reference/plot_violin.html">plot_violin</a></span><span class="op">(</span>dataframe1D <span class="op">=</span> <span class="va">dataframe1D</span>, labels <span class="op">=</span> <span class="va">labels</span>, colors <span class="op">=</span> <span class="cn">NULL</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu">ggplot2</span><span class="fu">::</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">labs</a></span><span class="op">(</span>title <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="va">vname</span>, <span class="st">"\n"</span>, <span class="st">"B cell (CD79A, CD22, ...)"</span><span class="op">)</span>,
                x <span class="op">=</span> <span class="st">"Cluster (CO &amp; MSigDB)"</span>, y <span class="op">=</span> <span class="st">"Sign score"</span>, fill <span class="op">=</span> <span class="st">"Cluster"</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu">ggplot2</span><span class="fu">::</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html" class="external-link">theme_classic</a></span><span class="op">(</span>base_size <span class="op">=</span> <span class="fl">20</span><span class="op">)</span></code></pre></div>
<p><img src="ASURAT_files/figure-html/unnamed-chunk-41-1.png" width="700"></p>
<p><br></p>
</div>
<div class="section level3">
<h3 id="investigate-significant-genes">Investigate significant genes<a class="anchor" aria-label="anchor" href="#investigate-significant-genes"></a>
</h3>
<div class="section level4">
<h4 id="use-seurat-function">Use Seurat function<a class="anchor" aria-label="anchor" href="#use-seurat-function"></a>
</h4>
<p>To date (December, 2021), one of the most useful methods of multiple statistical tests in scRNA-seq data analysis is to use a Seurat function <code>FindAllMarkers()</code>.</p>
<p>If there is gene expression data in <code>altExp(sce)</code>, one can investigate differentially expressed genes by using Seurat functions in the similar manner as described before.</p>
<div class="sourceCode" id="cb41"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/base/Random.html" class="external-link">set.seed</a></span><span class="op">(</span><span class="fl">1</span><span class="op">)</span>
<span class="va">pbmc_surt</span> <span class="op">&lt;-</span> <span class="fu">Seurat</span><span class="fu">::</span><span class="fu"><a href="https://satijalab.org/seurat/reference/as.Seurat.html" class="external-link">as.Seurat</a></span><span class="op">(</span><span class="va">pbmcs</span><span class="op">$</span><span class="va">CM</span>, counts <span class="op">=</span> <span class="st">"counts"</span>, data <span class="op">=</span> <span class="st">"counts"</span><span class="op">)</span>
<span class="va">pbmc_surt</span><span class="op">[[</span><span class="st">"GEM"</span><span class="op">]</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="fu">Seurat</span><span class="fu">::</span><span class="fu"><a href="https://satijalab.org/seurat/reference/CreateAssayObject.html" class="external-link">CreateAssayObject</a></span><span class="op">(</span>
  counts <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/matrix.html" class="external-link">as.matrix</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">assay</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/SingleCellExperiment/man/altExps.html" class="external-link">altExp</a></span><span class="op">(</span><span class="va">pbmcs</span><span class="op">$</span><span class="va">CM</span><span class="op">)</span>, <span class="st">"counts"</span><span class="op">)</span><span class="op">)</span><span class="op">)</span>
<span class="fu">Seurat</span><span class="fu">::</span><span class="fu"><a href="https://satijalab.org/seurat/reference/DefaultAssay.html" class="external-link">DefaultAssay</a></span><span class="op">(</span><span class="va">pbmc_surt</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="st">"GEM"</span>
<span class="va">pbmc_surt</span> <span class="op">&lt;-</span> <span class="fu">Seurat</span><span class="fu">::</span><span class="fu"><a href="https://satijalab.org/seurat/reference/Idents.html" class="external-link">SetIdent</a></span><span class="op">(</span><span class="va">pbmc_surt</span>, value <span class="op">=</span> <span class="st">"seurat_clusters"</span><span class="op">)</span>
<span class="va">res</span> <span class="op">&lt;-</span> <span class="fu">Seurat</span><span class="fu">::</span><span class="fu"><a href="https://satijalab.org/seurat/reference/FindAllMarkers.html" class="external-link">FindAllMarkers</a></span><span class="op">(</span><span class="va">pbmc_surt</span>, only.pos <span class="op">=</span> <span class="cn">TRUE</span>,
                              min.pct <span class="op">=</span> <span class="fl">0.25</span>, logfc.threshold <span class="op">=</span> <span class="fl">0.25</span><span class="op">)</span>
<span class="fu">metadata</span><span class="op">(</span><span class="va">pbmcs</span><span class="op">$</span><span class="va">CM</span><span class="op">)</span><span class="op">$</span><span class="va">marker_genes</span><span class="op">$</span><span class="va">all</span> <span class="op">&lt;-</span> <span class="va">res</span></code></pre></div>
<p>ASURAT function <code>plog_violin()</code> shows violin plots showing gene expression distributions (see <a href="#violin">here</a> for details).</p>
<div class="sourceCode" id="cb42"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">vname</span> <span class="op">&lt;-</span> <span class="st">"BANK1"</span>
<span class="va">expr_sub</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/SingleCellExperiment/man/altExps.html" class="external-link">altExp</a></span><span class="op">(</span><span class="va">pbmcs</span><span class="op">$</span><span class="va">CM</span>, <span class="st">"logcounts"</span><span class="op">)</span>
<span class="va">expr_sub</span> <span class="op">&lt;-</span> <span class="va">expr_sub</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">expr_sub</span><span class="op">)</span> <span class="op">==</span> <span class="va">vname</span><span class="op">]</span>
<span class="va">labels</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">colData</a></span><span class="op">(</span><span class="va">pbmcs</span><span class="op">$</span><span class="va">CM</span><span class="op">)</span><span class="op">$</span><span class="va">seurat_clusters</span>
<span class="va">dataframe1D</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/as.data.frame.html" class="external-link">as.data.frame</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/t.html" class="external-link">t</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">assay</a></span><span class="op">(</span><span class="va">expr_sub</span>, <span class="st">"counts"</span><span class="op">)</span><span class="op">)</span><span class="op">)</span>
<span class="fu"><a href="../reference/plot_violin.html">plot_violin</a></span><span class="op">(</span>dataframe1D <span class="op">=</span> <span class="va">dataframe1D</span>, labels <span class="op">=</span> <span class="va">labels</span>, colors <span class="op">=</span> <span class="cn">NULL</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu">ggplot2</span><span class="fu">::</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">labs</a></span><span class="op">(</span>title <span class="op">=</span> <span class="va">vname</span>, x <span class="op">=</span> <span class="st">"Cluster (CO &amp; MSigDB)"</span>,
                y <span class="op">=</span> <span class="st">"Normalized expression"</span>, fill <span class="op">=</span> <span class="st">"Cluster"</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu">ggplot2</span><span class="fu">::</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html" class="external-link">theme_classic</a></span><span class="op">(</span>base_size <span class="op">=</span> <span class="fl">20</span><span class="op">)</span></code></pre></div>
<p><img src="ASURAT_files/figure-html/unnamed-chunk-43-1.png" width="700"></p>
<p><br></p>
</div>
</div>
</div>
<div class="section level2">
<h2 id="multilayered">Multifaceted sign analysis<a class="anchor" aria-label="anchor" href="#multilayered"></a>
</h2>
<p>Simultaneously analyze multiple sign-by-sample matrices, which helps us characterize individual samples (cells) from multiple biological aspects.</p>
<p>ASURAT function <code><a href="../reference/plot_multiheatmaps.html">plot_multiheatmaps()</a></code> shows heatmaps (ComplexHeatmap object) of sign scores and gene expression levels (if there are), where rows and columns stand for sign (or gene) and sample (cell), respectively (see <a href="#heatmap">here</a> for details).</p>
<p>First, manually curate signs by setting keywords, followed by selecting top significant signs and genes for the clustering results with respect to separation index and p-value, respectively.</p>
<div class="sourceCode" id="cb43"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># Significant signs</span>
<span class="va">marker_signs</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="op">)</span>
<span class="va">keywords</span> <span class="op">&lt;-</span> <span class="st">"GLIA|glia"</span>
<span class="kw">for</span><span class="op">(</span><span class="va">i</span> <span class="kw">in</span> <span class="fu"><a href="https://rdrr.io/r/base/seq.html" class="external-link">seq_len</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/length.html" class="external-link">length</a></span><span class="op">(</span><span class="va">pbmcs</span><span class="op">)</span><span class="op">)</span><span class="op">)</span><span class="op">{</span>
  <span class="va">marker_signs</span><span class="op">[[</span><span class="va">i</span><span class="op">]</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="fu">metadata</span><span class="op">(</span><span class="va">pbmcs</span><span class="op">[[</span><span class="va">i</span><span class="op">]</span><span class="op">]</span><span class="op">)</span><span class="op">$</span><span class="va">marker_signs</span><span class="op">$</span><span class="va">all</span>
  <span class="va">marker_signs</span><span class="op">[[</span><span class="va">i</span><span class="op">]</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="va">marker_signs</span><span class="op">[[</span><span class="va">i</span><span class="op">]</span><span class="op">]</span><span class="op">[</span><span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/grep.html" class="external-link">grepl</a></span><span class="op">(</span><span class="va">keywords</span>, <span class="va">marker_signs</span><span class="op">[[</span><span class="va">i</span><span class="op">]</span><span class="op">]</span><span class="op">$</span><span class="va">Description</span><span class="op">)</span>, <span class="op">]</span>
  <span class="va">marker_signs</span><span class="op">[[</span><span class="va">i</span><span class="op">]</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/group_by.html" class="external-link">group_by</a></span><span class="op">(</span><span class="va">marker_signs</span><span class="op">[[</span><span class="va">i</span><span class="op">]</span><span class="op">]</span>, <span class="va">Ident_1</span><span class="op">)</span>
  <span class="va">marker_signs</span><span class="op">[[</span><span class="va">i</span><span class="op">]</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/slice.html" class="external-link">slice_max</a></span><span class="op">(</span><span class="va">marker_signs</span><span class="op">[[</span><span class="va">i</span><span class="op">]</span><span class="op">]</span>, <span class="va">sepI</span>, n <span class="op">=</span> <span class="fl">1</span><span class="op">)</span>
  <span class="va">marker_signs</span><span class="op">[[</span><span class="va">i</span><span class="op">]</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/slice.html" class="external-link">slice_min</a></span><span class="op">(</span><span class="va">marker_signs</span><span class="op">[[</span><span class="va">i</span><span class="op">]</span><span class="op">]</span>, <span class="va">SignID</span>, n <span class="op">=</span> <span class="fl">1</span><span class="op">)</span>
<span class="op">}</span>
<span class="co"># Significant genes</span>
<span class="va">marker_genes_CM</span> <span class="op">&lt;-</span> <span class="fu">metadata</span><span class="op">(</span><span class="va">pbmcs</span><span class="op">$</span><span class="va">CM</span><span class="op">)</span><span class="op">$</span><span class="va">marker_genes</span><span class="op">$</span><span class="va">all</span>
<span class="va">marker_genes_CM</span> <span class="op">&lt;-</span> <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/group_by.html" class="external-link">group_by</a></span><span class="op">(</span><span class="va">marker_genes_CM</span>, <span class="va">cluster</span><span class="op">)</span>
<span class="va">marker_genes_CM</span> <span class="op">&lt;-</span> <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/slice.html" class="external-link">slice_min</a></span><span class="op">(</span><span class="va">marker_genes_CM</span>, <span class="va">p_val_adj</span>, n <span class="op">=</span> <span class="fl">5</span><span class="op">)</span>
<span class="va">marker_genes_CM</span> <span class="op">&lt;-</span> <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/slice.html" class="external-link">slice_max</a></span><span class="op">(</span><span class="va">marker_genes_CM</span>, <span class="va">avg_log2FC</span>, n <span class="op">=</span> <span class="fl">1</span><span class="op">)</span></code></pre></div>
<p>Then, prepare arguments.</p>
<div class="sourceCode" id="cb44"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># ssm_list</span>
<span class="va">pbmcs_sub</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="op">)</span> ; <span class="va">ssm_list</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="op">)</span>
<span class="kw">for</span><span class="op">(</span><span class="va">i</span> <span class="kw">in</span> <span class="fu"><a href="https://rdrr.io/r/base/seq.html" class="external-link">seq_len</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/length.html" class="external-link">length</a></span><span class="op">(</span><span class="va">pbmcs</span><span class="op">)</span><span class="op">)</span><span class="op">)</span><span class="op">{</span>
  <span class="va">pbmcs_sub</span><span class="op">[[</span><span class="va">i</span><span class="op">]</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="va">pbmcs</span><span class="op">[[</span><span class="va">i</span><span class="op">]</span><span class="op">]</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">pbmcs</span><span class="op">[[</span><span class="va">i</span><span class="op">]</span><span class="op">]</span><span class="op">)</span> <span class="op"><a href="https://rdrr.io/r/base/match.html" class="external-link">%in%</a></span> <span class="va">marker_signs</span><span class="op">[[</span><span class="va">i</span><span class="op">]</span><span class="op">]</span><span class="op">$</span><span class="va">SignID</span>, <span class="op">]</span>
  <span class="va">ssm_list</span><span class="op">[[</span><span class="va">i</span><span class="op">]</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">assay</a></span><span class="op">(</span><span class="va">pbmcs_sub</span><span class="op">[[</span><span class="va">i</span><span class="op">]</span><span class="op">]</span>, <span class="st">"counts"</span><span class="op">)</span>
<span class="op">}</span>
<span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">ssm_list</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"SSM_COMSig"</span>, <span class="st">"SSM_GO"</span>, <span class="st">"SSM_KEGG"</span><span class="op">)</span>
<span class="co"># gem_list</span>
<span class="va">expr_sub</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/SingleCellExperiment/man/altExps.html" class="external-link">altExp</a></span><span class="op">(</span><span class="va">pbmcs</span><span class="op">$</span><span class="va">CM</span>, <span class="st">"logcounts"</span><span class="op">)</span>
<span class="va">expr_sub</span> <span class="op">&lt;-</span> <span class="va">expr_sub</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">expr_sub</span><span class="op">)</span> <span class="op"><a href="https://rdrr.io/r/base/match.html" class="external-link">%in%</a></span> <span class="va">marker_genes_CM</span><span class="op">$</span><span class="va">gene</span><span class="op">]</span>
<span class="va">gem_list</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>GeneExpr <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">assay</a></span><span class="op">(</span><span class="va">expr_sub</span>, <span class="st">"counts"</span><span class="op">)</span><span class="op">)</span>
<span class="co"># ssmlabel_list</span>
<span class="va">labels</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="op">)</span> ; <span class="va">ssmlabel_list</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="op">)</span>
<span class="kw">for</span><span class="op">(</span><span class="va">i</span> <span class="kw">in</span> <span class="fu"><a href="https://rdrr.io/r/base/seq.html" class="external-link">seq_len</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/length.html" class="external-link">length</a></span><span class="op">(</span><span class="va">pbmcs</span><span class="op">)</span><span class="op">)</span><span class="op">)</span><span class="op">{</span>
  <span class="va">labels</span><span class="op">[[</span><span class="va">i</span><span class="op">]</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html" class="external-link">data.frame</a></span><span class="op">(</span>label <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">colData</a></span><span class="op">(</span><span class="va">pbmcs_sub</span><span class="op">[[</span><span class="va">i</span><span class="op">]</span><span class="op">]</span><span class="op">)</span><span class="op">$</span><span class="va">seurat_clusters</span><span class="op">)</span>
  <span class="va">labels</span><span class="op">[[</span><span class="va">i</span><span class="op">]</span><span class="op">]</span><span class="op">$</span><span class="va">color</span> <span class="op">&lt;-</span> <span class="cn">NA</span>
  <span class="va">ssmlabel_list</span><span class="op">[[</span><span class="va">i</span><span class="op">]</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="va">labels</span><span class="op">[[</span><span class="va">i</span><span class="op">]</span><span class="op">]</span>
<span class="op">}</span>
<span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">ssmlabel_list</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"Label_COMSig"</span>, <span class="st">"Label_GO"</span>, <span class="st">"Label_KEGG"</span><span class="op">)</span>
<span class="co"># gemlabel_list</span>
<span class="va">label_CC</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html" class="external-link">data.frame</a></span><span class="op">(</span>label <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">colData</a></span><span class="op">(</span><span class="va">pbmcs</span><span class="op">$</span><span class="va">CM</span><span class="op">)</span><span class="op">$</span><span class="va">Phase</span>, color <span class="op">=</span> <span class="cn">NA</span><span class="op">)</span>
<span class="va">gemlabel_list</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>CellCycle <span class="op">=</span> <span class="va">label_CC</span><span class="op">)</span></code></pre></div>
<p>Finally, plot heatmaps for the selected signs and genes.</p>
<div class="sourceCode" id="cb45"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/base/Random.html" class="external-link">set.seed</a></span><span class="op">(</span><span class="fl">1</span><span class="op">)</span>
<span class="fu"><a href="../reference/plot_multiheatmaps.html">plot_multiheatmaps</a></span><span class="op">(</span>ssm_list <span class="op">=</span> <span class="va">ssm_list</span>, gem_list <span class="op">=</span> <span class="va">gem_list</span>,
                   ssmlabel_list <span class="op">=</span> <span class="va">ssmlabel_list</span>, gemlabel_list <span class="op">=</span> <span class="va">gemlabel_list</span>,
                   nSamples <span class="op">=</span> <span class="fl">100</span>, show_row_names <span class="op">=</span> <span class="cn">TRUE</span>, title <span class="op">=</span> <span class="st">"PBMC"</span><span class="op">)</span></code></pre></div>
<p><img src="ASURAT_files/figure-html/unnamed-chunk-46-1.png" width="700"></p>
<p><br></p>
</div>
<div class="section level2">
<h2 id="visualization">Appendix. Visualization<a class="anchor" aria-label="anchor" href="#visualization"></a>
</h2>
<p>In this section, ASURAT functions for visualization are introduced.</p>
<p><br></p>
<div class="section level3">
<h3 id="violin">Violin plot<a class="anchor" aria-label="anchor" href="#violin"></a>
</h3>
<p>ASURAT function <code><a href="../reference/plot_violin.html">plot_violin()</a></code> shows violin plots (ggplot objects) showing sign score or gene expression distributions.</p>
<p>The arguments are</p>
<ul>
<li>
<code>dataframe1D</code>: a dataframe with one column,</li>
<li>
<code>labels</code>: a vector of labels of all the samples, corresponding to colors,</li>
<li>
<code>colors</code>: a vector of colors of all the samples, corresponding to labels; if <code>colors = NULL</code>, default colors are used.</li>
</ul>
<p><br></p>
</div>
<div class="section level3">
<h3 id="visualization_lowdim">Low dimensional representation<a class="anchor" aria-label="anchor" href="#visualization_lowdim"></a>
</h3>
<p>ASURAT function <code><a href="../reference/plot_dataframe2D.html">plot_dataframe2D()</a></code> visualizes two-dimensional data with or without labels and colors.</p>
<p>The arguments are</p>
<ul>
<li>
<code>dataframe2D</code>: a dataframe with two columns,</li>
<li>
<code>labels</code>: a vector of labels of all the samples, corresponding to colors,</li>
<li>
<code>colors</code>: a vector of colors of all the samples, corresponding to labels; if <code>colors = NULL</code>, default colors are used.</li>
</ul>
<p><br></p>
<p>ASURAT function <code><a href="../reference/plot_dataframe3D.html">plot_dataframe3D()</a></code> visualizes three-dimensional data with or without labels and colors.</p>
<p>The arguments are</p>
<ul>
<li>
<code>dataframe3D</code>: a dataframe with three columns,</li>
<li>
<code>labels</code>: a vector of labels of all the samples, corresponding to colors,</li>
<li>
<code>colors</code>: a vector of colors of all the samples, corresponding to labels; if <code>colors = NULL</code>, default colors are used,</li>
<li>
<code>theta</code> and <code>phi</code>: angles in degree measure,</li>
<li>
<code>title</code>, <code>title_size</code>, <code>xlabel</code>, <code>ylabel</code>, and <code>zlabel</code>.</li>
</ul>
<p><br></p>
</div>
<div class="section level3">
<h3 id="heatmap">Heatmap<a class="anchor" aria-label="anchor" href="#heatmap"></a>
</h3>
<p>ASURAT function <code>plot_heatmaps()</code> shows heatmaps (<code>ComplexHeatmap</code> objects) of sign scores and gene expression levels (if there are), where rows and columns stand for sign (or gene) and sample (cell), respectively. Here, the column dendrogram is computed by the first component of <code>ssm_list</code> based on ComplexHeatmap functions.</p>
<p>The arguments are</p>
<ul>
<li>
<code>ssm_list</code>: list of sign-by-sample matrices,</li>
<li>
<code>gem_list</code>: list of gene-by-sample matrices,</li>
<li>
<code>ssmlabel_list</code>: <code>NULL</code> or a list of dataframes of sample (cell) labels and colors; the length of the list must be as same as that of <code>ssm_list</code>, and the order of labels in each list must be as same as those in <code>ssm_list</code>,</li>
<li>
<code>gemlabel_list</code>: <code>NULL</code> or a list of dataframes of sample (cell) annotations and colors; the length of the list must be as same as that of <code>gem_list</code>, and the order of labels in each list must be as same as those in <code>gem_list</code>,</li>
<li>
<code>nSamples</code>: number of samples (cells) used for random sampling,</li>
<li>
<code>show_row_names</code>: <code>TRUE</code> or <code>FALSE</code>: if <code>TRUE</code>, row names are shown unless the maximum number of rows exceeds a certain value,</li>
<li>
<code>title</code>: Title.</li>
</ul>
<p><br></p>
</div>
</div>
<div class="section level2">
<h2 id="session-information">Session information<a class="anchor" aria-label="anchor" href="#session-information"></a>
</h2>
<div class="sourceCode" id="cb46"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/utils/sessionInfo.html" class="external-link">sessionInfo</a></span><span class="op">(</span><span class="op">)</span>
<span class="co">#&gt; R version 4.0.4 (2021-02-15)</span>
<span class="co">#&gt; Platform: x86_64-apple-darwin17.0 (64-bit)</span>
<span class="co">#&gt; Running under: macOS Big Sur 10.16</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; Matrix products: default</span>
<span class="co">#&gt; BLAS:   /Library/Frameworks/R.framework/Versions/4.0/Resources/lib/libRblas.dylib</span>
<span class="co">#&gt; LAPACK: /Library/Frameworks/R.framework/Versions/4.0/Resources/lib/libRlapack.dylib</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; locale:</span>
<span class="co">#&gt; [1] en_US.UTF-8/en_US.UTF-8/en_US.UTF-8/C/en_US.UTF-8/en_US.UTF-8</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; attached base packages:</span>
<span class="co">#&gt; [1] parallel  stats4    stats     graphics  grDevices utils     datasets </span>
<span class="co">#&gt; [8] methods   base     </span>
<span class="co">#&gt; </span>
<span class="co">#&gt; other attached packages:</span>
<span class="co">#&gt;  [1] TENxPBMCData_1.8.0          HDF5Array_1.18.1           </span>
<span class="co">#&gt;  [3] rhdf5_2.34.0                DelayedArray_0.16.3        </span>
<span class="co">#&gt;  [5] Matrix_1.3-4                SingleCellExperiment_1.12.0</span>
<span class="co">#&gt;  [7] SummarizedExperiment_1.20.0 Biobase_2.50.0             </span>
<span class="co">#&gt;  [9] GenomicRanges_1.42.0        GenomeInfoDb_1.26.7        </span>
<span class="co">#&gt; [11] IRanges_2.24.1              S4Vectors_0.28.1           </span>
<span class="co">#&gt; [13] BiocGenerics_0.36.1         MatrixGenerics_1.2.1       </span>
<span class="co">#&gt; [15] matrixStats_0.61.0          ASURAT_0.99.1              </span>
<span class="co">#&gt; </span>
<span class="co">#&gt; loaded via a namespace (and not attached):</span>
<span class="co">#&gt;   [1] utf8_1.2.2                    reticulate_1.22              </span>
<span class="co">#&gt;   [3] tidyselect_1.1.1              RSQLite_2.2.8                </span>
<span class="co">#&gt;   [5] AnnotationDbi_1.52.0          htmlwidgets_1.5.4            </span>
<span class="co">#&gt;   [7] grid_4.0.4                    BiocParallel_1.24.1          </span>
<span class="co">#&gt;   [9] Rtsne_0.15                    munsell_0.5.0                </span>
<span class="co">#&gt;  [11] codetools_0.2-18              ragg_1.1.3                   </span>
<span class="co">#&gt;  [13] ica_1.0-2                     future_1.23.0                </span>
<span class="co">#&gt;  [15] miniUI_0.1.1.1                misc3d_0.9-1                 </span>
<span class="co">#&gt;  [17] withr_2.4.3                   colorspace_2.0-2             </span>
<span class="co">#&gt;  [19] highr_0.9                     knitr_1.36                   </span>
<span class="co">#&gt;  [21] Seurat_4.0.5                  ROCR_1.0-11                  </span>
<span class="co">#&gt;  [23] tensor_1.5                    listenv_0.8.0                </span>
<span class="co">#&gt;  [25] labeling_0.4.2                GenomeInfoDbData_1.2.4       </span>
<span class="co">#&gt;  [27] polyclip_1.10-0               bit64_4.0.5                  </span>
<span class="co">#&gt;  [29] farver_2.1.0                  rprojroot_2.0.2              </span>
<span class="co">#&gt;  [31] parallelly_1.28.1             vctrs_0.3.8                  </span>
<span class="co">#&gt;  [33] generics_0.1.1                xfun_0.27                    </span>
<span class="co">#&gt;  [35] BiocFileCache_1.14.0          R6_2.5.1                     </span>
<span class="co">#&gt;  [37] clue_0.3-60                   locfit_1.5-9.4               </span>
<span class="co">#&gt;  [39] bitops_1.0-7                  rhdf5filters_1.2.1           </span>
<span class="co">#&gt;  [41] spatstat.utils_2.2-0          cachem_1.0.6                 </span>
<span class="co">#&gt;  [43] assertthat_0.2.1              promises_1.2.0.1             </span>
<span class="co">#&gt;  [45] scales_1.1.1                  gtable_0.3.0                 </span>
<span class="co">#&gt;  [47] Cairo_1.5-12.2                globals_0.14.0               </span>
<span class="co">#&gt;  [49] goftest_1.2-3                 rlang_0.4.12                 </span>
<span class="co">#&gt;  [51] systemfonts_1.0.2             GlobalOptions_0.1.2          </span>
<span class="co">#&gt;  [53] splines_4.0.4                 lazyeval_0.2.2               </span>
<span class="co">#&gt;  [55] spatstat.geom_2.3-0           BiocManager_1.30.16          </span>
<span class="co">#&gt;  [57] yaml_2.2.1                    reshape2_1.4.4               </span>
<span class="co">#&gt;  [59] abind_1.4-5                   httpuv_1.6.3                 </span>
<span class="co">#&gt;  [61] tools_4.0.4                   tcltk_4.0.4                  </span>
<span class="co">#&gt;  [63] ggplot2_3.3.5                 ellipsis_0.3.2               </span>
<span class="co">#&gt;  [65] spatstat.core_2.3-1           jquerylib_0.1.4              </span>
<span class="co">#&gt;  [67] RColorBrewer_1.1-2            ggridges_0.5.3               </span>
<span class="co">#&gt;  [69] Rcpp_1.0.7                    plyr_1.8.6                   </span>
<span class="co">#&gt;  [71] zlibbioc_1.36.0               purrr_0.3.4                  </span>
<span class="co">#&gt;  [73] RCurl_1.98-1.5                rpart_4.1-15                 </span>
<span class="co">#&gt;  [75] deldir_1.0-6                  bayNorm_1.8.0                </span>
<span class="co">#&gt;  [77] pbapply_1.5-0                 GetoptLong_1.0.5             </span>
<span class="co">#&gt;  [79] cowplot_1.1.1                 zoo_1.8-9                    </span>
<span class="co">#&gt;  [81] SeuratObject_4.0.2            ggrepel_0.9.1                </span>
<span class="co">#&gt;  [83] cluster_2.1.2                 fs_1.5.0                     </span>
<span class="co">#&gt;  [85] magrittr_2.0.1                data.table_1.14.2            </span>
<span class="co">#&gt;  [87] scattermore_0.7               circlize_0.4.13              </span>
<span class="co">#&gt;  [89] lmtest_0.9-38                 RANN_2.6.1                   </span>
<span class="co">#&gt;  [91] fitdistrplus_1.1-6            patchwork_1.1.1              </span>
<span class="co">#&gt;  [93] mime_0.12                     evaluate_0.14                </span>
<span class="co">#&gt;  [95] xtable_1.8-4                  gridExtra_2.3                </span>
<span class="co">#&gt;  [97] shape_1.4.6                   compiler_4.0.4               </span>
<span class="co">#&gt;  [99] tibble_3.1.5                  KernSmooth_2.23-20           </span>
<span class="co">#&gt; [101] crayon_1.4.2                  htmltools_0.5.2              </span>
<span class="co">#&gt; [103] mgcv_1.8-38                   later_1.3.0                  </span>
<span class="co">#&gt; [105] snow_0.4-4                    tidyr_1.1.4                  </span>
<span class="co">#&gt; [107] DBI_1.1.1                     ExperimentHub_1.16.1         </span>
<span class="co">#&gt; [109] dbplyr_2.1.1                  ComplexHeatmap_2.6.2         </span>
<span class="co">#&gt; [111] MASS_7.3-54                   rappdirs_0.3.3               </span>
<span class="co">#&gt; [113] igraph_1.2.7                  pkgconfig_2.0.3              </span>
<span class="co">#&gt; [115] pkgdown_2.0.2                 plotly_4.10.0                </span>
<span class="co">#&gt; [117] spatstat.sparse_2.0-0         foreach_1.5.1                </span>
<span class="co">#&gt; [119] bslib_0.3.1                   XVector_0.30.0               </span>
<span class="co">#&gt; [121] stringr_1.4.0                 digest_0.6.28                </span>
<span class="co">#&gt; [123] sctransform_0.3.2             RcppAnnoy_0.0.19             </span>
<span class="co">#&gt; [125] spatstat.data_2.1-0           rmarkdown_2.11               </span>
<span class="co">#&gt; [127] leiden_0.3.9                  uwot_0.1.10                  </span>
<span class="co">#&gt; [129] curl_4.3.2                    shiny_1.7.1                  </span>
<span class="co">#&gt; [131] rjson_0.2.20                  nlme_3.1-153                 </span>
<span class="co">#&gt; [133] lifecycle_1.0.1               jsonlite_1.7.2               </span>
<span class="co">#&gt; [135] Rhdf5lib_1.12.1               limma_3.46.0                 </span>
<span class="co">#&gt; [137] desc_1.4.0                    viridisLite_0.4.0            </span>
<span class="co">#&gt; [139] fansi_0.5.0                   pillar_1.6.4                 </span>
<span class="co">#&gt; [141] lattice_0.20-45               fastmap_1.1.0                </span>
<span class="co">#&gt; [143] httr_1.4.2                    survival_3.2-13              </span>
<span class="co">#&gt; [145] interactiveDisplayBase_1.28.0 glue_1.4.2                   </span>
<span class="co">#&gt; [147] png_0.1-7                     iterators_1.0.13             </span>
<span class="co">#&gt; [149] plot3D_1.4                    BiocVersion_3.12.0           </span>
<span class="co">#&gt; [151] bit_4.0.4                     stringi_1.7.5                </span>
<span class="co">#&gt; [153] sass_0.4.0                    blob_1.2.2                   </span>
<span class="co">#&gt; [155] textshaping_0.3.6             org.Hs.eg.db_3.12.0          </span>
<span class="co">#&gt; [157] AnnotationHub_2.22.1          doSNOW_1.0.19                </span>
<span class="co">#&gt; [159] memoise_2.0.0                 dplyr_1.0.7                  </span>
<span class="co">#&gt; [161] irlba_2.3.3                   future.apply_1.8.1</span></code></pre></div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p></p>
<p>Developed by Keita Iida.</p>
</div>

<div class="pkgdown">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.2.</p>
</div>

      </footer>
</div>

  


  

  </body>
</html>
